{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import library\n",
    "from langchain_community.document_loaders import PyPDFLoader\n",
    "\n",
    "openai_api_key = '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'\n",
    "\n",
    "loader = PyPDFLoader('./documents/ISLP_website.pdf')\n",
    "data = loader.load()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "topic = \"Machine learning\"\n",
    "f\"\"\"You are given the following topic: {topic}. Based on the additional data provided, \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Document(page_content='(BSFUI\\x01+BNFT\\x01t\\x01%BOJFMB\\x018JUUFO\\x01t\\x015SFWPS\\x01)BTUJF3PCFSU\\x015JCTIJSBOJ\\x01t\\x01+POBUIBO\\x015BZMPS\"O\\x01*OUSPEVDUJPOUP\\x014UBUJTUJDBM\\x01-FBSOJOHXJUI\\x01\"QQMJDBUJPOT\\x01JO\\x011ZUIPO', metadata={'source': './documents/ISLP_website.pdf', 'page': 0}), Document(page_content='To our parents:Alison and Michael JamesChiara Nappi and Edward WittenValerie and Patrick HastieVera and Sami TibshiraniJohn and Brenda Taylorand to our families:Michael, Daniel, and CatherineTessa, Theo, Otto, and AriSamantha, Timothy, and LyndaCharlie, Ryan, Julie, and CherylLee-Ann and Isobel', metadata={'source': './documents/ISLP_website.pdf', 'page': 1}), Document(page_content='Preface\\nStatistical learning refers to a set of tools for making sense of complex\\ndatasets. In recent years, we have seen a staggering increase in the scale and\\nscope of data collection across virtually all areas of science and industry.\\nAs a result, statistical learning has become a critical toolkit for anyone who\\nwishes to understand data — and as more and more of today’s jobs involve\\ndata, this means that statistical learning is fast becoming a critical toolkit\\nforeveryone .\\nOne of the first books on statistical learning — The Elements of Statisti-\\ncal Learning (ESL, by Hastie, Tibshirani, and Friedman) — was published\\nin 2001, with a second edition in 2009. ESL has become a popular text not\\nonly in statistics but also in related fields. One of the reasons for ESL’s\\npopularity is its relatively accessible style. But ESL is best-suited for indi-\\nviduals with advanced training in the mathematical sciences.\\nAn Introduction to Statistical Learning, With Applications in R (ISLR)\\n— first published in 2013, with a second edition in 2021 — arose from\\nthe clear need for a broader and less technical treatment of the key topics\\nin statistical learning. In addition to a review of linear regression, ISLR\\ncovers many of today’s most important statistical and machine learning\\napproaches, including resampling, sparse methods for classification and re-\\ngression, generalized additive models, tree-based methods, support vector\\nmachines, deep learning, survival analysis, clustering, and multiple testing.\\nSince it was published in 2013, ISLR has become a mainstay of un-\\ndergraduate and graduate classrooms worldwide, as well as an important\\nreference book for data scientists. One of the keys to its success has been\\nthat, beginning with Chapter 2, each chapter contains an Rlab illustrating\\nhow to implement the statistical learning methods seen in that chapter,\\nproviding the reader with valuable hands-on experience.\\nHowever, in recent years Pythonhas become an increasingly popular lan-\\nguage for data science, and there has been increasing demand for a Python-\\nvii', metadata={'source': './documents/ISLP_website.pdf', 'page': 2}), Document(page_content='viiibased alternative to ISLR. Hence, this book,An Introduction to StatisticalLearning, With Applications in Python(ISLP), covers the same materialsas ISLR but with labs implemented inPython— a feat accomplished by theaddition of a new co-author, Jonathan Taylor. Several of the labs make useof theISLP Pythonpackage, which we have written to facilitate carrying outthe statistical learning methods covered in each chapter inPython. Theselabs will be useful both forPythonnovices, as well as experienced users.The intention behind ISLP (and ISLR) is to concentrate more on theapplications of the methods and less on the mathematical details, so it isappropriate for advanced undergraduates or master’s students in statisticsor related quantitative fields, or for individuals in other disciplines whowish to use statistical learning tools to analyze their data. It can be usedas a textbook for a course spanning two semesters.We are grateful to these readers for providing valuable comments on thefirst edition of ISLR: Pallavi Basu, Alexandra Chouldechova, Patrick Dana-her, Will Fithian, Luella Fu, Sam Gross, Max Grazier G’Sell, CourtneyPaulson, Xinghao Qiao, Elisa Sheng, Noah Simon, Kean Ming Tan, Xin LuTan.WethankthesereadersforhelpfulinputonthesecondeditionofISLR:Alan Agresti, Iain Carmichael, Yiqun Chen, Erin Craig, Daisy Ding, LucyGao, Ismael Lemhadri, Bryan Martin, Anna Neufeld, Geoff Tims, CarstenVoelkmann, Steve Yadlowsky, and James Zou. We are immensely gratefulto Balasubramanian “Naras” Narasimhan for his assistance on both ISLRand ISLP.It has been an honor and a privilege for us to see the considerable impactthatISLRhashadonthewayinwhichstatisticallearningispracticed,bothin and out of the academic setting. We hope that this newPythoneditionwill continue to give today’s and tomorrow’s applied statisticians and datascientists the tools they need for success in a data-driven world.It’s tough to make predictions, especially about the future.-Yogi BerraPreface', metadata={'source': './documents/ISLP_website.pdf', 'page': 3}), Document(page_content='Contents\\nPreface vii\\n1 Introduction 1\\n2 Statistical Learning 15\\n2.1 What Is Statistical Learning? . . . . . . . . . . . . . . . . . 15\\n2.1.1 Why Estimate f? . . . . . . . . . . . . . . . . . . . 17\\n2.1.2 How Do We Estimate f? . . . . . . . . . . . . . . . 20\\n2.1.3 The Trade-Off Between Prediction Accuracy\\nand Model Interpretability . . . . . . . . . . . . . . 23\\n2.1.4 Supervised Versus Unsupervised Learning . . . . . 25\\n2.1.5 Regression Versus Classification Problems . . . . . 27\\n2.2 Assessing Model Accuracy . . . . . . . . . . . . . . . . . . 27\\n2.2.1 Measuring the Quality of Fit . . . . . . . . . . . . 28\\n2.2.2 The Bias-Variance Trade-Off . . . . . . . . . . . . . 31\\n2.2.3 The Classification Setting . . . . . . . . . . . . . . 34\\n2.3 Lab: Introduction to Python . . . . . . . . . . . . . . . . . 40\\n2.3.1 Getting Started . . . . . . . . . . . . . . . . . . . . 40\\n2.3.2 Basic Commands . . . . . . . . . . . . . . . . . . . 40\\n2.3.3 Introduction to Numerical Python . . . . . . . . . 42\\n2.3.4 Graphics . . . . . . . . . . . . . . . . . . . . . . . . 48\\n2.3.5 Sequences and Slice Notation . . . . . . . . . . . . 51\\n2.3.6 Indexing Data . . . . . . . . . . . . . . . . . . . . . 51\\n2.3.7 Loading Data . . . . . . . . . . . . . . . . . . . . . 55\\n2.3.8 For Loops . . . . . . . . . . . . . . . . . . . . . . . 59\\n2.3.9 Additional Graphical and Numerical Summaries . . 61\\n2.4 Exercises . . . . . . . . . . . . . . . . . . . . . . . . . . . . 63\\n3 Linear Regression 69\\n3.1 Simple Linear Regression . . . . . . . . . . . . . . . . . . . 70\\n3.1.1 Estimating the Coefficients . . . . . . . . . . . . . 71\\n3.1.2 Assessing the Accuracy of the Coefficient\\nEstimates . . . . . . . . . . . . . . . . . . . . . . . 72\\n3.1.3 Assessing the Accuracy of the Model . . . . . . . . 77\\n3.2 Multiple Linear Regression . . . . . . . . . . . . . . . . . . 80\\n3.2.1 Estimating the Regression Coefficients . . . . . . . 81\\nix', metadata={'source': './documents/ISLP_website.pdf', 'page': 4}), Document(page_content='x Contents\\n3.2.2 Some Important Questions . . . . . . . . . . . . . . 83\\n3.3 Other Considerations in the Regression Model . . . . . . . 91\\n3.3.1 Qualitative Predictors . . . . . . . . . . . . . . . . 91\\n3.3.2 Extensions of the Linear Model . . . . . . . . . . . 94\\n3.3.3 Potential Problems . . . . . . . . . . . . . . . . . . 100\\n3.4 The Marketing Plan . . . . . . . . . . . . . . . . . . . . . . 109\\n3.5 Comparison of Linear Regression with K-Nearest\\nNeighbors . . . . . . . . . . . . . . . . . . . . . . . . . . . . 111\\n3.6 Lab: Linear Regression . . . . . . . . . . . . . . . . . . . . 116\\n3.6.1 Importing packages . . . . . . . . . . . . . . . . . . 116\\n3.6.2 Simple Linear Regression . . . . . . . . . . . . . . . 117\\n3.6.3 Multiple Linear Regression . . . . . . . . . . . . . . 122\\n3.6.4 Multivariate Goodness of Fit . . . . . . . . . . . . 123\\n3.6.5 Interaction Terms . . . . . . . . . . . . . . . . . . . 124\\n3.6.6 Non-linear Transformations of the Predictors . . . 125\\n3.6.7 Qualitative Predictors . . . . . . . . . . . . . . . . 126\\n3.7 Exercises . . . . . . . . . . . . . . . . . . . . . . . . . . . . 127\\n4 Classification 135\\n4.1 An Overview of Classification . . . . . . . . . . . . . . . . . 135\\n4.2 Why Not Linear Regression? . . . . . . . . . . . . . . . . . 136\\n4.3 Logistic Regression . . . . . . . . . . . . . . . . . . . . . . 138\\n4.3.1 The Logistic Model . . . . . . . . . . . . . . . . . . 139\\n4.3.2 Estimating the Regression Coefficients . . . . . . . 140\\n4.3.3 Making Predictions . . . . . . . . . . . . . . . . . . 141\\n4.3.4 Multiple Logistic Regression . . . . . . . . . . . . . 142\\n4.3.5 Multinomial Logistic Regression . . . . . . . . . . . 144\\n4.4 Generative Models for Classification . . . . . . . . . . . . . 146\\n4.4.1 Linear Discriminant Analysis for p=1 . . . . . . . 147\\n4.4.2 Linear Discriminant Analysis for p>1 . . . . . . . 150\\n4.4.3 Quadratic Discriminant Analysis . . . . . . . . . . 156\\n4.4.4 Naive Bayes . . . . . . . . . . . . . . . . . . . . . . 158\\n4.5 A Comparison of Classification Methods . . . . . . . . . . 161\\n4.5.1 An Analytical Comparison . . . . . . . . . . . . . . 161\\n4.5.2 An Empirical Comparison . . . . . . . . . . . . . . 164\\n4.6 Generalized Linear Models . . . . . . . . . . . . . . . . . . 167\\n4.6.1 Linear Regression on the Bikeshare Data . . . . . . 167\\n4.6.2 Poisson Regression on the Bikeshare Data . . . . . 169\\n4.6.3 Generalized Linear Models in Greater Generality . 172\\n4.7 Lab: Logistic Regression, LDA, QDA, and KNN . . . . . . 173\\n4.7.1 The Stock Market Data . . . . . . . . . . . . . . . 173\\n4.7.2 Logistic Regression . . . . . . . . . . . . . . . . . . 174\\n4.7.3 Linear Discriminant Analysis . . . . . . . . . . . . 179\\n4.7.4 Quadratic Discriminant Analysis . . . . . . . . . . 181\\n4.7.5 Naive Bayes . . . . . . . . . . . . . . . . . . . . . . 182\\n4.7.6 K-Nearest Neighbors . . . . . . . . . . . . . . . . . 183\\n4.7.7 Linear and Poisson Regression on the Bikeshare Data188\\n4.8 Exercises . . . . . . . . . . . . . . . . . . . . . . . . . . . . 193', metadata={'source': './documents/ISLP_website.pdf', 'page': 5}), Document(page_content='Contents xi\\n5 Resampling Methods 201\\n5.1 Cross-Validation . . . . . . . . . . . . . . . . . . . . . . . . 202\\n5.1.1 The Validation Set Approach . . . . . . . . . . . . 202\\n5.1.2 Leave-One-Out Cross-Validation . . . . . . . . . . 204\\n5.1.3k-Fold Cross-Validation . . . . . . . . . . . . . . . 206\\n5.1.4 Bias-Variance Trade-Off for k-Fold\\nCross-Validation . . . . . . . . . . . . . . . . . . . 208\\n5.1.5 Cross-Validation on Classification Problems . . . . 209\\n5.2 The Bootstrap . . . . . . . . . . . . . . . . . . . . . . . . . 212\\n5.3 Lab: Cross-Validation and the Bootstrap . . . . . . . . . . 215\\n5.3.1 The Validation Set Approach . . . . . . . . . . . . 216\\n5.3.2 Cross-Validation . . . . . . . . . . . . . . . . . . . 217\\n5.3.3 The Bootstrap . . . . . . . . . . . . . . . . . . . . 220\\n5.4 Exercises . . . . . . . . . . . . . . . . . . . . . . . . . . . . 224\\n6 Linear Model Selection and Regularization 229\\n6.1 Subset Selection . . . . . . . . . . . . . . . . . . . . . . . . 231\\n6.1.1 Best Subset Selection . . . . . . . . . . . . . . . . . 231\\n6.1.2 Stepwise Selection . . . . . . . . . . . . . . . . . . 233\\n6.1.3 Choosing the Optimal Model . . . . . . . . . . . . 235\\n6.2 Shrinkage Methods . . . . . . . . . . . . . . . . . . . . . . 240\\n6.2.1 Ridge Regression . . . . . . . . . . . . . . . . . . . 240\\n6.2.2 The Lasso . . . . . . . . . . . . . . . . . . . . . . . 244\\n6.2.3 Selecting the Tuning Parameter . . . . . . . . . . . 252\\n6.3 Dimension Reduction Methods . . . . . . . . . . . . . . . . 253\\n6.3.1 Principal Components Regression . . . . . . . . . . 254\\n6.3.2 Partial Least Squares . . . . . . . . . . . . . . . . . 260\\n6.4 Considerations in High Dimensions . . . . . . . . . . . . . 262\\n6.4.1 High-Dimensional Data . . . . . . . . . . . . . . . . 262\\n6.4.2 What Goes Wrong in High Dimensions? . . . . . . 263\\n6.4.3 Regression in High Dimensions . . . . . . . . . . . 265\\n6.4.4 Interpreting Results in High Dimensions . . . . . . 266\\n6.5 Lab: Linear Models and Regularization Methods . . . . . . 267\\n6.5.1 Subset Selection Methods . . . . . . . . . . . . . . 268\\n6.5.2 Ridge Regression and the Lasso . . . . . . . . . . . 273\\n6.5.3 PCR and PLS Regression . . . . . . . . . . . . . . 280\\n6.6 Exercises . . . . . . . . . . . . . . . . . . . . . . . . . . . . 283\\n7 Moving Beyond Linearity 289\\n7.1 Polynomial Regression . . . . . . . . . . . . . . . . . . . . . 290\\n7.2 Step Functions . . . . . . . . . . . . . . . . . . . . . . . . . 292\\n7.3 Basis Functions . . . . . . . . . . . . . . . . . . . . . . . . 293\\n7.4 Regression Splines . . . . . . . . . . . . . . . . . . . . . . . 294\\n7.4.1 Piecewise Polynomials . . . . . . . . . . . . . . . . 294\\n7.4.2 Constraints and Splines . . . . . . . . . . . . . . . 296\\n7.4.3 The Spline Basis Representation . . . . . . . . . . 296\\n7.4.4 Choosing the Number and Locations\\nof the Knots . . . . . . . . . . . . . . . . . . . . . . 297\\n7.4.5 Comparison to Polynomial Regression . . . . . . . 299', metadata={'source': './documents/ISLP_website.pdf', 'page': 6}), Document(page_content='xii Contents\\n7.5 Smoothing Splines . . . . . . . . . . . . . . . . . . . . . . . 300\\n7.5.1 An Overview of Smoothing Splines . . . . . . . . . 300\\n7.5.2 Choosing the Smoothing Parameter λ. . . . . . . 301\\n7.6 Local Regression . . . . . . . . . . . . . . . . . . . . . . . . 303\\n7.7 Generalized Additive Models . . . . . . . . . . . . . . . . . 305\\n7.7.1 GAMs for Regression Problems . . . . . . . . . . . 306\\n7.7.2 GAMs for Classification Problems . . . . . . . . . . 308\\n7.8 Lab: Non-Linear Modeling . . . . . . . . . . . . . . . . . . 309\\n7.8.1 Polynomial Regression and Step Functions . . . . . 310\\n7.8.2 Splines . . . . . . . . . . . . . . . . . . . . . . . . . 315\\n7.8.3 Smoothing Splines and GAMs . . . . . . . . . . . . 317\\n7.8.4 Local Regression . . . . . . . . . . . . . . . . . . . 324\\n7.9 Exercises . . . . . . . . . . . . . . . . . . . . . . . . . . . . 325\\n8 Tree-Based Methods 331\\n8.1 The Basics of Decision Trees . . . . . . . . . . . . . . . . . 331\\n8.1.1 Regression Trees . . . . . . . . . . . . . . . . . . . 331\\n8.1.2 Classification Trees . . . . . . . . . . . . . . . . . . 337\\n8.1.3 Trees Versus Linear Models . . . . . . . . . . . . . 341\\n8.1.4 Advantages and Disadvantages of Trees . . . . . . . 341\\n8.2 Bagging, Random Forests, Boosting, and Bayesian Additive\\nRegression Trees . . . . . . . . . . . . . . . . . . . . . . . . 343\\n8.2.1 Bagging . . . . . . . . . . . . . . . . . . . . . . . . 343\\n8.2.2 Random Forests . . . . . . . . . . . . . . . . . . . . 346\\n8.2.3 Boosting . . . . . . . . . . . . . . . . . . . . . . . . 347\\n8.2.4 Bayesian Additive Regression Trees . . . . . . . . . 350\\n8.2.5 Summary of Tree Ensemble Methods . . . . . . . . 353\\n8.3 Lab: Tree-Based Methods . . . . . . . . . . . . . . . . . . . 354\\n8.3.1 Fitting Classification Trees . . . . . . . . . . . . . . 355\\n8.3.2 Fitting Regression Trees . . . . . . . . . . . . . . . 358\\n8.3.3 Bagging and Random Forests . . . . . . . . . . . . 360\\n8.3.4 Boosting . . . . . . . . . . . . . . . . . . . . . . . . 361\\n8.3.5 Bayesian Additive Regression Trees . . . . . . . . . 362\\n8.4 Exercises . . . . . . . . . . . . . . . . . . . . . . . . . . . . 363\\n9 Support Vector Machines 367\\n9.1 Maximal Margin Classifier . . . . . . . . . . . . . . . . . . 367\\n9.1.1 What Is a Hyperplane? . . . . . . . . . . . . . . . . 368\\n9.1.2 Classification Using a Separating Hyperplane . . . 368\\n9.1.3 The Maximal Margin Classifier . . . . . . . . . . . 370\\n9.1.4 Construction of the Maximal Margin Classifier . . 372\\n9.1.5 The Non-separable Case . . . . . . . . . . . . . . . 372\\n9.2 Support Vector Classifiers . . . . . . . . . . . . . . . . . . . 373\\n9.2.1 Overview of the Support Vector Classifier . . . . . 373\\n9.2.2 Details of the Support Vector Classifier . . . . . . . 374\\n9.3 Support Vector Machines . . . . . . . . . . . . . . . . . . . 377\\n9.3.1 Classification with Non-Linear Decision\\nBoundaries . . . . . . . . . . . . . . . . . . . . . . 378\\n9.3.2 The Support Vector Machine . . . . . . . . . . . . 379', metadata={'source': './documents/ISLP_website.pdf', 'page': 7}), Document(page_content='Contents xiii\\n9.3.3 An Application to the Heart Disease Data . . . . . 382\\n9.4 SVMs with More than Two Classes . . . . . . . . . . . . . 383\\n9.4.1 One-Versus-One Classification . . . . . . . . . . . . 384\\n9.4.2 One-Versus-All Classification . . . . . . . . . . . . 384\\n9.5 Relationship to Logistic Regression . . . . . . . . . . . . . 384\\n9.6 Lab: Support Vector Machines . . . . . . . . . . . . . . . . 387\\n9.6.1 Support Vector Classifier . . . . . . . . . . . . . . . 387\\n9.6.2 Support Vector Machine . . . . . . . . . . . . . . . 390\\n9.6.3 ROC Curves . . . . . . . . . . . . . . . . . . . . . . 392\\n9.6.4 SVM with Multiple Classes . . . . . . . . . . . . . 393\\n9.6.5 Application to Gene Expression Data . . . . . . . . 394\\n9.7 Exercises . . . . . . . . . . . . . . . . . . . . . . . . . . . . 395\\n10 Deep Learning 399\\n10.1 Single Layer Neural Networks . . . . . . . . . . . . . . . . 400\\n10.2 Multilayer Neural Networks . . . . . . . . . . . . . . . . . . 402\\n10.3 Convolutional Neural Networks . . . . . . . . . . . . . . . . 406\\n10.3.1 Convolution Layers . . . . . . . . . . . . . . . . . . 407\\n10.3.2 Pooling Layers . . . . . . . . . . . . . . . . . . . . 410\\n10.3.3 Architecture of a Convolutional Neural Network . . 410\\n10.3.4 Data Augmentation . . . . . . . . . . . . . . . . . . 411\\n10.3.5 Results Using a Pretrained Classifier . . . . . . . . 412\\n10.4 Document Classification . . . . . . . . . . . . . . . . . . . . 413\\n10.5 Recurrent Neural Networks . . . . . . . . . . . . . . . . . . 416\\n10.5.1 Sequential Models for Document Classification . . 418\\n10.5.2 Time Series Forecasting . . . . . . . . . . . . . . . 420\\n10.5.3 Summary of RNNs . . . . . . . . . . . . . . . . . . 424\\n10.6 When to Use Deep Learning . . . . . . . . . . . . . . . . . 425\\n10.7 Fitting a Neural Network . . . . . . . . . . . . . . . . . . . 427\\n10.7.1 Backpropagation . . . . . . . . . . . . . . . . . . . 428\\n10.7.2 Regularization and Stochastic Gradient Descent . . 429\\n10.7.3 Dropout Learning . . . . . . . . . . . . . . . . . . . 431\\n10.7.4 Network Tuning . . . . . . . . . . . . . . . . . . . . 431\\n10.8 Interpolation and Double Descent . . . . . . . . . . . . . . 432\\n10.9 Lab: Deep Learning . . . . . . . . . . . . . . . . . . . . . . 435\\n10.9.1 Single Layer Network on Hitters Data . . . . . . . 437\\n10.9.2 Multilayer Network on the MNIST Digit Data . . . 444\\n10.9.3 Convolutional Neural Networks . . . . . . . . . . . 448\\n10.9.4 Using Pretrained CNN Models . . . . . . . . . . . 452\\n10.9.5 IMDB Document Classification . . . . . . . . . . . 454\\n10.9.6 Recurrent Neural Networks . . . . . . . . . . . . . 458\\n10.10 Exercises . . . . . . . . . . . . . . . . . . . . . . . . . . . . 465\\n11 Survival Analysis and Censored Data 469\\n11.1 Survival and Censoring Times . . . . . . . . . . . . . . . . 470\\n11.2 A Closer Look at Censoring . . . . . . . . . . . . . . . . . . 470\\n11.3 The Kaplan–Meier Survival Curve . . . . . . . . . . . . . . 472\\n11.4 The Log-Rank Test . . . . . . . . . . . . . . . . . . . . . . 474\\n11.5 Regression Models With a Survival Response . . . . . . . . 476', metadata={'source': './documents/ISLP_website.pdf', 'page': 8}), Document(page_content='xiv Contents\\n11.5.1 The Hazard Function . . . . . . . . . . . . . . . . . 476\\n11.5.2 Proportional Hazards . . . . . . . . . . . . . . . . . 478\\n11.5.3 Example: Brain Cancer Data . . . . . . . . . . . . 482\\n11.5.4 Example: Publication Data . . . . . . . . . . . . . 482\\n11.6 Shrinkage for the Cox Model . . . . . . . . . . . . . . . . . 484\\n11.7 Additional Topics . . . . . . . . . . . . . . . . . . . . . . . 486\\n11.7.1 Area Under the Curve for Survival Analysis . . . . 486\\n11.7.2 Choice of Time Scale . . . . . . . . . . . . . . . . . 487\\n11.7.3 Time-Dependent Covariates . . . . . . . . . . . . . 488\\n11.7.4 Checking the Proportional Hazards Assumption . . 488\\n11.7.5 Survival Trees . . . . . . . . . . . . . . . . . . . . . 488\\n11.8 Lab: Survival Analysis . . . . . . . . . . . . . . . . . . . . . 489\\n11.8.1 Brain Cancer Data . . . . . . . . . . . . . . . . . . 489\\n11.8.2 Publication Data . . . . . . . . . . . . . . . . . . . 493\\n11.8.3 Call Center Data . . . . . . . . . . . . . . . . . . . 494\\n11.9 Exercises . . . . . . . . . . . . . . . . . . . . . . . . . . . . 498\\n12 Unsupervised Learning 503\\n12.1 The Challenge of Unsupervised Learning . . . . . . . . . . 503\\n12.2 Principal Components Analysis . . . . . . . . . . . . . . . . 504\\n12.2.1 What Are Principal Components? . . . . . . . . . . 505\\n12.2.2 Another Interpretation of Principal Components . 508\\n12.2.3 The Proportion of Variance Explained . . . . . . . 510\\n12.2.4 More on PCA . . . . . . . . . . . . . . . . . . . . . 512\\n12.2.5 Other Uses for Principal Components . . . . . . . . 515\\n12.3 Missing Values and Matrix Completion . . . . . . . . . . . 515\\n12.4 Clustering Methods . . . . . . . . . . . . . . . . . . . . . . 520\\n12.4.1K-Means Clustering . . . . . . . . . . . . . . . . . 521\\n12.4.2 Hierarchical Clustering . . . . . . . . . . . . . . . . 525\\n12.4.3 Practical Issues in Clustering . . . . . . . . . . . . 532\\n12.5 Lab: Unsupervised Learning . . . . . . . . . . . . . . . . . 535\\n12.5.1 Principal Components Analysis . . . . . . . . . . . 535\\n12.5.2 Matrix Completion . . . . . . . . . . . . . . . . . . 539\\n12.5.3 Clustering . . . . . . . . . . . . . . . . . . . . . . . 542\\n12.5.4 NCI60 Data Example . . . . . . . . . . . . . . . . . 546\\n12.6 Exercises . . . . . . . . . . . . . . . . . . . . . . . . . . . . 552\\n13 Multiple Testing 557\\n13.1 A Quick Review of Hypothesis Testing . . . . . . . . . . . 558\\n13.1.1 Testing a Hypothesis . . . . . . . . . . . . . . . . . 558\\n13.1.2 Type I and Type II Errors . . . . . . . . . . . . . . 562\\n13.2 The Challenge of Multiple Testing . . . . . . . . . . . . . . 563\\n13.3 The Family-Wise Error Rate . . . . . . . . . . . . . . . . . 565\\n13.3.1 What is the Family-Wise Error Rate? . . . . . . . 565\\n13.3.2 Approaches to Control the Family-Wise Error Rate 567\\n13.3.3 Trade-Off Between the FWER and Power . . . . . 572\\n13.4 The False Discovery Rate . . . . . . . . . . . . . . . . . . . 573\\n13.4.1 Intuition for the False Discovery Rate . . . . . . . 573\\n13.4.2 The Benjamini–Hochberg Procedure . . . . . . . . 575', metadata={'source': './documents/ISLP_website.pdf', 'page': 9}), Document(page_content='Contents xv\\n13.5 A Re-Sampling Approach to p-Values and False Discovery\\nRates . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 577\\n13.5.1 A Re-Sampling Approach to the p-Value . . . . . . 578\\n13.5.2 A Re-Sampling Approach to the False Discovery Rate579\\n13.5.3 When Are Re-Sampling Approaches Useful? . . . . 581\\n13.6 Lab: Multiple Testing . . . . . . . . . . . . . . . . . . . . . 583\\n13.6.1 Review of Hypothesis Tests . . . . . . . . . . . . . 583\\n13.6.2 Family-Wise Error Rate . . . . . . . . . . . . . . . 585\\n13.6.3 False Discovery Rate . . . . . . . . . . . . . . . . . 588\\n13.6.4 A Re-Sampling Approach . . . . . . . . . . . . . . 590\\n13.7 Exercises . . . . . . . . . . . . . . . . . . . . . . . . . . . . 593\\nIndex 597', metadata={'source': './documents/ISLP_website.pdf', 'page': 10}), Document(page_content='1\\nIntroduction\\nAn Overview of Statistical Learning\\nStatistical learning referstoavastsetoftoolsfor understanding data .These\\ntools can be classified as supervised orunsupervised . Broadly speaking,\\nsupervised statistical learning involves building a statistical model for pre-\\ndicting, or estimating, an outputbased on one or more inputs. Problems of\\nthisnatureoccurinfieldsasdiverseasbusiness,medicine,astrophysics,and\\npublic policy. With unsupervised statistical learning, there are inputs but\\nno supervising output; nevertheless we can learn relationships and struc-\\nture from such data. To provide an illustration of some applications of\\nstatistical learning, we briefly discuss three real-world data sets that are\\nconsidered in this book.\\nWage Data\\nIn this application (which we refer to as the Wagedata set throughout this\\nbook), we examine a number of factors that relate to wages for a group of\\nmen from the Atlantic region of the United States. In particular, we wish\\nto understand the association between an employee’s ageandeducation , as\\nwell as the calendar year, on his wage. Consider, for example, the left-hand\\npanel of Figure 1.1, which displays wageversusagefor each of the individu-\\nals in the data set. There is evidence that wageincreases with agebut then\\ndecreases again after approximately age 60. The blue line, which provides\\nan estimate of the average wagefor a given age, makes this trend clearer.\\nGiven an employee’s age, we can use this curve to predicthiswage. However,\\nit is also clear from Figure 1.1that there is a significant amount of vari-\\nability associated with this average value, and so agealone is unlikely to\\nprovide an accurate prediction of a particular man’s wage.\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning , Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_1  1', metadata={'source': './documents/ISLP_website.pdf', 'page': 11}), Document(page_content='2 1. Introduction\\n20 40 60 8050 100 200 300AgeWage2003 2006 200950 100 200 300YearWage1234550 100 200 300Education LevelWage\\nFIGURE 1.1. Wagedata, which contains income survey information for men\\nfrom the central Atlantic region of the United States. Left:wageas a function of\\nage. On average, wageincreases with ageuntil about 60years of age, at which\\npoint it begins to decline. Center:wageas a function of year. There is a slow\\nbut steady increase of approximately $10,000in the average wagebetween 2003\\nand2009.Right:Boxplots displaying wageas a function of education , with1\\nindicating the lowest level (no high school diploma) and 5the highest level (an\\nadvanced graduate degree). On average, wageincreases with the level of education.\\nWe also have information regarding each employee’s education level and\\ntheyearin which the wagewas earned. The center and right-hand panels\\nof Figure 1.1, which display wageas a function of both yearandeducation ,\\nindicate that both of these factors are associated with wage. Wages increase\\nby approximately $10,000, in a roughly linear (or straight-line) fashion,\\nbetween 2003 and2009 , though this rise is very slight relative to the vari-\\nability in the data. Wages are also typically greater for individuals with\\nhigher education levels: men with the lowest education level (1) tend to\\nhave substantially lower wages than those with the highest education level\\n(5). Clearly, the most accurate prediction of a given man’s wagewill be\\nobtained by combining his age, hiseducation , and the year. In Chapter 3,\\nwe discuss linear regression, which can be used to predict wagefrom this\\ndata set. Ideally, we should predict wagein a way that accounts for the\\nnon-linear relationship between wageandage. In Chapter 7, we discuss a\\nclass of approaches for addressing this problem.\\nStock Market Data\\nTheWagedatainvolvespredictinga continuous orquantitative outputvalue.\\nThis is often referred to as a regression problem. However, in certain cases\\nwemayinsteadwishtopredictanon-numericalvalue—thatis,a categorical\\norqualitative output. For example, in Chapter 4we examine a stock market\\ndata set that contains the daily movements in the Standard & Poor’s 500\\n(S&P) stock index over a 5-year period between 2001 and2005 . We refer\\nto this as the Smarketdata. The goal is to predict whether the index will\\nincrease ordecrease on a given day, using the past 5 days’ percentage\\nchanges in the index. Here the statistical learning problem does not involve\\npredicting a numerical value. Instead it involves predicting whether a given', metadata={'source': './documents/ISLP_website.pdf', 'page': 12}), Document(page_content='1. Introduction 3\\nDown Up−4 −2 0 2 4 6Yesterday\\nToday’s DirectionPercentage change in S&PDown Up−4 −2 0 2 4 6Two Days Previous\\nToday’s DirectionPercentage change in S&PDown Up−4 −2 0 2 4 6Three Days Previous\\nToday’s DirectionPercentage change in S&P\\nFIGURE 1.2. Left:Boxplots of the previous day’s percentage change in the S&P\\nindex for the days for which the market increased or decreased, obtained from the\\nSmarket data.Center and Right: Same as left panel, but the percentage changes\\nfor 2 and 3 days previous are shown.\\nday’s stock market performance will fall into the Upbucket or the Down\\nbucket. This is known as a classification problem. A model that could\\naccurately predict the direction in which the market will move would be\\nvery useful!\\nThe left-hand panel of Figure 1.2displays two boxplots of the previous\\nday’s percentage changes in the stock index: one for the 648days for which\\nthe market increased on the subsequent day, and one for the 602days for\\nwhich the market decreased. The two plots look almost identical, suggest-\\ning that there is no simple strategy for using yesterday’s movement in the\\nS&P to predict today’s returns. The remaining panels, which display box-\\nplots for the percentage changes 2 and 3 days previous to today, similarly\\nindicate little association between past and present returns. Of course, this\\nlack of pattern is to be expected: in the presence of strong correlations be-\\ntween successive days’ returns, one could adopt a simple trading strategy\\nto generate profits from the market. Nevertheless, in Chapter 4, we explore\\nthesedatausingseveraldifferentstatisticallearningmethods.Interestingly,\\nthere are hints of some weak trends in the data that suggest that, at least\\nfor this 5-year period, it is possible to correctly predict the direction of\\nmovement in the market approximately 60% of the time (Figure 1.3).\\nGene Expression Data\\nThe previous two applications illustrate data sets with both input and\\noutput variables. However, another important class of problems involves\\nsituations in which we only observe input variables, with no corresponding\\noutput. For example, in a marketing setting, we might have demographic\\ninformationforanumberofcurrentorpotentialcustomers.Wemaywishto\\nunderstand which types of customers are similar to each other by grouping\\nindividuals according to their observed characteristics. This is known as a', metadata={'source': './documents/ISLP_website.pdf', 'page': 13}), Document(page_content='4 1. Introduction\\nDownUp0.46 0.48 0.50 0.52Today’s DirectionPredicted ProbabilityFIGURE 1.3. We fit a quadratic discriminant analysis model to the subset\\nof theSmarket data corresponding to the 2001–2004 time period, and predicted\\nthe probability of a stock market decrease using the 2005 data. On average, the\\npredicted probability of decrease is higher for the days in which the market does\\ndecrease. Based on these results, we are able to correctly predict the direction of\\nmovement in the market 60% of the time.\\nclustering problem. Unlike in the previous examples, here we are not trying\\nto predict an output variable.\\nWe devote Chapter 12to a discussion of statistical learning methods\\nfor problems in which no natural output variable is available. We consider\\ntheNCI60data set, which consists of 6,830gene expression measurements\\nfor each of 64cancer cell lines. Instead of predicting a particular output\\nvariable, we are interested in determining whether there are groups, or\\nclusters, among the cell lines based on their gene expression measurements.\\nThis is a difficult question to address, in part because there are thousands\\nof gene expression measurements per cell line, making it hard to visualize\\nthe data.\\nThe left-hand panel of Figure 1.4addresses this problem by represent-\\ning each of the 64cell lines using just two numbers, Z1andZ2. These\\nare the first two principal components of the data, which summarize the\\n6,830expression measurements for each cell line down to two numbers or\\ndimensions . While it is likely that this dimension reduction has resulted in\\nsome loss of information, it is now possible to visually examine the data\\nfor evidence of clustering. Deciding on the number of clusters is often a\\ndifficult problem. But the left-hand panel of Figure 1.4suggests at least\\nfour groups of cell lines, which we have represented using separate colors.\\nIn this particular data set, it turns out that the cell lines correspond\\nto14different types of cancer. (However, this information was not used\\nto create the left-hand panel of Figure 1.4.) The right-hand panel of Fig-\\nure1.4is identical to the left-hand panel, except that the 14 cancer types\\nare shown using distinct colored symbols. There is clear evidence that cell\\nlines with the same cancer type tend to be located near each other in this\\ntwo-dimensional representation. In addition, even though the cancer infor-\\nmationwasnotusedtoproducetheleft-handpanel,theclusteringobtained\\ndoes bear some resemblance to some of the actual cancer types observed\\nin the right-hand panel. This provides some independent verification of the\\naccuracy of our clustering analysis.', metadata={'source': './documents/ISLP_website.pdf', 'page': 14}), Document(page_content='1. Introduction 5\\n−40 −20 0 20 40 60−60 −40 −20 0 20−40 −20 0 20 40 60−60 −40 −20 0 20Z1Z1\\nZ2Z2\\nFIGURE 1.4.Left:Representation of theNCI60gene expression data set ina two-dimensional space,Z1andZ2. Each point corresponds to one of the64cell lines. There appear to be four groups of cell lines, which we have representedusing different colors.Right:Same as left panel except that we have representedeach of the14different types of cancer using a different colored symbol. Cell linescorresponding to the same cancer type tend to be nearby in the two-dimensionalspace.A Brief History of Statistical LearningThoughthetermstatistical learningisfairlynew,manyoftheconceptsthatunderlie the field were developed long ago. At the beginning of the nine-teenth century, the method ofleast squareswas developed, implementingthe earliest form of what is now known aslinear regression. The approachwas first successfully applied to problems in astronomy. Linear regressionis used for predicting quantitative values, such as an individual’s salary.In order to predict qualitative values, such as whether a patient survivesor dies, or whether the stock market increases or decreases,linear discrim-inant analysiswas proposed in 1936. In the 1940s, various authors putforth an alternative approach,logistic regression. In the early 1970s, thetermgeneralized linear modelwas developed to describe an entire class ofstatistical learning methods that include both linear and logistic regressionas special cases.By the end of the 1970s, many more techniques for learning from datawere available. However, they were almost exclusivelylinearmethods be-cause fittingnon-linearrelationships was computationally difficult at thetime. By the 1980s, computing technology had finally improved sufficientlythat non-linear methods were no longer computationally prohibitive. Inthe mid 1980s,classification and regression treeswere developed, followedshortly bygeneralized additive models.Neural networksgained popularityin the 1980s, andsupport vector machinesarose in the 1990s.Since that time, statistical learning has emerged as a new subfield instatistics,focusedonsupervisedandunsupervisedmodelingandprediction.In recent years, progress in statistical learning has been marked by theincreasingavailabilityofpowerfulandrelativelyuser-friendlysoftware,suchas the popular and freely availablePythonsystem. This has the potential tocontinue the transformation of the field from a set of techniques used and', metadata={'source': './documents/ISLP_website.pdf', 'page': 15}), Document(page_content='6 1. Introductiondeveloped by statisticians and computer scientists to an essential toolkitfor a much broader community.This BookThe Elements of Statistical Learning(ESL) by Hastie, Tibshirani, andFriedman was first published in 2001. Since that time, it has become animportant reference on the fundamentals of statistical machine learning.Its success derives from its comprehensive and detailed treatment of manyimportant topics in statistical learning, as well as the fact that (relative tomany upper-level statistics textbooks) it is accessible to a wide audience.However, the greatest factor behind the success of ESL has been its topicalnature. At the time of its publication, interest in the field of statisticallearning was starting to explode. ESL provided one of the first accessibleand comprehensive introductions to the topic.Since ESL was first published, the field of statistical learning has con-tinued to flourish. The field’s expansion has taken two forms. The mostobvious growth has involved the development of new and improved statis-tical learning approaches aimed at answering a range of scientific questionsacross a number of fields. However, the field of statistical learning hasalso expanded its audience. In the 1990s, increases in computational powergenerated a surge of interest in the field from non-statisticians who wereeager to use cutting-edge statistical tools to analyze their data. Unfortu-nately, the highly technical nature of these approaches meant that the usercommunity remained primarily restricted to experts in statistics, computerscience, and related fields with the training (and time) to understand andimplement them.In recent years, new and improved software packages have significantlyeased the implementation burden for many statistical learning methods.At the same time, there has been growing recognition across a number offields, from business to health care to genetics to the social sciences andbeyond, that statistical learning is a powerful tool with important practicalapplications.Asaresult,thefieldhasmovedfromoneofprimarilyacademicinterest to a mainstream discipline, with an enormous potential audience.This trend will surely continue with the increasing availability of enormousquantities of data and the software to analyze it.The purpose ofAn Introduction to Statistical Learning(ISL) is to facili-tate the transition of statistical learning from an academic to a mainstreamfield. ISL is not intended to replace ESL, which is a far more comprehen-sive text both in terms of the number of approaches considered and thedepth to which they are explored. We consider ESL to be an importantcompanion for professionals (with graduate degrees in statistics, machinelearning, or related fields) who need to understand the technical detailsbehind statistical learning approaches. However, the community of users ofstatistical learning techniques has expanded to include individuals with awider range of interests and backgrounds. Therefore, there is a place for aless technical and more accessible version of ESL.', metadata={'source': './documents/ISLP_website.pdf', 'page': 16}), Document(page_content='1. Introduction 7In teaching these topics over the years, we have discovered that they areof interest to master’s and PhD students in fields as disparate as businessadministration, biology, and computer science, as well as to quantitatively-oriented upper-division undergraduates. It is important for this diversegroup to be able to understand the models, intuitions, and strengths andweaknesses of the various approaches. But for this audience, many of thetechnical details behind statistical learning methods, such as optimiza-tion algorithms and theoretical properties, are not of primary interest.We believe that these students do not need a deep understanding of theseaspectsinordertobecomeinformedusersofthevariousmethodologies,andin order to contribute to their chosen fields through the use of statisticallearning tools.ISL is based on the following four premises.1.Many statistical learning methods are relevant and useful in a widerange of academic and non-academic disciplines, beyond just the sta-tistical sciences.Webelievethatmanycontemporarystatisticallearn-ing procedures should, and will, become as widely available and usedas is currently the case for classical methods such as linear regres-sion. As a result, rather than attempting to consider every possibleapproach (an impossible task), we have concentrated on presentingthe methods that we believe are most widely applicable.2.Statistical learning should not be viewed as a series of black boxes.Nosingle approach will perform well in all possible applications. With-out understanding all of the cogs inside the box, or the interactionbetween those cogs, it is impossible to select the best box. Hence, wehave attempted to carefully describe the model, intuition, assump-tions, and trade-offs behind each of the methods that we consider.3.While it is important to know what job is performed by each cog, it isnot necessary to have the skills to construct the machine inside thebox!Thus, we have minimized discussion of technical details relatedto fitting procedures and theoretical properties. We assume that thereader is comfortable with basic mathematical concepts, but we donot assume a graduate degree in the mathematical sciences. For in-stance, we have almost completely avoided the use of matrix algebra,and it is possible to understand the entire book without a detailedknowledge of matrices and vectors.4.We presume that the reader is interested in applying statistical learn-ing methods to real-world problems.In order to facilitate this, as wellas to motivate the techniques discussed, we have devoted a sectionwithin each chapter to computer labs. In each lab, we walk the readerthrougharealisticapplicationofthemethodsconsideredinthatchap-ter. When we have taught this material in our courses, we have al-located roughly one-third of classroom time to working through thelabs, and we have found them to be extremely useful. Many of theless computationally-oriented students who were initially intimidatedby the labs got the hang of things over the course of the quarter orsemester. This book originally appeared (2013, second edition 2021)', metadata={'source': './documents/ISLP_website.pdf', 'page': 17}), Document(page_content='8 1. Introduction\\nwith computer labs written in the Rlanguage. Since then, there has\\nbeen increasing demand for Pythonimplementations of the impor-\\ntant techniques in statistical learning. Consequently, this version has\\nlabs inPython. There are a rapidly growing number of Pythonpack-\\nages available, and by examination of the imports at the beginning of\\neach lab, readers will see that we have carefully selected and used the\\nmost appropriate. We have also supplied some additional code and\\nfunctionality in our package ISLP. However, the labs in ISL are self-\\ncontained, and can be skipped if the reader wishes to use a different\\nsoftware package or does not wish to apply the methods discussed to\\nreal-world problems.\\nWho Should Read This Book?\\nThis book is intended for anyone who is interested in using modern statis-\\ntical methods for modeling and prediction from data. This group includes\\nscientists, engineers, data analysts, data scientists, and quants, but also less\\ntechnical individuals with degrees in non-quantitative fields such as the so-\\ncial sciences or business. We expect that the reader will have had at least\\none elementary course in statistics. Background in linear regression is also\\nuseful, though not required, since we review the key concepts behind linear\\nregression in Chapter 3. The mathematical level of this book is modest,\\nand a detailed knowledge of matrix operations is not required. This book\\nprovides an introduction to Python. Previous exposure to a programming\\nlanguage, such as MATLABorR, is useful but not required.\\nThe first edition of this textbook has been used to teach master’s and\\nPhD students in business, economics, computer science, biology, earth sci-\\nences, psychology, and many other areas of the physical and social sciences.\\nIt has also been used to teach advanced undergraduates who have already\\ntaken a course on linear regression. In the context of a more mathemat-\\nically rigorous course in which ESL serves as the primary textbook, ISL\\ncould be used as a supplementary text for teaching computational aspects\\nof the various approaches.\\nNotation and Simple Matrix Algebra\\nChoosing notation for a textbook is always a difficult task. For the most\\npart we adopt the same notational conventions as ESL.\\nWe will use nto represent the number of distinct data points, or observa-\\ntions, in our sample. We will let pdenote the number of variables that are\\navailable for use in making predictions. For example, the Wagedata set con-\\nsists of 11 variables for 3,000people, so we have n=3,000observations and\\np= 11 variables (such as year,age,race, and more). Note that throughout\\nthis book, we indicate variable names using colored font: Variable Name .\\nIn some examples, pmight be quite large, such as on the order of thou-\\nsands or even millions; this situation arises quite often, for example, in the\\nanalysis of modern biological data or web-based advertising data.', metadata={'source': './documents/ISLP_website.pdf', 'page': 18}), Document(page_content='1. Introduction 9In general, we will letxijrepresent the value of thejth variable for theith observation, wherei=1,2,...,nandj=1,2,...,p. Throughout thisbook,iwill be used to index the samples or observations (from1ton) andjwill be used to index the variables (from1top). We letXdenote ann×pmatrix whose(i, j)th element isxij. That is,X=\\uf8eb\\uf8ec\\uf8ec\\uf8ec\\uf8edx11x12... x1px21x22... x2p............xn1xn2... xnp\\uf8f6\\uf8f7\\uf8f7\\uf8f7\\uf8f8.For readers who are unfamiliar with matrices, it is useful to visualizeXasa spreadsheet of numbers withnrows andpcolumns.At times we will be interested in the rows ofX, which we write asx1,x2,...,xn. Herexiis a vector of lengthp, containing thepvariablemeasurements for theith observation. That is,xi=\\uf8eb\\uf8ec\\uf8ec\\uf8ec\\uf8edxi1xi2...xip\\uf8f6\\uf8f7\\uf8f7\\uf8f7\\uf8f8.(1.1)(Vectors are by default represented as columns.) For example, for theWagedata,xiis a vector of length 11, consisting ofyear,age,race, and othervalues for theith individual. At other times we will instead be interestedin the columns ofX, which we write asx1,x2,...,xp. Each is a vector oflengthn. That is,xj=\\uf8eb\\uf8ec\\uf8ec\\uf8ec\\uf8edx1jx2j...xnj\\uf8f6\\uf8f7\\uf8f7\\uf8f7\\uf8f8.For example, for theWagedata,x1contains then=3,000values foryear.Using this notation, the matrixXcan be written asX=(x1x2···xp),orX=\\uf8eb\\uf8ec\\uf8ec\\uf8ec\\uf8edxT1xT2...xTn\\uf8f6\\uf8f7\\uf8f7\\uf8f7\\uf8f8.TheTnotationdenotesthetransposeofamatrixorvector.So,forexample,XT=\\uf8eb\\uf8ec\\uf8ec\\uf8ec\\uf8edx11x21... xn1x12x22... xn2.........x1px2p... xnp\\uf8f6\\uf8f7\\uf8f7\\uf8f7\\uf8f8,', metadata={'source': './documents/ISLP_website.pdf', 'page': 19}), Document(page_content='10 1. Introduction\\nwhile\\nxT\\ni=(xi1xi2···xip)\\n.\\nWe useyito denote the ith observation of the variable on which we\\nwish to make predictions, such as wage. Hence, we write the set of all n\\nobservations in vector form as\\ny=\\uf8eb\\n\\uf8ec\\uf8ec\\uf8ec\\uf8edy1\\ny2\\n...\\nyn\\uf8f6\\n\\uf8f7\\uf8f7\\uf8f7\\uf8f8.\\nThen our observed data consists of {(x1,y1),(x2,y2),...,(xn,yn)}, where\\neachxiis a vector of length p. (Ifp=1, thenxiis simply a scalar.)\\nIn this text, a vector of length nwill always be denoted in lower case\\nbold; e.g.\\na=\\uf8eb\\n\\uf8ec\\uf8ec\\uf8ec\\uf8eda1\\na2\\n...\\nan\\uf8f6\\n\\uf8f7\\uf8f7\\uf8f7\\uf8f8.\\nHowever, vectors that are not of length n(such as feature vectors of length\\np, as in (1.1)) will be denoted in lower case normal font , e.g.a. Scalars will\\nalso be denoted in lower case normal font , e.g.a. In the rare cases in which\\nthese two uses for lower case normal font lead to ambiguity, we will clarify\\nwhich use is intended. Matrices will be denoted using bold capitals , such\\nasA. Random variables will be denoted using capital normal font , e.g.A,\\nregardless of their dimensions.\\nOccasionallywewillwanttoindicatethedimensionofaparticularobject.\\nTo indicate that an object is a scalar, we will use the notation a∈R.T o\\nindicate that it is a vector of length k, we will use a∈Rk(ora∈Rnif\\nit is of length n). We will indicate that an object is an r×smatrix using\\nA∈Rr×s.\\nWe have avoided using matrix algebra whenever possible. However, in\\na few instances it becomes too cumbersome to avoid it entirely. In these\\nrare instances it is important to understand the concept of multiplying\\ntwo matrices. Suppose that A∈Rr×dandB∈Rd×s. Then the product\\nofAandBis denoted AB. The(i, j)th element of ABis computed by\\nmultiplying each element of the ith row of Aby the corresponding element\\nof thejth column of B. That is, (AB)ij=∑d\\nk=1aikbkj. As an example,\\nconsider\\nA=(12\\n34)\\nandB=(56\\n78)\\n.\\nThen\\nAB=(12\\n34)(56\\n78)\\n=(1×5+2 ×71 ×6+2 ×8\\n3×5+4 ×73 ×6+4 ×8)\\n=(19 22\\n43 50)\\n.\\nNote that this operation produces an r×smatrix. It is only possible to\\ncompute ABif the number of columns of Ais the same as the number of\\nrows ofB.', metadata={'source': './documents/ISLP_website.pdf', 'page': 20}), Document(page_content='1. Introduction 11\\nOrganization of This Book\\nChapter 2introduces the basic terminology and concepts behind statisti-\\ncal learning. This chapter also presents the K-nearest neighbor classifier, a\\nvery simple method that works surprisingly well on many problems. Chap-\\nters3and4cover classical linear methods for regression and classification.\\nIn particular, Chapter 3reviewslinear regression , the fundamental start-\\ning point for all regression methods. In Chapter 4we discuss two of the\\nmost important classical classification methods, logistic regression andlin-\\near discriminant analysis .\\nA central problem in all statistical learning situations involves choosing\\nthe best method for a given application. Hence, in Chapter 5we intro-\\nducecross-validation and thebootstrap , which can be used to estimate the\\naccuracy of a number of different methods in order to choose the best one.\\nMuch of the recent research in statistical learning has concentrated on\\nnon-linear methods. However, linear methods often have advantages over\\ntheir non-linear competitors in terms of interpretability and sometimes also\\naccuracy. Hence, in Chapter 6we consider a host of linear methods, both\\nclassical and more modern, which offer potential improvements over stan-\\ndard linear regression. These include stepwise selection ,ridge regression ,\\nprincipal components regression , and the lasso.\\nThe remaining chapters move into the world of non-linear statistical\\nlearning. We first introduce in Chapter 7a number of non-linear meth-\\nods that work well for problems with a single input variable. We then\\nshow how these methods can be used to fit non-linear additive models for\\nwhich there is more than one input. In Chapter 8, we investigate tree-based\\nmethods, including bagging,boosting, andrandom forests .Support vector\\nmachines , a set of approaches for performing both linear and non-linear\\nclassification, are discussed in Chapter 9. We cover deep learning , an ap-\\nproach for non-linear regression and classification that has received a lot\\nof attention in recent years, in Chapter 10. Chapter 11explores survival\\nanalysis, a regression approach that is specialized to the setting in which\\nthe output variable is censored, i.e. not fully observed.\\nIn Chapter 12, we consider the unsupervised setting in which we have\\ninput variables but no output variable. In particular, we present princi-\\npal components analysis ,K-means clustering , andhierarchical clustering .\\nFinally, in Chapter 13we cover the very important topic of multiple hy-\\npothesis testing.\\nAt the end of each chapter, we present one or more Pythonlab sections\\nin which we systematically work through applications of the various meth-\\nods discussed in that chapter. These labs demonstrate the strengths and\\nweaknesses of the various approaches, and also provide a useful reference\\nfor the syntax required to implement the various methods. The reader may\\nchoose to work through the labs at their own pace, or the labs may be the\\nfocus of group sessions as part of a classroom environment. Within each\\nPythonlab, we present the results that we obtained when we performed the\\nlab at the time of writing this book. However, new versions of Pythonare\\ncontinuously released, and over time, the packages called in the labs will\\nbe updated. Therefore, in the future, it is possible that the results shown in', metadata={'source': './documents/ISLP_website.pdf', 'page': 21}), Document(page_content='12 1. Introduction\\nName Description\\nAuto Gas mileage, horsepower, and other information for cars.\\nBikeshare Hourly usage of a bike sharing program in Washington, DC.\\nBoston Housing values and other information about Boston census tracts.\\nBrainCancer Survival times for patients diagnosed with brain cancer.\\nCaravan Information about individuals offered caravan insurance.\\nCarseats Information about car seat sales in 400 stores.\\nCollege Demographic characteristics, tuition, and more for USA colleges.\\nCredit Information about credit card debt for 400 customers.\\nDefault Customer default records for a credit card company.\\nFund Returns of 2,000 hedge fund managers over 50 months.\\nHitters Records and salaries for baseball players.\\nKhan Gene expression measurements for four cancer types.\\nNCI60 Gene expression measurements for 64 cancer cell lines.\\nNYSE Returns, volatility, and volume for the New York Stock Exchange.\\nOJ Sales information for Citrus Hill and Minute Maid orange juice.\\nPortfolio Past values of financial assets, for use in portfolio allocation.\\nPublication Time to publication for 244 clinical trials.\\nSmarket Daily percentage returns for S&P 500 over a 5-year period.\\nUSArrests Crime statistics per 100,000 residents in 50 states of USA.\\nWage Income survey data for men in central Atlantic region of USA.\\nWeekly 1,089 weekly stock market returns for 21 years.\\nTABLE 1.1. A list of data sets needed to perform the labs and exercises in this\\ntextbook. All data sets are available in the ISLPpackage, with the exception of\\nUSArrests , which is part of the base Rdistribution, but accessible from Python.\\nthe lab sections may no longer correspond precisely to the results obtained\\nby the reader who performs the labs. As necessary, we will post updates to\\nthe labs on the book website.\\nWe use the\\nsymbol to denote sections or exercises that contain more\\nchallenging concepts. These can be easily skipped by readers who do not\\nwish to delve as deeply into the material, or who lack the mathematical\\nbackground.\\nData Sets Used in Labs and Exercises\\nIn this textbook, we illustrate statistical learning methods using applica-\\ntions from marketing, finance, biology, and other areas. The ISLPpackage\\ncontains a number of data sets that are required in order to perform the\\nlabs and exercises associated with this book. One other data set is part of\\nthe base Rdistribution (the USArrests data), and we show how to access it\\nfromPythonin Section 12.5.1. Table1.1contains a summary of the data\\nsets required to perform the labs and exercises. A couple of these data sets\\nare also available as text files on the book website, for use in Chapter 2.', metadata={'source': './documents/ISLP_website.pdf', 'page': 22}), Document(page_content='1. Introduction 13\\nBook Website\\nThe website for this book is located at\\nwww.statlearning.com\\nIt contains a number of resources, including the Pythonpackage associated\\nwith this book, and some additional data sets.\\nAcknowledgements\\nA few of the plots in this book were taken from ESL: Figures 6.7,8.3,\\nand12.14. All other plots were produced for the Rversion of ISL, except\\nfor Figure 13.10which differs because of the Pythonsoftware supporting\\nthe plot.', metadata={'source': './documents/ISLP_website.pdf', 'page': 23}), Document(page_content='2\\nStatistical Learning\\n2.1 What Is Statistical Learning?\\nInordertomotivateourstudyofstatisticallearning,webeginwithasimple\\nexample. Suppose that we are statistical consultants hired by a client to\\ninvestigate the association between advertising and sales of a particular\\nproduct. The Advertising data set consists of the salesof that product\\nin200different markets, along with advertising budgets for the product in\\neach of those markets for three different media: TV,radio, andnewspaper .\\nThe data are displayed in Figure 2.1. It is not possible for our client to\\ndirectly increase sales of the product. On the other hand, they can control\\nthe advertising expenditure in each of the three media. Therefore, if we\\ndetermine that there is an association between advertising and sales, then\\nwe can instruct our client to adjust advertising budgets, thereby indirectly\\nincreasing sales. In other words, our goal is to develop an accurate model\\nthat can be used to predict sales on the basis of the three media budgets.\\nIn this setting, the advertising budgets are input variables whilesalesinput\\nvariableis anoutput variable . The input variables are typically denoted using the\\noutput\\nvariablesymbolX, with a subscript to distinguish them. So X1might be the TV\\nbudget,X2theradiobudget, and X3thenewspaper budget. The inputs\\ngo by different names, such as predictors ,independent variables ,features,predictor\\nindependent\\nvariable\\nfeatureor sometimes just variables . The output variable—in this case, sales—is\\nvariableoften called the response ordependent variable , and is typically denoted\\nresponse\\ndependent\\nvariableusing the symbol Y. Throughout this book, we will use all of these terms\\ninterchangeably.\\nMore generally, suppose that we observe a quantitative response Yandp\\ndifferent predictors, X1,X2,...,X p. We assume that there is some\\nrelationship between YandX=(X1,X2,...,X p), which can be written\\nin the very general form\\nY=f(X)+ϵ. (2.1)\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning , Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_2  15', metadata={'source': './documents/ISLP_website.pdf', 'page': 24}), Document(page_content='16 2. Statistical Learning\\n0 50 100 200 3005 10 15 20 25TVSales\\n0 10 20 30 40 505 10 15 20 25RadioSales\\n0 20 40 60 80 1005 10 15 20 25NewspaperSales\\nFIGURE 2.1. TheAdvertising data set. The plot displays sales, in thousands\\nof units, as a function of TV,radio, andnewspaper budgets, in thousands of\\ndollars, for 200different markets. In each plot we show the simple least squares\\nfit ofsalesto that variable, as described in Chapter 3. In other words, each blue\\nline represents a simple model that can be used to predict salesusingTV,radio,\\nandnewspaper , respectively.\\nHerefissomefixedbutunknownfunctionof X1,...,X p,andϵisarandom\\nerror term , which is independent of Xand has mean zero. In this formula-error termtion,frepresents the systematic information that Xprovides about Y.systematic\\n1012141618202220304050607080\\nYears of EducationIncome\\n1012141618202220304050607080\\nYears of EducationIncome\\nFIGURE 2.2. TheIncomedata set. Left:The red dots are the observed values\\nofincome(in thousands of dollars) and years of education for30individuals.\\nRight:The blue curve represents the true underlying relationship between income\\nandyears of education , which is generally unknown (but is known in this case\\nbecause the data were simulated). The black lines represent the error associated\\nwith each observation. Note that some errors are positive (if an observation lies\\nabove the blue curve) and some are negative (if an observation lies below the\\ncurve). Overall, these errors have approximately mean zero.\\nAs another example, consider the left-hand panel of Figure 2.2, a plot of\\nincomeversusyears of education for30individuals in the Incomedata set.\\nThe plot suggests that one might be able to predict incomeusingyears of\\neducation . However, the function fthat connects the input variable to the', metadata={'source': './documents/ISLP_website.pdf', 'page': 25}), Document(page_content='2.1 What Is Statistical Learning? 17\\noutput variable is in general unknown. In this situation one must estimate\\nfbased on the observed points. Since Incomeis a simulated data set, fis\\nknown and is shown by the blue curve in the right-hand panel of Figure 2.2.\\nThe vertical lines represent the error terms ϵ. We note that some of the\\n30 observations lie above the blue curve and some lie below it; overall, the\\nerrors have approximately mean zero.\\nIn general, the function fmay involve more than one input variable.\\nIn Figure 2.3we plot incomeas a function of years of education and\\nseniority . Herefis a two-dimensional surface that must be estimated\\nbased on the observed data.\\nIn essence, statistical learning refers to a set of approaches for estimating\\nf. In this chapter we outline some of the key theoretical concepts that arise\\nin estimating f, as well as tools for evaluating the estimates obtained.\\n2.1.1 Why Estimate f?\\nThere are two main reasons that we may wish to estimate f:prediction\\nandinference . We discuss each in turn.\\nPrediction\\nIn many situations, a set of inputs Xare readily available, but the output\\nYcannot be easily obtained. In this setting, since the error term averages\\nto zero, we can predict Yusing\\nˆY=ˆf(X), (2.2)\\nwhereˆfrepresents our estimate for f, andˆYrepresents the resulting pre-\\ndiction for Y. In this setting, ˆfis often treated as a black box , in the sense\\nthat one is not typically concerned with the exact form of ˆf, provided that\\nit yields accurate predictions for Y.\\nAs an example, suppose that X1,...,X pare characteristics of a patient’s\\nblood sample that can be easily measured in a lab, and Yis a variable\\nencoding the patient’s risk for a severe adverse reaction to a particular\\ndrug. It is natural to seek to predict YusingX, since we can then avoid\\ngiving the drug in question to patients who are at high risk of an adverse\\nreaction—that is, patients for whom the estimate of Yis high.\\nThe accuracy of ˆYas a prediction for Ydepends on two quantities,\\nwhich we will call the reducible error and theirreducible error . In general,reducible\\nerror\\nirreducible\\nerrorˆfwill not be a perfect estimate for f, and this inaccuracy will introduce\\nsome error. This error is reducible because we can potentially improve the\\naccuracyof ˆfbyusingthemostappropriatestatisticallearningtechniqueto\\nestimate f. However, even if it were possible to form a perfect estimate for\\nf, so that our estimated response took the form ˆY=f(X), our prediction\\nwould still have some error in it! This is because Yis also a function of\\nϵ, which, by definition, cannot be predicted using X. Therefore, variability\\nassociated with ϵalso affects the accuracy of our predictions. This is known\\nas theirreducible error, because no matter how well we estimate f,w e\\ncannot reduce the error introduced by ϵ.\\nWhy is the irreducible error larger than zero? The quantity ϵmay con-\\ntain unmeasured variables that are useful in predicting Y: since we don’t', metadata={'source': './documents/ISLP_website.pdf', 'page': 26}), Document(page_content='18 2. Statistical Learning\\nYears of EducationSeniorityIncome\\nFIGURE 2.3.The plot displaysincomeas a function ofyears of educationandseniorityin theIncomedata set. The blue surface represents the trueunderlying relationship betweenincomeandyears of educationandseniority,which is known since the data are simulated. The red dots indicate the observedvalues of these quantities for30individuals.measure them,fcannot use them for its prediction. The quantityϵmayalso contain unmeasurable variation. For example, the risk of an adversereaction might vary for a given patient on a given day, depending onmanufacturing variation in the drug itself or the patient’s general feelingof well-being on that day.Consider a given estimateˆfand a set of predictorsX, which yields thepredictionˆY=ˆf(X). Assume for a moment that bothˆfandXare fixed,so that the only variability comes fromϵ. Then, it is easy to show thatE(Y−ˆY)2= E[f(X)+ϵ−ˆf(X)]2=[f(X)−ˆf(X)]2\\ued19\\ued18\\ued17\\ued1aReducible+ Var(ϵ)\\ued19\\ued18\\ued17\\ued1aIrreducible,(2.3)whereE(Y−ˆY)2represents the average, orexpected value, of the squaredexpectedvaluedifference between the predicted and actual value ofY, andVar(ϵ)repre-sents thevarianceassociated with the error termϵ.varianceThe focus of this book is on techniques for estimatingfwith the aim ofminimizing the reducible error. It is important to keep in mind that theirreducible error will always provide an upper bound on the accuracy ofour prediction forY. This bound is almost always unknown in practice.InferenceWe are often interested in understanding the association betweenYandX1,...,Xp. In this situation we wish to estimatef, but our goal is notnecessarily to make predictions forY. Nowˆfcannot be treated as a blackbox, because we need to know its exact form. In this setting, one may beinterested in answering the following questions:', metadata={'source': './documents/ISLP_website.pdf', 'page': 27}), Document(page_content='2.1 What Is Statistical Learning? 19\\n•Which predictors are associated with the response? It is often the case\\nthat only a small fraction of the available predictors are substantially\\nassociated with Y. Identifying the few important predictors among a\\nlarge set of possible variables can be extremely useful, depending on\\nthe application.\\n•What is the relationship between the response and each predictor?\\nSome predictors may have a positive relationship with Y, in the sense\\nthat larger values of the predictor are associated with larger values of\\nY. Other predictors may have the opposite relationship. Depending\\non the complexity of f, the relationship between the response and a\\ngiven predictor may also depend on the values of the other predictors.\\n•Can the relationship between Yand each predictor be adequately sum-\\nmarized using a linear equation, or is the relationship more compli-\\ncated?Historically, most methods for estimating fhave taken a linear\\nform.Insomesituations,suchanassumptionisreasonableorevende-\\nsirable. But often the true relationship is more complicated, in which\\ncase a linear model may not provide an accurate representation of\\nthe relationship between the input and output variables.\\nInthisbook,wewillseeanumberofexamplesthatfallintotheprediction\\nsetting, the inference setting, or a combination of the two.\\nFor instance, consider a company that is interested in conducting a\\ndirect-marketing campaign. The goal is to identify individuals who are\\nlikely to respond positively to a mailing, based on observations of demo-\\ngraphic variables measured on each individual. In this case, the demo-\\ngraphic variables serve as predictors, and response to the marketing cam-\\npaign (either positive or negative) serves as the outcome. The company is\\nnot interested in obtaining a deep understanding of the relationships be-\\ntween each individual predictor and the response; instead, the company\\nsimply wants to accurately predict the response using the predictors. This\\nis an example of modeling for prediction.\\nIn contrast, consider the Advertising data illustrated in Figure 2.1. One\\nmay be interested in answering questions such as:\\n–Which media are associated with sales?\\n–Which media generate the biggest boost in sales? or\\n–How large of an increase in sales is associated with a given increase\\nin TV advertising?\\nThis situation falls into the inference paradigm. Another example involves\\nmodeling the brand of a product that a customer might purchase based on\\nvariables such as price, store location, discount levels, competition price,\\nand so forth. In this situation one might really be most interested in the\\nassociation between each variable and the probability of purchase. For in-\\nstance,to what extent is the product’s price associated with sales? This is\\nan example of modeling for inference.\\nFinally, some modeling could be conducted both for prediction and in-\\nference. For example, in a real estate setting, one may seek to relate values', metadata={'source': './documents/ISLP_website.pdf', 'page': 28}), Document(page_content='20 2. Statistical Learning\\nof homes to inputs such as crime rate, zoning, distance from a river, air\\nquality, schools, income level of community, size of houses, and so forth. In\\nthis case one might be interested in the association between each individ-\\nual input variable and housing price—for instance, how much extra will a\\nhouse be worth if it has a view of the river? This is an inference problem.\\nAlternatively, one may simply be interested in predicting the value of a\\nhome given its characteristics: is this house under- or over-valued? This is\\na prediction problem.\\nDepending on whether our ultimate goal is prediction, inference, or a\\ncombination of the two, different methods for estimating fmay be ap-\\npropriate. For example, linear models allow for relatively simple and in-linear modelterpretable inference, but may not yield as accurate predictions as some\\nother approaches. In contrast, some of the highly non-linear approaches\\nthat we discuss in the later chapters of this book can potentially provide\\nquite accurate predictions for Y, but this comes at the expense of a less\\ninterpretable model for which inference is more challenging.\\n2.1.2 How Do We Estimate f?\\nThroughout this book, we explore many linear and non-linear approaches\\nfor estimating f. However, these methods generally share certain charac-\\nteristics. We provide an overview of these shared characteristics in this\\nsection. We will always assume that we have observed a set of ndifferent\\ndata points. For example in Figure 2.2we observed n= 30 data points.\\nThese observations are called the training data because we will use thesetraining\\ndataobservations to train, or teach, our method how to estimate f. Letxij\\nrepresent the value of the jth predictor, or input, for observation i, where\\ni=1,2,...,n andj=1,2,...,p . Correspondingly, let yirepresent the\\nresponse variable for the ith observation. Then our training data consist of\\n{(x1,y1),(x2,y2),...,(xn,yn)}wherexi=(xi1,xi2,...,x ip)T.\\nOur goal is to apply a statistical learning method to the training data\\nin order to estimate the unknown function f. In other words, we want to\\nfind a function ˆfsuch that Y≈ˆf(X)for any observation (X,Y). Broadly\\nspeaking, most statistical learning methods for this task can be character-\\nized as either parametric ornon-parametric . We now briefly discuss theseparametric\\nnon-\\nparametrictwo types of approaches.\\nParametric Methods\\nParametric methods involve a two-step model-based approach.\\n1. First, we make an assumption about the functional form, or shape,\\noff. For example, one very simple assumption is that fis linear in\\nX:\\nf(X)=β0+β1X1+β2X2+···+βpXp. (2.4)\\nThis is a linear model , which will be discussed extensively in Chap-\\nter3. Once we have assumed that fis linear, the problem of estimat-\\ningfis greatly simplified. Instead of having to estimate an entirely\\narbitrary p-dimensional function f(X), one only needs to estimate\\nthep+1coefficients β0,β1,...,βp.', metadata={'source': './documents/ISLP_website.pdf', 'page': 29}), Document(page_content='2.1 What Is Statistical Learning? 21\\nYears of EducationSeniorityIncome\\nFIGURE 2.4. A linear model fit by least squares to the Incomedata from\\nFigure2.3. The observations are shown in red, and the yellow plane indicates the\\nleast squares fit to the data.\\n2. After a model has been selected, we need a procedure that uses the\\ntraining data to fitortrainthe model. In the case of the linear modelfit\\ntrain(2.4), we need to estimate the parameters β0,β1,...,βp. That is, we\\nwant to find values of these parameters such that\\nY≈β0+β1X1+β2X2+···+βpXp.\\nThe most common approach to fitting the model ( 2.4) is referred to\\nas(ordinary) least squares , which we discuss in Chapter 3. However,least squaresleast squares is one of many possible ways to fit the linear model. In\\nChapter 6, we discuss other approaches for estimating the parameters\\nin (2.4).\\nThe model-based approach just described is referred to as parametric ;\\nit reduces the problem of estimating fdown to one of estimating a set of\\nparameters. Assuming a parametric form for fsimplifies the problem of\\nestimating fbecause it is generally much easier to estimate a set of pa-\\nrameters, such as β0,β1,...,βpin the linear model ( 2.4), than it is to fit\\nan entirely arbitrary function f. The potential disadvantage of a paramet-\\nric approach is that the model we choose will usually not match the true\\nunknown form of f. If the chosen model is too far from the true f, then\\nour estimate will be poor. We can try to address this problem by choos-\\ningflexiblemodels that can fit many different possible functional formsflexibleforf. But in general, fitting a more flexible model requires estimating a\\ngreater number of parameters. These more complex models can lead to a\\nphenomenon known as overfitting the data, which essentially means theyoverfittingfollow the errors, or noise, too closely. These issues are discussed through-noiseout this book.\\nFigure2.4shows an example of the parametric approach applied to the\\nIncomedata from Figure 2.3. We have fit a linear model of the form\\nincome ≈β0+β1×education +β2×seniority .', metadata={'source': './documents/ISLP_website.pdf', 'page': 30}), Document(page_content='22 2. Statistical Learning\\nYears of EducationSeniorityIncome\\nFIGURE 2.5. A smooth thin-plate spline fit to the Incomedata from Figure 2.3\\nis shown in yellow; the observations are displayed in red. Splines are discussed in\\nChapter 7.\\nSince we have assumed a linear relationship between the response and the\\ntwo predictors, the entire fitting problem reduces to estimating β0,β1, and\\nβ2, which we do using least squares linear regression. Comparing Figure 2.3\\nto Figure 2.4, we can see that the linear fit given in Figure 2.4is not quite\\nright: the true fhas some curvature that is not captured in the linear fit.\\nHowever, the linear fit still appears to do a reasonable job of capturing the\\npositive relationship between years of education andincome, as well as the\\nslightly less positive relationship between seniority andincome. It may be\\nthat with such a small number of observations, this is the best we can do.\\nNon-Parametric Methods\\nNon-parametric methods do not make explicit assumptions about the func-\\ntional form of f. Instead they seek an estimate of fthat gets as close to the\\ndata points as possible without being too rough or wiggly. Such approaches\\ncan have a major advantage over parametric approaches: by avoiding the\\nassumption of a particular functional form for f, they have the potential\\nto accurately fit a wider range of possible shapes for f. Any parametric\\napproach brings with it the possibility that the functional form used to\\nestimate fis very different from the true f, in which case the resulting\\nmodel will not fit the data well. In contrast, non-parametric approaches\\ncompletely avoid this danger, since essentially no assumption about the\\nform offis made. But non-parametric approaches do suffer from a major\\ndisadvantage: since they do not reduce the problem of estimating fto a\\nsmall number of parameters, a very large number of observations (far more\\nthan is typically needed for a parametric approach) is required in order to\\nobtain an accurate estimate for f.\\nAn example of a non-parametric approach to fitting the Incomedata is\\nshown in Figure 2.5.Athin-plate spline is used to estimate f. This ap-thin-plate\\nsplineproach does not impose any pre-specified model on f. It instead attempts', metadata={'source': './documents/ISLP_website.pdf', 'page': 31}), Document(page_content='2.1 What Is Statistical Learning? 23\\nYears of EducationSeniorityIncome\\nFIGURE 2.6. A rough thin-plate spline fit to the Incomedata from Figure 2.3.\\nThis fit makes zero errors on the training data.\\nto produce an estimate for fthat is as close as possible to the observed\\ndata, subject to the fit—that is, the yellow surface in Figure 2.5—being\\nsmooth. In this case, the non-parametric fit has produced a remarkably ac-\\ncurateestimateofthetrue fshowninFigure 2.3.Inordertofitathin-plate\\nspline, the data analyst must select a level of smoothness. Figure 2.6shows\\nthe same thin-plate spline fit using a lower level of smoothness, allowing\\nfor a rougher fit. The resulting estimate fits the observed data perfectly!\\nHowever, the spline fit shown in Figure 2.6is far more variable than the\\ntrue function f, from Figure 2.3. This is an example of overfitting the\\ndata, which we discussed previously. It is an undesirable situation because\\nthe fit obtained will not yield accurate estimates of the response on new\\nobservations that were not part of the original training data set. We dis-\\ncuss methods for choosing the correctamount of smoothness in Chapter 5.\\nSplines are discussed in Chapter 7.\\nAs we have seen, there are advantages and disadvantages to parametric\\nand non-parametric methods for statistical learning. We explore both types\\nof methods throughout this book.\\n2.1.3 The Trade-Off Between Prediction Accuracy and Model\\nInterpretability\\nOf the many methods that we examine in this book, some are less flexible,\\nor more restrictive, in the sense that they can produce just a relatively\\nsmall range of shapes to estimate f. For example, linear regression is a\\nrelatively inflexible approach, because it can only generate linear functions\\nsuch as the lines shown in Figure 2.1or the plane shown in Figure 2.4.\\nOther methods, such as the thin plate splines shown in Figures 2.5and2.6,\\nare considerably more flexible because they can generate a much wider\\nrange of possible shapes to estimate f.', metadata={'source': './documents/ISLP_website.pdf', 'page': 32}), Document(page_content='24 2. Statistical Learning\\nFlexibilityInterpretability\\nLowHighLowHighSubset SelectionLassoLeast SquaresGeneralized Additive ModelsTreesBagging, BoostingSupport Vector MachinesDeep LearningFIGURE 2.7. A representation of the tradeoff between flexibility and inter-\\npretability, using different statistical learning methods. In general, as the flexibility\\nof a method increases, its interpretability decreases.\\nOne might reasonably ask the following question: why would we ever\\nchoose to use a more restrictive method instead of a very flexible approach?\\nThere are several reasons that we might prefer a more restrictive model.\\nIf we are mainly interested in inference, then restrictive models are much\\nmore interpretable. For instance, when inference is the goal, the linear\\nmodel may be a good choice since it will be quite easy to understand\\nthe relationship between YandX1,X2,...,X p. In contrast, very flexible\\napproaches, such as the splines discussed in Chapter 7and displayed in\\nFigures2.5and2.6, and the boosting methods discussed in Chapter 8, can\\nlead to such complicated estimates of fthat it is difficult to understand\\nhow any individual predictor is associated with the response.\\nFigure2.7provides an illustration of the trade-off between flexibility and\\ninterpretability for some of the methods that we cover in this book. Least\\nsquares linear regression, discussed in Chapter 3, is relatively inflexible but\\nis quite interpretable. The lasso, discussed in Chapter 6, relies upon thelassolinear model ( 2.4) but uses an alternative fitting procedure for estimating\\nthe coefficients β0,β1,...,βp. The new procedure is more restrictive in es-\\ntimating the coefficients, and sets a number of them to exactly zero. Hence\\nin this sense the lasso is a less flexible approach than linear regression.\\nIt is also more interpretable than linear regression, because in the final\\nmodel the response variable will only be related to a small subset of the\\npredictors—namely, those with nonzero coefficient estimates. Generalized\\nadditive models (GAMs), discussed in Chapter 7, instead extend the lin-generalized\\nadditive\\nmodelear model ( 2.4) to allow for certain non-linear relationships. Consequently,\\nGAMs are more flexible than linear regression. They are also somewhat\\nless interpretable than linear regression, because the relationship between\\neach predictor and the response is now modeled using a curve. Finally,\\nfully non-linear methods such as bagging,boosting,support vector machinesbagging\\nboostingwith non-linear kernels, and neural networks (deep learning), discussed in\\nsupport\\nvector\\nmachineChapters 8,9, and10, are highly flexible approaches that are harder to\\ninterpret.', metadata={'source': './documents/ISLP_website.pdf', 'page': 33}), Document(page_content='2.1 What Is Statistical Learning? 25\\nWe have established that when inference is the goal, there are clear ad-\\nvantages to using simple and relatively inflexible statistical learning meth-\\nods. In some settings, however, we are only interested in prediction, and\\nthe interpretability of the predictive model is simply not of interest. For\\ninstance, if we seek to develop an algorithm to predict the price of a\\nstock, our sole requirement for the algorithm is that it predict accurately—\\ninterpretability is not a concern. In this setting, we might expect that it\\nwill be best to use the most flexible model available. Surprisingly, this is\\nnot always the case! We will often obtain more accurate predictions using\\na less flexible method. This phenomenon, which may seem counterintu-\\nitive at first glance, has to do with the potential for overfitting in highly\\nflexible methods. We saw an example of overfitting in Figure 2.6. We will\\ndiscuss this very important concept further in Section 2.2and throughout\\nthis book.\\n2.1.4 Supervised Versus Unsupervised Learning\\nMost statistical learning problems fall into one of two categories: supervisedsupervisedorunsupervised . The examples that we have discussed so far in this chap-unsupervisedter all fall into the supervised learning domain. For each observation of the\\npredictor measurement(s) xi,i=1,...,n there is an associated response\\nmeasurement yi. We wish to fit a model that relates the response to the\\npredictors, with the aim of accurately predicting the response for future\\nobservations (prediction) or better understanding the relationship between\\nthe response and the predictors (inference). Many classical statistical learn-\\ning methods such as linear regression and logistic regression (Chapter 4), aslogistic\\nregressionwell as more modern approaches such as GAM, boosting, and support vec-\\ntor machines, operate in the supervised learning domain. The vast majority\\nof this book is devoted to this setting.\\nBy contrast, unsupervised learning describes the somewhat more chal-\\nlenging situation in which for every observation i=1,...,n , we observe\\na vector of measurements xibut no associated response yi. It is not pos-\\nsible to fit a linear regression model, since there is no response variable\\nto predict. In this setting, we are in some sense working blind; the sit-\\nuation is referred to as unsupervised because we lack a response vari-\\nable that can supervise our analysis. What sort of statistical analysis is\\npossible?Wecanseektounderstandtherelationshipsbetweenthevariables\\nor between the observations. One statistical learning tool that we may use\\nin this setting is cluster analysis , or clustering. The goal of cluster analysiscluster\\nanalysisis to ascertain, on the basis of x1,...,x n, whether the observations fall into\\nrelatively distinct groups. For example, in a market segmentation study we\\nmight observe multiple characteristics (variables) for potential customers,\\nsuch as zip code, family income, and shopping habits. We might believe\\nthat the customers fall into different groups, such as big spenders versus\\nlow spenders. If the information about each customer’s spending patterns\\nwere available, then a supervised analysis would be possible. However, this\\ninformation is not available—that is, we do not know whether each poten-\\ntial customer is a big spender or not. In this setting, we can try to cluster\\nthe customers on the basis of the variables measured, in order to identify', metadata={'source': './documents/ISLP_website.pdf', 'page': 34}), Document(page_content='26 2. Statistical Learning\\n0 2 4 6 8 10 122 4 6 8 10 1202462468X1X1\\nX2X2\\nFIGURE 2.8. A clustering data set involving three groups. Each group is\\nshown using a different colored symbol. Left:The three groups are well-separated.\\nIn this setting, a clustering approach should successfully identify the three groups.\\nRight:There is some overlap among the groups. Now the clustering task is more\\nchallenging.\\ndistinct groups of potential customers. Identifying such groups can be of\\ninterest because it might be that the groups differ with respect to some\\nproperty of interest, such as spending habits.\\nFigure2.8provides a simple illustration of the clustering problem. We\\nhave plotted 150 observations with measurements on two variables, X1\\nandX2. Each observation corresponds to one of three distinct groups. For\\nillustrative purposes, we have plotted the members of each group using\\ndifferent colors and symbols. However, in practice the group memberships\\nare unknown, and the goal is to determine the group to which each obser-\\nvation belongs. In the left-hand panel of Figure 2.8, this is a relatively easy\\ntask because the groups are well-separated. By contrast, the right-hand\\npanel illustrates a more challenging setting in which there is some overlap\\nbetween the groups. A clustering method could not be expected to assign\\nall of the overlapping points to their correct group (blue, green, or orange).\\nIn the examples shown in Figure 2.8, there are only two variables, and\\nso one can simply visually inspect the scatterplots of the observations in\\norder to identify clusters. However, in practice, we often encounter data\\nsets that contain many more than two variables. In this case, we cannot\\neasily plot the observations. For instance, if there are pvariables in our\\ndata set, then p(p−1)/2distinct scatterplots can be made, and visual\\ninspection is simply not a viable way to identify clusters. For this reason,\\nautomated clustering methods are important. We discuss clustering and\\nother unsupervised learning approaches in Chapter 12.\\nMany problems fall naturally into the supervised or unsupervised learn-\\ning paradigms. However, sometimes the question of whether an analysis\\nshould be considered supervised or unsupervised is less clear-cut. For in-\\nstance, suppose that we have a set of nobservations. For mof the observa-\\ntions, where m<n , we have both predictor measurements and a response', metadata={'source': './documents/ISLP_website.pdf', 'page': 35}), Document(page_content='2.2 Assessing Model Accuracy 27\\nmeasurement. For the remaining n−mobservations, we have predictor\\nmeasurements but no response measurement. Such a scenario can arise if\\nthe predictors can be measured relatively cheaply but the corresponding\\nresponses are much more expensive to collect. We refer to this setting as\\nasemi-supervised learning problem. In this setting, we wish to use a sta-semi-\\nsupervised\\nlearningtistical learning method that can incorporate the mobservations for which\\nresponse measurements are available as well as the n−mobservations for\\nwhich they are not. Although this is an interesting topic, it is beyond the\\nscope of this book.\\n2.1.5 Regression Versus Classification Problems\\nVariables can be characterized as either quantitative orqualitative (alsoquantitative\\nqualitativeknown as categorical ). Quantitative variables take on numerical values. Ex-\\ncategoricalamples include a person’s age, height, or income, the value of a house, and\\nthe price of a stock. In contrast, qualitative variables take on values in\\none ofKdifferent classes, or categories. Examples of qualitative variablesclassinclude a person’s marital status (married or not), the brand of product\\npurchased (brand A, B, or C), whether a person defaults on a debt (yes\\nor no), or a cancer diagnosis (Acute Myelogenous Leukemia, Acute Lym-\\nphoblastic Leukemia, or No Leukemia). We tend to refer to problems with\\na quantitative response as regression problems, while those involving aregressionqualitative response are often referred to as classification problems. How-classificationever, the distinction is not always that crisp. Least squares linear regression\\n(Chapter 3)isusedwithaquantitativeresponse,whereaslogisticregression\\n(Chapter 4) is typically used with a qualitative (two-class, or binary) re-binarysponse.Thus,despiteitsname,logisticregressionisaclassificationmethod.\\nBut since it estimates class probabilities, it can be thought of as a regres-\\nsion method as well. Some statistical methods, such as K-nearest neighbors\\n(Chapters 2and4) and boosting (Chapter 8), can be used in the case of\\neither quantitative or qualitative responses.\\nWe tend to select statistical learning methods on the basis of whether\\nthe response is quantitative or qualitative; i.e. we might use linear regres-\\nsion when quantitative and logistic regression when qualitative. However,\\nwhether the predictors are qualitative or quantitative is generally consid-\\nered less important. Most of the statistical learning methods discussed in\\nthis book can be applied regardless of the predictor variable type, provided\\nthat any qualitative predictors are properly codedbefore the analysis is\\nperformed. This is discussed in Chapter 3.\\n2.2 Assessing Model Accuracy\\nOne of the key aims of this book is to introduce the reader to a wide range\\nof statistical learning methods that extend far beyond the standard linear\\nregression approach. Why is it necessary to introduce so many different\\nstatisticallearningapproaches,ratherthanjustasingle bestmethod? There\\nis no free lunch in statistics: no one method dominates all others over all\\npossible data sets. On a particular data set, one specific method may work', metadata={'source': './documents/ISLP_website.pdf', 'page': 36}), Document(page_content='28 2. Statistical Learning\\nbest, but some other method may work better on a similar but different\\ndata set. Hence it is an important task to decide for any given set of data\\nwhich method produces the best results. Selecting the best approach can\\nbe one of the most challenging parts of performing statistical learning in\\npractice.\\nIn this section, we discuss some of the most important concepts that\\narise in selecting a statistical learning procedure for a specific data set. As\\nthe book progresses, we will explain how the concepts presented here can\\nbe applied in practice.\\n2.2.1 Measuring the Quality of Fit\\nIn order to evaluate the performance of a statistical learning method on\\na given data set, we need some way to measure how well its predictions\\nactually match the observed data. That is, we need to quantify the extent\\nto which the predicted response value for a given observation is close to\\nthe true response value for that observation. In the regression setting, the\\nmost commonly-used measure is the mean squared error (MSE), given bymean\\nsquared\\nerrorMSE =1\\nnn∑\\ni=1(yi−ˆf(xi))2, (2.5)\\nwhereˆf(xi)is the prediction that ˆfgives for the ith observation. The MSE\\nwill be small if the predicted responses are very close to the true responses,\\nand will be large if for some of the observations, the predicted and true\\nresponses differ substantially.\\nThe MSE in ( 2.5) is computed using the training data that was used to\\nfit the model, and so should more accurately be referred to as the training\\nMSE. But in general, we do not really care how well the method workstraining\\nMSEon the training data. Rather, we are interested in the accuracy of the pre-\\ndictions that we obtain when we apply our method to previously unseen\\ntest data. Why is this what we care about? Suppose that we are interestedtest datain developing an algorithm to predict a stock’s price based on previous\\nstock returns. We can train the method using stock returns from the past\\n6months.Butwedon’treallycarehowwellourmethodpredictslastweek’s\\nstock price. We instead care about how well it will predict tomorrow’s price\\nor next month’s price. On a similar note, suppose that we have clinical\\nmeasurements (e.g. weight, blood pressure, height, age, family history of\\ndisease)foranumberofpatients,aswellasinformationaboutwhethereach\\npatient has diabetes. We can use these patients to train a statistical learn-\\ning method to predict risk of diabetes based on clinical measurements. In\\npractice, we want this method to accurately predict diabetes risk for future\\npatients based on their clinical measurements. We are not very interested\\nin whether or not the method accurately predicts diabetes risk for patients\\nused to train the model, since we already know which of those patients\\nhave diabetes.\\nTo state it more mathematically, suppose that we fit our statistical learn-\\ning method on our training observations {(x1,y1),(x2,y2),...,(xn,yn)},\\nandweobtaintheestimate ˆf.Wecanthencompute ˆf(x1),ˆf(x2),..., ˆf(xn).', metadata={'source': './documents/ISLP_website.pdf', 'page': 37}), Document(page_content='2.2 Assessing Model Accuracy 29\\n0 20 40 60 80 1002 4 6 8 10 12XY\\n2 5 10 200.0 0.5 1.0 1.5 2.0 2.5FlexibilityMean Squared Error\\nFIGURE 2.9. Left:Data simulated from f, shown in black. Three estimates of\\nfare shown: the linear regression line (orange curve), and two smoothing spline\\nfits (blue and green curves). Right:Training MSE (grey curve), test MSE (red\\ncurve), and minimum possible test MSE over all methods (dashed line). Squares\\nrepresent the training and test MSEs for the three fits shown in the left-hand\\npanel.\\nIf these are approximately equal to y1,y2,...,y n, then the training MSE\\ngiven by ( 2.5) is small. However, we are really not interested in whether\\nˆf(xi)≈yi; instead, we want to know whether ˆf(x0)is approximately equal\\ntoy0, where(x0,y0)is apreviously unseen test observation not used to train\\nthe statistical learning method . We want to choose the method that gives\\nthelowest test MSE ,asopposedtothelowesttrainingMSE.Inotherwords,test MSEif we had a large number of test observations, we could compute\\nAve(y0−ˆf(x0))2, (2.6)\\nthe average squared prediction error for these test observations (x0,y0).\\nWe’d like to select the model for which this quantity is as small as possible.\\nHow can we go about trying to select a method that minimizes the test\\nMSE? In some settings, we may have a test data set available—that is,\\nwe may have access to a set of observations that were not used to train\\nthe statistical learning method. We can then simply evaluate ( 2.6) on the\\ntest observations, and select the learning method for which the test MSE is\\nsmallest. But what if no test observations are available? In that case, one\\nmightimaginesimplyselectingastatisticallearningmethodthatminimizes\\nthe training MSE ( 2.5). This seems like it might be a sensible approach,\\nsince the training MSE and the test MSE appear to be closely related.\\nUnfortunately, there is a fundamental problem with this strategy: there\\nis no guarantee that the method with the lowest training MSE will also\\nhave the lowest test MSE. Roughly speaking, the problem is that many\\nstatistical methods specifically estimate coefficients so as to minimize the\\ntraining set MSE. For these methods, the training set MSE can be quite\\nsmall, but the test MSE is often much larger.\\nFigure2.9illustrates this phenomenon on a simple example. In the left-\\nhand panel of Figure 2.9, we have generated observations from ( 2.1) with', metadata={'source': './documents/ISLP_website.pdf', 'page': 38}), Document(page_content='30 2. Statistical Learning\\nthetruefgivenbythe blackcurve.The orange, blue andgreen curvesillus-\\ntrate three possible estimates for fobtained using methods with increasing\\nlevels of flexibility. The orange line is the linear regression fit, which is rela-\\ntively inflexible. The blue and green curves were produced using smoothing\\nsplines, discussed in Chapter 7, with different levels of smoothness. It issmoothing\\nsplineclear that as the level of flexibility increases, the curves fit the observed\\ndata more closely. The green curve is the most flexible and matches the\\ndata very well; however, we observe that it fits the true f(shown in black)\\npoorly because it is too wiggly. By adjusting the level of flexibility of the\\nsmoothing spline fit, we can produce many different fits to this data.\\nWe now move on to the right-hand panel of Figure 2.9. The grey curve\\ndisplays the average training MSE as a function of flexibility, or more\\nformally the degrees of freedom , for a number of smoothing splines. Thedegrees of\\nfreedomdegrees of freedom is a quantity that summarizes the flexibility of a curve;\\nit is discussed more fully in Chapter 7. The orange, blue and green squares\\nindicate the MSEs associated with the corresponding curves in the left-\\nhand panel. A more restricted and hence smoother curve has fewer degrees\\nof freedom than a wiggly curve—note that in Figure 2.9, linear regression\\nis at the most restrictive end, with two degrees of freedom. The training\\nMSE declines monotonically as flexibility increases. In this example the\\ntruefis non-linear, and so the orange linear fit is not flexible enough to\\nestimate fwell. The green curve has the lowest training MSE of all three\\nmethods, since it corresponds to the most flexible of the three curves fit in\\nthe left-hand panel.\\nIn this example, we know the true function f, and so we can also com-\\npute the test MSE over a very large test set, as a function of flexibility. (Of\\ncourse, in general fis unknown, so this will not be possible.) The test MSE\\nis displayed using the red curve in the right-hand panel of Figure 2.9. As\\nwith the training MSE, the test MSE initially declines as the level of flex-\\nibility increases. However, at some point the test MSE levels off and then\\nstarts to increase again. Consequently, the orange and green curves both\\nhave high test MSE. The blue curve minimizes the test MSE, which should\\nnot be surprising given that visually it appears to estimate fthe best in the\\nleft-hand panel of Figure 2.9. The horizontal dashed line indicates Var (ϵ),\\nthe irreducible error in ( 2.3), which corresponds to the lowest achievable\\ntest MSE among all possible methods. Hence, the smoothing spline repre-\\nsented by the blue curve is close to optimal.\\nIn the right-hand panel of Figure 2.9, as the flexibility of the statistical\\nlearning method increases, we observe a monotone decrease in the training\\nMSE and a U-shape in the test MSE. This is a fundamental property of\\nstatistical learning that holds regardless of the particular data set at hand\\nand regardless of the statistical method being used. As model flexibility\\nincreases, the training MSE will decrease, but the test MSE may not. When\\na given method yields a small training MSE but a large test MSE, we are\\nsaidtobe overfitting thedata.Thishappensbecauseourstatisticallearning\\nprocedure is working too hard to find patterns in the training data, and\\nmay be picking up some patterns that are just caused by random chance\\nrather than by true properties of the unknown function f. When we overfit\\nthe training data, the test MSE will be very large because the supposed', metadata={'source': './documents/ISLP_website.pdf', 'page': 39}), Document(page_content='2.2 Assessing Model Accuracy 31\\n0 20 40 60 80 1002 4 6 8 10 12XY\\n2 5 10 200.0 0.5 1.0 1.5 2.0 2.5FlexibilityMean Squared Error\\nFIGURE 2.10. Details are as in Figure 2.9, using a different true fthat is\\nmuch closer to linear. In this setting, linear regression provides a very good fit to\\nthe data.\\npatterns that the method found in the training data simply don’t exist\\nin the test data. Note that regardless of whether or not overfitting has\\noccurred, we almost always expect the training MSE to be smaller than\\nthe test MSE because most statistical learning methods either directly or\\nindirectly seek to minimize the training MSE. Overfitting refers specifically\\nto the case in which a less flexible model would have yielded a smaller\\ntest MSE.\\nFigure2.10provides another example in which the true fis approxi-\\nmately linear. Again we observe that the training MSE decreases mono-\\ntonically as the model flexibility increases, and that there is a U-shape in\\nthe test MSE. However, because the truth is close to linear, the test MSE\\nonly decreases slightly before increasing again, so that the orange least\\nsquares fit is substantially better than the highly flexible green curve. Fi-\\nnally, Figure 2.11displays an example in which fis highly non-linear. The\\ntraining and test MSE curves still exhibit the same general patterns, but\\nnow there is a rapid decrease in both curves before the test MSE starts to\\nincrease slowly.\\nIn practice, one can usually compute the training MSE with relative\\nease, but estimating the test MSE is considerably more difficult because\\nusuallynotestdata areavailable.Asthe previousthreeexamples illustrate,\\nthe flexibility level corresponding to the model with the minimal test MSE\\ncan vary considerably among data sets. Throughout this book, we discuss a\\nvarietyofapproachesthatcanbeusedinpracticetoestimatethisminimum\\npoint. One important method is cross-validation (Chapter 5), which is across-\\nvalidation method for estimating the test MSE using the training data.\\n2.2.2 The Bias-Variance Trade-Off\\nThe U-shape observed in the test MSE curves (Figures 2.9–2.11) turns out\\nto be the result of two competing properties of statistical learning methods.', metadata={'source': './documents/ISLP_website.pdf', 'page': 40}), Document(page_content='32 2. Statistical Learning\\n0 20 40 60 80 100−10 0 10 20XY\\n2 5 10 200 5 10 15 20FlexibilityMean Squared Error\\nFIGURE 2.11. Details are as in Figure 2.9, using a different fthat is far from\\nlinear. In this setting, linear regression provides a very poor fit to the data.\\nThough the mathematical proof is beyond the scope of this book, it is\\npossible to show that the expected test MSE, for a given value x0, can\\nalways be decomposed into the sum of three fundamental quantities: the\\nvariance ofˆf(x0), the squared biasofˆf(x0)and the variance of the errorvariance\\nbiastermsϵ. That is,\\nE(\\ny0−ˆf(x0))2\\n=Var(ˆf(x0)) + [ Bias(ˆf(x0))]2+Var(ϵ). (2.7)\\nHere the notation E(\\ny0−ˆf(x0))2\\ndefines the expected test MSE atx0,expected\\ntest MSE and refers to the average test MSE that we would obtain if we repeatedly\\nestimated fusingalargenumberoftrainingsets,andtestedeachat x0.The\\noverall expected test MSE can be computed by averaging E(\\ny0−ˆf(x0))2\\nover all possible values of x0in the test set.\\nEquation 2.7tells us that in order to minimize the expected test error,\\nwe need to select a statistical learning method that simultaneously achieves\\nlow variance andlow bias. Note that variance is inherently a nonnegative\\nquantity, and squared bias is also nonnegative. Hence, we see that the\\nexpected test MSE can never lie below Var (ϵ), the irreducible error from\\n(2.3).\\nWhat do we mean by the variance andbiasof a statistical learning\\nmethod? Variance refers to the amount by which ˆfwould change if we\\nestimated it using a different training data set. Since the training data\\nare used to fit the statistical learning method, different training data sets\\nwill result in a different ˆf. But ideally the estimate for fshould not vary\\ntoo much between training sets. However, if a method has high variance\\nthen small changes in the training data can result in large changes in ˆf. In\\ngeneral,moreflexiblestatisticalmethodshavehighervariance.Considerthe\\ngreen and orange curves in Figure 2.9. The flexible green curve is following\\nthe observations very closely. It has high variance because changing any\\none of these data points may cause the estimate ˆfto change considerably.', metadata={'source': './documents/ISLP_website.pdf', 'page': 41}), Document(page_content='2.2 Assessing Model Accuracy 33\\n2 5 10 200.0 0.5 1.0 1.5 2.0 2.5Flexibility2 5 10 200.0 0.5 1.0 1.5 2.0 2.5Flexibility2 5 10 2005 10 15 20\\nFlexibilityMSEBiasVar\\nFIGURE 2.12. Squared bias (blue curve), variance (orange curve), Var (ϵ)\\n(dashed line), and test MSE (red curve) for the three data sets in Figures 2.9–2.11.\\nThe vertical dotted line indicates the flexibility level corresponding to the smallest\\ntest MSE.\\nIn contrast, the orange least squares line is relatively inflexible and has low\\nvariance, because moving any single observation will likely cause only a\\nsmall shift in the position of the line.\\nOn the other hand, biasrefers to the error that is introduced by approxi-\\nmatingareal-lifeproblem,whichmaybeextremelycomplicated,byamuch\\nsimpler model. For example, linear regression assumes that there is a linear\\nrelationship between YandX1,X2,...,X p. It is unlikely that any real-life\\nproblem truly has such a simple linear relationship, and so performing lin-\\near regression will undoubtedly result in some bias in the estimate of f. In\\nFigure2.11, the true fis substantially non-linear, so no matter how many\\ntraining observations we are given, it will not be possible to produce an\\naccurate estimate using linear regression. In other words, linear regression\\nresults in high bias in this example. However, in Figure 2.10the true f\\nis very close to linear, and so given enough data, it should be possible for\\nlinear regression to produce an accurate estimate. Generally, more flexible\\nmethods result in less bias.\\nAs a general rule, as we use more flexible methods, the variance will\\nincrease and the bias will decrease. The relative rate of change of these\\ntwo quantities determines whether the test MSE increases or decreases. As\\nwe increase the flexibility of a class of methods, the bias tends to initially\\ndecrease faster than the variance increases. Consequently, the expected\\ntest MSE declines. However, at some point increasing flexibility has little\\nimpact on the bias but starts to significantly increase the variance. When\\nthis happens the test MSE increases. Note that we observed this pattern\\nof decreasing test MSE followed by increasing test MSE in the right-hand\\npanels of Figures 2.9–2.11.\\nThe three plots in Figure 2.12illustrate Equation 2.7for the examples in\\nFigures2.9–2.11. In each case the blue solid curve represents the squared\\nbias, for different levels of flexibility, while the orange curve corresponds to\\nthe variance. The horizontal dashed line represents Var (ϵ), the irreducible\\nerror. Finally, the red curve, corresponding to the test set MSE, is the sum', metadata={'source': './documents/ISLP_website.pdf', 'page': 42}), Document(page_content='34 2. Statistical Learning\\nof these three quantities. In all three cases, the variance increases and the\\nbias decreases as the method’s flexibility increases. However, the flexibility\\nlevel corresponding to the optimal test MSE differs considerably among the\\nthree data sets, because the squared bias and variance change at different\\nrates in each of the data sets. In the left-hand panel of Figure 2.12, the\\nbias initially decreases rapidly, resulting in an initial sharp decrease in the\\nexpected test MSE. On the other hand, in the center panel of Figure 2.12\\nthe truefis close to linear, so there is only a small decrease in bias as flex-\\nibility increases, and the test MSE only declines slightly before increasing\\nrapidly as the variance increases. Finally, in the right-hand panel of Fig-\\nure2.12, as flexibility increases, there is a dramatic decline in bias because\\nthe true fis very non-linear. There is also very little increase in variance\\nas flexibility increases. Consequently, the test MSE declines substantially\\nbefore experiencing a small increase as model flexibility increases.\\nThe relationship between bias, variance, and test set MSE given in Equa-\\ntion2.7and displayed in Figure 2.12is referred to as the bias-variance\\ntrade-off. Good test set performance of a statistical learning method re-bias-variance\\ntrade-offquires low variance as well as low squared bias. This is referred to as a\\ntrade-off because it is easy to obtain a method with extremely low bias but\\nhigh variance (for instance, by drawing a curve that passes through every\\nsingle training observation) or a method with very low variance but high\\nbias (by fitting a horizontal line to the data). The challenge lies in finding\\na method for which both the variance and the squared bias are low. This\\ntrade-off is one of the most important recurring themes in this book.\\nIn a real-life situation in which fis unobserved, it is generally not pos-\\nsible to explicitly compute the test MSE, bias, or variance for a statistical\\nlearning method. Nevertheless, one should always keep the bias-variance\\ntrade-off in mind. In this book we explore methods that are extremely\\nflexible and hence can essentially eliminate bias. However, this does not\\nguarantee that they will outperform a much simpler method such as linear\\nregression. To take an extreme example, suppose that the true fis linear.\\nIn this situation linear regression will have no bias, making it very hard\\nfor a more flexible method to compete. In contrast, if the true fis highly\\nnon-linear and we have an ample number of training observations, then\\nwe may do better using a highly flexible approach, as in Figure 2.11. In\\nChapter 5we discuss cross-validation, which is a way to estimate the test\\nMSE using the training data.\\n2.2.3 The Classification Setting\\nThus far, our discussion of model accuracy has been focused on the regres-\\nsion setting. But many of the concepts that we have encountered, such\\nas the bias-variance trade-off, transfer over to the classification setting\\nwith only some modifications due to the fact that yiis no longer quan-\\ntitative. Suppose that we seek to estimate fon the basis of training obser-\\nvations{(x1,y1),...,(xn,yn)}, where now y1,...,y nare qualitative. The\\nmost common approach for quantifying the accuracy of our estimate ˆfis\\nthetraining error rate ,theproportionofmistakesthataremadeifweapplyerror rate', metadata={'source': './documents/ISLP_website.pdf', 'page': 43}), Document(page_content='2.2 Assessing Model Accuracy 35\\nour estimate ˆfto the training observations:\\n1\\nnn∑\\ni=1I(yi̸=ˆyi). (2.8)\\nHereˆyiis the predicted class label for the ith observation using ˆf. And\\nI(yi̸=ˆyi)is anindicator variable that equals 1ifyi̸=ˆyiand zero if yi=ˆyi.indicator\\nvariableIfI(yi̸=ˆyi)=0 then the ith observation was classified correctly by our\\nclassification method; otherwise it was misclassified. Hence Equation 2.8\\ncomputes the fraction of incorrect classifications.\\nEquation 2.8is referred to as the training error rate because it is com-training\\nerrorputed based on the data that was used to train our classifier. As in the\\nregression setting, we are most interested in the error rates that result from\\napplying our classifier to test observations that were not used in training.\\nThetest error rate associated with a set of test observations of the formtest error(x0,y0)is given by\\nAve(I(y0̸=ˆy0)), (2.9)\\nwhereˆy0is the predicted class label that results from applying the classifier\\nto the test observation with predictor x0.Agoodclassifier is one for which\\nthe test error ( 2.9) is smallest.\\nThe Bayes Classifier\\nIt is possible to show (though the proof is outside of the scope of this\\nbook) that the test error rate given in ( 2.9) is minimized, on average, by a\\nvery simple classifier that assigns each observation to the most likely class,\\ngiven its predictor values . In other words, we should simply assign a test\\nobservation with predictor vector x0to the class jfor which\\nPr(Y=j|X=x0) (2.10)\\nis largest. Note that ( 2.10) is aconditional probability : it is the probabilityconditional\\nprobabilitythatY=j, given the observed predictor vector x0. This very simple clas-\\nsifier is called the Bayes classifier . In a two-class problem where there areBayes\\nclassifieronly two possible response values, say class 1orclass 2, the Bayes classifier\\ncorresponds to predicting class one if Pr(Y=1|X=x0)>0.5, and class\\ntwo otherwise.\\nFigure2.13provides an example using a simulated data set in a two-\\ndimensional space consisting of predictors X1andX2. The orange and\\nblue circles correspond to training observations that belong to two different\\nclasses. For each value of X1andX2, there is a different probability of the\\nresponse being orange or blue. Since this is simulated data, we know how\\nthe data were generated and we can calculate the conditional probabilities\\nfor each value of X1andX2. The orange shaded region reflects the set of\\npoints for which Pr(Y=orange|X)is greater than 50 % , while the blue\\nshaded region indicates the set of points for which the probability is below\\n50 % . The purple dashed line represents the points where the probability\\nis exactly 50 % . This is called the Bayes decision boundary . The BayesBayes\\ndecision\\nboundaryclassifier’s prediction is determined by the Bayes decision boundary; an\\nobservation that falls on the orange side of the boundary will be assigned', metadata={'source': './documents/ISLP_website.pdf', 'page': 44}), Document(page_content='36 2. Statistical Learning\\nooooo\\noooooo\\noooo\\noo\\noooo\\nooooooooooooooooooo\\noooo\\noooooooooooooooooooooooooooooooooooooooooooooooo\\nooooooooo\\nooooooooooooooooooooooo\\noooooooooooo\\nooooooooooooooooooooooooooooooooooooooooooo\\nooooooooooooooo\\noooooo\\nX1X2\\nFIGURE 2.13. A simulated data set consisting of 100observations in each of\\ntwo groups, indicated in blue and in orange. The purple dashed line represents\\nthe Bayes decision boundary. The orange background grid indicates the region\\nin which a test observation will be assigned to the orange class, and the blue\\nbackground grid indicates the region in which a test observation will be assigned\\nto the blue class.\\nto the orange class, and similarly an observation on the blue side of the\\nboundary will be assigned to the blue class.\\nThe Bayes classifier produces the lowest possible test error rate, called\\ntheBayes error rate . Since the Bayes classifier will always choose the classBayes error\\nratefor which ( 2.10) is largest, the error rate will be 1−max jPr(Y=j|X=x0)\\natX=x0. In general, the overall Bayes error rate is given by\\n1−E(\\nmax\\njPr(Y=j|X))\\n, (2.11)\\nwhere the expectation averages the probability over all possible values of\\nX. For our simulated data, the Bayes error rate is 0.133. It is greater than\\nzero, because the classes overlap in the true population, which implies that\\nmax jPr(Y=j|X=x0)<1for some values of x0. The Bayes error rate is\\nanalogous to the irreducible error, discussed earlier.\\nK-Nearest Neighbors\\nIn theory we would always like to predict qualitative responses using the\\nBayes classifier. But for real data, we do not know the conditional distri-\\nbution of YgivenX, and so computing the Bayes classifier is impossi-\\nble. Therefore, the Bayes classifier serves as an unattainable gold standard\\nagainst which to compare other methods. Many approaches attempt to\\nestimate the conditional distribution of YgivenX, and then classify a\\ngiven observation to the class with highest estimated probability. One such\\nmethod is the K-nearest neighbors (KNN) classifier. Given a positive in-K-nearest\\nneighbors', metadata={'source': './documents/ISLP_website.pdf', 'page': 45}), Document(page_content='2.2 Assessing Model Accuracy 37\\ntegerKand a test observation x0, the KNN classifier first identifies the\\nKpoints in the training data that are closest to x0, represented by N0.\\nIt then estimates the conditional probability for class jas the fraction of\\npoints in N0whose response values equal j:\\nPr(Y=j|X=x0)=1\\nK∑\\ni∈N0I(yi=j). (2.12)\\nFinally, KNN classifies the test observation x0to the class with the largest\\nprobability from ( 2.12).\\nFigure2.14provides an illustrative example of the KNN approach. In\\nthe left-hand panel, we have plotted a small training data set consisting of\\nsix blue and six orange observations. Our goal is to make a prediction for\\nthe point labeled by the black cross. Suppose that we choose K=3. Then\\nKNN will first identify the three observations that are closest to the cross.\\nThis neighborhood is shown as a circle. It consists of two blue points and\\none orange point, resulting in estimated probabilities of 2/3for the blue\\nclass and 1/3for the orange class. Hence KNN will predict that the black\\ncross belongs to the blue class. In the right-hand panel of Figure 2.14we\\nhave applied the KNN approach with K=3at all of the possible values for\\nX1andX2, and have drawn in the corresponding KNN decision boundary.\\nDespite the fact that it is a very simple approach, KNN can often pro-\\nduce classifiers that are surprisingly close to the optimal Bayes classifier.\\nFigure2.15displays the KNN decision boundary, using K= 10 , when ap-\\nplied to the larger simulated data set from Figure 2.13. Notice that even\\nthough the true distribution is not known by the KNN classifier, the KNN\\ndecision boundary is very close to that of the Bayes classifier. The test error\\nrate using KNN is 0.1363 , which is close to the Bayes error rate of 0.1304 .\\nThe choice of Khas a drastic effect on the KNN classifier obtained.\\nFigure2.16displays two KNN fits to the simulated data from Figure 2.13,\\nusingK=1andK= 100 . WhenK=1, the decision boundary is overly\\nflexible and finds patterns in the data that don’t correspond to the Bayes\\ndecision boundary. This corresponds to a classifier that has low bias but\\nvery high variance. As Kgrows, the method becomes less flexible and\\nproduces a decision boundary that is close to linear. This corresponds to\\na low-variance but high-bias classifier. On this simulated data set, neither\\nK=1 norK= 100 give good predictions: they have test error rates of\\n0.1695 and0.1925 , respectively.\\nJust as in the regression setting, there is not a strong relationship be-\\ntween the training error rate and the test error rate. With K=1, the\\nKNN training error rate is 0, but the test error rate may be quite high. In\\ngeneral, as we use more flexible classification methods, the training error\\nrate will decline but the test error rate may not. In Figure 2.17, we have\\nplotted the KNN test and training errors as a function of 1/K. As1/Kin-\\ncreases, the method becomes more flexible. As in the regression setting, the\\ntrainingerrorrateconsistentlydeclinesastheflexibilityincreases.However,\\nthe test error exhibits a characteristic U-shape, declining at first (with a\\nminimum at approximately K= 10 ) before increasing again when the\\nmethod becomes excessively flexible and overfits.', metadata={'source': './documents/ISLP_website.pdf', 'page': 46}), Document(page_content='38 2. Statistical Learning\\noo\\nooooooo\\nooooo\\nooooooo\\noooFIGURE 2.14. The KNN approach, using K=3, is illustrated in a simple\\nsituation with six blue observations and six orange observations. Left:a test\\nobservation at which a predicted class label is desired is shown as a black cross.\\nThe three closest points to the test observation are identified, and it is predicted\\nthat the test observation belongs to the most commonly-occurring class, in this\\ncase blue. Right:The KNN decision boundary for this example is shown in black.\\nThe blue grid indicates the region in which a test observation will be assigned to\\nthe blue class, and the orange grid indicates the region in which it will be assigned\\nto the orange class.\\nooooo\\noooooo\\noooo\\noo\\noooo\\nooooooooooooooooooo\\noooo\\noooooooooooooooooooooooooooooooooooooooooooooooo\\nooooooooo\\nooooooooooooooooooooooo\\noooooooooooo\\nooooooooooooooooooooooooooooooooooooooooooo\\nooooooooooooooo\\noooooo\\nX1X2KNN: K=10\\nFIGURE 2.15. The black curve indicates the KNN decision boundary on the\\ndata from Figure 2.13, usingK= 10 . The Bayes decision boundary is shown as\\na purple dashed line. The KNN and Bayes decision boundaries are very similar.\\nIn both the regression and classification settings, choosing the correct\\nlevelofflexibilityiscriticaltothesuccessofanystatisticallearningmethod.\\nThe bias-variance tradeoff, and the resulting U-shape in the test error, can\\nmake this a difficult task. In Chapter 5, we return to this topic and discuss', metadata={'source': './documents/ISLP_website.pdf', 'page': 47}), Document(page_content='2.2 Assessing Model Accuracy 39\\nooooo\\noooooo\\noooo\\noo\\noooo\\noooooooooo\\nooooooooo\\noooo\\noooooooooooooooooooooo\\nooooooooooo\\noooooooo\\nooooooo\\nooo\\noooooo\\nooooooooooooooooooooooo\\noooooooooooo\\nooooooooooooooooo\\noo\\noooooooooooooooooooooooo\\nooooooooooooooo\\nooooooooooo\\noooooo\\noooo\\noo\\noooo\\noooooooooo\\nooooooooo\\noooo\\noooooooooooooooooooooo\\nooooooooooo\\noooooooo\\nooooooo\\nooo\\noooooo\\nooooooooooooooooooooooo\\noooooooooooo\\nooooooooooooooooo\\noo\\noooooooooooooooooooooooo\\nooooooooooooooo\\nooooooKNN: K=1KNN: K=100\\nFIGURE 2.16. A comparison of the KNN decision boundaries (solid black\\ncurves) obtained using K=1andK= 100 on the data from Figure 2.13. With\\nK=1, the decision boundary is overly flexible, while with K= 100 it is not\\nsufficiently flexible. The Bayes decision boundary is shown as a purple dashed\\nline.\\n0.01 0.020.05 0.10 0.200.50 1.000.000.050.100.150.20\\n1/KError Rate\\nTraining ErrorsTest ErrorsFIGURE 2.17. The KNN training error rate (blue, 200 observations) and test\\nerror rate (orange, 5,000 observations) on the data from Figure 2.13, as the level\\nof flexibility (assessed using 1/Kon the log scale) increases, or equivalently as\\nthe number of neighbors Kdecreases. The black dashed line indicates the Bayes\\nerror rate. The jumpiness of the curves is due to the small size of the training\\ndata set.\\nvarious methods for estimating test error rates and thereby choosing the\\noptimal level of flexibility for a given statistical learning method.', metadata={'source': './documents/ISLP_website.pdf', 'page': 48}), Document(page_content=\"40 2. Statistical Learning\\n2.3 Lab: Introduction to Python\\n2.3.1 Getting Started\\nTo run the labs in this book, you will need two things:\\n1. An installation of Python3, which is the specific version of Pythonused\\nin the labs.\\n2. Access to Jupyter, a very popular Pythoninterface that runs code\\nthrough a file called a notebook.notebook\\nYou can download and install Python3by following the instructions avail-\\nable atanaconda.com .\\nThere are a number of ways to get access to Jupyter. Here are just a few:\\n1. Using Google’s Colaboratory service:colab.research.google.com/ .\\n2. Using JupyterHub , available at jupyter.org/hub .\\n3. Using your own jupyter installation. Installation instructions are\\navailable at jupyter.org/install .\\nPlease see the Pythonresources page on the book website statlearning.com\\nfor up-to-date information about getting PythonandJupyterworking on\\nyour computer.\\nYou will need to install the ISLPpackage, which provides access to the\\ndatasets and custom-built functions that we provide. Inside a macOS or\\nLinux terminal type pip install ISLP ; this also installs most other pack-\\nages needed in the labs. The Pythonresources page has a link to the ISLP\\ndocumentation website.\\nTo run this lab, download the file Ch2-statlearn-lab.ipynb from the\\nPythonresources page. Now run the following code at the command line:\\njupyter lab Ch2-statlearn-lab.ipynb .\\nIf you’re using Windows, you can use the start menu to access anaconda ,\\nand follow the links. For example, to install ISLPand run this lab, you can\\nrun the same code above in an anaconda shell.\\n2.3.2 Basic Commands\\nIn this lab, we will introduce some simple Pythoncommands. For more\\nresources about Pythonin general, readers may want to consult the tutorial\\natdocs.python.org/3/tutorial/ .\\nLike most programming languages, Pythonusesfunctions to perform op-functionerations. To run a function called fun, we type fun(input1,input2) , where\\nthe inputs (or arguments )input1andinput2tellPythonhow to run the\\nfunction. A function can have any number of inputs. For example, theargumentprint()function outputs a text representation of all of its arguments toprint()the console.\\nIn[1]: print('fit a model with' ,1 1 ,'variables' )\", metadata={'source': './documents/ISLP_website.pdf', 'page': 49}), Document(page_content='2.3 Lab: Introduction to Python 41\\nfit a model with 11 variables\\nThe following command will provide information about the print()func-\\ntion.\\nIn[2]: print?\\nAdding two integers in Pythonis pretty intuitive.\\nIn[3]: 3+5\\nOut[3]: 8\\nInPython, textual data is handled using strings. For instance, \"hello\"andstring\\'hello\\'are strings. We can concatenate them using the addition +symbol.\\nIn[4]: \"hello\" +\"\"+\"world\"\\nOut[4]: \\'hello world\\'\\nA string is actually a type of sequence: this is a generic term for an orderedsequence\\nlist. The three most important types of sequences are lists, tuples, and\\nstrings. We introduce lists now.\\nThe following command instructs Pythonto join together the numbers 3,\\n4, and 5, and to save them as a listnamedx. When we type x, it gives uslistback the list.\\nIn[5]: x=[ 3 ,4 ,5 ]\\nx\\nOut[5]: [3, 4, 5]\\nNote that we used the brackets []to construct this list.\\nWe will often want to add two sets of numbers together. It is reasonable\\nto try the following code, though it will not produce the desired results.\\nIn[6]: y=[ 4 ,9 ,7 ]\\nx+y\\nOut[6]: [3, 4, 5, 4, 9, 7]\\nThe result may appear slightly counterintuitive: why did Pythonnot add\\nthe entries of the lists element-by-element? In Python, lists hold arbitrary\\nobjects, and are added using concatenation . In fact, concatenation is theconcatenat-\\nion behavior that we saw earlier when we entered \"hello\" + \" \" + \"world\" .\\nThis example reflects the fact that Pythonis a general-purpose program-\\nming language. Much of Python’s data-specific functionality comes from\\notherpackages,notably numpyandpandas.Inthenextsection,wewillintro-\\nducethe numpypackage.See docs.scipy.org/doc/numpy/user/quickstart.html\\nfor more information about numpy.', metadata={'source': './documents/ISLP_website.pdf', 'page': 50}), Document(page_content='42 2. Statistical Learning\\n2.3.3 Introduction to Numerical Python\\nAs mentioned earlier, this book makes use of functionality that is contained\\nin thenumpylibrary, orpackage. A package is a collection of modules thatnumpy\\npackagearenotnecessarilyincludedinthebase Pythondistribution.Thename numpy\\nis an abbreviation for numerical Python .\\nTo access numpy, we must first importit.import\\nIn[7]: import numpy as np\\nIn the previous line, we named the numpymodule np; an abbreviation formoduleeasier referencing.\\nInnumpy,anarrayisagenerictermforamultidimensionalsetofnumbers.array\\nWeusethe np.array() functiontodefine xandy,whichareone-dimensionalnp.array()arrays, i.e. vectors.\\nIn[8]: x=n p . a r r a y ( [ 3 ,4 ,5 ] )\\ny=n p . a r r a y ( [ 4 ,9 ,7 ] )\\nNote that if you forgot to run the import numpy as np command earlier,\\nthen you will encounter an error in calling the np.array() function in\\nthe previous line. The syntax np.array() indicates that the function be-\\ning called is part of the numpypackage, which we have abbreviated as np.\\nSincexandyhave been defined using np.array() , we get a sensible result\\nwhen we add them together. Compare this to our results in the previous\\nsection, when we tried to add two lists without using numpy.\\nIn[9]: x+y\\nOut[9]: array([ 7, 13, 12])\\nInnumpy, matrices are typically represented as two-dimensional arrays,\\nand vectors as one-dimensional arrays.1We can create a two-dimensional\\narray as follows.\\nIn[10]: x=n p . a r r a y ( [ [ 1 ,2 ] ,[ 3 ,4 ] ] )\\nx\\nOut[10]: array([[1, 2],\\n[3, 4]])\\nThe object xhas several attributes , or associated objects. To access anattributeattribute of x, we type x.attribute , where we replace attribute with the\\nname of the attribute. For instance, we can access the ndimattribute of xndimas follows.\\nIn[11]: x.ndim\\nOut[11]: 2\\nThe output indicates that xis a two-dimensional array. Similarly, x.dtype\\nis thedata type attribute of the object x. This indicates that xis compriseddata typeof 64-bit integers:\\n1While it is also possible to create matrices using np.matrix() , we will use\\nnp.array() throughout the labs in this book.', metadata={'source': './documents/ISLP_website.pdf', 'page': 51}), Document(page_content=\"2.3 Lab: Introduction to Python 43In[12]:x.dtypeOut[12]:dtype('int64')Why isxcomprised of integers? This is because we createdxby passingin exclusively integers to thenp.array()function. If we had passed in anydecimals, then we would have obtained an array offloating point numbersfloatingpoint(i.e. real-valued numbers).In[13]:np.array([[1, 2], [3.0, 4]]).dtypeOut[13]:dtype('float64')Typingfun?will causePythonto display documentation associated withthe functionfun, if it exists. We can try this fornp.array().In[14]:np.array?This documentation indicates that we could create a floating point arrayby passing adtypeargument intonp.array().dtypeIn[15]:np.array([[1, 2], [3, 4]], float).dtypeOut[15]:dtype('float64')The arrayxis two-dimensional. We can find out the number of rows andcolumns by looking at itsshapeattribute.shapeIn[16]:x.shapeOut[16]:(2, 2)Amethodis a function that is associated with an object. For instance,methodgiven an arrayx, the expressionx.sum()sums all of its elements, using thesum()method for arrays. The callx.sum()automatically providesxas the.sum()first argument to itssum()method.In[17]:x=n p . a r r a y ( [ 1 ,2 ,3 ,4 ] )x.sum()Out[17]:10We could also sum the elements ofxby passing inxas an argument to thenp.sum()function.np.sum()In[18]:x=n p . a r r a y ( [ 1 ,2 ,3 ,4 ] )np.sum(x)Out[18]:10As another example, thereshape()method returns a new array with the.reshape()same elements asx, but a different shape. We do this by passing in atupletuple\", metadata={'source': './documents/ISLP_website.pdf', 'page': 52}), Document(page_content=\"44 2. Statistical Learning\\nin our call to reshape() , in this case (2, 3). This tuple specifies that we\\nwould like to create a two-dimensional array with 2rows and 3columns.2\\nIn what follows, the \\\\ncharacter creates a new line.\\nIn[19]: x=n p . a r r a y ( [ 1 ,2 ,3 ,4 ,5 ,6 ] )\\nprint('beginning x:\\\\n' ,x )\\nx_reshape = x.reshape((2, 3))\\nprint('reshaped x:\\\\n' ,x _ r e s h a p e )\\nbeginning x:\\n[1 2 3 4 5 6]\\nreshaped x:\\n[[1 2 3]\\n[4 5 6]]\\nThe previous output reveals that numpyarrays are specified as a sequence\\nofrows. This is called row-major ordering , as opposed to column-major\\nordering.\\nPython(and hence numpy) uses 0-based indexing. This means that to\\naccess the top left element of x_reshape , we type in x_reshape[0,0] .\\nIn[20]: x_reshape[0, 0]\\nOut[20]: 1\\nSimilarly, x_reshape[1,2] yields the element in the second row and the\\nthird column of x_reshape .\\nIn[21]: x_reshape[1, 2]\\nOut[21]: 6\\nSimilarly, x[2]yields the third entry of x.\\nNow, let’s modify the top left element of x_reshape . To our surprise, we\\ndiscover that the first element of xhas been modified as well!\\nIn[22]: print('x before we modify x_reshape:\\\\n' ,x )\\nprint('x_reshape before we modify x_reshape:\\\\n' ,x _ r e s h a p e )\\nx_reshape[0, 0] = 5\\nprint('x_reshape after we modify its top left element:\\\\n' ,\\nx_reshape)\\nprint('x after we modify top left element of x_reshape:\\\\n' ,x )\\nOut[22]: xb e f o r ew em o d i f yx _ r e s h a p e :\\n[1 2 3 4 5 6]\\nx_reshape before we modify x_reshape:\\n[[1 2 3]\\n[4 5 6]]\\nx_reshape after we modify its top left element:\\n[[5 2 3]\\n2Like lists, tuples represent a sequence of objects. Why do we need more than one way\\nto create a sequence? There are a few differences between tuples and lists, but perhaps\\nthe most important is that elements of a tuple cannot be modified, whereas elements of\\na list can be.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 53}), Document(page_content=\"2.3 Lab: Introduction to Python 45[4 5 6]]xa f t e rw em o d i f yt o pl e f te l e m e n to fx _ r e s h a p e :[5 2 3 4 5 6]Modifyingx_reshapealso modifiedxbecause the two objects occupy thesame space in memory.Wejustsawthatwecanmodifyanelementofanarray.Canwealsomod-ify a tuple? It turns out that we cannot — and trying to do so introducesanexception, or error.exceptionIn[23]:my_tuple = (3, 4, 5)my_tuple[0] = 2TypeError:'tuple'object does not support item assignmentWenowbrieflymentionsomeattributesofarraysthatwillcomeinhandy.An array’sshapeattribute contains its dimension; this is always a tuple.Thendimattribute yields the number of dimensions, andTprovides itstranspose.In[24]:x_reshape.shape, x_reshape.ndim, x_reshape.TOut[24]:((2, 3),2,array([[5, 4],[2, 5],[3, 6]]))Notice that the three individual outputs(2,3),2, andarray([[5, 4],[2,5], [3,6]])are themselves output as a tuple.We will often want to apply functions to arrays. For instance, we cancompute the square root of the entries using thenp.sqrt()function:np.sqrt()In[25]:np.sqrt(x)Out[25]:array([2.24, 1.41, 1.73, 2., 2.24, 2.45])We can also square the elements:In[26]:x**2Out[26]:array([25, 4, 9, 16, 25, 36])We can compute the square roots using the same notation, raising to thepower of1/2instead of 2.In[27]:x**0.5Out[27]:array([2.24, 1.41, 1.73, 2., 2.24, 2.45])Throughout this book, we will often want to generate random data. Thenp.random.normal()functiongeneratesavectorofrandomnormalvariables.np.random.normal()Wecanlearnmoreaboutthisfunctionbylookingatthehelppage,viaacalltonp.random.normal?. The first line of the help page readsnormal(loc=0.0,scale=1.0, size=None). Thissignatureline tells us that the function’s ar-signature\", metadata={'source': './documents/ISLP_website.pdf', 'page': 54}), Document(page_content='46 2. Statistical Learning\\nguments are loc,scale, andsize. These are keyword arguments, whichkeywordmeans that when they are passed into the function, they can be referred\\nto by name (in any order).3By default, this function will generate random\\nnormal variable(s) with mean ( loc)0and standard deviation ( scale)1; fur-\\nthermore, a single random variable will be generated unless the argument\\ntosizeis changed.\\nWe now generate 50 independent random variables from a N(0,1)dis-\\ntribution.\\nIn[28]: x=n p . r a n d o m . n o r m a l ( s i z e = 5 0 )\\nx\\nOut[28]: array([-1.19, 0.41, 0.9 , -0.44, -0.9 , -0.38, 0.13, 1.87,\\n-0.35, 1.16, 0.79, -0.97, -1.21, 0.06, -1.62, -0.6 ,\\n-0.77, -2.12, 0.38, -1.22, -0.06, -1.97, -1.74, -0.56,\\n1.7 , -0.95, 0.56, 0.35, 0.87, 0.88, -1.66, -0.32,\\n-0.3 , -1.36, 0.92, -0.31, 1.28, -1.94, 1.07, 0.07,\\n0.79, -0.46, 2.19, -0.27, -0.64, 0.85, 0.13, 0.46,\\n-0.09, 0.7 ])\\nWe create an array yby adding an independent N(50,1)random variable\\nto each element of x.\\nIn[29]: y=x+n p . r a n d o m . n o r m a l ( l o c = 5 0 ,s c a l e = 1 ,s i z e = 5 0 )\\nThenp.corrcoef() functioncomputesthecorrelationmatrixbetween xandnp.corrcoef()y. The off-diagonal elements give the correlation between xandy.\\nIn[30]: np.corrcoef(x, y)\\nOut[30]: array([[1. , 0.69],\\n[0.69, 1. ]])\\nIf you’re following along in your own Jupyternotebook, then you probably\\nnoticed that you got a different set of results when you ran the past few\\ncommands. In particular, each time we call np.random.normal() , we will get\\na different answer, as shown in the following example.\\nIn[31]: print(np.random.normal(scale=5, size=2))\\nprint(np.random.normal(scale=5, size=2))\\nOut[31]: [4.28 2.59]\\n[4.62 -2.54]\\nIn order to ensure that our code provides exactly the same results each\\ntime it is run, we can set a random seed using the np.random.default_rng()random seed\\nnp.random.\\ndefault_rng()function. This function takes an arbitrary, user-specified integer argument.\\nIf we set a random seed before generating random data, then re-running\\nour code will yield the same results. The object rnghas essentially all the\\n3Python also uses positional arguments. Positional arguments do not need to use a\\nkeyword. To see an example, type in np.sum? . We see that ais a positional argument,\\ni.e. this function assumes that the first unnamed argument that it receives is the array\\nto be summed. By contrast, axisanddtypeare keyword arguments: the position in\\nwhich these arguments are entered into np.sum() does not matter.', metadata={'source': './documents/ISLP_website.pdf', 'page': 55}), Document(page_content='2.3 Lab: Introduction to Python 47randomnumbergeneratingmethodsfoundinnp.random.Hence,togeneratenormal data we userng.normal().In[32]:rng = np.random.default_rng(1303)print(rng.normal(scale=5, size=2))rng2 = np.random.default_rng(1303)print(rng2.normal(scale=5, size=2))Out[32]:[4.09 -1.07 ][4.09 -1.07 ]Throughout the labs in this book, we usenp.random.default_rng()when-ever we perform calculations involving random quantities withinnumpy. Inprinciple, this should enable the reader to exactly reproduce the statedresults. However, as new versions ofnumpybecome available, it is possiblethat some small discrepancies may occur between the output in the labsand the output fromnumpy.Thenp.mean(),np.var(), andnp.std()functions can be used to computenp.mean()np.var()np.std()the mean, variance, and standard deviation of arrays. These functions arealso available as methods on the arrays.In[33]:rng = np.random.default_rng(3)y=r n g . s t a n d a r d _ n o r m a l ( 1 0 )np.mean(y), y.mean()Out[33]:(-0.11, -0.11)In[34]:np.var(y), y.var(), np.mean((y - y.mean())**2)Out[34]:(2.72, 2.72, 2.72)Notice that by defaultnp.var()divides by the sample sizenrather thann−1; see theddofargument innp.var?.In[35]:np.sqrt(np.var(y)), np.std(y)Out[35]:(1.65, 1.65)Thenp.mean(),np.var(), andnp.std()functions can also be applied to therows and columns of a matrix. To see this, we construct a10×3matrix ofN(0,1)random variables, and consider computing its row sums.In[36]:X=r n g . s t a n d a r d _ n o r m a l ( ( 1 0 ,3 ) )XOut[36]:array([[ 0.23, -0.35, -0.28],[-0.67, -1.06, -0.39],[0 . 4 8 ,- 0 . 2 4 , 0 . 9 6 ] ,[-0.2 , 0.02, 1.55],[0 . 5 5 ,- 0 . 5 1 ,- 0 . 1 8 ] ,[0 . 5 4 , 1 . 9 4 ,- 0 . 2 7 ] ,[-0.24, 1. , -0.89],[-0.29, 0.88, 0.58],[0 . 0 9 , 0 . 6 7 ,- 2 . 8 3 ] ,[1 . 0 2 ,- 0 . 9 6 ,- 1 . 6 7 ] ] )', metadata={'source': './documents/ISLP_website.pdf', 'page': 56}), Document(page_content='48 2. Statistical Learning\\nSince arrays are row-major ordered, the first axis, i.e. axis=0, refers to its\\nrows. We pass this argument into the mean()method for the object X..mean()\\nIn[37]: X.mean(axis=0)\\nOut[37]: array([0.15, 0.14, -0.34])\\nThe following yields the same result.\\nIn[38]: X.mean(0)\\nOut[38]: array([0.15, 0.14, -0.34])\\n2.3.4 Graphics\\nInPython, common practice is to use the library matplotlib for graphics.matplotlibHowever, since Pythonwas not written with data analysis in mind, the no-\\ntion of plotting is not intrinsic to the language. We will use the subplots()\\nfunction from matplotlib.pyplot to create a figure and the axes onto which\\nwe plot our data. For many more examples of how to make plots in Python,\\nreaders are encouraged to visit matplotlib.org/stable/gallery/ .\\nInmatplotlib , a plot consists of a figureand one or more axes. You canfigure\\naxesthink of the figure as the blank canvas upon which one or more plots will\\nbe displayed: it is the entire plotting window. The axescontain important\\ninformation about each plot, such as its x- andy-axis labels, title, and\\nmore. (Note that in matplotlib , the word axesis not the plural of axis:a\\nplot’saxescontains much more information than just the x-axis and the\\ny-axis.)\\nWe begin by importing the subplots() function from matplotlib .W esubplots()use this function throughout when creating figures. The function returns\\na tuple of length two: a figure object as well as the relevant axes object.\\nWe will typically pass figsizeas a keyword argument. Having created our\\naxes, we attempt our first plot using its plot()method. To learn more.plot()about it, type ax.plot? .\\nIn[39]: from matplotlib.pyplot import subplots\\nfig, ax = subplots(figsize=(8, 8))\\nx=r n g . s t a n d a r d _ n o r m a l ( 1 0 0 )\\ny=r n g . s t a n d a r d _ n o r m a l ( 1 0 0 )\\nax.plot(x, y);\\nWe pause here to note that we have unpacked the tuple of length two re-\\nturned by subplots() into the two distinct variables figandax. Unpacking\\nis typically preferred to the following equivalent but slightly more verbose\\ncode:\\nIn[40]: output = subplots(figsize=(8, 8))\\nfig = output[0]\\nax = output[1]\\nWe see that our earlier cell produced a line plot, which is the default.\\nTo create a scatterplot, we provide an additional argument to ax.plot() ,\\nindicating that circles should be displayed.', metadata={'source': './documents/ISLP_website.pdf', 'page': 57}), Document(page_content='2.3 Lab: Introduction to Python 49In[41]:fig, ax = subplots(figsize=(8, 8))ax.plot(x, y,\\'o\\');Differentvaluesofthisadditionalargumentcanbeusedtoproducedifferentcolored lines as well as different linestyles.As an alternative, we could use theax.scatter()function to create a.scatter()scatterplot.In[42]:fig, ax = subplots(figsize=(8, 8))ax.scatter(x, y, marker=\\'o\\');Notice that in the code blocks above, we have ended the last line with asemicolon. This preventsax.plot(x, y)from printing text to the notebook.However, it does not prevent a plot from being produced. If we omit thetrailing semi-colon, then we obtain the following output:In[43]:fig, ax = subplots(figsize=(8, 8))ax.scatter(x, y, marker=\\'o\\')Out[43]:<matplotlib.collections.PathCollection at 0x7fb3d9c8f310>Figure(432x288)In what follows, we will use trailing semicolons whenever the text thatwould be output is not germane to the discussion at hand.To label our plot, we make use of theset_xlabel(),set_ylabel(), and.set_xlabel().set_ylabel()set_title()methods ofax..set_title()In[44]:fig, ax = subplots(figsize=(8, 8))ax.scatter(x, y, marker=\\'o\\')ax.set_xlabel(\"this is the x-axis\")ax.set_ylabel(\"this is the y-axis\")ax.set_title(\"Plot of X vs Y\");Having access to the figure objectfigitself means that we can go in andchange some aspects and then redisplay it. Here, we change the size from(8, 8)to(12, 3).fig.set_size_inches(12,3)figOccasionally we will want to create several plots within a figure. Thiscan be achieved by passing additional arguments tosubplots(). Below, wecreate a2×3grid of plots in a figure of size determined by thefigsizeargument. In such situations, there is often a relationship between theaxes in the plots. For example, all plots may have a commonx-axis. Thesubplots()function can automatically handle this situation when passedthe keyword argumentsharex=True. Theaxesobject below is an arraypointing to different plots in the figure.In[45]:fig, axes = subplots(nrows=2,ncols=3,figsize=(15, 5))We now produce a scatter plot with\\'o\\'in the second column of the firstrow and a scatter plot with\\'+\\'in the third column of the second row.', metadata={'source': './documents/ISLP_website.pdf', 'page': 58}), Document(page_content='50 2. Statistical LearningIn[46]:axes[0,1].plot(x, y,\\'o\\')axes[1,2].scatter(x, y, marker=\\'+\\')figTypesubplots?to learn more aboutsubplots().To save the output offig, we call itssavefig()method. The argument.savefig()dpiis the dots per inch, used to determine how large the figure will be inpixels.In[47]:fig.savefig(\"Figure.png\",d p i = 4 0 0 )fig.savefig(\"Figure.pdf\",d p i = 2 0 0 ) ;We can continue to modifyfigusing step-by-step updates; for example,wecanmodifytherangeofthex-axis,re-savethefigure,andevenre-displayit.In[48]:axes[0,1].set_xlim([-1,1])fig.savefig(\"Figure_updated.jpg\")figWe now create some more sophisticated plots. Theax.contour()method.contour()producesacontourplotinordertorepresentthree-dimensionaldata,similarcontour plotto a topographical map. It takes three arguments:• A vector ofxvalues (the first dimension),• A vector ofyvalues (the second dimension), and• A matrix whose elements correspond to thezvalue (the third dimen-sion) for each pair of(x,y)coordinates.To createxandy, we’ll use the commandnp.linspace(a, b, n), whichnp.linspace()returns a vector ofnnumbers starting ataand ending atb.In[49]:fig, ax = subplots(figsize=(8, 8))x=n p . l i n s p a c e ( - n p . p i ,n p . p i ,5 0 )y=xf=n p . m u l t i p l y . o u t e r ( n p . c o s ( y ) ,1/( 1+x * * 2 ) )ax.contour(x, y, f);We can increase the resolution by adding more levels to the image.In[50]:fig, ax = subplots(figsize=(8, 8))ax.contour(x, y, f, levels=45);To fine-tune the output of theax.contour()function, take a look at thehelp file by typing?plt.contour.Theax.imshow()method is similar toax.contour(), except that it pro-.imshow()duces a color-coded plot whose colors depend on thezvalue. This is knownas aheatmap, and is sometimes used to plot temperature in weather fore-heatmapcasts.In[51]:fig, ax = subplots(figsize=(8, 8))ax.imshow(f);', metadata={'source': './documents/ISLP_website.pdf', 'page': 59}), Document(page_content='2.3 Lab: Introduction to Python 512.3.5 Sequences and Slice NotationAs seen above, the functionnp.linspace()can be used to create a sequenceof numbers.In[52]:seq1 = np.linspace(0, 10, 11)seq1Out[52]:array([ 0., 1., 2., 3., 4., 5., 6., 7., 8., 9., 10.])The functionnp.arange()returns a sequence of numbers spaced out bynp.arange()step. Ifstepis not specified, then a default value of1is used. Let’s createa sequence that starts at0and ends at10.In[53]:seq2 = np.arange(0, 10)seq2Out[53]:array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])Why isn’t10output above? This has to do withslicenotation inPython.sliceSlice notation is used to index sequences such as lists, tuples and arrays.Suppose we want to retrieve the fourth through sixth (inclusive) entries ofa string. We obtain a slice of the string using the indexing notation[3:6].In[54]:\"hello world\"[3:6]Out[54]:\\'lo \\'In the code block above, the notation3:6is shorthand forslice(3,6)whenused inside[].In[55]:\"hello world\"[slice(3,6)]Out[55]:\\'lo \\'You might have expectedslice(3,6)to output the fourth through seventhcharacters in the text string (recalling thatPythonbegins its indexing atzero), but instead it output the fourth through sixth. This also explainswhy the earliernp.arange(0, 10)command output only the integers from0to9. See the documentationslice?for useful options in creating slices.2.3.6 Indexing DataTo begin, we create a two-dimensionalnumpyarray.In[56]:A=n p . a r r a y ( n p . a r a n g e ( 1 6 ) ) . r e s h a p e ( ( 4 ,4 ) )AOut[56]:array([[ 0, 1, 2, 3],[4 , 5 , 6 , 7 ] ,[8 , 9 ,1 0 ,1 1 ] ,[12, 13, 14, 15]])TypingA[1,2]retrieves the element corresponding to the second row andthird column. (As usual,Pythonindexes from0.)', metadata={'source': './documents/ISLP_website.pdf', 'page': 60}), Document(page_content='52 2. Statistical LearningIn[57]:A[1,2]Out[57]:6The first number after the open-bracket symbol[refers to the row, andthe second number refers to the column.Indexing Rows, Columns, and SubmatricesTo select multiple rows at a time, we can pass in a list specifying ourselection. For instance,[1,3]will retrieve the second and fourth rows:In[58]:A[[1,3]]Out[58]:array([[ 4, 5, 6, 7],[12, 13, 14, 15]])To select the first and third columns, we pass in[0,2]as the second ar-gument in the square brackets. In this case we need to supply the firstargument:which selects all rows.In[59]:A[:,[0,2]]Out[59]:array([[ 0, 2],[4 , 6 ] ,[8 ,1 0 ] ,[12, 14]])Now, suppose that we want to select the submatrix made up of the secondand fourth rows as well as the first and third columns. This is where index-ing gets slightly tricky. It is natural to try to use lists to retrieve the rowsand columns:In[60]:A[[1,3],[0,2]]Out[60]:array([ 4, 14])Oops — what happened? We got a one-dimensional array of length twoidentical toIn[61]:np.array([A[1,0],A[3,2]])Out[61]:array([ 4, 14])Similarly, the following code fails to extract the submatrix comprised ofthe second and fourth rows and the first, third, and fourth columns:In[62]:A[[1,3],[0,2,3]]IndexError: shape mismatch: indexing arrays could not be broadcasttogether with shapes (2,) (3,)', metadata={'source': './documents/ISLP_website.pdf', 'page': 61}), Document(page_content='2.3 Lab: Introduction to Python 53We can see what has gone wrong here. When supplied with two indexinglists, thenumpyinterpretation is that these provide pairs ofi, jindices fora series of entries. That is why the pair of lists must have the same length.However, that was not our intent, since we are looking for a submatrix.One easy way to do this is as follows. We first create a submatrix bysubsetting the rows ofA, and then on the fly we make a further submatrixby subsetting its columns.In[63]:A[[1,3]][:,[0,2]]Out[63]:array([[ 4, 6],[12, 14]])There are more efficient ways of achieving the same result.Theconvenience functionnp.ix_()allowsustoextractasubmatrixusingconveniencefunctionnp.ix_()lists, by creating an intermediatemeshobject.meshIn[64]:idx = np.ix_([1,3],[0,2,3])A[idx]Out[64]:array([[ 4, 6, 7],[12, 14, 15]])Alternatively, we can subset matrices efficiently using slices. The slice1:4:2captures the second and fourth items of a sequence, while the slice0:3:2captures the first and third items (the third element in a slice se-quence is the step size).In[65]:A[1:4:2,0:3:2]Out[65]:array([[ 4, 6],[12, 14]])Whyareweabletoretrieveasubmatrixdirectlyusingslicesbutnotusinglists? Its because they are differentPythontypes, and are treated differentlybynumpy. Slices can be used to extract objects from arbitrary sequences,such as strings, lists, and tuples, while the use of lists for indexing is morelimited.Boolean IndexingInnumpy,aBooleanis a type that equals eitherTrueorFalse(also rep-Booleanresented as1and0, respectively). The next line creates a vector of0’s,represented as Booleans, of length equal to the first dimension ofA.In[66]:keep_rows = np.zeros(A.shape[0], bool)keep_rowsOut[66]:array([False, False, False, False])We now set two of the elements toTrue.In[67]:keep_rows[[1,3]] = Truekeep_rows', metadata={'source': './documents/ISLP_website.pdf', 'page': 62}), Document(page_content='54 2. Statistical LearningOut[67]:array([False, True, False, True])Note that the elements ofkeep_rows, when viewed as integers, are the sameasthe valuesofnp.array([0,1,0,1]).Below,weuse==toverifytheir equal-ity. When applied to two arrays, the==operation is applied elementwise.In[68]:np.all(keep_rows == np.array([0,1,0,1]))Out[68]:True(Here, the functionnp.all()has checked whether all entries of an arraynp.all()areTrue. A similar function,np.any(), can be used to check whether anynp.any()entries of an array areTrue.)However, even thoughnp.array([0,1,0,1])andkeep_rowsare equal ac-cording to==, they index different sets of rows! The former retrieves thefirst, second, first, and second rows ofA.In[69]:A[np.array([0,1,0,1])]Out[69]:array([[0, 1, 2, 3],[4, 5, 6, 7],[0, 1, 2, 3],[4, 5, 6, 7]])By contrast,keep_rowsretrieves only the second and fourth rows ofA—i.e. the rows for which the Boolean equalsTRUE.In[70]:A[keep_rows]Out[70]:array([[ 4, 5, 6, 7],[12, 13, 14, 15]])This example shows that Booleans and integers are treated differently bynumpy.We again make use of thenp.ix_()function to create a mesh containingthe second and fourth rows, and the first, third, and fourth columns. Thistime, we apply the function to Booleans, rather than lists.In[71]:keep_cols = np.zeros(A.shape[1], bool)keep_cols[[0, 2, 3]] = Trueidx_bool = np.ix_(keep_rows, keep_cols)A[idx_bool]Out[71]:array([[ 4, 6, 7],[12, 14, 15]])We can also mix a list with an array of Booleans in the arguments tonp.ix_():In[72]:idx_mixed = np.ix_([1,3], keep_cols)A[idx_mixed]Out[72]:array([[ 4, 6, 7],[12, 14, 15]])For more details on indexing innumpy, readers are referred to thenumpytutorial mentioned earlier.', metadata={'source': './documents/ISLP_website.pdf', 'page': 63}), Document(page_content=\"2.3 Lab: Introduction to Python 552.3.7 Loading DataData sets often contain different types of data, and may have names as-sociated with the rows or columns. For these reasons, they typically arebest accommodated using adata frame. We can think of a data frame asdata framea sequence of arrays of identical length; these are the columns. Entries inthe different arrays can be combined to form a row. Thepandaslibrary canbe used to create and work with data frame objects.Reading in a Data SetThe first step of most analyses involves importing a data set intoPython.Before attempting to load a data set, we must make sure thatPythonknowswhere to find the file containing it. If the file is in the same location as thisnotebook file, then we are all set. Otherwise, the commandos.chdir()canos.chdir()be used tochange directory. (You will need to callimport osbefore callingos.chdir().)We will begin by reading inAuto.csv, available on the book website. Thisis a comma-separated file, and can be read in usingpd.read_csv():pd.read_csv()In[73]:import pandas as pdAuto = pd.read_csv('Auto.csv')AutoThe book website also has a whitespace-delimited version of this data,calledAuto.data. This can be read in as follows:In[74]:Auto = pd.read_csv('Auto.data',d e l i m _ w h i t e s p a c e = T r u e )BothAuto.csvandAuto.dataare simply text files. Before loading data intoPython, it is a good idea to view it using a text editor or other software,such as Microsoft Excel.We now take a look at the column ofAutocorresponding to the variablehorsepower:In[75]:Auto['horsepower']Out[75]:01 3 0 . 011 6 5 . 021 5 0 . 031 5 0 . 041 4 0 . 0...392 86.00393 52.00394 84.00395 79.00396 82.00Name: horsepower, Length: 397, dtype: objectWe see that thedtypeof this column isobject. It turns out that all valuesof thehorsepowercolumn were interpreted as strings when reading in thedata. We can find out why by looking at the unique values.In[76]:np.unique(Auto['horsepower'])\", metadata={'source': './documents/ISLP_website.pdf', 'page': 64}), Document(page_content=\"56 2. Statistical LearningTo save space, we have omitted the output of the previous code block. Wesee the culprit is the value?, which is being used to encode missing values.To fix the problem, we must providepd.read_csv()with an argumentcalledna_values. Now, each instance of?in the file is replaced with thevaluenp.nan, which meansnot a number:In[77]:Auto = pd.read_csv('Auto.data',na_values=['?'],delim_whitespace=True)Auto['horsepower'].sum()Out[77]:40952.0TheAuto.shapeattribute tells us that the data has 397 observations, orrows, and nine variables, or columns.In[78]:Auto.shapeOut[78]:(397, 9)There are various ways to deal with missing data. In this case, sinceonly five of the rows contain missing observations, we choose to use theAuto.dropna()method to simply remove these rows..dropna()In[79]:Auto_new = Auto.dropna()Auto_new.shapeOut[79]:(392, 9)Basics of Selecting Rows and ColumnsWe can useAuto.columnsto check the variable names.In[80]:Auto = Auto_new#o v e r w r i t et h ep r e v i o u sv a l u eAuto.columnsOut[80]:Index(['mpg','cylinders','displacement','horsepower','weight','acceleration','year','origin','name'],dtype='object')Accessing the rows and columns of a data frame is similar, but not iden-tical, to accessing the rows and columns of an array. Recall that the firstargument to the[]method is always applied to the rows of the array. Sim-ilarly, passing in a slice to the[]method creates a data frame whoserowsare determined by the slice:In[81]:Auto[:3]Out[81]:mpg cylinders displacement horsepower weight ...01 8 . 08307.0 130.0 3504.0 ...11 5 . 08350.0 165.0 3693.0 ...21 8 . 08318.0 150.0 3436.0 ...Similarly, an array of Booleans can be used to subset the rows:\", metadata={'source': './documents/ISLP_website.pdf', 'page': 65}), Document(page_content=\"2.3 Lab: Introduction to Python 57In[82]:idx_80 = Auto['year']>8 0Auto[idx_80]However, if we pass in a list of strings to the[]method, then we obtain adata frame containing the corresponding set ofcolumns.In[83]:Auto[['mpg','horsepower']]Out[83]:mpg horsepower01 8 . 0 1 3 0 . 011 5 . 0 1 6 5 . 021 8 . 0 1 5 0 . 0316.0 150.041 7 . 0 1 4 0 . 0... ... ...392 27.0 86.0393 44.0 52.0394 32.0 84.0395 28.0 79.0396 31.0 82.0392 rows x 2 columnsSince we did not specify anindexcolumn when we loaded our data frame,the rows are labeled using integers 0 to 396.In[84]:Auto.indexOut[84]:Int64Index([ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9,...387, 388, 389, 390, 391, 392, 393, 394, 395, 396],dtype='int64',l e n g t h = 3 9 2 )We can use theset_index()method to re-name the rows using the contents.set_index()ofAuto['name'].In[85]:Auto_re = Auto.set_index('name')Auto_reOut[85]:mpg cylinders displacement ...namechevrolet chevelle malibu 18.08307.0 ...buick skylark 32 15.08350.0 ...plymouth satellite 18.08318.0 ...amc rebel sst 16.08304.0 ...In[86]:Auto_re.columnsOut[86]:Index(['mpg','cylinders','displacement','horsepower','weight','acceleration','year','origin'],dtype='object')We see that the column'name'is no longer there.Now that the index has been set toname, we can access rows of the dataframe bynameusing theloc[]method ofAuto:.loc[]\", metadata={'source': './documents/ISLP_website.pdf', 'page': 66}), Document(page_content=\"58 2. Statistical LearningIn[87]:rows = ['amc rebel sst','ford torino']Auto_re.loc[rows]Out[87]:mpg cylinders displacement horsepower ...nameamc rebel sst 16.08304.0 150.0 ...ford torino 17.08302.0 140.0 ...As an alternative to using the index name, we could retrieve the 4th and5th rows ofAutousing theiloc[]method:.iloc[]In[88]:Auto_re.iloc[[3,4]]We can also use it to retrieve the 1st, 3rd and and 4th columns ofAuto_re:In[89]:Auto_re.iloc[:,[0,2,3]]We can extract the 4th and 5th rows, as well as the 1st, 3rd and 4thcolumns, using a single call toiloc[]:In[90]:Auto_re.iloc[[3,4],[0,2,3]]Out[90]:mpg displacement horsepowernameamc rebel sst 16.0304.0 150.0ford torino 17.0302.0 140.0Index entries need not be unique: there are several cars in the data framenamedford galaxie 500.In[91]:Auto_re.loc['ford galaxie 500',['mpg','origin']]Out[91]:mpg originnameford galaxie 500 15.01ford galaxie 500 14.01ford galaxie 500 14.01More on Selecting Rows and ColumnsSuppose now that we want to create a data frame consisting of theweightandoriginof the subset of cars withyeargreater than 80 — i.e. thosebuilt after 1980. To do this, we first create a Boolean array that indexesthe rows. Theloc[]method allows for Boolean entries as well as strings:In[92]:idx_80 = Auto_re['year']>8 0Auto_re.loc[idx_80, ['weight','origin']]To do this more concisely, we can use an anonymous function called alambda:lambdaIn[93]:Auto_re.loc[lambda df: df['year']>8 0 ,['weight','origin']]Thelambdacall creates a function that takes a single argument, heredf,and returnsdf['year']>80. Since it is created inside theloc[]method for\", metadata={'source': './documents/ISLP_website.pdf', 'page': 67}), Document(page_content=\"2.3 Lab: Introduction to Python 59the dataframeAuto_re, that dataframe will be the argument supplied. Asanother example of using alambda, suppose that we want all cars built after1980 that achieve greater than 30 miles per gallon:In[94]:Auto_re.loc[lambda df: (df['year']>8 0 )&( d f ['mpg']>3 0 ) ,['weight','origin']]The symbol&computes an element-wiseandoperation. As another ex-ample, suppose that we want to retrieve allFordandDatsuncars withdisplacementless than 300. We check whether eachnameentry containseither the stringfordordatsunusing thestr.contains()method of the.str.contains()indexattribute of of the dataframe:In[95]:Auto_re.loc[lambda df: (df['displacement']<3 0 0 )&( d f . i n d e x . s t r . c o n t a i n s ('ford')|d f . i n d e x . s t r . c o n t a i n s ('datsun')),['weight','origin']]Here, the symbol|computes an element-wiseoroperation.In summary, a powerful set of operations is available to index the rowsand columns of data frames. For integer based queries, use theiloc[]method. For string and Boolean selections, use theloc[]method. Forfunctional queries that filter rows, use theloc[]method with a function(typically alambda) in the rows argument.2.3.8 For LoopsAforloop is a standard tool in many languages that repeatedly evaluatesforsome chunk of code while varying different values inside the code. Forexample, suppose we loop over elements of a list and compute their sum.In[96]:total = 0for value in [3,2,19]:total += valueprint('Total is: {0}'.format(total))Total is: 24The indented code beneath the line with theforstatement is run for eachvalue in the sequence specified in theforstatement. The loop ends eitherwhen the cell ends or when code is indented at the same level as the originalforstatement. We see that the final line above which prints the total isexecuted only once after the for loop has terminated. Loops can be nestedby additional indentation.In[97]:total = 0for value in [2,3,19]:for weight in [3, 2, 1]:total += value * weightprint('Total is: {0}'.format(total))Total is: 144\", metadata={'source': './documents/ISLP_website.pdf', 'page': 68}), Document(page_content='60 2. Statistical LearningAbove, we summed over each combination ofvalueandweight. We alsotook advantage of theincrementnotation inPython: the expressiona += bincrementis equivalent toa=a+b. Besides being a convenient notation, this cansave time in computationally heavy tasks in which the intermediate valueofa+bneed not be explicitly created.Perhaps a more common task would be to sum over(value, weight)pairs. For instance, to compute the average value of a random variablethat takes on possible values 2, 3 or 19 with probability 0.2, 0.3, 0.5 respec-tively we would compute the weighted sum. Tasks such as this can often beaccomplished using thezip()function that loops over a sequence of tuples.zip()In[98]:total = 0for value, weight in zip([2,3,19],[0.2,0.3,0.5]):total += weight * valueprint(\\'Weighted average is: {0}\\'.format(total))Weighted average is: 10.8String FormattingIn the code chunk above we also printed a string displaying the total.However, the objecttotalis an integer and not a string. Inserting thevalue of something into a string is a common task, made simple using someof the powerful string formatting tools inPython. Many data cleaning tasksinvolve manipulating and programmatically producing strings.For example we may want to loop over the columns of a data frameand print the percent missing in each column. Let’s create a data frameDwith columns in which 20% of the entries are missing i.e. set tonp.nan.np.nanWe’ll create the values inDfrom a normal distribution with mean 0 andvariance 1 usingrng.standard_normal()and then overwrite some randomentries usingrng.choice().In[99]:rng = np.random.default_rng(1)A=r n g . s t a n d a r d _ n o r m a l ( ( 1 2 7 ,5 ) )M=r n g . c h o i c e ( [ 0 ,n p . n a n ] ,p = [ 0 . 8 , 0 . 2 ] ,s i z e = A . s h a p e )A+ =MD=p d . D a t a F r a m e ( A ,c o l u m n s = [\\'food\\',\\'bar\\',\\'pickle\\',\\'snack\\',\\'popcorn\\'])D[:3]Out[99]:food bar pickle snack popcorn00 . 3 4 5 5 8 40 . 8 2 1 6 1 8 0 . 3 3 0 4 3 7 - 1 . 3 0 3 1 5 7 N a N1N a N - 0 . 5 3 6 9 5 3 0 . 5 8 1 1 1 8 0 . 3 6 4 5 7 2 0 . 2 9 4 1 3 22N a N 0 . 5 4 6 7 1 3 N a N - 0 . 1 6 2 9 1 0 - 0 . 4 8 2 1 1 9In[100]:for col in D.columns:template =\\'Column \"{0}\" has {1:.2%} missing values\\'print(template.format(col,np.isnan(D[col]).mean()))', metadata={'source': './documents/ISLP_website.pdf', 'page': 69}), Document(page_content='2.3 Lab: Introduction to Python 61\\nColumn \"food\" has 16.54% missing values\\nColumn \"bar\"has 25.98% missing values\\nColumn \"pickle\" has 29.13% missing values\\nColumn \"snack\" has 21.26% missing values\\nColumn \"popcorn\" has 22.83% missing values\\nWe see that the template.format() method expects two arguments {0}and\\n{1:.2%}, and the latter includes some formatting information. In particular,\\nit specifies that the second argument should be expressed as a percent with\\ntwo decimal digits.\\nThereference docs.python.org/3/library/string.html includesmanyhelp-\\nful and more complex examples.\\n2.3.9 Additional Graphical and Numerical Summaries\\nWe can use the ax.plot() orax.scatter() functions to display the quan-\\ntitative variables. However, simply typing the variable names will produce\\nan error message, because Pythondoes not know to look in the Autodata\\nset for those variables.\\nIn[101]: fig, ax = subplots(figsize=(8, 8))\\nax.plot(horsepower, mpg, \\'o\\');\\nNameError: name \\'horsepower\\' is not defined\\nWe can address this by accessing the columns directly:\\nIn[102]: fig, ax = subplots(figsize=(8, 8))\\nax.plot(Auto[ \\'horsepower\\' ], Auto[ \\'mpg\\'],\\'o\\');\\nAlternatively, we can use the plot()method with the call Auto.plot() . Us-.plot()ing this method, the variables can be accessed by name. The plot methods\\nof a data frame return a familiar object: an axes. We can use it to update\\nthe plot as we did previously:\\nIn[103]: ax = Auto.plot.scatter( \\'horsepower\\' ,\\'mpg\\');\\nax.set_title( \\'Horsepower vs. MPG\\' )\\nIf we want to save the figure that contains a given axes, we can find the\\nrelevant figure by accessing the figureattribute:\\nIn[104]: fig = ax.figure\\nfig.savefig( \\'horsepower_mpg.png\\' );\\nWecanfurtherinstructthedataframetoplottoaparticularaxesobject.\\nIn this case the corresponding plot()method will return the modified axes\\nwe passed in as an argument. Note that when we request a one-dimensional\\ngrid of plots, the object axesis similarly one-dimensional. We place our\\nscatter plot in the middle plot of a row of three plots within a figure.\\nIn[105]: fig, axes = subplots(ncols=3, figsize=(15, 5))\\nAuto.plot.scatter( \\'horsepower\\' ,\\'mpg\\',a x = a x e s [ 1 ] ) ;\\nNote also that the columns of a data frame can be accessed as attributes:\\ntry typing in Auto.horsepower .', metadata={'source': './documents/ISLP_website.pdf', 'page': 70}), Document(page_content=\"62 2. Statistical LearningWenowconsiderthecylindersvariable.TypinginAuto.cylinders.dtypereveals that it is being treated as a quantitative variable. However, sincethere is only a small number of possible values for this variable, we maywish to treat it as qualitative. Below, we replace thecylinderscolumn witha categorical version ofAuto.cylinders. The functionpd.Series()owes itspd.Series()name to the fact thatpandasis often used in time series applications.In[106]:Auto.cylinders = pd.Series(Auto.cylinders, dtype='category')Auto.cylinders.dtypeNow thatcylindersis qualitative, we can display it using theboxplot().boxplot()method.In[107]:fig, ax = subplots(figsize=(8, 8))Auto.boxplot('mpg',b y ='cylinders',a x = a x ) ;Thehist()method can be used to plot ahistogram..hist()In[108]:fig, ax = subplots(figsize=(8, 8))Auto.hist('mpg',a x = a x ) ;The color of the bars and the number of bins can be changed:In[109]:fig, ax = subplots(figsize=(8, 8))Auto.hist('mpg',c o l o r ='red',b i n s = 1 2 ,a x = a x ) ;SeeAuto.hist?for more plotting options.We can use thepd.plotting.scatter_matrix()function to create ascat-pd.plotting.scatter_matrix()terplot matrixto visualize all of the pairwise relationships between thecolumns in a data frame.In[110]:pd.plotting.scatter_matrix(Auto);We can also produce scatterplots for a subset of the variables.In[111]:pd.plotting.scatter_matrix(Auto[['mpg','displacement','weight']]);Thedescribe()method produces a numerical summary of each column in.describe()a data frame.In[112]:Auto[['mpg','weight']].describe()We can also produce a summary of just a single column.In[113]:Auto['cylinders'].describe()Auto['mpg'].describe()To exitJupyter, selectFile / Close and Halt.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 71}), Document(page_content='2.4 Exercises 632.4 ExercisesConceptual1. Foreachofparts(a)through(d),indicatewhetherwewouldgenerallyexpect the performance of a flexible statistical learning method to bebetter or worse than an inflexible method. Justify your answer.(a) The sample sizenis extremely large, and the number of predic-torspis small.(b) The number of predictorspis extremely large, and the numberof observationsnis small.(c) The relationship between the predictors and response is highlynon-linear.(d) The variance of the error terms, i.e.σ2=Var(ϵ), is extremelyhigh.2. Explain whether each scenario is a classification or regression prob-lem, and indicate whether we are most interested in inference or pre-diction. Finally, providenandp.(a) We collect a set of data on the top 500 firms in the US. For eachfirm we record profit, number of employees, industry and theCEO salary. We are interested in understanding which factorsaffect CEO salary.(b) We are considering launching a new product and wish to knowwhether it will be asuccessor afailure. We collect data on 20similar products that were previously launched. For each prod-uct we have recorded whether it was a success or failure, pricecharged for the product, marketing budget, competition price,and ten other variables.(c) We are interested in predicting the % change in the USD/Euroexchange rate in relation to the weekly changes in the worldstock markets. Hence we collect weekly data for all of 2012. Foreach week we record the % change in the USD/Euro, the %change in the US market, the % change in the British market,and the % change in the German market.3. We now revisit the bias-variance decomposition.(a) Provide a sketch of typical (squared) bias, variance, training er-ror, test error, and Bayes (or irreducible) error curves, on a sin-gle plot, as we go from less flexible statistical learning methodstowards more flexible approaches. Thex-axis should representthe amount of flexibility in the method, and they-axis shouldrepresent the values for each curve. There should be five curves.Make sure to label each one.(b) Explain why each of the five curves has the shape displayed inpart (a).', metadata={'source': './documents/ISLP_website.pdf', 'page': 72}), Document(page_content='64 2. Statistical Learning4. You will now think of some real-life applications for statistical learn-ing.(a) Describe three real-life applications in whichclassificationmightbe useful. Describe the response, as well as the predictors. Is thegoal of each application inference or prediction? Explain youranswer.(b) Describe three real-life applications in whichregressionmightbe useful. Describe the response, as well as the predictors. Is thegoal of each application inference or prediction? Explain youranswer.(c) Describe three real-life applications in whichcluster analysismight be useful.5. What are the advantages and disadvantages of a very flexible (versusa less flexible) approach for regression or classification? Under whatcircumstances might a more flexible approach be preferred to a lessflexible approach? When might a less flexible approach be preferred?6. Describe the differences between a parametric and a non-parametricstatistical learning approach. What are the advantages of a para-metric approach to regression or classification (as opposed to a non-parametric approach)? What are its disadvantages?7. The table below provides a training data set containing six observa-tions, three predictors, and one qualitative response variable.Obs.X1X2X3Y1030Red2200Red3013Red4012Green5−101Green6111RedSuppose we wish to use this data set to make a prediction forYwhenX1=X2=X3=0usingK-nearest neighbors.(a) Compute the Euclidean distance between each observation andthe test point,X1=X2=X3=0.(b) What is our prediction withK=1? Why?(c) What is our prediction withK=3? Why?(d) If the Bayes decision boundary in this problem is highly non-linear, then would we expect thebestvalue forKto be large orsmall? Why?', metadata={'source': './documents/ISLP_website.pdf', 'page': 73}), Document(page_content=\"2.4 Exercises 65Applied8. This exercise relates to theCollegedata set, which can be found inthe fileCollege.csvon the book website. It contains a number ofvariables for 777 different universities and colleges in the US. Thevariables are•Private: Public/private indicator•Apps: Number of applications received•Accept: Number of applicants accepted•Enroll: Number of new students enrolled•Top10perc: New students from top 10% of high school class•Top25perc: New students from top 25% of high school class•F.Undergrad: Number of full-time undergraduates•P.Undergrad: Number of part-time undergraduates•Outstate: Out-of-state tuition•Room.Board: Room and board costs•Books: Estimated book costs•Personal: Estimated personal spending•PhD: Percent of faculty with Ph.D.s•Terminal: Percent of faculty with terminal degree•S.F.Ratio: Student/faculty ratio•perc.alumni: Percent of alumni who donate•Expend: Instructional expenditure per student•Grad.Rate: Graduation rateBefore reading the data intoPython, it can be viewed in Excel or atext editor.(a) Usethepd.read_csv()functiontoreadthedataintoPython.Callthe loaded datacollege. Make sure that you have the directoryset to the correct location for the data.(b) Look at the data used in the notebook by creating and runninga new cell with just the codecollegein it. You should noticethat the first column is just the name of each university in acolumn named something likeUnnamed: 0. We don’t really wantpandasto treat this as data. However, it may be handy to havethese names for later. Try the following commands and similarlylook at the resulting data frames:college2 = pd.read_csv('College.csv',i n d e x _ c o l = 0 )college3 = college.rename({'Unnamed: 0':'College'},axis=1)college3 = college3.set_index('College')\", metadata={'source': './documents/ISLP_website.pdf', 'page': 74}), Document(page_content=\"66 2. Statistical LearningThis has used the first column in the file as anindexfor thedata frame. This means thatpandashas given each row a namecorresponding to the appropriate university. Now you should seethat the first data column isPrivate. Note that the names ofthe colleges appear on the left of the table. We also introduceda new python object above: adictionary, which is specified bydictionary(key, value)pairs. Keep your modified version of the data withthe following:college = college3(c) Use thedescribe()method of to produce a numerical summaryof the variables in the data set.(d) Use thepd.plotting.scatter_matrix()function to produce ascatterplotmatrixofthefirstcolumns[Top10perc, Apps, Enroll].Recall that you can reference a listCof columns of a data frameAusingA[C].(e) Use theboxplot()method ofcollegeto produce side-by-sideboxplots ofOutstateversusPrivate.(f) Create a new qualitative variable, calledElite,b ybinningtheTop10percvariable into two groups based on whether or not theproportion of students coming from the top 10% of their highschool classes exceeds 50%.college['Elite']=p d . c u t ( c o l l e g e ['Top10perc'],[0,0.5,1],labels=['No','Yes'])Use thevalue_counts()method ofcollege['Elite']to see howmanyeliteuniversitiesthereare.Finally,usetheboxplot()methodagain to produce side-by-side boxplots ofOutstateversusElite.(g) Use theplot.hist()method ofcollegeto produce some his-tograms with differing numbers of bins for a few of the quanti-tative variables. The commandplt.subplots(2, 2)may be use-ful: it will divide the plot window into four regions so that fourplots can be made simultaneously. By changing the argumentsyou can divide the screen up in other combinations.(h) Continue exploring the data, and provide a brief summary ofwhat you discover.9. This exercise involves theAutodata set studied in the lab. Make surethat the missing values have been removed from the data.(a) Which of the predictors are quantitative, and which are quali-tative?(b) What is therangeof each quantitative predictor? You can an-swer this using themin()andmax()methods innumpy..min().max()(c) What is the mean and standard deviation of each quantitativepredictor?\", metadata={'source': './documents/ISLP_website.pdf', 'page': 75}), Document(page_content='2.4 Exercises 67(d) Now remove the 10th through 85th observations. What is therange, mean, and standard deviation of each predictor in thesubset of the data that remains?(e) Using the full data set, investigate the predictors graphically,usingscatterplotsorothertoolsofyourchoice.Createsomeplotshighlighting the relationships among the predictors. Commenton your findings.(f) Suppose that we wish to predict gas mileage (mpg) on the basisof the other variables. Do your plots suggest that any of theother variables might be useful in predictingmpg? Justify youranswer.10. This exercise involves theBostonhousing data set.(a) To begin, load in theBostondata set, which is part of theISLPlibrary.(b) How many rows are in this data set? How many columns? Whatdo the rows and columns represent?(c) Make some pairwise scatterplots of the predictors (columns) inthis data set. Describe your findings.(d) Are any of the predictors associated with per capita crime rate?If so, explain the relationship.(e) Do any of the suburbs of Boston appear to have particularlyhigh crime rates? Tax rates? Pupil-teacher ratios? Comment onthe range of each predictor.(f) How many of the suburbs in this data set bound the Charlesriver?(g) What is the median pupil-teacher ratio among the towns in thisdata set?(h) Which suburb of Boston has lowest median value of owner-occupied homes? What are the values of the other predictorsfor that suburb, and how do those values compare to the overallranges for those predictors? Comment on your findings.(i) In this data set, how many of the suburbs average more thanseven rooms per dwelling? More than eight rooms per dwelling?Comment on the suburbs that average more than eight roomsper dwelling.', metadata={'source': './documents/ISLP_website.pdf', 'page': 76}), Document(page_content='3\\nLinear Regression\\nThis chapter is about linear regression , a very simple approach for super-\\nvised learning. In particular, linear regression is a useful tool for predicting\\na quantitative response. It has been around for a long time and is the topic\\nof innumerable textbooks. Though it may seem somewhat dull compared to\\nsome of the more modern statistical learning approaches described in later\\nchapters of this book, linear regression is still a useful and widely used sta-\\ntistical learning method. Moreover, it serves as a good jumping-off point for\\nnewer approaches: as we will see in later chapters, many fancy statistical\\nlearning approaches can be seen as generalizations or extensions of linear\\nregression. Consequently, the importance of having a good understanding\\nof linear regression before studying more complex learning methods cannot\\nbe overstated. In this chapter, we review some of the key ideas underlying\\nthe linear regression model, as well as the least squares approach that is\\nmost commonly used to fit this model.\\nRecall the Advertising data from Chapter 2. Figure 2.1displays sales\\n(in thousands of units) for a particular product as a function of advertis-\\ning budgets (in thousands of dollars) for TV,radio, andnewspaper media.\\nSuppose that in our role as statistical consultants we are asked to suggest,\\non the basis of this data, a marketing plan for next year that will result in\\nhigh product sales. What information would be useful in order to provide\\nsuch a recommendation? Here are a few important questions that we might\\nseek to address:\\n1.Is there a relationship between advertising budget and sales?\\nOur first goal should be to determine whether the data provide evi-\\ndence of an association between advertising expenditure and sales. If\\nthe evidence is weak, then one might argue that no money should be\\nspent on advertising!\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning , Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_3  69', metadata={'source': './documents/ISLP_website.pdf', 'page': 77}), Document(page_content='70 3. Linear Regression\\n2.How strong is the relationship between advertising budget and sales?\\nAssuming that there is a relationship between advertising and sales,\\nwe would like to know the strength of this relationship. Does knowl-\\nedge of the advertising budget provide a lot of information about\\nproduct sales?\\n3.Which media are associated with sales?\\nAre all three media—TV, radio, and newspaper—associated with\\nsales, or are just one or two of the media associated? To answer this\\nquestion, we must find a way to separate out the individual contribu-\\ntion of each medium to sales when we have spent money on all three\\nmedia.\\n4.How large is the association between each medium and sales?\\nFor every dollar spent on advertising in a particular medium, by\\nwhat amount will sales increase? How accurately can we predict this\\namount of increase?\\n5.How accurately can we predict future sales?\\nForanygivenleveloftelevision,radio,ornewspaperadvertising,what\\nisourpredictionforsales,andwhatistheaccuracyofthisprediction?\\n6.Is the relationship linear?\\nIfthereisapproximatelyastraight-linerelationshipbetweenadvertis-\\ning expenditure in the various media and sales, then linear regression\\nis an appropriate tool. If not, then it may still be possible to trans-\\nform the predictor or the response so that linear regression can be\\nused.\\n7.Is there synergy among the advertising media?\\nPerhapsspending $50,000ontelevisionadvertisingand $50,000onra-\\ndioadvertisingisassociatedwithhighersalesthanallocating $100 ,000\\nto either television or radio individually. In marketing, this is known\\nas asynergyeffect, while in statistics it is called an interaction effect.synergy\\ninteractionIt turns out that linear regression can be used to answer each of these\\nquestions. We will first discuss all of these questions in a general context,\\nand then return to them in this specific context in Section 3.4.\\n3.1 Simple Linear Regression\\nSimple linear regression lives up to its name: it is a very straightforwardsimple linear\\nregressionapproach for predicting a quantitative response Yon the basis of a sin-\\ngle predictor variable X. It assumes that there is approximately a linear\\nrelationship between XandY. Mathematically, we can write this linear\\nrelationship as\\nY≈β0+β1X. (3.1)\\nYou might read “ ≈” as“is approximately modeled as” . We will sometimes\\ndescribe ( 3.1) by saying that we are regressing YonX(orYontoX).', metadata={'source': './documents/ISLP_website.pdf', 'page': 78}), Document(page_content='3.1 Simple Linear Regression 71\\nFor example, Xmay represent TVadvertising and Ymay represent sales.\\nThen we can regress salesontoTVby fitting the model\\nsales ≈β0+β1×TV.\\nIn Equation 3.1,β0andβ1are two unknown constants that represent\\ntheintercept andslopeterms in the linear model. Together, β0andβ1areintercept\\nslopeknown as the model coefficients orparameters . Once we have used our\\ncoefficient\\nparametertraining data to produce estimates ˆβ0andˆβ1for the model coefficients, we\\ncan predict future sales on the basis of a particular value of TV advertising\\nby computing\\nˆy=ˆβ0+ˆβ1x, (3.2)\\nwhereˆyindicates a prediction of Yon the basis of X=x. Here we use a\\nhatsymbol, ˆ, to denote the estimated value for an unknown parameter\\nor coefficient, or to denote the predicted value of the response.\\n3.1.1 Estimating the Coefficients\\nIn practice, β0andβ1are unknown. So before we can use ( 3.1) to make\\npredictions, we must use data to estimate the coefficients. Let\\n(x1,y1),(x2,y2),..., (xn,yn)\\nrepresent nobservation pairs, each of which consists of a measurement of\\nXand a measurement of Y. In the Advertising example, this data set con-\\nsists of the TV advertising budget and product sales in n= 200 different\\nmarkets. (Recall that the data are displayed in Figure 2.1.) Our goal is to\\nobtain coefficient estimates ˆβ0andˆβ1such that the linear model ( 3.1) fits\\nthe available data well—that is, so that yi≈ˆβ0+ˆβ1xifori=1,...,n . In\\nother words, we want to find an intercept ˆβ0and a slope ˆβ1such that the\\nresulting line is as close as possible to the n= 200 data points. There are\\na number of ways of measuring closeness . However, by far the most com-\\nmon approach involves minimizing the least squares criterion, and we takeleast squaresthat approach in this chapter. Alternative approaches will be considered in\\nChapter 6.\\nLetˆyi=ˆβ0+ˆβ1xibe the prediction for Ybased on the ith value of X.\\nThenei=yi−ˆyirepresents the ithresidual—this is the difference betweenresidualtheith observed response value and the ith response value that is predicted\\nby our linear model. We define the residual sum of squares (RSS) asresidual sum\\nof squaresRSS = e2\\n1+e2\\n2+···+e2\\nn,\\nor equivalently as\\nRSS = ( y1−ˆβ0−ˆβ1x1)2+(y2−ˆβ0−ˆβ1x2)2+···+(yn−ˆβ0−ˆβ1xn)2.(3.3)\\nThe least squares approach chooses ˆβ0andˆβ1to minimize the RSS. Using\\nsome calculus, one can show that the minimizers are\\nˆβ1=∑n\\ni=1(xi−¯x)(yi−¯y)∑n\\ni=1(xi−¯x)2,\\nˆβ0=¯y−ˆβ1¯x,(3.4)', metadata={'source': './documents/ISLP_website.pdf', 'page': 79}), Document(page_content='72 3. Linear Regression\\n0 50 100 150 200 250 3005 10 15 20 25TVSales\\nFIGURE 3.1. For theAdvertising data, the least squares fit for the regression\\nofsalesontoTVis shown. The fit is found by minimizing the residual sum of\\nsquares. Each grey line segment represents a residual. In this case a linear fit\\ncaptures the essence of the relationship, although it overestimates the trend in the\\nleft of the plot.\\nwhere¯y≡1\\nn∑n\\ni=1yiand¯x≡1\\nn∑n\\ni=1xiare the sample means. In other\\nwords, (3.4) defines the least squares coefficient estimates for simple linear\\nregression.\\nFigure3.1displays the simple linear regression fit to the Advertising\\ndata, where ˆβ0=7.03andˆβ1=0.0475 . In other words, according to\\nthis approximation, an additional $1,000spent on TV advertising is asso-\\nciated with selling approximately 47.5additional units of the product. In\\nFigure3.2, we have computed RSS for a number of values of β0andβ1,\\nusing the advertising data with salesas the response and TVas the predic-\\ntor. In each plot, the red dot represents the pair of least squares estimates\\n(ˆβ0,ˆβ1)given by ( 3.4). These values clearly minimize the RSS.\\n3.1.2 Assessing the Accuracy of the Coefficient Estimates\\nRecall from ( 2.1) that we assume that the truerelationship between Xand\\nYtakes the form Y=f(X)+ϵfor some unknown function f, whereϵ\\nis a mean-zero random error term. If fis to be approximated by a linear\\nfunction, then we can write this relationship as\\nY=β0+β1X+ϵ. (3.5)\\nHereβ0istheinterceptterm—thatis,theexpectedvalueof YwhenX=0,\\nandβ1is the slope—the average increase in Yassociated with a one-unit\\nincrease in X. The error term is a catch-all for what we miss with this\\nsimple model: the true relationship is probably not linear, there may be\\nother variables that cause variation in Y, and there may be measurement\\nerror. We typically assume that the error term is independent of X.', metadata={'source': './documents/ISLP_website.pdf', 'page': 80}), Document(page_content='3.1 Simple Linear Regression 73\\nFIGURE 3.2. Contour and three-dimensional plots of the RSS on the\\nAdvertising data, using salesas the response and TVas the predictor. The\\nred dots correspond to the least squares estimates ˆβ0andˆβ1, given by ( 3.4).\\nThe model given by ( 3.5) defines the population regression line , whichpopulation\\nregression\\nlineis the best linear approximation to the true relationship between Xand\\nY.1The least squares regression coefficient estimates ( 3.4) characterize the\\nleast squares line (3.2). The left-hand panel of Figure 3.3displays theseleast squares\\nlinetwo lines in a simple simulated example. We created 100randomXs, and\\ngenerated 100corresponding Ys from the model\\nY=2+3 X+ϵ, (3.6)\\nwhereϵwas generated from a normal distribution with mean zero. The\\nred line in the left-hand panel of Figure 3.3displays the truerelationship,\\nf(X)=2 + 3 X, while the blue line is the least squares estimate based\\non the observed data. The true relationship is generally not known for\\nreal data, but the least squares line can always be computed using the\\ncoefficient estimates given in ( 3.4). In other words, in real applications,\\nwe have access to a set of observations from which we can compute the\\nleast squares line; however, the population regression line is unobserved.\\nIn the right-hand panel of Figure 3.3we have generated ten different data\\nsets from the model given by ( 3.6) and plotted the corresponding ten least\\nsquares lines. Notice that different data sets generated from the same true\\nmodel result in slightly different least squares lines, but the unobserved\\npopulation regression line does not change.\\nAt first glance, the difference between the population regression line and\\nthe least squares line may seem subtle and confusing. We only have one\\ndata set, and so what does it mean that two different lines describe the\\nrelationship between the predictor and the response? Fundamentally, the\\nconcept of these two lines is a natural extension of the standard statistical\\napproachofusinginformationfromasampletoestimatecharacteristicsofa\\nlarge population. For example, suppose that we are interested in knowing\\n1The assumption of linearity is often a useful working model. However, despite what\\nmany textbooks might tell us, we seldom believe that the true relationship is linear.β0β1 2.11  2.15  2.2  2.3  2.5  2.5 \\n 3  3 \\n567890.03 0.04 0.05 0.06RSSβ1β0', metadata={'source': './documents/ISLP_website.pdf', 'page': 81}), Document(page_content='74 3. Linear Regression\\n−2 −1 0 1 2−10 −5 0 5 10XY\\n−2 −1 0 1 2−10 −5 0 5 10XY\\nFIGURE 3.3. A simulated data set. Left:The red line represents the true\\nrelationship, f(X)=2+3 X, which is known as the population regression line.\\nThe blue line is the least squares line; it is the least squares estimate for f(X)\\nbased on the observed data, shown in black. Right:The population regression line\\nis again shown in red, and the least squares line in dark blue. In light blue, ten\\nleast squares lines are shown, each computed on the basis of a separate random\\nset of observations. Each least squares line is different, but on average, the least\\nsquares lines are quite close to the population regression line.\\nthe population mean µof some random variable Y. Unfortunately, µis\\nunknown, but we do have access to nobservations from Y,y1,...,y n,\\nwhich we can use to estimate µ. A reasonable estimate is ˆµ=¯y, where\\n¯y=1\\nn∑n\\ni=1yiis the sample mean. The sample mean and the population\\nmean are different, but in general the sample mean will provide a good\\nestimate of the population mean. In the same way, the unknown coefficients\\nβ0andβ1in linear regression define the population regression line. We seek\\nto estimate these unknown coefficients using ˆβ0andˆβ1given in ( 3.4). These\\ncoefficient estimates define the least squares line.\\nThe analogy between linear regression and estimation of the mean of a\\nrandom variable is an apt one based on the concept of bias. If we use thebiassample mean ˆµto estimate µ, this estimate is unbiased, in the sense thatunbiasedon average, we expect ˆµto equalµ. What exactly does this mean? It means\\nthat on the basis of one particular set of observations y1,...,y n,ˆµmight\\noverestimate µ, and on the basis of another set of observations, ˆµmight\\nunderestimate µ. But if we could average a huge number of estimates of\\nµobtained from a huge number of sets of observations, then this average\\nwouldexactlyequalµ.Hence,anunbiasedestimatordoesnot systematically\\nover- or under-estimate the true parameter. The property of unbiasedness\\nholds for the least squares coefficient estimates given by ( 3.4) as well: if\\nwe estimate β0andβ1on the basis of a particular data set, then our\\nestimates won’t be exactly equal to β0andβ1. But if we could average\\nthe estimates obtained over a huge number of data sets, then the average\\nof these estimates would be spot on! In fact, we can see from the right-\\nhand panel of Figure 3.3that the average of many least squares lines, each', metadata={'source': './documents/ISLP_website.pdf', 'page': 82}), Document(page_content='3.1 Simple Linear Regression 75\\nestimated from a separate data set, is pretty close to the true population\\nregression line.\\nWe continue the analogy with the estimation of the population mean\\nµof a random variable Y. A natural question is as follows: how accurate\\nis the sample mean ˆµas an estimate of µ? We have established that the\\naverage of ˆµ’s over many data sets will be very close to µ, but that a\\nsingle estimate ˆµmay be a substantial underestimate or overestimate of µ.\\nHow far off will that single estimate of ˆµbe? In general, we answer this\\nquestion by computing the standard error ofˆµ, written as SE(ˆµ). We havestandard\\nerrorthe well-known formula\\nVar(ˆµ) = SE(ˆ µ)2=σ2\\nn, (3.7)\\nwhereσis the standard deviation of each of the realizations yiofY.2\\nRoughly speaking, the standard error tells us the average amount that this\\nestimate ˆµdiffers from the actual value of µ. Equation 3.7also tells us how\\nthis deviation shrinks with n—the more observations we have, the smaller\\nthe standard error of ˆµ. In a similar vein, we can wonder how close ˆβ0\\nandˆβ1are to the true values β0andβ1. To compute the standard errors\\nassociated with ˆβ0andˆβ1, we use the following formulas:\\nSE(ˆβ0)2=σ2[1\\nn+¯x2\\n∑n\\ni=1(xi−¯x)2]\\n,SE(ˆβ1)2=σ2\\n∑n\\ni=1(xi−¯x)2,(3.8)\\nwhereσ2=Var(ϵ). For these formulas to be strictly valid, we need to\\nassumethattheerrors ϵiforeachobservationhavecommonvariance σ2and\\nare uncorrelated. This is clearly not true in Figure 3.1, but the formula still\\nturns out to be a good approximation. Notice in the formula that SE(ˆβ1)is\\nsmaller when the xiare more spread out; intuitively we have more leverage\\nto estimate a slope when this is the case. We also see that SE(ˆβ0)would be\\nthe same as SE(ˆµ)if¯xwere zero (in which case ˆβ0would be equal to ¯y). In\\ngeneral,σ2isnotknown,butcanbeestimatedfromthedata.Thisestimate\\nofσis known as the residual standard error , and is given by the formularesidual\\nstandard\\nerrorRSE =√\\nRSS/(n−2). Strictly speaking, when σ2is estimated from the\\ndata we should write ˆSE(ˆβ1)to indicate that an estimate has been made,\\nbut for simplicity of notation we will drop this extra “hat”.\\nStandard errors can be used to compute confidence intervals . A 95%confidence\\nintervalconfidence interval is defined as a range of values such that with 95%\\nprobability, the range will contain the true unknown value of the param-\\neter. The range is defined in terms of lower and upper limits computed\\nfrom the sample of data. A 95% confidence interval has the following prop-\\nerty: if we take repeated samples and construct the confidence interval for\\neach sample, 95% of the intervals will contain the true unknown value of\\nthe parameter. For linear regression, the 95% confidence interval for β1\\napproximately takes the form\\nˆβ1±2·SE(ˆβ1). (3.9)\\n2This formula holds provided that the nobservations are uncorrelated.', metadata={'source': './documents/ISLP_website.pdf', 'page': 83}), Document(page_content='76 3. Linear Regression\\nThat is, there is approximately a 95% chance that the interval\\n[\\nˆβ1−2·SE(ˆβ1),ˆβ1+2·SE(ˆβ1)]\\n(3.10)\\nwill contain the true value of β1.3Similarly, a confidence interval for β0\\napproximately takes the form\\nˆβ0±2·SE(ˆβ0). (3.11)\\nIn the case of the advertising data, the 95% confidence interval for β0\\nis[6.130,7.935] and the 95% confidence interval for β1is[0.042,0.053].\\nTherefore,wecanconcludethatintheabsenceofanyadvertising,saleswill,\\non average, fall somewhere between 6,130and7,935units. Furthermore,\\nfor each $1,000increase in television advertising, there will be an average\\nincrease in sales of between 42and53units.\\nStandard errors can also be used to perform hypothesis tests on thehypothesis\\ntestcoefficients. The most common hypothesis test involves testing the null\\nhypothesis ofnull\\nhypothesisH0:There is no relationship between XandY (3.12)\\nversus the alternative hypothesisalternative\\nhypothesisHa:There is some relationship between XandY. (3.13)\\nMathematically, this corresponds to testing\\nH0:β1=0\\nversus\\nHa:β1̸=0,\\nsince ifβ1=0 then the model ( 3.5) reduces to Y=β0+ϵ, andXis\\nnot associated with Y. To test the null hypothesis, we need to determine\\nwhether ˆβ1, our estimate for β1, is sufficiently far from zero that we can\\nbe confident that β1is non-zero. How far is far enough? This of course\\ndepends on the accuracy of ˆβ1—that is, it depends on SE(ˆβ1). IfSE(ˆβ1)is\\nsmall, then even relatively small values of ˆβ1may provide strong evidence\\nthatβ1̸=0, and hence that there is a relationship between XandY. In\\ncontrast, if SE(ˆβ1)is large, then ˆβ1must be large in absolute value in order\\nfor us to reject the null hypothesis. In practice, we compute a t-statistic ,t-statisticgiven by\\nt=ˆβ1−0\\nSE(ˆβ1), (3.14)\\n3Approximately for several reasons. Equation 3.10relies on the assumption that the\\nerrors are Gaussian. Also, the factor of 2in front of the SE(ˆβ1)term will vary slightly\\ndepending on the number of observations nin the linear regression. To be precise, rather\\nthan the number 2, ( 3.10) should contain the 97.5% quantile of a t-distribution with\\nn−2degrees of freedom. Details of how to compute the 95% confidence interval precisely\\ninRwill be provided later in this chapter.', metadata={'source': './documents/ISLP_website.pdf', 'page': 84}), Document(page_content='3.1 Simple Linear Regression 77\\nCoefficient Std. error t-statistic p-value\\nIntercept 7.0325 0.4578 15.36 <0.0001\\nTV 0.0475 0.0027 17.67 <0.0001\\nTABLE 3.1. For theAdvertising data, coefficients of the least squares model\\nfor the regression of number of units sold on TV advertising budget. An increase\\nof$1,000in the TV advertising budget is associated with an increase in sales by\\naround 50 units. (Recall that the salesvariable is in thousands of units, and the\\nTVvariable is in thousands of dollars.)\\nwhichmeasuresthenumberofstandarddeviationsthat ˆβ1isawayfrom 0.If\\nthere really is no relationship between XandY, then we expect that ( 3.14)\\nwill have a t-distribution with n−2degrees of freedom. The t-distribution\\nhas a bell shape and for values of ngreater than approximately 30it is\\nquite similar to the standard normal distribution. Consequently, it is a\\nsimple matter to compute the probability of observing any number equal\\nto|t|or larger in absolute value, assuming β1=0. We call this probability\\nthep-value. Roughly speaking, we interpret the p-value as follows: a smallp-valuep-value indicates that it is unlikely to observe such a substantial association\\nbetween the predictor and the response due to chance, in the absence of\\nany real association between the predictor and the response. Hence, if we\\nsee a small p-value, then we can infer that there is an association between\\nthe predictor and the response. We reject the null hypothesis —that is, we\\ndeclare a relationship to exist between XandY—if the p-value is small\\nenough. Typical p-value cutoffs for rejecting the null hypothesis are 5% or\\n1%, although this topic will be explored in much greater detail in Chap-\\nter13. Whenn= 30, these correspond to t-statistics ( 3.14) of around 2\\nand 2.75, respectively.\\nTable3.1provides details of the least squares model for the regression of\\nnumber of units sold on TV advertising budget for the Advertising data.\\nNotice that the coefficients for ˆβ0andˆβ1are very large relative to their\\nstandard errors, so the t-statistics are also large; the probabilities of seeing\\nsuch values if H0is true are virtually zero. Hence we can conclude that\\nβ0̸=0andβ1̸=0.4\\n3.1.3 Assessing the Accuracy of the Model\\nOnce we have rejected the null hypothesis ( 3.12) in favor of the alternative\\nhypothesis ( 3.13), it is natural to want to quantify the extent to which the\\nmodel fits the data . The quality of a linear regression fit is typically assessed\\nusing two related quantities: the residual standard error (RSE) and the R2\\nR2\\nstatistic.\\n4In Table 3.1, a small p-value for the intercept indicates that we can reject the null\\nhypothesis that β0=0, and a small p-value for TVindicates that we can reject the null\\nhypothesis that β1=0. Rejecting the latter null hypothesis allows us to conclude that\\nthere is a relationship between TVandsales. Rejecting the former allows us to conclude\\nthat in the absence of TVexpenditure, salesare non-zero.', metadata={'source': './documents/ISLP_website.pdf', 'page': 85}), Document(page_content='78 3. Linear Regression\\nQuantity Value\\nResidual standard error 3.26\\nR20.612\\nF-statistic 312.1\\nTABLE 3.2. For the Advertising data, more information about the least\\nsquares model for the regression of number of units sold on TV advertising budget.\\nTable3.2displays the RSE, the R2statistic, and the F-statistic (to be\\ndescribed in Section 3.2.2) for the linear regression of number of units sold\\non TV advertising budget.\\nResidual Standard Error\\nRecall from the model ( 3.5) that associated with each observation is an\\nerror term ϵ. Due to the presence of these error terms, even if we knew the\\ntrue regression line (i.e. even if β0andβ1were known), we would not be\\nable to perfectly predict YfromX. The RSE is an estimate of the standard\\ndeviationof ϵ.Roughlyspeaking,itistheaverageamountthattheresponse\\nwill deviate from the true regression line. It is computed using the formula\\nRSE =√\\n1\\nn−2RSS =\\ued6a\\ued6b\\ued6b√1\\nn−2n∑\\ni=1(yi−ˆyi)2. (3.15)\\nNote that RSS was defined in Section 3.1.1, and is given by the formula\\nRSS =n∑\\ni=1(yi−ˆyi)2. (3.16)\\nIn the case of the advertising data, we see from the linear regression\\noutput in Table 3.2that the RSE is 3.26. In other words, actual sales in\\neach market deviate from the true regression line by approximately 3,260\\nunits, on average. Another way to think about this is that even if the\\nmodel were correct and the true values of the unknown coefficients β0\\nandβ1were known exactly, any prediction of sales on the basis of TV\\nadvertising would still be off by about 3,260units on average. Of course,\\nwhether or not 3,260units is an acceptable prediction error depends on the\\nproblem context. In the advertising data set, the mean value of salesover\\nall markets is approximately 14,000units, and so the percentage error is\\n3,260/14,000 = 23 %.\\nThe RSE is considered a measure of the lack of fit of the model ( 3.5) to\\nthe data. If the predictions obtained using the model are very close to the\\ntrue outcome values—that is, if ˆyi≈yifori=1,...,n —then (3.15) will\\nbe small, and we can conclude that the model fits the data very well. On\\nthe other hand, if ˆyiis very far from yifor one or more observations, then\\nthe RSE may be quite large, indicating that the model doesn’t fit the data\\nwell.\\nR2Statistic\\nThe RSE provides an absolute measure of lack of fit of the model ( 3.5)\\nto the data. But since it is measured in the units of Y, it is not always', metadata={'source': './documents/ISLP_website.pdf', 'page': 86}), Document(page_content='3.1 Simple Linear Regression 79\\nclear what constitutes a good RSE. The R2statistic provides an alternative\\nmeasure of fit. It takes the form of a proportion —the proportion of variance\\nexplained—and so it always takes on a value between 0and1, and is\\nindependent of the scale of Y.\\nTo calculate R2, we use the formula\\nR2=TSS −RSS\\nTSS=1−RSS\\nTSS(3.17)\\nwhereTSS =∑(yi−¯y)2is thetotal sum of squares , and RSS is definedtotal sum of\\nsquaresin (3.16). TSS measures the total variance in the response Y, and can be\\nthought of as the amount of variability inherent in the response before the\\nregressionisperformed.Incontrast,RSSmeasurestheamountofvariability\\nthat is left unexplained after performing the regression. Hence, TSS −RSS\\nmeasures the amount of variability in the response that is explained (or\\nremoved) by performing the regression, and R2measures the proportion of\\nvariability in Ythat can be explained using X. AnR2statistic that is close\\nto1indicates that a large proportion of the variability in the response is\\nexplained by the regression. A number near 0indicates that the regression\\ndoes not explain much of the variability in the response; this might occur\\nbecause the linear model is wrong, or the error variance σ2is high, or both.\\nIn Table 3.2, theR2was0.61, and so just under two-thirds of the variability\\ninsalesis explained by a linear regression on TV.\\nTheR2statistic ( 3.17) has an interpretational advantage over the RSE\\n(3.15), since unlike the RSE, it always lies between 0 and 1. However, it can\\nstill be challenging to determine what is a goodR2value, and in general,\\nthis will depend on the application. For instance, in certain problems in\\nphysics, we may know that the data truly comes from a linear model with\\na small residual error. In this case, we would expect to see an R2value that\\nisextremelycloseto 1,andasubstantiallysmaller R2valuemightindicatea\\nserious problem with the experiment in which the data were generated. On\\nthe other hand, in typical applications in biology, psychology, marketing,\\nand other domains, the linear model ( 3.5) is at best an extremely rough\\napproximation to the data, and residual errors due to other unmeasured\\nfactors are often very large. In this setting, we would expect only a very\\nsmall proportion of the variance in the response to be explained by the\\npredictor, and an R2value well below 0.1might be more realistic!\\nTheR2statistic is a measure of the linear relationship between Xand\\nY. Recall that correlation , defined ascorrelation\\nCor(X,Y)=∑n\\ni=1(xi−x)(yi−y)√∑n\\ni=1(xi−x)2√∑n\\ni=1(yi−y)2, (3.18)\\nis also a measure of the linear relationship between XandY.5This sug-\\ngests that we might be able to use r=Cor(X,Y)instead of R2in order to\\nassess the fit of the linear model. In fact, it can be shown that in the simple\\n5We note that in fact, the right-hand side of ( 3.18) is the sample correlation; thus,\\nit would be more correct to write ˆCor(X,Y); however, we omit the “hat” for ease of\\nnotation.', metadata={'source': './documents/ISLP_website.pdf', 'page': 87}), Document(page_content='80 3. Linear Regression\\nSimple regression of salesonradio\\nCoefficient Std. error t-statistic p-value\\nIntercept 9.312 0.563 16.54 <0.0001\\nradio 0.203 0.020 9.92 <0.0001\\nSimple regression of salesonnewspaper\\nCoefficient Std. error t-statistic p-value\\nIntercept 12.351 0.621 19.88 <0.0001\\nnewspaper 0.055 0.017 3.30 0.00115\\nTABLE 3.3. More simple linear regression models for the Advertising data.\\nCoefficients of the simple linear regression model for number of units sold on Top:\\nradio advertising budget and Bottom: newspaper advertising budget. A $ 1,000in-\\ncrease in spending on radio advertising is associated with an average increase in\\nsales by around 203 units, while the same increase in spending on newspaper ad-\\nvertising is associated with an average increase in sales by around 55 units. (Note\\nthat the salesvariable is in thousands of units, and the radioandnewspaper\\nvariables are in thousands of dollars.)\\nlinear regression setting, R2=r2. In other words, the squared correlation\\nand the R2statistic are identical. However, in the next section we will\\ndiscuss the multiple linear regression problem, in which we use several pre-\\ndictors simultaneously to predict the response. The concept of correlation\\nbetween the predictors and the response does not extend automatically to\\nthis setting, since correlation quantifies the association between a single\\npair of variables rather than between a larger number of variables. We will\\nsee that R2fills this role.\\n3.2 Multiple Linear Regression\\nSimplelinearregressionisausefulapproachforpredictingaresponseonthe\\nbasis of a single predictor variable. However, in practice we often have more\\nthanonepredictor.Forexample,inthe Advertising data,wehaveexamined\\nthe relationship between sales and TV advertising. We also have data for\\nthe amount of money spent advertising on the radio and in newspapers,\\nand we may want to know whether either of these two media is associated\\nwith sales. How can we extend our analysis of the advertising data in order\\nto accommodate these two additional predictors?\\nOne option is to run three separate simple linear regressions, each of\\nwhich uses a different advertising medium as a predictor. For instance,\\nwe can fit a simple linear regression to predict sales on the basis of the\\namount spent on radio advertisements. Results are shown in Table 3.3(top\\ntable). We find that a $ 1,000increase in spending on radio advertising is\\nassociated with an increase in sales of around 203 units. Table 3.3(bottom\\ntable) contains the least squares coefficients for a simple linear regression of\\nsales onto newspaper advertising budget. A $1,000increase in newspaper\\nadvertising budget is associated with an increase in sales of approximately\\n55 units.', metadata={'source': './documents/ISLP_website.pdf', 'page': 88}), Document(page_content='3.2 Multiple Linear Regression 81\\nHowever,theapproachoffittingaseparatesimplelinearregressionmodel\\nforeachpredictorisnotentirelysatisfactory.Firstofall,itisunclearhowto\\nmake a single prediction of sales given the three advertising media budgets,\\nsince each of the budgets is associated with a separate regression equation.\\nSecond, each of the three regression equations ignores the other two media\\nin forming estimates for the regression coefficients. We will see shortly that\\nif the media budgets are correlated with each other in the 200 markets\\nin our data set, then this can lead to very misleading estimates of the\\nassociation between each media budget and sales.\\nInstead of fitting a separate simple linear regression model for each pre-\\ndictor, a better approach is to extend the simple linear regression model\\n(3.5) so that it can directly accommodate multiple predictors. We can do\\nthis by giving each predictor a separate slope coefficient in a single model.\\nIn general, suppose that we have pdistinct predictors. Then the multiple\\nlinear regression model takes the form\\nY=β0+β1X1+β2X2+···+βpXp+ϵ, (3.19)\\nwhereXjrepresents the jth predictor and βjquantifies the association\\nbetween that variable and the response. We interpret βjas theaverage\\neffect on Yof a one unit increase in Xj,holding all other predictors fixed .\\nIn the advertising example, ( 3.19) becomes\\nsales=β0+β1×TV+β2×radio+β3×newspaper +ϵ. (3.20)\\n3.2.1 Estimating the Regression Coefficients\\nAs was the case in the simple linear regression setting, the regression coef-\\nficientsβ0,β1,...,βpin (3.19) are unknown, and must be estimated. Given\\nestimates ˆβ0,ˆβ1,..., ˆβp, we can make predictions using the formula\\nˆy=ˆβ0+ˆβ1x1+ˆβ2x2+···+ˆβpxp. (3.21)\\nTheparametersareestimatedusingthesameleastsquaresapproachthat\\nwe saw in the context of simple linear regression. We choose β0,β1,...,βp\\nto minimize the sum of squared residuals\\nRSS =n∑\\ni=1(yi−ˆyi)2\\n=n∑\\ni=1(yi−ˆβ0−ˆβ1xi1−ˆβ2xi2−···−ˆβpxip)2.(3.22)\\nThevalues ˆβ0,ˆβ1,..., ˆβpthatminimize( 3.22)arethemultipleleastsquares\\nregression coefficient estimates. Unlike the simple linear regression esti-\\nmates given in ( 3.4), the multiple regression coefficient estimates have\\nsomewhat complicated forms that are most easily represented using ma-\\ntrix algebra. For this reason, we do not provide them here. Any statistical\\nsoftware package can be used to compute these coefficient estimates, and\\nlater in this chapter we will show how this can be done in R. Figure 3.4', metadata={'source': './documents/ISLP_website.pdf', 'page': 89}), Document(page_content='82 3. Linear Regression\\nX1X2Y\\nFIGURE 3.4. In a three-dimensional setting, with two predictors and one\\nresponse, the least squares regression line becomes a plane. The plane is chosen\\nto minimize the sum of the squared vertical distances between each observation\\n(shown in red) and the plane.\\nillustrates an example of the least squares fit to a toy data set with p=2\\npredictors.\\nTable3.4displays the multiple regression coefficient estimates when TV,\\nradio, and newspaper advertising budgets are used to predict product sales\\nusingthe Advertising data.Weinterprettheseresultsasfollows:foragiven\\namountofTVandnewspaperadvertising,spendinganadditional $1,000on\\nradio advertising is associated with approximately 189 units of additional\\nsales. Comparing these coefficient estimates to those displayed in Tables 3.1\\nand3.3, we notice that the multiple regression coefficient estimates for\\nTVandradioare pretty similar to the simple linear regression coefficient\\nestimates. However, while the newspaper regression coefficient estimate in\\nTable3.3was significantly non-zero, the coefficient estimate for newspaper\\nin the multiple regression model is close to zero, and the corresponding p-\\nvalue is no longer significant, with a value around 0.86. This illustrates that\\nthe simple and multiple regression coefficients can be quite different. This\\ndifference stems from the fact that in the simple regression case, the slope\\nterm represents the average increase in product sales associated with a\\n$1,000increase in newspaper advertising, ignoring other predictors such as\\nTVandradio. By contrast, in the multiple regression setting, the coefficient\\nfornewspaper represents the average increase in product sales associated\\nwith increasing newspaper spending by $1,000while holding TVandradio\\nfixed.\\nDoes it make sense for the multiple regression to suggest no relationship\\nbetween salesandnewspaper while the simple linear regression implies the', metadata={'source': './documents/ISLP_website.pdf', 'page': 90}), Document(page_content='3.2 Multiple Linear Regression 83\\nCoefficient Std. error t-statistic p-value\\nIntercept 2.939 0.3119 9.42 <0.0001\\nTV 0.046 0.0014 32.81 <0.0001\\nradio 0.189 0.0086 21.89 <0.0001\\nnewspaper −0.001 0.0059 −0.18 0.8599\\nTABLE 3.4. For theAdvertising data, least squares coefficient estimates of the\\nmultiple linear regression of number of units sold on TV, radio, and newspaper\\nadvertising budgets.\\nTV radio newspaper sales\\nTV 1.0000 0.0548 0.0567 0.7822\\nradio 1.0000 0.3541 0.5762\\nnewspaper 1.0000 0.2283\\nsales 1.0000\\nTABLE 3.5. Correlation matrix for TV,radio,newspaper , andsalesfor the\\nAdvertising data.\\nopposite? In fact it does. Consider the correlation matrix for the three\\npredictor variables and response variable, displayed in Table 3.5. Notice\\nthat the correlation between radioandnewspaper is0.35. This indicates\\nthat markets with high newspaper advertising tend to also have high ra-\\ndio advertising. Now suppose that the multiple regression is correct and\\nnewspaper advertising is not associated with sales, but radio advertising\\nis associated with sales. Then in markets where we spend more on radio\\nour sales will tend to be higher, and as our correlation matrix shows, we\\nalso tend to spend more on newspaper advertising in those same mar-\\nkets. Hence, in a simple linear regression which only examines salesversus\\nnewspaper , we will observe that higher values of newspaper tend to be as-\\nsociated with higher values of sales, even though newspaper advertising is\\nnot directly associated with sales. So newspaper advertising is a surrogate\\nforradioadvertising; newspaper gets “credit” for the association between\\nradioonsales.\\nThis slightly counterintuitive result is very common in many real life\\nsituations. Consider an absurd example to illustrate the point. Running\\na regression of shark attacks versus ice cream sales for data collected at\\na given beach community over a period of time would show a positive\\nrelationship, similar to that seen between salesandnewspaper . Of course\\nno one has (yet) suggested that ice creams should be banned at beaches\\nto reduce shark attacks. In reality, higher temperatures cause more people\\nto visit the beach, which in turn results in more ice cream sales and more\\nshark attacks. A multiple regression of shark attacks onto ice cream sales\\nand temperature reveals that, as intuition implies, ice cream sales is no\\nlonger a significant predictor after adjusting for temperature.\\n3.2.2 Some Important Questions\\nWhen we perform multiple linear regression, we usually are interested in\\nanswering a few important questions.', metadata={'source': './documents/ISLP_website.pdf', 'page': 91}), Document(page_content='84 3. Linear Regression\\n1.Is at least one of the predictors X1,X2,...,X puseful in predicting\\nthe response?\\n2.Do all the predictors help to explain Y, or is only a subset of the\\npredictors useful?\\n3.How well does the model fit the data?\\n4.Given a set of predictor values, what response value should we predict,\\nand how accurate is our prediction?\\nWe now address each of these questions in turn.\\nOne: Is There a Relationship Between the Response and Predictors?\\nRecall that in the simple linear regression setting, in order to determine\\nwhether there is a relationship between the response and the predictor we\\ncan simply check whether β1=0. In the multiple regression setting with p\\npredictors, we need to ask whether all of the regression coefficients are zero,\\ni.e. whether β1=β2=···=βp=0. As in the simple linear regression\\nsetting, we use a hypothesis test to answer this question. We test the null\\nhypothesis,\\nH0:β1=β2=···=βp=0\\nversus the alternative\\nHa:at least one βjis non-zero.\\nThis hypothesis test is performed by computing the F-statistic ,F-statistic\\nF=(TSS −RSS) /p\\nRSS/(n−p−1), (3.23)\\nwhere, as with simple linear regression, TSS =∑(yi−¯y)2andRSS =∑(yi−ˆyi)2. If the linear model assumptions are correct, one can show that\\nE{RSS/(n−p−1)}=σ2\\nand that, provided H0is true,\\nE{(TSS −RSS) /p}=σ2.\\nHence, when there is no relationship between the response and predictors,\\none would expect the F-statistic to take on a value close to 1. On the other\\nhand, if Hais true, then E{(TSS −RSS) /p}>σ2, so we expect Fto be\\ngreater than 1.\\nTheF-statistic for the multiple linear regression model obtained by re-\\ngressing salesontoradio,TV, andnewspaper is shown in Table 3.6. In this\\nexample the F-statistic is 570. Since this is far larger than 1, it provides\\ncompelling evidence against the null hypothesis H0. In other words, the\\nlargeF-statistic suggests that at least one of the advertising media must\\nbe related to sales. However, what if the F-statistic had been closer to\\n1? How large does the F-statistic need to be before we can reject H0and', metadata={'source': './documents/ISLP_website.pdf', 'page': 92}), Document(page_content='3.2 Multiple Linear Regression 85\\nQuantity Value\\nResidual standard error 1.69\\nR20.897\\nF-statistic 570\\nTABLE 3.6. More information about the least squares model for the regression\\nof number of units sold on TV, newspaper, and radio advertising budgets in the\\nAdvertising data. Other information about this model was displayed in Table 3.4.\\nconclude that there is a relationship? It turns out that the answer depends\\non the values of nandp. Whennis large, an F-statistic that is just a\\nlittle larger than 1might still provide evidence against H0. In contrast,\\na larger F-statistic is needed to reject H0ifnis small. When H0is true\\nand the errors ϵihave a normal distribution, the F-statistic follows an\\nF-distribution.6For any given value of nandp, any statistical software\\npackage can be used to compute the p-value associated with the F-statistic\\nusing this distribution. Based on this p-value, we can determine whether\\nor not to reject H0. For the advertising data, the p-value associated with\\ntheF-statistic in Table 3.6is essentially zero, so we have extremely strong\\nevidence that at least one of the media is associated with increased sales.\\nIn (3.23) we are testing H0that all the coefficients are zero. Sometimes\\nwe want to test that a particular subset of qof the coefficients are zero.\\nThis corresponds to a null hypothesis\\nH0:βp−q+1=βp−q+2=···=βp=0,\\nwhere for convenience we have put the variables chosen for omission at the\\nend of the list. In this case we fit a second model that uses all the variables\\nexceptthose last q. Suppose that the residual sum of squares for that model\\nisRSS 0. Then the appropriate F-statistic is\\nF=(RSS 0−RSS) /q\\nRSS/(n−p−1). (3.24)\\nNotice that in Table 3.4, for each individual predictor a t-statistic and\\nap-value were reported. These provide information about whether each\\nindividual predictor is related to the response, after adjusting for the other\\npredictors. It turns out that each of these is exactly equivalent7to theF-\\ntest that omits that single variable from the model, leaving all the others\\nin—i.e.q=1 in (3.24). So it reports the partial effect of adding that variable\\nto the model. For instance, as we discussed earlier, these p-values indicate\\nthatTVandradioare related to sales, but that there is no evidence that\\nnewspaper is associated with sales, whenTVandradioare held fixed.\\nGiven these individual p-values for each variable, why do we need to look\\nat the overall F-statistic? After all, it seems likely that if any one of the\\np-values for the individual variables is very small, then at least one of the\\npredictors is related to the response . However, this logic is flawed, especially\\nwhen the number of predictors pis large.\\n6Even if the errors are not normally-distributed, the F-statistic approximately follows\\nanF-distribution provided that the sample size nis large.\\n7The square of each t-statistic is the corresponding F-statistic.', metadata={'source': './documents/ISLP_website.pdf', 'page': 93}), Document(page_content='86 3. Linear Regression\\nFor instance, consider an example in which p= 100 andH0:β1=β2=\\n···=βp=0is true, so no variable is truly associated with the response. In\\nthis situation, about 5% of the p-values associated with each variable (of\\nthe type shown in Table 3.4) will be below 0.05by chance. In other words,\\nwe expect to see approximately five smallp-values even in the absence of\\nany true association between the predictors and the response.8In fact, it\\nis likely that we will observe at least one p-value below 0.05by chance!\\nHence, if we use the individual t-statistics and associated p-values in order\\nto decide whether or not there is any association between the variables and\\nthe response, there is a very high chance that we will incorrectly conclude\\nthat there is a relationship. However, the F-statistic does not suffer from\\nthis problem because it adjusts for the number of predictors. Hence, if H0\\nis true, there is only a 5% chance that the F-statistic will result in a p-\\nvalue below 0.05, regardless of the number of predictors or the number of\\nobservations.\\nThe approach of using an F-statistic to test for any association between\\nthe predictors and the response works when pis relatively small, and cer-\\ntainly small compared to n. However, sometimes we have a very large num-\\nber of variables. If p>n then there are more coefficients βjto estimate\\nthan observations from which to estimate them. In this case we cannot\\neven fit the multiple linear regression model using least squares, so the F-\\nstatistic cannot be used, and neither can most of the other concepts that\\nwe have seen so far in this chapter. When pis large, some of the approaches\\ndiscussed in the next section, such as forward selection , can be used. This\\nhigh-dimensional setting is discussed in greater detail in Chapter 6.high-\\ndimensional\\nTwo: Deciding on Important Variables\\nAs discussed in the previous section, the first step in a multiple regression\\nanalysis is to compute the F-statistic and to examine the associated p-\\nvalue. If we conclude on the basis of that p-value that at least one of the\\npredictors is related to the response, then it is natural to wonder whichare\\nthe guilty ones! We could look at the individual p-values as in Table 3.4,\\nbut as discussed (and as further explored in Chapter 13), ifpis large we\\nare likely to make some false discoveries.\\nIt is possible that all of the predictors are associated with the response,\\nbut it is more often the case that the response is only associated with\\na subset of the predictors. The task of determining which predictors are\\nassociated with the response, in order to fit a single model involving only\\nthose predictors, is referred to as variable selection . The variable selectionvariable\\nselectionproblem is studied extensively in Chapter 6, and so here we will provide\\nonly a brief outline of some classical approaches.\\nIdeally, we would like to perform variable selection by trying out a lot of\\ndifferent models, each containing a different subset of the predictors. For\\ninstance, if p=2, then we can consider four models: (1) a model contain-\\ning no variables, (2) a model containing X1only, (3) a model containing\\n8This is related to the important concept of multiple testing , which is the focus of\\nChapter 13.', metadata={'source': './documents/ISLP_website.pdf', 'page': 94}), Document(page_content='3.2 Multiple Linear Regression 87\\nX2only, and (4) a model containing both X1andX2. We can then se-\\nlect thebestmodel out of all of the models that we have considered. How\\ndo we determine which model is best? Various statistics can be used to\\njudge the quality of a model. These include Mallow’s Cp,Akaike informa-Mallow’s Cption criterion (AIC),Bayesian information criterion (BIC), and adjusted\\nAkaike\\ninformation\\ncriterion\\nBayesian\\ninformation\\ncriterionR2. These are discussed in more detail in Chapter 6. We can also deter-\\nadjusted R2mine which model is best by plotting various model outputs, such as the\\nresiduals, in order to search for patterns.\\nUnfortunately, there are a total of 2pmodels that contain subsets of p\\nvariables. This means that even for moderate p, trying out every possible\\nsubsetofthepredictorsisinfeasible.Forinstance,wesawthatif p=2,then\\nthere are 22=4models to consider. But if p= 30 , then we must consider\\n230=1,073,741,824models!Thisisnotpractical.Therefore,unless pisvery\\nsmall, we cannot consider all 2pmodels, and instead we need an automated\\nand efficient approach to choose a smaller set of models to consider. There\\nare three classical approaches for this task:\\n•Forward selection . We begin with the null model —a model that con-forward\\nselection\\nnull modeltains an intercept but no predictors. We then fit psimple linear re-\\ngressions and add to the null model the variable that results in the\\nlowest RSS. We then add to that model the variable that results\\nin the lowest RSS for the new two-variable model. This approach is\\ncontinued until some stopping rule is satisfied.\\n•Backward selection . We start with all variables in the model, andbackward\\nselectionremove the variable with the largest p-value—that is, the variable\\nthat is the least statistically significant. The new (p−1)-variable\\nmodel is fit, and the variable with the largest p-value is removed. This\\nprocedure continues until a stopping rule is reached. For instance, we\\nmay stop when all remaining variables have a p-value below some\\nthreshold.\\n•Mixed selection . This is a combination of forward and backward se-mixed\\nselectionlection. We start with no variables in the model, and as with forward\\nselection, we add the variable that provides the best fit. We con-\\ntinue to add variables one-by-one. Of course, as we noted with the\\nAdvertising example, the p-values for variables can become larger as\\nnew predictors are added to the model. Hence, if at any point the\\np-value for one of the variables in the model rises above a certain\\nthreshold, then we remove that variable from the model. We con-\\ntinue to perform these forward and backward steps until all variables\\nin the model have a sufficiently low p-value, and all variables outside\\nthe model would have a large p-value if added to the model.\\nBackward selection cannot be used if p>n , while forward selection can\\nalways be used. Forward selection is a greedy approach, and might include\\nvariables early that later become redundant. Mixed selection can remedy\\nthis.', metadata={'source': './documents/ISLP_website.pdf', 'page': 95}), Document(page_content='88 3. Linear Regression\\nThree: Model Fit\\nTwo of the most common numerical measures of model fit are the RSE and\\nR2, the fraction of variance explained. These quantities are computed and\\ninterpreted in the same fashion as for simple linear regression.\\nRecall that in simple regression, R2is the square of the correlation of the\\nresponse and the variable. In multiple linear regression, it turns out that it\\nequals Cor (Y,ˆY)2, the square of the correlation between the response and\\nthe fitted linear model; in fact one property of the fitted linear model is\\nthat it maximizes this correlation among all possible linear models.\\nAnR2value close to 1indicates that the model explains a large por-\\ntion of the variance in the response variable. As an example, we saw in\\nTable3.6that for the Advertising data, the model that uses all three ad-\\nvertising media to predict saleshas anR2of0.8972 . On the other hand,\\nthe model that uses only TVandradioto predict saleshas anR2value\\nof0.89719 . In other words, there is a smallincrease in R2if we include\\nnewspaper advertising in the model that already contains TV and radio\\nadvertising, even though we saw earlier that the p-value for newspaper ad-\\nvertising in Table 3.4is not significant. It turns out that R2will always\\nincrease when more variables are added to the model, even if those vari-\\nables are only weakly associated with the response. This is due to the fact\\nthat adding another variable always results in a decrease in the residual\\nsum of squares on the training data (though not necessarily the testing\\ndata). Thus, the R2statistic, which is also computed on the training data,\\nmust increase. The fact that adding newspaper advertising to the model\\ncontaining only TV and radio advertising leads to just a tiny increase in\\nR2provides additional evidence that newspaper can be dropped from the\\nmodel. Essentially, newspaper provides no real improvement in the model\\nfit to the training samples, and its inclusion will likely lead to poor results\\non independent test samples due to overfitting.\\nBy contrast, the model containing only TVas a predictor had an R2of\\n0.61(Table3.2). Adding radioto the model leads to a substantial improve-\\nment inR2. This implies that a model that uses TV and radio expenditures\\nto predict sales is substantially better than one that uses only TV advertis-\\ning. We could further quantify this improvement by looking at the p-value\\nfor the radiocoefficient in a model that contains only TVandradioas\\npredictors.\\nThe model that contains only TVandradioas predictors has an RSE\\nof 1.681, and the model that also contains newspaper as a predictor has\\nan RSE of 1.686 (Table 3.6). In contrast, the model that contains only TV\\nhas an RSE of 3.26(Table3.2). This corroborates our previous conclusion\\nthat a model that uses TV and radio expenditures to predict sales is much\\nmore accurate (on the training data) than one that only uses TV spending.\\nFurthermore, given that TV and radio expenditures are used as predictors,\\nthere is no point in also using newspaper spending as a predictor in the\\nmodel. The observant reader may wonder how RSE can increase when\\nnewspaper is added to the model given that RSS must decrease. In general\\nRSE is defined as\\nRSE =√1\\nn−p−1RSS, (3.25)', metadata={'source': './documents/ISLP_website.pdf', 'page': 96}), Document(page_content='3.2 Multiple Linear Regression 89Sales\\nRadioTVFIGURE 3.5. For theAdvertising data, a linear regression fit to salesusing\\nTVandradioas predictors. From the pattern of the residuals, we can see that\\nthere is a pronounced non-linear relationship in the data. The positive residuals\\n(those visible above the surface), tend to lie along the 45-degree line, where TV\\nand Radio budgets are split evenly. The negative residuals (most not visible), tend\\nto lie away from this line, where budgets are more lopsided.\\nwhich simplifies to ( 3.15) for a simple linear regression. Thus, models with\\nmore variables can have higher RSE if the decrease in RSS is small relative\\nto the increase in p.\\nIn addition to looking at the RSE and R2statistics just discussed, it\\ncan be useful to plot the data. Graphical summaries can reveal problems\\nwith a model that are not visible from numerical statistics. For example,\\nFigure3.5displays a three-dimensional plot of TVandradioversussales.\\nWe see that some observations lie above and some observations lie below\\nthe least squares regression plane. In particular, the linear model seems to\\noverestimate salesfor instances in which most of the advertising money\\nwas spent exclusively on either TVorradio. It underestimates salesfor\\ninstances where the budget was split between the two media. This pro-\\nnounced non-linear pattern suggests a synergyorinteraction effect betweeninteractionthe advertising media, whereby combining the media together results in a\\nbigger boost to sales than using any single medium. In Section 3.3.2,w e\\nwill discuss extending the linear model to accommodate such synergistic\\neffects through the use of interaction terms.\\nFour: Predictions\\nOnce we have fit the multiple regression model, it is straightforward to\\napply (3.21) in order to predict the response Yon the basis of a set of\\nvalues for the predictors X1,X2,...,X p. However, there are three sorts of\\nuncertainty associated with this prediction.\\n1. Thecoefficientestimates ˆβ0,ˆβ1,..., ˆβpareestimatesfor β0,β1,...,βp.\\nThat is, the least squares plane\\nˆY=ˆβ0+ˆβ1X1+···+ˆβpXp', metadata={'source': './documents/ISLP_website.pdf', 'page': 97}), Document(page_content='90 3. Linear Regression\\nis only an estimate for the true population regression plane\\nf(X)=β0+β1X1+···+βpXp.\\nThe inaccuracy in the coefficient estimates is related to the reducible\\nerrorfrom Chapter 2. We can compute a confidence interval in order\\nto determine how close ˆYwill be to f(X).\\n2. Of course, in practice assuming a linear model for f(X)is almost\\nalways an approximation of reality, so there is an additional source of\\npotentially reducible error which we call model bias . So when we use a\\nlinear model, we are in fact estimating the best linear approximation\\nto the true surface. However, here we will ignore this discrepancy,\\nand operate as if the linear model were correct.\\n3. Even if we knew f(X)—that is, even if we knew the true values\\nforβ0,β1,...,βp—the response value cannot be predicted perfectly\\nbecause of the random error ϵin the model ( 3.20). In Chapter 2,w e\\nreferred to this as the irreducible error . How much will Yvary from\\nˆY? We use prediction intervals to answer this question. Prediction\\nintervals are always wider than confidence intervals, because they\\nincorporate both the error in the estimate for f(X)(the reducible\\nerror) and the uncertainty as to how much an individual point will\\ndiffer from the population regression plane (the irreducible error).\\nWe use a confidence interval to quantify the uncertainty surroundingconfidence\\nintervaltheaverage salesover a large number of cities. For example, given that\\n$100 ,000isspenton TVadvertisingand $20,000isspenton radioadvertising\\nin each city, the 95% confidence interval is [10,985,11,528]. We interpret\\nthis to mean that 95% of intervals of this form will contain the true value of\\nf(X).9On the other hand, a prediction interval can be used to quantify theprediction\\nintervaluncertainty surrounding salesfor aparticular city. Given that $100 ,000is\\nspenton TVadvertisingand $20,000isspenton radioadvertisinginthatcity\\nthe 95% prediction interval is [7,930,14,580]. We interpret this to mean\\nthat 95% of intervals of this form will contain the true value of Yfor this\\ncity.Notethatbothintervalsarecenteredat 11,256,butthattheprediction\\ninterval is substantially wider than the confidence interval, reflecting the\\nincreased uncertainty about salesfor a given city in comparison to the\\naverage salesover many locations.\\n9In other words, if we collect a large number of data sets like the Advertising data\\nset, and we construct a confidence interval for the average saleson the basis of each\\ndata set (given $100 ,000inTVand$20,000inradioadvertising), then 95% of these\\nconfidence intervals will contain the true value of average sales.', metadata={'source': './documents/ISLP_website.pdf', 'page': 98}), Document(page_content='3.3 Other Considerations in the Regression Model 91\\n3.3 Other Considerations in the Regression Model\\n3.3.1 Qualitative Predictors\\nIn our discussion so far, we have assumed that all variables in our linear\\nregression model are quantitative . But in practice, this is not necessarily\\nthe case; often some predictors are qualitative .\\nFor example, the Creditdata set displayed in Figure 3.6records variables\\nfor a number of credit card holders. The response is balance(average credit\\ncard debt for each individual) and there are several quantitative predictors:\\nage,cards(number of credit cards), education (years of education), income\\n(in thousands of dollars), limit(credit limit), and rating(credit rating).\\nEach panel of Figure 3.6is a scatterplot for a pair of variables whose iden-\\ntities are given by the corresponding row and column labels. For example,\\nthe scatterplot directly to the right of the word “Balance” depicts balance\\nversusage, while the plot directly to the right of “Age” corresponds to\\nageversuscards. In addition to these quantitative variables, we also have\\nfour qualitative variables: own(house ownership), student(student status),\\nstatus(marital status), and region(East, West or South).\\nPredictors with Only Two Levels\\nSuppose that we wish to investigate differences in credit card balance be-\\ntween those who own a house and those who don’t, ignoring the other vari-\\nables for the moment. If a qualitative predictor (also known as a factor)factoronly has two levels, or possible values, then incorporating it into a regres-levelsion model is very simple. We simply create an indicator or dummy variabledummy\\nvariablethat takes on two possible numerical values.10For example, based on the\\nownvariable, we can create a new variable that takes the form\\nxi={\\n1 ifith person owns a house\\n0 ifith person does not own a house ,(3.26)\\nand use this variable as a predictor in the regression equation. This results\\nin the model\\nyi=β0+β1xi+ϵi={\\nβ0+β1+ϵiifith person owns a house\\nβ0+ϵi ifith person does not .(3.27)\\nNowβ0can be interpreted as the average credit card balance among those\\nwho do not own, β0+β1as the average credit card balance among those\\nwho do own their house, and β1as the average difference in credit card\\nbalance between owners and non-owners.\\nTable3.7displays the coefficient estimates and other information asso-\\nciated with the model ( 3.27). The average credit card debt for non-owners\\nis estimated to be $509 .80, whereas owners are estimated to carry $19.73\\nin additional debt for a total of $509 .80 + $19 .73 = $529 .53. However, we\\n10In the machine learning community, the creation of dummy variables to handle\\nqualitative predictors is known as “one-hot encoding”.', metadata={'source': './documents/ISLP_website.pdf', 'page': 99}), Document(page_content='92 3. Linear RegressionBalance20 40 60 80 1005 10 15 202000 8000 14000\\n0 500 150020 40 60 80 100AgeCards\\n24685 10 15 20EducationIncome\\n50 100 1502000 8000 14000Limit\\n0 500 1500246850 100 150200 600 1000\\n200 600 1000RatingFIGURE 3.6. TheCreditdata set contains information about balance,age,\\ncards,education ,income,limit, andratingfor a number of potential cus-\\ntomers.\\nnotice that the p-value for the dummy variable is very high. This indicates\\nthat there is no statistical evidence of a difference in average credit card\\nbalance based on house ownership.\\nThe decision to code owners as 1 and non-owners as 0 in ( 3.27) is ar-\\nbitrary, and has no effect on the regression fit, but does alter the inter-\\npretation of the coefficients. If we had coded non-owners as 1and own-\\ners as0, then the estimates for β0andβ1would have been 529.53and\\n−19.73, respectively, leading once again to a prediction of credit card debt\\nof$529 .53−$19.73 = $509 .80for non-owners and a prediction of $529 .53\\nfor owners. Alternatively, instead of a 0/1coding scheme, we could create\\na dummy variable\\nxi={\\n1 ifith person owns a house\\n−1 ifith person does not own a house\\nand use this variable in the regression equation. This results in the model\\nyi=β0+β1xi+ϵi={\\nβ0+β1+ϵiifith person owns a house\\nβ0−β1+ϵiifith person does not own a house.', metadata={'source': './documents/ISLP_website.pdf', 'page': 100}), Document(page_content='3.3 Other Considerations in the Regression Model 93\\nCoefficient Std. error t-statistic p-value\\nIntercept 509.80 33.13 15.389 <0.0001\\nown[Yes] 19.73 46.05 0.429 0.6690\\nTABLE 3.7. Least squares coefficient estimates associated with the regression\\nofbalance ontoownin theCreditdata set. The linear model is given in ( 3.27).\\nThat is, ownership is encoded as a dummy variable, as in ( 3.26).\\nNowβ0can be interpreted as the overall average credit card balance (ig-\\nnoring the house ownership effect), and β1is the amount by which house\\nowners and non-owners have credit card balances that are above and below\\nthe average, respectively.11In this example, the estimate for β0is$519.665,\\nhalfway between the non-owner and owner averages of $509.80and$529.53.\\nThe estimate for β1is$9.865, which is half of $19.73, the average difference\\nbetween owners and non-owners. It is important to note that the final pre-\\ndictions for the credit balances of owners and non-owners will be identical\\nregardless of the coding scheme used. The only difference is in the way that\\nthe coefficients are interpreted.\\nQualitative Predictors with More than Two Levels\\nWhen a qualitative predictor has more than two levels, a single dummy\\nvariablecannotrepresentallpossiblevalues.Inthissituation, wecan create\\nadditional dummy variables. For example, for the regionvariable we create\\ntwo dummy variables. The first could be\\nxi1={\\n1ifith person is from the South\\n0ifith person is not from the South ,(3.28)\\nand the second could be\\nxi2={\\n1ifith person is from the West\\n0ifith person is not from the West .(3.29)\\nThen both of these variables can be used in the regression equation, in\\norder to obtain the model\\nyi=β0+β1xi1+β2xi2+ϵi=\\uf8f1\\n\\uf8f4\\uf8f2\\n\\uf8f4\\uf8f3β0+β1+ϵiifith person is from the South\\nβ0+β2+ϵiifith person is from the West\\nβ0+ϵiifith person is from the East .\\n(3.30)\\nNowβ0canbeinterpretedastheaveragecreditcardbalanceforindividuals\\nfromtheEast, β1canbeinterpretedasthedifferenceintheaveragebalance\\nbetween people from the South versus the East, and β2can be interpreted\\nas the difference in the average balance between those from the West versus\\nthe East. There will always be one fewer dummy variable than the number\\nof levels. The level with no dummy variable—East in this example—is\\nknown as the baseline.baseline\\n11Technically β0is half the sum of the average debt for house owners and the average\\ndebt for non-house owners. Hence, β0is exactly equal to the overall average only if the\\ntwo groups have an equal number of members.', metadata={'source': './documents/ISLP_website.pdf', 'page': 101}), Document(page_content='94 3. Linear Regression\\nCoefficient Std. error t-statistic p-value\\nIntercept 531.00 46.32 11.464 <0.0001\\nregion[South] −12.50 56.68 −0.221 0.8260\\nregion[West] −18.69 65.02 −0.287 0.7740\\nTABLE 3.8. Least squares coefficient estimates associated with the regression of\\nbalance ontoregionin theCreditdata set. The linear model is given in ( 3.30).\\nThat is, region is encoded via two dummy variables ( 3.28) and (3.29).\\nFrom Table 3.8, we see that the estimated balancefor the baseline, East,\\nis$531.00. It is estimated that those in the South will have $18.69less\\ndebt than those in the East, and that those in the West will have $12.50\\nless debt than those in the East. However, the p-values associated with the\\ncoefficient estimates for the two dummy variables are very large, suggesting\\nno statistical evidence of a real difference in average credit card balance\\nbetween South and East or between West and East.12Once again, the\\nlevel selected as the baseline category is arbitrary, and the final predictions\\nfor each group will be the same regardless of this choice. However, the\\ncoefficients and their p-values do depend on the choice of dummy variable\\ncoding.Ratherthanrelyontheindividualcoefficients,wecanusean F-test\\nto testH0:β1=β2=0; this does not depend on the coding. This F-test\\nhas ap-value of 0.96, indicating that we cannot reject the null hypothesis\\nthat there is no relationship between balanceandregion.\\nUsing this dummy variable approach presents no difficulties when in-\\ncorporating both quantitative and qualitative predictors. For example, to\\nregressbalanceon both a quantitative variable such as incomeand a qual-\\nitative variable such as student, we must simply create a dummy variable\\nforstudentand then fit a multiple regression model using incomeand the\\ndummy variable as predictors for credit card balance.\\nThere are many different ways of coding qualitative variables besides\\nthe dummy variable approach taken here. All of these approaches lead to\\nequivalent model fits, but the coefficients are different and have different\\ninterpretations,andaredesignedtomeasureparticular contrasts .Thistopiccontrastis beyond the scope of the book.\\n3.3.2 Extensions of the Linear Model\\nThe standard linear regression model ( 3.19) provides interpretable results\\nand works quite well on many real-world problems. However, it makes sev-\\neral highly restrictive assumptions that are often violated in practice. Two\\nof the most important assumptions state that the relationship between the\\npredictors and response are additiveandlinear. The additivity assumptionadditive\\nlinearmeans that the association between a predictor Xjand the response Ydoes\\nnot depend on the values of the other predictors. The linearity assumption\\nstates that the change in the response Yassociated with a one-unit change\\ninXjis constant, regardless of the value of Xj. In later chapters of this\\nbook, we examine a number of sophisticated methods that relax these two\\n12There could still in theory be a difference between South and West, although the\\ndata here does not suggest any difference.', metadata={'source': './documents/ISLP_website.pdf', 'page': 102}), Document(page_content='3.3 Other Considerations in the Regression Model 95\\nassumptions. Here, we briefly examine some common classical approaches\\nfor extending the linear model.\\nRemoving the Additive Assumption\\nIn our previous analysis of the Advertising data, we concluded that both TV\\nandradioseem to be associated with sales. The linear models that formed\\nthe basis for this conclusion assumed that the effect on salesof increasing\\none advertising medium is independent of the amount spent on the other\\nmedia.Forexample,thelinearmodel( 3.20)statesthattheaverageincrease\\ninsalesassociated with a one-unit increase in TVis always β1, regardless\\nof the amount spent on radio.\\nHowever, this simple model may be incorrect. Suppose that spending\\nmoney on radio advertising actually increases the effectiveness of TV ad-\\nvertising, so that the slope term for TVshould increase as radioincreases.\\nIn this situation, given a fixed budget of $100 ,000, spending half on radio\\nand half on TVmay increase salesmore than allocating the entire amount\\nto either TVor toradio. In marketing, this is known as a synergyeffect,\\nand in statistics it is referred to as an interaction effect. Figure 3.5sug-\\ngests that such an effect may be present in the advertising data. Notice\\nthat when levels of either TVorradioare low, then the true salesare lower\\nthan predicted by the linear model. But when advertising is split between\\nthe two media, then the model tends to underestimate sales.\\nConsider the standard linear regression model with two variables,\\nY=β0+β1X1+β2X2+ϵ.\\nAccording to this model, a one-unit increase in X1is associated with an\\naverage increase in Yofβ1units. Notice that the presence of X2does\\nnot alter this statement—that is, regardless of the value of X2, a one-\\nunit increase in X1is associated with a β1-unit increase in Y. One way of\\nextending this model is to include a third predictor, called an interaction\\nterm, which is constructed by computing the product of X1andX2. This\\nresults in the model\\nY=β0+β1X1+β2X2+β3X1X2+ϵ. (3.31)\\nHow does inclusion of this interaction term relax the additive assumption?\\nNotice that ( 3.31) can be rewritten as\\nY=β0+(β1+β3X2)X1+β2X2+ϵ (3.32)\\n=β0+˜β1X1+β2X2+ϵ\\nwhere˜β1=β1+β3X2. Since˜β1is now a function of X2, the association\\nbetween X1andYis no longer constant: a change in the value of X2will\\nchange the association between X1andY. A similar argument shows that\\na change in the value of X1changes the association between X2andY.\\nFor example, suppose that we are interested in studying the productiv-\\nity of a factory. We wish to predict the number of unitsproduced on the\\nbasis of the number of production linesand the total number of workers.\\nIt seems likely that the effect of increasing the number of production lines', metadata={'source': './documents/ISLP_website.pdf', 'page': 103}), Document(page_content='96 3. Linear Regression\\nCoefficient Std. error t-statistic p-value\\nIntercept 6.7502 0.248 27.23 <0.0001\\nTV 0.0191 0.002 12.70 <0.0001\\nradio 0.0289 0.009 3.24 0.0014\\nTV×radio 0.0011 0.000 20.73 <0.0001\\nTABLE 3.9. For theAdvertising data, least squares coefficient estimates asso-\\nciated with the regression of salesontoTVandradio, with an interaction term,\\nas in (3.33).\\nwill depend on the number of workers, since if no workers are available\\nto operate the lines, then increasing the number of lines will not increase\\nproduction. This suggests that it would be appropriate to include an inter-\\naction term between linesandworkersin a linear model to predict units.\\nSuppose that when we fit the model, we obtain\\nunits≈1.2+3.4×lines+0.22×workers+1.4×(lines×workers)\\n=1.2 + (3.4+1.4×workers)×lines+0.22×workers.\\nIn other words, adding an additional line will increase the number of units\\nproduced by 3.4+1.4×workers. Hence the more workers we have, the\\nstronger will be the effect of lines.\\nWe now return to the Advertising example. A linear model that uses\\nradio,TV, and an interaction between the two to predict salestakes the\\nform\\nsales=β0+β1×TV+β2×radio+β3×(radio×TV)+ϵ\\n=β0+(β1+β3×radio)×TV+β2×radio+ϵ.(3.33)\\nWe can interpret β3as the increase in the effectiveness of TV advertising\\nassociated with a one-unit increase in radio advertising (or vice-versa). The\\ncoefficients that result from fitting the model ( 3.33) are given in Table 3.9.\\nTheresults in Table 3.9stronglysuggest that the modelthat includes the\\ninteraction term is superior to the model that contains only main effects .main effectThep-value for the interaction term, TV×radio, is extremely low, indicating\\nthat there is strong evidence for Ha:β3̸=0. In other words, it is clear that\\nthe true relationship is not additive. The R2for the model ( 3.33) is 96.8%,\\ncompared to only 89.7% for the model that predicts salesusingTVand\\nradiowithout an interaction term. This means that (96.8−89.7)/(100−\\n89.7) = 69% of the variability in salesthat remains after fitting the ad-\\nditive model has been explained by the interaction term. The coefficient\\nestimatesinTable 3.9suggestthatanincreaseinTVadvertisingof $1,000is\\nassociated with increased sales of (ˆβ1+ˆβ3×radio)×1,000 = 19+1 .1×radio\\nunits. And an increase in radio advertising of $1,000will be associated with\\nan increase in sales of (ˆβ2+ˆβ3×TV)×1,000 = 29 + 1 .1×TVunits.\\nIn this example, the p-values associated with TV,radio, and the interac-\\ntion term all are statistically significant (Table 3.9), and so it is obvious\\nthat all three variables should be included in the model. However, it is\\nsometimes the case that an interaction term has a very small p-value, but\\nthe associated main effects (in this case, TVandradio) do not. The hier-\\narchical principle states that if we include an interaction in a model, wehierarchical\\nprinciple', metadata={'source': './documents/ISLP_website.pdf', 'page': 104}), Document(page_content='3.3 Other Considerations in the Regression Model 97\\nshould also include the main effects, even if the p-values associated with\\ntheir coefficients are not significant. In other words, if the interaction be-\\ntweenX1andX2seems important, then we should include both X1and\\nX2in the model even if their coefficient estimates have large p-values. The\\nrationale for this principle is that if X1×X2is related to the response,\\nthen whether or not the coefficients of X1orX2are exactly zero is of lit-\\ntle interest. Also X1×X2is typically correlated with X1andX2, and so\\nleaving them out tends to alter the meaning of the interaction.\\nIn the previous example, we considered an interaction between TVand\\nradio, both of which are quantitative variables. However, the concept of\\ninteractions applies just as well to qualitative variables, or to a combination\\nof quantitative and qualitative variables. In fact, an interaction between\\na qualitative variable and a quantitative variable has a particularly nice\\ninterpretation.Considerthe CreditdatasetfromSection 3.3.1,andsuppose\\nthatwewishtopredict balanceusingthe income(quantitative)and student\\n(qualitative) variables. In the absence of an interaction term, the model\\ntakes the form\\nbalance i≈β0+β1×income i+{\\nβ2 ifith person is a student\\n0 ifith person is not a student\\n=β1×income i+{\\nβ0+β2 ifith person is a student\\nβ0 ifith person is not a student .\\n(3.34)\\nNotice that this amounts to fitting two parallel lines to the data, one for\\nstudents and one for non-students. The lines for students and non-students\\nhave different intercepts, β0+β2versusβ0, but the same slope, β1. This\\nis illustrated in the left-hand panel of Figure 3.7. The fact that the lines\\nare parallel means that the average effect on balanceof a one-unit increase\\ninincomedoes not depend on whether or not the individual is a student.\\nThis represents a potentially serious limitation of the model, since in fact a\\nchange in incomemay have a very different effect on the credit card balance\\nof a student versus a non-student.\\nThis limitation can be addressed by adding an interaction variable, cre-\\nated by multiplying incomewith the dummy variable for student. Our\\nmodel now becomes\\nbalance i≈β0+β1×income i+{\\nβ2+β3×income iif student\\n0 if not student\\n={\\n(β0+β2)+( β1+β3)×income iif student\\nβ0+β1×income i if not student .\\n(3.35)\\nOnce again, we have two different regression lines for the students and\\nthe non-students. But now those regression lines have different intercepts,\\nβ0+β2versusβ0,aswellasdifferentslopes, β1+β3versusβ1.Thisallowsfor\\nthe possibility that changes in income may affect the credit card balances\\nof students and non-students differently. The right-hand panel of Figure 3.7', metadata={'source': './documents/ISLP_website.pdf', 'page': 105}), Document(page_content='98 3. Linear Regression\\n0 50 100 150200 600 1000 1400IncomeBalance0 50 100 150200 600 1000 1400IncomeBalancestudentnon−student\\nFIGURE 3.7. For theCreditdata, the least squares lines are shown for pre-\\ndiction of balance fromincomefor students and non-students. Left:The model\\n(3.34) was fit. There is no interaction between incomeandstudent.Right:The\\nmodel (3.35) was fit. There is an interaction term between incomeandstudent.\\nshows the estimated relationships between incomeandbalancefor students\\nand non-students in the model ( 3.35). We note that the slope for students\\nis lower than the slope for non-students. This suggests that increases in\\nincome are associated with smaller increases in credit card balance among\\nstudents as compared to non-students.\\nNon-linear Relationships\\nAs discussed previously, the linear regression model ( 3.19) assumes a linear\\nrelationship between the response and predictors. But in some cases, the\\ntrue relationship between the response and the predictors may be non-\\nlinear.Herewepresentaverysimplewaytodirectlyextendthelinearmodel\\nto accommodate non-linear relationships, using polynomial regression . Inpolynomial\\nregressionlater chapters, we will present more complex approaches for performing\\nnon-linear fits in more general settings.\\nConsider Figure 3.8, in which the mpg(gas mileage in miles per gallon)\\nversushorsepower is shown for a number of cars in the Autodata set. The\\norange line represents the linear regression fit. There is a pronounced rela-\\ntionship between mpgandhorsepower , but it seems clear that this relation-\\nship is in fact non-linear: the data suggest a curved relationship. A simple\\napproach for incorporating non-linear associations in a linear model is to\\ninclude transformed versions of the predictors. For example, the points in\\nFigure3.8seem to have a quadratic shape, suggesting that a model of thequadraticform\\nmpg=β0+β1×horsepower +β2×horsepower2+ϵ (3.36)\\nmay provide a better fit. Equation 3.36involves predicting mpgusing a\\nnon-linear function of horsepower .But it is still a linear model! That is,\\n(3.36) is simply a multiple linear regression model with X1=horsepower\\nandX2=horsepower2. So we can use standard linear regression software to\\nestimate β0,β1, andβ2in order to produce a non-linear fit. The blue curve\\nin Figure 3.8shows the resulting quadratic fit to the data. The quadratic', metadata={'source': './documents/ISLP_website.pdf', 'page': 106}), Document(page_content='3.3 Other Considerations in the Regression Model 99\\n5010015020010 20 30 40 50HorsepowerMiles per gallonLinearDegree 2Degree 5\\nFIGURE 3.8. TheAutodata set. For a number of cars, mpgandhorsepower are\\nshown. The linear regression fit is shown in orange. The linear regression fit for a\\nmodel that includes horsepower2is shown as a blue curve. The linear regression\\nfit for a model that includes all polynomials of horsepower up to fifth-degree is\\nshown in green.\\nCoefficient Std. error t-statistic p-value\\nIntercept 56.9001 1.8004 31.6 <0.0001\\nhorsepower −0.4662 0.0311 −15.0<0.0001\\nhorsepower20.0012 0.0001 10.1 <0.0001\\nTABLE 3.10. FortheAutodataset,leastsquarescoefficientestimatesassociated\\nwith the regression of mpgontohorsepower andhorsepower2.\\nfit appears to be substantially better than the fit obtained when just the\\nlinear term is included. The R2of the quadratic fit is 0.688, compared to\\n0.606for the linear fit, and the p-value in Table 3.10for the quadratic term\\nis highly significant.\\nIf including horsepower2led to such a big improvement in the model, why\\nnotinclude horsepower3,horsepower4,oreven horsepower5?Thegreencurve\\nin Figure 3.8displays the fit that results from including all polynomials up\\nto fifth degree in the model ( 3.36). The resulting fit seems unnecessarily\\nwiggly—that is, it is unclear that including the additional terms really has\\nled to a better fit to the data.\\nThe approach that we have just described for extending the linear model\\nto accommodate non-linear relationships is known as polynomial regres-\\nsion, since we have included polynomial functions of the predictors in the\\nregression model. We further explore this approach and other non-linear\\nextensions of the linear model in Chapter 7.', metadata={'source': './documents/ISLP_website.pdf', 'page': 107}), Document(page_content='100 3. Linear Regression3.3.3 Potential ProblemsWhen we fit a linear regression model to a particular data set, many prob-lems may occur. Most common among these are the following:1.Non-linearity of the response-predictor relationships.2.Correlation of error terms.3.Non-constant variance of error terms.4.Outliers.5.High-leverage points.6.Collinearity.In practice, identifying and overcoming these problems is as much anart as a science. Many pages in countless books have been written on thistopic. Since the linear regression model is not our primary focus here, wewill provide only a brief summary of some key points.1. Non-linearity of the Data\\n5 10 15 20 25 30−15 −10 −5 0 5 10 15 20Fitted valuesResidualsResidual Plot for Linear Fit323330334\\n15 20 25 30 35−15 −10 −5 0 5 10 15Fitted valuesResidualsResidual Plot for Quadratic Fit334323\\n155FIGURE 3.9.Plots of residuals versus predicted (or fitted) values for theAutodata set. In each plot, the red line is a smooth fit to the residuals, intended to makeit easier to identify a trend.Left:A linear regression ofmpgonhorsepower.Astrong pattern in the residuals indicates non-linearity in the data.Right:A linearregression ofmpgonhorsepowerandhorsepower2. There is little pattern in theresiduals.The linear regression model assumes that there is a straight-line rela-tionship between the predictors and the response. If the true relationshipis far from linear, then virtually all of the conclusions that we draw fromthe fit are suspect. In addition, the prediction accuracy of the model canbe significantly reduced.Residual plotsare a useful graphical tool for identifying non-linearity.residual plotGiven a simple linear regression model, we can plot the residuals,ei=', metadata={'source': './documents/ISLP_website.pdf', 'page': 108}), Document(page_content='3.3 Other Considerations in the Regression Model 101\\nyi−ˆyi, versus the predictor xi. In the case of a multiple regression model,\\nsince there are multiple predictors, we instead plot the residuals versus\\nthe predicted (or fitted) values ˆyi. Ideally, the residual plot will show nofitteddiscernible pattern. The presence of a pattern may indicate a problem with\\nsome aspect of the linear model.\\nThe left panel of Figure 3.9displays a residual plot from the linear re-\\ngression of mpgontohorsepower on theAutodata set that was illustrated in\\nFigure3.8.Theredlineisasmoothfittotheresiduals,whichisdisplayedin\\norder to make it easier to identify any trends. The residuals exhibit a clear\\nU-shape, which provides a strong indication of non-linearity in the data.\\nIn contrast, the right-hand panel of Figure 3.9displays the residual plot\\nthat results from the model ( 3.36), which contains a quadratic term. There\\nappears to be little pattern in the residuals, suggesting that the quadratic\\nterm improves the fit to the data.\\nIf the residual plot indicates that there are non-linear associations in the\\ndata, then a simple approach is to use non-linear transformations of the\\npredictors, such as logX,√\\nX, andX2, in the regression model. In the\\nlater chapters of this book, we will discuss other more advanced non-linear\\napproaches for addressing this issue.\\n2. Correlation of Error Terms\\nAn important assumption of the linear regression model is that the error\\nterms,ϵ1,ϵ2,...,ϵn, are uncorrelated. What does this mean? For instance,\\nif the errors are uncorrelated, then the fact that ϵiis positive provides\\nlittle or no information about the sign of ϵi+1. The standard errors that\\nare computed for the estimated regression coefficients or the fitted values\\nare based on the assumption of uncorrelated error terms. If in fact there is\\ncorrelation among the error terms, then the estimated standard errors will\\ntend to underestimate the true standard errors. As a result, confidence and\\nprediction intervals will be narrower than they should be. For example,\\na 95% confidence interval may in reality have a much lower probability\\nthan0.95of containing the true value of the parameter. In addition, p-\\nvalues associated with the model will be lower than they should be; this\\ncould cause us to erroneously conclude that a parameter is statistically\\nsignificant. In short, if the error terms are correlated, we may have an\\nunwarranted sense of confidence in our model.\\nAs an extreme example, suppose we accidentally doubled our data, lead-\\ning to observations and error terms identical in pairs. If we ignored this, our\\nstandard error calculations would be as if we had a sample of size 2n, when\\nin fact we have only nsamples. Our estimated parameters would be the\\nsame for the 2nsamples as for the nsamples, but the confidence intervals\\nwould be narrower by a factor of√\\n2!\\nWhy might correlations among the error terms occur? Such correlations\\nfrequently occur in the context of time series data, which consists of ob-time seriesservations for which measurements are obtained at discrete points in time.\\nIn many cases, observations that are obtained at adjacent time points will\\nhave positively correlated errors. In order to determine if this is the case for\\na given data set, we can plot the residuals from our model as a function of\\ntime. If the errors are uncorrelated, then there should be no discernible pat-', metadata={'source': './documents/ISLP_website.pdf', 'page': 109}), Document(page_content='102 3. Linear Regression\\n020406080100−3 −1 0 1 2 3ρ=0.0Residual\\n020406080100−4 −2 0 1 2ρ=0.5Residual\\n020406080100−1.5 −0.5 0.5 1.5ρ=0.9ResidualObservationFIGURE 3.10. Plots of residuals from simulated time series data sets generated\\nwith differing levels of correlation ρbetween error terms for adjacent time points.\\ntern. On the other hand, if the error terms are positively correlated, then\\nwe may see tracking in the residuals—that is, adjacent residuals may havetrackingsimilar values. Figure 3.10provides an illustration. In the top panel, we see\\nthe residuals from a linear regression fit to data generated with uncorre-\\nlated errors. There is no evidence of a time-related trend in the residuals.\\nIn contrast, the residuals in the bottom panel are from a data set in which\\nadjacent errors had a correlation of 0.9. Now there is a clear pattern in the\\nresiduals—adjacent residuals tend to take on similar values. Finally, the\\ncenter panel illustrates a more moderate case in which the residuals had a\\ncorrelation of 0.5. There is still evidence of tracking, but the pattern is less\\nclear.\\nMany methods have been developed to properly take account of corre-\\nlations in the error terms in time series data. Correlation among the error\\nterms can also occur outside of time series data. For instance, consider a\\nstudy in which individuals’ heights are predicted from their weights. The\\nassumption of uncorrelated errors could be violated if some of the indi-\\nviduals in the study are members of the same family, eat the same diet,\\nor have been exposed to the same environmental factors. In general, the\\nassumption of uncorrelated errors is extremely important for linear regres-\\nsion as well as for other statistical methods, and good experimental design\\nis crucial in order to mitigate the risk of such correlations.', metadata={'source': './documents/ISLP_website.pdf', 'page': 110}), Document(page_content='3.3 Other Considerations in the Regression Model 103\\n10 15 20 25 30−10 −5 0 5 10 15Fitted valuesResidualsResponse Y998975845\\n2.4 2.6 2.8 3.0 3.2 3.4−0.8 −0.6 −0.4 −0.2 0.0 0.2 0.4Fitted valuesResidualsResponse log(Y)\\n437671605FIGURE 3.11. Residual plots. In each plot, the red line is a smooth fit to the\\nresiduals, intended to make it easier to identify a trend. The blue lines track the\\nouter quantiles of the residuals, and emphasize patterns. Left:The funnel shape\\nindicates heteroscedasticity. Right:The response has been log transformed, and\\nthere is now no evidence of heteroscedasticity.\\n3. Non-constant Variance of Error Terms\\nAnother important assumption of the linear regression model is that the\\nerror terms have a constant variance, Var (ϵi)=σ2. The standard errors,\\nconfidence intervals, and hypothesis tests associated with the linear model\\nrely upon this assumption.\\nUnfortunately,itisoftenthecasethatthevariancesoftheerrortermsare\\nnon-constant. For instance, the variances of the error terms may increase\\nwith the value of the response. One can identify non-constant variances in\\nthe errors, or heteroscedasticity , from the presence of a funnel shape inhetero-\\nscedasticitytheresidualplot.Anexampleisshownintheleft-handpanelofFigure 3.11,\\nin which the magnitude of the residuals tends to increase with the fitted\\nvalues. When faced with this problem, one possible solution is to trans-\\nform the response Yusing a concave function such as logYor√\\nY. Such\\na transformation results in a greater amount of shrinkage of the larger re-\\nsponses, leading to a reduction in heteroscedasticity. The right-hand panel\\nof Figure 3.11displays the residual plot after transforming the response\\nusinglogY. The residuals now appear to have constant variance, though\\nthere is some evidence of a slight non-linear relationship in the data.\\nSometimes we have a good idea of the variance of each response. For\\nexample, the ith response could be an average of niraw observations. If\\neach of these raw observations is uncorrelated with variance σ2, then their\\naverage has variance σ2\\ni=σ2/ni. In this case a simple remedy is to fit our\\nmodel by weighted least squares , with weights proportional to the inverseweighted\\nleast squaresvariances—i.e. wi=niin this case. Most linear regression software allows\\nfor observation weights.\\n4. Outliers\\nAnoutlieris a point for which yiis far from the value predicted by theoutlier', metadata={'source': './documents/ISLP_website.pdf', 'page': 111}), Document(page_content='104 3. Linear Regression\\n−2 −1 0 1 2−4 −2 0 2 4 620\\n−2 0 2 4 6−1 0 1 2 3 4Fitted ValuesResiduals20\\n−2 0 2 4 60246Fitted ValuesStudentized Residuals20\\nXYFIGURE 3.12. Left:The least squares regression line is shown in red, and the\\nregression line after removing the outlier is shown in blue. Center:The residual\\nplot clearly identifies the outlier. Right:The outlier has a studentized residual of\\n6; typically we expect values between −3and3.\\nmodel.Outlierscanariseforavarietyofreasons,suchasincorrectrecording\\nof an observation during data collection.\\nThe red point (observation 20) in the left-hand panel of Figure 3.12\\nillustrates a typical outlier. The red solid line is the least squares regression\\nfit, while the blue dashed line is the least squares fit after removal of the\\noutlier.Inthiscase,removingtheoutlierhaslittleeffectontheleastsquares\\nline: it leads to almost no change in the slope, and a miniscule reduction\\nin the intercept. It is typical for an outlier that does not have an unusual\\npredictor value to have little effect on the least squares fit. However, even\\nif an outlier does not have much effect on the least squares fit, it can cause\\nother problems. For instance, in this example, the RSE is 1.09when the\\noutlier is included in the regression, but it is only 0.77when the outlier\\nis removed. Since the RSE is used to compute all confidence intervals and\\np-values, such a dramatic increase caused by a single data point can have\\nimplications for the interpretation of the fit. Similarly, inclusion of the\\noutlier causes the R2to decline from 0.892to0.805.\\nResidual plots can be used to identify outliers. In this example, the out-\\nlier is clearly visible in the residual plot illustrated in the center panel of\\nFigure3.12. But in practice, it can be difficult to decide how large a resid-\\nual needs to be before we consider the point to be an outlier. To address\\nthis problem, instead of plotting the residuals, we can plot the studentized\\nresiduals , computed by dividing each residual eiby its estimated standardstudentized\\nresidualerror. Observations whose studentized residuals are greater than 3 in abso-\\nlute value are possible outliers. In the right-hand panel of Figure 3.12, the\\noutlier’s studentized residual exceeds 6, while all other observations have\\nstudentized residuals between −2and2.\\nIf we believe that an outlier has occurred due to an error in data collec-\\ntion or recording, then one solution is to simply remove the observation.\\nHowever, care should be taken, since an outlier may instead indicate a\\ndeficiency with the model, such as a missing predictor.\\n5. High Leverage Points\\nWe just saw that outliers are observations for which the response yiis\\nunusual given the predictor xi. In contrast, observations with high leveragehigh\\nleveragehave an unusual value for xi. For example, observation 41in the left-hand', metadata={'source': './documents/ISLP_website.pdf', 'page': 112}), Document(page_content='3.3 Other Considerations in the Regression Model 105\\n−2 −1 0 1 2 3 40 5 102041\\n−2 −1 0 1 2−2 −1 0 1 20.00 0.05 0.10 0.15 0.20 0.25−1 0 1 2 3 4 5LeverageStudentized Residuals2041\\nXY\\nX1X2FIGURE 3.13. Left:Observation 41 is a high leverage point, while 20 is not.\\nThe red line is the fit to all the data, and the blue line is the fit with observation\\n41 removed. Center:The red observation is not unusual in terms of its X1value\\nor itsX2value, but still falls outside the bulk of the data, and hence has high\\nleverage. Right:Observation 41has a high leverage and a high residual.\\npanel of Figure 3.13has high leverage, in that the predictor value for this\\nobservation is large relative to the other observations. (Note that the data\\ndisplayed in Figure 3.13are the same as the data displayed in Figure 3.12,\\nbut with the addition of a single high leverage observation.) The red solid\\nline is the least squares fit to the data, while the blue dashed line is the\\nfit produced when observation 41is removed. Comparing the left-hand\\npanels of Figures 3.12and3.13, we observe that removing the high leverage\\nobservation has a much more substantial impact on the least squares line\\nthan removing the outlier. In fact, high leverage observations tend to have\\na sizable impact on the estimated regression line. It is cause for concern if\\nthe least squares line is heavily affected by just a couple of observations,\\nbecause any problems with these points may invalidate the entire fit. For\\nthis reason, it is important to identify high leverage observations.\\nIn a simple linear regression, high leverage observations are fairly easy to\\nidentify, since we can simply look for observations for which the predictor\\nvalue is outside of the normal range of the observations. But in a multiple\\nlinear regression with many predictors, it is possible to have an observation\\nthat is well within the range of each individual predictor’s values, but that\\nis unusual in terms of the full set of predictors. An example is shown in\\nthe center panel of Figure 3.13, for a data set with two predictors, X1and\\nX2. Most of the observations’ predictor values fall within the blue dashed\\nellipse, but the red observation is well outside of this range. But neither its\\nvalue for X1nor its value for X2is unusual. So if we examine just X1or\\njustX2, we will fail to notice this high leverage point. This problem is more\\npronounced in multiple regression settings with more than two predictors,\\nbecause then there is no simple way to plot all dimensions of the data\\nsimultaneously.\\nIn order to quantify an observation’s leverage, we compute the leverage\\nstatistic. A large value of this statistic indicates an observation with highleverage\\nstatisticleverage. For a simple linear regression,\\nhi=1\\nn+(xi−¯x)2\\n∑n\\ni′=1(xi′−¯x)2. (3.37)', metadata={'source': './documents/ISLP_website.pdf', 'page': 113}), Document(page_content='106 3. Linear Regression\\n2000 4000 6000 8000 1200030 40 50 60 70 80LimitAge\\n2000 4000 6000 8000 12000200 400 600 800LimitRating\\nFIGURE 3.14. Scatterplots of the observations from the Creditdata set. Left:\\nA plot of ageversuslimit. These two variables are not collinear. Right:A plot\\nofratingversuslimit. There is high collinearity.\\nItisclearfromthisequationthat hiincreaseswiththedistanceof xifrom¯x.\\nThere is a simple extension of hito the case of multiple predictors, though\\nwe do not provide the formula here. The leverage statistic hiis always\\nbetween 1/nand1, and the average leverage for all the observations is\\nalways equal to (p+ 1)/n. So if a given observation has a leverage statistic\\nthat greatly exceeds (p+1)/n, then we may suspect that the corresponding\\npoint has high leverage.\\nThe right-hand panel of Figure 3.13provides a plot of the studentized\\nresiduals versus hifor the data in the left-hand panel of Figure 3.13. Ob-\\nservation 41stands out as having a very high leverage statistic as well as a\\nhigh studentized residual. In other words, it is an outlier as well as a high\\nleverage observation. This is a particularly dangerous combination! This\\nplot also reveals the reason that observation 20 had relatively little effect\\non the least squares fit in Figure 3.12: it has low leverage.\\n6. Collinearity\\nCollinearity refers to the situation in which two or more predictor variablescollinearityare closely related to one another. The concept of collinearity is illustrated\\nin Figure 3.14using the Creditdata set. In the left-hand panel of Fig-\\nure3.14, the two predictors limitandageappear to have no obvious rela-\\ntionship. In contrast, in the right-hand panel of Figure 3.14, the predictors\\nlimitandratingare very highly correlated with each other, and we say\\nthat they are collinear. The presence of collinearity can pose problems in\\nthe regression context, since it can be difficult to separate out the indi-\\nvidual effects of collinear variables on the response. In other words, since\\nlimitandratingtend to increase or decrease together, it can be difficult to\\ndeterminehoweachoneseparatelyisassociatedwiththeresponse, balance.\\nFigure3.15illustratessomeofthedifficultiesthatcanresultfromcollinear-\\nity. The left-hand panel of Figure 3.15is a contour plot of the RSS ( 3.22)\\nassociated with different possible coefficient estimates for the regression\\nofbalanceonlimitandage. Each ellipse represents a set of coefficients', metadata={'source': './documents/ISLP_website.pdf', 'page': 114}), Document(page_content='3.3 Other Considerations in the Regression Model 107\\n 21.25  21.5  21.8 \\n0.16 0.17 0.18 0.19−5 −4 −3 −2 −1 0 21.5  21.8 \\n−0.1 0.0 0.1 0.2012345βLimitβLimitβAge\\nβRating\\nFIGURE 3.15. Contour plots for the RSS values as a function of the parameters\\nβfor various regressions involving the Creditdata set. In each plot, the black\\ndots represent the coefficient values corresponding to the minimum RSS. Left:\\nA contour plot of RSS for the regression of balance ontoageandlimit. The\\nminimum value is well defined. Right:A contour plot of RSS for the regression\\nofbalance ontoratingandlimit. Because of the collinearity, there are many\\npairs(βLimit,βRating )with a similar value for RSS.\\nthat correspond to the same RSS, with ellipses nearest to the center tak-\\ning on the lowest values of RSS. The black dots and associated dashed\\nlines represent the coefficient estimates that result in the smallest possible\\nRSS—in other words, these are the least squares estimates. The axes for\\nlimitandagehave been scaled so that the plot includes possible coeffi-\\ncient estimates that are up to four standard errors on either side of the\\nleast squares estimates. Thus the plot includes all plausible values for the\\ncoefficients. For example, we see that the true limitcoefficient is almost\\ncertainly somewhere between 0.15and0.20.\\nIn contrast, the right-hand panel of Figure 3.15displays contour plots\\nof the RSS associated with possible coefficient estimates for the regression\\nofbalanceontolimitandrating, which we know to be highly collinear.\\nNow the contours run along a narrow valley; there is a broad range of\\nvalues for the coefficient estimates that result in equal values for RSS.\\nHence a small change in the data could cause the pair of coefficient values\\nthat yield the smallest RSS—that is, the least squares estimates—to move\\nanywhere along this valley. This results in a great deal of uncertainty in the\\ncoefficientestimates.Noticethatthescaleforthe limitcoefficientnowruns\\nfrom roughly −0.2to0.2; this is an eight-fold increase over the plausible\\nrange of the limitcoefficient in the regression with age. Interestingly, even\\nthough the limitandratingcoefficients now have much more individual\\nuncertainty, they will almost certainly lie somewhere in this contour valley.\\nFor example, we would not expect the true value of the limitandrating\\ncoefficients to be −0.1and1respectively, even though such a value is\\nplausible for each coefficient individually.\\nSince collinearity reduces the accuracy of the estimates of the regression\\ncoefficients, it causes the standard error for ˆβjto grow. Recall that the\\nt-statistic for each predictor is calculated by dividing ˆβjby its standard', metadata={'source': './documents/ISLP_website.pdf', 'page': 115}), Document(page_content='108 3. Linear Regression\\nCoefficient Std. error t-statistic p-value\\nIntercept −173.411 43.828 −3.957<0.0001\\nModel 1 age −2.292 0.672 −3.407 0.0007\\nlimit 0.173 0.005 34.496 <0.0001\\nIntercept −377.537 45.254 −8.343<0.0001\\nModel 2 rating 2.202 0.952 2.312 0.0213\\nlimit 0.025 0.064 0.384 0.7012\\nTABLE 3.11. The results for two multiple regression models involving the Credit\\ndata set are shown. Model 1 is a regression of balance onageandlimit, and\\nModel 2 a regression of balance onratingandlimit. The standard error of\\nˆβlimitincreases 12-fold in the second regression, due to collinearity.\\nerror. Consequently, collinearity results in a decline in the t-statistic. As a\\nresult, in the presence of collinearity, we may fail to reject H0:βj=0. This\\nmeans that the powerof the hypothesis test—the probability of correctlypower\\ndetecting a non-zero coefficient—is reduced by collinearity.\\nTable3.11compares the coefficient estimates obtained from two separate\\nmultiple regression models. The first is a regression of balanceonageand\\nlimit, and the second is a regression of balanceonratingandlimit. In the\\nfirst regression, both ageandlimitare highly significant with very small p-\\nvalues. In the second, the collinearity between limitandratinghas caused\\nthe standard error for the limitcoefficient estimate to increase by a factor\\nof12and thep-value to increase to 0.701. In other words, the importance\\nof thelimitvariable has been masked due to the presence of collinearity.\\nTo avoid such a situation, it is desirable to identify and address potential\\ncollinearity problems while fitting the model.\\nA simple way to detect collinearity is to look at the correlation matrix\\nof the predictors. An element of this matrix that is large in absolute value\\nindicates a pair of highly correlated variables, and therefore a collinearity\\nproblem in the data. Unfortunately, not all collinearity problems can be\\ndetected by inspection of the correlation matrix: it is possible for collinear-\\nity to exist between three or more variables even if no pair of variables\\nhas a particularly high correlation. We call this situation multicollinearity .multi-\\ncollinearityInstead of inspecting the correlation matrix, a better way to assess multi-\\ncollinearity is to compute the variance inflation factor (VIF). The VIF isvariance\\ninflation\\nfactorthe ratio of the variance of ˆβjwhen fitting the full model divided by the\\nvariance of ˆβjif fit on its own. The smallest possible value for VIF is 1,\\nwhich indicates the complete absence of collinearity. Typically in practice\\nthere is a small amount of collinearity among the predictors. As a rule of\\nthumb, a VIF value that exceeds 5or10indicates a problematic amount of\\ncollinearity. The VIF for each variable can be computed using the formula\\nVIF(ˆβj)=1\\n1−R2\\nXj|X−j,\\nwhereR2\\nXj|X−jis theR2from a regression of Xjonto all of the other\\npredictors. If R2\\nXj|X−jis close to one, then collinearity is present, and so\\nthe VIF will be large.', metadata={'source': './documents/ISLP_website.pdf', 'page': 116}), Document(page_content='3.4 The Marketing Plan 109\\nIn theCreditdata, a regression of balanceonage,rating, andlimit\\nindicates that the predictors have VIF values of 1.01, 160.67, and 160.59.\\nAs we suspected, there is considerable collinearity in the data!\\nWhen faced with the problem of collinearity, there are two simple solu-\\ntions. The first is to drop one of the problematic variables from the regres-\\nsion. This can usually be done without much compromise to the regression\\nfit, since the presence of collinearity implies that the information that this\\nvariable provides about the response is redundant in the presence of the\\nother variables. For instance, if we regress balance ontoageandlimit,\\nwithout the ratingpredictor, then the resulting VIF values are close to\\nthe minimum possible value of 1, and the R2drops from 0.754to0.75.\\nSo dropping ratingfrom the set of predictors has effectively solved the\\ncollinearity problem without compromising the fit. The second solution is\\nto combine the collinear variables together into a single predictor. For in-\\nstance, we might take the average of standardized versions of limitand\\nratingin order to create a new variable that measures credit worthiness .\\n3.4 The Marketing Plan\\nWe now briefly return to the seven questions about the Advertising data\\nthat we set out to answer at the beginning of this chapter.\\n1.Is there a relationship between sales and advertising budget?\\nThis question can be answered by fitting a multiple regression model\\nofsalesontoTV,radio, andnewspaper , as in (3.20), and testing the\\nhypothesis H0:βTV=βradio=βnewspaper =0. In Section 3.2.2,\\nwe showed that the F-statistic can be used to determine whether\\nor not we should reject this null hypothesis. In this case the p-value\\ncorresponding to the F-statistic in Table 3.6is very low, indicating\\nclear evidence of a relationship between advertising and sales.\\n2.How strong is the relationship?\\nWe discussed two measures of model accuracy in Section 3.1.3. First,\\nthe RSE estimates the standard deviation of the response from the\\npopulation regression line. For the Advertising data, the RSE is 1.69\\nunits while the mean value for the response is 14.022, indicating a\\npercentage error of roughly 12%. Second, the R2statistic records\\nthe percentage of variability in the response that is explained by\\nthe predictors. The predictors explain almost 90% of the variance in\\nsales. The RSE and R2statistics are displayed in Table 3.6.\\n3.Which media are associated with sales?\\nTo answer this question, we can examine the p-values associated with\\neach predictor’s t-statistic (Section 3.1.2). In the multiple linear re-\\ngression displayed in Table 3.4, thep-values for TVandradioare low,\\nbut thep-value for newspaper is not. This suggests that only TVand\\nradioare related to sales. In Chapter 6we explore this question in\\ngreater detail.', metadata={'source': './documents/ISLP_website.pdf', 'page': 117}), Document(page_content='110 3. Linear Regression\\n4.How large is the association between each medium and sales?\\nWe saw in Section 3.1.2that the standard error of ˆβjcan be used to\\nconstruct confidence intervals for βj. For the Advertising data, we\\ncan use the results in Table 3.4to compute the 95% confidence inter-\\nvals for the coefficients in a multiple regression model using all three\\nmedia budgets as predictors. The confidence intervals are as follows:\\n(0.043,0.049) forTV,(0.172,0.206) forradio, and(−0.013,0.011) for\\nnewspaper . The confidence intervals for TVandradioare narrow and\\nfar from zero, providing evidence that these media are related to\\nsales. But the interval for newspaper includes zero, indicating that\\nthe variable is not statistically significant given the values of TVand\\nradio.\\nWe saw in Section 3.3.3that collinearity can result in very wide stan-\\ndard errors. Could collinearity be the reason that the confidence in-\\nterval associated with newspaper is so wide? The VIF scores are 1.005,\\n1.145, and1.145forTV,radio, andnewspaper , suggesting no evidence\\nof collinearity.\\nIn order to assess the association of each medium individually on\\nsales, we can perform three separate simple linear regressions. Re-\\nsults are shown in Tables 3.1and3.3. There is evidence of an ex-\\ntremely strong association between TVandsalesand between radio\\nandsales. There is evidence of a mild association between newspaper\\nandsales, when the values of TVandradioare ignored.\\n5.How accurately can we predict future sales?\\nThe response can be predicted using ( 3.21). The accuracy associ-\\nated with this estimate depends on whether we wish to predict an\\nindividual response, Y=f(X)+ϵ, or the average response, f(X)\\n(Section 3.2.2). If the former, we use a prediction interval, and if\\nthe latter, we use a confidence interval. Prediction intervals will al-\\nways be wider than confidence intervals because they account for the\\nuncertainty associated with ϵ, the irreducible error.\\n6.Is the relationship linear?\\nIn Section 3.3.3, we saw that residual plots can be used in order to\\nidentify non-linearity. If the relationships are linear, then the residual\\nplots should display no pattern. In the case of the Advertising data,\\nwe observe a non-linear effect in Figure 3.5, though this effect could\\nalso be observed in a residual plot. In Section 3.3.2, we discussed the\\ninclusion of transformations of the predictors in the linear regression\\nmodel in order to accommodate non-linear relationships.\\n7.Is there synergy among the advertising media?\\nThe standard linear regression model assumes an additive relation-\\nship between the predictors and the response. An additive model\\nis easy to interpret because the association between each predictor\\nand the response is unrelated to the values of the other predictors.\\nHowever, the additive assumption may be unrealistic for certain data\\nsets. In Section 3.3.2, we showed how to include an interaction term', metadata={'source': './documents/ISLP_website.pdf', 'page': 118}), Document(page_content='111\\nin the regression model in order to accommodate non-additive rela-\\ntionships. A small p-value associated with the interaction term indi-\\ncates the presence of such relationships. Figure 3.5suggested that the\\nAdvertising data may not be additive. Including an interaction term\\nin the model results in a substantial increase in R2, from around 90%\\nto almost 97%.\\n3.5 Comparison of Linear Regression\\nwithK-Nearest Neighbors\\nAs discussed in Chapter 2, linear regression is an example of a parametric\\napproach because it assumes a linear functional form for f(X). Parametric\\nmethods have several advantages. They are often easy to fit, because one\\nneed estimate only a small number of coefficients. In the case of linear re-\\ngression, the coefficients have simple interpretations, and tests of statistical\\nsignificance can be easily performed. But parametric methods do have a\\ndisadvantage: by construction, they make strong assumptions about the\\nform off(X). If the specified functional form is far from the truth, and\\nprediction accuracy is our goal, then the parametric method will perform\\npoorly. For instance, if we assume a linear relationship between XandY\\nbut the true relationship is far from linear, then the resulting model will\\nprovide a poor fit to the data, and any conclusions drawn from it will be\\nsuspect.\\nIn contrast, non-parametric methods do not explicitly assume a para-\\nmetric form for f(X), and thereby provide an alternative and more flexi-\\nble approach for performing regression. We discuss various non-parametric\\nmethods in this book. Here we consider one of the simplest and best-known\\nnon-parametricmethods, K-nearest neighbors regression (KNNregression).K-nearest\\nneighbors\\nregressionThe KNN regression method is closely related to the KNN classifier dis-\\ncussed in Chapter 2. Given a value for Kand a prediction point x0, KNN\\nregression first identifies the Ktraining observations that are closest to\\nx0, represented by N0. It then estimates f(x0)using the average of all the\\ntraining responses in N0. In other words,\\nˆf(x0)=1\\nK∑\\nxi∈N0yi.\\nFigure3.16illustratestwoKNNfitsonadatasetwith p=2predictors.The\\nfit withK=1is shown in the left-hand panel, while the right-hand panel\\ncorresponds to K=9. We see that when K=1, the KNN fit perfectly\\ninterpolates the training observations, and consequently takes the form\\nof a step function. When K=9, the KNN fit still is a step function, but\\naveraging over nine observations results in much smaller regions of constant\\nprediction, and consequently a smoother fit. In general, the optimal value\\nforKwill depend on the bias-variance tradeoff , which we introduced in\\nChapter 2. A small value for Kprovides the most flexible fit, which will\\nhave low bias but high variance. This variance is due to the fact that the\\nprediction in a given region is entirely dependent on just one observation.3.5ComparisonofLinearRegressionwithK-NearestNeighbors', metadata={'source': './documents/ISLP_website.pdf', 'page': 119}), Document(page_content='112 3. Linear Regression\\nyyx1x1x2x2yyy\\nFIGURE 3.16. Plots of ˆf(X)using KNN regression on a two-dimensional\\ndata set with 64observations (orange dots). Left:K=1results in a rough step\\nfunction fit. Right:K=9produces a much smoother fit.\\nIn contrast, larger values of Kprovide a smoother and less variable fit; the\\nprediction in a region is an average of several points, and so changing one\\nobservation has a smaller effect. However, the smoothing may cause bias by\\nmasking some of the structure in f(X). In Chapter 5, we introduce several\\napproaches for estimating test error rates. These methods can be used to\\nidentify the optimal value of Kin KNN regression.\\nInwhatsettingwillaparametricapproachsuchasleastsquareslinearre-\\ngression outperform a non-parametric approach such as KNN regression?\\nThe answer is simple: the parametric approach will outperform the non-\\nparametric approach if the parametric form that has been selected is close\\nto the true form of f. Figure3.17provides an example with data generated\\nfrom a one-dimensional linear regression model. The black solid lines rep-\\nresentf(X), while the blue curves correspond to the KNN fits using K=1\\nandK=9. In this case, the K=1predictions are far too variable, while\\nthe smoother K=9 fit is much closer to f(X). However, since the true\\nrelationship is linear, it is hard for a non-parametric approach to compete\\nwith linear regression: a non-parametric approach incurs a cost in variance\\nthat is not offset by a reduction in bias. The blue dashed line in the left-\\nhand panel of Figure 3.18represents the linear regression fit to the same\\ndata. It is almost perfect. The right-hand panel of Figure 3.18reveals that\\nlinear regression outperforms KNN for this data. The green solid line, plot-\\nted as a function of 1/K, represents the test set mean squared error (MSE)\\nfor KNN. The KNN errors are well above the black dashed line, which is\\nthe test MSE for linear regression. When the value of Kis large, then KNN\\nperforms only a little worse than least squares regression in terms of MSE.\\nIt performs far worse when Kis small.\\nIn practice, the true relationship between XandYis rarely exactly lin-\\near. Figure 3.19examines the relative performances of least squares regres-\\nsion and KNN under increasing levels of non-linearity in the relationship\\nbetween XandY. In the top row, the true relationship is nearly linear.\\nIn this case we see that the test MSE for linear regression is still superior', metadata={'source': './documents/ISLP_website.pdf', 'page': 120}), Document(page_content='3.5ComparisonofLinearRegressionwithK-NearestNeighbors113\\n−1.0 −0.5 0.0 0.5 1.01234−1.0 −0.5 0.0 0.5 1.01234yyxxFIGURE 3.17. Plots of ˆf(X)using KNN regression on a one-dimensional data\\nset with 50observations. The true relationship is given by the black solid line.\\nLeft:The blue curve corresponds to K=1and interpolates (i.e. passes directly\\nthrough) the training data. Right:The blue curve corresponds to K=9, and\\nrepresents a smoother fit.\\n−1.0 −0.5 0.0 0.5 1.012340.20.5 1.00.00 0.05 0.10 0.15Mean Squared Erroryx1/KFIGURE 3.18. The same data set shown in Figure 3.17is investigated further.\\nLeft:The blue dashed line is the least squares fit to the data. Since f(X)is in\\nfact linear (displayed as the black line), the least squares regression line provides\\na very good estimate of f(X).Right:The dashed horizontal line represents the\\nleast squares test set MSE, while the green solid line corresponds to the MSE\\nfor KNN as a function of 1/K(on the log scale). Linear regression achieves a\\nlower test MSE than does KNN regression, since f(X)is in fact linear. For KNN\\nregression, the best results occur with a very large value of K, corresponding to a\\nsmall value of 1/K.', metadata={'source': './documents/ISLP_website.pdf', 'page': 121}), Document(page_content='114 3. Linear Regression\\n−1.0 −0.5 0.0 0.5 1.00.5 1.0 1.5 2.0 2.5 3.0 3.50.20.51.00.00 0.02 0.04 0.06 0.08Mean Squared Error\\n−1.0 −0.5 0.0 0.5 1.01.0 1.5 2.0 2.5 3.0 3.50.20.51.00.00 0.05 0.10 0.15Mean Squared Erroryy\\nxx\\n1/K1/K\\nFIGURE 3.19. Top Left: In a setting with a slightly non-linear relationship\\nbetween XandY(solid black line), the KNN fits with K=1(blue) and K=9\\n(red) are displayed. Top Right: For the slightly non-linear data, the test set MSE\\nfor least squares regression (horizontal black) and KNN with various values of\\n1/K(green) are displayed. Bottom Left and Bottom Right: As in the top panel,\\nbut with a strongly non-linear relationship between XandY.\\nto that of KNN for low values of K. However, for K≥4, KNN out-\\nperforms linear regression. The second row illustrates a more substantial\\ndeviation from linearity. In this situation, KNN substantially outperforms\\nlinear regression for all values of K. Note that as the extent of non-linearity\\nincreases, there is little change in the test set MSE for the non-parametric\\nKNN method, but there is a large increase in the test set MSE of linear\\nregression.\\nFigures3.18and3.19display situations in which KNN performs slightly\\nworse than linear regression when the relationship is linear, but much bet-\\nter than linear regression for nonlinear situations. In a real life situation\\nin which the true relationship is unknown, one might suspect that KNN\\nshould be favored over linear regression because it will at worst be slightly\\ninferior to linear regression if the true relationship is linear, and may give\\nsubstantially better results if the true relationship is non-linear. But in re-\\nality, even when the true relationship is highly non-linear, KNN may still\\nprovide inferior results to linear regression. In particular, both Figures 3.18', metadata={'source': './documents/ISLP_website.pdf', 'page': 122}), Document(page_content='3.5 Comparison of Linear Regression with K-Nearest Neighbors 115\\n0.2 0.5 1.00.0 0.2 0.4 0.6 0.8 1.0p=1\\n0.2 0.5 1.00.0 0.2 0.4 0.6 0.8 1.0p=2\\n0.2 0.5 1.00.0 0.2 0.4 0.6 0.8 1.0p=3\\n0.2 0.5 1.00.0 0.2 0.4 0.6 0.8 1.0p=4\\n0.2 0.5 1.00.0 0.2 0.4 0.6 0.8 1.0p=10\\n0.2 0.5 1.00.0 0.2 0.4 0.6 0.8 1.0p=20Mean Squared Error1/KFIGURE 3.20. Test MSE for linear regression (black dashed lines) and KNN\\n(green curves) as the number of variables pincreases. The true function is non-\\nlinear in the first variable, as in the lower panel in Figure 3.19, and does not\\ndepend on the additional variables. The performance of linear regression deteri-\\norates slowly in the presence of these additional noise variables, whereas KNN’s\\nperformance degrades much more quickly as pincreases.\\nand3.19illustrate settings with p=1predictor. But in higher dimensions,\\nKNN often performs worse than linear regression.\\nFigure3.20considers the same strongly non-linear situation as in the\\nsecond row of Figure 3.19, except that we have added additional noise\\npredictors that are not associated with the response. When p=1orp=2,\\nKNN outperforms linear regression. But for p=3 the results are mixed,\\nand forp≥4linear regression is superior to KNN. In fact, the increase in\\ndimension has only caused a small deterioration in the linear regression test\\nset MSE, but it has caused more than a ten-fold increase in the MSE for\\nKNN.Thisdecreaseinperformanceasthedimensionincreasesisacommon\\nproblem for KNN, and results from the fact that in higher dimensions\\nthere is effectively a reduction in sample size. In this data set there are\\n50training observations; when p=1, this provides enough information to\\naccurately estimate f(X). However, spreading 50observations over p= 20\\ndimensions results in a phenomenon in which a given observation has no\\nnearby neighbors —this is the so-called curse of dimensionality . That is,curse of di-\\nmensionalitytheKobservations that are nearest to a given test observation x0may be\\nvery far away from x0inp-dimensional space when pis large, leading to a\\nvery poor prediction of f(x0)and hence a poor KNN fit. As a general rule,\\nparametric methods will tend to outperform non-parametric approaches\\nwhen there is a small number of observations per predictor.\\nEven when the dimension is small, we might prefer linear regression to\\nKNN from an interpretability standpoint. If the test MSE of KNN is only\\nslightly lower than that of linear regression, we might be willing to forego\\na little bit of prediction accuracy for the sake of a simple model that can\\nbe described in terms of just a few coefficients, and for which p-values are\\navailable.', metadata={'source': './documents/ISLP_website.pdf', 'page': 123}), Document(page_content=\"116 3. Linear Regression3.6 Lab: Linear Regression3.6.1 Importing packagesWe import our standard libraries at this top level.In[1]:import numpy as npimport pandas as pdfrom matplotlib.pyplot import subplotsNew importsThroughoutthislabwewillintroducenewfunctionsandlibraries.However,we will import them here to emphasize these are the new code objects inthis lab. Keeping imports near the top of a notebook makes the code morereadable, since scanning the first few lines tells us what libraries are used.In[2]:import statsmodels.api as smWe will provide relevant details about the functions below as they areneeded.Besides importing whole modules, it is also possible to import only afew items from a given module. This will help keep thenamespaceclean.namespaceWe will use a few specific objects from thestatsmodelspackage which westatsmodelsimport here.In[3]:from statsmodels.stats.outliers_influence \\\\import variance_inflation_factor as VIFfrom statsmodels.stats.anova import anova_lmAs one of the import statements above is quite a long line, we inserted aline break\\\\to ease readability.We will also use some functions written for the labs in this book in theISLPpackage.In[4]:from ISLP import load_datafrom ISLP.models import (ModelSpec as MS,summarize,poly)Inspecting Objects and NamespacesThe functiondir()provides a list of objects in a namespace.dir()In[5]:dir()Out[5]:['In','MS','_','__','___','__builtin__','__builtins__',...\", metadata={'source': './documents/ISLP_website.pdf', 'page': 124}), Document(page_content='3.6 Lab: Linear Regression 117\\'poly\\',\\'quit\\',\\'sm\\',\\'summarize\\']This shows you everything thatPythoncan find at the top level. Thereare certain objects like__builtins__that contain references to built-infunctions likeprint().Every python object has its own notion of namespace, also accessiblewithdir(). This will include both the attributes of the object as well asany methods associated with it. For instance, we see\\'sum\\'in the listingfor an array.In[6]:A=n p . a r r a y ( [ 3 , 5 , 1 1 ] )dir(A)Out[6]:...\\'strides\\',\\'sum\\',\\'swapaxes\\',...This indicates that the objectA.sumexists. In this case it is a method thatcan be used to compute the sum of the arrayAas can be seen by typingA.sum?.In[7]:A.sum()Out[7]:193.6.2 Simple Linear RegressionIn this section we will construct model matrices (also called design matri-ces) using theModelSpec()transform fromISLP.models.We will use theBostonhousing data set, which is contained in theISLPpackage. TheBostondataset recordsmedv(median house value) for506neighborhoods around Boston. We will build a regression model to pre-dictmedvusing13predictors such asrmvar(average number of rooms perhouse),age(proportion of owner-occupied units built prior to 1940), andlstat(percent of households with low socioeconomic status). We will usestatsmodelsfor this task, aPythonpackage that implements several com-monly used regression methods.Wehaveincludedasimpleloadingfunctionload_data()intheISLPpack-load_data()age:In[8]:Boston = load_data(\"Boston\")Boston.columnsOut[8]:Index([\\'crim\\',\\'zn\\',\\'indus\\',\\'chas\\',\\'nox\\',\\'rm\\',\\'age\\',\\'dis\\',\\'rad\\',\\'tax\\',\\'ptratio\\',\\'black\\',\\'lstat\\',\\'medv\\'],dtype=\\'object\\')', metadata={'source': './documents/ISLP_website.pdf', 'page': 125}), Document(page_content=\"118 3. Linear RegressionTypeBoston?to find out more about these data.We start by using thesm.OLS()function to fit a simple linear regressionsm.OLS()model. Our response will bemedvandlstatwill be the single predictor.For this model, we can create the model matrix by hand.In[9]:X=p d . D a t a F r a m e ( {'intercept':n p . o n e s ( B o s t o n . s h a p e [ 0 ] ) ,'lstat':B o s t o n ['lstat']})X[:4]Out[9]:intercept lstat01 . 0 4 . 9 811 . 0 9 . 1 421 . 0 4 . 0 331 . 0 2 . 9 4We extract the response, and fit the model.In[10]:y=B o s t o n ['medv']model = sm.OLS(y, X)results = model.fit()Note thatsm.OLS()does not fit the model; it specifies the model, and thenmodel.fit()does the actual fitting.OurISLPfunctionsummarize()produces a simple table of the parame-summarize()ter estimates, their standard errors, t-statistics and p-values. The functiontakes a single argument, such as the objectresultsreturned here by thefitmethod, and returns such a summary.In[11]:summarize(results)Out[11]:coef std err t P>|t|intercept 34.5538 0.563 61.415 0.0lstat -0.9500 0.039 -24.528 0.0Before we describe other methods for working with fitted models, weoutline a more useful and general framework for constructing a model ma-trixX.Using Transformations: Fit and TransformOur model above has a single predictor, and constructingXwas straight-forward. In practice we often fit models with more than one predictor,typically selected from an array or data frame. We may wish to introducetransformations to the variables before fitting the model, specify interac-tions between variables, and expand some particular variables into sets ofvariables (e.g. polynomials). Thesklearnpackage has a particular notionsklearnfor this type of task: atransform. A transform is an object that is createdwith some parameters as arguments. The object has two main methods:fit()andtransform()..fit().transform()We provide a general approach for specifying models and constructingthe model matrix through the transformModelSpec()in theISLPlibrary.ModelSpec()ModelSpec()(renamedMS()in the preamble) creates a transform object,and then a pair of methodstransform()andfit()are used to construct acorresponding model matrix.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 126}), Document(page_content=\"3.6 Lab: Linear Regression 119We first describe this process for our simple regression model using asingle predictorlstatin theBostondata frame, but will use it repeatedlyin more complex tasks in this and other labs in this book. In our case thetransform is created by the expressiondesign = MS(['lstat']).Thefit()method takes the original array and may do some initial com-putations on it, as specified in the transform object. For example, it maycompute means and standard deviations for centering and scaling. Thetransform()method applies the fitted transformation to the array of data,and produces the model matrix.In[12]:design = MS(['lstat'])design = design.fit(Boston)X=d e s i g n . t r a n s f o r m ( B o s t o n )X[:4]Out[12]:intercept lstat01 . 0 4 . 9 811 . 0 9 . 1 421 . 0 4 . 0 331 . 0 2 . 9 4In this simple case, thefit()method does very little; it simply checks thatthe variable'lstat'specified indesignexists inBoston. Thentransform()constructs the model matrix with two columns: aninterceptand the vari-ablelstat.Thesetwooperationscanbecombinedwiththefit_transform()method..fit_transform()In[13]:design = MS(['lstat'])X=d e s i g n . f i t _ t r a n s f o r m ( B o s t o n )X[:4]Out[13]:intercept lstat01 . 0 4 . 9 811 . 0 9 . 1 421 . 0 4 . 0 331 . 0 2 . 9 4Note that, as in the previous code chunk when the two steps were doneseparately, thedesignobject is changed as a result of thefit()operation.The power of this pipeline will become clearer when we fit more complexmodels that involve interactions and transformations.Let’sreturntoourfittedregressionmodel.Theobjectresultshasseveralmethods that can be used for inference. We already presented a functionsummarize()for showing the essentials of the fit. For a full and somewhatexhaustive summary of the fit, we can use thesummary()method (outputnot shown).In[14]:results.summary()The fitted coefficients can also be retrieved as theparamsattribute ofresults.In[15]:results.params\", metadata={'source': './documents/ISLP_website.pdf', 'page': 127}), Document(page_content=\"120 3. Linear RegressionOut[15]:intercept 34.553841lstat -0.950049dtype: float64Theget_prediction()method can be used to obtain predictions, and.get_prediction()produce confidence intervals and prediction intervals for the prediction ofmedvfor given values oflstat.We first create a new data frame, in this case containing only the vari-ablelstat, with the values for this variable at which we wish to makepredictions. We then use thetransform()method ofdesignto create thecorresponding model matrix.In[16]:new_df = pd.DataFrame({'lstat':[5, 10, 15]})newX = design.transform(new_df)newXOut[16]:intercept lstat01 . 0511 . 0 1 021 . 0 1 5Next we compute the predictions atnewX, and view them by extractingthepredicted_meanattribute.In[17]:new_predictions = results.get_prediction(newX);new_predictions.predicted_meanOut[17]:array([29.80359411, 25.05334734, 20.30310057])We can produce confidence intervals for the predicted values.In[18]:new_predictions.conf_int(alpha=0.05)Out[18]:array([[29.00741194, 30.59977628],[24.47413202, 25.63256267],[19.73158815, 20.87461299]])Prediction intervals are computing by settingobs=True:In[19]:new_predictions.conf_int(obs=True, alpha=0.05)Out[19]:array([[17.56567478, 42.04151344],[12.82762635, 37.27906833],[8 . 0 7 7 7 4 2 1,3 2 . 5 2 8 4 5 9 0 5 ] ] )For instance, the 95% confidence interval associated with anlstatvalue of10 is (24.47, 25.63), and the 95% prediction interval is (12.82, 37.28). Asexpected, the confidence and prediction intervals are centered around thesame point (a predicted value of 25.05 formedvwhenlstatequals 10), butthe latter are substantially wider.Next we will plotmedvandlstatusingDataFrame.plot.scatter(), and.plot.scatter()wish to add the regression line to the resulting plot.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 128}), Document(page_content='3.6 Lab: Linear Regression 121\\nDefining Functions\\nWhile there is a function within the ISLPpackage that adds a line to an\\nexisting plot, we take this opportunity to define our first function to do so.def\\nIn[20]: def abline(ax, b, m):\\n\"Add a line with slope m and intercept b to ax\"\\nxlim = ax.get_xlim()\\nylim = [m * xlim[0] + b, m * xlim[1] + b]\\nax.plot(xlim, ylim)\\nA few things are illustrated above. First we see the syntax for defining a\\nfunction: def funcname(...) . The function has arguments ax, b, m where\\naxis an axis object for an exisiting plot, bis the intercept and mis the slope\\nof the desired line. Other plotting options can be passed on to ax.plotby\\nincluding additional optional arguments as follows:\\nIn[21]: def abline(ax, b, m, *args, **kwargs):\\n\"Add a line with slope m and intercept b to ax\"\\nxlim = ax.get_xlim()\\nylim = [m * xlim[0] + b, m * xlim[1] + b]\\nax.plot(xlim, ylim, *args, **kwargs)\\nThe addition of *argsallows any number of non-named arguments to\\nabline, while *kwargs allows any number of named arguments (such as\\nlinewidth=3 ) toabline. In our function, we pass these arguments verbatim\\ntoax.plotabove. Readers interested in learning more about functions are\\nreferred to the section on defining functions in docs.python.org/tutorial .\\nLet’s use our new function to add this regression line to a plot of medv\\nvs.lstat.\\nIn[22]: ax = Boston.plot.scatter( \\'lstat\\' ,\\'medv\\')\\nabline(ax,\\nresults.params[0],\\nresults.params[1],\\n\\'r--\\',\\nlinewidth=3)\\nThus,thefinalcallto ax.plot() isax.plot(xlim, ylim, \\'r--\\', linewidth=3) .\\nWe have used the argument \\'r--\\'to produce a red dashed line, and added\\nan argument to make it of width 3. There is some evidence for non-linearity\\nin the relationship between lstatandmedv. We will explore this issue later\\nin this lab.\\nAs mentioned above, there is an existing function to add a line to a plot\\n—ax.axline() — but knowing how to write such functions empowers us\\nto create more expressive displays.\\nNext we examine some diagnostic plots, several of which were discussed\\nin Section 3.3.3. We can find the fitted values and residuals of the fit as\\nattributes of the resultsobject. Various influence measures describing the\\nregression model are computed with the get_influence() method. As we.get_\\ninfluence() will not use the figcomponent returned as the first value from subplots() ,\\nwe simply capture the second returned value in axbelow.\\nIn[23]: ax = subplots(figsize=(8,8))[1]', metadata={'source': './documents/ISLP_website.pdf', 'page': 129}), Document(page_content=\"122 3. Linear Regressionax.scatter(results.fittedvalues, results.resid)ax.set_xlabel('Fitted value')ax.set_ylabel('Residual')ax.axhline(0, c='k',l s ='--');We add a horizontal line at 0 for reference using theax.axhline()method,.axhline()indicating it should be black (c='k') and have a dashed linestyle (ls='--').On the basis of the residual plot (not shown), there is some evidenceof non-linearity. Leverage statistics can be computed for any number ofpredictors using thehat_matrix_diagattribute of the value returned by theget_influence()method.In[24]:infl = results.get_influence()ax = subplots(figsize=(8,8))[1]ax.scatter(np.arange(X.shape[0]), infl.hat_matrix_diag)ax.set_xlabel('Index')ax.set_ylabel('Leverage')np.argmax(infl.hat_matrix_diag)Out[24]:374Thenp.argmax()function identifies the index of the largest element of annp.argmax()array, optionally computed over an axis of the array. In this case, we maxi-mized over the entire array to determine which observation has the largestleverage statistic.3.6.3 Multiple Linear RegressionIn order to fit a multiple linear regression model using least squares, weagainusetheModelSpec()transformtoconstructtherequiredmodelmatrixand response. The arguments toModelSpec()can be quite general, but inthis case a list of column names suffice. We consider a fit here with the twovariableslstatandage.In[25]:X=M S ( ['lstat','age']).fit_transform(Boston)model1 = sm.OLS(y, X)results1 = model1.fit()summarize(results1)Out[25]:coef std err t P>|t|intercept 33.2228 0.731 45.458 0.000lstat -1.0321 0.048 -21.416 0.000age 0.0345 0.012 2.826 0.005Notice how we have compacted the first line into a succinct expressiondescribing the construction ofX.TheBostondatasetcontains12variables,andsoitwouldbecumbersometo have to type all of these in order to perform a regression using all of thepredictors. Instead, we can use the following short-hand:.columns.drop()In[26]:terms = Boston.columns.drop('medv')terms\", metadata={'source': './documents/ISLP_website.pdf', 'page': 130}), Document(page_content=\"3.6 Lab: Linear Regression 123\\nOut[26]: Index([ 'crim','zn','indus' ,'chas','nox','rm','age','dis',\\n'rad','tax','ptratio' ,'lstat' ],\\ndtype='object' )\\nWe can now fit the model with all the variables in termsusing the same\\nmodel matrix builder.\\nIn[27]: X=M S ( t e r m s ) . f i t _ t r a n s f o r m ( B o s t o n )\\nmodel = sm.OLS(y, X)\\nresults = model.fit()\\nsummarize(results)\\nOut[27]: coef std err t P>|t|\\nintercept 41.6173 4.936 8.431 0.000\\ncrim -0.1214 0.033 -3.678 0.000\\nzn 0.0470 0.014 3.384 0.001\\nindus 0.0135 0.062 0.217 0.829\\nchas 2.8400 0.870 3.264 0.001\\nnox -18.7580 3.851 -4.870 0.000\\nrm 3.6581 0.420 8.705 0.000\\nage 0.0036 0.013 0.271 0.787\\ndis -1.4908 0.202 -7.394 0.000\\nrad 0.2894 0.067 4.325 0.000\\ntax -0.0127 0.004 -3.337 0.001\\nptratio -0.9375 0.132 -7.091 0.000\\nlstat -0.5520 0.051 -10.897 0.000\\nWhatifwewouldliketoperformaregressionusingallofthevariablesbut\\none? For example, in the above regression output, agehas a high p-value.\\nSo we may wish to run a regression excluding this predictor. The following\\nsyntax results in a regression using all predictors except age(output not\\nshown).\\nIn[28]: minus_age = Boston.columns.drop([ 'medv','age'])\\nXma = MS(minus_age).fit_transform(Boston)\\nmodel1 = sm.OLS(y, Xma)\\nsummarize(model1.fit())\\n3.6.4 Multivariate Goodness of Fit\\nWecanaccesstheindividualcomponentsof resultsbyname( dir(results)\\nshows us what is available). Hence results.rsquared gives us the R2, and\\nnp.sqrt(results.scale) gives us the RSE.\\nVariance inflation factors (section 3.3.3) are sometimes useful to assess\\nthe effect of collinearity in the model matrix of a regression model. We will\\ncompute the VIFs in our multiple regression fit, and use the opportunity\\nto introduce the idea of list comprehension .list compre-\\nhension\\nList Comprehension\\nOften we encounter a sequence of objects which we would like to transform\\nfor some other task. Below, we compute the VIF for each feature in our X\\nmatrix and produce a data frame whose index agrees with the columns of\\nX. The notion of list comprehension can often make such a task easier.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 131}), Document(page_content='124 3. Linear RegressionList comprehensions are simple and powerful ways to form lists ofPythonobjects. The language also supports dictionary andgeneratorcomprehen-sion, though these are beyond our scope here. Let’s look at an example.We compute the VIF for each of the variables in the model matrixX, usingthe functionvariance_inflation_factor().variance_inflation_factor()In[29]:vals = [VIF(X, i)for i in range(1, X.shape[1])]vif = pd.DataFrame({\\'vif\\':vals},index=X.columns[1:])vifOut[29]:vifcrim 1.767zn 2.298indus 3.987chas 1.071nox 4.369rm 1.913age 3.088dis 3.954rad 7.445tax 9.002ptratio 1.797lstat 2.871The functionVIF()takes two arguments: a dataframe or array, and a vari-ablecolumnindex.InthecodeabovewecallVIF()ontheflyforallcolumnsinX. We have excluded column 0 above (the intercept), which is not of in-terest. In this case the VIFs are not that exciting.The objectvalsabove could have been constructed with the followingfor loop:In[30]:vals = []for i in range(1, X.values.shape[1]):vals.append(VIF(X.values, i))List comprehension allows us to perform such repetitive operations in amore straightforward way.3.6.5 Interaction TermsIt is easy to include interaction terms in a linear model usingModelSpec().Including a tuple(\"lstat\",\"age\")tells the model matrix builder to includean interaction term betweenlstatandage.In[31]:X=M S ( [\\'lstat\\',\\'age\\',(\\'lstat\\',\\'age\\')]).fit_transform(Boston)model2 = sm.OLS(y, X)summarize(model2.fit())Out[31]:coef std err t P>|t|intercept 36.0885 1.470 24.553 0.000lstat -1.3921 0.167 -8.313 0.000', metadata={'source': './documents/ISLP_website.pdf', 'page': 132}), Document(page_content=\"3.6 Lab: Linear Regression 125\\nage -0.0007 0.020 -0.036 0.971\\nlstat:age 0.0042 0.002 2.244 0.025\\n3.6.6 Non-linear Transformations of the Predictors\\nThemodelmatrixbuildercanincludetermsbeyondjustcolumnnamesand\\ninteractions. For instance, the poly()function supplied in ISLPspecifiespoly()that columns representing polynomial functions of its first argument are\\nadded to the model matrix.\\nIn[32]: X=M S ( [ p o l y ( 'lstat' ,d e g r e e = 2 ) , 'age']).fit_transform(Boston)\\nmodel3 = sm.OLS(y, X)\\nresults3 = model3.fit()\\nsummarize(results3)\\nOut[32]: coef std err t P>|t|\\nintercept 17.7151 0.781 22.681 0.000\\npoly(lstat, degree=2)[0] -179.2279 6.733 -26.620 0.000\\npoly(lstat, degree=2)[1] 72.9908 5.482 13.315 0.000\\nage 0.0703 0.011 6.471 0.000\\nThe effectively zero p-value associated with the quadratic term (i.e. the\\nthird row above) suggests that it leads to an improved model.\\nBy default, poly()creates a basis matrix for inclusion in the model ma-\\ntrix whose columns are orthogonal polynomials , which are designed for sta-orthogonal\\npolynomialble least squares computations.13Alternatively, had we included an argu-\\nmentraw=True in the above call to poly(), the basis matrix would consist\\nsimplyof lstatandlstat**2 .Sinceeitherofthesebasesrepresentquadratic\\npolynomials, the fitted values would not change in this case, just the poly-\\nnomial coefficients. Also by default, the columns created by poly()do not\\ninclude an intercept column as that is automatically added by MS().\\nWe use the anova_lm() function to further quantify the extent to whichanova_lm()the quadratic fit is superior to the linear fit.\\nIn[33]: anova_lm(results1, results3)\\nOut[33]: df_resid ssr df_diff ss_diff F Pr(>F)\\n05 0 3 . 0 1 9 1 6 8 . 1 3 0 . 0 N a N N a NN a N\\n15 0 2 . 0 1 4 1 6 5 . 6 1 1 . 0 5 0 0 2 . 5 2 1 7 7 . 2 8 7 . 4 7 e - 3 5\\nHereresults1 represents the linear submodel containing predictors lstat\\nandage, while results3 corresponds to the larger model above with a\\nquadratic term in lstat. Theanova_lm() function performs a hypothesis\\ntest comparing the two models. The null hypothesis is that the quadratic\\nterm in the bigger model is not needed, and the alternative hypothesis is\\nthat the bigger model is superior. Here the F-statistic is 177.28 and the\\nassociated p-value is zero. In this case the F-statistic is the square of the\\nt-statistic for the quadratic term in the linear model summary for results3\\n—aconsequenceofthefactthatthesenestedmodelsdifferbyonedegreeof\\n13Actually, poly() is a wrapper for the workhorse and standalone function Poly()\\nthat does the work in building the model matrix.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 133}), Document(page_content=\"126 3. Linear Regressionfreedom. This provides very clear evidence that the quadratic polynomialinlstatimproves the linear model. This is not surprising, since earlier wesaw evidence for non-linearity in the relationship betweenmedvandlstat.The functionanova_lm()can take more than two nested models as input,inwhichcaseitcompareseverysuccessivepairofmodels.Thatalsoexplainswhy their areNaNs in the first row above, since there is no previous modelwith which to compare the first.In[34]:ax = subplots(figsize=(8,8))[1]ax.scatter(results3.fittedvalues, results3.resid)ax.set_xlabel('Fitted value')ax.set_ylabel('Residual')ax.axhline(0, c='k',l s ='--')We see that when the quadratic term is included in the model, there islittle discernible pattern in the residuals. In order to create a cubic orhigher-degree polynomial fit, we can simply change the degree argument topoly().3.6.7 Qualitative PredictorsHere we use theCarseatsdata, which is included in theISLPpackage. Wewill attempt to predictSales(child car seat sales) in 400 locations basedon a number of predictors.In[35]:Carseats = load_data('Carseats')Carseats.columnsOut[35]:Index(['Sales','CompPrice','Income','Advertising','Population','Price','ShelveLoc','Age','Education','Urban','US'],dtype='object')TheCarseatsdata includes qualitative predictors such asShelveLoc, anindicator of the quality of the shelving location — that is, the space withina store in which the car seat is displayed. The predictorShelveLoctakes onthree possible values,Bad,Medium, andGood. Given a qualitative variablesuch asShelveLoc,ModelSpec()generates dummy variables automatically.Thesevariablesareoftenreferredtoasaone-hot encodingofthecategoricalone-hotencodingfeature. Their columns sum to one, so to avoid collinearity with an inter-cept, the first column is dropped. Below we see the columnShelveLoc[Bad]has been dropped, sinceBadis the first level ofShelveLoc. Below we fit amultiple regression model that includes some interaction terms.In[36]:allvars = list(Carseats.columns.drop('Sales'))y=C a r s e a t s ['Sales']final = allvars + [('Income','Advertising'),('Price','Age')]X=M S ( f i n a l ) . f i t _ t r a n s f o r m ( C a r s e a t s )model = sm.OLS(y, X)summarize(model.fit())Out[36]:coef std err t P>|t|intercept 6.5756 1.009 6.519 0.000\", metadata={'source': './documents/ISLP_website.pdf', 'page': 134}), Document(page_content='3.7 Exercises 127\\nCompPrice 0.0929 0.004 22.567 0.000\\nIncome 0.0109 0.003 4.183 0.000\\nAdvertising 0.0702 0.023 3.107 0.002\\nPopulation 0.0002 0.000 0.433 0.665\\nPrice -0.1008 0.007 -13.549 0.000\\nShelveLoc[Good] 4.8487 0.153 31.724 0.000\\nShelveLoc[Medium] 1.9533 0.126 15.531 0.000\\nAge -0.0579 0.016 -3.633 0.000\\nEducation -0.0209 0.020 -1.063 0.288\\nUrban[Yes] 0.1402 0.112 1.247 0.213\\nUS[Yes] -0.1576 0.149 -1.058 0.291\\nIncome:Advertising 0.0008 0.000 2.698 0.007\\nPrice:Age 0.0001 0.000 0.801 0.424\\nIn the first line above, we made allvarsa list, so that we could add the\\ninteraction terms two lines down. Our model-matrix builder has created a\\nShelveLoc[Good] dummy variable that takes on a value of 1 if the shelving\\nlocation is good, and 0 otherwise. It has also created a ShelveLoc[Medium]\\ndummy variable that equals 1 if the shelving location is medium, and 0 oth-\\nerwise. A bad shelving location corresponds to a zero for each of the two\\ndummy variables. The fact that the coefficient for ShelveLoc[Good] in the\\nregression output is positive indicates that a good shelving location is asso-\\nciated with high sales (relative to a bad location). And ShelveLoc[Medium]\\nhas a smaller positive coefficient, indicating that a medium shelving loca-\\ntion leads to higher sales than a bad shelving location, but lower sales than\\na good shelving location.\\n3.7 Exercises\\nConceptual\\n1. Describe the null hypotheses to which the p-values given in Table 3.4\\ncorrespond. Explain what conclusions you can draw based on these\\np-values. Your explanation should be phrased in terms of sales,TV,\\nradio, andnewspaper , rather than in terms of the coefficients of the\\nlinear model.\\n2. CarefullyexplainthedifferencesbetweentheKNNclassifierandKNN\\nregression methods.\\n3. Suppose we have a data set with five predictors, X1=GPA,X2=\\nIQ,X3=Level (1for College and 0for High School), X4=Interac-\\ntion between GPA and IQ, and X5=Interaction between GPA and\\nLevel. The response is starting salary after graduation (in thousands\\nof dollars). Suppose we use least squares to fit the model, and get\\nˆβ0= 50 ,ˆβ1= 20 ,ˆβ2=0.07,ˆβ3= 35 ,ˆβ4=0.01,ˆβ5=−10.\\n(a) Which answer is correct, and why?\\ni. For a fixed value of IQ and GPA, high school graduates earn\\nmore, on average, than college graduates.\\nii. For a fixed value of IQ and GPA, college graduates earn\\nmore, on average, than high school graduates.', metadata={'source': './documents/ISLP_website.pdf', 'page': 135}), Document(page_content='128 3. Linear Regressioniii. For a fixed value of IQ and GPA, high school graduates earnmore, on average, than college graduates provided that theGPA is high enough.iv. For a fixed value of IQ and GPA, college graduates earnmore, on average, than high school graduates provided thatthe GPA is high enough.(b) Predict the salary of a college graduate with IQ of110and aGPA of4.0.(c) True or false: Since the coefficient for the GPA/IQ interactionterm is very small, there is very little evidence of an interactioneffect. Justify your answer.4. I collect a set of data (n= 100observations) containing a singlepredictor and a quantitative response. I then fit a linear regressionmodel to the data, as well as a separate cubic regression, i.e.Y=β0+β1X+β2X2+β3X3+ϵ.(a) Suppose that the true relationship between X and Y is linear,i.e.Y=β0+β1X+ϵ. Consider the training residual sum ofsquares (RSS) for the linear regression, and also the trainingRSS for the cubic regression. Would we expect one to be lowerthan the other, would we expect them to be the same, or is therenot enough information to tell? Justify your answer.(b) Answer (a) using test rather than training RSS.(c) SupposethatthetruerelationshipbetweenXandYisnotlinear,butwedon’tknowhowfaritisfromlinear.ConsiderthetrainingRSS for the linear regression, and also the training RSS for thecubic regression. Would we expect one to be lower than theother, would we expect them to be the same, or is there notenough information to tell? Justify your answer.(d) Answer (c) using test rather than training RSS.5. Consider the fitted values that result from performing linear regres-sion without an intercept. In this setting, theith fitted value takesthe formˆyi=xiˆβ,whereˆβ=(n∑i=1xiyi)/(n∑i′=1x2i′).(3.38)Show that we can writeˆyi=n∑i′=1ai′yi′.What isai′?Note: We interpret this result by saying that the fitted values fromlinear regression arelinear combinationsof the response values.', metadata={'source': './documents/ISLP_website.pdf', 'page': 136}), Document(page_content='3.7 Exercises 129\\n6. Using ( 3.4), argue that in the case of simple linear regression, the\\nleast squares line always passes through the point (¯x,¯y).\\n7. It is claimed in the text that in the case of simple linear regression\\nofYontoX, theR2statistic ( 3.17) is equal to the square of the\\ncorrelation between XandY(3.18). Prove that this is the case. For\\nsimplicity, you may assume that ¯x=¯y=0.\\nApplied\\n8. This question involves the use of simple linear regression on the Auto\\ndata set.\\n(a) Use the sm.OLS() function to perform a simple linear regression\\nwithmpgas the response and horsepower as the predictor. Use\\nthesummarize() function to print the results. Comment on the\\noutput. For example:\\ni. Is there a relationship between the predictor and the re-\\nsponse?\\nii. How strong is the relationship between the predictor and\\nthe response?\\niii. Is the relationship between the predictor and the response\\npositive or negative?\\niv. What is the predicted mpgassociated with a horsepower of\\n98? What are the associated 95% confidence and prediction\\nintervals?\\n(b) Plot the response and the predictor in a new set of axes ax. Use\\ntheax.axline() method or the abline() function defined in the\\nlab to display the least squares regression line.\\n(c) Produce some of diagnostic plots of the least squares regression\\nfit as described in the lab. Comment on any problems you see\\nwith the fit.\\n9. This question involves the use of multiple linear regression on the\\nAutodata set.\\n(a) Produce a scatterplot matrix which includes all of the variables\\nin the data set.\\n(b) Compute the matrix of correlations between the variables using\\ntheDataFrame.corr() method..corr()\\n(c) Usethe sm.OLS() functiontoperformamultiplelinearregression\\nwithmpgas the response and all other variables except nameas\\nthe predictors. Use the summarize() function to print the results.\\nComment on the output. For instance:\\ni. Is there a relationship between the predictors and the re-\\nsponse? Use the anova_lm() function from statsmodels to\\nanswer this question.', metadata={'source': './documents/ISLP_website.pdf', 'page': 137}), Document(page_content='130 3. Linear Regressionii. Which predictors appear to have a statistically significantrelationship to the response?iii. What does the coefficient for theyearvariable suggest?(d) Produce some of diagnostic plots of the linear regression fit asdescribed in the lab. Comment on any problems you see with thefit. Do the residual plots suggest any unusually large outliers?Does the leverage plot identify any observations with unusuallyhigh leverage?(e) Fit some models with interactions as described in the lab. Doany interactions appear to be statistically significant?(f) Try a few different transformations of the variables, such aslog(X),√X,X2. Comment on your findings.10. This question should be answered using theCarseatsdata set.(a) Fit a multiple regression model to predictSalesusingPrice,Urban, andUS.(b) Provide an interpretation of each coefficient in the model. Becareful—some of the variables in the model are qualitative!(c) Write out the model in equation form, being careful to handlethe qualitative variables properly.(d) For which of the predictors can you reject the null hypothesisH0:βj=0?(e) On the basis of your response to the previous question, fit asmaller model that only uses the predictors for which there isevidence of association with the outcome.(f) How well do the models in (a) and (e) fit the data?(g) Using the model from (e), obtain 95% confidence intervals forthe coefficient(s).(h) Is there evidence of outliers or high leverage observations in themodel from (e)?11. In this problem we will investigate thet-statistic for the null hypoth-esisH0:β=0in simple linear regression without an intercept. Tobegin, we generate a predictorxand a responseyas follows.rng = np.random.default_rng(1)x=r n g . n o r m a l ( s i z e = 1 0 0 )y=2*x+r n g . n o r m a l ( s i z e = 1 0 0 )(a) Perform a simple linear regression ofyontox,withoutan in-tercept. Report the coefficient estimateˆβ, the standard error ofthis coefficient estimate, and thet-statistic andp-value associ-ated with the null hypothesisH0:β=0. Comment on theseresults. (You can perform regression without an intercept usingthe keywords argumentintercept=FalsetoModelSpec().)', metadata={'source': './documents/ISLP_website.pdf', 'page': 138}), Document(page_content='3.7 Exercises 131\\n(b) Now perform a simple linear regression of xontoywithout an\\nintercept, and report the coefficient estimate, its standard error,\\nand the corresponding t-statistic and p-values associated with\\nthe null hypothesis H0:β=0. Comment on these results.\\n(c) What is the relationship between the results obtained in (a) and\\n(b)?\\n(d) For the regression of YontoXwithout an intercept, the t-\\nstatistic for H0:β=0 takes the form ˆβ/SE(ˆβ), where ˆβis\\ngiven by ( 3.38), and where\\nSE(ˆβ)=√∑n\\ni=1(yi−xiˆβ)2\\n(n−1)∑n\\ni′=1x2\\ni′.\\n(These formulas are slightly different from those given in Sec-\\ntions3.1.1and3.1.2, since here we are performing regression\\nwithout an intercept.) Show algebraically, and confirm numeri-\\ncally in R, that the t-statistic can be written as\\n(√n−1)∑n\\ni=1xiyi√\\n(∑n\\ni=1x2\\ni)(∑n\\ni′=1y2\\ni′)−(∑n\\ni′=1xi′yi′)2.\\n(e) Using the results from (d), argue that the t-statistic for the re-\\ngressionof yontoxisthesameasthe t-statisticfortheregression\\nofxontoy.\\n(f) InR, show that when regression is performed withan intercept,\\nthet-statistic for H0:β1=0is the same for the regression of y\\nontoxas it is for the regression of xontoy.\\n12. This problem involves simple linear regression without an intercept.\\n(a) Recall that the coefficient estimate ˆβfor the linear regression of\\nYontoXwithout an intercept is given by ( 3.38). Under what\\ncircumstance is the coefficient estimate for the regression of X\\nontoYthe same as the coefficient estimate for the regression of\\nYontoX?\\n(b) Generate an example in Pythonwithn= 100 observations in\\nwhich the coefficient estimate for the regression of XontoY\\nisdifferent from the coefficient estimate for the regression of Y\\nontoX.\\n(c) Generate an example in Pythonwithn= 100 observations in\\nwhich the coefficient estimate for the regression of XontoYis\\nthe same as the coefficient estimate for the regression of Yonto\\nX.\\n13. In this exercise you will create some simulated data and will fit simple\\nlinear regression models to it. Make sure to use the default random\\nnumber generator with seed set to 1 prior to starting part (a) to\\nensure consistent results.', metadata={'source': './documents/ISLP_website.pdf', 'page': 139}), Document(page_content='132 3. Linear Regression\\n(a) Using the normal() method of your random number generator,\\ncreate a vector, x, containing 100 observations drawn from a\\nN(0,1)distribution. This represents a feature, X.\\n(b) Using the normal() method, create a vector, eps, containing 100\\nobservations drawn from a N(0,0.25)distribution—a normal\\ndistribution with mean zero and variance 0.25.\\n(c) Using xandeps, generate a vector yaccording to the model\\nY=−1+0 .5X+ϵ. (3.39)\\nWhat is the length of the vector y? What are the values of β0\\nandβ1in this linear model?\\n(d) Create a scatterplot displaying the relationship between xand\\ny. Comment on what you observe.\\n(e) Fit a least squares linear model to predict yusingx. Comment\\non the model obtained. How do ˆβ0andˆβ1compare to β0and\\nβ1?\\n(f) Display the least squares line on the scatterplot obtained in (d).\\nDraw the population regression line on the plot, in a different\\ncolor. Use the legend() method of the axes to create an appro-\\npriate legend.\\n(g) Now fit a polynomial regression model that predicts yusingx\\nandx2. Is there evidence that the quadratic term improves the\\nmodel fit? Explain your answer.\\n(h) Repeat (a)–(f) after modifying the data generation process in\\nsuch a way that there is lessnoise in the data. The model ( 3.39)\\nshould remain the same. You can do this by decreasing the vari-\\nance of the normal distribution used to generate the error term\\nϵin (b). Describe your results.\\n(i) Repeat (a)–(f) after modifying the data generation process in\\nsuch a way that there is morenoise in the data. The model\\n(3.39) should remain the same. You can do this by increasing\\nthe variance of the normal distribution used to generate the\\nerror term ϵin (b). Describe your results.\\n(j) What are the confidence intervals for β0andβ1based on the\\noriginal data set, the noisier data set, and the less noisy data\\nset? Comment on your results.\\n14. This problem focuses on the collinearity problem.\\n(a) Perform the following commands in Python:\\nrng = np.random.default_rng(10)\\nx1 = rng.uniform(0, 1, size=100)\\nx2 = 0.5 * x1 + rng.normal(size=100) / 10\\ny=2+2*x 1+0 . 3*x 2+r n g . n o r m a l ( s i z e = 1 0 0 )\\nThe last line corresponds to creating a linear model in which yis\\na function of x1andx2. Write out the form of the linear model.\\nWhat are the regression coefficients?', metadata={'source': './documents/ISLP_website.pdf', 'page': 140}), Document(page_content='3.7 Exercises 133(b) What is the correlation betweenx1andx2? Create a scatterplotdisplaying the relationship between the variables.(c) Using this data, fit a least squares regression to predictyusingx1andx2. Describe the results obtained. What areˆβ0,ˆβ1, andˆβ2? How do these relate to the trueβ0,β1, andβ2? Can youreject the null hypothesisH0:β1=0? How about the nullhypothesisH0:β2=0?(d) Now fit a least squares regression to predictyusing onlyx1.Comment on your results. Can you reject the null hypothesisH0:β1=0?(e) Now fit a least squares regression to predictyusing onlyx2.Comment on your results. Can you reject the null hypothesisH0:β1=0?(f) Dotheresultsobtainedin(c)–(e)contradicteachother?Explainyour answer.(g) Suppose we obtain one additional observation, which was unfor-tunately mismeasured. We use the functionnp.concatenate()tonp.conca-tenate()add this additional observation to each ofx1,x2andy.x1 = np.concatenate([x1, [0.1]])x2 = np.concatenate([x2, [0.8]])y=n p . c o n c a t e n a t e ( [ y ,[ 6 ] ] )Re-fitthelinearmodelsfrom(c)to(e)usingthisnewdata.Whateffect does this new observation have on the each of the models?In each model, is this observation an outlier? A high-leveragepoint? Both? Explain your answers.15. This problem involves theBostondata set, which we saw in the labfor this chapter. We will now try to predict per capita crime rateusing the other variables in this data set. In other words, per capitacrime rate is the response, and the other variables are the predictors.(a) For each predictor, fit a simple linear regression model to predictthe response. Describe your results. In which of the models isthere a statistically significant association between the predictorand the response? Create some plots to back up your assertions.(b) Fit a multiple regression model to predict the response usingall of the predictors. Describe your results. For which predictorscan we reject the null hypothesisH0:βj=0?(c) How do your results from (a) compare to your results from (b)?Create a plot displaying the univariate regression coefficientsfrom (a) on thex-axis, and the multiple regression coefficientsfrom (b) on they-axis. That is, each predictor is displayed as asingle point in the plot. Its coefficient in a simple linear regres-sion model is shown on thex-axis, and its coefficient estimatein the multiple linear regression model is shown on they-axis.', metadata={'source': './documents/ISLP_website.pdf', 'page': 141}), Document(page_content='134 3. Linear Regression(d) Is there evidence of non-linear association between any of thepredictors and the response? To answer this question, for eachpredictorX, fit a model of the formY=β0+β1X+β2X2+β3X3+ϵ.', metadata={'source': './documents/ISLP_website.pdf', 'page': 142}), Document(page_content='4\\nClassification\\nThe linear regression model discussed in Chapter 3assumes that the re-\\nsponse variable Yis quantitative. But in many situations, the response\\nvariable is instead qualitative . For example, eye color is qualitative. Of-qualitativeten qualitative variables are referred to as categorical ; we will use these\\nterms interchangeably. In this chapter, we study approaches for predicting\\nqualitative responses, a process that is known as classification . Predictingclassificationa qualitative response for an observation can be referred to as classifying\\nthat observation, since it involves assigning the observation to a category,\\nor class. On the other hand, often the methods used for classification first\\npredict the probability that the observation belongs to each of the cate-\\ngories of a qualitative variable, as the basis for making the classification.\\nIn this sense they also behave like regression methods.\\nThere are many possible classification techniques, or classifiers , that oneclassifiermight use to predict a qualitative response. We touched on some of these\\nin Sections 2.1.5and2.2.3. In this chapter we discuss some widely-used\\nclassifiers: logistic regression ,linear discriminant analysis ,quadratic dis-logistic\\nregression\\nlinear\\ndiscriminant\\nanalysiscriminant analysis ,naive Bayes , andK-nearest neighbors . The discussion\\nquadratic\\ndiscriminant\\nanalysis\\nnaive Bayes\\nK-nearest\\nneighborsof logistic regression is used as a jumping-off point for a discussion of gen-\\neralized linear models , and in particular, Poisson regression . We discuss\\ngeneralized\\nlinear\\nmodels\\nPoisson\\nregressionmore computer-intensive classification methods in later chapters: these in-\\nclude generalized additive models (Chapter 7); trees, random forests, and\\nboosting (Chapter 8); and support vector machines (Chapter 9).\\n4.1 An Overview of Classification\\nClassification problems occur often, perhaps even more so than regression\\nproblems. Some examples include:\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning , Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_4  135', metadata={'source': './documents/ISLP_website.pdf', 'page': 143}), Document(page_content='136 4. Classification\\n1. A person arrives at the emergency room with a set of symptoms\\nthat could possibly be attributed to one of three medical conditions.\\nWhich of the three conditions does the individual have?\\n2. An online banking service must be able to determine whether or not\\na transaction being performed on the site is fraudulent, on the basis\\nof the user’s IP address, past transaction history, and so forth.\\n3. On the basis of DNA sequence data for a number of patients with\\nand without a given disease, a biologist would like to figure out which\\nDNA mutations are deleterious (disease-causing) and which are not.\\nJust as in the regression setting, in the classification setting we have a\\nset of training observations (x1,y1),...,(xn,yn)that we can use to build\\na classifier. We want our classifier to perform well not only on the training\\ndata, but also on test observations that were not used to train the classifier.\\nIn this chapter, we will illustrate the concept of classification using the\\nsimulated Default data set. We are interested in predicting whether an\\nindividual will default on his or her credit card payment, on the basis of\\nannual income and monthly credit card balance. The data set is displayed\\nin Figure 4.1. In the left-hand panel of Figure 4.1, we have plotted annual\\nincomeand monthly credit card balancefor a subset of 10,000individuals.\\nThe individuals who defaulted in a given month are shown in orange, and\\nthose who did not in blue. (The overall default rate is about 3%, so we\\nhave plotted only a fraction of the individuals who did not default.) It\\nappears that individuals who defaulted tended to have higher credit card\\nbalances than those who did not. In the center and right-hand panels of\\nFigure4.1, two pairs of boxplots are shown. The first shows the distribution\\nofbalancesplit by the binary defaultvariable; the second is a similar plot\\nforincome.Inthischapter,welearnhowtobuildamodeltopredict default\\n(Y) for any given value of balance(X1) andincome(X2). SinceYis not\\nquantitative, the simple linear regression model of Chapter 3is not a good\\nchoice: we will elaborate on this further in Section 4.2.\\nIt is worth noting that Figure 4.1displays a very pronounced relation-\\nship between the predictor balanceand the response default. In most real\\napplications, the relationship between the predictor and the response will\\nnot be nearly so strong. However, for the sake of illustrating the classifica-\\ntion procedures discussed in this chapter, we use an example in which the\\nrelationship between the predictor and the response is somewhat exagger-\\nated.\\n4.2 Why Not Linear Regression?\\nWe have stated that linear regression is not appropriate in the case of a\\nqualitative response. Why not?\\nSuppose that we are trying to predict the medical condition of a patient\\nin the emergency room on the basis of her symptoms. In this simplified\\nexample, there are three possible diagnoses: stroke,drug overdose , and', metadata={'source': './documents/ISLP_website.pdf', 'page': 144}), Document(page_content='4.2 Why Not Linear Regression? 137\\n0 500 1000 1500 2000 25000 20000 40000 60000BalanceIncome\\nNo Yes0 500 1000 1500 2000 2500DefaultBalance\\nNo Yes020000 40000 60000DefaultIncome\\nFIGURE 4.1.TheDefaultdata set.Left:The annual incomes and monthlycredit card balances of a number of individuals. The individuals who defaulted ontheir credit card payments are shown in orange, and those who did not are shownin blue.Center:Boxplots ofbalanceas a function ofdefaultstatus.Right:Boxplots ofincomeas a function ofdefaultstatus.epileptic seizure. We could consider encoding these values as a quantita-tive response variable,Y, as follows:Y=\\uf8f1\\uf8f4\\uf8f2\\uf8f4\\uf8f31ifstroke;2ifdrug overdose;3ifepileptic seizure.Usingthiscoding,leastsquarescouldbeusedtofitalinearregressionmodelto predictYon the basis of a set of predictorsX1,...,Xp. Unfortunately,this coding implies an ordering on the outcomes, puttingdrug overdoseinbetweenstrokeandepileptic seizure, and insisting that the differencebetweenstrokeanddrug overdoseis the same as the difference betweendrug overdoseandepileptic seizure. In practice there is no particularreason that this needs to be the case. For instance, one could choose anequally reasonable coding,Y=\\uf8f1\\uf8f4\\uf8f2\\uf8f4\\uf8f31ifepileptic seizure;2ifstroke;3ifdrug overdose,which would imply a totally different relationship among the three condi-tions. Each of these codings would produce fundamentally different linearmodels that would ultimately lead to different sets of predictions on testobservations.If the response variable’s values did take on a natural ordering, such asmild,moderate, andsevere, and we felt the gap between mild and moderatewas similar to the gap between moderate and severe, then a 1, 2, 3 codingwould be reasonable. Unfortunately, in general there is no natural way to', metadata={'source': './documents/ISLP_website.pdf', 'page': 145}), Document(page_content='138 4. Classification\\nconvert a qualitative response variable with more than two levels into a\\nquantitative response that is ready for linear regression.\\nFor abinary(two level) qualitative response, the situation is better. Forbinaryinstance, perhaps there are only two possibilities for the patient’s medical\\ncondition: strokeanddrug overdose . We could then potentially use the\\ndummyvariable approachfromSection 3.3.1tocodetheresponseasfollows:\\nY={\\n0ifstroke;\\n1ifdrug overdose .\\nWe could then fit a linear regression to this binary response, and predict\\ndrug overdose ifˆY>0.5andstrokeotherwise. In the binary case it is not\\nhard to show that even if we flip the above coding, linear regression will\\nproduce the same final predictions.\\nFor a binary response with a 0/1 coding as above, regression by least\\nsquares is not completely unreasonable: it can be shown that the Xˆβob-\\ntained using linear regression is in fact an estimate of Pr(drug overdose |X)\\nin this special case. However, if we use linear regression, some of our es-\\ntimates might be outside the [0,1]interval (see Figure 4.2), making them\\nhard to interpret as probabilities! Nevertheless, the predictions provide an\\nordering and can be interpreted as crude probability estimates. Curiously,\\nit turns out that the classifications that we get if we use linear regression\\nto predict a binary response will be the same as for the linear discriminant\\nanalysis (LDA) procedure we discuss in Section 4.4.\\nTo summarize, there are at least two reasons not to perform classifica-\\ntion using a regression method: (a) a regression method cannot accommo-\\ndate a qualitative response with more than two classes; (b) a regression\\nmethod will not provide meaningful estimates of Pr(Y|X), even with just\\ntwo classes. Thus, it is preferable to use a classification method that is\\ntruly suited for qualitative response values. In the next section, we present\\nlogistic regression, which is well-suited for the case of a binary qualita-\\ntive response; in later sections we will cover classification methods that are\\nappropriate when the qualitative response has two or more classes.\\n4.3 Logistic Regression\\nConsider again the Defaultdata set, where the response defaultfalls into\\none of two categories, YesorNo. Rather than modeling this response Y\\ndirectly, logistic regression models the probability thatYbelongs to a par-\\nticular category.\\nFortheDefaultdata,logisticregressionmodelstheprobabilityofdefault.\\nFor example, the probability of default given balancecan be written as\\nPr(default=Yes|balance).\\nThevaluesof Pr(default=Yes|balance),whichweabbreviate p(balance),\\nwillrangebetween0and1.Thenforanygivenvalueof balance,aprediction\\ncan be made for default. For example, one might predict default =Yes', metadata={'source': './documents/ISLP_website.pdf', 'page': 146}), Document(page_content='4.3 Logistic Regression 139\\n0 500 1000 1500 2000 25000.0 0.2 0.4 0.6 0.8 1.0BalanceProbability of Default|| |||| | |||||||||||||||||| | |||||| |||||||| | |||||||||||||||| | ||| ||||||||| |||||| | |||||| | | ||||||| |||||||||||||||| | |||||||||||||||||||||||| ||||\\n|||||| ||||| |||| ||||||||||||||||||| | ||\\n||||||||||||||||||||||||||||\\n|||||\\n|||\\n||||||||||||||||||||||||||||||||\\n||\\n||||| ||||| ||||||||||\\n||||||| | | | |||||||||||||||||||| ||||||||| | | | |||||||||||| | | |||||||||||||||||||| ||\\n||||\\n|| ||\\n||||||||\\n||||||||||||||||| ||||||||||||||||||||||||||||||||\\n|| ||||||||||| | | |||||||||||||||||||\\n|||||||||||| ||||||||||||||||| | |||||||||||||||||\\n||||||||||||||||||||| ||||||||||| ||||||||||||||||| | | ||\\n|||||\\n|||||| ||||| ||| || |||||||||||||||\\n|||||\\n|||||||||||||||||||||||||||||||| |||| | ||||||||||||||||||||| ||\\n||||||||||\\n|| | |||||||| | | |||||||| | ||||||||||||||||||||||||| | | | | | ||||||||||\\n||||||||||||||||| ||||||||||\\n|||||| ||| ||||||||| |||\\n|| | ||| ||||||||| |||||||||||||| ||| ||||||||||\\n||||| |||||||||||| ||||| ||||||||\\n||||||||| |||||||||| ||||||||||||| ||\\n||||| |||| ||||||||||||||||||||||||||||||||||||||||||||\\n||||||||||||\\n||| | ||||||||||||||||| ||| | | ||||||||||||||||\\n||| ||||\\n|| ||||||||||||| |||\\n|||||||||||||||||||\\n||| ||\\n||||||||||| | ||||||||\\n|| ||||||||| | |||||||||||||||||||||||||||||||||||||||||||\\n||||||||| | | ||||||||||| ||||||||||||||||\\n|||||||\\n|||||||||| |||||||\\n||||||| |||||||||||||||| ||| | | |||| |||||||| |||||| |||\\n||||||\\n|| ||| | ||||||||||| ||||||||| ||||||||||||||\\n|| |||||||||| ||||||| | ||||||||||||||||| ||||||||||||||||||||||||||| | |||||||||||||| | ||||||||||| | ||||| ||| ||||\\n||\\n||||||| | ||||||||||||||||| |||||||||\\n||||||||| |||||| |||||||||||||||||| |||||||||||||| | ||\\n||\\n|||||||| | | | | | | | | |||||||||||||||| |||||\\n|||\\n|||||||||\\n||||||\\n|||||||||||||| | |||||| | |||||| |||| ||||||||| | | |||\\n|||||||||||||||||||| | ||||||| ||||||||||||||\\n||| ||||||||||||||||\\n||||||||||||||||\\n||||||||||||||||||||||\\n||||||||||||||||||||| | | ||| |||| ||||| |||||||||| ||||||||||\\n|||||||\\n||||| |||||| | |||||||||||||||||||||||||||\\n||||||||||||||||||||||||||||||||\\n|||||||||||||| |||| ||||||||||| |||||||| | | |||||||||\\n||||\\n|||||||| ||||\\n|||| |||||||||| | |||\\n|||||| | |||||||\\n|| | ||||||| ||| | | |||||\\n|||||||| | |||| ||||| | |||| |||||||||||||||| | |||\\n||\\n|| ||||||| ||||||||||||||||||||||||||||||||| ||||||||| |||||| ||||\\n||||| |||\\n||||||||||||||||| ||||||\\n||||||||||||||||\\n||||||| | | ||||||\\n|||||||||||||||||||||||||||\\n|||||\\n||||||||||||||||| |||||||||||||\\n|||||||| |||||||| | | |||| | |||||||| ||\\n|| |||| ||||||||\\n|| ||| ||\\n|||| ||||| ||||||||||||\\n||||\\n|||| | |||||||\\n|||||||||||||||||||||||\\n|||||||| |||||| |||| |||||||||| ||||| ||| | | |||||\\n|| ||| ||||||||||||| | | | | | |||| |||||||||||||\\n||| ||||||||| | ||||\\n||||||||||||||\\n|||| |||||||||||||||||||||||||||| |||||||||||||||||||||||||||||||||||||||| ||||||||||||||||||||||\\n||||||||||||\\n|| ||||| | |||| | |||| ||||||||||||||||||||||||||| |||||||||||||||||||||||||| ||||||||\\n||||| |||||||\\n|| ||||||||||||||| | | | ||||||||||||||| |||||||||||||||||||\\n||||||||||||||||||||| |||||||||||||||||||| ||||||||\\n|||||||||||||||| | ||||||||||||||| | ||||||||\\n||| ||| |||||||||||||||||||||||||||||| | |||||| |||| |||| | ||| |||\\n|||||||||||| |||||||| ||||||||||||\\n||||| | | | ||||||||\\n|||||||||||||||||||||||||| |||||||||||||||||||||||| |||| | | | |||||||||| | |||||||||| |||||||||||\\n|| |||||||||||||||||||||||||||||| | | ||||||\\n|||| ||||||||||||||||||||||||||||||||| ||||||| ||||| |||||||||| |||| | | | ||||||||||||||||||||||| | | ||||||||||||||||||||\\n||||| | | | |||||\\n|||||| ||||||||||| ||||||||||||||||| | ||||||| | |||||||\\n|||||| | |||||||\\n||||||\\n|||||| | |||||| | ||||||||| ||| | | | ||||\\n|||| ||\\n|||||||||||||||||||\\n|| |||| ||\\n|||| |||| ||||||||||| | ||\\n|||| ||||| |||||||||||| | |||||\\n||||||| |||\\n||||||||||||||| | |||||| | ||||||||||||||\\n||||||||||||\\n|||||||||\\n||||||||\\n|||||||||\\n|||||||| ||||| | ||||||||||||||||||||||||||||| |||| ||||||\\n|||\\n||||||\\n||||||\\n|||||||||||||||| | ||||||||| ||||||||||||||||||||\\n||||||||||||||||||||| ||||||||||||||||||||||||||||\\n|| ||\\n||||| |||||||||||||||||||| ||||| | ||||||||||| |||||||||||| ||||||||||||||||||\\n|||||| | |||||| | | |||||||||| |||||| ||||\\n||||||||| ||| |||\\n|||||||||||| | ||||||| | ||||||||\\n|||||||| | |||||||||||||||| ||||| ||||||||||||||||| | | ||||||||||||||||||| | | |||\\n||||\\n|||| || ||||||||||||||||||||||||| ||||||||| | |||||||| |||||||||||||||||\\n|||| |||||||| | |||||||||||||||||||||||||||||||||| |||| ||||||||| |||||||||||\\n||||||||||||||||||||||||||\\n||||| ||| | |||||||| |||||||||||||||\\n||| ||| ||\\n|| ||||||||||||| ||||| ||||||||||||||||\\n||||||||||||\\n|| |||||||||||||||||||||||||||||||||| || |||||||| ||||||||||||||||||||| |||||||||||||\\n|||||||| | ||\\n|||| | ||||||||\\n||||\\n||\\n||||||||||||||||||| |||||||||||||\\n|||||||||||||||||||||||||||||| | ||\\n||||||||||| ||||\\n||||||||\\n|||||||||||||||||||||||||||||||||||||||||||\\n||||||||||| |||\\n|||||||\\n||||| ||||||||||||| |||| |||||||||||| | |||||||||| | |||||| ||||||||||||| ||| | | |||||||\\n||||||||||||| ||| | |||||||||||||||||||||||||| |||||||||||||| | |||||\\n|||||| | ||||||| ||||||| ||||||||||||||||\\n|||||||||\\n|||||| ||| | |||\\n||||| |||||||||||||||||||||||| ||||\\n||||||||| |||||\\n|||||||||||||||| ||| | ||||| |||||||\\n||||\\n|||||||||||||||||||| |||||||| | | ||||||||||| | |||||||||||||\\n||||||||||||| |||||||||\\n||||\\n|| |||||||||||||||||| |||||||||||| ||| ||||||||||||||||| |||||||||||| || |||| |||| ||||| ||||||||\\n||\\n||||||||||||||\\n||||||||||||\\n|| |||||||||| |||||\\n|| |||||||||||||||||||||||||| | |||||||||| |||||| | ||\\n|||||||||||||||| | | | |||||||||||||||||||||| ||||||||||| ||||||||||||||||| | |||||||||||||| ||||||||||||||||||||||||||||||||||||||||||||||||| |||| | |||| | | ||||||||||||| ||||||\\n|||| | | | ||||\\n||||\\n||| |||||||||\\n||| |||||||||||||||\\n||||||||||||||||||||||||||||||||\\n|||||||||||||||| | | |||| ||\\n|||||||||||\\n||||\\n||||||||||||||||||||||||||||||||||| |||\\n|||||||||||||||||||||||||||| |||||||||||||||||||||||||||||||||||||| ||||\\n|||||| ||||| | ||||||||||| | ||||\\n|||||||||||||| ||||||\\n||||| ||| ||||||||||||| |||||||||||||||||| |||||||||||||||||||||||||||||| | ||||||\\n|||||||| |||||||| | | | | ||||||||| |||||||||||||||||||||||||| | | | |||||||| |||\\n|| |||| |||| ||||| |||| | ||||| ||||||||||| |||| ||\\n|| |||||||||||| ||||||| | ||||||||||| ||\\n||||||||\\n||||||||||||||||||||||\\n||||||||||||||||||||||| ||||||||||| ||||| ||||||\\n||||| | |||||| | | | |||| | | | | | | |||||||||| | | ||| | | | ||||||||||||||| ||||| ||| ||\\n|| | ||| ||| ||||||| |||||||||||||||||||||||||||||||||||||| | ||||| | | ||||||| | ||||||||\\n||||||| ||| |||\\n|||||||||\\n||||||||||||||||||||| ||||||||| | ||| ||||||| | ||||| | |||| ||| ||| ||||||||||| | |||||||||| |||||||||||||||| ||||||||||||\\n|| ||| |||||||||||||||||||||||||||||||||\\n||||||| |||||||||||||||| |||| ||||||||||\\n|||||||||||||||||||||||| |||||||||\\n||||||||||||| |||||||||| ||||| |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\\n|| ||||||||| |||||| ||||||| | ||| |||| | |||||\\n||||||| ||||||\\n|||| |||||||| |||||||||||| | ||||||||||||||||||\\n||\\n||||||||||||||||| | ||\\n|| | ||||||||| ||||||| |||||\\n||||||||||||||||\\n|||||||||||||||||||| ||||||||||||||| ||||| |||\\n|||||||||||\\n||| |||| |||| |||| ||||||||||||||||| ||||||||| ||||||||||\\n|||||||\\n||||||||||||\\n|||||||||||| |||||||| |||||||||||||||||||||| | |||||| ||| | | |||||||||| |||||||||\\n|| | | |||| ||||||||||| | | ||||| | ||||||||||\\n||||||| |||||||||||||||||||||\\n|||| ||||||||||||||||||\\n||||||||||| ||||||| ||||||||||\\n||||| |||||||\\n|||||||||||||\\n||||||||||||||||||||||||||||||||||||||||||||||\\n|||||||| |||||||||||||\\n||\\n||| | | | | |||||| ||||||||| |||| |||||||||||| | ||||||| |||||||\\n|| |||\\n|||||||||||\\n|||||||||||||| | |||||||\\n|||||||||\\n|| ||| | |||||||||||||||||||||||||||||||||||||\\n|||||||||||| |||||||| ||| |||||||||||||| | |||||||||||||||||||||||||||||||\\n|||||||||||||||||||||||||||\\n||||||||| |||\\n|||||||||\\n||||||||| | ||||| ||||||||| |||\\n||||||||| | | | |||||\\n|||\\n||||||||||| |||| |||||||\\n||||||||||\\n|||||| |||||||||||||| | ||||||||||||||||||||\\n|||| ||||||| | | ||||||||||| |||||||||||||||||| ||||\\n||| ||||||||| ||||\\n||||| |||||| ||||||||| | ||||||||\\n||||| | | |||||||||||||||||||||||||\\n|||||||||||||| | |||||||||||||||||||||| | | ||| | ||||\\n|||||||||||\\n||||||||||| ||||||| ||||||||||||||||||\\n||||||||||||||||||||||||||||| ||| |||||||||||||||||||||||||||\\n|||||||||||||||||| ||||||||||||| | |||||||||||||||\\n|||||||| |||| ||||| ||||||||||||||||||||||||||||||||||||||||||||| |||||||||||| ||||||||||||| ||\\n||\\n|||||| |||||||||||||||| |||\\n|||||||||||||||| ||||||||||||||||||||||||||| ||||| ||||| ||||||| ||| |||\\n|||||||||\\n|| ||||||| | | ||||| ||||||||||||||||||||||||||| ||||||\\n|||\\n||||||| | |||||||||||||||||| |||||||||\\n||||||||||||| | | |||||||||||||\\n|||||||||||||| | |||| | | |||||||| |||||||||||||||||||||||||||||||||||||||| ||||||| | |||||||||||||| ||||||||||||||||||\\n|||||||||||||||||||| | | |||||| |||| |||||| |||||||||||| | | ||| ||| | | | ||| ||||||\\n|||||||| |||||||||| | ||||||||||||||||| | ||||||||||||||||\\n||||||||||||||\\n|||||||||||||\\n||\\n||\\n||||||||||||\\n||||||||||||||||||||||||||||||||||| |||||||\\n||||||||| || |||||||||||||||||||||||\\n|||||||||||||||||||||||||| ||||||||||| | ||||||||||| |||||| ||||||||||||||||||||||| ||||||||||||||||||||||||||||||||||| |||||||||||||||||||| |||| ||| | ||||| | | ||||||||||||||||||||||||||||||| |||||||||||||||||| ||||||||| ||| | | ||||||\\n||||||||||||||||||||\\n||| ||||||||| | ||| |||| ||| | | | |||||||||| ||| |||||||||||\\n||||\\n|||||||||||||||||||||||||||||||||||||||||||||||||\\n||||||| ||| ||||| ||||||\\n||||||||||||||||||||||||| ||||||||||||||||||||\\n||||||||||||||||||| | ||| |||||| ||||\\n|||||||||||| ||\\n||||||||\\n||||||| ||||||||||||||||||||||||||||\\n|||||||||||\\n|| |||| |||||||||||\\n||||| |||| ||||||| | ||||||||||||\\n|||\\n||||\\n|||||| ||||||||||\\n||| |||||||\\n|| | |||\\n||||||||||||||||||||||| ||||||||||||||||||||||||\\n|||||||||||||||||||||||||| ||| | ||\\n|||| | | | | ||||||||||||||||||| | | | |||||||||||| ||||| |||| | | | ||||||||| | |||||||||||||||||||||||||| ||||| |||||||||||||||| | ||| ||||||||||||||||\\n|||| |||||||||||||||\\n|||||||||||||\\n|||||||||||||||||||||||||||||| |||||||||||||||| ||||||| ||\\n||||||| ||||||| ||||||||||||| |||||||||||||\\n|||\\n||||||||||||||||||||||| | ||| | |||||||||||||||||||||||||||||| | |||||||||||||\\n||||||||||||\\n|||||||||||||||||| | |||||||||||||||||||||| | |||||||||| |||||||||||||||||||||\\n|||||||||||||||||| | |||| |||||||||||||||||||||||||\\n||| ||\\n||||||||| ||||||||||||||||\\n|||||||||\\n|||||\\n|| ||||| |||||||||||||||||| |||||||||||||||||||||||||||||||||||| ||||||||||||||| ||||||||||||||| ||||\\n|||||||||||||||||||||||||||||||| | | | ||| ||||||||||||||||||| ||||||||||||||||||||\\n||| | ||||||||| ||\\n||||||||||||||||||||| |||\\n||||| ||||||||||||||||||| | | | ||||||\\n|||||||||||||||| ||||||||||||||||||||||| ||\\n|||||||||| ||||||||||| ||||||||||||||||||||||||||||||||||||||||||\\n|| | ||\\n|||| |||||||\\n||||||||||\\n|||||||||||||||||||\\n||||||||\\n|||||||| | | |||||\\n|| ||||||||\\n|| | ||\\n|||||||\\n||||||||| |||||||\\n||||||||||||||||||||||| ||||||||||||||||||| | ||||||||| ||||||||||||||||||||\\n|||||| | ||||||||||||||||| |||| |||||||||| | | | |||\\n|||||| ||| ||| | ||||||||| | ||||||||| | ||||||||||||||||||||||| ||||||||||| | |||||||||||| | |||||||||| |||||||||||||||||| ||||||||||||||||| |||\\n|||\\n||||||||||\\n||||| ||||||||||||\\n|||||||||||||||||||||||||||||| |||||||||||||\\n|||||||||||||||||||||| |||||\\n||| |||||||\\n||||||||| | | | |||| | ||\\n|||||| |||\\n||||||| ||| | ||||||||| | ||||| ||\\n||\\n||| |||||||||||||| ||||| | ||||\\n||| | |||||||||||||| |||0 500 1000 1500 2000 25000.0 0.2 0.4 0.6 0.8 1.0BalanceProbability of Default|| |||| | |||||||||||||||||| | |||||| |||||||| | |||||||||||||||| | ||| ||||||||| |||||| | |||||| | | ||||||| |||||||||||||||| | |||||||||||||||||||||||| ||||\\n|||||| ||||| |||| ||||||||||||||||||| | ||\\n||||||||||||||||||||||||||||\\n|||||\\n|||\\n||||||||||||||||||||||||||||||||\\n||\\n||||| ||||| ||||||||||\\n||||||| | | | |||||||||||||||||||| ||||||||| | | | |||||||||||| | | |||||||||||||||||||| ||\\n||||\\n|| ||\\n||||||||\\n||||||||||||||||| ||||||||||||||||||||||||||||||||\\n|| ||||||||||| | | |||||||||||||||||||\\n|||||||||||| ||||||||||||||||| | |||||||||||||||||\\n||||||||||||||||||||| ||||||||||| ||||||||||||||||| | | ||\\n|||||\\n|||||| ||||| ||| || |||||||||||||||\\n|||||\\n|||||||||||||||||||||||||||||||| |||| | ||||||||||||||||||||| ||\\n||||||||||\\n|| | |||||||| | | |||||||| | ||||||||||||||||||||||||| | | | | | ||||||||||\\n||||||||||||||||| ||||||||||\\n|||||| ||| ||||||||| |||\\n|| | ||| ||||||||| |||||||||||||| ||| ||||||||||\\n||||| |||||||||||| ||||| ||||||||\\n||||||||| |||||||||| ||||||||||||| ||\\n||||| |||| ||||||||||||||||||||||||||||||||||||||||||||\\n||||||||||||\\n||| | ||||||||||||||||| ||| | | ||||||||||||||||\\n||| ||||\\n|| ||||||||||||| |||\\n|||||||||||||||||||\\n||| ||\\n||||||||||| | ||||||||\\n|| ||||||||| | |||||||||||||||||||||||||||||||||||||||||||\\n||||||||| | | ||||||||||| ||||||||||||||||\\n|||||||\\n|||||||||| |||||||\\n||||||| |||||||||||||||| ||| | | |||| |||||||| |||||| |||\\n||||||\\n|| ||| | ||||||||||| ||||||||| ||||||||||||||\\n|| |||||||||| ||||||| | ||||||||||||||||| ||||||||||||||||||||||||||| | |||||||||||||| | ||||||||||| | ||||| ||| ||||\\n||\\n||||||| | ||||||||||||||||| |||||||||\\n||||||||| |||||| |||||||||||||||||| |||||||||||||| | ||\\n||\\n|||||||| | | | | | | | | |||||||||||||||| |||||\\n|||\\n|||||||||\\n||||||\\n|||||||||||||| | |||||| | |||||| |||| ||||||||| | | |||\\n|||||||||||||||||||| | ||||||| ||||||||||||||\\n||| ||||||||||||||||\\n||||||||||||||||\\n||||||||||||||||||||||\\n||||||||||||||||||||| | | ||| |||| ||||| |||||||||| ||||||||||\\n|||||||\\n||||| |||||| | |||||||||||||||||||||||||||\\n||||||||||||||||||||||||||||||||\\n|||||||||||||| |||| ||||||||||| |||||||| | | |||||||||\\n||||\\n|||||||| ||||\\n|||| |||||||||| | |||\\n|||||| | |||||||\\n|| | ||||||| ||| | | |||||\\n|||||||| | |||| ||||| | |||| |||||||||||||||| | |||\\n||\\n|| ||||||| ||||||||||||||||||||||||||||||||| ||||||||| |||||| ||||\\n||||| |||\\n||||||||||||||||| ||||||\\n||||||||||||||||\\n||||||| | | ||||||\\n|||||||||||||||||||||||||||\\n|||||\\n||||||||||||||||| |||||||||||||\\n|||||||| |||||||| | | |||| | |||||||| ||\\n|| |||| ||||||||\\n|| ||| ||\\n|||| ||||| ||||||||||||\\n||||\\n|||| | |||||||\\n|||||||||||||||||||||||\\n|||||||| |||||| |||| |||||||||| ||||| ||| | | |||||\\n|| ||| ||||||||||||| | | | | | |||| |||||||||||||\\n||| ||||||||| | ||||\\n||||||||||||||\\n|||| |||||||||||||||||||||||||||| |||||||||||||||||||||||||||||||||||||||| ||||||||||||||||||||||\\n||||||||||||\\n|| ||||| | |||| | |||| ||||||||||||||||||||||||||| |||||||||||||||||||||||||| ||||||||\\n||||| |||||||\\n|| ||||||||||||||| | | | ||||||||||||||| |||||||||||||||||||\\n||||||||||||||||||||| |||||||||||||||||||| ||||||||\\n|||||||||||||||| | ||||||||||||||| | ||||||||\\n||| ||| |||||||||||||||||||||||||||||| | |||||| |||| |||| | ||| |||\\n|||||||||||| |||||||| ||||||||||||\\n||||| | | | ||||||||\\n|||||||||||||||||||||||||| |||||||||||||||||||||||| |||| | | | |||||||||| | |||||||||| |||||||||||\\n|| |||||||||||||||||||||||||||||| | | ||||||\\n|||| ||||||||||||||||||||||||||||||||| ||||||| ||||| |||||||||| |||| | | | ||||||||||||||||||||||| | | ||||||||||||||||||||\\n||||| | | | |||||\\n|||||| ||||||||||| ||||||||||||||||| | ||||||| | |||||||\\n|||||| | |||||||\\n||||||\\n|||||| | |||||| | ||||||||| ||| | | | ||||\\n|||| ||\\n|||||||||||||||||||\\n|| |||| ||\\n|||| |||| ||||||||||| | ||\\n|||| ||||| |||||||||||| | |||||\\n||||||| |||\\n||||||||||||||| | |||||| | ||||||||||||||\\n||||||||||||\\n|||||||||\\n||||||||\\n|||||||||\\n|||||||| ||||| | ||||||||||||||||||||||||||||| |||| ||||||\\n|||\\n||||||\\n||||||\\n|||||||||||||||| | ||||||||| ||||||||||||||||||||\\n||||||||||||||||||||| ||||||||||||||||||||||||||||\\n|| ||\\n||||| |||||||||||||||||||| ||||| | ||||||||||| |||||||||||| ||||||||||||||||||\\n|||||| | |||||| | | |||||||||| |||||| ||||\\n||||||||| ||| |||\\n|||||||||||| | ||||||| | ||||||||\\n|||||||| | |||||||||||||||| ||||| ||||||||||||||||| | | ||||||||||||||||||| | | |||\\n||||\\n|||| || ||||||||||||||||||||||||| ||||||||| | |||||||| |||||||||||||||||\\n|||| |||||||| | |||||||||||||||||||||||||||||||||| |||| ||||||||| |||||||||||\\n||||||||||||||||||||||||||\\n||||| ||| | |||||||| |||||||||||||||\\n||| ||| ||\\n|| ||||||||||||| ||||| ||||||||||||||||\\n||||||||||||\\n|| |||||||||||||||||||||||||||||||||| || |||||||| ||||||||||||||||||||| |||||||||||||\\n|||||||| | ||\\n|||| | ||||||||\\n||||\\n||\\n||||||||||||||||||| |||||||||||||\\n|||||||||||||||||||||||||||||| | ||\\n||||||||||| ||||\\n||||||||\\n|||||||||||||||||||||||||||||||||||||||||||\\n||||||||||| |||\\n|||||||\\n||||| ||||||||||||| |||| |||||||||||| | |||||||||| | |||||| ||||||||||||| ||| | | |||||||\\n||||||||||||| ||| | |||||||||||||||||||||||||| |||||||||||||| | |||||\\n|||||| | ||||||| ||||||| ||||||||||||||||\\n|||||||||\\n|||||| ||| | |||\\n||||| |||||||||||||||||||||||| ||||\\n||||||||| |||||\\n|||||||||||||||| ||| | ||||| |||||||\\n||||\\n|||||||||||||||||||| |||||||| | | ||||||||||| | |||||||||||||\\n||||||||||||| |||||||||\\n||||\\n|| |||||||||||||||||| |||||||||||| ||| ||||||||||||||||| |||||||||||| || |||| |||| ||||| ||||||||\\n||\\n||||||||||||||\\n||||||||||||\\n|| |||||||||| |||||\\n|| |||||||||||||||||||||||||| | |||||||||| |||||| | ||\\n|||||||||||||||| | | | |||||||||||||||||||||| ||||||||||| ||||||||||||||||| | |||||||||||||| ||||||||||||||||||||||||||||||||||||||||||||||||| |||| | |||| | | ||||||||||||| ||||||\\n|||| | | | ||||\\n||||\\n||| |||||||||\\n||| |||||||||||||||\\n||||||||||||||||||||||||||||||||\\n|||||||||||||||| | | |||| ||\\n|||||||||||\\n||||\\n||||||||||||||||||||||||||||||||||| |||\\n|||||||||||||||||||||||||||| |||||||||||||||||||||||||||||||||||||| ||||\\n|||||| ||||| | ||||||||||| | ||||\\n|||||||||||||| ||||||\\n||||| ||| ||||||||||||| |||||||||||||||||| |||||||||||||||||||||||||||||| | ||||||\\n|||||||| |||||||| | | | | ||||||||| |||||||||||||||||||||||||| | | | |||||||| |||\\n|| |||| |||| ||||| |||| | ||||| ||||||||||| |||| ||\\n|| |||||||||||| ||||||| | ||||||||||| ||\\n||||||||\\n||||||||||||||||||||||\\n||||||||||||||||||||||| ||||||||||| ||||| ||||||\\n||||| | |||||| | | | |||| | | | | | | |||||||||| | | ||| | | | ||||||||||||||| ||||| ||| ||\\n|| | ||| ||| ||||||| |||||||||||||||||||||||||||||||||||||| | ||||| | | ||||||| | ||||||||\\n||||||| ||| |||\\n|||||||||\\n||||||||||||||||||||| ||||||||| | ||| ||||||| | ||||| | |||| ||| ||| ||||||||||| | |||||||||| |||||||||||||||| ||||||||||||\\n|| ||| |||||||||||||||||||||||||||||||||\\n||||||| |||||||||||||||| |||| ||||||||||\\n|||||||||||||||||||||||| |||||||||\\n||||||||||||| |||||||||| ||||| |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\\n|| ||||||||| |||||| ||||||| | ||| |||| | |||||\\n||||||| ||||||\\n|||| |||||||| |||||||||||| | ||||||||||||||||||\\n||\\n||||||||||||||||| | ||\\n|| | ||||||||| ||||||| |||||\\n||||||||||||||||\\n|||||||||||||||||||| ||||||||||||||| ||||| |||\\n|||||||||||\\n||| |||| |||| |||| ||||||||||||||||| ||||||||| ||||||||||\\n|||||||\\n||||||||||||\\n|||||||||||| |||||||| |||||||||||||||||||||| | |||||| ||| | | |||||||||| |||||||||\\n|| | | |||| ||||||||||| | | ||||| | ||||||||||\\n||||||| |||||||||||||||||||||\\n|||| ||||||||||||||||||\\n||||||||||| ||||||| ||||||||||\\n||||| |||||||\\n|||||||||||||\\n||||||||||||||||||||||||||||||||||||||||||||||\\n|||||||| |||||||||||||\\n||\\n||| | | | | |||||| ||||||||| |||| |||||||||||| | ||||||| |||||||\\n|| |||\\n|||||||||||\\n|||||||||||||| | |||||||\\n|||||||||\\n|| ||| | |||||||||||||||||||||||||||||||||||||\\n|||||||||||| |||||||| ||| |||||||||||||| | |||||||||||||||||||||||||||||||\\n|||||||||||||||||||||||||||\\n||||||||| |||\\n|||||||||\\n||||||||| | ||||| ||||||||| |||\\n||||||||| | | | |||||\\n|||\\n||||||||||| |||| |||||||\\n||||||||||\\n|||||| |||||||||||||| | ||||||||||||||||||||\\n|||| ||||||| | | ||||||||||| |||||||||||||||||| ||||\\n||| ||||||||| ||||\\n||||| |||||| ||||||||| | ||||||||\\n||||| | | |||||||||||||||||||||||||\\n|||||||||||||| | |||||||||||||||||||||| | | ||| | ||||\\n|||||||||||\\n||||||||||| ||||||| ||||||||||||||||||\\n||||||||||||||||||||||||||||| ||| |||||||||||||||||||||||||||\\n|||||||||||||||||| ||||||||||||| | |||||||||||||||\\n|||||||| |||| ||||| ||||||||||||||||||||||||||||||||||||||||||||| |||||||||||| ||||||||||||| ||\\n||\\n|||||| |||||||||||||||| |||\\n|||||||||||||||| ||||||||||||||||||||||||||| ||||| ||||| ||||||| ||| |||\\n|||||||||\\n|| ||||||| | | ||||| ||||||||||||||||||||||||||| ||||||\\n|||\\n||||||| | |||||||||||||||||| |||||||||\\n||||||||||||| | | |||||||||||||\\n|||||||||||||| | |||| | | |||||||| |||||||||||||||||||||||||||||||||||||||| ||||||| | |||||||||||||| ||||||||||||||||||\\n|||||||||||||||||||| | | |||||| |||| |||||| |||||||||||| | | ||| ||| | | | ||| ||||||\\n|||||||| |||||||||| | ||||||||||||||||| | ||||||||||||||||\\n||||||||||||||\\n|||||||||||||\\n||\\n||\\n||||||||||||\\n||||||||||||||||||||||||||||||||||| |||||||\\n||||||||| || |||||||||||||||||||||||\\n|||||||||||||||||||||||||| ||||||||||| | ||||||||||| |||||| ||||||||||||||||||||||| ||||||||||||||||||||||||||||||||||| |||||||||||||||||||| |||| ||| | ||||| | | ||||||||||||||||||||||||||||||| |||||||||||||||||| ||||||||| ||| | | ||||||\\n||||||||||||||||||||\\n||| ||||||||| | ||| |||| ||| | | | |||||||||| ||| |||||||||||\\n||||\\n|||||||||||||||||||||||||||||||||||||||||||||||||\\n||||||| ||| ||||| ||||||\\n||||||||||||||||||||||||| ||||||||||||||||||||\\n||||||||||||||||||| | ||| |||||| ||||\\n|||||||||||| ||\\n||||||||\\n||||||| ||||||||||||||||||||||||||||\\n|||||||||||\\n|| |||| |||||||||||\\n||||| |||| ||||||| | ||||||||||||\\n|||\\n||||\\n|||||| ||||||||||\\n||| |||||||\\n|| | |||\\n||||||||||||||||||||||| ||||||||||||||||||||||||\\n|||||||||||||||||||||||||| ||| | ||\\n|||| | | | | ||||||||||||||||||| | | | |||||||||||| ||||| |||| | | | ||||||||| | |||||||||||||||||||||||||| ||||| |||||||||||||||| | ||| ||||||||||||||||\\n|||| |||||||||||||||\\n|||||||||||||\\n|||||||||||||||||||||||||||||| |||||||||||||||| ||||||| ||\\n||||||| ||||||| ||||||||||||| |||||||||||||\\n|||\\n||||||||||||||||||||||| | ||| | |||||||||||||||||||||||||||||| | |||||||||||||\\n||||||||||||\\n|||||||||||||||||| | |||||||||||||||||||||| | |||||||||| |||||||||||||||||||||\\n|||||||||||||||||| | |||| |||||||||||||||||||||||||\\n||| ||\\n||||||||| ||||||||||||||||\\n|||||||||\\n|||||\\n|| ||||| |||||||||||||||||| |||||||||||||||||||||||||||||||||||| ||||||||||||||| ||||||||||||||| ||||\\n|||||||||||||||||||||||||||||||| | | | ||| ||||||||||||||||||| ||||||||||||||||||||\\n||| | ||||||||| ||\\n||||||||||||||||||||| |||\\n||||| ||||||||||||||||||| | | | ||||||\\n|||||||||||||||| ||||||||||||||||||||||| ||\\n|||||||||| ||||||||||| ||||||||||||||||||||||||||||||||||||||||||\\n|| | ||\\n|||| |||||||\\n||||||||||\\n|||||||||||||||||||\\n||||||||\\n|||||||| | | |||||\\n|| ||||||||\\n|| | ||\\n|||||||\\n||||||||| |||||||\\n||||||||||||||||||||||| ||||||||||||||||||| | ||||||||| ||||||||||||||||||||\\n|||||| | ||||||||||||||||| |||| |||||||||| | | | |||\\n|||||| ||| ||| | ||||||||| | ||||||||| | ||||||||||||||||||||||| ||||||||||| | |||||||||||| | |||||||||| |||||||||||||||||| ||||||||||||||||| |||\\n|||\\n||||||||||\\n||||| ||||||||||||\\n|||||||||||||||||||||||||||||| |||||||||||||\\n|||||||||||||||||||||| |||||\\n||| |||||||\\n||||||||| | | | |||| | ||\\n|||||| |||\\n||||||| ||| | ||||||||| | ||||| ||\\n||\\n||| |||||||||||||| ||||| | ||||\\n||| | |||||||||||||| |||FIGURE 4.2. Classification using the Default data.Left:Estimated probability\\nofdefault using linear regression. Some estimated probabilities are negative! The\\norangeticks indicatethe 0/1values codedfor default(NoorYes).Right:Predicted\\nprobabilities of default using logistic regression. All probabilities lie between 0\\nand1.\\nfor any individual for whom p(balance )>0.5. Alternatively, if a company\\nwishes to be conservative in predicting individuals who are at risk for de-\\nfault, then they may choose to use a lower threshold, such as p(balance )>\\n0.1.\\n4.3.1 The Logistic Model\\nHow should we model the relationship between p(X) = Pr( Y=1|X)and\\nX? (For convenience we are using the generic 0/1 coding for the response.)\\nIn Section 4.2we considered using a linear regression model to represent\\nthese probabilities:\\np(X)=β0+β1X. (4.1)\\nIf we use this approach to predict default=Yesusingbalance, then we\\nobtain the model shown in the left-hand panel of Figure 4.2. Here we see\\nthe problem with this approach: for balances close to zero we predict a\\nnegative probability of default; if we were to predict for very large balances,\\nwe would get values bigger than 1. These predictions are not sensible, since\\nof course the true probability of default, regardless of credit card balance,\\nmust fall between 0and1. This problem is not unique to the credit default\\ndata. Any time a straight line is fit to a binary response that is coded as\\n0or1, in principle we can always predict p(X)<0for some values of X\\nandp(X)>1for others (unless the range of Xis limited).\\nTo avoid this problem, we must model p(X)using a function that gives\\noutputs between 0and1for all values of X. Many functions meet this\\ndescription. In logistic regression, we use the logistic function ,logistic\\nfunction\\np(X)=eβ0+β1X\\n1+eβ0+β1X. (4.2)\\nTo fit the model ( 4.2), we use a method called maximum likelihood , whichmaximum\\nlikelihoodwediscussinthenextsection.Theright-handpanelofFigure 4.2illustrates\\nthe fit of the logistic regression model to the Defaultdata. Notice that for', metadata={'source': './documents/ISLP_website.pdf', 'page': 147}), Document(page_content='140 4. Classification\\nlow balances we now predict the probability of default as close to, but never\\nbelow, zero. Likewise, for high balances we predict a default probability\\nclose to, but never above, one. The logistic function will always produce\\nanS-shaped curve of this form, and so regardless of the value of X,w e\\nwill obtain a sensible prediction. We also see that the logistic model is\\nbetter able to capture the range of probabilities than is the linear regression\\nmodel in the left-hand plot. The average fitted probability in both cases is\\n0.0333 (averaged over the training data), which is the same as the overall\\nproportion of defaulters in the data set.\\nAfter a bit of manipulation of ( 4.2), we find that\\np(X)\\n1−p(X)=eβ0+β1X. (4.3)\\nThe quantity p(X)/[1−p(X)]is called the odds, and can take on any valueoddsbetween 0and∞. Values of the odds close to 0and∞indicate very low\\nand very high probabilities of default, respectively. For example, on average\\n1in5people with an odds of 1/4will default, since p(X)=0 .2implies an\\nodds of0.2\\n1−0.2=1/4. Likewise, on average nine out of every ten people with\\nan odds of 9will default, since p(X)=0 .9implies an odds of0.9\\n1−0.9=9.\\nOdds are traditionally used instead of probabilities in horse-racing, since\\nthey relate more naturally to the correct betting strategy.\\nBy taking the logarithm of both sides of ( 4.3), we arrive at\\nlog(p(X)\\n1−p(X))\\n=β0+β1X. (4.4)\\nThe left-hand side is called the log odds orlogit. We see that the logisticlog odds\\nlogitregression model ( 4.2) has a logit that is linear in X.\\nRecall from Chapter 3that in a linear regression model, β1gives the\\naverage change in Yassociated with a one-unit increase in X. By contrast,\\nin a logistic regression model, increasing Xby one unit changes the log\\nodds byβ1(4.4). Equivalently, it multiplies the odds by eβ1(4.3). However,\\nbecause the relationship between p(X)andXin (4.2) is not a straight line,\\nβ1doesnotcorrespond to the change in p(X)associated with a one-unit\\nincrease in X. The amount that p(X)changes due to a one-unit change in\\nXdepends on the current value of X. But regardless of the value of X, if\\nβ1is positive then increasing Xwill be associated with increasing p(X),\\nand ifβ1is negative then increasing Xwill be associated with decreasing\\np(X). The fact that there is not a straight-line relationship between p(X)\\nandX, and the fact that the rate of change in p(X)per unit change in X\\ndepends on the current value of X, can also be seen by inspection of the\\nright-hand panel of Figure 4.2.\\n4.3.2 Estimating the Regression Coefficients\\nThe coefficients β0andβ1in (4.2) are unknown, and must be estimated\\nbasedontheavailabletrainingdata.InChapter 3,weusedtheleastsquares\\napproach to estimate the unknown linear regression coefficients. Although\\nwe could use (non-linear) least squares to fit the model ( 4.4), the more\\ngeneral method of maximum likelihood is preferred, since it has better sta-\\ntistical properties. The basic intuition behind using maximum likelihood', metadata={'source': './documents/ISLP_website.pdf', 'page': 148}), Document(page_content='4.3 Logistic Regression 141\\nto fit a logistic regression model is as follows: we seek estimates for β0and\\nβ1such that the predicted probability ˆp(xi)of default for each individual,\\nusing (4.2), corresponds as closely as possible to the individual’s observed\\ndefault status. In other words, we try to find ˆβ0andˆβ1such that plugging\\nthese estimates into the model for p(X), given in ( 4.2), yields a number\\nclose to one for all individuals who defaulted, and a number close to zero\\nfor all individuals who did not. This intuition can be formalized using a\\nmathematical equation called a likelihood function :likelihood\\nfunction\\nℓ(β0,β1)=∏\\ni:yi=1p(xi)∏\\ni′:yi′=0(1−p(xi′)). (4.5)\\nThe estimates ˆβ0andˆβ1are chosen to maximize this likelihood function.\\nMaximum likelihood is a very general approach that is used to fit many\\nof the non-linear models that we examine throughout this book. In the\\nlinear regression setting, the least squares approach is in fact a special case\\nof maximum likelihood. The mathematical details of maximum likelihood\\nare beyond the scope of this book. However, in general, logistic regression\\nand other models can be easily fit using statistical software such as R, and\\nso we do not need to concern ourselves with the details of the maximum\\nlikelihood fitting procedure.\\nTable4.1shows the coefficient estimates and related information that\\nresult from fitting a logistic regression model on the Defaultdata in order\\nto predict the probability of default=Yesusingbalance. We see that ˆβ1=\\n0.0055 ; this indicates that an increase in balance is associated with an\\nincrease in the probability of default. To be precise, a one-unit increase in\\nbalanceis associated with an increase in the log odds of defaultby0.0055\\nunits.\\nMany aspects of the logistic regression output shown in Table 4.1are\\nsimilar to the linear regression output of Chapter 3. For example, we can\\nmeasure the accuracy of the coefficient estimates by computing their stan-\\ndarderrors.The z-statisticinTable 4.1playsthesameroleasthe t-statistic\\nin the linear regression output, for example in Table 3.1on page 77. For\\ninstance, the z-statistic associated with β1is equal to ˆβ1/SE(ˆβ1), and so a\\nlarge (absolute) value of the z-statistic indicates evidence against the null\\nhypothesis H0:β1=0. This null hypothesis implies that p(X)=eβ0\\n1+eβ0: in\\nother words, that the probability of defaultdoes not depend on balance.\\nSince the p-value associated with balancein Table 4.1is tiny, we can reject\\nH0. In other words,we conclude that there is indeed an association between\\nbalanceand probability of default. The estimated intercept in Table 4.1\\nis typically not of interest; its main purpose is to adjust the average fitted\\nprobabilities to the proportion of ones in the data (in this case, the overall\\ndefault rate).\\n4.3.3 Making Predictions\\nOnce the coefficients have been estimated, we can compute the probability\\nofdefaultfor any given credit card balance. For example, using the coeffi-\\ncient estimates given in Table 4.1, we predict that the default probability', metadata={'source': './documents/ISLP_website.pdf', 'page': 149}), Document(page_content='142 4. Classification\\nCoefficient Std. error z-statistic p-value\\nIntercept −10.6513 0.3612 −29.5<0.0001\\nbalance 0.0055 0.0002 24.9 <0.0001\\nTABLE 4.1. For theDefault data, estimated coefficients of the logistic regres-\\nsion model that predicts the probability of default usingbalance. A one-unit\\nincrease in balance is associated with an increase in the log odds of default by\\n0.0055units.\\nCoefficient Std. error z-statistic p-value\\nIntercept −3.5041 0.0707 −49.55<0.0001\\nstudent[Yes] 0.4049 0.1150 3.52 0.0004\\nTABLE 4.2. FortheDefault data,estimatedcoefficientsofthelogisticregression\\nmodel that predicts the probability of default using student status. Student status\\nis encoded as a dummy variable, with a value of 1for a student and a value of 0\\nfor a non-student, and represented by the variable student[Yes] in the table.\\nfor an individual with a balanceof$1,000is\\nˆp(X)=eˆβ0+ˆβ1X\\n1+eˆβ0+ˆβ1X=e−10.6513+0.0055×1,000\\n1+e−10.6513+0.0055×1,000=0.00576,\\nwhich is below 1%. In contrast, the predicted probability of default for an\\nindividual with a balance of $2,000is much higher, and equals 0.586or\\n58.6%.\\nOne can use qualitative predictors with the logistic regression model us-\\ning the dummy variable approach from Section 3.3.1. As an example, the\\nDefaultdata set contains the qualitative variable student. To fit a model\\nthat uses student status as a predictor variable, we simply create a dummy\\nvariable that takes on a value of 1for students and 0for non-students. The\\nlogistic regression model that results from predicting probability of default\\nfrom student status can be seen in Table 4.2. The coefficient associated\\nwith the dummy variable is positive, and the associated p-value is statisti-\\ncally significant. This indicates that students tend to have higher default\\nprobabilities than non-students:\\nˆPr(default=Yes|student=Yes)=e−3.5041+0.4049×1\\n1+e−3.5041+0.4049×1=0.0431,\\nˆPr(default=Yes|student=No)=e−3.5041+0.4049×0\\n1+e−3.5041+0.4049×0=0.0292.\\n4.3.4 Multiple Logistic Regression\\nWenowconsidertheproblemofpredictingabinaryresponseusingmultiple\\npredictors. By analogy with the extension from simple to multiple linear\\nregression in Chapter 3, we can generalize ( 4.4) as follows:\\nlog(p(X)\\n1−p(X))\\n=β0+β1X1+···+βpXp, (4.6)\\nwhereX=(X1,...,X p)areppredictors. Equation 4.6can be rewritten as\\np(X)=eβ0+β1X1+···+βpXp\\n1+eβ0+β1X1+···+βpXp. (4.7)', metadata={'source': './documents/ISLP_website.pdf', 'page': 150}), Document(page_content='4.3 Logistic Regression 143\\nCoefficient Std. error z-statistic p-value\\nIntercept −10.8690 0.4923 −22.08<0.0001\\nbalance 0.0057 0.0002 24.74 <0.0001\\nincome 0.0030 0.0082 0.37 0.7115\\nstudent[Yes] −0.6468 0.2362 −2.74 0.0062\\nTABLE 4.3. FortheDefault data,estimatedcoefficientsofthelogisticregression\\nmodel that predicts the probability of default usingbalance,income, and student\\nstatus. Student status is encoded as a dummy variable student[Yes] , with a value\\nof1for a student and a value of 0for a non-student. In fitting this model, income\\nwas measured in thousands of dollars.\\nJustasinSection 4.3.2,weusethemaximumlikelihoodmethodtoestimate\\nβ0,β1,...,βp.\\nTable4.3shows the coefficient estimates for a logistic regression model\\nthat uses balance,income(in thousands of dollars), and studentstatus to\\npredict probability of default. There is a surprising result here. The p-\\nvalues associated with balanceand the dummy variable for studentstatus\\nare very small, indicating that each of these variables is associated with\\nthe probability of default. However, the coefficient for the dummy variable\\nis negative, indicating that students are less likely to default than non-\\nstudents. In contrast, the coefficient for the dummy variable is positive in\\nTable4.2. How is it possible for student status to be associated with an\\nincrease in probability of default in Table 4.2and adecrease in probability\\nof default in Table 4.3? The left-hand panel of Figure 4.3provides a graph-\\nical illustration of this apparent paradox. The orange and blue solid lines\\nshow the average default rates for students and non-students, respectively,\\nas a function of credit card balance. The negative coefficient for studentin\\nthe multiple logistic regression indicates that for a fixed value of balance\\nandincome, a student is less likely to default than a non-student. Indeed,\\nwe observe from the left-hand panel of Figure 4.3that the student default\\nrate is at or below that of the non-student default rate for every value of\\nbalance. But the horizontal broken lines near the base of the plot, which\\nshow the default rates for students and non-students averaged over all val-\\nues ofbalanceandincome, suggest the opposite effect: the overall student\\ndefaultrateishigherthanthenon-studentdefaultrate.Consequently,there\\nis a positive coefficient for studentin the single variable logistic regression\\noutput shown in Table 4.2.\\nThe right-hand panel of Figure 4.3provides an explanation for this dis-\\ncrepancy. The variables studentandbalanceare correlated. Students tend\\nto hold higher levels of debt, which is in turn associated with higher prob-\\nability of default. In other words, students are more likely to have large\\ncredit card balances, which, as we know from the left-hand panel of Fig-\\nure4.3, tend to be associated with high default rates. Thus, even though\\nan individual student with a given credit card balance will tend to have a\\nlower probability of default than a non-student with the same credit card\\nbalance, the fact that students on the whole tend to have higher credit card\\nbalances means that overall, students tend to default at a higher rate than\\nnon-students. This is an important distinction for a credit card company\\nthat is trying to determine to whom they should offer credit. A student is\\nriskier than a non-student if no information about the student’s credit card', metadata={'source': './documents/ISLP_website.pdf', 'page': 151}), Document(page_content='144 4. Classification\\n500 1000 1500 20000.0 0.2 0.4 0.6 0.8Credit Card BalanceDefault RateNoYes0 500 1000 1500 2000 2500Student StatusCredit Card BalanceFIGURE 4.3. Confounding in the Default data.Left:Default rates are shown\\nfor students (orange) and non-students (blue). The solid lines display default rate\\nas a function of balance, while the horizontal broken lines display the overall\\ndefault rates. Right:Boxplots of balance for students (orange) and non-students\\n(blue) are shown.\\nbalance is available. However, that student is less risky than a non-student\\nwith the same credit card balance !\\nThis simple example illustrates the dangers and subtleties associated\\nwith performing regressions involving only a single predictor when other\\npredictors may also be relevant. As in the linear regression setting, the\\nresults obtained using one predictor may be quite different from those ob-\\ntained using multiple predictors, especially when there is correlation among\\nthe predictors. In general, the phenomenon seen in Figure 4.3is known as\\nconfounding .confoundingBy substituting estimates for the regression coefficients from Table 4.3\\ninto (4.7), we can make predictions. For example, a student with a credit\\ncard balance of $1,500and an income of $40,000has an estimated proba-\\nbility of default of\\nˆp(X)=e−10.869+0 .00574 ×1,500+0 .003×40−0.6468 ×1\\n1+e−10.869+0 .00574 ×1,500+0 .003×40−0.6468 ×1=0.058. (4.8)\\nA non-student with the same balance and income has an estimated prob-\\nability of default of\\nˆp(X)=e−10.869+0 .00574 ×1,500+0 .003×40−0.6468 ×0\\n1+e−10.869+0 .00574 ×1,500+0 .003×40−0.6468 ×0=0.105. (4.9)\\n(Here we multiply the incomecoefficient estimate from Table 4.3by 40,\\nrather than by 40,000, because in that table the model was fit with income\\nmeasured in units of $1,000.)\\n4.3.5 Multinomial Logistic Regression\\nWe sometimes wish to classify a response variable that has more than two\\nclasses. For example, in Section 4.2we had three categories of medical con-\\ndition in the emergency room: stroke,drug overdose ,epileptic seizure .\\nHowever, the logistic regression approach that we have seen in this section\\nonly allows for K=2classes for the response variable.', metadata={'source': './documents/ISLP_website.pdf', 'page': 152}), Document(page_content='4.3 Logistic Regression 145\\nIt turns out that it is possible to extend the two-class logistic regression\\napproach to the setting of K> 2classes. This extension is sometimes\\nknown as multinomial logistic regression . To do this, we first select a singlemultinomial\\nlogistic\\nregressionclass to serve as the baseline; without loss of generality, we select the Kth\\nclass for this role. Then we replace the model ( 4.7) with the model\\nPr(Y=k|X=x)=eβk0+βk1x1+···+βkpxp\\n1+∑K−1\\nl=1eβl0+βl1x1+···+βlpxp(4.10)\\nfork=1,...,K −1, and\\nPr(Y=K|X=x)=1\\n1+∑K−1\\nl=1eβl0+βl1x1+···+βlpxp. (4.11)\\nIt is not hard to show that for k=1,...,K −1,\\nlog(Pr(Y=k|X=x)\\nPr(Y=K|X=x))\\n=βk0+βk1x1+···+βkpxp. (4.12)\\nNoticethat( 4.12)isquitesimilarto( 4.6).Equation 4.12indicatesthatonce\\nagain, the log odds between any pair of classes is linear in the features.\\nIt turns out that in ( 4.10)–(4.12), the decision to treat the Kth class as\\nthebaselineisunimportant.Forexample,whenclassifyingemergencyroom\\nvisits into stroke,drug overdose , andepileptic seizure , suppose that we\\nfit two multinomial logistic regression models: one treating strokeas the\\nbaseline, another treating drug overdose as the baseline. The coefficient\\nestimates will differ between the two fitted models due to the differing\\nchoice of baseline, but the fitted values (predictions), the log odds between\\nany pair of classes, and the other key model outputs will remain the same.\\nNonetheless, interpretation of the coefficients in a multinomial logistic\\nregression model must be done with care, since it is tied to the choice\\nof baseline. For example, if we set epileptic seizure to be the baseline,\\nthen we can interpret βstroke 0as the log odds of strokeversusepileptic\\nseizure, given that x1=···=xp=0. Furthermore, a one-unit increase\\ninXjis associated with a βstroke jincrease in the log odds of strokeover\\nepileptic seizure . Stated another way, if Xjincreases by one unit, then\\nPr(Y=stroke|X=x)\\nPr(Y=epileptic seizure |X=x)\\nincreases by eβstroke j.\\nWe now briefly present an alternative coding for multinomial logistic\\nregression, known as the softmaxcoding. The softmax coding is equivalentsoftmaxto the coding just described in the sense that the fitted values, log odds\\nbetween any pair of classes, and other key model outputs will remain the\\nsame, regardless of coding. But the softmax coding is used extensively in\\nsome areas of the machine learning literature (and will appear again in\\nChapter 10), so it is worth being aware of it. In the softmax coding, rather\\nthan selecting a baseline class, we treat all Kclasses symmetrically, and\\nassume that for k=1,...,K ,\\nPr(Y=k|X=x)=eβk0+βk1x1+···+βkpxp\\n∑K\\nl=1eβl0+βl1x1+···+βlpxp. (4.13)', metadata={'source': './documents/ISLP_website.pdf', 'page': 153}), Document(page_content='146 4. Classification\\nThus, rather than estimating coefficients for K−1classes, we actually\\nestimate coefficients for all Kclasses. It is not hard to see that as a result\\nof (4.13), the log odds ratio between the kth andk′th classes equals\\nlog(Pr(Y=k|X=x)\\nPr(Y=k′|X=x))\\n=(βk0−βk′0)+(βk1−βk′1)x1+···+(βkp−βk′p)xp.\\n(4.14)\\n4.4 Generative Models for Classification\\nLogistic regression involves directly modeling Pr(Y=k|X=x)using the\\nlogistic function, given by ( 4.7) for the case of two response classes. In\\nstatistical jargon, we model the conditional distribution of the response Y,\\ngiven the predictor(s) X. We now consider an alternative and less direct\\napproach to estimating these probabilities. In this new approach, we model\\nthe distribution of the predictors Xseparately in each of the response\\nclasses (i.e. for each value of Y). We then use Bayes’ theorem to flip these\\naround into estimates for Pr(Y=k|X=x). When the distribution of X\\nwithin each class is assumed to be normal, it turns out that the model is\\nvery similar in form to logistic regression.\\nWhy do we need another method, when we have logistic regression?\\nThere are several reasons:\\n• When there is substantial separation between the two classes, the\\nparameter estimates for the logistic regression model are surprisingly\\nunstable. The methods that we consider in this section do not suffer\\nfrom this problem.\\n• If the distribution of the predictors Xis approximately normal in\\neach of the classes and the sample size is small, then the approaches\\nin this section may be more accurate than logistic regression.\\n• The methods in this section can be naturally extended to the case\\nof more than two response classes. (In the case of more than two\\nresponse classes, we can also use multinomial logistic regression from\\nSection4.3.5.)\\nSuppose that we wish to classify an observation into one of Kclasses,\\nwhereK≥2. In other words, the qualitative response variable Ycan take\\nonKpossible distinct and unordered values. Let πkrepresent the overall\\norpriorprobability that a randomly chosen observation comes from thepriorkth class. Let fk(X)≡Pr(X|Y=k)1denote the density function ofXdensity\\nfunctionfor an observation that comes from the kth class. In other words, fk(x)is\\nrelatively large if there is a high probability that an observation in the kth\\nclass has X≈x, andfk(x)is small if it is very unlikely that an observation\\nin thekth class has X≈x. ThenBayes’ theorem states thatBayes’\\ntheorem\\n1Technically, this definition is only correct if Xis a qualitative random variable. If\\nXis quantitative, then fk(x)dxcorresponds to the probability of Xfalling in a small\\nregiondxaroundx.', metadata={'source': './documents/ISLP_website.pdf', 'page': 154}), Document(page_content='4.4 Generative Models for Classification 147\\nPr(Y=k|X=x)=πkfk(x)∑K\\nl=1πlfl(x). (4.15)\\nInaccordancewithourearliernotation,wewillusetheabbreviation pk(x)=\\nPr(Y=k|X=x); this is the posterior probability that an observationposteriorX=xbelongs to the kth class. That is, it is the probability that the\\nobservation belongs to the kth class, giventhe predictor value for that\\nobservation.\\nEquation 4.15suggests that instead of directly computing the posterior\\nprobability pk(x)as in Section 4.3.1, we can simply plug in estimates of πk\\nandfk(x)into (4.15). In general, estimating πkis easy if we have a random\\nsample from the population: we simply compute the fraction of the training\\nobservations that belong to the kth class. However, estimating the density\\nfunction fk(x)is much more challenging. As we will see, to estimate fk(x),\\nwe will typically have to make some simplifying assumptions.\\nWe know from Chapter 2that the Bayes classifier, which classifies an\\nobservation xto the class for which pk(x)is largest, has the lowest possible\\nerror rate out of all classifiers. (Of course, this is only true if all of the\\nterms in ( 4.15) are correctly specified.) Therefore, if we can find a way to\\nestimate fk(x), then we can plug it into ( 4.15) in order to approximate the\\nBayes classifier.\\nIn the following sections, we discuss three classifiers that use different\\nestimates of fk(x)in (4.15) to approximate the Bayes classifier: linear dis-\\ncriminant analysis, quadratic discriminant analysis, andnaive Bayes .\\n4.4.1 Linear Discriminant Analysis for p=1\\nFor now, assume that p=1—that is, we have only one predictor. We would\\nlike to obtain an estimate for fk(x)that we can plug into ( 4.15) in order to\\nestimate pk(x). We will then classify an observation to the class for which\\npk(x)is greatest. To estimate fk(x), we will first make some assumptions\\nabout its form.\\nIn particular, we assume that fk(x)isnormalorGaussian . In the one-normal\\nGaussiandimensional setting, the normal density takes the form\\nfk(x)=1√\\n2πσkexp(\\n−1\\n2σ2\\nk(x−µk)2)\\n, (4.16)\\nwhereµkandσ2\\nkare the mean and variance parameters for the kth class.\\nFor now, let us further assume that σ2\\n1=···=σ2\\nK: that is, there is a shared\\nvariance term across all Kclasses, which for simplicity we can denote by\\nσ2. Plugging ( 4.16) into (4.15), we find that\\npk(x)=πk1√\\n2πσexp(\\n−1\\n2σ2(x−µk)2)\\n∑K\\nl=1πl1√\\n2πσexp(\\n−1\\n2σ2(x−µl)2). (4.17)\\n(Note that in ( 4.17),πkdenotes the prior probability that an observation\\nbelongs to the kth class, not to be confused with π≈3.14159 , the math-\\nematical constant.) The Bayes classifier2involves assigning an observation\\n2Recall that the Bayes classifier assigns an observation to the class for which pk(x)\\nis largest. This is different from Bayes’ theorem in (4.15), which allows us to manipulate\\nconditional distributions.', metadata={'source': './documents/ISLP_website.pdf', 'page': 155}), Document(page_content='148 4. Classification\\n−4 −2 0 2 4−3 −2 −1 0 1 2 3 4012345FIGURE 4.4. Left:Two one-dimensional normal density functions are shown.\\nThe dashed vertical line represents the Bayes decision boundary. Right:20 obser-\\nvations were drawn from each of the two classes, and are shown as histograms.\\nThe Bayes decision boundary is again shown as a dashed vertical line. The solid\\nvertical line represents the LDA decision boundary estimated from the training\\ndata.\\nX=xto the class for which ( 4.17) is largest. Taking the logof (4.17) and\\nrearranging the terms, it is not hard to show3that this is equivalent to\\nassigning the observation to the class for which\\nδk(x)=x·µk\\nσ2−µ2\\nk\\n2σ2+ log( πk) (4.18)\\nis largest. For instance, if K=2 andπ1=π2, then the Bayes classifier\\nassigns an observation to class 1if2x(µ1−µ2)>µ2\\n1−µ2\\n2, and to class\\n2 otherwise. The Bayes decision boundary is the point for which δ1(x)=\\nδ2(x); one can show that this amounts to\\nx=µ2\\n1−µ2\\n2\\n2(µ1−µ2)=µ1+µ2\\n2. (4.19)\\nAnexampleisshownintheleft-handpanelofFigure 4.4.Thetwonormal\\ndensity functions that are displayed, f1(x)andf2(x), represent two distinct\\nclasses. The mean and variance parameters for the two density functions\\nareµ1=−1.25,µ2=1.25, andσ2\\n1=σ2\\n2=1. The two densities overlap,\\nand so given that X=x, there is some uncertainty about the class to which\\nthe observation belongs. If we assume that an observation is equally likely\\nto come from either class—that is, π1=π2=0.5—then by inspection of\\n(4.19), we see that the Bayes classifier assigns the observation to class 1\\nifx< 0and class 2 otherwise. Note that in this case, we can compute\\nthe Bayes classifier because we know that Xis drawn from a Gaussian\\ndistribution within each class, and we know all of the parameters involved.\\nIn a real-life situation, we are not able to calculate the Bayes classifier.\\nIn practice, even if we are quite certain of our assumption that Xis\\ndrawn from a Gaussian distribution within each class, to apply the Bayes\\nclassifier we still have to estimate the parameters µ1,...,µ K,π1,...,πK,\\nandσ2. Thelinear discriminant analysis (LDA) method approximates thelinear\\ndiscriminant\\nanalysisBayes classifier by plugging estimates for πk,µk, andσ2into (4.18). In\\n3See Exercise 2 at the end of this chapter.', metadata={'source': './documents/ISLP_website.pdf', 'page': 156}), Document(page_content='4.4 Generative Models for Classification 149\\nparticular, the following estimates are used:\\nˆµk=1\\nnk∑\\ni:yi=kxi\\nˆσ2=1\\nn−KK∑\\nk=1∑\\ni:yi=k(xi−ˆµk)2(4.20)\\nwherenis the total number of training observations, and nkis the number\\nof training observations in the kth class. The estimate for µkis simply the\\naverage of all the training observations from the kth class, while ˆσ2can\\nbe seen as a weighted average of the sample variances for each of the K\\nclasses. Sometimes we have knowledge of the class membership probabili-\\ntiesπ1,...,πK, which can be used directly. In the absence of any additional\\ninformation, LDA estimates πkusing the proportion of the training obser-\\nvations that belong to the kth class. In other words,\\nˆπk=nk/n. (4.21)\\nTheLDAclassifierplugstheestimatesgivenin( 4.20)and(4.21)into(4.18),\\nand assigns an observation X=xto the class for which\\nˆδk(x)=x·ˆµk\\nˆσ2−ˆµ2\\nk\\n2ˆσ2+ log(ˆ πk) (4.22)\\nis largest. The word linearin the classifier’s name stems from the fact\\nthat the discriminant functions ˆδk(x)in (4.22) are linear functions of x(asdiscriminant\\nfunctionopposed to a more complex function of x).\\nThe right-hand panel of Figure 4.4displays a histogram of a random\\nsample of 20observations from each class. To implement LDA, we began\\nbyestimating πk,µk,andσ2using(4.20)and(4.21).Wethencomputedthe\\ndecision boundary, shown as a black solid line, that results from assigning\\nan observation to the class for which ( 4.22) is largest. All points to the left\\nof this line will be assigned to the green class, while points to the right of\\nthis line are assigned to the purple class. In this case, since n1=n2= 20 ,\\nwe have ˆπ1=ˆπ2. As a result, the decision boundary corresponds to the\\nmidpoint between the sample means for the two classes, (ˆµ1+ˆµ2)/2. The\\nfigure indicates that the LDA decision boundary is slightly to the left of\\nthe optimal Bayes decision boundary, which instead equals (µ1+µ2)/2=\\n0. How well does the LDA classifier perform on this data? Since this is\\nsimulateddata,wecangeneratealargenumberoftestobservationsinorder\\nto compute the Bayes error rate and the LDA test error rate. These are\\n10.6% and11.1%, respectively. In other words, the LDA classifier’s error\\nrate is only 0.5% above the smallest possible error rate! This indicates that\\nLDA is performing pretty well on this data set.\\nTo reiterate, the LDA classifier results from assuming that the obser-\\nvations within each class come from a normal distribution with a class-\\nspecific mean and a common variance σ2, and plugging estimates for these\\nparameters into the Bayes classifier. In Section 4.4.3, we will consider a less\\nstringent set of assumptions, by allowing the observations in the kth class\\nto have a class-specific variance, σ2\\nk.', metadata={'source': './documents/ISLP_website.pdf', 'page': 157}), Document(page_content='150 4. Classification\\nx1x1x2x2FIGURE 4.5. Two multivariate Gaussian density functions are shown, with\\np=2.Left:The two predictors are uncorrelated. Right:The two variables have\\na correlation of 0.7.\\n4.4.2 Linear Discriminant Analysis for p>1\\nWe now extend the LDA classifier to the case of multiple predictors. To\\ndo this, we will assume that X=(X1,X2,...,X p)is drawn from a multi-\\nvariate Gaussian (or multivariate normal) distribution, with a class-specificmultivariate\\nGaussianmeanvectorandacommoncovariancematrix.Webeginwithabriefreview\\nof this distribution.\\nThemultivariateGaussiandistributionassumesthateachindividualpre-\\ndictorfollowsaone-dimensionalnormaldistribution,asin( 4.16),withsome\\ncorrelation between each pair of predictors. Two examples of multivariate\\nGaussian distributions with p=2 are shown in Figure 4.5. The height of\\nthe surface at any particular point represents the probability that both X1\\nandX2fall in a small region around that point. In either panel, if the sur-\\nfaceiscutalongthe X1axisoralongthe X2axis,theresultingcross-section\\nwill have the shape of a one-dimensional normal distribution. The left-hand\\npanel of Figure 4.5illustrates an example in which Var (X1)=Var(X2)and\\nCor(X1,X2)=0 ; this surface has a characteristic bell shape . However, the\\nbell shape will be distorted if the predictors are correlated or have unequal\\nvariances, as is illustrated in the right-hand panel of Figure 4.5. In this\\nsituation, the base of the bell will have an elliptical, rather than circular,\\nshape. To indicate that a p-dimensional random variable Xhas a multi-\\nvariate Gaussian distribution, we write X∼N(µ,Σ). HereE(X)=µis\\nthe mean of X(a vector with pcomponents), and Cov (X)= Σis the\\np×pcovariance matrix of X. Formally, the multivariate Gaussian density\\nis defined as\\nf(x)=1\\n(2π)p/2|Σ|1/2exp(\\n−1\\n2(x−µ)TΣ−1(x−µ))\\n. (4.23)\\nIn the case of p> 1predictors, the LDA classifier assumes that the\\nobservations in the kth class are drawn from a multivariate Gaussian dis-\\ntribution N(µk,Σ), whereµkis a class-specific mean vector, and Σis a\\ncovariance matrix that is common to all Kclasses. Plugging the density\\nfunction for the kth class, fk(X=x), into (4.15) and performing a little\\nbit of algebra reveals that the Bayes classifier assigns an observation X=x', metadata={'source': './documents/ISLP_website.pdf', 'page': 158}), Document(page_content='4.4 Generative Models for Classification 151\\n−4 −2 0 2 4−4 −2 0 2 4−4 −2 0 2 4−4 −2 0 2 4X1X1\\nX2X2\\nFIGURE 4.6. An example with three classes. The observations from each class\\nare drawn from a multivariate Gaussian distribution with p=2, with a class-spe-\\ncific mean vector and a common covariance matrix. Left:Ellipses that contain\\n95% of the probability for each of the three classes are shown. The dashed lines\\nare the Bayes decision boundaries. Right:20observations were generated from\\neach class, and the corresponding LDA decision boundaries are indicated using\\nsolid black lines. The Bayes decision boundaries are once again shown as dashed\\nlines.\\nto the class for which\\nδk(x)=xTΣ−1µk−1\\n2µT\\nkΣ−1µk+ log πk (4.24)\\nis largest. This is the vector/matrix version of ( 4.18).\\nAn example is shown in the left-hand panel of Figure 4.6. Three equally-\\nsized Gaussian classes are shown with class-specific mean vectors and a\\ncommon covariance matrix. The three ellipses represent regions that con-\\ntain95% of the probability for each of the three classes. The dashed lines\\nare the Bayes decision boundaries. In other words, they represent the set\\nof values xfor which δk(x)=δℓ(x); i.e.\\nxTΣ−1µk−1\\n2µT\\nkΣ−1µk=xTΣ−1µl−1\\n2µT\\nlΣ−1µl (4.25)\\nfork̸=l. (Thelogπkterm from ( 4.24) has disappeared because each of\\nthe three classes has the same number of training observations; i.e. πkis\\nthe same for each class.) Note that there are three lines representing the\\nBayes decision boundaries because there are three pairs of classes among\\nthe three classes. That is, one Bayes decision boundary separates class 1\\nfrom class 2, one separates class 1 from class 3, and one separates class 2\\nfrom class 3. These three Bayes decision boundaries divide the predictor\\nspace into three regions. The Bayes classifier will classify an observation\\naccording to the region in which it is located.\\nOnce again, we need to estimate the unknown parameters µ1,...,µ K,\\nπ1,...,πK, andΣ; the formulas are similar to those used in the one-\\ndimensional case, given in ( 4.20). To assign a new observation X=x,\\nLDA plugs these estimates into ( 4.24) to obtain quantities ˆδk(x), and clas-\\nsifies to the class for which ˆδk(x)is largest. Note that in ( 4.24)δk(x)is\\na linear function of x; that is, the LDA decision rule depends on xonly', metadata={'source': './documents/ISLP_website.pdf', 'page': 159}), Document(page_content='152 4. Classification\\nTrue default status\\nNo Yes Total\\nPredicted No9644 252 9896\\ndefault status Yes23 81 104\\nTotal9667 333 10000\\nTABLE 4.4. A confusion matrix compares the LDA predictions to the true\\ndefault statuses for the 10,000training observations in the Default data set.\\nElementsonthediagonalofthematrixrepresentindividualswhosedefaultstatuses\\nwere correctly predicted, while off-diagonal elements represent individuals that\\nwere misclassified. LDA made incorrect predictions for 23individuals who did\\nnot default and for 252individuals who did default.\\nthrough a linear combination of its elements. As previously discussed, this\\nis the reason for the word linearin LDA.\\nIn the right-hand panel of Figure 4.6,20observations drawn from each of\\nthe three classes are displayed, and the resulting LDA decision boundaries\\nare shown as solid black lines. Overall, the LDA decision boundaries are\\npretty close to the Bayes decision boundaries, shown again as dashed lines.\\nThetesterrorratesfortheBayesandLDAclassifiersare 0.0746and0.0770,\\nrespectively. This indicates that LDA is performing well on this data.\\nWe can perform LDA on the Defaultdata in order to predict whether\\nor not an individual will default on the basis of credit card balance and\\nstudent status.4The LDA model fit to the 10,000training samples results\\nin atrainingerror rate of 2.75%. This sounds like a low error rate, but two\\ncaveats must be noted.\\n• First of all, training error rates will usually be lower than test error\\nrates, which are the real quantity of interest. In other words, we\\nmight expect this classifier to perform worse if we use it to predict\\nwhether or not a new set of individuals will default. The reason is\\nthat we specifically adjust the parameters of our model to do well on\\nthe training data. The higher the ratio of parameters pto number\\nof samples n, the more we expect this overfitting to play a role. Foroverfittingthese data we don’t expect this to be a problem, since p=2and\\nn= 10,000.\\n• Second, since only 3.33% of the individuals in the training sample\\ndefaulted, a simple but useless classifier that always predicts that\\nan individual will not default, regardless of his or her credit card\\nbalance and student status, will result in an error rate of 3.33%. In\\nother words, the trivial nullclassifier will achieve an error rate thatnullis only a bit higher than the LDA training set error rate.\\nIn practice, a binary classifier such as this one can make two types of\\nerrors: it can incorrectly assign an individual who defaults to the no default\\ncategory, or it can incorrectly assign an individual who does not default to\\n4The careful reader will notice that student status is qualitative — thus, the normality\\nassumption made by LDA is clearly violated in this example! However, LDA is often\\nremarkably robust to model violations, as this example shows. Naive Bayes, discussed in\\nSection4.4.4, provides an alternative to LDA that does not assume normally distributed\\npredictors.', metadata={'source': './documents/ISLP_website.pdf', 'page': 160}), Document(page_content='4.4 Generative Models for Classification 153\\nthedefaultcategory. It is often of interest to determine which of these two\\ntypes of errors are being made. A confusion matrix , shown for the Defaultconfusion\\nmatrixdata in Table 4.4, is a convenient way to display this information. The\\ntable reveals that LDA predicted that a total of 104people would default.\\nOf these people, 81actually defaulted and 23did not. Hence only 23out\\nof9,667of the individuals who did not default were incorrectly labeled.\\nThis looks like a pretty low error rate! However, of the 333individuals who\\ndefaulted, 252(or75.7%) were missed by LDA. So while the overall error\\nrate is low, the error rate among individuals who defaulted is very high.\\nFrom the perspective of a credit card company that is trying to identify\\nhigh-risk individuals, an error rate of 252/333 = 75 .7% among individuals\\nwho default may well be unacceptable.\\nClass-specific performance is also important in medicine and biology,\\nwhere the terms sensitivity andspecificity characterize the performance ofsensitivity\\nspecificitya classifier or screening test. In this case the sensitivity is the percent-\\nage of true defaulters that are identified; it equals 24.3%. The specificity\\nis the percentage of non-defaulters that are correctly identified; it equals\\n(1−23/9667) = 99 .8%.\\nWhy does LDA do such a poor job of classifying the customers who de-\\nfault? In other words, why does it have such low sensitivity? As we have\\nseen, LDA is trying to approximate the Bayes classifier, which has the low-\\nesttotalerror rate out of all classifiers. That is, the Bayes classifier will\\nyield the smallest possible total number of misclassified observations, re-\\ngardlessoftheclassfromwhichtheerrorsstem.Somemisclassificationswill\\nresult from incorrectly assigning a customer who does not default to the\\ndefault class, and others will result from incorrectly assigning a customer\\nwho defaults to the non-default class. In contrast, a credit card company\\nmight particularly wish to avoid incorrectly classifying an individual who\\nwill default, whereas incorrectly classifying an individual who will not de-\\nfault, though still to be avoided, is less problematic. We will now see that it\\nis possible to modify LDA in order to develop a classifier that better meets\\nthe credit card company’s needs.\\nThe Bayes classifier works by assigning an observation to the class for\\nwhich the posterior probability pk(X)is greatest. In the two-class case, this\\namounts to assigning an observation to the defaultclass if\\nPr(default =Yes|X=x)>0.5. (4.26)\\nThus, the Bayes classifier, and by extension LDA, uses a threshold of 50%\\nfor the posterior probability of default in order to assign an observation\\nto thedefaultclass. However, if we are concerned about incorrectly pre-\\ndicting the default status for individuals who default, then we can consider\\nlowering this threshold. For instance, we might label any customer with a\\nposterior probability of default above 20% to the defaultclass. In other\\nwords, instead of assigning an observation to the defaultclass if ( 4.26)\\nholds, we could instead assign an observation to this class if\\nPr(default =Yes|X=x)>0.2. (4.27)\\nTheerrorratesthatresultfromtakingthisapproachareshowninTable 4.5.\\nNow LDA predicts that 430individuals will default. Of the 333individuals\\nwho default, LDA correctly predicts all but 138, or41.4%. This is a vast', metadata={'source': './documents/ISLP_website.pdf', 'page': 161}), Document(page_content='154 4. Classification\\nTrue default status\\nNo Yes Total\\nPredicted No9432 138 9570\\ndefault status Yes235 195 430\\nTotal9667 333 10000\\nTABLE 4.5. A confusion matrix compares the LDA predictions to the true\\ndefault statuses for the 10,000training observations in the Default data set,\\nusing a modified threshold value that predicts default for any individuals whose\\nposterior default probability exceeds 20%.\\n0.00.10.20.30.40.50.0 0.2 0.4 0.6ThresholdError RateFIGURE 4.7.For theDefaultdata set, error rates are shown as a function ofthe threshold value for the posterior probability that is used to perform the assign-ment. The black solid line displays the overall error rate. The blue dashed linerepresents the fraction of defaulting customers that are incorrectly classified, andthe orange dotted line indicates the fraction of errors among the non-defaultingcustomers.improvement over the error rate of75.7% that resulted from using thethreshold of50%. However, this improvement comes at a cost: now235individuals who do not default are incorrectly classified. As a result, theoverallerrorratehasincreasedslightlyto3.73%.Butacreditcardcompanymayconsiderthis slightincreasein thetotalerror ratetobe asmallprice topay for more accurate identification of individuals who do indeed default.Figure4.7illustratesthetrade-offthatresultsfrommodifyingthethresh-old value for the posterior probability of default. Various error rates areshown as a function of the threshold value. Using a threshold of0.5, as in(4.26), minimizes the overall error rate, shown as a black solid line. Thisis to be expected, since the Bayes classifier uses a threshold of0.5and isknown to have the lowest overall error rate. But when a threshold of0.5isused, the error rate among the individuals who default is quite high (bluedashed line). As the threshold is reduced, the error rate among individualswho default decreases steadily, but the error rate among the individualswho do not default increases. How can we decide which threshold value isbest? Such a decision must be based ondomain knowledge, such as detailedinformation about the costs associated with default.TheROC curveis a popular graphic for simultaneously displaying theROC curvetwo types of errors for all possible thresholds. The name “ROC” is historic,and comes from communications theory. It is an acronym forreceiver op-erating characteristics. Figure4.8displays the ROC curve for the LDAclassifier on the training data. The overall performance of a classifier, sum-', metadata={'source': './documents/ISLP_website.pdf', 'page': 162}), Document(page_content='4.4 Generative Models for Classification 155ROC Curve\\nFalse positive rateTrue positive rate0.0 0.2 0.4 0.6 0.8 1.00.0 0.2 0.4 0.6 0.8 1.0FIGURE 4.8. A ROC curve for the LDA classifier on the Default data. It\\ntraces out two types of error as we vary the threshold value for the posterior\\nprobability of default. The actual thresholds are not shown. The true positive rate\\nis the sensitivity: the fraction of defaulters that are correctly identified, using\\na given threshold value. The false positive rate is 1-specificity: the fraction of\\nnon-defaulters that we classify incorrectly as defaulters, using that same threshold\\nvalue. The ideal ROC curve hugs the top left corner, indicating a high true positive\\nrate and a low false positive rate. The dotted line represents the “no information”\\nclassifier; this is what we would expect if student status and credit card balance\\nare not associated with probability of default.\\nmarized over all possible thresholds, is given by the area under the (ROC)\\ncurve(AUC). AnidealROCcurvewillhugthetopleftcorner,sothelargerarea under\\nthe (ROC)\\ncurvethe AUC the better the classifier. For this data the AUC is 0.95, which is\\nclose to the maximum of 1.0, so would be considered very good. We expect\\na classifier that performs no better than chance to have an AUC of 0.5\\n(when evaluated on an independent test set not used in model training).\\nROC curves are useful for comparing different classifiers, since they take\\ninto account all possible thresholds. It turns out that the ROC curve for\\nthe logistic regression model of Section 4.3.4fit to these data is virtually\\nindistinguishable from this one for the LDA model, so we do not display it\\nhere.\\nAs we have seen above, varying the classifier threshold changes its true\\npositive and false positive rate. These are also called the sensitivity and onesensitivityminus the specificity of our classifier. Since there is an almost bewilderingspecificityarray of terms used in this context, we now give a summary. Table 4.6\\nshows the possible results when applying a classifier (or diagnostic test)\\nto a population. To make the connection with the epidemiology literature,\\nwe think of “+” as the “disease” that we are trying to detect, and “ −” as\\nthe “non-disease” state. To make the connection to the classical hypothesis\\ntesting literature, we think of “ −” as the null hypothesis and “+” as the', metadata={'source': './documents/ISLP_website.pdf', 'page': 163}), Document(page_content='156 4. Classification\\nTrue class\\n−or Null +or Non-null Total\\nPredicted −or Null True Neg. (TN) False Neg. (FN) N∗\\nclass +or Non-null False Pos. (FP) True Pos. (TP) P∗\\nTotal N P\\nTABLE 4.6. Possible results when applying a classifier or diagnostic test to a\\npopulation.\\nName Definition Synonyms\\nFalse Pos. rate FP/NType I error, 1 −Specificity\\nTrue Pos. rate TP/P1−Type II error, power, sensitivity, recall\\nPos. Pred. value TP/P∗Precision, 1 −false discovery proportion\\nNeg. Pred. value TN/N∗\\nTABLE 4.7. Importantmeasuresforclassificationanddiagnostictesting,derived\\nfrom quantities in Table 4.6.\\nalternative (non-null) hypothesis. In the context of the Defaultdata, “+”\\nindicates an individual who defaults, and “ −” indicates one who does not.\\nTable4.7listsmanyofthepopularperformancemeasuresthatareusedin\\nthis context. The denominators for the false positive and true positive rates\\naretheactualpopulationcountsineachclass.Incontrast,thedenominators\\nfor the positive predictive value and the negative predictive value are the\\ntotal predicted counts for each class.\\n4.4.3 Quadratic Discriminant Analysis\\nAs we have discussed, LDA assumes that the observations within each class\\nare drawn from a multivariate Gaussian distribution with a class-specific\\nmean vector and a covariance matrix that is common to all Kclasses.\\nQuadratic discriminant analysis (QDA) provides an alternative approach.quadratic\\ndiscriminant\\nanalysisLike LDA, the QDA classifier results from assuming that the observations\\nfrom each class are drawn from a Gaussian distribution, and plugging es-\\ntimates for the parameters into Bayes’ theorem in order to perform pre-\\ndiction. However, unlike LDA, QDA assumes that each class has its own\\ncovariance matrix. That is, it assumes that an observation from the kth\\nclass is of the form X∼N(µk,Σk), whereΣkis a covariance matrix for\\nthekth class. Under this assumption, the Bayes classifier assigns an obser-\\nvationX=xto the class for which\\nδk(x)=−1\\n2(x−µk)TΣ−1\\nk(x−µk)−1\\n2log|Σk|+ logπk\\n=−1\\n2xTΣ−1\\nkx+xTΣ−1\\nkµk−1\\n2µT\\nkΣ−1\\nkµk−1\\n2log|Σk|+ logπk\\n(4.28)\\nis largest. So the QDA classifier involves plugging estimates for Σk,µk,\\nandπkinto (4.28), and then assigning an observation X=xto the class\\nfor which this quantity is largest. Unlike in ( 4.24), the quantity xappears\\nas aquadratic function in ( 4.28). This is where QDA gets its name.\\nWhy does it matter whether or not we assume that the Kclasses share a\\ncommon covariance matrix? In other words, why would one prefer LDA to', metadata={'source': './documents/ISLP_website.pdf', 'page': 164}), Document(page_content='4.4 Generative Models for Classification 157\\n−4 −2 0 2 4−4 −3 −2 −1 0 1 2−4 −2 0 2 4−4 −3 −2 −1 0 1 2X1X1\\nX2X2\\nFIGURE 4.9. Left:The Bayes (purple dashed), LDA (black dotted), and QDA\\n(green solid) decision boundaries for a two-class problem with Σ1=Σ2. The\\nshading indicates the QDA decision rule. Since the Bayes decision boundary is\\nlinear, it is more accurately approximated by LDA than by QDA. Right:Details\\nare as given in the left-hand panel, except that Σ1̸=Σ2. Since the Bayes decision\\nboundary is non-linear, it is more accurately approximated by QDA than by LDA.\\nQDA, or vice-versa? The answer lies in the bias-variance trade-off. When\\nthere are ppredictors, then estimating a covariance matrix requires esti-\\nmatingp(p+1)/2parameters. QDA estimates a separate covariance matrix\\nfor each class, for a total of Kp(p+1)/2parameters. With 50 predictors this\\nis some multiple of 1,275, which is a lot of parameters. By instead assum-\\ning that the Kclasses share a common covariance matrix, the LDA model\\nbecomes linear in x, which means there are Kplinear coefficients to esti-\\nmate. Consequently, LDA is a much less flexible classifier than QDA, and\\nso has substantially lower variance. This can potentially lead to improved\\nprediction performance. But there is a trade-off: if LDA’s assumption that\\ntheKclasses share a common covariance matrix is badly off, then LDA\\ncan suffer from high bias. Roughly speaking, LDA tends to be a better bet\\nthan QDA if there are relatively few training observations and so reducing\\nvariance is crucial. In contrast, QDA is recommended if the training set is\\nvery large, so that the variance of the classifier is not a major concern, or if\\nthe assumption of a common covariance matrix for the Kclasses is clearly\\nuntenable.\\nFigure4.9illustratestheperformancesofLDAandQDAintwoscenarios.\\nIn the left-hand panel, the two Gaussian classes have a common correla-\\ntion of0.7between X1andX2. As a result, the Bayes decision boundary\\nis linear and is accurately approximated by the LDA decision boundary.\\nThe QDA decision boundary is inferior, because it suffers from higher vari-\\nance without a corresponding decrease in bias. In contrast, the right-hand\\npanel displays a situation in which the orange class has a correlation of 0.7\\nbetween the variables and the blue class has a correlation of −0.7. Now\\nthe Bayes decision boundary is quadratic, and so QDA more accurately\\napproximates this boundary than does LDA.', metadata={'source': './documents/ISLP_website.pdf', 'page': 165}), Document(page_content='158 4. Classification\\n4.4.4 Naive Bayes\\nIn previous sections, we used Bayes’ theorem ( 4.15) to develop the LDA\\nand QDA classifiers. Here, we use Bayes’ theorem to motivate the popular\\nnaive Bayes classifier.naive BayesRecall that Bayes’ theorem ( 4.15) provides an expression for the pos-\\nterior probability pk(x) = Pr( Y=k|X=x)in terms of π1,...,πKand\\nf1(x),...,f K(x).Touse( 4.15)inpractice,weneedestimatesfor π1,...,πK\\nandf1(x),...,f K(x). As we saw in previous sections, estimating the prior\\nprobabilities π1,...,πKis typically straightforward: for instance, we can\\nestimate ˆπkas the proportion of training observations belonging to the kth\\nclass, for k=1,...,K .\\nHowever, estimating f1(x),...,f K(x)is more subtle. Recall that fk(x)\\nis thep-dimensional density function for an observation in the kth class,\\nfork=1,...,K . In general, estimating a p-dimensional density function is\\nchallenging. In LDA, we make a very strong assumption that greatly sim-\\nplifies the task: we assume that fkis the density function for a multivariate\\nnormal random variable with class-specific mean µk, and shared covariance\\nmatrixΣ. By contrast, in QDA, we assume that fkis the density function\\nfor a multivariate normal random variable with class-specific mean µk, and\\nclass-specific covariance matrix Σk. By making these very strong assump-\\ntions, we are able to replace the very challenging problem of estimating K\\np-dimensional density functions with the much simpler problem of estimat-\\ningKp-dimensional mean vectors and one (in the case of LDA) or K(in\\nthe case of QDA) (p×p)-dimensional covariance matrices.\\nThe naive Bayes classifier takes a different tack for estimating f1(x),...,\\nfK(x). Instead of assuming that these functions belong to a particular\\nfamily of distributions (e.g. multivariate normal), we instead make a single\\nassumption:\\nWithin the kth class, the ppredictors are independent.\\nStated mathematically, this assumption means that for k=1,...,K ,\\nfk(x)=fk1(x1)×fk2(x2)×···×fkp(xp), (4.29)\\nwherefkjis the density function of the jth predictor among observations\\nin thekth class.\\nWhy is this assumption so powerful? Essentially, estimating a p-dimen-\\nsional density function is challenging because we must consider not only\\nthemarginal distribution of each predictor — that is, the distribution ofmarginal\\ndistributioneach predictor on its own — but also the joint distribution of the predictors\\njoint\\ndistribution— that is, the association between the different predictors. In the case of\\na multivariate normal distribution, the association between the different\\npredictors is summarized by the off-diagonal elements of the covariance\\nmatrix. However, in general, this association can be very hard to charac-\\nterize, and exceedingly challenging to estimate. But by assuming that the\\npcovariates are independent within each class, we completely eliminate the\\nneed to worry about the association between the ppredictors, because we\\nhave simply assumed that there is noassociation between the predictors!\\nDo we really believe the naive Bayes assumption that the pcovariates\\nare independent within each class? In most settings, we do not. But even\\nthough this modeling assumption is made for convenience, it often leads to', metadata={'source': './documents/ISLP_website.pdf', 'page': 166}), Document(page_content='4.4 Generative Models for Classification 159\\nprettydecentresults, especially in settings where nis not large enough rela-\\ntive topfor us to effectively estimate the joint distribution of the predictors\\nwithin each class. In fact, since estimating a joint distribution requires such\\na huge amount of data, naive Bayes is a good choice in a wide range of set-\\ntings. Essentially, the naive Bayes assumption introduces some bias, but\\nreduces variance, leading to a classifier that works quite well in practice as\\na result of the bias-variance trade-off.\\nOnce we have made the naive Bayes assumption, we can plug ( 4.29) into\\n(4.15) to obtain an expression for the posterior probability,\\nPr(Y=k|X=x)=πk×fk1(x1)×fk2(x2)×···×fkp(xp)∑K\\nl=1πl×fl1(x1)×fl2(x2)×···×flp(xp)(4.30)\\nfork=1,...,K .\\nTo estimate the one-dimensional density function fkjusing training data\\nx1j,...,x nj, we have a few options.\\n• IfXjisquantitative,thenwecanassumethat Xj|Y=k∼N(µjk,σ2\\njk).\\nIn other words, we assume that within each class, the jth predictor is\\ndrawn from a (univariate) normal distribution. While this may sound\\na bit like QDA, there is one key difference, in that here we are assum-\\ning that the predictors are independent; this amounts to QDA with\\nan additional assumption that the class-specific covariance matrix is\\ndiagonal.\\n• IfXjis quantitative, then another option is to use a non-parametric\\nestimate for fkj. A very simple way to do this is by making a his-\\ntogram for the observations of the jth predictor within each class.\\nThen we can estimate fkj(xj)as the fraction of the training obser-\\nvations in the kth class that belong to the same histogram bin as\\nxj. Alternatively, we can use a kernel density estimator , which iskernel\\ndensity\\nestimatoressentially a smoothed version of a histogram.\\n• IfXjis qualitative, then we can simply count the proportion of train-\\ning observations for the jth predictor corresponding to each class. For\\ninstance, suppose that Xj∈{1,2,3}, and we have 100observations\\nin thekth class. Suppose that the jth predictor takes on values of 1,\\n2, and3in 32, 55, and 13 of those observations, respectively. Then\\nwe can estimate fkjas\\nˆfkj(xj)=\\uf8f1\\n\\uf8f4\\uf8f2\\n\\uf8f4\\uf8f30.32 ifxj=1\\n0.55 ifxj=2\\n0.13 ifxj=3.\\nWe now consider the naive Bayes classifier in a toy example with p=3\\npredictors and K=2 classes. The first two predictors are quantitative,\\nand the third predictor is qualitative with three levels. Suppose further\\nthatˆπ1=ˆπ2=0.5. The estimated density functions ˆfkjfork=1,2\\nandj=1,2,3are displayed in Figure 4.10. Now suppose that we wish\\nto classify a new observation, x∗= (0 .4,1.5,1)T. It turns out that in this', metadata={'source': './documents/ISLP_website.pdf', 'page': 167}), Document(page_content='160 4. Classification\\nDensity estimates for class k=1\\nˆf11ˆf12ˆf13\\n−4 −2 0 2 4\\nFrequency−2 0 2 4123Density estimates for class k=2\\nˆf21ˆf22ˆf23\\n−4 −2 0 2 4\\nFrequency−2 0 2 4123FIGURE 4.10. In the toy example in Section 4.4.4, we generate data with p=3\\npredictors and K=2classes. The first two predictors are quantitative, and the\\nthird predictor is qualitative with three levels. In each class, the estimated density\\nfor each of the three predictors is displayed. If the prior probabilities for the two\\nclasses are equal, then the observation x∗= (0.4,1.5,1)Thas a94.4%posterior\\nprobability of belonging to the first class.\\nTrue default status\\nNo Yes Total\\nPredicted No9621 244 9865\\ndefault status Yes46 89 135\\nTotal9667 333 10000\\nTABLE 4.8. Comparison of the naive Bayes predictions to the true default\\nstatus for the 10,000training observations in the Default data set, when we\\npredict default for any observation for which P(Y=default|X=x)>0.5.\\nexample, ˆf11(0.4) = 0.368,ˆf12(1.5) = 0.484,ˆf13(1) = 0.226, andˆf21(0.4) =\\n0.030,ˆf22(1.5) = 0.130,ˆf23(1) = 0.616.Pluggingtheseestimatesinto( 4.30)\\nresults in posterior probability estimates of Pr(Y=1|X=x∗)=0.944and\\nPr(Y=2|X=x∗)=0.056.\\nTable4.8providestheconfusionmatrixresultingfromapplyingthenaive\\nBayes classifier to the Defaultdata set, where we predict a default if the\\nposterior probability of a default — that is, P(Y=default|X=x)— ex-\\nceeds0.5. Comparing this to the results for LDA in Table 4.4, our findings\\nare mixed. While LDA has a slightly lower overall error rate, naive Bayes', metadata={'source': './documents/ISLP_website.pdf', 'page': 168}), Document(page_content='4.5 A Comparison of Classification Methods 161\\nTrue default status\\nNo Yes Total\\nPredicted No9339 130 9469\\ndefault status Yes328 203 531\\nTotal9667 333 10000\\nTABLE 4.9. Comparison of the naive Bayes predictions to the true default\\nstatus for the 10,000training observations in the Default data set, when we\\npredict default for any observation for which P(Y=default|X=x)>0.2.\\ncorrectly predicts a higher fraction of the true defaulters. In this implemen-\\ntation of naive Bayes, we have assumed that each quantitative predictor is\\ndrawn from a Gaussian distribution (and, of course, that within each class,\\neach predictor is independent).\\nJust as with LDA, we can easily adjust the probability threshold for\\npredicting a default. For example, Table 4.9provides the confusion matrix\\nresulting from predicting a default if P(Y=default|X=x)>0.2. Again,\\nthe results are mixed relative to LDA with the same threshold (Table 4.5).\\nNaiveBayeshasahighererrorrate,butcorrectlypredictsalmosttwo-thirds\\nof the true defaults.\\nIn this example, it should not be too surprising that naive Bayes does\\nnot convincingly outperform LDA: this data set has n= 10,000andp=2,\\nand so the reduction in variance resulting from the naive Bayes assumption\\nis not necessarily worthwhile. We expect to see a greater pay-off to using\\nnaive Bayes relative to LDA or QDA in instances where pis larger or nis\\nsmaller, so that reducing the variance is very important.\\n4.5 A Comparison of Classification Methods\\n4.5.1 An Analytical Comparison\\nWenowperforman analytical (ormathematical)comparisonofLDA,QDA,\\nnaive Bayes, and logistic regression. We consider these approaches in a\\nsetting with Kclasses, so that we assign an observation to the class that\\nmaximizes Pr(Y=k|X=x). Equivalently, we can set Kas thebaseline\\nclass and assign an observation to the class that maximizes\\nlog(Pr(Y=k|X=x)\\nPr(Y=K|X=x))\\n(4.31)\\nfork=1,...,K. Examining the specific form of ( 4.31) for each method\\nprovides a clear understanding of their similarities and differences.\\nFirst, for LDA, we can make use of Bayes’ theorem ( 4.15) as well as\\nthe assumption that the predictors within each class are drawn from a\\nmultivariate normal density ( 4.23) with class-specific mean and shared co-', metadata={'source': './documents/ISLP_website.pdf', 'page': 169}), Document(page_content='162 4. Classification\\nvariance matrix in order to show that\\nlog(Pr(Y=k|X=x)\\nPr(Y=K|X=x))\\n= log(πkfk(x)\\nπKfK(x))\\n= log(\\nπkexp(\\n−1\\n2(x−µk)TΣ−1(x−µk))\\nπKexp(\\n−1\\n2(x−µK)TΣ−1(x−µK)))\\n= log(πk\\nπK)\\n−1\\n2(x−µk)TΣ−1(x−µk)\\n+1\\n2(x−µK)TΣ−1(x−µK)\\n= log(πk\\nπK)\\n−1\\n2(µk+µK)TΣ−1(µk−µK)\\n+xTΣ−1(µk−µK)\\n=ak+p∑\\nj=1bkjxj, (4.32)\\nwhereak= log(\\nπk\\nπK)\\n−1\\n2(µk+µK)TΣ−1(µk−µK)andbkjis thejth\\ncomponent of Σ−1(µk−µK). Hence LDA, like logistic regression, assumes\\nthat the log odds of the posterior probabilities is linear in x.\\nUsing similar calculations, in the QDA setting ( 4.31) becomes\\nlog(Pr(Y=k|X=x)\\nPr(Y=K|X=x))\\n=ak+p∑\\nj=1bkjxj+p∑\\nj=1p∑\\nl=1ckjlxjxl,(4.33)\\nwhereak,bkj, andckjlare functions of πk,πK,µk,µK,ΣkandΣK. Again,\\nas the name suggests, QDA assumes that the log odds of the posterior\\nprobabilities is quadratic in x.\\nFinally, we examine ( 4.31) in the naive Bayes setting. Recall that in\\nthis setting, fk(x)is modeled as a product of pone-dimensional functions\\nfkj(xj)forj=1,...,p . Hence,\\nlog(Pr(Y=k|X=x)\\nPr(Y=K|X=x))\\n= log(πkfk(x)\\nπKfK(x))\\n= log(\\nπk∏p\\nj=1fkj(xj)\\nπK∏p\\nj=1fKj(xj))\\n= log(πk\\nπK)\\n+p∑\\nj=1log(fkj(xj)\\nfKj(xj))\\n=ak+p∑\\nj=1gkj(xj), (4.34)\\nwhereak= log(\\nπk\\nπK)\\nandgkj(xj) = log(\\nfkj(xj)\\nfKj(xj))\\n. Hence, the right-hand\\nside of (4.34) takes the form of a generalized additive model , a topic that is\\ndiscussed further in Chapter 7.', metadata={'source': './documents/ISLP_website.pdf', 'page': 170}), Document(page_content='4.5 A Comparison of Classification Methods 163\\nInspection of ( 4.32), (4.33), and (4.34) yields the following observations\\nabout LDA, QDA, and naive Bayes:\\n• LDA is a special case of QDA with ckjl=0 for allj=1,...,p ,\\nl=1,...,p , andk=1,...,K . (Of course, this is not surprising, since\\nLDAissimplyarestrictedversionofQDAwith Σ1=···=ΣK=Σ.)\\n• Anyclassifierwithalineardecisionboundaryisaspecialcaseofnaive\\nBayes with gkj(xj)= bkjxj. In particular, this means that LDA is\\na special case of naive Bayes! This is not at all obvious from the\\ndescriptions of LDA and naive Bayes earlier in this chapter, since\\neach method makes very different assumptions: LDA assumes that\\nthe features are normally distributed with a common within-class\\ncovariance matrix, and naive Bayes instead assumes independence of\\nthe features.\\n• If we model fkj(xj)in the naive Bayes classifier using a one-dimensio-\\nnal Gaussian distribution N(µkj,σ2\\nj), then we end up with gkj(xj)=\\nbkjxjwherebkj=(µkj−µKj)/σ2\\nj.Inthiscase,naiveBayesisactually\\na special case of LDA with Σrestricted to be a diagonal matrix with\\njth diagonal element equal to σ2\\nj.\\n• Neither QDA nor naive Bayes is a special case of the other. Naive\\nBayes can produce a more flexible fit, since any choice can be made\\nforgkj(xj). However, it is restricted to a purely additive fit, in the\\nsense that in ( 4.34), a function of xjisaddedto a function of xl, for\\nj̸=l; however, these terms are never multiplied. By contrast, QDA\\nincludes multiplicative terms of the form ckjlxjxl. Therefore, QDA\\nhas the potential to be more accurate in settings where interactions\\namongthepredictorsareimportantindiscriminatingbetweenclasses.\\nNone of these methods uniformly dominates the others: in any setting, the\\nchoice of method will depend on the true distribution of the predictors in\\neach of the Kclasses, as well as other considerations, such as the values of\\nnandp. The latter ties into the bias-variance trade-off.\\nHow does logistic regression tie into this story? Recall from ( 4.12) that\\nmultinomial logistic regression takes the form\\nlog(Pr(Y=k|X=x)\\nPr(Y=K|X=x))\\n=βk0+p∑\\nj=1βkjxj.\\nThis is identical to the linear form of LDA ( 4.32): in both cases,\\nlog(\\nPr(Y=k|X=x)\\nPr(Y=K|X=x))\\nis a linear function of the predictors. In LDA, the co-\\nefficients in this linear function are functions of estimates for πk,πK,µk,\\nµK, andΣobtained by assuming that X1,...,X pfollow a normal distri-\\nbution within each class. By contrast, in logistic regression, the coefficients\\nare chosen to maximize the likelihood function ( 4.5). Thus, we expect LDA\\nto outperform logistic regression when the normality assumption (approxi-\\nmately) holds, and we expect logistic regression to perform better when it\\ndoes not.', metadata={'source': './documents/ISLP_website.pdf', 'page': 171}), Document(page_content='164 4. Classification\\nWe close with a brief discussion of K-nearest neighbors (KNN), intro-\\nduced in Chapter 2. Recall that KNN takes a completely different approach\\nfrom the classifiers seen in this chapter. In order to make a prediction for\\nan observation X=x, the training observations that are closest to xare\\nidentified. Then Xis assigned to the class to which the plurality of these\\nobservations belong. Hence KNN is a completely non-parametric approach:\\nno assumptions are made about the shape of the decision boundary. We\\nmake the following observations about KNN:\\n• Because KNN is completely non-parametric, we can expect this ap-\\nproach to dominate LDA and logistic regression when the decision\\nboundary is highly non-linear, provided that nis very large and pis\\nsmall.\\n• In order to provide accurate classification, KNN requires a lotof ob-\\nservationsrelativetothenumberofpredictors—thatis, nmuchlarger\\nthanp. This has to do with the fact that KNN is non-parametric, and\\nthus tends to reduce the bias while incurring a lot of variance.\\n• In settings where the decision boundary is non-linear but nis only\\nmodest, or pis not very small, then QDA may be preferred to KNN.\\nThis is because QDA can provide a non-linear decision boundary\\nwhile taking advantage of a parametric form, which means that it\\nrequires a smaller sample size for accurate classification, relative to\\nKNN.\\n• Unlike logistic regression, KNN does not tell us which predictors are\\nimportant: we don’t get a table of coefficients as in Table 4.3.\\n4.5.2 An Empirical Comparison\\nWe now compare the empirical (practical) performance of logistic regres-\\nsion, LDA, QDA, naive Bayes, and KNN. We generated data from six dif-\\nferent scenarios, each of which involves a binary (two-class) classification\\nproblem. In three of the scenarios, the Bayes decision boundary is linear,\\nand in the remaining scenarios it is non-linear. For each scenario, we pro-\\nduced100random training data sets. On each of these training sets, we\\nfit each method to the data and computed the resulting test error rate on\\na large test set. Results for the linear scenarios are shown in Figure 4.11,\\nand the results for the non-linear scenarios are in Figure 4.12. The KNN\\nmethod requires selection of K, the number of neighbors (not to be con-\\nfused with the number of classes in earlier sections of this chapter). We\\nperformed KNN with two values of K:K=1, and a value of Kthat was\\nchosen automatically using an approach called cross-validation , which we\\ndiscuss further in Chapter 5. We applied naive Bayes assuming univariate\\nGaussian densities for the features within each class (and, of course — since\\nthis is the key characteristic of naive Bayes — assuming independence of\\nthe features).\\nIn each of the six scenarios, there were p=2 quantitative predictors.\\nThe scenarios were as follows:', metadata={'source': './documents/ISLP_website.pdf', 'page': 172}), Document(page_content='4.5 A Comparison of Classification Methods 165KNN−1KNN−CVLDALogisticNBayesQDASCENARIO 10.25 0.30 0.35 0.40 0.45\\nKNN−1KNN−CVLDALogisticNBayesQDASCENARIO 20.15 0.20 0.25 0.30\\nKNN−1KNN−CVLDALogisticNBayesQDASCENARIO 30.20 0.25 0.30 0.35 0.40 0.45 0.50FIGURE 4.11. Boxplots of the test error rates for each of the linear scenarios\\ndescribed in the main text.\\nScenario 1: There were 20training observations in each of two classes. The\\nobservations within each class were uncorrelated random normal variables\\nwithadifferentmeanineachclass.Theleft-handpanelofFigure 4.11shows\\nthat LDA performed well in this setting, as one would expect since this is\\nthe model assumed by LDA. Logistic regression also performed quite well,\\nsinceitassumesalineardecisionboundary.KNNperformedpoorlybecause\\nitpaidapriceintermsofvariancethatwasnotoffsetbyareductioninbias.\\nQDA also performed worse than LDA, since it fit a more flexible classifier\\nthan necessary. The performance of naive Bayes was slightly better than\\nQDA, because the naive Bayes assumption of independent predictors is\\ncorrect.\\nScenario 2: Details are as in Scenario 1, except that within each class, the\\ntwo predictors had a correlation of −0.5. The center panel of Figure 4.11\\nindicates that the performance of most methods is similar to the previ-\\nous scenario. The notable exception is naive Bayes, which performs very\\npoorly here, since the naive Bayes assumption of independent predictors is\\nviolated.\\nScenario 3: As in the previous scenario, there is substantial negative cor-\\nrelation between the predictors within each class. However, this time we\\ngenerated X1andX2fromthe t-distribution ,with50observationsperclass.t-\\ndistribution Thet-distribution has a similar shape to the normal distribution, but it\\nhas a tendency to yield more extreme points—that is, more points that are\\nfar from the mean. In this setting, the decision boundary was still linear,\\nand so fit into the logistic regression framework. The set-up violated the\\nassumptions of LDA, since the observations were not drawn from a normal\\ndistribution. The right-hand panel of Figure 4.11shows that logistic regres-\\nsion outperformed LDA, though both methods were superior to the other\\napproaches. In particular, the QDA results deteriorated considerably as a\\nconsequence of non-normality. Naive Bayes performed very poorly because\\nthe independence assumption is violated.\\nScenario 4: The data were generated from a normal distribution, with a\\ncorrelation of 0.5between the predictors in the first class, and correlation of\\n−0.5betweenthepredictorsinthesecondclass.Thissetupcorrespondedto\\nthe QDA assumption, and resulted in quadratic decision boundaries. The\\nleft-hand panel of Figure 4.12shows that QDA outperformed all of the', metadata={'source': './documents/ISLP_website.pdf', 'page': 173}), Document(page_content='166 4. ClassificationKNN−1KNN−CVLDALogisticNBayesQDASCENARIO 40.300.350.40\\nKNN−1KNN−CVLDALogisticNBayesQDASCENARIO 50.18 0.20 0.22 0.24 0.26 0.28 0.30 0.32\\nKNN−1KNN−CVLDALogisticNBayesQDASCENARIO 60.15 0.20 0.25 0.30 0.35 0.40 0.45FIGURE 4.12. Boxplots of the test error rates for each of the non-linear\\nscenarios described in the main text.\\nother approaches. The naive Bayes assumption of independent predictors\\nis violated, so naive Bayes performs poorly.\\nScenario 5: The data were generated from a normal distribution with un-\\ncorrelated predictors. Then the responses were sampled from the logistic\\nfunction applied to a complicated non-linear function of the predictors. The\\ncenter panel of Figure 4.12shows that both QDA and naive Bayes gave\\nslightly better results than the linear methods, while the much more flexi-\\nble KNN-CV method gave the best results. But KNN with K=1gave the\\nworstresultsoutofallmethods.Thishighlightsthefactthatevenwhenthe\\ndata exhibits a complex non-linear relationship, a non-parametric method\\nsuch as KNN can still give poor results if the level of smoothness is not\\nchosen correctly.\\nScenario 6: The observations were generated from a normal distribution\\nwith a different diagonal covariance matrix for each class. However, the\\nsample size was verysmall: just n=6in each class. Naive Bayes performed\\nvery well, because its assumptions are met. LDA and logistic regression\\nperformed poorly because the true decision boundary is non-linear, due to\\nthe unequal covariance matrices. QDA performed a bit worse than naive\\nBayes, because given the very small sample size, the former incurred too\\nmuch variance in estimating the correlation between the predictors within\\neach class. KNN’s performance also suffered due to the very small sample\\nsize.\\nThese six examples illustrate that no one method will dominate the oth-\\ners in every situation. When the true decision boundaries are linear, then\\ntheLDAandlogisticregressionapproacheswilltendtoperformwell.When\\nthe boundaries are moderately non-linear, QDA or naive Bayes may give\\nbetter results. Finally, for much more complicated decision boundaries, a\\nnon-parametric approach such as KNN can be superior. But the level of\\nsmoothness for a non-parametric approach must be chosen carefully. In the\\nnext chapter we examine a number of approaches for choosing the correct\\nlevel of smoothness and, in general, for selecting the best overall method.\\nFinally,recallfromChapter 3thatintheregressionsettingwecanaccom-\\nmodate a non-linear relationship between the predictors and the response\\nby performing regression using transformations of the predictors. A similar\\napproach could be taken in the classification setting. For instance, we could', metadata={'source': './documents/ISLP_website.pdf', 'page': 174}), Document(page_content='4.6 Generalized Linear Models 167\\nCoefficient Std. error t-statistic p-value\\nIntercept 73.60 5.13 14.34 0.00\\nworkingday 1.27 1.78 0.71 0.48\\ntemp 157.21 10.26 15.32 0.00\\nweathersit[cloudy/misty] -12.89 1.96 -6.56 0.00\\nweathersit[light rain/snow] -66.49 2.97 -22.43 0.00\\nweathersit[heavy rain/snow] -109.75 76.67 -1.43 0.15\\nTABLE 4.10. Results for a least squares linear model fit to predict bikersin\\ntheBikeshare data. The predictors mnthandhrare omitted from this table due\\nto space constraints, and can be seen in Figure 4.13. For the qualitative variable\\nweathersit , the baseline level corresponds to clear skies.\\ncreate a more flexible version of logistic regression by including X2,X3,\\nand even X4as predictors. This may or may not improve logistic regres-\\nsion’s performance, depending on whether the increase in variance due to\\nthe added flexibility is offset by a sufficiently large reduction in bias. We\\ncould do the same for LDA. If we added all possible quadratic terms and\\ncross-products to LDA, the form of the model would be the same as the\\nQDA model, although the parameter estimates would be different. This\\ndevice allows us to move somewhere between an LDA and a QDA model.\\n4.6 Generalized Linear Models\\nIn Chapter 3, we assumed that the response Yis quantitative, and ex-\\nplored the use of least squares linear regression to predict Y. Thus far in\\nthis chapter, we have instead assumed that Yis qualitative. However, we\\nmay sometimes be faced with situations in which Yis neither qualitative\\nnor quantitative, and so neither linear regression from Chapter 3nor the\\nclassification approaches covered in this chapter is applicable.\\nAs a concrete example, we consider the Bikeshare data set. The response\\nisbikers, the number of hourly users of a bike sharing program in Wash-\\nington, DC. This response value is neither qualitative nor quantitative:\\ninstead, it takes on non-negative integer values, or counts. We will considercountspredicting bikersusing the covariates mnth(month of the year), hr(hour\\nof the day, from 0 to 23), workingday (an indicator variable that equals 1if\\nit is neither a weekend nor a holiday), temp(the normalized temperature,\\nin Celsius), and weathersit (a qualitative variable that takes on one of four\\npossible values: clear; misty or cloudy; light rain or light snow; or heavy\\nrain or heavy snow.)\\nIn the analyses that follow, we will treat mnth,hr, andweathersit as\\nqualitative variables.\\n4.6.1 Linear Regression on the Bikeshare Data\\nTo begin, we consider predicting bikersusing linear regression. The results\\nare shown in Table 4.10.\\nWe see, for example, that a progression of weather from clear to cloudy\\nresults in, on average, 12.89fewer bikers per hour; however, if the weather\\nprogresses further to rain or snow, then this further results in 53.60fewer\\nbikers per hour. Figure 4.13displays the coefficients associated with mnth', metadata={'source': './documents/ISLP_website.pdf', 'page': 175}), Document(page_content='168 4. Classification\\nFIGURE 4.13. A least squares linear regression model was fit to predict bikers\\nin theBikeshare data set. Left:The coefficients associated with the month of the\\nyear. Bike usage is highest in the spring and fall, and lowest in the winter. Right:\\nThe coefficients associated with the hour of the day. Bike usage is highest during\\npeak commute times, and lowest overnight.\\nand the coefficients associated with hr. We see that bike usage is highest in\\nthe spring and fall, and lowest during the winter months. Furthermore, bike\\nusage is greatest around rush hour (9 AM and 6 PM), and lowest overnight.\\nThus, at first glance, fitting a linear regression model to the Bikeshare data\\nset seems to provide reasonable and intuitive results.\\nBut upon more careful inspection, some issues become apparent. For\\nexample, 9.6% of the fitted values in the Bikeshare data set are negative:\\nthat is, the linear regression model predicts a negative number of users\\nduring 9.6% of the hours in the data set. This calls into question our ability\\nto perform meaningful predictions on the data, and it also raises concerns\\nabout the accuracy of the coefficient estimates, confidence intervals, and\\nother outputs of the regression model.\\nFurthermore, it is reasonable to suspect that when the expected value\\nofbikersis small, the variance of bikersshould be small as well. For\\ninstance, at 2 AM during a heavy December snow storm, we expect that\\nextremely few people will use a bike, and moreover that there should be\\nlittle variance associated with the number of users during those conditions.\\nThis is borne out in the data: between 1 AM and 4 AM, in December,\\nJanuary, and February, when it is raining, there are 5.05users, on average,\\nwith a standard deviation of 3.73. By contrast, between 7 AM and 10 AM,\\nin April, May, and June, when skies are clear, there are 243.59users, on\\naverage,withastandarddeviationof 131.7.Themean-variancerelationship\\nis displayed in the left-hand panel of Figure 4.14. This is a major violation\\nof the assumptions of a linear model, which state that Y=∑p\\nj=1Xjβj+ϵ,\\nwhereϵis a mean-zero error term with variance σ2that isconstant, and\\nnot a function of the covariates. Therefore, the heteroscedasticity of the\\ndata calls into question the suitability of a linear regression model.\\nFinally, the response bikersis integer-valued. But under a linear model,\\nY=β0+∑p\\nj=1Xjβj+ϵ, whereϵis a continuous-valued error term. This\\nmeans that in a linear model, the response Yis necessarily continuous-\\nvalued (quantitative). Thus, the integer nature of the response bikerssug-\\ngests that a linear regression model is not entirely satisfactory for this data\\nset./uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF\\n/uni25CF\\n/uni25CF/uni25CF/uni25CF/uni25CF\\n/uni25CF\\n/uni25CF−40 −20 0 20\\nMonthCoefficient\\nJF M A MJJAS O N D/uni25CF\\n/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF\\n/uni25CF\\n/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF\\n/uni25CF\\n/uni25CF\\n/uni25CF\\n/uni25CF\\n/uni25CF\\n/uni25CF\\n51 0 1 5 2 0−100 0 50 100 200\\nHourCoefficient', metadata={'source': './documents/ISLP_website.pdf', 'page': 176}), Document(page_content='4.6 Generalized Linear Models 169\\n5 10 15 200 100 200 300 400 500 600HourNumber of Bikers5 10 15 200123456HourLog(Number of Bikers)FIGURE 4.14. Left:On theBikeshare dataset, the number of bikers is dis-\\nplayed on the y-axis, and the hour of the day is displayed on the x-axis. Jitter\\nwas applied for ease of visualization. For the most part, as the mean number of\\nbikers increases, so does the variance in the number of bikers. A smoothing spline\\nfit is shown in green. Right:The log of the number of bikers is now displayed on\\nthey-axis.\\nSome of the problems that arise when fitting a linear regression model\\nto theBikeshare data can be overcome by transforming the response; for\\ninstance, we can fit the model\\nlog(Y)=p∑\\nj=1Xjβj+ϵ.\\nTransforming the response avoids the possibility of negative predictions,\\nand it overcomes much of the heteroscedasticity in the untransformed data,\\nas is shown in the right-hand panel of Figure 4.14. However, it is not quite\\na satisfactory solution, since predictions and inference are made in terms of\\nthe log of the response, rather than the response. This leads to challenges\\nin interpretation, e.g. “a one-unit increase in Xjis associated with an\\nincrease in the mean of the log of Yby an amount βj”. Furthermore, a\\nlog transformation of the response cannot be applied in settings where the\\nresponse can take on a value of 0. Thus, while fitting a linear model to\\na transformation of the response may be an adequate approach for some\\ncount-valued data sets, it often leaves something to be desired. We will see\\nin the next section that a Poisson regression model provides a much more\\nnatural and elegant approach for this task.\\n4.6.2 Poisson Regression on the Bikeshare Data\\nToovercometheinadequaciesoflinearregressionforanalyzingthe Bikeshare\\ndata set, we will make use of an alternative approach, called Poisson\\nregression . Before we can talk about Poisson regression, we must first in-Poisson\\nregressiontroduce the Poisson distribution .\\nPoisson\\ndistributionSuppose that a random variable Ytakes on nonnegative integer values,\\ni.e.Y∈{0,1,2,...}. IfYfollows the Poisson distribution, then\\nPr(Y=k)=e−λλk\\nk!fork=0,1,2,.... (4.35)', metadata={'source': './documents/ISLP_website.pdf', 'page': 177}), Document(page_content='170 4. Classification\\nHere,λ>0is the expected value of Y, i.e.E(Y). It turns out that λalso\\nequals the variance of Y, i.e.λ= E( Y) = Var( Y). This means that if Y\\nfollows the Poisson distribution, then the larger the mean of Y, the larger\\nits variance. (In ( 4.35), the notation k!, pronounced “k factorial”, is defined\\nask!=k×(k−1)×(k−2)×...×3×2×1.)\\nThe Poisson distribution is typically used to model counts; this is a nat-\\nural choice for a number of reasons, including the fact that counts, like\\nthe Poisson distribution, take on nonnegative integer values. To see how\\nwe might use the Poisson distribution in practice, let Ydenote the num-\\nber of users of the bike sharing program during a particular hour of the\\nday, under a particular set of weather conditions, and during a particu-\\nlar month of the year. We might model Yas a Poisson distribution with\\nmeanE(Y)= λ=5. This means that the probability of no users dur-\\ning this particular hour is Pr(Y= 0) =e−550\\n0!=e−5=0.0067 (where\\n0! = 1 by convention). The probability that there is exactly one user\\nisPr(Y= 1) =e−551\\n1!=5e−5=0.034, the probability of two users is\\nPr(Y= 2) =e−552\\n2!=0.084, and so on.\\nOf course, in reality, we expect the mean number of users of the bike\\nsharing program, λ= E( Y), to vary as a function of the hour of the day,\\nthe month of the year, the weather conditions, and so forth. So rather\\nthan modeling the number of bikers, Y, as a Poisson distribution with a\\nfixed mean value like λ=5, we would like to allow the mean to vary as a\\nfunction of the covariates. In particular, we consider the following model\\nfor the mean λ= E( Y), which we now write as λ(X1,...,X p)to emphasize\\nthat it is a function of the covariates X1,...,X p:\\nlog(λ(X1,...,X p)) = β0+β1X1+···+βpXp (4.36)\\nor equivalently\\nλ(X1,...,X p)=eβ0+β1X1+···+βpXp. (4.37)\\nHere,β0,β1,...,βpare parameters to be estimated. Together, ( 4.35) and\\n(4.36) define the Poisson regression model. Notice that in ( 4.36), we take\\nthelogofλ(X1,...,X p)to be linear in X1,...,X p, rather than having\\nλ(X1,...,X p)itselfbelinearin X1,...,X p;thisensuresthat λ(X1,...,X p)\\ntakes on nonnegative values for all values of the covariates.\\nTo estimate the coefficients β0,β1,...,βp, we use the same maximum\\nlikelihood approach that we adopted for logistic regression in Section 4.3.2.\\nSpecifically, given nindependent observations from the Poisson regression\\nmodel, the likelihood takes the form\\nℓ(β0,β1,...,βp)=n∏\\ni=1e−λ(xi)λ(xi)yi\\nyi!, (4.38)\\nwhereλ(xi)= eβ0+β1xi1+···+βpxip, due to ( 4.37). We estimate the coef-\\nficients that maximize the likelihood ℓ(β0,β1,...,βp), i.e. that make the\\nobserved data as likely as possible.\\nWe now fit a Poisson regression model to the Bikeshare data set. The\\nresults are shown in Table 4.11and Figure 4.15. Qualitatively, the results\\nare similar to those from linear regression in Section 4.6.1. We again see\\nthat bike usage is highest in the spring and fall and during rush hour,', metadata={'source': './documents/ISLP_website.pdf', 'page': 178}), Document(page_content='4.6 Generalized Linear Models 171\\nCoefficient Std. error z-statistic p-value\\nIntercept 4.12 0.01 683.96 0.00\\nworkingday 0.01 0.00 7.5 0.00\\ntemp 0.79 0.01 68.43 0.00\\nweathersit[cloudy/misty] -0.08 0.00 -34.53 0.00\\nweathersit[light rain/snow] -0.58 0.00 -141.91 0.00\\nweathersit[heavy rain/snow] -0.93 0.17 -5.55 0.00\\nTABLE 4.11. Results for a Poisson regression model fit to predict bikersin\\ntheBikeshare data. The predictors mnthandhrare omitted from this table due\\nto space constraints, and can be seen in Figure 4.15. For the qualitative variable\\nweathersit , the baseline corresponds to clear skies.\\nFIGURE 4.15. A Poisson regression model was fit to predict bikersin the\\nBikeshare data set. Left:The coefficients associated with the month of the year.\\nBike usage is highest in the spring and fall, and lowest in the winter. Right:The\\ncoefficients associated with the hour of the day. Bike usage is highest during peak\\ncommute times, and lowest overnight.\\nand lowest during the winter and in the early morning hours. Moreover,\\nbike usage increases as the temperature increases, and decreases as the\\nweather worsens. Interestingly, the coefficient associated with workingday\\nisstatisticallysignificantunderthePoissonregressionmodel,butnotunder\\nthe linear regression model.\\nSome important distinctions between the Poisson regression model and\\nthe linear regression model are as follows:\\n•Interpretation: To interpret the coefficients in the Poisson regression\\nmodel, we must pay close attention to ( 4.37), which states that an\\nincrease in Xjby one unit is associated with a change in E(Y)=λ\\nby a factor of exp(βj). For example, a change in weather from clear\\nto cloudy skies is associated with a change in mean bike usage by a\\nfactor of exp(−0.08) = 0 .923, i.e. on average, only 92.3% as many\\npeople will use bikes when it is cloudy relative to when it is clear.\\nIf the weather worsens further and it begins to rain, then the mean\\nbike usage will further change by a factor of exp(−0.5) = 0 .607, i.e.\\non average only 60.7% as many people will use bikes when it is rainy\\nrelative to when it is cloudy./uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF\\n/uni25CF/uni25CF/uni25CF/uni25CF\\n/uni25CF\\n/uni25CF−0.6−0.4−0.2 0.0 0.2\\nMonthCoefficient\\nJF M A MJJAS O N D/uni25CF\\n/uni25CF\\n/uni25CF\\n/uni25CF\\n/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF\\n/uni25CF\\n/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF\\n/uni25CF\\n/uni25CF\\n/uni25CF\\n/uni25CF\\n/uni25CF\\n51 0 1 5 2 0−2 −10 1\\nHourCoefficient', metadata={'source': './documents/ISLP_website.pdf', 'page': 179}), Document(page_content='172 4. Classification\\n•Mean-variance relationship: As mentioned earlier, under the Poisson\\nmodel,λ= E( Y) = Var( Y). Thus, by modeling bike usage with a\\nPoisson regression, we implicitly assume that mean bike usage in a\\ngiven hour equals the variance of bike usage during that hour. By\\ncontrast, under a linear regression model, the variance of bike usage\\nalways takes on a constant value. Recall from Figure 4.14that in the\\nBikeshare data, when biking conditions are favorable, both the mean\\nandthe variance in bike usage are much higher than when conditions\\nare unfavorable. Thus, the Poisson regression model is able to handle\\nthe mean-variance relationship seen in the Bikeshare data in a way\\nthat the linear regression model is not.5\\noverdispersion\\n•nonnegative fitted values: There are no negative predictions using the\\nPoisson regression model. This is because the Poisson model itself\\nonly allows for nonnegative values; see ( 4.35). By contrast, when we\\nfit a linear regression model to the Bikeshare data set, almost 10% of\\nthe predictions were negative.\\n4.6.3 Generalized Linear Models in Greater Generality\\nWe have now discussed three types of regression models: linear, logistic and\\nPoisson. These approaches share some common characteristics:\\n1. Each approach uses predictors X1,...,X pto predict a response Y.\\nWe assume that, conditional on X1,...,X p,Ybelongs to a certain\\nfamily of distributions. Forlinear regression, wetypicallyassume that\\nYfollows a Gaussian or normal distribution. For logistic regression,\\nweassumethat YfollowsaBernoullidistribution.Finally,forPoisson\\nregression, we assume that Yfollows a Poisson distribution.\\n2. Each approach models the mean of Yas a function of the predictors.\\nIn linear regression, the mean of Ytakes the form\\nE(Y|X1,...,X p)=β0+β1X1+···+βpXp, (4.39)\\ni.e. it is a linear function of the predictors. For logistic regression, the\\nmean instead takes the form\\nE(Y|X1,...,X p) = Pr( Y=1|X1,...,X p)\\n=eβ0+β1X1+···+βpXp\\n1+eβ0+β1X1+···+βpXp, (4.40)\\nwhile for Poisson regression it takes the form\\nE(Y|X1,...,X p)=λ(X1,...,X p)=eβ0+β1X1+···+βpXp.(4.41)\\nEquations ( 4.39)–(4.41) can be expressed using a link function ,η, whichlink function\\n5In fact, the variance in the Bikeshare data appears to be much higher than the\\nmean, a situation referred to as overdispersion . This causes the Z-values to be inflated\\nin Table 4.11. A more careful analysis should account for this overdispersion to obtain\\nmore accurate Z-values, and there are a variety of methods for doing this. But they are\\nbeyond the scope of this book.', metadata={'source': './documents/ISLP_website.pdf', 'page': 180}), Document(page_content='4.7 Lab: Logistic Regression, LDA, QDA, and KNN 173\\napplies a transformation to E(Y|X1,...,X p)so that the transformed mean\\nis a linear function of the predictors. That is,\\nη(E(Y|X1,...,X p)) = β0+β1X1+···+βpXp. (4.42)\\nThe link functions for linear, logistic and Poisson regression are η(µ)=µ,\\nη(µ) = log( µ/(1−µ)), andη(µ) = log( µ), respectively.\\nThe Gaussian, Bernoulli and Poisson distributions are all members of a\\nwider class of distributions, known as the exponential family . Other well-exponential\\nfamilyknown members of this family are the exponential distribution, the Gamma\\nexponential\\nGammadistribution, and the negative binomial distribution. In general, we can per-\\nnegative\\nbinomialform a regression by modeling the response Yas coming from a particular\\nmember of the exponential family, and then transforming the mean of the\\nresponse so that the transformed mean is a linear function of the predictors\\nvia (4.42). Any regression approach that follows this very general recipe is\\nknownas a generalized linear model (GLM). Thus,linear regression, logisticgeneralized\\nlinear modelregression, and Poisson regression are three examples of GLMs. Other ex-\\namples not covered here include Gamma regression andnegative binomial\\nregression .\\n4.7 Lab: Logistic Regression, LDA, QDA, and\\nKNN\\n4.7.1 The Stock Market Data\\nIn this lab we will examine the Smarket data, which is part of the ISLP\\nlibrary. This data set consists of percentage returns for the S&P 500 stock\\nindex over 1,250 days, from the beginning of 2001 until the end of 2005.\\nFor each date, we have recorded the percentage returns for each of the five\\nprevious trading days, Lag1through Lag5. We have also recorded Volume\\n(the number of shares traded on the previous day, in billions), Today(the\\npercentage return on the date in question) and Direction (whether the\\nmarket was UporDownon this date).\\nWe start by importing our libraries at this top level; these are all imports\\nwe have seen in previous labs.\\nIn[1]: import numpy as np\\nimport pandas as pd\\nfrom matplotlib.pyplot import subplots\\nimport statsmodels.api as sm\\nfrom ISLP import load_data\\nfrom ISLP.models import (ModelSpec as MS,\\nsummarize)\\nWe also collect together the new imports needed for this lab.\\nIn[2]: from ISLP import confusion_table\\nfrom ISLP.models import contrast\\nfrom sklearn.discriminant_analysis import \\\\\\n(LinearDiscriminantAnalysis as LDA,\\nQuadraticDiscriminantAnalysis as QDA)\\nfrom sklearn.naive_bayes import GaussianNB\\nfrom sklearn.neighbors import KNeighborsClassifier\\nfrom sklearn.preprocessing import StandardScaler', metadata={'source': './documents/ISLP_website.pdf', 'page': 181}), Document(page_content=\"174 4. Classificationfrom sklearn.model_selection import train_test_splitfrom sklearn.linear_model import LogisticRegressionNow we are ready to load theSmarketdata.In[3]:Smarket = load_data('Smarket')SmarketThis gives a truncated listing of the data, which we do not show here. Wecan see what the variable names are.In[4]:Smarket.columnsOut[4]:Index(['Year','Lag1','Lag2','Lag3','Lag4','Lag5','Volume','Today','Direction'],dtype='object')We compute the correlation matrix using thecorr()method for data.corr()frames, which produces a matrix that contains all of the pairwise corre-lations among the variables. (We suppress the output here.) Thepandaslibrary does not report a correlation for theDirectionvariable because itis qualitative.In[5]:Smarket.corr()As one would expect, the correlations between the lagged return variablesand today’s return are close to zero. The only substantial correlation isbetweenYearandVolume. By plotting the data we see thatVolumeis in-creasing over time. In other words, the average number of shares tradeddaily increased from 2001 to 2005.In[6]:Smarket.plot(y='Volume');4.7.2 Logistic RegressionNext, we will fit a logistic regression model in order to predictDirectionusingLag1throughLag5andVolume. Thesm.GLM()function fitsgener-sm.GLM()alized linear models, a class of models that includes logistic regression.generalizedlinear modelAlternatively, the functionsm.Logit()fits a logistic regression model di-rectly. The syntax ofsm.GLM()is similar to that ofsm.OLS(), except thatwe must pass in the argumentfamily=sm.families.Binomial()in order totellstatsmodelsto run a logistic regression rather than some other type ofgeneralized linear model.In[7]:allvars = Smarket.columns.drop(['Today','Direction','Year'])design = MS(allvars)X=d e s i g n . f i t _ t r a n s f o r m ( S m a r k e t )y=S m a r k e t . D i r e c t i o n= ='Up'glm = sm.GLM(y,X,family=sm.families.Binomial())results = glm.fit()summarize(results)\", metadata={'source': './documents/ISLP_website.pdf', 'page': 182}), Document(page_content='4.7 Lab: Logistic Regression, LDA, QDA, and KNN 175Out[7]:coef std err z P>|z|intercept -0.1260 0.241 -0.523 0.601Lag1 -0.0731 0.050 -1.457 0.145Lag2 -0.0423 0.050 -0.845 0.398Lag3 0.0111 0.050 0.222 0.824Lag4 0.0094 0.050 0.187 0.851Lag5 0.0103 0.050 0.208 0.835Volume 0.1354 0.158 0.855 0.392The smallestp-value here is associated withLag1. The negative coefficientforthispredictorsuggeststhatifthemarkethadapositivereturnyesterday,then it is less likely to go up today. However, at a value of 0.15, thep-valueis still relatively large, and so there is no clear evidence of a real associationbetweenLag1andDirection.We use theparamsattribute ofresultsin order to access just the coeffi-cients for this fitted model.In[8]:results.paramsOut[8]:intercept -0.126000Lag1 -0.073074Lag2 -0.042301Lag30.011085Lag40.009359Lag50.010313Volume 0.135441dtype: float64Likewise we can use thepvaluesattribute to access thep-values for thecoefficients (not shown).In[9]:results.pvaluesThepredict()method ofresultscan be used to predict the probabilitythat the market will go up, given values of the predictors. This methodreturns predictions on the probability scale. If no data set is supplied tothepredict()function, then the probabilities are computed for the train-ing data that was used to fit the logistic regression model. As with linearregression, one can pass an optionalexogargument consistent with a designmatrix if desired. Here we have printed only the first ten probabilities.In[10]:probs = results.predict()probs[:10]Out[10]:array([0.5070841, 0.4814679, 0.4811388, 0.5152223, 0.5107812,0.5069565, 0.4926509, 0.5092292, 0.5176135, 0.4888378])In order to make a prediction as to whether the market will go up ordown on a particular day, we must convert these predicted probabilitiesinto class labels,UporDown. The following two commands create a vectorof class predictions based on whether the predicted probability of a marketincrease is greater than or less than 0.5.In[11]:labels = np.array([\\'Down\\']*1250)labels[probs>0.5] =\"Up\"', metadata={'source': './documents/ISLP_website.pdf', 'page': 183}), Document(page_content='176 4. ClassificationTheconfusion_table()function from theISLPpackage summarizes theseconfusion_table()predictions, showing how many observations were correctly or incorrectlyclassified. Our function, which is adapted from a similar function in themodulesklearn.metrics, transposes the resulting matrix and includes rowand column labels. Theconfusion_table()function takes as first argumentthe predicted labels, and second argument the true labels.In[12]:confusion_table(labels, Smarket.Direction)Out[12]:Truth Down UpPredictedDown 145 141Up 457 507The diagonal elements of the confusion matrix indicate correct predictions,while the off-diagonals represent incorrect predictions. Hence our modelcorrectly predicted that the market would go up on 507 days and thatit would go down on 145 days, for a total of 507 + 145 = 652 correctpredictions. Thenp.mean()function can be used to compute the fraction ofdays for which the prediction was correct. In this case, logistic regressioncorrectly predicted the movement of the market 52.2% of the time.In[13]:(507+145)/1250, np.mean(labels == Smarket.Direction)Out[13]:(0.5216, 0.5216)At first glance, it appears that the logistic regression model is workinga little better than random guessing. However, this result is misleadingbecause we trained and tested the model on the same set of 1,250 obser-vations. In other words,100−52.2 = 47.8%is thetrainingerror rate. Aswe have seen previously, the training error rate is often overly optimistic— it tends to underestimate the test error rate. In order to better assessthe accuracy of the logistic regression model in this setting, we can fit themodel using part of the data, and then examine how well it predicts theheld outdata. This will yield a more realistic error rate, in the sense that inpractice we will be interested in our model’s performance not on the datathat we used to fit the model, but rather on days in the future for whichthe market’s movements are unknown.To implement this strategy, we first create a Boolean vector correspond-ing to the observations from 2001 through 2004. We then use this vectorto create a held out data set of observations from 2005.In[14]:train = (Smarket.Year < 2005)Smarket_train = Smarket.loc[train]Smarket_test = Smarket.loc[∼train]Smarket_test.shapeOut[14]:(252, 9)The objecttrainis a vector of 1,250 elements, corresponding to theobservations in our data set. The elements of the vector that correspondto observations that occurred before 2005 are set toTrue, whereas thosethat correspond to observations in 2005 are set toFalse. Hencetrainis abooleanarray, since its elements areTrueandFalse. Boolean arrays can beused to obtain a subset of the rows or columns of a data frame using the', metadata={'source': './documents/ISLP_website.pdf', 'page': 184}), Document(page_content=\"4.7 Lab: Logistic Regression, LDA, QDA, and KNN 177locmethod. For instance, the commandSmarket.loc[train]would pickout a submatrix of the stock market data set, corresponding only to thedates before 2005, since those are the ones for which the elements oftrainareTrue. The∼symbol can be used to negate all of the elements of aBoolean vector. That is,∼trainis a vector similar totrain, except thatthe elements that areTrueintrainget swapped toFalsein∼train, andvice versa. Therefore,Smarket.loc[∼train]yields a subset of the rows ofthe data frame of the stock market data containing only the observationsfor whichtrainisFalse. The output above indicates that there are 252such observations.We now fit a logistic regression model using only the subset of the ob-servations that correspond to dates before 2005. We then obtain predictedprobabilities of the stock market going up for each of the days in our testset — that is, for the days in 2005.In[15]:X_train, X_test = X.loc[train], X.loc[∼train]y_train, y_test = y.loc[train], y.loc[∼train]glm_train = sm.GLM(y_train,X_train,family=sm.families.Binomial())results = glm_train.fit()probs = results.predict(exog=X_test)Notice that we have trained and tested our model on two completelyseparate data sets: training was performed using only the dates before2005, and testing was performed using only the dates in 2005.Finally, we compare the predictions for 2005 to the actual movements ofthe market over that time period. We will first store the test and traininglabels (recally_testis binary).In[16]:D=S m a r k e t . D i r e c t i o nL_train, L_test = D.loc[train], D.loc[∼train]Nowwethresholdthefittedprobabilityat50%toformourpredictedlabels.In[17]:labels = np.array(['Down']*252)labels[probs>0.5] ='Up'confusion_table(labels, L_test)Out[17]:Truth Down UpPredictedDown 77 97Up 34 44The test accuracy is about 48% while the error rate is about 52%In[18]:np.mean(labels == L_test), np.mean(labels != L_test)Out[18]:(0.4802, 0.5198)The!=notation meansnot equal to, and so the last command computes!=the test set error rate. The results are rather disappointing: the test errorrate is 52%, which is worse than random guessing! Of course this resultis not all that surprising, given that one would not generally expect to beable to use previous days’ returns to predict future market performance.(After all, if it were possible to do so, then the authors of this book wouldbe out striking it rich rather than writing a statistics textbook.)\", metadata={'source': './documents/ISLP_website.pdf', 'page': 185}), Document(page_content=\"178 4. ClassificationWe recall that the logistic regression model had very underwhelmingp-values associated with all of the predictors, and that the smallestp-value,though not very small, corresponded toLag1. Perhaps by removing thevariables that appear not to be helpful in predictingDirection, we canobtain a more effective model. After all, using predictors that have norelationship with the response tends to cause a deterioration in the testerror rate (since such predictors cause an increase in variance without acorresponding decrease in bias), and so removing such predictors may inturn yield an improvement. Below we refit the logistic regression using justLag1andLag2, which seemed to have the highest predictive power in theoriginal logistic regression model.In[19]:model = MS(['Lag1','Lag2']).fit(Smarket)X=m o d e l . t r a n s f o r m ( S m a r k e t )X_train, X_test = X.loc[train], X.loc[∼train]glm_train = sm.GLM(y_train,X_train,family=sm.families.Binomial())results = glm_train.fit()probs = results.predict(exog=X_test)labels = np.array(['Down']*252)labels[probs>0.5] ='Up'confusion_table(labels, L_test)Out[19]:Truth Down UpPredictedDown 35 35Up 76 106Let’s evaluate the overall accuracy as well as the accuracy within thedays when logistic regression predicts an increase.In[20]:(35+106)/252,106/(106+76)Out[20]:(0.5595, 0.5824)Now the results appear to be a little better: 56% of the daily movementshave been correctly predicted. It is worth noting that in this case, a muchsimpler strategy of predicting that the market will increase every day willalso be correct 56% of the time! Hence, in terms of overall error rate, thelogistic regression method is no better than the naive approach. However,the confusion matrix shows that on days when logistic regression predictsan increase in the market, it has a 58% accuracy rate. This suggests apossible trading strategy of buying on days when the model predicts an in-creasing market, and avoiding trades on days when a decrease is predicted.Of course one would need to investigate more carefully whether this smallimprovement was real or just due to random chance.Suppose that we want to predict the returns associated with particularvalues ofLag1andLag2. In particular, we want to predictDirectionon aday whenLag1andLag2equal1.2and1.1, respectively, and on a day whenthey equal1.5and−0.8. We do this using thepredict()function.In[21]:newdata = pd.DataFrame({'Lag1':[1.2, 1.5],'Lag2':[1.1, -0.8]});\", metadata={'source': './documents/ISLP_website.pdf', 'page': 186}), Document(page_content=\"4.7 Lab: Logistic Regression, LDA, QDA, and KNN 179\\nnewX = model.transform(newdata)\\nresults.predict(newX)\\nOut[21]: 00 . 4 7 9 1\\n10 . 4 9 6 1\\ndtype: float64\\n4.7.3 Linear Discriminant Analysis\\nWe begin by performing LDA on the Smarket data, using the function\\nLinearDiscriminantAnalysis() , which we have abbreviated LDA(). We fitLinear\\nDiscriminant\\nAnalysis()the model using only the observations before 2005.\\nIn[22]: lda = LDA(store_covariance=True)\\nSince the LDAestimator automatically adds an intercept, we should re-\\nmovethecolumncorrespondingtotheinterceptinboth X_trainandX_test.\\nWe can also directly use the labels rather than the Boolean vectors y_train.\\nIn[23]: X_train, X_test = [M.drop(columns=[ 'intercept' ])\\nfor M in [X_train, X_test]]\\nlda.fit(X_train, L_train)\\nOut[23]: LinearDiscriminantAnalysis(store_covariance=True)\\nHere we have used the list comprehensions introduced in Section 3.6.4.\\nLooking at our first line above, we see that the right-hand side is a list of\\nlength two. This is because the code for M in [X_train, X_test] iterates\\nover a list of length two. While here we loop over a list, the list comprehen-\\nsion method works when looping over any iterable object. We then apply\\nthedrop()method to each element in the iteration, collecting the result.drop()in a list. The left-hand side tells Pythonto unpack this list of length two,\\nassigning its elements to the variables X_trainandX_test. Of course, this\\noverwrites the previous values of X_trainandX_test.\\nHaving fit the model, we can extract the means in the two classes with\\nthemeans_attribute. These are the average of each predictor within each\\nclass, and are used by LDA as estimates of µk. These suggest that there is\\na tendency for the previous 2 days’ returns to be negative on days when\\nthe market increases, and a tendency for the previous days’ returns to be\\npositive on days when the market declines.\\nIn[24]: lda.means_\\nOut[24]: array([[ 0.04, 0.03],\\n[-0.04, -0.03]])\\nTheestimated prior probabilities are stored in the priors_attribute.The\\npackage sklearntypicallyusesthistrailing _todenoteaquantityestimated\\nwhen using the fit()method. We can be sure of which entry corresponds\\nto which label by looking at the classes_ attribute.\\nIn[25]: lda.classes_\\nOut[25]: array([ 'Down','Up'], dtype= '<U4')\", metadata={'source': './documents/ISLP_website.pdf', 'page': 187}), Document(page_content=\"180 4. Classification\\nThe LDA output indicates that ˆπDown=0.492andˆπUp=0.508.\\nIn[26]: lda.priors_\\nOut[26]: array([0.492, 0.508])\\nThe linear discriminant vectors can be found in the scalings_ attribute:\\nIn[27]: lda.scalings_\\nOut[27]: array([[-0.642],\\n[-0.513]])\\nThese values provide the linear combination of Lag1andLag2that are used\\nto form the LDA decision rule. In other words, these are the multipliers of\\nthe elements of X=xin (4.24). If−0.64×Lag1−0.51×Lag2is large,\\nthen the LDA classifier will predict a market increase, and if it is small,\\nthen the LDA classifier will predict a market decline.\\nIn[28]: lda_pred = lda.predict(X_test)\\nAs we observed in our comparison of classification methods (Section 4.5),\\nthe LDA and logistic regression predictions are almost identical.\\nIn[29]: confusion_table(lda_pred, L_test)\\nOut[29]: Truth Down Up\\nPredicted\\nDown 35 35\\nUp 76 106\\nWe can also estimate the probability of each class for each point in a\\ntraining set. Applying a 50% threshold to the posterior probabilities of be-\\ning in class one allows us to recreate the predictions contained in lda_pred .\\nIn[30]: lda_prob = lda.predict_proba(X_test)\\nnp.all(\\nnp.where(lda_prob[:,1] >= 0.5, 'Up','Down')= =l d a _ p r e d\\n)\\nOut[30]: True\\nAbove, we used the np.where() function that creates an array with valuenp.where()'Up'for indices where the second column of lda_prob (the estimated poste-\\nrior probability of 'Up') is greater than 0.5. For problems with more than\\ntwo classes the labels are chosen as the class whose posterior probability is\\nhighest:\\nIn[31]: np.all(\\n[lda.classes_[i] for i in np.argmax(lda_prob, 1)] ==\\nlda_pred\\n)\\nOut[31]: True\\nIf we wanted to use a posterior probability threshold other than 50% in\\norder to make predictions, then we could easily do so. For instance, suppose\\nthatwewishtopredictamarketdecreaseonlyifweareverycertainthatthe\", metadata={'source': './documents/ISLP_website.pdf', 'page': 188}), Document(page_content='4.7 Lab: Logistic Regression, LDA, QDA, and KNN 181\\nmarket will indeed decrease on that day — say, if the posterior probability\\nis at least 90%. We know that the first column of lda_prob corresponds to\\nthe label Downafter having checked the classes_ attribute, hence we use\\nthe column index 0 rather than 1 as we did above.\\nIn[32]: np.sum(lda_prob[:,0] > 0.9)\\nOut[32]: 0\\nNo days in 2005 meet that threshold! In fact, the greatest posterior prob-\\nability of decrease in all of 2005 was 52.02%.\\nTheLDAclassifieraboveisthefirstclassifierfromthe sklearnlibrary.We\\nwilluseseveralotherobjectsfromthislibrary.Theobjectsfollowacommon\\nstructure that simplifies tasks such as cross-validation, which we will see in\\nChapter 5. Specifically, the methods first create a generic classifier without\\nreferring to any data. This classifier is then fit to data with the fit()\\nmethod and predictions are always produced with the predict() method.\\nThis pattern of first instantiating the classifier, followed by fitting it, and\\nthen producing predictions is an explicit design choice of sklearn. This\\nuniformity makes it possible to cleanly copy the classifier so that it can be\\nfit on different data; e.g. different training sets arising in cross-validation.\\nThis standard pattern also allows for a predictable formation of workflows.\\n4.7.4 Quadratic Discriminant Analysis\\nWe will now fit a QDA model to the Smarketdata. QDA is implemented\\nviaQuadraticDiscriminantAnalysis() in thesklearnpackage, which we ab-Quadratic\\nDiscriminant\\nAnalysis()breviate to QDA(). The syntax is very similar to LDA().\\nIn[33]: qda = QDA(store_covariance=True)\\nqda.fit(X_train, L_train)\\nOut[33]: QuadraticDiscriminantAnalysis(store_covariance=True)\\nTheQDA()function will again compute means_andpriors_.\\nIn[34]: qda.means_, qda.priors_\\nOut[34]: (array([[ 0.04279022, 0.03389409],\\n[-0.03954635, -0.03132544]]),\\narray([0.49198397, 0.50801603]))\\nTheQDA()classifier will estimate one covariance per class. Here is the\\nestimated covariance in the first class:\\nIn[35]: qda.covariance_[0]\\nOut[35]: array([[ 1.50662277, -0.03924806],\\n[-0.03924806, 1.53559498]])\\nThe output contains the group means. But it does not contain the coef-\\nficients of the linear discriminants, because the QDA classifier involves a\\nquadratic, rather than a linear, function of the predictors. The predict()\\nfunction works in exactly the same fashion as for LDA.', metadata={'source': './documents/ISLP_website.pdf', 'page': 189}), Document(page_content=\"182 4. ClassificationIn[36]:qda_pred = qda.predict(X_test)confusion_table(qda_pred, L_test)Out[36]:Truth Down UpPredictedDown 30 20Up 81 121Interestingly, the QDA predictions are accurate almost 60% of the time,even though the 2005 data was not used to fit the model.In[37]:np.mean(qda_pred == L_test)Out[37]:0.599This level of accuracy is quite impressive for stock market data, whichis known to be quite hard to model accurately. This suggests that thequadratic form assumed by QDA may capture the true relationship moreaccurately than the linear forms assumed by LDA and logistic regression.However, we recommend evaluating this method’s performance on a largertest set before betting that this approach will consistently beat the market!4.7.5 Naive BayesNext, we fit a naive Bayes model to theSmarketdata. The syntax is similarto that ofLDA()andQDA(). By default, this implementationGaussianNB()ofGaussianNB()the naive Bayes classifier models each quantitative feature using a Gaussiandistribution. However, a kernel density method can also be used to estimatethe distributions.In[38]:NB = GaussianNB()NB.fit(X_train, L_train)Out[38]:GaussianNB()The classes are stored asclasses_.In[39]:NB.classes_Out[39]:array(['Down','Up'], dtype='<U4')The class prior probabilities are stored in theclass_prior_attribute.In[40]:NB.class_prior_Out[40]:array([0.49, 0.51])The parameters of the features can be found in thetheta_andvar_attributes. The number of rows is equal to the number of classes, while thenumber of columns is equal to the number of features. We see below thatthe mean for featureLag1in theDownclass is 0.043.In[41]:NB.theta_Out[41]:array([[ 0.043, 0.034],[-0.040, -0.031]])\", metadata={'source': './documents/ISLP_website.pdf', 'page': 190}), Document(page_content=\"4.7 Lab: Logistic Regression, LDA, QDA, and KNN 183\\nIts variance is 1.503.\\nIn[42]: NB.var_\\nOut[42]: array([[1.503, 1.532],\\n[1.514, 1.487]])\\nHow do we know the names of these attributes? We use NB?(or?NB).\\nWe can easily verify the mean computation:\\nIn[43]: X_train[L_train == 'Down'].mean()\\nOut[43]: Lag1 0.042790\\nLag2 0.033894\\ndtype: float64\\nSimilarly for the variance:\\nIn[44]: X_train[L_train == 'Down'].var(ddof=0)\\nOut[44]: Lag1 1.503554\\nLag2 1.532467\\ndtype: float64\\nTheGaussianNB() function calculates variances using the 1/nformula.6\\nSinceNB()is a classifier in the sklearnlibrary, making predictions uses the\\nsame syntax as for LDA()andQDA()above.\\nIn[45]: nb_labels = NB.predict(X_test)\\nconfusion_table(nb_labels, L_test)\\nOut[45]: Truth Down Up\\nPredicted\\nDown 29 20\\nUp 82 121\\nNaive Bayes performs well on these data, with accurate predictions over\\n59% of the time. This is slightly worse than QDA, but much better than\\nLDA.\\nAs forLDA, thepredict_proba() method estimates the probability that\\neach observation belongs to a particular class.\\nIn[46]: NB.predict_proba(X_test)[:5]\\nOut[46]: array([[0.4873, 0.5127],\\n[0.4762, 0.5238],\\n[0.4653, 0.5347],\\n[0.4748, 0.5252],\\n[0.4902, 0.5098]])\\n4.7.6 K-Nearest Neighbors\\nWewillnowperformKNNusingthe KNeighborsClassifier() function.ThisKNeighbors\\nClassifier()\\n6There are two formulas for computing the sample variance of nobservations\\nx1,...,x n:1\\nn∑n\\ni=1(xi−¯x)2and1\\nn−1∑n\\ni=1(xi−¯x)2where¯xis the sample mean.\\nIn most cases the distinction is not important.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 191}), Document(page_content=\"184 4. Classificationfunction works similarly to the other model-fitting functions that we haveencountered thus far.As is the case for LDA and QDA, we fit the classifier using thefitmethod.Newpredictionsareformedusingthepredictmethodoftheobjectreturned byfit().In[47]:knn1 = KNeighborsClassifier(n_neighbors=1)knn1.fit(X_train, L_train)knn1_pred = knn1.predict(X_test)confusion_table(knn1_pred, L_test)Out[47]:Truth Down UpPredictedDown 43 58Up 68 83The results usingK=1are not very good, since only50%of the observa-tions are correctly predicted. Of course, it may be thatK=1results in anoverly-flexible fit to the data.In[48]:(83+43)/252, np.mean(knn1_pred == L_test)Out[48]:(0.5, 0.5)We repeat the analysis below usingK=3.In[49]:knn3 = KNeighborsClassifier(n_neighbors=3)knn3_pred = knn3.fit(X_train, L_train).predict(X_test)np.mean(knn3_pred == L_test)Out[49]:0.532The results have improved slightly. But increasingKfurther provides nofurther improvements. It appears that for these data, and this train/testsplit, QDA gives the best results of the methods that we have examined sofar.KNNdoesnotperformwellontheSmarketdata,butitoftendoesprovideimpressive results. As an example we will apply the KNN approach to theCaravandata set, which is part of theISLPlibrary. This data set includes 85predictors that measure demographic characteristics for 5,822 individuals.The response variable isPurchase, which indicates whether or not a givenindividual purchases a caravan insurance policy. In this data set, only 6%of people purchased caravan insurance.In[50]:Caravan = load_data('Caravan')Purchase = Caravan.PurchasePurchase.value_counts()Out[50]:No 5474Yes 348Name: Purchase, dtype: int64The methodvalue_counts()takes apd.Seriesorpd.DataFrameand re-turns apd.Serieswith the corresponding counts for each unique element.In this casePurchasehas onlyYesandNovalues and returns how manyvalues of each there are.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 192}), Document(page_content=\"4.7 Lab: Logistic Regression, LDA, QDA, and KNN 185In[51]:348 / 5822Out[51]:0.0598Our features will include all columns exceptPurchase.In[52]:feature_df = Caravan.drop(columns=['Purchase'])Because the KNN classifier predicts the class of a given test observa-tion by identifying the observations that are nearest to it, the scale ofthe variables matters. Any variables that are on a large scale will have amuch larger effect on thedistancebetween the observations, and hence onthe KNN classifier, than variables that are on a small scale. For instance,imagine a data set that contains two variables,salaryandage(measuredin dollars and years, respectively). As far as KNN is concerned, a differenceof 1,000 USD in salary is enormous compared to a difference of 50 years inage. Consequently,salarywill drive the KNN classification results, andagewill have almost no effect. This is contrary to our intuition that a salarydifference of 1,000 USD is quite small compared to an age difference of 50years. Furthermore, the importance of scale to the KNN classifier leads toanother issue: if we measuredsalaryin Japanese yen, or if we measuredagein minutes, then we’d get quite different classification results from what weget if these two variables are measured in dollars and years.A good way to handle this problem is tostandardizethe data so that allstandardizevariables are given a mean of zero and a standard deviation of one. Thenall variables will be on a comparable scale. This is accomplished using theStandardScaler()transformation.StandardScaler()In[53]:scaler = StandardScaler(with_mean=True,with_std=True,copy=True)The argumentwith_meanindicates whether or not we should subtract themean, whilewith_stdindicates whether or not we should scale the columnsto have standard deviation of 1 or not. Finally, the argumentcopy=Truein-dicates that we will always copy data, rather than trying to do calculationsin place where possible.This transformation can be fit and then applied to arbitrary data. In thefirst line below, the parameters for the scaling are computed and storedinscaler, while the second line actually constructs the standardized set offeatures.In[54]:scaler.fit(feature_df)X_std = scaler.transform(feature_df)Now every column offeature_stdbelow has a standard deviation of oneand a mean of zero.In[55]:feature_std = pd.DataFrame(X_std,columns=feature_df.columns);feature_std.std()Out[55]:MOSTYPE 1.000086MAANTHUI 1.000086\", metadata={'source': './documents/ISLP_website.pdf', 'page': 193}), Document(page_content='186 4. Classification\\nMGEMOMV 1.000086\\nMGEMLEEF 1.000086\\nMOSHOOFD 1.000086\\n...\\nAZEILPL 1.000086\\nAPLEZIER 1.000086\\nAFIETS 1.000086\\nAINBOED 1.000086\\nABYSTAND 1.000086\\nLength: 85, dtype: float64\\nNotice that the standard deviations are not quite 1here; this is again due\\nto some procedures using the 1/nconvention for variances (in this case\\nscaler() ), while others use 1/(n−1)(thestd()method). See the footnote.std()on page183. In this case it does not matter, as long as the variables are all\\non the same scale.\\nUsingthefunction train_test_split() wenowsplittheobservationsintotrain_test_\\nsplit() a test set, containing 1000 observations, and a training set containing the\\nremaining observations. The argument random_state=0 ensures that we get\\nthe same split each time we rerun the code.\\nIn[56]: (X_train,\\nX_test,\\ny_train,\\ny_test) = train_test_split(feature_std,\\nPurchase,\\ntest_size=1000,\\nrandom_state=0)\\n?train_test_split reveals that the non-keyword arguments can be lists,\\narrays,pandas dataframes etc that all have the same length ( shape[0] ) and\\nhence are indexable . In this case they are the dataframe feature_std andindexablethe response variable Purchase . We fit a KNN model on the training data\\nusingK=1, and evaluate its performance on the test data.\\nIn[57]: knn1 = KNeighborsClassifier(n_neighbors=1)\\nknn1_pred = knn1.fit(X_train, y_train).predict(X_test)\\nnp.mean(y_test != knn1_pred), np.mean(y_test != \"No\")\\nOut[57]: (0.111, 0.067)\\nThe KNN error rate on the 1,000 test observations is about 11%. At first\\nglance, this may appear to be fairly good. However, since just over 6% of\\ncustomers purchased insurance, we could get the error rate down to almost\\n6% by always predicting Noregardless of the values of the predictors! This\\nis known as the null rate.null rateSuppose that there is some non-trivial cost to trying to sell insurance\\nto a given individual. For instance, perhaps a salesperson must visit each\\npotential customer. If the company tries to sell insurance to a random\\nselection of customers, then the success rate will be only 6%, which may\\nbe far too low given the costs involved. Instead, the company would like\\nto try to sell insurance only to customers who are likely to buy it. So the\\noverall error rate is not of interest. Instead, the fraction of individuals that\\nare correctly predicted to buy insurance is of interest.\\nIn[58]: confusion_table(knn1_pred, y_test)', metadata={'source': './documents/ISLP_website.pdf', 'page': 194}), Document(page_content=\"4.7 Lab: Logistic Regression, LDA, QDA, and KNN 187\\nOut[58]: Truth No Yes\\nPredicted\\nNo 880 58\\nYes 53 9\\nIt turns out that KNN with K=1does far better than random guessing\\namong the customers that are predicted to buy insurance. Among 62 such\\ncustomers, 9, or 14.5%, actually do purchase insurance. This is double the\\nrate that one would obtain from random guessing.\\nIn[59]: 9/(53+9)\\nOut[59]: 0.145\\nTuning Parameters\\nThe number of neighbors in KNN is referred to as a tuning parameter , alsotuning\\nparameterreferred to as a hyperparameter . We do not know a priori what value to\\nhyper-\\nparameteruse. It is therefore of interest to see how the classifier performs on test\\ndata as we vary these parameters. This can be achieved with a forloop,\\ndescribed in Section 2.3.8. Here we use a for loop to look at the accuracy of\\nour classifier in the group predicted to purchase insurance as we vary the\\nnumber of neighbors from 1 to 5:\\nIn[60]: for K in range(1,6):\\nknn = KNeighborsClassifier(n_neighbors=K)\\nknn_pred = knn.fit(X_train, y_train).predict(X_test)\\nC=c o n f u s i o n _ t a b l e ( k n n _ p r e d ,y _ t e s t )\\ntempl = ( 'K={0:d}: # predicted to rent: {1:>2},' +\\n'# w h o d i d r e n t { 2 : d } , a c c u r a c y { 3 : . 1 % } ' )\\npred = C.loc[ 'Yes'].sum()\\ndid_rent = C.loc[ 'Yes','Yes']\\nprint(templ.format(\\nK,\\npred,\\ndid_rent,\\ndid_rent / pred))\\nK=1:#p r e d i c t e dt or e n t :6 2 , #w h od i dr e n t9 ,a c c u r a c y1 4 . 5 %\\nK=2:#p r e d i c t e dt or e n t :6 , #w h od i dr e n t1 ,a c c u r a c y1 6 . 7 %\\nK=3:#p r e d i c t e dt or e n t :2 0 , #w h od i dr e n t3 ,a c c u r a c y1 5 . 0 %\\nK=4:#p r e d i c t e dt or e n t :3 , #w h od i dr e n t0 ,a c c u r a c y0 . 0 %\\nK=5:#p r e d i c t e dt or e n t :7 , #w h od i dr e n t1 ,a c c u r a c y1 4 . 3 %\\nWe see some variability — the numbers for K=4are very different from the\\nrest.\\nComparison to Logistic Regression\\nAs a comparison, we can also fit a logistic regression model to the data.\\nThis can also be done with sklearn, though by default it fits something\\nlike theridge regression version of logistic regression, which we introduce\\nin Chapter 6. This can be modified by appropriately setting the argument\\nCbelow. Its default value is 1 but by setting it to a very large number,\\nthe algorithm converges to the same solution as the usual (unregularized)\\nlogistic regression estimator discussed above.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 195}), Document(page_content=\"188 4. Classification\\nUnlike the statsmodels package, sklearnfocuses less on inference and\\nmore on classification. Hence, the summary methods seen in statsmodels\\nand our simplified version seen with summarize are not generally available\\nfor the classifiers in sklearn.\\nIn[61]: logit = LogisticRegression(C=1e10, solver= 'liblinear' )\\nlogit.fit(X_train, y_train)\\nlogit_pred = logit.predict_proba(X_test)\\nlogit_labels = np.where(logit_pred[:,1] > 5, 'Yes','No')\\nconfusion_table(logit_labels, y_test)\\nOut[61]: Truth No Yes\\nPredicted\\nNo 933 67\\nYes 0 0\\nWe used the argument solver='liblinear' above to avoid a warning with\\nthe default solver which would indicate that the algorithm does not con-\\nverge.\\nIf we use 0.5as the predicted probability cut-off for the classifier, then\\nwe have a problem: none of the test observations are predicted to purchase\\ninsurance. However, we are not required to use a cut-off of 0.5. If we instead\\npredict a purchase any time the predicted probability of purchase exceeds\\n0.25, we get much better results: we predict that 29 people will purchase\\ninsurance, and we are correct for about 31% of these people. This is almost\\nfive times better than random guessing!\\nIn[62]: logit_labels = np.where(logit_pred[:,1]>0.25, 'Yes','No')\\nconfusion_table(logit_labels, y_test)\\nOut[62]: Truth No Yes\\nPredicted\\nNo 913 58\\nYes 20 9\\nIn[63]: 9/(20+9)\\nOut[63]: 0.310\\n4.7.7 Linear and Poisson Regression on the Bikeshare Data\\nHere we fit linear and Poisson regression models to the Bikeshare data, as\\ndescribed in Section 4.6. The response bikersmeasures the number of bike\\nrentals per hour in Washington, DC in the period 2010–2012.\\nIn[64]: Bike = load_data( 'Bikeshare' )\\nLet’s have a peek at the dimensions and names of the variables in this\\ndataframe.\\nIn[65]: Bike.shape, Bike.columns\", metadata={'source': './documents/ISLP_website.pdf', 'page': 196}), Document(page_content=\"4.7 Lab: Logistic Regression, LDA, QDA, and KNN 189\\nOut[65]: ((8645, 15),\\nIndex([ 'season' ,'mnth','day','hr','holiday' ,'weekday' ,\\n'workingday' ,'weathersit' ,'temp','atemp' ,'hum',\\n'windspeed' ,'casual' ,'registered' ,'bikers' ],\\ndtype='object' ))\\nLinear Regression\\nWe begin by fitting a linear regression model to the data.\\nIn[66]: X=M S ( [ 'mnth',\\n'hr',\\n'workingday' ,\\n'temp',\\n'weathersit' ]).fit_transform(Bike)\\nY=B i k e [ 'bikers' ]\\nM_lm = sm.OLS(Y, X).fit()\\nsummarize(M_lm)\\nOut[66]: coef std err t P>|t|\\nintercept -68.6317 5.307 -12.932 0.000\\nmnth[Feb] 6.8452 4.287 1.597 0.110\\nmnth[March] 16.5514 4.301 3.848 0.000\\nmnth[April] 41.4249 4.972 8.331 0.000\\nmnth[May] 72.5571 5.641 12.862 0.000\\nmnth[June] 67.8187 6.544 10.364 0.000\\nmnth[July] 45.3245 7.081 6.401 0.000\\nmnth[Aug] 53.2430 6.640 8.019 0.000\\nmnth[Sept] 66.6783 5.925 11.254 0.000\\nmnth[Oct] 75.8343 4.950 15.319 0.000\\nmnth[Nov] 60.3100 4.610 13.083 0.000\\nmnth[Dec] 46.4577 4.271 10.878 0.000\\nhr[1] -14.5793 5.699 -2.558 0.011\\nhr[2] -21.5791 5.733 -3.764 0.000\\nhr[3] -31.1408 5.778 -5.389 0.000\\n..... ....... ..... ..... .....\\nThere are 24 levels in hrand 40 rows in all, so we have truncated the\\nsummary. In M_lm, the first levels hr[0]andmnth[Jan] are treated as the\\nbaseline values, and so no coefficient estimates are provided for them: im-\\nplicitly,theircoefficientestimatesarezero,andallotherlevelsaremeasured\\nrelative to these baselines. For example, the Feb coefficient of 6.845signi-\\nfies that, holding all other variables constant, there are on average about\\n7 more riders in February than in January. Similarly there are about 16.5\\nmore riders in March than in January.\\nThe results seen in Section 4.6.1used a slightly different coding of the\\nvariables hrandmnth, as follows:\\nIn[67]: hr_encode = contrast( 'hr','sum')\\nmnth_encode = contrast( 'mnth','sum')\\nRefitting again:\\nIn[68]: X2 = MS([mnth_encode,\\nhr_encode,\\n'workingday' ,\\n'temp',\", metadata={'source': './documents/ISLP_website.pdf', 'page': 197}), Document(page_content=\"190 4. Classification\\n'weathersit' ]).fit_transform(Bike)\\nM2_lm = sm.OLS(Y, X2).fit()\\nS2 = summarize(M2_lm)\\nS2\\nOut[68]: coef std err t P>|t|\\nintercept 73.5974 5.132 14.340 0.000\\nmnth[Jan] -46.0871 4.085 -11.281 0.000\\nmnth[Feb] -39.2419 3.539 -11.088 0.000\\nmnth[March] -29.5357 3.155 -9.361 0.000\\nmnth[April] -4.6622 2.741 -1.701 0.089\\nmnth[May] 26.4700 2.851 9.285 0.000\\nmnth[June] 21.7317 3.465 6.272 0.000\\nmnth[July] -0.7626 3.908 -0.195 0.845\\nmnth[Aug] 7.1560 3.535 2.024 0.043\\nmnth[Sept] 20.5912 3.046 6.761 0.000\\nmnth[Oct] 29.7472 2.700 11.019 0.000\\nmnth[Nov] 14.2229 2.860 4.972 0.000\\nhr[0] -96.1420 3.955 -24.307 0.000\\nhr[1] -110.7213 3.966 -27.916 0.000\\nhr[2] -117.7212 4.016 -29.310 0.000\\n..... ....... ..... ...... .....\\nWhat is the difference between the two codings? In M2_lm, a coefficient esti-\\nmate is reported for all but level 23ofhrand level Decofmnth. Importantly,\\ninM2_lm, the (unreported) coefficient estimate for the last level of mnthis\\nnot zero: instead, it equals the negative of the sum of the coefficient esti-\\nmates for all of the other levels. Similarly, in M2_lm, the coefficient estimate\\nfor the last level of hris the negative of the sum of the coefficient estimates\\nfor all of the other levels. This means that the coefficients of hrandmnth\\ninM2_lmwill always sum to zero, and can be interpreted as the difference\\nfrom the mean level. For example, the coefficient for January of −46.087\\nindicates that, holding all other variables constant, there are typically 46\\nfewer riders in January relative to the yearly average.\\nIt is important to realize that the choice of coding really does not matter,\\nprovided that we interpret the model output correctly in light of the coding\\nused. For example, we see that the predictions from the linear model are\\nthe same regardless of coding:\\nIn[69]: np.sum((M_lm.fittedvalues - M2_lm.fittedvalues)**2)\\nOut[69]: 1.53e-20\\nThe sum of squared differences is zero. We can also see this using the\\nnp.allclose() function:np.allclose()\\nIn[70]: np.allclose(M_lm.fittedvalues, M2_lm.fittedvalues)\\nOut[70]: True\\nTo reproduce the left-hand side of Figure 4.13we must first obtain the\\ncoefficient estimates associated with mnth. The coefficients for January\\nthrough November can be obtained directly from the M2_lmobject. The\\ncoefficient for December must be explicitly computed as the negative sum\\nof all the other months. We first extract all the coefficients for month from\\nthe coefficients of M2_lm.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 198}), Document(page_content=\"4.7 Lab: Logistic Regression, LDA, QDA, and KNN 191\\nIn[71]: coef_month = S2[S2.index.str.contains( 'mnth')]['coef']\\ncoef_month\\nOut[71]: mnth[Jan] -46.0871\\nmnth[Feb] -39.2419\\nmnth[March] -29.5357\\nmnth[April] -4.6622\\nmnth[May] 26.4700\\nmnth[June] 21.7317\\nmnth[July] -0.7626\\nmnth[Aug] 7.1560\\nmnth[Sept] 20.5912\\nmnth[Oct] 29.7472\\nmnth[Nov] 14.2229\\nName: coef, dtype: float64\\nNext, we append Decas the negative of the sum of all other months.\\nIn[72]: months = Bike[ 'mnth'].dtype.categories\\ncoef_month = pd.concat([\\ncoef_month,\\npd.Series([-coef_month.sum()],\\nindex=[ 'mnth[Dec]'\\n])\\n])\\ncoef_month\\nOut[72]: mnth[Jan] -46.0871\\nmnth[Feb] -39.2419\\nmnth[March] -29.5357\\nmnth[April] -4.6622\\nmnth[May] 26.4700\\nmnth[June] 21.7317\\nmnth[July] -0.7626\\nmnth[Aug] 7.1560\\nmnth[Sept] 20.5912\\nmnth[Oct] 29.7472\\nmnth[Nov] 14.2229\\nmnth[Dec] 0.3705\\nName: coef, dtype: float64\\nFinally, to make the plot neater, we’ll just use the first letter of each month,\\nwhich is the 6th entry of each of the labels in the index.\\nIn[73]: fig_month, ax_month = subplots(figsize=(8,8))\\nx_month = np.arange(coef_month.shape[0])\\nax_month.plot(x_month, coef_month, marker= 'o',m s = 1 0 )\\nax_month.set_xticks(x_month)\\nax_month.set_xticklabels([l[5] for l in coef_month.index], fontsize\\n=20)\\nax_month.set_xlabel( 'Month' ,f o n t s i z e = 2 0 )\\nax_month.set_ylabel( 'Coefficient' ,f o n t s i z e = 2 0 ) ;\\nReproducing the right-hand plot in Figure 4.13follows a similar process.\\nIn[74]: coef_hr = S2[S2.index.str.contains( 'hr')]['coef']\\ncoef_hr = coef_hr.reindex([ 'hr[{0}]' .format(h) for h in range(23)])\\ncoef_hr = pd.concat([coef_hr,\", metadata={'source': './documents/ISLP_website.pdf', 'page': 199}), Document(page_content=\"192 4. Classification\\npd.Series([-coef_hr.sum()], index=[ 'hr[23]' ])\\n])\\nWe now make the hour plot.\\nIn[75]: fig_hr, ax_hr = subplots(figsize=(8,8))\\nx_hr = np.arange(coef_hr.shape[0])\\nax_hr.plot(x_hr, coef_hr, marker= 'o',m s = 1 0 )\\nax_hr.set_xticks(x_hr[::2])\\nax_hr.set_xticklabels(range(24)[::2], fontsize=20)\\nax_hr.set_xlabel( 'Hour',f o n t s i z e = 2 0 )\\nax_hr.set_ylabel( 'Coefficient' ,f o n t s i z e = 2 0 ) ;\\nPoisson Regression\\nNow we fit instead a Poisson regression model to the Bikeshare data. Very\\nlittle changes, except that we now use the function sm.GLM() with the Pois-\\nson family specified:\\nIn[76]: M_pois = sm.GLM(Y, X2, family=sm.families.Poisson()).fit()\\nWe can plot the coefficients associated with mnthandhr, in order to\\nreproduce Figure 4.15. We first complete these coefficients as before.\\nIn[77]: S_pois = summarize(M_pois)\\ncoef_month = S_pois[S_pois.index.str.contains( 'mnth')]['coef']\\ncoef_month = pd.concat([coef_month,\\npd.Series([-coef_month.sum()],\\nindex=[ 'mnth[Dec]' ])])\\ncoef_hr = S_pois[S_pois.index.str.contains( 'hr')]['coef']\\ncoef_hr = pd.concat([coef_hr,\\npd.Series([-coef_hr.sum()],\\nindex=[ 'hr[23]' ])])\\nThe plotting is as before.\\nIn[78]: fig_pois, (ax_month, ax_hr) = subplots(1, 2, figsize=(16,8))\\nax_month.plot(x_month, coef_month, marker= 'o',m s = 1 0 )\\nax_month.set_xticks(x_month)\\nax_month.set_xticklabels([l[5] for l in coef_month.index], fontsize\\n=20)\\nax_month.set_xlabel( 'Month' ,f o n t s i z e = 2 0 )\\nax_month.set_ylabel( 'Coefficient' ,f o n t s i z e = 2 0 )\\nax_hr.plot(x_hr, coef_hr, marker= 'o',m s = 1 0 )\\nax_hr.set_xticklabels(range(24)[::2], fontsize=20)\\nax_hr.set_xlabel( 'Hour',f o n t s i z e = 2 0 )\\nax_hr.set_ylabel( 'Coefficient' ,f o n t s i z e = 2 0 ) ;\\nWecomparethefittedvaluesofthetwomodels.Thefittedvaluesarestored\\nin thefittedvalues attribute returned by the fit()method for both the\\nlinear regression and the Poisson fits. The linear predictors are stored as\\nthe attribute lin_pred .\\nIn[79]: fig, ax = subplots(figsize=(8, 8))\\nax.scatter(M2_lm.fittedvalues,\\nM_pois.fittedvalues,\\ns=20)\\nax.set_xlabel( 'Linear Regression Fit' ,f o n t s i z e = 2 0 )\", metadata={'source': './documents/ISLP_website.pdf', 'page': 200}), Document(page_content=\"4.8 Exercises 193\\nax.set_ylabel( 'Poisson Regression Fit' ,f o n t s i z e = 2 0 )\\nax.axline([0,0], c= 'black' ,l i n e w i d t h = 3 ,\\nlinestyle= '--',s l o p e = 1 ) ;\\nThe predictions from the Poisson regression model are correlated with\\nthose from the linear model; however, the former are non-negative. As a\\nresult the Poisson regression predictions tend to be larger than those from\\nthe linear model for either very low or very high levels of ridership.\\nIn this section, we fit Poisson regression models using the sm.GLM() func-\\ntion with the argument family=sm.families.Poisson() . Earlier in this lab\\nwe used the sm.GLM() function with family=sm.families.Binomial() to per-\\nform logistic regression. Other choices for the familyargument can be used\\nto fit other types of GLMs. For instance, family=sm.families.Gamma() fits\\na Gamma regression model.\\n4.8 Exercises\\nConceptual\\n1. Using a little bit of algebra, prove that ( 4.2) is equivalent to ( 4.3). In\\nother words, the logistic function representation and logit represen-\\ntation for the logistic regression model are equivalent.\\n2. It was stated in the text that classifying an observation to the class\\nfor which ( 4.17) is largest is equivalent to classifying an observation\\nto the class for which ( 4.18) is largest. Prove that this is the case. In\\nother words, under the assumption that the observations in the kth\\nclass are drawn from a N(µk,σ2)distribution, the Bayes classifier\\nassignsanobservationtotheclassforwhichthediscriminantfunction\\nis maximized.\\n3. This problem relates to the QDA model, in which the observations\\nwithin each class are drawn from a normal distribution with a class-\\nspecific mean vector and a class specific covariance matrix. We con-\\nsider the simple case where p=1; i.e. there is only one feature.\\nSuppose that we have Kclasses, and that if an observation belongs\\nto thekth class then Xcomes from a one-dimensional normal dis-\\ntribution, X∼N(µk,σ2\\nk). Recall that the density function for the\\none-dimensional normal distribution is given in ( 4.16). Prove that in\\nthis case, the Bayes classifier is notlinear. Argue that it is in fact\\nquadratic.\\nHint: For this problem, you should follow the arguments laid out in\\nSection4.4.1, but without making the assumption that σ2\\n1=···=σ2\\nK.\\n4. When the number of features pis large, there tends to be a deteri-\\noration in the performance of KNN and other localapproaches that\\nperform prediction using only observations that are nearthe test ob-\\nservation for which a prediction must be made. This phenomenon is\\nknown as thecurse of dimensionality, and it ties into the fact thatcurse of di-mensionalitynon-parametric approaches often perform poorly whenpis large. Wewill now investigate this curse.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 201}), Document(page_content='194 4. Classification(a) Suppose that we have a set of observations, each with measure-ments onp=1feature,X. We assume thatXis uniformly(evenly) distributed on[0,1]. Associated with each observationis a response value. Suppose that we wish to predict a test obser-vation’s response using only observations that are within 10% ofthe range ofXclosest to that test observation. For instance, inordertopredicttheresponseforatestobservationwithX=0.6,we will use observations in the range[0.55,0.65]. On average,what fraction of the available observations will we use to makethe prediction?(b) Now suppose that we have a set of observations, each withmeasurements onp=2features,X1andX2. We assume that(X1,X2)are uniformly distributed on[0,1]×[0,1]. We wish topredictatestobservation’sresponseusingonlyobservationsthatare within 10% of the range ofX1andwithin 10% of the rangeofX2closest to that test observation. For instance, in order topredict the response for a test observation withX1=0.6andX2=0.35, we will use observations in the range[0.55,0.65]forX1and in the range[0.3,0.4]forX2. On average, what fractionof the available observations will we use to make the prediction?(c) Now suppose that we have a set of observations onp= 100fea-tures. Again the observations are uniformly distributed on eachfeature, and again each feature ranges in value from 0 to 1. Wewish to predict a test observation’s response using observationswithin the 10% of each feature’s range that is closest to that testobservation. What fraction of the available observations will weuse to make the prediction?(d) Using your answers to parts (a)–(c), argue that a drawback ofKNN whenpis large is that there are very few training obser-vations “near” any given test observation.(e) Now suppose that we wish to make a prediction for a test obser-vation by creating ap-dimensional hypercube centered aroundthe test observation that contains, on average, 10% of the train-ingobservations.Forp=1,2,and100,whatisthelengthofeachside of the hypercube? Comment on your answer.Note: A hypercube is a generalization of a cube to an arbitrarynumber of dimensions. Whenp=1, a hypercube is simply a linesegment, whenp=2it is a square, and whenp= 100it is a100-dimensional cube.5. We now examine the differences between LDA and QDA.(a) If the Bayes decision boundary is linear, do we expect LDA orQDA to perform better on the training set? On the test set?(b) If the Bayes decision boundary is non-linear, do we expect LDAor QDA to perform better on the training set? On the test set?', metadata={'source': './documents/ISLP_website.pdf', 'page': 202}), Document(page_content='4.8 Exercises 195(c) In general, as the sample sizenincreases, do we expect the testprediction accuracy of QDA relative to LDA to improve, decline,or be unchanged? Why?(d) True or False: Even if the Bayes decision boundary for a givenproblem is linear, we will probably achieve a superior test er-ror rate using QDA rather than LDA because QDA is flexibleenough to model a linear decision boundary. Justify your an-swer.6. Suppose we collect data for a group of students in a statistics classwith variablesX1=hours studied,X2=undergrad GPA, andY=receive an A. We fit a logistic regression and produce estimatedcoefficient,ˆβ0=−6,ˆβ1=0.05,ˆβ2=1.(a) Estimate the probability that a student who studies for40h andhas an undergrad GPA of3.5gets an A in the class.(b) How many hours would the student in part (a) need to study tohave a 50% chance of getting an A in the class?7. Suppose that we wish to predict whether a given stock will issue adividend this year (“Yes” or “No”) based onX, last year’s percentprofit.Weexaminealargenumberofcompaniesanddiscoverthatthemean value ofXfor companies that issued a dividend was¯X= 10,while the mean for those that didn’t was¯X=0. In addition, thevariance ofXfor these two sets of companies wasˆσ2= 36. Finally,80% of companies issued dividends. Assuming thatXfollows a nor-mal distribution, predict the probability that a company will issuea dividend this year given that its percentage profit wasX=4lastyear.Hint: Recall that the density function for a normal random variableisf(x)=1√2πσ2e−(x−µ)2/2σ2. You will need to use Bayes’ theorem.8. Suppose that we take a data set, divide it into equally-sized trainingand test sets, and then try out two different classification procedures.First we use logistic regression and get an error rate of 20% on thetraining data and 30% on the test data. Next we use 1-nearest neigh-bors (i.e.K=1) and get an average error rate (averaged over bothtestand training data sets) of 18%. Based on these results, whichmethod should we prefer to use for classification of new observations?Why?9. This problem has to do withodds.(a) On average, what fraction of people with an odds of 0.37 ofdefaulting on their credit card payment will in fact default?(b) Suppose that an individual has a 16% chance of defaulting onher credit card payment. What are the odds that she will de-fault?', metadata={'source': './documents/ISLP_website.pdf', 'page': 203}), Document(page_content='196 4. Classification\\n10. Equation 4.32derived an expression for log(\\nPr(Y=k|X=x)\\nPr(Y=K|X=x))\\nin the\\nsetting where p>1, so that the mean for the kth class, µk, is ap-\\ndimensional vector, and the shared covariance Σis ap×pmatrix.\\nHowever, in the setting with p=1,(4.32) takes a simpler form, since\\nthe means µ1,...,µ Kand the variance σ2are scalars. In this simpler\\nsetting, repeat the calculation in ( 4.32), and provide expressions for\\nakandbkjin terms of πk,πK,µk,µK, andσ2.\\n11. Workoutthedetailedformsof ak,bkj,andbkjlin(4.33).Youranswer\\nshould involve πk,πK,µk,µK,Σk, andΣK.\\n12. Suppose that you wish to classify an observation X∈Rintoapples\\nandoranges. You fit a logistic regression model and find that\\nˆPr(Y=orange |X=x)=exp( ˆβ0+ˆβ1x)\\n1 + exp( ˆβ0+ˆβ1x).\\nYour friend fits a logistic regression model to the same data using the\\nsoftmaxformulation in ( 4.13), and finds that\\nˆPr(Y=orange |X=x)=\\nexp(ˆαorange 0+ˆαorange 1x)\\nexp(ˆαorange 0+ˆαorange 1x) + exp(ˆ αapple 0+ˆαapple 1x).\\n(a) What is the log odds of orangeversusapplein your model?\\n(b) What is the log odds of orangeversusapplein your friend’s\\nmodel?\\n(c) Suppose that in your model, ˆβ0=2 andˆβ1=−1. What are\\nthe coefficient estimates in your friend’s model? Be as specific\\nas possible.\\n(d) Now suppose that you and your friend fit the same two models\\non a different data set. This time, your friend gets the coefficient\\nestimates ˆαorange 0=1.2,ˆαorange 1=−2,ˆαorange 0=3,ˆαorange 1=\\n0.6. What are the coefficient estimates in your model?\\n(e) Finally, suppose you apply both models from (d) to a data set\\nwith 2,000 test observations. What fraction of the time do you\\nexpect the predicted class labels from your model to agree with\\nthose from your friend’s model? Explain your answer.\\nApplied\\n13. This question should be answered using the Weeklydata set, which\\nis part of the ISLPpackage. This data is similar in nature to the\\nSmarketdata from this chapter’s lab, except that it contains 1,089\\nweekly returns for 21 years, from the beginning of 1990 to the end of\\n2010.\\n(a) Produce some numerical and graphical summaries of the Weekly\\ndata. Do there appear to be any patterns?', metadata={'source': './documents/ISLP_website.pdf', 'page': 204}), Document(page_content=\"4.8 Exercises 197(b) Use the full data set to perform a logistic regression withDirectionas the response and the five lag variables plusVolumeas predictors. Use the summary function to print the results. Doany of the predictors appear to be statistically significant? If so,which ones?(c) Compute the confusion matrix and overall fraction of correctpredictions. Explain what the confusion matrix is telling youabout the types of mistakes made by logistic regression.(d) Nowfitthelogisticregressionmodelusingatrainingdataperiodfrom 1990 to 2008, withLag2as the only predictor. Compute theconfusion matrix and the overall fraction of correct predictionsfor the held out data (that is, the data from 2009 and 2010).(e) Repeat (d) using LDA.(f) Repeat (d) using QDA.(g) Repeat (d) using KNN withK=1.(h) Repeat (d) using naive Bayes.(i) Which of these methods appears to provide the best results onthis data?(j) Experiment with different combinations of predictors, includ-ing possible transformations and interactions, for each of themethods. Report the variables, method, and associated confu-sion matrix that appears to provide the best results on the heldout data. Note that you should also experiment with values forKin the KNN classifier.14. In this problem, you will develop a model to predict whether a givencar gets high or low gas mileage based on theAutodata set.(a) Create a binary variable,mpg01, that contains a 1 ifmpgcontainsa value above its median, and a 0 ifmpgcontains a value belowits median. You can compute the median using themedian()method of the data frame. Note you may find it helpful to adda columnmpg01to the data frame by assignment. Assuming youhave stored the data frame asAuto, this can be done as follows:Auto['mpg01']=m p g 0 1(b) Explore the data graphically in order to investigate the associ-ation betweenmpg01and the other features. Which of the otherfeatures seem most likely to be useful in predictingmpg01? Scat-terplots and boxplots may be useful tools to answer this ques-tion. Describe your findings.(c) Split the data into a training set and a test set.(d) Perform LDA on the training data in order to predictmpg01using the variables that seemed most associated withmpg01in(b). What is the test error of the model obtained?\", metadata={'source': './documents/ISLP_website.pdf', 'page': 205}), Document(page_content='198 4. Classification(e) Perform QDA on the training data in order to predictmpg01using the variables that seemed most associated withmpg01in(b). What is the test error of the model obtained?(f) Perform logistic regression on the training data in order to pre-dictmpg01using the variables that seemed most associated withmpg01in (b). What is the test error of the model obtained?(g) Perform naive Bayes on the training data in order to predictmpg01usingthevariablesthatseemedmostassociatedwithmpg01in (b). What is the test error of the model obtained?(h) Perform KNN on the training data, with several values ofK, inorder to predictmpg01. Use only the variables that seemed mostassociated withmpg01in (b). What test errors do you obtain?Which value ofKseems to perform the best on this data set?15. This problem involves writing functions.(a) Write a function,Power(), that prints out the result of raising2to the3rd power. In other words, your function should compute23and print out the results.Hint: Recall thatx**araisesxto the powera. Use theprint()function to display the result.(b) Create a new function,Power2(), that allows you to passanytwo numbers,xanda, and prints out the value ofx**a. You cando this by beginning your function with the linedef Power2(x, a):Youshouldbeabletocallyourfunctionbyentering,forinstance,Power2(3, 8)onthecommandline.Thisshouldoutputthevalueof38,namely,6,561.(c) Using thePower2()function that you just wrote, compute103,817, and1313.(d) Now create a new function,Power3(), that actuallyreturnstheresultx**aas aPythonobject, rather than simply printing itto the screen. That is, if you store the valuex**ain an objectcalledresultwithin your function, then you can simplyreturnreturnthis result, using the following line:return resultNote that the line above should be the last line in your function,and it should be indented 4 spaces.(e) Now using thePower3()function, create a plot off(x)=x2.Thex-axis should display a range of integers from1to10, andthey-axis should displayx2. Label the axes appropriately, anduse an appropriate title for the figure. Consider displaying eitherthex-axis, they-axis, or both on the log-scale. You can do thisby using theax.set_xscale()andax.set_yscale()methods of.set_xscale().set_yscale()the axes you are plotting to.', metadata={'source': './documents/ISLP_website.pdf', 'page': 206}), Document(page_content='4.8 Exercises 199(f) Create a function,PlotPower(), that allows you to create a plotofxagainstx**afor a fixedaand a sequence of values ofx. Forinstance, if you callPlotPower(np.arange(1, 11), 3)then a plot should be created with anx-axis taking on values1,2,...,10, and ay-axis taking on values13,23,...,103.16. Using theBostondata set, fit classification models in order to predictwhether a given suburb has a crime rate above or below the median.Explorelogisticregression,LDA,naiveBayes,andKNNmodelsusingvarious subsets of the predictors. Describe your findings.Hint: You will have to create the response variable yourself, using thevariables that are contained in theBostondata set.', metadata={'source': './documents/ISLP_website.pdf', 'page': 207}), Document(page_content='5\\nResampling Methods\\nResampling methods are an indispensable tool in modern statistics. They\\ninvolverepeatedlydrawingsamplesfromatrainingsetandrefittingamodel\\nof interest on each sample in order to obtain additional information about\\nthefittedmodel.Forexample,inordertoestimatethevariabilityofalinear\\nregression fit, we can repeatedly draw different samples from the training\\ndata, fit a linear regression to each new sample, and then examine the\\nextent to which the resulting fits differ. Such an approach may allow us to\\nobtain information that would not be available from fitting the model only\\nonce using the original training sample.\\nResampling approaches can be computationally expensive, because they\\ninvolve fitting the same statistical method multiple times using different\\nsubsets of the training data. However, due to recent advances in computing\\npower, the computational requirements of resampling methods generally\\nare not prohibitive. In this chapter, we discuss two of the most commonly\\nusedresamplingmethods, cross-validation andthebootstrap .Bothmethods\\nare important tools in the practical application of many statistical learning\\nprocedures. For example, cross-validation can be used to estimate the test\\nerrorassociatedwithagivenstatisticallearningmethodinordertoevaluate\\nits performance, or to select the appropriate level of flexibility. The process\\nofevaluatingamodel’sperformanceisknownas model assessment ,whereasmodel\\nassessmenttheprocess of selecting theproper levelof flexibilityfora model is knownas\\nmodel selection . The bootstrap is used in several contexts, most commonlymodel\\nselectionto provide a measure of accuracy of a parameter estimate or of a given\\nstatistical learning method.\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning , Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_5  201', metadata={'source': './documents/ISLP_website.pdf', 'page': 208}), Document(page_content='202 5. Resampling Methods\\n5.1 Cross-Validation\\nIn Chapter 2we discuss the distinction between the test error rate and the\\ntrainingerrorrate .Thetesterroristheaverageerrorthatresultsfromusing\\nastatisticallearningmethodtopredicttheresponseonanewobservation—\\nthat is, a measurement that was not used in training the method. Given\\na data set, the use of a particular statistical learning method is warranted\\nif it results in a low test error. The test error can be easily calculated if a\\ndesignated test set is available. Unfortunately, this is usually not the case.\\nIn contrast, the training error can be easily calculated by applying the\\nstatistical learning method to the observations used in its training. But as\\nwe saw in Chapter 2, the training error rate often is quite different from the\\ntest error rate, and in particular the former can dramatically underestimate\\nthe latter.\\nIn the absence of a very large designated test set that can be used to\\ndirectly estimate the test error rate, a number of techniques can be used\\nto estimate this quantity using the available training data. Some methods\\nmake a mathematical adjustment to the training error rate in order to\\nestimate the test error rate. Such approaches are discussed in Chapter 6.\\nIn this section, we instead consider a class of methods that estimate the\\ntest error rate by holding out a subset of the training observations from the\\nfitting process, and then applying the statistical learning method to those\\nheld out observations.\\nIn Sections 5.1.1–5.1.4, for simplicity we assume that we are interested\\nin performing regression with a quantitative response. In Section 5.1.5we\\nconsider the case of classification with a qualitative response. As we will\\nsee, the key concepts remain the same regardless of whether the response\\nis quantitative or qualitative.\\n5.1.1 The Validation Set Approach\\nSuppose that we would like to estimate the test error associated with fit-\\nting a particular statistical learning method on a set of observations. The\\nvalidation set approach , displayed in Figure 5.1, is a very simple strategyvalidation\\nset approachfor this task. It involves randomly dividing the available set of observa-\\ntions into two parts, a training set and avalidation set orhold-out set . Thevalidation\\nset\\nhold-out setmodel is fit on the training set, and the fitted model is used to predict the\\nresponsesfortheobservationsinthevalidationset.Theresultingvalidation\\nset error rate—typically assessed using MSE in the case of a quantitative\\nresponse—provides an estimate of the test error rate.\\nWeillustratethevalidationsetapproachonthe Autodataset.Recallfrom\\nChapter 3that there appears to be a non-linear relationship between mpg\\nandhorsepower , and that a model that predicts mpgusinghorsepower and\\nhorsepower2gives better results than a model that uses only a linear term.\\nIt is natural to wonder whether a cubic or higher-order fit might provide\\neven better results. We answer this question in Chapter 3by looking at\\nthe p-values associated with a cubic term and higher-order polynomial\\nterms in a linear regression. But we could also answer this question using\\nthe validation method. We randomly split the 392observations into two', metadata={'source': './documents/ISLP_website.pdf', 'page': 209}), Document(page_content='5.1 Cross-Validation 203/g3/g1/g4/g1/g5/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g2/g1/g6/g1/g1/g4/g4/g1/g1/g3/g5/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g7/g3/g1FIGURE 5.1. A schematic display of the validation set approach. A set of n\\nobservations are randomly split into a training set (shown in blue, containing\\nobservations 7, 22, and 13, among others) and a validation set (shown in beige,\\nand containing observation 91, among others). The statistical learning method is\\nfit on the training set, and its performance is evaluated on the validation set.\\nsets, a training set containing 196of the data points, and a validation set\\ncontaining the remaining 196observations. The validation set error rates\\nthat result from fitting various regression models on the training sample\\nand evaluating their performance on the validation sample, using MSE\\nas a measure of validation set error, are shown in the left-hand panel of\\nFigure5.2. The validation set MSE for the quadratic fit is considerably\\nsmallerthanforthelinearfit.However,thevalidationsetMSEforthecubic\\nfit is actually slightly larger than for the quadratic fit. This implies that\\nincluding a cubic term in the regression does not lead to better prediction\\nthan simply using a quadratic term.\\nRecall that in order to create the left-hand panel of Figure 5.2, we ran-\\ndomly divided the data set into two parts, a training set and a validation\\nset. If we repeat the process of randomly splitting the sample set into two\\nparts, we will get a somewhat different estimate for the test MSE. As an\\nillustration, the right-hand panel of Figure 5.2displays ten different vali-\\ndation set MSE curves from the Autodata set, produced using ten different\\nrandom splits of the observations into training and validation sets. All ten\\ncurves indicate that the model with a quadratic term has a dramatically\\nsmaller validation set MSE than the model with only a linear term. Fur-\\nthermore, all ten curves indicate that there is not much benefit in including\\ncubic or higher-order polynomial terms in the model. But it is worth noting\\nthat each of the ten curves results in a different test MSE estimate for each\\nof the ten regression models considered. And there is no consensus among\\nthe curves as to which model results in the smallest validation set MSE.\\nBased on the variability among these curves, all that we can conclude with\\nany confidence is that the linear fit is not adequate for this data.\\nThe validation set approach is conceptually simple and is easy to imple-\\nment. But it has two potential drawbacks:\\n1. As is shown in the right-hand panel of Figure 5.2, the validation esti-\\nmate of the test error rate can be highly variable, depending on pre-\\ncisely which observations are included in the training set and which\\nobservations are included in the validation set.\\n2. In the validation approach, only a subset of the observations—those\\nthat are included in the training set rather than in the validation\\nset—are used to fit the model. Since statistical methods tend to per-\\nform worse when trained on fewer observations, this suggests that the', metadata={'source': './documents/ISLP_website.pdf', 'page': 210}), Document(page_content='204 5. Resampling Methods\\n2 4 6 8 1016 18 20 22 24 26 28Degree of PolynomialMean Squared Error2 4 6 8 1016 18 20 22 24 26 28Degree of PolynomialMean Squared ErrorFIGURE 5.2. The validation set approach was used on the Autodata set in\\norder to estimate the test error that results from predicting mpgusing polynomial\\nfunctions of horsepower .Left:Validation error estimates for a single split into\\ntraining and validation data sets. Right:The validation method was repeated ten\\ntimes, each time using a different random split of the observations into a training\\nset and a validation set. This illustrates the variability in the estimated test MSE\\nthat results from this approach.\\nvalidation set error rate may tend to overestimate the test error rate\\nfor the model fit on the entire data set.\\nIn the coming subsections, we will present cross-validation , a refinement of\\nthe validation set approach that addresses these two issues.\\n5.1.2 Leave-One-Out Cross-Validation\\nLeave-one-out cross-validation (LOOCV)iscloselyrelatedtothevalidationleave-one-\\nout\\ncross-\\nvalidationset approach of Section 5.1.1, but it attempts to address that method’s\\ndrawbacks.\\nLike the validation set approach, LOOCV involves splitting the set of\\nobservations into two parts. However, instead of creating two subsets of\\ncomparable size, a single observation (x1,y1)is used for the validation\\nset, and the remaining observations {(x2,y2),...,(xn,yn)}make up the\\ntraining set. The statistical learning method is fit on the n−1training\\nobservations, and a prediction ˆy1is made for the excluded observation,\\nusingitsvalue x1.Since(x1,y1)wasnotusedinthefittingprocess, MSE 1=\\n(y1−ˆy1)2provides an approximately unbiased estimate for the test error.\\nBut even though MSE 1is unbiased for the test error, it is a poor estimate\\nbecause it is highly variable, since it is based upon a single observation\\n(x1,y1).\\nWe can repeat the procedure by selecting (x2,y2)for the validation\\ndata, training the statistical learning procedure on the n−1observations\\n{(x1,y1),(x3,y3),...,(xn,yn)},andcomputing MSE 2=(y2−ˆy2)2.Repeat-\\ning this approach ntimes produces nsquared errors, MSE 1,..., MSE n.\\nThe LOOCV estimate for the test MSE is the average of these ntest error\\nestimates:\\nCV (n)=1\\nnn∑\\ni=1MSE i. (5.1)', metadata={'source': './documents/ISLP_website.pdf', 'page': 211}), Document(page_content='5.1 Cross-Validation 205/g4/g1/g5/g1/g6/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g2/g1/g4/g1/g5/g1/g6/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g2/g1/g4/g1/g5/g1/g6/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g2/g1/g4/g1/g5/g1/g6/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g2/g1/g4/g1/g5/g1/g6/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g2/g1/g3/g1/g3/g1/g3/g1FIGURE 5.3. A schematic display of LOOCV. A set of ndata points is repeat-\\nedly split into a training set (shown in blue) containing all but one observation,\\nand a validation set that contains only that observation (shown in beige). The test\\nerror is then estimated by averaging the nresulting MSEs. The first training set\\ncontains all but observation 1, the second training set contains all but observation\\n2, and so forth.\\nA schematic of the LOOCV approach is illustrated in Figure 5.3.\\nLOOCV has a couple of major advantages over the validation set ap-\\nproach. First, it has far less bias. In LOOCV, we repeatedly fit the sta-\\ntistical learning method using training sets that contain n−1observa-\\ntions, almost as many as are in the entire data set. This is in contrast to\\nthe validation set approach, in which the training set is typically around\\nhalf the size of the original data set. Consequently, the LOOCV approach\\ntends not to overestimate the test error rate as much as the validation\\nset approach does. Second, in contrast to the validation approach which\\nwill yield different results when applied repeatedly due to randomness in\\nthe training/validation set splits, performing LOOCV multiple times will\\nalways yield the same results: there is no randomness in the training/vali-\\ndation set splits.\\nWe used LOOCV on the Autodata set in order to obtain an estimate\\nof the test set MSE that results from fitting a linear regression model to\\npredictmpgusingpolynomialfunctionsof horsepower .Theresultsareshown\\nin the left-hand panel of Figure 5.4.\\nLOOCV has the potential to be expensive to implement, since the model\\nhas to be fit ntimes. This can be very time consuming if nis large, and if\\neach individual model is slow to fit. With least squares linear or polynomial\\nregression, an amazing shortcut makes the cost of LOOCV the same as that\\nof a single model fit! The following formula holds:\\nCV (n)=1\\nnn∑\\ni=1(yi−ˆyi\\n1−hi)2\\n, (5.2)', metadata={'source': './documents/ISLP_website.pdf', 'page': 212}), Document(page_content='206 5. Resampling Methods\\n2 4 6 8 1016 18 20 22 24 26 28LOOCV\\nDegree of PolynomialMean Squared Error2 4 6 8 1016 18 20 22 24 26 2810−fold CV\\nDegree of PolynomialMean Squared ErrorFIGURE 5.4. Cross-validation was used on the Autodata set in order to\\nestimate the test error that results from predicting mpgusing polynomial functions\\nofhorsepower .Left:The LOOCV error curve. Right:10-fold CV was run nine\\nseparate times, each with a different random split of the data into ten parts. The\\nfigure shows the nine slightly different CV error curves.\\nwhereˆyiis theith fitted value from the original least squares fit, and hiis\\nthe leverage defined in ( 3.37) on page 105.1This is like the ordinary MSE,\\nexcept the ith residual is divided by 1−hi. The leverage lies between 1/n\\nand1, and reflects the amount that an observation influences its own fit.\\nHence the residuals for high-leverage points are inflated in this formula by\\nexactly the right amount for this equality to hold.\\nLOOCV is a very general method, and can be used with any kind of\\npredictive modeling. For example we could use it with logistic regression\\nor linear discriminant analysis, or any of the methods discussed in later\\nchapters. The magic formula ( 5.2) does not hold in general, in which case\\nthe model has to be refit ntimes.\\n5.1.3k-Fold Cross-Validation\\nAn alternative to LOOCV is k-fold CV . This approach involves randomlyk-fold CVdividing the set of observations into kgroups, or folds, of approximately\\nequal size. The first fold is treated as a validation set, and the method\\nis fit on the remaining k−1folds. The mean squared error, MSE 1, is\\nthen computed on the observations in the held-out fold. This procedure is\\nrepeated ktimes; each time, a different group of observations is treated\\nas a validation set. This process results in kestimates of the test error,\\nMSE 1,MSE 2,...,MSE k. Thek-fold CV estimate is computed by averaging\\nthese values,\\nCV (k)=1\\nkk∑\\ni=1MSE i. (5.3)\\nFigure5.5illustrates the k-fold CV approach.\\n1In the case of multiple linear regression, the leverage takes a slightly more compli-\\ncated form than ( 3.37), but (5.2) still holds.', metadata={'source': './documents/ISLP_website.pdf', 'page': 213}), Document(page_content='5.1 Cross-Validation 207/g4/g1/g5/g1/g6/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g2/g1/g4/g4/g1/g10/g9/g1/g8/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g7/g10/g1/g4/g4/g1/g10/g9/g1/g8/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g7/g10/g1/g4/g4/g1/g10/g9/g1/g8/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g7/g10/g1/g4/g4/g1/g10/g9/g1/g8/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g7/g10/g1/g4/g4/g1/g10/g9/g1/g8/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g7/g10/g1/g4/g4/g1/g10/g9/g1/g8/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g7/g10/g1FIGURE 5.5. A schematic display of 5-fold CV. A set of nobservations is\\nrandomly split into five non-overlapping groups. Each of these fifths acts as a\\nvalidation set (shown in beige), and the remainder as a training set (shown in\\nblue). The test error is estimated by averaging the five resulting MSE estimates.\\nIt is not hard to see that LOOCV is a special case of k-fold CV in which k\\nis set to equal n. In practice, one typically performs k-fold CV using k=5\\nork= 10 . What is the advantage of using k=5 ork= 10 rather than\\nk=n? The most obvious advantage is computational. LOOCV requires\\nfitting the statistical learning method ntimes. This has the potential to be\\ncomputationally expensive (except for linear models fit by least squares,\\nin which case formula ( 5.2) can be used). But cross-validation is a very\\ngeneral approach that can be applied to almost any statistical learning\\nmethod. Some statistical learning methods have computationally intensive\\nfitting procedures, and so performing LOOCV may pose computational\\nproblems, especially if nis extremely large. In contrast, performing 10-fold\\nCV requires fitting the learning procedure only ten times, which may be\\nmuch more feasible. As we see in Section 5.1.4, there also can be other\\nnon-computational advantages to performing 5-fold or10-fold CV, which\\ninvolve the bias-variance trade-off.\\nThe right-hand panel of Figure 5.4displays nine different 10-fold CV\\nestimates for the Autodata set, each resulting from a different random split\\nof the observations into ten folds. As we can see from the figure, there is\\nsome variability in the CV estimates as a result of the variability in how\\nthe observations are divided into ten folds. But this variability is typically\\nmuch lower than the variability in the test error estimates that results from\\nthe validation set approach (right-hand panel of Figure 5.2).\\nWhen we examine real data, we do not know the truetest MSE, and\\nso it is difficult to determine the accuracy of the cross-validation estimate.\\nHowever, if we examine simulated data, then we can compute the true\\ntest MSE, and can thereby evaluate the accuracy of our cross-validation\\nresults. In Figure 5.6, we plot the cross-validation estimates and true test\\nerror rates that result from applying smoothing splines to the simulated\\ndata sets illustrated in Figures 2.9–2.11of Chapter 2. The true test MSE\\nis displayed in blue. The black dashed and orange solid lines respectively\\nshow the estimated LOOCV and 10-fold CV estimates. In all three plots,\\nthe two cross-validation estimates are very similar. In the right-hand panel', metadata={'source': './documents/ISLP_website.pdf', 'page': 214}), Document(page_content='208 5. Resampling Methods\\n25 10 200.0 0.5 1.0 1.5 2.0 2.5 3.0FlexibilityMean Squared Error25 10 200.0 0.5 1.0 1.5 2.02.5 3.0\\nFlexibilityMean Squared Error25 10 2005101520\\nFlexibilityMean Squared ErrorFIGURE 5.6. True and estimated test MSE for the simulated data sets in\\nFigures2.9(left),2.10(center), and2.11(right). The true test MSE is shown\\nin blue, the LOOCV estimate is shown as a black dashed line, and the 10-fold\\nCV estimate is shown in orange. The crosses indicate the minimum of each of\\nthe MSE curves.\\nof Figure 5.6, the true test MSE and the cross-validation curves are almost\\nidentical. In the center panel of Figure 5.6, the two sets of curves are similar\\nat the lower degrees of flexibility, while the CV curves overestimate the test\\nsetMSEforhigherdegreesofflexibility.Intheleft-handpanelofFigure 5.6,\\nthe CV curves have the correct general shape, but they underestimate the\\ntrue test MSE.\\nWhen we perform cross-validation, our goal might be to determine how\\nwell a given statistical learning procedure can be expected to perform on\\nindependent data; in this case, the actual estimate of the test MSE is\\nof interest. But at other times we are interested only in the location of\\ntheminimum point in the estimated test MSE curve . This is because we\\nmight be performing cross-validation on a number of statistical learning\\nmethods, or on a single method using different levels of flexibility, in order\\ntoidentifythemethodthatresultsinthelowesttesterror.Forthispurpose,\\nthe location of the minimum point in the estimated test MSE curve is\\nimportant, but the actual value of the estimated test MSE is not. We find\\nin Figure 5.6that despite the fact that they sometimes underestimate the\\ntrue test MSE, all of the CV curves come close to identifying the correct\\nlevelofflexibility—thatis,theflexibilitylevelcorrespondingtothesmallest\\ntest MSE.\\n5.1.4 Bias-Variance Trade-Off for k-Fold Cross-Validation\\nWe mentioned in Section 5.1.3thatk-fold CV with k<n has a compu-\\ntational advantage to LOOCV. But putting computational issues aside,\\na less obvious but potentially more important advantage of k-fold CV is\\nthat it often gives more accurate estimates of the test error rate than does\\nLOOCV. This has to do with a bias-variance trade-off.\\nIt was mentioned in Section 5.1.1that the validation set approach can\\nlead to overestimates of the test error rate, since in this approach the\\ntraining set used to fit the statistical learning method contains only half\\nthe observations of the entire data set. Using this logic, it is not hard to see', metadata={'source': './documents/ISLP_website.pdf', 'page': 215}), Document(page_content='5.1 Cross-Validation 209\\nthat LOOCV will give approximately unbiased estimates of the test error,\\nsince each training set contains n−1observations, which is almost as many\\nas the number of observations in the full data set. And performing k-fold\\nCV for, say, k=5 ork= 10 will lead to an intermediate level of bias,\\nsince each training set contains approximately (k−1)n/kobservations—\\nfewer than in the LOOCV approach, but substantially more than in the\\nvalidation set approach. Therefore, from the perspective of bias reduction,\\nit is clear that LOOCV is to be preferred to k-fold CV.\\nHowever, we know that bias is not the only source for concern in an esti-\\nmating procedure; we must also consider the procedure’s variance. It turns\\noutthatLOOCV hashigher variancethan does k-foldCVwith k<n .Why\\nis this the case? When we perform LOOCV, we are in effect averaging the\\noutputs of nfitted models, each of which is trained on an almost identical\\nset of observations; therefore, these outputs are highly (positively) corre-\\nlated with each other. In contrast, when we perform k-fold CV with k<n ,\\nwe are averaging the outputs of kfitted models that are somewhat less\\ncorrelated with each other, since the overlap between the training sets in\\neach model is smaller. Since the mean of many highly correlated quantities\\nhas higher variance than does the mean of many quantities that are not\\nas highly correlated, the test error estimate resulting from LOOCV tends\\nto have higher variance than does the test error estimate resulting from\\nk-fold CV.\\nTo summarize, there is a bias-variance trade-off associated with the\\nchoice of kink-fold cross-validation. Typically, given these considerations,\\none performs k-fold cross-validation using k=5ork= 10 , as these values\\nhave been shown empirically to yield test error rate estimates that suffer\\nneither from excessively high bias nor from very high variance.\\n5.1.5 Cross-Validation on Classification Problems\\nIn this chapter so far, we have illustrated the use of cross-validation in the\\nregression setting where the outcome Yis quantitative, and so have used\\nMSE to quantify test error. But cross-validation can also be a very useful\\napproach in the classification setting when Yis qualitative. In this setting,\\ncross-validation works just as described earlier in this chapter, except that\\nrather than using MSE to quantify test error, we instead use the number\\nof misclassified observations. For instance, in the classification setting, the\\nLOOCV error rate takes the form\\nCV (n)=1\\nnn∑\\ni=1Err i, (5.4)\\nwhereErr i=I(yi̸=ˆyi). Thek-fold CV error rate and validation set error\\nrates are defined analogously.\\nAs an example, we fit various logistic regression models on the two-\\ndimensional classification data displayed in Figure 2.13. In the top-left\\npanelofFigure 5.7,theblacksolidlineshowstheestimateddecisionbound-\\nary resulting from fitting a standard logistic regression model to this data\\nset. Since this is simulated data, we can compute the truetest error rate,\\nwhich takes a value of 0.201and so is substantially larger than the Bayes', metadata={'source': './documents/ISLP_website.pdf', 'page': 216}), Document(page_content='210 5. Resampling MethodsDegree=1ooooooooooo\\noooo\\noo\\nooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo\\nooooooooo\\nooooooooooooooooooooooo\\noooooooooooo\\noooooooooooooooooooooooooooooooooooooooooooooooooooooooooo\\nooooooDegree=2ooooooooooo\\noooo\\noo\\nooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo\\nooooooooo\\nooooooooooooooooooooooo\\noooooooooooo\\noooooooooooooooooooooooooooooooooooooooooooooooooooooooooo\\noooooo\\nDegree=3ooooooooooo\\noooo\\noo\\nooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo\\nooooooooo\\nooooooooooooooooooooooo\\noooooooooooo\\noooooooooooooooooooooooooooooooooooooooooooooooooooooooooo\\nooooooDegree=4ooooooooooo\\noooo\\noo\\nooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo\\nooooooooo\\nooooooooooooooooooooooo\\noooooooooooo\\noooooooooooooooooooooooooooooooooooooooooooooooooooooooooo\\noooooo\\nFIGURE 5.7. Logistic regression fits on the two-dimensional classification data\\ndisplayed in Figure 2.13. The Bayes decision boundary is represented using a\\npurple dashed line. Estimated decision boundaries from linear, quadratic, cubic\\nand quartic (degrees 1–4) logistic regressions are displayed in black. The test error\\nrates for the four logistic regression fits are respectively 0.201,0.197,0.160, and\\n0.162, while the Bayes error rate is 0.133.\\nerror rate of 0.133. Clearly logistic regression does not have enough flexi-\\nbility to model the Bayes decision boundary in this setting. We can easily\\nextend logistic regression to obtain a non-linear decision boundary by using\\npolynomial functions of the predictors, as we did in the regression setting in\\nSection3.3.2. For example, we can fit a quadratic logistic regression model,\\ngiven by\\nlog(p\\n1−p)\\n=β0+β1X1+β2X2\\n1+β3X2+β4X2\\n2. (5.5)\\nThe top-right panel of Figure 5.7displays the resulting decision boundary,\\nwhichisnowcurved.However,thetesterrorratehasimprovedonlyslightly,\\nto0.197. A much larger improvement is apparent in the bottom-left panel', metadata={'source': './documents/ISLP_website.pdf', 'page': 217}), Document(page_content='5.1 Cross-Validation 211\\n2 4 6 8 100.12 0.14 0.16 0.18 0.20Order of Polynomials UsedError Rate0.01 0.02 0.05 0.10 0.20 0.50 1.000.12 0.14 0.16 0.18 0.201/KError RateFIGURE 5.8. Test error (brown), training error (blue), and 10-fold CV error\\n(black) on the two-dimensional classification data displayed in Figure 5.7.Left:\\nLogistic regression using polynomial functions of the predictors. The order of\\nthe polynomials used is displayed on the x-axis.Right:The KNN classifier with\\ndifferent values of K, the number of neighbors used in the KNN classifier.\\nof Figure 5.7, in which we have fit a logistic regression model involving\\ncubic polynomials of the predictors. Now the test error rate has decreased\\nto0.160. Going to a quartic polynomial (bottom-right) slightly increases\\nthe test error.\\nIn practice, for real data, the Bayes decision boundary and the test er-\\nror rates are unknown. So how might we decide between the four logistic\\nregression models displayed in Figure 5.7? We can use cross-validation in\\norder to make this decision. The left-hand panel of Figure 5.8displays in\\nblack the 10-fold CV error rates that result from fitting ten logistic regres-\\nsion models to the data, using polynomial functions of the predictors up\\nto tenth order. The true test errors are shown in brown, and the training\\nerrors are shown in blue. As we have seen previously, the training error\\ntends to decrease as the flexibility of the fit increases. (The figure indicates\\nthat though the training error rate doesn’t quite decrease monotonically,\\nit tends to decrease on the whole as the model complexity increases.) In\\ncontrast, the test error displays a characteristic U-shape. The 10-fold CV\\nerror rate provides a pretty good approximation to the test error rate.\\nWhile it somewhat underestimates the error rate, it reaches a minimum\\nwhen fourth-order polynomials are used, which is very close to the min-\\nimum of the test curve, which occurs when third-order polynomials are\\nused. In fact, using fourth-order polynomials would likely lead to good test\\nset performance, as the true test error rate is approximately the same for\\nthird, fourth, fifth, and sixth-order polynomials.\\nThe right-hand panel of Figure 5.8displays the same three curves us-\\ning the KNN approach for classification, as a function of the value of K\\n(which in this context indicates the number of neighbors used in the KNN\\nclassifier, rather than the number of CV folds used). Again the training\\nerror rate declines as the method becomes more flexible, and so we see that\\nthe training error rate cannot be used to select the optimal value for K.\\nThough the cross-validation error curve slightly underestimates the test\\nerror rate, it takes on a minimum very close to the best value for K.', metadata={'source': './documents/ISLP_website.pdf', 'page': 218}), Document(page_content='212 5. Resampling Methods\\n5.2 The Bootstrap\\nThebootstrap is a widely applicable and extremely powerful statistical toolbootstrapthat can be used to quantify the uncertainty associated with a given esti-\\nmator or statistical learning method. As a simple example, the bootstrap\\ncan be used to estimate the standard errors of the coefficients from a linear\\nregression fit. In the specific case of linear regression, this is not particularly\\nuseful, since we saw in Chapter 3that standard statistical software such as\\nRoutputs such standard errors automatically. However, the power of the\\nbootstrap lies in the fact that it can be easily applied to a wide range of\\nstatistical learning methods, including some for which a measure of vari-\\nability is otherwise difficult to obtain and is not automatically output by\\nstatistical software.\\nIn this section we illustrate the bootstrap on a toy example in which we\\nwish to determine the best investment allocation under a simple model.\\nIn Section 5.3we explore the use of the bootstrap to assess the variability\\nassociated with the regression coefficients in a linear model fit.\\nSuppose that we wish to invest a fixed sum of money in two financial\\nassets that yield returns of XandY, respectively, where XandYare\\nrandom quantities. We will invest a fraction αof our money in X, and will\\ninvest the remaining 1−αinY. Since there is variability associated with\\nthe returns on these two assets, we wish to choose αto minimize the total\\nrisk, or variance, of our investment. In other words, we want to minimize\\nVar(αX+ (1 −α)Y). One can show that the value that minimizes the risk\\nis given by\\nα=σ2\\nY−σXY\\nσ2\\nX+σ2\\nY−2σXY, (5.6)\\nwhereσ2\\nX= Var( X),σ2\\nY= Var( Y), andσXY= Cov( X,Y).\\nInreality,thequantities σ2\\nX,σ2\\nY,andσXYareunknown.Wecancompute\\nestimates for these quantities, ˆσ2\\nX,ˆσ2\\nY, andˆσXY, using a data set that\\ncontains past measurements for XandY. We can then estimate the value\\nofαthat minimizes the variance of our investment using\\nˆα=ˆσ2\\nY−ˆσXY\\nˆσ2\\nX+ˆσ2\\nY−2ˆσXY. (5.7)\\nFigure5.9illustrates this approach for estimating αon a simulated data\\nset. In each panel, we simulated 100pairs of returns for the investments\\nXandY. We used these returns to estimate σ2\\nX,σ2\\nY, andσXY, which we\\nthen substituted into ( 5.7) in order to obtain estimates for α. The value of\\nˆαresulting from each simulated data set ranges from 0.532to0.657.\\nIt is natural to wish to quantify the accuracy of our estimate of α.T o\\nestimate the standard deviation of ˆα, we repeated the process of simu-\\nlating100paired observations of XandY, and estimating αusing (5.7),\\n1,000 times. We thereby obtained 1,000 estimates for α, which we can call\\nˆα1,ˆα2,..., ˆα1,000. The left-hand panel of Figure 5.10displays a histogram\\nof the resulting estimates. For these simulations the parameters were set to\\nσ2\\nX=1,σ2\\nY=1.25, andσXY=0.5, and so we know that the true value of\\nαis0.6. We indicated this value using a solid vertical line on the histogram.', metadata={'source': './documents/ISLP_website.pdf', 'page': 219}), Document(page_content='5.2 The Bootstrap 213\\n−2 −1 0 1 2−2 −1 0 1 2XY\\n−2 −1 0 1 2−2 −1 0 1 2XY\\n−3 −2 −1 0 1 2−3 −2 −1 0 1 2XY\\n−2 −1 0 1 2 3−3 −2 −1 0 1 2XYFIGURE 5.9. Each panel displays 100 simulated returns for investments\\nXandY. From left to right and top to bottom, the resulting estimates for α\\nare0.576,0.532,0.657, and0.651.\\nThe mean over all 1,000 estimates for αis\\n¯α=1\\n10001000∑\\nr=1ˆαr=0.5996 ,\\nvery close to α=0.6, and the standard deviation of the estimates is\\n\\ued6a\\ued6b\\ued6b√1\\n1000 −11000∑\\nr=1(ˆαr−¯α)2=0.083.\\nThis gives us a very good idea of the accuracy of ˆα:SE(ˆα)≈0.083. So\\nroughly speaking, for a random sample from the population, we would\\nexpectˆαto differ from αby approximately 0.08, on average.\\nIn practice, however, the procedure for estimating SE(ˆα)outlined above\\ncannot be applied, because for real data we cannot generate new samples\\nfrom the original population. However, the bootstrap approach allows us\\nto use a computer to emulate the process of obtaining new sample sets,\\nso that we can estimate the variability of ˆαwithout generating additional\\nsamples. Rather than repeatedly obtaining independent data sets from the\\npopulation, we instead obtain distinct data sets by repeatedly sampling\\nobservations from the original data set .\\nThis approach is illustrated in Figure 5.11on a simple data set, which\\nwe callZ, that contains only n=3 observations. We randomly select n\\nobservations from the data set in order to produce a bootstrap data set,', metadata={'source': './documents/ISLP_website.pdf', 'page': 220}), Document(page_content='214 5. Resampling Methods\\n0.4 0.5 0.6 0.7 0.8 0.90 50 100 150 2000.3 0.4 0.5 0.6 0.7 0.8 0.90 50 100 150 200TrueBootstrap0.3 0.4 0.5 0.6 0.7 0.8 0.9αα\\nαFIGURE 5.10. Left:A histogram of the estimates of αobtained by generating\\n1,000 simulated data sets from the true population. Center:A histogram of the\\nestimates of αobtained from 1,000 bootstrap samples from a single data set.\\nRight:The estimates of αdisplayed in the left and center panels are shown as\\nboxplots. In each panel, the pink line indicates the true value of α.\\nZ∗1. The sampling is performed with replacement , which means that thewith\\nreplacementsame observation can occur more than once in the bootstrap data set. In\\nthis example, Z∗1contains the third observation twice, the first observation\\nonce,andnoinstancesofthesecondobservation.Notethatifanobservation\\nis contained in Z∗1, then both its XandYvalues are included. We can use\\nZ∗1to produce a new bootstrap estimate for α, which we call ˆα∗1. This\\nprocedureisrepeated Btimesforsomelargevalueof B,inordertoproduce\\nBdifferent bootstrap data sets, Z∗1,Z∗2,...,Z∗B, andBcorresponding α\\nestimates, ˆα∗1,ˆα∗2,..., ˆα∗B. We can compute the standard error of these\\nbootstrap estimates using the formula\\nSEB(ˆα)=\\ued6a\\ued6b\\ued6b√1\\nB−1B∑\\nr=1(\\nˆα∗r−1\\nBB∑\\nr′=1ˆα∗r′)2\\n. (5.8)\\nThis serves as an estimate of the standard error of ˆαestimated from the\\noriginal data set.\\nThe bootstrap approach is illustrated in the center panel of Figure 5.10,\\nwhich displays a histogram of 1,000 bootstrap estimates of α, each com-\\nputed using a distinct bootstrap data set. This panel was constructed on\\nthe basis of a single data set, and hence could be created using real data.\\nNote that the histogram looks very similar to the left-hand panel, which\\ndisplays the idealized histogram of the estimates of αobtained by generat-\\ning 1,000 simulated data sets from the true population. In particular the\\nbootstrap estimate SE(ˆα)from (5.8) is0.087, very close to the estimate of\\n0.083obtained using 1,000 simulated data sets. The right-hand panel dis-\\nplays the information in the center and left panels in a different way, via\\nboxplots of the estimates for αobtained by generating 1,000 simulated data\\nsets from the true population and using the bootstrap approach. Again, the\\nboxplots have similar spreads, indicating that the bootstrap approach can\\nbe used to effectively estimate the variability associated with ˆα.', metadata={'source': './documents/ISLP_website.pdf', 'page': 221}), Document(page_content='5.3 Lab: Cross-Validation and the Bootstrap 215\\n2.8 5.3 3 1.1 2.1 2 2.4 4.3 1 Y X Obs 2.8 5.3 3 2.4 4.3 1 2.8 5.3 3 Y X Obs \\n2.4 4.3 1 2.8 5.3 3 1.1 2.1 2 Y X Obs \\n2.4 4.3 1 1.1 2.1 2 1.1 2.1 2 Y X Obs Original Data (Z) 1*Z2*ZZ*B1*ˆ\\n2*ˆˆ*B/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1/g1FIGURE 5.11.A graphical illustration of the bootstrap approach on a smallsample containingn=3observations. Each bootstrap data set containsnobser-vations, sampled with replacement from the original data set. Each bootstrap dataset is used to obtain an estimate ofα.5.3 Lab: Cross-Validation and the BootstrapIn this lab, we explore the resampling techniques covered in this chap-ter. Some of the commands in this lab may take a while to run on yourcomputer.We again begin by placing most of our imports at this top level.In[1]:import numpy as npimport statsmodels.api as smfrom ISLP import load_datafrom ISLP.models import (ModelSpec as MS,summarize,poly)from sklearn.model_selection import train_test_splitThere are several new imports needed for this lab.In[2]:from functools import partialfrom sklearn.model_selection import \\\\(cross_validate,KFold,ShuffleSplit)from sklearn.base import clonefrom ISLP.models import sklearn_sm\\x01\\x01\\x01', metadata={'source': './documents/ISLP_website.pdf', 'page': 222}), Document(page_content=\"216 5. Resampling Methods5.3.1 The Validation Set ApproachWe explore the use of the validation set approach in order to estimate thetest error rates that result from fitting various linear models on theAutodata set.We use the functiontrain_test_split()to split the data into trainingtrain_test_split()and validation sets. As there are 392 observations, we split into two equalsets of size 196 using the argumenttest_size=196. It is generally a goodidea to set a random seed when performing operations like this that containan element of randomness, so that the results obtained can be reproducedprecisely at a later time. We set the random seed of the splitter with theargumentrandom_state=0.In[3]:Auto = load_data('Auto')Auto_train, Auto_valid = train_test_split(Auto,test_size=196,random_state=0)Now we can fit a linear regression using only the observations corre-sponding to the training setAuto_train.In[4]:hp_mm = MS(['horsepower'])X_train = hp_mm.fit_transform(Auto_train)y_train = Auto_train['mpg']model = sm.OLS(y_train, X_train)results = model.fit()Wenowusethepredict()methodofresultsevaluatedonthemodelma-trix for this model created using the validation data set. We also calculatethe validation MSE of our model.In[5]:X_valid = hp_mm.transform(Auto_valid)y_valid = Auto_valid['mpg']valid_pred = results.predict(X_valid)np.mean((y_valid - valid_pred)**2)Out[5]:23.6166Hence our estimate for the validation MSE of the linear regression fit is23.62.We can also estimate the validation error for higher-degree polynomialregressions. We first provide a functionevalMSE()that takes a model stringas well as a training and test set and returns the MSE on the test set.In[6]:def evalMSE(terms,response,train,test):mm = MS(terms)X_train = mm.fit_transform(train)y_train = train[response]X_test = mm.transform(test)y_test = test[response]\", metadata={'source': './documents/ISLP_website.pdf', 'page': 223}), Document(page_content=\"5.3 Lab: Cross-Validation and the Bootstrap 217results = sm.OLS(y_train, X_train).fit()test_pred = results.predict(X_test)return np.mean((y_test - test_pred)**2)Let’s use this function to estimate the validation MSE using linear,quadratic and cubic fits. We use theenumerate()function here, which givesenumerate()both the values and indices of objects as one iterates over a for loop.In[7]:MSE = np.zeros(3)for idx, degree in enumerate(range(1, 4)):MSE[idx] = evalMSE([poly('horsepower',d e g r e e ) ] ,'mpg',Auto_train,Auto_valid)MSEOut[7]:array([23.62, 18.76, 18.80])These error rates are23.62,18.76, and18.80, respectively. If we choose adifferent training/validation split instead, then we can expect somewhatdifferent errors on the validation set.In[8]:Auto_train, Auto_valid = train_test_split(Auto,test_size=196,random_state=3)MSE = np.zeros(3)for idx, degree in enumerate(range(1, 4)):MSE[idx] = evalMSE([poly('horsepower',d e g r e e ) ] ,'mpg',Auto_train,Auto_valid)MSEOut[8]:array([20.76, 16.95, 16.97])Using this split of the observations into a training set and a validationset, we find that the validation set error rates for the models with linear,quadratic, and cubic terms are20.76,16.95, and16.97, respectively.These results are consistent with our previous findings: a model thatpredictsmpgusing a quadratic function ofhorsepowerperforms better thana model that involves only a linear function ofhorsepower, and there is noevidence of an improvement in using a cubic function ofhorsepower.5.3.2 Cross-ValidationIn theory, the cross-validation estimate can be computed for any general-ized linear model. In practice, however, the simplest way to cross-validatein Python is to usesklearn, which has a different interface or API thanstatsmodels, the code we have been using to fit GLMs.Thisisaproblemwhichoftenconfrontsdatascientists:“Ihaveafunctionto do taskA, and need to feed it into something that performs taskB, sothat I can computeB(A(D)), whereDis my data.” WhenAandBdon’tnaturally speak to each other, this requires the use of awrapper. In theISLPwrapper\", metadata={'source': './documents/ISLP_website.pdf', 'page': 224}), Document(page_content=\"218 5. Resampling Methods\\npackage, we provide a wrapper, sklearn_sm() , that enables us to easily usesklearn_sm()the cross-validation tools of sklearnwith models fit by statsmodels .\\nTheclass sklearn_sm() hasasitsfirstargumentamodelfrom statsmodels .\\nItcantaketwoadditional optionalarguments: model_str whichcanbeused\\nto specify a formula, and model_args which should be a dictionary of addi-\\ntional arguments used when fitting the model. For example, to fit a logistic\\nregression model we have to specify a familyargument. This is passed as\\nmodel_args={'family':sm.families.Binomial()} .\\nHere is our wrapper in action:\\nIn[9]: hp_model = sklearn_sm(sm.OLS,\\nMS(['horsepower' ]))\\nX, Y = Auto.drop(columns=[ 'mpg']), Auto[ 'mpg']\\ncv_results = cross_validate(hp_model,\\nX,\\nY,\\ncv=Auto.shape[0])\\ncv_err = np.mean(cv_results[ 'test_score' ])\\ncv_err\\nOut[9]: 24.2315\\nThe arguments to cross_validate() are as follows: an object with the ap-\\npropriate fit(),predict() ,andscore()methods,anarrayoffeatures Xand\\naresponse Y.Wealsoincludedanadditionalargument cvtocross_validate() ;\\nspecifyinganinteger Kresultsin K-foldcross-validation.Wehaveprovided\\na value corresponding to the total number of observations, which results\\nin leave-one-out cross-validation (LOOCV). The cross_validate() func-cross_\\nvalidate() tion produces a dictionary with several components; we simply want the\\ncross-validated test score here (MSE), which is estimated to be 24.23.\\nWe can repeat this procedure for increasingly complex polynomial fits.\\nTo automate the process, we again use a for loop which iteratively fits\\npolynomial regressions of degree 1 to 5, computes the associated cross-\\nvalidation error, and stores it in the ith element of the vector cv_error .\\nThe variable din the for loop corresponds to the degree of the polynomial.\\nWe begin by initializing the vector. This command may take a couple of\\nseconds to run.\\nIn[10]: cv_error = np.zeros(5)\\nH=n p . a r r a y ( A u t o [ 'horsepower' ])\\nM=s k l e a r n _ s m ( s m . O L S )\\nfor i, d in enumerate(range(1,6)):\\nX=n p . p o w e r . o u t e r ( H ,n p . a r a n g e ( d + 1 ) )\\nM_CV = cross_validate(M,\\nX,\\nY,\\ncv=Auto.shape[0])\\ncv_error[i] = np.mean(M_CV[ 'test_score' ])\\ncv_error\\nOut[10]: array([24.2315, 19.2482, 19.3350, 19.4244, 19.0332])\\nAs in Figure 5.4, we see a sharp drop in the estimated test MSE between\\nthe linear and quadratic fits, but then no clear improvement from using\\nhigher-degree polynomials.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 225}), Document(page_content=\"5.3 Lab: Cross-Validation and the Bootstrap 219\\nAbove we introduced the outer() method of the np.power() function..outer()\\nnp.power()Theouter()method is applied to an operation that has two arguments,\\nsuch as add(),min(), orpower(). It has two arrays as arguments, and then\\nforms a larger array where the operation is applied to each pair of elements\\nof the two arrays.\\nIn[11]: A=n p . a r r a y ( [ 3 ,5 ,9 ] )\\nB=n p . a r r a y ( [ 2 ,4 ] )\\nnp.add.outer(A, B)\\nOut[11]: array([[ 5, 7],\\n[7 , 9 ] ,\\n[11, 13]])\\nIn the CV example above, we used K=n, but of course we can also use\\nK<n. The code is very similar to the above (and is significantly faster).\\nHere we use KFold()to partition the data into K= 10random groups. WeKFold()userandom_state to set a random seed and initialize a vector cv_error in\\nwhich we will store the CV errors corresponding to the polynomial fits of\\ndegrees one to five.\\nIn[12]: cv_error = np.zeros(5)\\ncv = KFold(n_splits=10,\\nshuffle=True,\\nrandom_state=0) #u s es a m es p l i t sf o re a c hd e g r e e\\nfor i, d in enumerate(range(1,6)):\\nX=n p . p o w e r . o u t e r ( H ,n p . a r a n g e ( d + 1 ) )\\nM_CV = cross_validate(M,\\nX,\\nY,\\ncv=cv)\\ncv_error[i] = np.mean(M_CV[ 'test_score' ])\\ncv_error\\nOut[12]: array([24.2077, 19.1853, 19.2763, 19.4785, 19.1372])\\nNotice that the computation time is much shorter than that of LOOCV.\\n(In principle, the computation time for LOOCV for a least squares linear\\nmodel should be faster than for K-fold CV, due to the availability of the\\nformula ( 5.2) for LOOCV; however, the generic cross_validate() function\\ndoes not make use of this formula.) We still see little evidence that using\\ncubic or higher-degree polynomial terms leads to a lower test error than\\nsimply using a quadratic fit.\\nThecross_validate() function is flexible and can take different splitting\\nmechanisms as an argument. For instance, one can use the ShuffleSplit()Shuffle\\nSplit() funtion to implement the validation set approach just as easily as K-fold\\ncross-validation.\\nIn[13]: validation = ShuffleSplit(n_splits=1,\\ntest_size=196,\\nrandom_state=0)\\nresults = cross_validate(hp_model,\\nAuto.drop([ 'mpg'], axis=1),\\nAuto['mpg'],\\ncv=validation);\\nresults[ 'test_score' ]\", metadata={'source': './documents/ISLP_website.pdf', 'page': 226}), Document(page_content=\"220 5. Resampling Methods\\nOut[13]: array([23.6166])\\nOnecanestimatethevariabilityinthetesterrorbyrunningthefollowing:\\nIn[14]: validation = ShuffleSplit(n_splits=10,\\ntest_size=196,\\nrandom_state=0)\\nresults = cross_validate(hp_model,\\nAuto.drop([ 'mpg'], axis=1),\\nAuto['mpg'],\\ncv=validation)\\nresults[ 'test_score' ].mean(), results[ 'test_score' ].std()\\nOut[14]: (23.8022, 1.4218)\\nNote that this standard deviation is not a valid estimate of the sam-\\npling variability of the mean test score or the individual scores, since the\\nrandomly-selected training samples overlap and hence introduce correla-\\ntions. But it does give an idea of the Monte Carlo variation incurred by\\npicking different random folds.\\n5.3.3 The Bootstrap\\nWe illustrate the use of the bootstrap in the simple example of Section 5.2,\\nas well as on an example involving estimating the accuracy of the linear\\nregression model on the Autodata set.\\nEstimating the Accuracy of a Statistic of Interest\\nOne of the great advantages of the bootstrap approach is that it can be\\napplied in almost all situations. No complicated mathematical calculations\\nare required. While there are several implementations of the bootstrap in\\nPython,itsuseforestimatingstandarderrorissimpleenoughthatwewrite\\nour own function below for the case when our data is stored in a dataframe.\\nToillustratethebootstrap,westartwithasimpleexample.The Portfolio\\ndata set in the ISLPpackage is described in Section 5.2. The goal is to es-\\ntimate the sampling variance of the parameter αgiven in formula ( 5.7).\\nWe will create a function alpha_func() , which takes as input a dataframe D\\nassumed to have columns XandY, as well as a vector idxindicating which\\nobservations should be used to estimate α. The function then outputs the\\nestimate for αbased on the selected observations.\\nIn[15]: Portfolio = load_data( 'Portfolio' )\\ndef alpha_func(D, idx):\\ncov_ = np.cov(D[[ 'X','Y']].loc[idx], rowvar=False)\\nreturn ((cov_[1,1] - cov_[0,1]) /\\n(cov_[0,0]+cov_[1,1]-2*cov_[0,1]))\\nThis function returns an estimate for αbased on applying the minimum\\nvarianceformula( 5.7)totheobservationsindexedbytheargument idx.For\\ninstance, the following command estimates αusing all 100 observations.\\nIn[16]: alpha_func(Portfolio, range(100))\", metadata={'source': './documents/ISLP_website.pdf', 'page': 227}), Document(page_content='5.3 Lab: Cross-Validation and the Bootstrap 221Out[16]:0.5758Next we randomly select 100 observations fromrange(100), with replace-ment. This is equivalent to constructing a new bootstrap data set andrecomputingˆαbased on the new data set.In[17]:rng = np.random.default_rng(0)alpha_func(Portfolio,rng.choice(100,100,replace=True))Out[17]:0.6074This process can be generalized to create a simple functionboot_SE()forcomputing the bootstrap standard error for arbitrary functions that takeonly a data frame as an argument.In[18]:def boot_SE(func,D,n=None,B=1000,seed=0):rng = np.random.default_rng(seed)first_, second_ = 0, 0n=no rD . s h a p e [ 0 ]for _ in range(B):idx = rng.choice(D.index,n,replace=True)value = func(D, idx)first_ += valuesecond_ += value**2return np.sqrt(second_ / B - (first_ / B)**2)Notice the use of_as a loop variable infor _ in range(B). This is oftenused if the value of the counter is unimportant and simply makes sure theloop is executedBtimes.Let’s use our function to evaluate the accuracy of our estimate ofαusingB=1,000bootstrap replications.In[19]:alpha_SE = boot_SE(alpha_func,Portfolio,B=1000,seed=0)alpha_SEOut[19]:0.0912The final output shows that the bootstrap estimate forSE(ˆα)is0.0912.Estimating the Accuracy of a Linear Regression ModelThe bootstrap approach can be used to assess the variability of the coef-ficient estimates and predictions from a statistical learning method. Herewe use the bootstrap approach in order to assess the variability of the', metadata={'source': './documents/ISLP_website.pdf', 'page': 228}), Document(page_content=\"222 5. Resampling Methods\\nestimates for β0andβ1, the intercept and slope terms for the linear regres-\\nsion model that uses horsepower to predict mpgin theAutodata set. We\\nwill compare the estimates obtained using the bootstrap to those obtained\\nusing the formulas for SE(ˆβ0)andSE(ˆβ1)described in Section 3.1.2.\\nTo use our boot_SE() function, we must write a function (its first argu-\\nment) that takes a data frame Dand indices idxas its only arguments. But\\nhere we want to bootstrap a specific regression model, specified by a model\\nformula and data. We show how to do this in a few simple steps.\\nWe start by writing a generic function boot_OLS() for bootstrapping a\\nregression model that takes a formula to define the corresponding regres-\\nsion. We use the clone()function to make a copy of the formula that canclone()be refit to the new dataframe. This means that any derived features such\\nas those defined by poly()(which we will see shortly), will be re-fit on the\\nresampled data frame.\\nIn[20]: def boot_OLS(model_matrix, response, D, idx):\\nD_ = D.loc[idx]\\nY_ = D_[response]\\nX_ = clone(model_matrix).fit_transform(D_)\\nreturn sm.OLS(Y_, X_).fit().params\\nThisisnotquitewhatisneededasthefirstargumentto boot_SE() .Thefirst\\ntwo arguments which specify the model will not change in the bootstrap\\nprocess, and we would like to freezethem. The function partial() from thepartial()functools module does precisely this: it takes a function as an argument,\\nand freezes some of its arguments, starting from the left. We use it to freeze\\nthe first two model-formula arguments of boot_OLS() .\\nIn[21]: hp_func = partial(boot_OLS, MS([ 'horsepower' ]),'mpg')\\nTyping hp_func? will show that it has two arguments Dandidx— it is a\\nversion of boot_OLS() with the first two arguments frozen — and hence is\\nideal as the first argument for boot_SE() .\\nThehp_func() function can now be used in order to create bootstrap\\nestimates for the intercept and slope terms by randomly sampling from\\namong the observations with replacement. We first demonstrate its utility\\non 10 bootstrap samples.\\nIn[22]: rng = np.random.default_rng(0)\\nnp.array([hp_func(Auto,\\nrng.choice(392,\\n392,\\nreplace=True)) for _ in range(10)])\\nOut[22]: array([[39.8806, -0.1568],\\n[38.733 , -0.147 ],\\n[38.3173, -0.1444],\\n[39.9145, -0.1578],\\n[39.4335, -0.1507],\\n[40.3663, -0.1591],\\n[39.6233, -0.1545],\\n[39.0581, -0.1495],\\n[38.6669, -0.1452],\\n[39.6428, -0.1556]])\", metadata={'source': './documents/ISLP_website.pdf', 'page': 229}), Document(page_content=\"5.3 Lab: Cross-Validation and the Bootstrap 223\\nNext, we use the boot_SE() function to compute the standard errors of\\n1,000 bootstrap estimates for the intercept and slope terms.\\nIn[23]: hp_se = boot_SE(hp_func,\\nAuto,\\nB=1000,\\nseed=10)\\nhp_se\\nOut[23]: intercept 0.8488\\nhorsepower 0.0074\\ndtype: float64\\nThis indicates that the bootstrap estimate for SE(ˆβ0)is 0.85, and that\\nthe bootstrap estimate for SE(ˆβ1)is 0.0074. As discussed in Section 3.1.2,\\nstandard formulas can be used to compute the standard errors for the\\nregression coefficients in a linear model. These can be obtained using the\\nsummarize() function from ISLP.sm.\\nIn[24]: hp_model.fit(Auto, Auto[ 'mpg'])\\nmodel_se = summarize(hp_model.results_)[ 'std err' ]\\nmodel_se\\nOut[24]: intercept 0.717\\nhorsepower 0.006\\nName: std err, dtype: float64\\nThe standard error estimates for ˆβ0andˆβ1obtained using the formulas\\nfrom Section 3.1.2are 0.717 for the intercept and 0.006 for the slope. Inter-\\nestingly, these are somewhat different from the estimates obtained using\\nthe bootstrap. Does this indicate a problem with the bootstrap? In fact,\\nit suggests the opposite. Recall that the standard formulas given in Equa-\\ntion3.8on page75rely on certain assumptions. For example, they depend\\non the unknown parameter σ2, the noise variance. We then estimate σ2\\nusing the RSS. Now although the formula for the standard errors do not\\nrely on the linear model being correct, the estimate for σ2does. We see in\\nFigure3.8on page 99that there is a non-linear relationship in the data,\\nand so the residuals from a linear fit will be inflated, and so will ˆσ2. Sec-\\nondly, the standard formulas assume (somewhat unrealistically) that the\\nxiare fixed, and all the variability comes from the variation in the errors\\nϵi. The bootstrap approach does not rely on any of these assumptions, and\\nso it is likely giving a more accurate estimate of the standard errors of ˆβ0\\nandˆβ1than the results from sm.OLS.\\nBelow we compute the bootstrap standard error estimates and the stan-\\ndardlinearregressionestimatesthatresultfromfittingthequadraticmodel\\nto the data. Since this model provides a good fit to the data (Figure 3.8),\\nthere is now a better correspondence between the bootstrap estimates and\\nthe standard estimates of SE(ˆβ0),SE(ˆβ1)andSE(ˆβ2).\\nIn[25]: quad_model = MS([poly( 'horsepower' ,2 ,r a w = T r u e ) ] )\\nquad_func = partial(boot_OLS,\\nquad_model,\\n'mpg')\\nboot_SE(quad_func, Auto, B=1000)\", metadata={'source': './documents/ISLP_website.pdf', 'page': 230}), Document(page_content=\"224 5. Resampling Methods\\nOut[25]: intercept 2.067840\\npoly(horsepower, 2, raw=True)[0] 0.033019\\npoly(horsepower, 2, raw=True)[1] 0.000120\\ndtype: float64\\nWe compare the results to the standard errors computed using sm.OLS() .\\nIn[26]: M=s m . O L S ( A u t o [ 'mpg'],\\nquad_model.fit_transform(Auto))\\nsummarize(M.fit())[ 'std err' ]\\nOut[26]: intercept 1.800\\npoly(horsepower, 2, raw=True)[0] 0.031\\npoly(horsepower, 2, raw=True)[1] 0.000\\nName: std err, dtype: float64\\n5.4 Exercises\\nConceptual\\n1. Using basic statistical properties of the variance, as well as single-\\nvariable calculus, derive ( 5.6). In other words, prove that αgiven by\\n(5.6) does indeed minimize Var(αX+ (1−α)Y).\\n2. We will now derive the probability that a given observation is part\\nof a bootstrap sample. Suppose that we obtain a bootstrap sample\\nfrom a set of nobservations.\\n(a) What is the probability that the first bootstrap observation is\\nnotthejth observation from the original sample? Justify your\\nanswer.\\n(b) What is the probability that the second bootstrap observation\\nisnotthejth observation from the original sample?\\n(c) Argue that the probability that the jth observation is notin the\\nbootstrap sample is (1−1/n)n.\\n(d) When n=5, what is the probability that the jth observation is\\nin the bootstrap sample?\\n(e) When n= 100, what is the probability that the jth observation\\nis in the bootstrap sample?\\n(f) When n= 10,000, what is the probability that the jth observa-\\ntion is in the bootstrap sample?\\n(g) Create a plot that displays, for each integer value of nfrom1\\nto100,000, the probability that the jth observation is in the\\nbootstrap sample. Comment on what you observe.\\n(h) We will now investigate numerically the probability that a boot-\\nstrap sample of size n= 100contains the jth observation. Here\\nj=4. We first create an array storewith values that will subse-\\nquently be overwritten using the function np.empty() . We thennp.empty()\", metadata={'source': './documents/ISLP_website.pdf', 'page': 231}), Document(page_content='5.4 Exercises 225\\nrepeatedly create bootstrap samples, and each time we record\\nwhetherornotthefifthobservationiscontainedinthebootstrap\\nsample.\\nrng = np.random.default_rng(10)\\nstore = np.empty(10000)\\nfor i in range(10000):\\nstore[i] = np.sum(rng.choice(100, replace=True) == 4)\\n>0\\nnp.mean(store)\\nComment on the results obtained.\\n3. We now review k-fold cross-validation.\\n(a) Explain how k-fold cross-validation is implemented.\\n(b) What are the advantages and disadvantages of k-fold cross-\\nvalidation relative to:\\ni. The validation set approach?\\nii. LOOCV?\\n4. Suppose that we use some statistical learning method to make a pre-\\ndiction for the response Yfor a particular value of the predictor X.\\nCarefully describe how we might estimate the standard deviation of\\nour prediction.\\nApplied\\n5. In Chapter 4, we used logistic regression to predict the probability of\\ndefaultusingincomeandbalanceon theDefaultdata set. We will\\nnow estimate the test error of this logistic regression model using the\\nvalidation set approach. Do not forget to set a random seed before\\nbeginning your analysis.\\n(a) Fit a logistic regression model that uses incomeandbalanceto\\npredict default.\\n(b) Using the validation set approach, estimate the test error of this\\nmodel. In order to do this, you must perform the following steps:\\ni. Split the sample set into a training set and a validation set.\\nii. Fit a multiple logistic regression model using only the train-\\ning observations.\\niii. Obtain a prediction of default status for each individual in\\nthe validation set by computing the posterior probability of\\ndefault for that individual, and classifying the individual to\\nthedefaultcategory if the posterior probability is greater\\nthan 0.5.\\niv. Compute the validation set error, which is the fraction of\\nthe observations in the validation set that are misclassified.\\n(c) Repeat the process in (b) three times, using three different splits\\nof the observations into a training set and a validation set. Com-\\nment on the results obtained.', metadata={'source': './documents/ISLP_website.pdf', 'page': 232}), Document(page_content='226 5. Resampling Methods\\n(d) Now consider a logistic regression model that predicts the prob-\\nability of defaultusingincome,balance, and a dummy variable\\nforstudent. Estimate the test error for this model using the val-\\nidation set approach. Comment on whether or not including a\\ndummy variable for studentleads to a reduction in the test error\\nrate.\\n6. We continue to consider the use of a logistic regression model to\\npredict the probability of defaultusingincomeandbalanceon the\\nDefaultdata set. In particular, we will now compute estimates for the\\nstandard errors of the incomeandbalancelogistic regression coeffi-\\ncients in two different ways: (1) using the bootstrap, and (2) using the\\nstandard formula for computing the standard errors in the sm.GLM()\\nfunction. Do not forget to set a random seed before beginning your\\nanalysis.\\n(a) Using the summarize() andsm.GLM() functions, determine the\\nestimated standard errors for the coefficients associated with\\nincomeandbalancein a multiple logistic regression model that\\nuses both predictors.\\n(b) Writeafunction, boot_fn() ,thattakesasinputthe Defaultdata\\nset as well as an index of the observations, and that outputs\\nthe coefficient estimates for incomeandbalancein the multiple\\nlogistic regression model.\\n(c) Following the bootstrap example in the lab, use your boot_fn()\\nfunction to estimate the standard errors of the logistic regression\\ncoefficients for incomeandbalance.\\n(d) Comment on the estimated standard errors obtained using the\\nsm.GLM() function and using the bootstrap.\\n7. InSections 5.1.2and5.1.3,wesawthatthe cross_validate() function\\ncan be used in order to compute the LOOCV test error estimate.\\nAlternatively, one could compute those quantities using just sm.GLM()\\nand the predict() method of the fitted model within a for loop. You\\nwill now take this approach in order to compute the LOOCV error\\nfor a simple logistic regression model on the Weeklydata set. Recall\\nthat in the context of classification problems, the LOOCV error is\\ngiven in ( 5.4).\\n(a) Fitalogisticregressionmodelthatpredicts Direction usingLag1\\nandLag2.\\n(b) Fitalogisticregressionmodelthatpredicts Direction usingLag1\\nandLag2using all but the first observation .\\n(c) Usethe model from (b) to predict the direction of the first obser-\\nvation. You can do this by predicting that the first observation\\nwill go up if P(Direction = \"Up\" |Lag1,Lag2)>0.5. Was this\\nobservation correctly classified?\\n(d) Write a for loop from i=1toi=n, wherenis the number of\\nobservations in the data set, that performs each of the following\\nsteps:', metadata={'source': './documents/ISLP_website.pdf', 'page': 233}), Document(page_content='5.4 Exercises 227i. Fit a logistic regression model using all but theith obser-vation to predictDirectionusingLag1andLag2.ii. Compute the posterior probability of the market moving upfor theith observation.iii. Usetheposteriorprobabilityfortheithobservationinorderto predict whether or not the market moves up.iv. Determine whether or not an error was made in predictingthe direction for theith observation. If an error was made,then indicate this as a1, and otherwise indicate it as a0.(e) Take the average of thennumbers obtained in (d)iv in order toobtain the LOOCV estimate for the test error. Comment on theresults.8. We will now perform cross-validation on a simulated data set.(a) Generate a simulated data set as follows:rng = np.random.default_rng(1)x=r n g . n o r m a l ( s i z e = 1 0 0 )y=x-2*x * * 2+r n g . n o r m a l ( s i z e = 1 0 0 )In this data set, what isnand what isp? Write out the modelused to generate the data in equation form.(b) Createa scatterplot ofXagainstY.Commenton what youfind.(c) Set a random seed, and then compute the LOOCV errors thatresult from fitting the following four models using least squares:i.Y=β0+β1X+ϵii.Y=β0+β1X+β2X2+ϵiii.Y=β0+β1X+β2X2+β3X3+ϵiv.Y=β0+β1X+β2X2+β3X3+β4X4+ϵ.Note you may find it helpful to use thedata.frame()functionto create a single data set containing bothXandY.(d) Repeat (c) using another random seed, and report your results.Are your results the same as what you got in (c)? Why?(e) Which of the models in (c) had the smallest LOOCV error? Isthis what you expected? Explain your answer.(f) Comment on the statistical significance of the coefficient esti-mates that results from fitting each of the models in (c) usingleast squares. Do these results agree with the conclusions drawnbased on the cross-validation results?9. We will now consider theBostonhousing data set, from theISLPlibrary.(a) Based on this data set, provide an estimate for the populationmean ofmedv. Call this estimateˆµ.', metadata={'source': './documents/ISLP_website.pdf', 'page': 234}), Document(page_content=\"228 5. Resampling Methods\\n(b) Provide an estimate of the standard error of ˆµ. Interpret this\\nresult.\\nHint: We can compute the standard error of the sample mean by\\ndividing the sample standard deviation by the square root of the\\nnumber of observations.\\n(c) Now estimate the standard error of ˆµusing the bootstrap. How\\ndoes this compare to your answer from (b)?\\n(d) Based on your bootstrap estimate from (c), provide a 95% con-\\nfidence interval for the mean of medv. Compare it to the results\\nobtained by using Boston['medv'].std() and the two standard\\nerror rule ( 3.9).\\nHint: You can approximate a 95% confidence interval using the\\nformula[ˆµ−2SE(ˆ µ),ˆµ+ 2SE(ˆ µ)].\\n(e) Basedonthisdataset,provideanestimate, ˆµmed,forthemedian\\nvalue of medvin the population.\\n(f) Wenowwouldliketoestimatethestandarderrorof ˆµmed.Unfor-\\ntunately, there is no simple formula for computing the standard\\nerror of the median. Instead, estimate the standard error of the\\nmedian using the bootstrap. Comment on your findings.\\n(g) Based on this data set, provide an estimate for the tenth per-\\ncentile of medvin Boston census tracts. Call this quantity ˆµ0.1.\\n(You can use the np.percentile() function.)np.\\npercentile() (h) Use the bootstrap to estimate the standard error of ˆµ0.1. Com-\\nment on your findings.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 235}), Document(page_content='6\\nLinear Model Selection\\nand Regularization\\nIn the regression setting, the standard linear model\\nY=β0+β1X1+···+βpXp+ϵ (6.1)\\nis commonly used to describe the relationship between a response Yand\\na set of variables X1,X2,...,X p. We have seen in Chapter 3that one\\ntypically fits this model using least squares.\\nIn the chapters that follow, we consider some approaches for extending\\nthe linear model framework. In Chapter 7we generalize ( 6.1) in order to\\naccommodate non-linear, but still additive, relationships, while in Chap-\\nters8and10we consider even more general non-linear models. However,\\nthe linear model has distinct advantages in terms of inference and, on real-\\nworld problems, is often surprisingly competitive in relation to non-linear\\nmethods. Hence, before moving to the non-linear world, we discuss in this\\nchaptersomewaysinwhichthesimplelinearmodelcanbeimproved,byre-\\nplacing plain least squares fitting with some alternative fitting procedures.\\nWhy might we want to use another fitting procedure instead of least\\nsquares? As we will see, alternative fitting procedures can yield better pre-\\ndiction accuracy andmodel interpretability .\\n•Prediction Accuracy : Provided that the true relationship between the\\nresponse and the predictors is approximately linear, the least squares\\nestimates will have low bias. If n≫p—that is, if n, the number of\\nobservations,ismuchlargerthan p,thenumberofvariables—thenthe\\nleast squares estimates tend to also have low variance, and hence will\\nperform well on test observations. However, if nis not much larger\\nthanp, then there can be a lot of variability in the least squares fit,\\nresulting in overfitting and consequently poor predictions on future\\nobservationsnotusedinmodeltraining.Andif p>n ,thenthereisno\\nlonger a unique least squares coefficient estimate: there are infinitely\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning , Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_6  229', metadata={'source': './documents/ISLP_website.pdf', 'page': 236}), Document(page_content='230 6. Linear Model Selection and Regularization\\nmany solutions. Each of these least squares solutions gives zero error\\non the training data, but typically very poor test set performance\\ndue to extremely high variance.1Byconstraining orshrinking the\\nestimated coefficients, we can often substantially reduce the variance\\nat the cost of a negligible increase in bias. This can lead to substantial\\nimprovements in the accuracy with which we can predict the response\\nfor observations not used in model training.\\n•Model Interpretability : It is often the case that some or many of the\\nvariables used in a multiple regression model are in fact not associ-\\nated with the response. Including such irrelevant variables leads to\\nunnecessary complexity in the resulting model. By removing these\\nvariables—that is, by setting the corresponding coefficient estimates\\nto zero—we can obtain a model that is more easily interpreted. Now\\nleast squares is extremely unlikely to yield any coefficient estimates\\nthat are exactly zero. In this chapter, we see some approaches for au-\\ntomatically performing feature selection orvariable selection —that is,feature\\nselection\\nvariable\\nselectionfor excluding irrelevant variables from a multiple regression model.\\nThere are many alternatives, both classical and modern, to using least\\nsquares to fit ( 6.1). In this chapter, we discuss three important classes of\\nmethods.\\n•Subset Selection . This approach involves identifying a subset of the p\\npredictors that we believe to be related to the response. We then fit\\na model using least squares on the reduced set of variables.\\n•Shrinkage . This approach involves fitting a model involving all ppre-\\ndictors.However,theestimatedcoefficientsareshrunkentowardszero\\nrelative to the least squares estimates. This shrinkage (also known as\\nregularization )hastheeffectofreducingvariance.Dependingonwhat\\ntype of shrinkage is performed, some of the coefficients may be esti-\\nmated to be exactly zero. Hence, shrinkage methods can also perform\\nvariable selection.\\n•Dimension Reduction . This approach involves projecting theppredic-\\ntors into an M-dimensional subspace, where M<p . This is achieved\\nby computing Mdifferent linear combinations , orprojections , of the\\nvariables. Then these Mprojections are used as predictors to fit a\\nlinear regression model by least squares.\\nIn the following sections we describe each of these approaches in greater de-\\ntail, along with their advantages and disadvantages. Although this chapter\\ndescribes extensions and modifications to the linear model for regression\\nseen in Chapter 3, the same concepts apply to other methods, such as the\\nclassification models seen in Chapter 4.\\n1Whenp≫n, the least squares solution that has the smallest sum of squared coeffi-\\ncients can sometimes perform quite well. See Section 10.8for a more detailed discussion.', metadata={'source': './documents/ISLP_website.pdf', 'page': 237}), Document(page_content='6.1 Subset Selection 231\\n6.1 Subset Selection\\nIn this section we consider some methods for selecting subsets of predictors.\\nThese include best subset and stepwise model selection procedures.\\n6.1.1 Best Subset Selection\\nTo perform best subset selection , we fit a separate least squares regressionbest subset\\nselectionforeachpossiblecombinationofthe ppredictors.Thatis,wefitall pmodels\\nthat contain exactly one predictor, all(p\\n2)\\n=p(p−1)/2models that contain\\nexactly two predictors, and so forth. We then look at all of the resulting\\nmodels, with the goal of identifying the one that is best.\\nThe problem of selecting the best model from among the 2ppossibilities\\nconsidered by best subset selection is not trivial. This is usually broken up\\ninto two stages, as described in Algorithm 6.1.\\nAlgorithm 6.1 Best subset selection\\n1. LetM0denote the null model , which contains no predictors. This\\nmodel simply predicts the sample mean for each observation.\\n2. Fork=1,2,...p :\\n(a) Fit all(p\\nk)\\nmodels that contain exactly kpredictors.\\n(b) Pick the best among these(p\\nk)\\nmodels, and call it Mk. Herebest\\nis defined as having the smallest RSS, or equivalently largest R2.\\n3. Select a single best model from among M0,...,Mpusing using the\\nprediction error on a validation set, Cp(AIC), BIC, or adjusted R2.\\nOr use the cross-validation method.\\nIn Algorithm 6.1, Step 2 identifies the best model (on the training data)\\nfor each subset size, in order to reduce the problem from one of 2ppossible\\nmodels to one of p+1possible models. In Figure 6.1, these models form\\nthe lower frontier depicted in red.\\nNow in order to select a single best model, we must simply choose among\\nthesep+1options. This task must be performed with care, because the\\nRSS of these p+1models decreases monotonically, and the R2increases\\nmonotonically, as the number of features included in the models increases.\\nTherefore, if we use these statistics to select the best model, then we will\\nalways end up with a model involving all of the variables. The problem is\\nthat a low RSS or a high R2indicates a model with a low training error,\\nwhereas we wish to choose a model that has a low testerror. (As shown in\\nChapter 2in Figures 2.9–2.11, training error tends to be quite a bit smaller\\nthan test error, and a low training error by no means guarantees a low test\\nerror.) Therefore, in Step 3, weuse the error on a validationset, Cp, BIC, or\\nadjusted R2in order to select among M0,M1,...,Mp. If cross-validation\\nis used to select the best model, then Step 2 is repeated on each training\\nfold, and the validation errors are averaged to select the best value of k.', metadata={'source': './documents/ISLP_website.pdf', 'page': 238}), Document(page_content='232 6. Linear Model Selection and Regularization\\n2 4 6 8 102e+07 4e+07 6e+07 8e+07Number of PredictorsResidual Sum of Squares2 4 6 8 100.0 0.2 0.4 0.6 0.8 1.0Number of PredictorsR2\\nFIGURE 6.1. For each possible model containing a subset of the ten predictors\\nin theCreditdata set, the RSS and R2are displayed. The red frontier tracks the\\nbestmodel for a given number of predictors, according to RSS and R2. Though\\nthe data set contains only ten predictors, the x-axis ranges from 1to11, since one\\nof the variables is categorical and takes on three values, leading to the creation of\\ntwo dummy variables.\\nThen the model Mkfit on the full training set is delivered for the chosen\\nk. These approaches are discussed in Section 6.1.3.\\nAn application of best subset selection is shown in Figure 6.1. Each\\nplotted point corresponds to a least squares regression model fit using a\\ndifferent subset of the 10 predictors in the Creditdata set, discussed in\\nChapter 3. Here the variable regionis a three-level qualitative variable,\\nand so is represented by two dummy variables, which are selected sepa-\\nrately in this case. Hence, there are a total of 11 possible variables which\\ncan be included in the model. We have plotted the RSS and R2statistics\\nfor each model, as a function of the number of variables. The red curves\\nconnect the best models for each model size, according to RSS or R2. The\\nfigure shows that, as expected, these quantities improve as the number of\\nvariablesincreases;however,fromthethree-variablemodelon,thereislittle\\nimprovement in RSS and R2as a result of including additional predictors.\\nAlthough we have presented best subset selection here for least squares\\nregression, the same ideas apply to other types of models, such as logistic\\nregression. In the case of logistic regression, instead of ordering models by\\nRSS in Step 2 of Algorithm 6.1, we instead use the deviance, a measuredeviancethat plays the role of RSS for a broader class of models. The deviance is\\nnegative two times the maximized log-likelihood; the smaller the deviance,\\nthe better the fit.\\nWhile best subset selection is a simple and conceptually appealing ap-\\nproach, it suffers from computational limitations. The number of possible\\nmodels that must be considered grows rapidly as pincreases. In general,\\nthere are 2pmodels that involve subsets of ppredictors. So if p= 10 ,\\nthen there are approximately 1,000 possible models to be considered, and if\\np= 20 ,thenthereareoveronemillionpossibilities!Consequently,bestsub-\\nsetselectionbecomescomputationallyinfeasibleforvaluesof pgreaterthan', metadata={'source': './documents/ISLP_website.pdf', 'page': 239}), Document(page_content='6.1 Subset Selection 233\\nAlgorithm 6.2 Forward stepwise selection\\n1. LetM0denote the nullmodel, which contains no predictors.\\n2. Fork=0,...,p −1:\\n(a) Consider all p−kmodels that augment the predictors in Mk\\nwith one additional predictor.\\n(b) Choose the bestamong these p−kmodels, and call it Mk+1.\\nHerebestis defined as having smallest RSS or highest R2.\\n3. Select a single best model from among M0,...,Mpusing the pre-\\ndiction error on a validation set, Cp(AIC), BIC, or adjusted R2. Or\\nuse the cross-validation method.\\naround40, even with extremely fast modern computers. There are compu-\\ntational shortcuts—so called branch-and-bound techniques—for eliminat-\\ning some choices, but these have their limitations as pgets large. They also\\nonly work for least squares linear regression. We present computationally\\nefficient alternatives to best subset selection next.\\n6.1.2 Stepwise Selection\\nFor computational reasons, best subset selection cannot be applied with\\nvery large p. Best subset selection may also suffer from statistical problems\\nwhenpislarge.Thelargerthesearchspace,thehigherthechanceoffinding\\nmodels that look good on the training data, even though they might not\\nhave any predictive power on future data. Thus an enormous search space\\ncan lead to overfitting and high variance of the coefficient estimates.\\nFor both of these reasons, stepwise methods, which explore a far more\\nrestricted set of models, are attractive alternatives to best subset selection.\\nForward Stepwise Selection\\nForward stepwise selection is a computationally efficient alternative to bestforward\\nstepwise\\nselectionsubset selection. While the best subset selection procedure considers all\\n2ppossible models containing subsets of the ppredictors, forward step-\\nwise considers a much smaller set of models. Forward stepwise selection\\nbegins with a model containing no predictors, and then adds predictors\\nto the model, one-at-a-time, until all of the predictors are in the model.\\nIn particular, at each step the variable that gives the greatest additional\\nimprovement to the fit is added to the model. More formally, the forward\\nstepwise selection procedure is given in Algorithm 6.2.\\nUnlike best subset selection, which involved fitting 2pmodels, forward\\nstepwise selection involves fitting one null model, along with p−kmodels\\nin thekth iteration, for k=0,...,p −1. This amounts to a total of 1+∑p−1\\nk=0(p−k)=1 + p(p+1)/2models. This is a substantial difference: when', metadata={'source': './documents/ISLP_website.pdf', 'page': 240}), Document(page_content='234 6. Linear Model Selection and Regularization\\n# Variables Best subset Forward stepwise\\nOne rating rating\\nTwo rating,income rating,income\\nThree rating,income,student rating ,income,student\\nFour cards,income rating,income,\\nstudent,limit student,limit\\nTABLE 6.1. The first four selected models for best subset selection and forward\\nstepwise selection on the Creditdata set. The first three models are identical but\\nthe fourth models differ.\\np= 20 , best subset selection requires fitting 1,048,576models, whereas\\nforward stepwise selection requires fitting only 211models.2\\nIn Step 2(b) of Algorithm 6.2, we must identify the bestmodel from\\namong those p−kthat augment Mkwith one additional predictor. We can\\ndo this by simply choosing the model with the lowest RSS or the highest\\nR2. However, in Step 3, we must identify the best model among a set of\\nmodels with different numbers of variables. This is more challenging, and\\nis discussed in Section 6.1.3.\\nForward stepwise selection’s computational advantage over best subset\\nselection is clear. Though forward stepwise tends to do well in practice,\\nit is not guaranteed to find the best possible model out of all 2pmod-\\nels containing subsets of the ppredictors. For instance, suppose that in a\\ngiven data set with p=3predictors, the best possible one-variable model\\ncontains X1, and the best possible two-variable model instead contains X2\\nandX3. Then forward stepwise selection will fail to select the best possible\\ntwo-variable model, because M1will contain X1, soM2must also contain\\nX1together with one additional variable.\\nTable6.1, which shows the first four selected models for best subset\\nand forward stepwise selection on the Creditdata set, illustrates this phe-\\nnomenon. Both best subset selection and forward stepwise selection choose\\nratingfor the best one-variable model and then include incomeandstudent\\nfor the two- and three-variable models. However, best subset selection re-\\nplacesratingbycardsin the four-variable model, while forward stepwise\\nselection must maintain ratingin its four-variable model. In this example,\\nFigure6.1indicates that there is not much difference between the three-\\nand four-variable models in terms of RSS, so either of the four-variable\\nmodels will likely be adequate.\\nForward stepwise selection can be applied even in the high-dimensional\\nsetting where n<p ; however, in this case, it is possible to construct sub-\\nmodelsM0,...,Mn−1only, since each submodel is fit using least squares,\\nwhich will not yield a unique solution if p≥n.\\nBackward Stepwise Selection\\nLike forward stepwise selection, backward stepwise selection provides anbackward\\nstepwise\\nselection 2Though forward stepwise selection considers p(p+ 1)/2+1 models, it performs a\\nguidedsearch over model space, and so the effective model space considered contains\\nsubstantially more than p(p+ 1)/2+1 models.', metadata={'source': './documents/ISLP_website.pdf', 'page': 241}), Document(page_content='6.1 Subset Selection 235\\nefficient alternative to best subset selection. However, unlike forward step-\\nwise selection, it begins with the full least squares model containing all p\\npredictors, and then iteratively removes the least useful predictor, one-at-\\na-time. Details are given in Algorithm 6.3.\\nAlgorithm 6.3 Backward stepwise selection\\n1. LetMpdenote the fullmodel, which contains all ppredictors.\\n2. Fork=p, p−1,...,1:\\n(a) Consider all kmodels that contain all but one of the predictors\\ninMk, for a total of k−1predictors.\\n(b) Choose the bestamong these kmodels, and call it Mk−1. Here\\nbestis defined as having smallest RSS or highest R2.\\n3. Select a single best model from among M0,...,Mpusing the pre-\\ndiction error on a validation set, Cp(AIC), BIC, or adjusted R2. Or\\nuse the cross-validation method.\\nLikeforwardstepwiseselection,thebackwardselectionapproachsearches\\nthrough only 1+p(p+1)/2models, and so can be applied in settings where\\npis too large to apply best subset selection.3Also like forward stepwise\\nselection, backward stepwise selection is not guaranteed to yield the best\\nmodel containing a subset of the ppredictors.\\nBackward selection requires that the number of samples nis larger than\\nthe number of variables p(so that the full model can be fit). In contrast,\\nforward stepwise can be used even when n<p , and so is the only viable\\nsubset method when pis very large.\\nHybrid Approaches\\nThe best subset, forward stepwise, and backward stepwise selection ap-\\nproaches generally give similar but not identical models. As another al-\\nternative, hybrid versions of forward and backward stepwise selection are\\navailable, in which variables are added to the model sequentially, in analogy\\nto forward selection. However, after adding each new variable, the method\\nmay also remove any variables that no longer provide an improvement in\\nthe model fit. Such an approach attempts to more closely mimic best sub-\\nset selection while retaining the computational advantages of forward and\\nbackward stepwise selection.\\n6.1.3 Choosing the Optimal Model\\nBest subset selection, forward selection, and backward selection result in\\nthe creation of a set of models, each of which contains a subset of the p\\n3Likeforwardstepwiseselection, backwardstepwiseselection performs a guidedsearch\\nover model space, and so effectively considers substantially more than 1+p(p+ 1)/2\\nmodels.', metadata={'source': './documents/ISLP_website.pdf', 'page': 242}), Document(page_content='236 6. Linear Model Selection and Regularization\\npredictors. To apply these methods, we need a way to determine which of\\nthese models is best. As we discussed in Section 6.1.1, the model containing\\nall of the predictors will always have the smallest RSS and the largest R2,\\nsince these quantities are related to the training error. Instead, we wish to\\nchoose a model with a low test error. As is evident here, and as we show\\nin Chapter 2, the training error can be a poor estimate of the test error.\\nTherefore, RSS and R2are not suitable for selecting the best model among\\na collection of models with different numbers of predictors.\\nIn order to select the best model with respect to test error, we need to\\nestimate this test error. There are two common approaches:\\n1. We can indirectly estimate test error by making an adjustment to the\\ntraining error to account for the bias due to overfitting.\\n2. We can directlyestimate the test error, using either a validation set\\napproach or a cross-validation approach, as discussed in Chapter 5.\\nWe consider both of these approaches below.\\nCp, AIC, BIC, and Adjusted R2\\nWe show in Chapter 2that the training set MSE is generally an under-\\nestimate of the test MSE. (Recall that MSE = RSS /n.) This is because\\nwhen we fit a model to the training data using least squares, we specifi-\\ncally estimate the regression coefficients such that the training RSS (but\\nnot the test RSS) is as small as possible. In particular, the training error\\nwill decrease as more variables are included in the model, but the test error\\nmay not. Therefore, training set RSS and training set R2cannot be used\\nto select from among a set of models with different numbers of variables.\\nHowever, a number of techniques for adjusting the training error for the\\nmodelsizeareavailable.Theseapproachescanbeusedtoselectamongaset\\nof models with different numbers of variables. We now consider four such\\napproaches: Cp,Akaike information criterion (AIC),Bayesian informationCp\\nAkaike\\ninformation\\ncriterioncriterion (BIC),and adjusted R2.Figure6.2displaysCp,BIC,andadjusted\\nBayesian\\ninformation\\ncriterion\\nadjusted R2R2for the best model of each size produced by best subset selection on the\\nCreditdata set.\\nFor a fitted least squares model containing dpredictors, the Cpestimate\\nof test MSE is computed using the equation\\nCp=1\\nn(\\nRSS + 2 dˆσ2)\\n, (6.2)\\nwhereˆσ2is an estimate of the variance of the error ϵassociated with each\\nresponse measurement in ( 6.1).4Typically ˆσ2is estimated using the full\\nmodel containing all predictors. Essentially, the Cpstatistic adds a penalty\\nof2dˆσ2to the training RSS in order to adjust for the fact that the training\\nerror tends to underestimate the test error. Clearly, the penalty increases as\\nthe number of predictors in the model increases; this is intended to adjust\\n4Mallow’s Cpis sometimes defined as C′\\np= RSS /ˆσ2+2d−n. This is equivalent to\\nthe definition given above in the sense that Cp=1\\nnˆσ2(C′\\np+n), and so the model with\\nsmallest Cpalso has smallest C′\\np.', metadata={'source': './documents/ISLP_website.pdf', 'page': 243}), Document(page_content='6.1 Subset Selection 237\\n2 4 6 8 1010000 15000 20000 25000 30000Number of PredictorsCp\\n2 4 6 8 1010000 15000 20000 25000 30000Number of PredictorsBIC\\n2 4 6 8 100.86 0.88 0.90 0.92 0.94 0.96Number of PredictorsAdjusted R2 \\nFIGURE 6.2. Cp, BIC, and adjusted R2are shown for the best models of each\\nsize for the Creditdata set (the lower frontier in Figure 6.1).Cpand BIC are\\nestimates of test MSE. In the middle plot we see that the BIC estimate of test\\nerror shows an increase after four variables are selected. The other two plots are\\nrather flat after four variables are included.\\nfor the corresponding decrease in training RSS. Though it is beyond the\\nscope of this book, one can show that if ˆσ2is an unbiased estimate of σ2in\\n(6.2), thenCpis an unbiased estimate of test MSE. As a consequence, the\\nCpstatistic tends to take on a small value for models with a low test error,\\nso when determining which of a set of models is best, we choose the model\\nwith the lowest Cpvalue. In Figure 6.2,Cpselects the six-variable model\\ncontaining the predictors income,limit,rating,cards,ageandstudent.\\nThe AIC criterion is defined for a large class of models fit by maximum\\nlikelihood. In the case of the model ( 6.1) with Gaussian errors, maximum\\nlikelihood and least squares are the same thing. In this case AIC is given by\\nAIC =1\\nn(\\nRSS + 2 dˆσ2)\\n,\\nwhere, for simplicity, we have omitted irrelevant constants.5Hence for least\\nsquares models, Cpand AIC are proportional to each other, and so only\\nCpis displayed in Figure 6.2.\\nBIC is derived from a Bayesian point of view, but ends up looking similar\\ntoCp(and AIC) as well. For the least squares model with dpredictors, the\\nBIC is, up to irrelevant constants, given by\\nBIC=1\\nn(\\nRSS + log( n)dˆσ2)\\n. (6.3)\\nLikeCp, the BIC will tend to take on a small value for a model with a\\nlow test error, and so generally we select the model that has the lowest\\nBIC value. Notice that BIC replaces the 2dˆσ2used byCpwith alog(n)dˆσ2\\nterm, where nis the number of observations. Since logn>2for anyn>7,\\n5There are two formulas for AIC for least squares regression. The formula that we\\nprovide here requires an expression for σ2, which we obtain using the full model con-\\ntaining all predictors. The second formula is appropriate when σ2is unknown and we do\\nnot want to explicitly estimate it; that formula has a log(RSS)term instead of an RSS\\nterm. Detailed derivations of these two formulas are outside of the scope of this book.', metadata={'source': './documents/ISLP_website.pdf', 'page': 244}), Document(page_content='238 6. Linear Model Selection and Regularization\\nthe BIC statistic generally places a heavier penalty on models with many\\nvariables, and hence results in the selection of smaller models than Cp.\\nIn Figure 6.2, we see that this is indeed the case for the Creditdata set;\\nBIC chooses a model that contains only the four predictors income,limit,\\ncards, andstudent. In this case the curves are very flat and so there does\\nnot appear to be much difference in accuracy between the four-variable and\\nsix-variable models.\\nTheadjusted R2statisticisanotherpopularapproachforselectingamong\\na set of models that contain different numbers of variables. Recall from\\nChapter 3that the usual R2is defined as 1−RSS/TSS, where TSS =∑(yi−y)2is thetotal sum of squares for the response. Since RSS always\\ndecreases as more variables are added to the model, the R2always increases\\nas more variables are added. For a least squares model with dvariables,\\nthe adjusted R2statistic is calculated as\\nAdjusted R2=1−RSS/(n−d−1)\\nTSS/(n−1). (6.4)\\nUnlikeCp, AIC, and BIC, for which a smallvalue indicates a model with\\na low test error, a largevalue of adjusted R2indicates a model with a\\nsmall test error. Maximizing the adjusted R2is equivalent to minimizing\\nRSS\\nn−d−1.WhileRSSalwaysdecreasesasthenumberofvariablesinthemodel\\nincreases,RSS\\nn−d−1may increase or decrease, due to the presence of din the\\ndenominator.\\nThe intuition behind the adjusted R2is that once all of the correct\\nvariableshavebeenincludedinthemodel,addingadditional noisevariables\\nwill lead to only a very small decrease in RSS. Since adding noise variables\\nleads to an increase in d, such variables will lead to an increase inRSS\\nn−d−1,\\nand consequently a decrease in the adjusted R2. Therefore, in theory, the\\nmodel with the largest adjusted R2will have only correct variables and\\nno noise variables. Unlike the R2statistic, the adjusted R2statistic pays\\na pricefor the inclusion of unnecessary variables in the model. Figure 6.2\\ndisplays the adjusted R2for theCreditdata set. Using this statistic results\\nin the selection of a model that contains seven variables, adding ownto the\\nmodel selected by Cpand AIC.\\nCp, AIC, and BIC all have rigorous theoretical justifications that are\\nbeyond the scope of this book. These justifications rely on asymptotic ar-\\nguments (scenarios where the sample size nis very large). Despite its pop-\\nularity, and even though it is quite intuitive, the adjusted R2is not as well\\nmotivated in statistical theory as AIC, BIC, and Cp. All of these measures\\nare simple to use and compute. Here we have presented their formulas in\\nthe case of a linear model fit using least squares; however, AIC and BIC\\ncan also be defined for more general types of models.\\nValidation and Cross-Validation\\nAs an alternative to the approaches just discussed, we can directly esti-\\nmate the test error using the validation set and cross-validation methods\\ndiscussed in Chapter 5. We can compute the validation set error or the\\ncross-validation error for each model under consideration, and then select', metadata={'source': './documents/ISLP_website.pdf', 'page': 245}), Document(page_content='6.1 Subset Selection 239\\n2 4 6 8 10100 120 140 160 180 200 220Number of PredictorsSquare Root of BIC2 4 6 8 10100 120 140 160 180 200 220Number of PredictorsValidation Set Error2 4 6 8 10100 120 140 160 180 200 220Number of PredictorsCross−Validation ErrorFIGURE 6.3. For theCreditdata set, three quantities are displayed for the\\nbest model containing dpredictors, for dranging from 1to11. The overall best\\nmodel, based on each of these quantities, is shown as a blue cross. Left:Square\\nroot of BIC. Center:Validation set errors. Right:Cross-validation errors.\\nthe model for which the resulting estimated test error is smallest. This pro-\\ncedure has an advantage relative to AIC, BIC, Cp, and adjusted R2, in that\\nit provides a direct estimate of the test error, and makes fewer assumptions\\nabout the true underlying model. It can also be used in a wider range of\\nmodel selection tasks, even in cases where it is hard to pinpoint the model\\ndegrees of freedom (e.g. the number of predictors in the model) or hard\\nto estimate the error variance σ2. Note that when cross-validation is used,\\nthe sequence of models Mkin Algorithms 6.1–6.3is determined separately\\nfor each training fold, and the validation errors are averaged over all folds\\nfor each model size k. This means, for example with best-subset regression,\\nthatMk, the best subset of size k, can differ across the folds. Once the\\nbest size kis chosen, we find the best model of that size on the full data\\nset.\\nIn the past, performing cross-validation was computationally prohibitive\\nfor many problems with large pand/or large n, and so AIC, BIC, Cp,\\nand adjusted R2were more attractive approaches for choosing among a\\nset of models. However, nowadays with fast computers, the computations\\nrequired to perform cross-validation are hardly ever an issue. Thus, cross-\\nvalidation is a very attractive approach for selecting from among a number\\nof models under consideration.\\nFigure6.3displays, as a function of d, the BIC, validation set errors, and\\ncross-validation errors on the Creditdata, for the best d-variable model.\\nThe validation errors were calculated by randomly selecting three-quarters\\nof the observations as the training set, and the remainder as the valida-\\ntion set. The cross-validation errors were computed using k= 10 folds.\\nIn this case, the validation and cross-validation methods both result in a\\nsix-variable model. However, all three approaches suggest that the four-,\\nfive-, and six-variable models are roughly equivalent in terms of their test\\nerrors.\\nIn fact, the estimated test error curves displayed in the center and right-\\nhandpanelsofFigure 6.3arequiteflat.Whileathree-variablemodelclearly\\nhaslowerestimated test error than a two-variablemodel, the estimated test\\nerrors of the 3- to 11-variable models are quite similar. Furthermore, if we', metadata={'source': './documents/ISLP_website.pdf', 'page': 246}), Document(page_content='240 6. Linear Model Selection and Regularization\\nrepeated the validation set approach using a different split of the data into\\na training set and a validation set, or if we repeated cross-validation using\\na different set of cross-validation folds, then the precise model with the\\nlowest estimated test error would surely change. In this setting, we can\\nselect a model using the one-standard-error rule . We first calculate theone-\\nstandard-\\nerror\\nrulestandard error of the estimated test MSE for each model size, and then\\nselect the smallest model for which the estimated test error is within one\\nstandard error of the lowest point on the curve. The rationale here is that\\nif a set of models appear to be more or less equally good, then we might\\nas well choose the simplest model—that is, the model with the smallest\\nnumber of predictors. In this case, applying the one-standard-error rule\\nto the validation set or cross-validation approach leads to selection of the\\nthree-variable model.\\n6.2 Shrinkage Methods\\nThe subset selection methods described in Section 6.1involve using least\\nsquares to fit a linear model that contains a subset of the predictors. As an\\nalternative, we can fit a model containing all ppredictors using a technique\\nthatconstrains orregularizes the coefficient estimates, or equivalently, that\\nshrinksthe coefficient estimates towards zero. It may not be immediately\\nobvious why such a constraint should improve the fit, but it turns out that\\nshrinking the coefficient estimates can significantly reduce their variance.\\nThe two best-known techniques for shrinking the regression coefficients\\ntowards zero are ridge regression and thelasso.\\n6.2.1 Ridge Regression\\nRecall from Chapter 3that the least squares fitting procedure estimates\\nβ0,β1,...,βpusing the values that minimize\\nRSS =n∑\\ni=1\\uf8eb\\n\\uf8edyi−β0−p∑\\nj=1βjxij\\uf8f6\\n\\uf8f82\\n.\\nRidge regression is very similar to least squares, except that the coefficientsridge\\nregressionare estimated by minimizing a slightly different quantity. In particular, the\\nridge regression coefficient estimates ˆβRare the values that minimize\\nn∑\\ni=1\\uf8eb\\n\\uf8edyi−β0−p∑\\nj=1βjxij\\uf8f6\\n\\uf8f82\\n+λp∑\\nj=1β2\\nj= RSS + λp∑\\nj=1β2\\nj, (6.5)\\nwhereλ≥0is atuning parameter , to be determined separately. Equa-tuning\\nparametertion6.5trades off two different criteria. As with least squares, ridge regres-\\nsion seeks coefficient estimates that fit the data well, by making the RSS\\nsmall. However, the second term, λ∑\\njβ2\\nj, called a shrinkage penalty , isshrinkage\\npenaltysmall when β1,...,βpare close to zero, and so it has the effect of shrinking\\nthe estimates of βjtowards zero. The tuning parameter λserves to control', metadata={'source': './documents/ISLP_website.pdf', 'page': 247}), Document(page_content='6.2 Shrinkage Methods 241\\n1e−02 1e+00 1e+02 1e+04−300 −100 0 100 200 300 400Standardized CoefficientsIncomeLimitRatingStudent\\n0.0 0.2 0.4 0.6 0.8 1.0−300 −100 0 100 200 300 400Standardized Coefficientsλ∥ˆβRλ∥2/∥ˆβ∥2FIGURE 6.4. The standardized ridge regression coefficients are displayed for\\ntheCreditdata set, as a function of λand∥ˆβR\\nλ∥2/∥ˆβ∥2.\\nthe relative impact of these two terms on the regression coefficient esti-\\nmates. When λ=0, the penalty term has no effect, and ridge regression\\nwill produce the least squares estimates. However, as λ→∞, the impact of\\nthe shrinkage penalty grows, and the ridge regression coefficient estimates\\nwill approach zero. Unlike least squares, which generates only one set of co-\\nefficient estimates, ridge regression will produce a different set of coefficient\\nestimates, ˆβR\\nλ, for each value of λ. Selecting a good value for λis critical;\\nwe defer this discussion to Section 6.2.3, where we use cross-validation.\\nNote that in ( 6.5), the shrinkage penalty is applied to β1,...,βp, but\\nnot to the intercept β0. We want to shrink the estimated association of\\neach variable with the response; however, we do not want to shrink the\\nintercept, which is simply a measure of the mean value of the response\\nwhenxi1=xi2=...=xip=0. If we assume that the variables—that is,\\nthe columns of the data matrix X—have been centered to have mean zero\\nbefore ridge regression is performed, then the estimated intercept will take\\nthe form ˆβ0=¯y=∑n\\ni=1yi/n.\\nAn Application to the Credit Data\\nIn Figure 6.4, the ridge regression coefficient estimates for the Creditdata\\nset are displayed. In the left-hand panel, each curve corresponds to the\\nridge regression coefficient estimate for one of the ten variables, plotted\\nas a function of λ. For example, the black solid line represents the ridge\\nregression estimate for the incomecoefficient, as λis varied. At the extreme\\nleft-hand side of the plot, λis essentially zero, and so the corresponding\\nridge coefficient estimates are the same as the usual least squares esti-\\nmates. But as λincreases, the ridge coefficient estimates shrink towards\\nzero. When λis extremely large, then all of the ridge coefficient estimates\\nare basically zero; this corresponds to the null model that contains no pre-\\ndictors. In this plot, the income,limit,rating, andstudentvariables are\\ndisplayed in distinct colors, since these variables tend to have by far the\\nlargest coefficient estimates. While the ridge coefficient estimates tend to\\ndecrease in aggregate as λincreases, individual coefficients, such as rating\\nandincome, may occasionally increase as λincreases.', metadata={'source': './documents/ISLP_website.pdf', 'page': 248}), Document(page_content='242 6. Linear Model Selection and Regularization\\nThe right-hand panel of Figure 6.4displays the same ridge coefficient\\nestimates as the left-hand panel, but instead of displaying λon thex-axis,\\nwe now display ∥ˆβR\\nλ∥2/∥ˆβ∥2, where ˆβdenotes the vector of least squares\\ncoefficient estimates. The notation ∥β∥2denotes the ℓ2norm(pronouncedℓ2norm\\n“ell 2”) of a vector, and is defined as ∥β∥2=√∑p\\nj=1βj2. It measures the\\ndistance of βfrom zero. As λincreases, the ℓ2norm of ˆβR\\nλwillalways\\ndecrease, and so will ∥ˆβR\\nλ∥2/∥ˆβ∥2. The latter quantity ranges from 1 (when\\nλ=0, in which case the ridge regression coefficient estimate is the same\\nas the least squares estimate, and so their ℓ2norms are the same) to 0\\n(whenλ=∞, in which case the ridge regression coefficient estimate is a\\nvector of zeros, with ℓ2norm equal to zero). Therefore, we can think of the\\nx-axis in the right-hand panel of Figure 6.4as the amount that the ridge\\nregression coefficient estimates have been shrunken towards zero; a small\\nvalue indicates that they have been shrunken very close to zero.\\nThe standard least squares coefficient estimates discussed in Chapter 3\\narescale equivariant : multiplying Xjby a constant csimply leads to ascale\\nequivariantscaling of the least squares coefficient estimates by a factor of 1/c. In other\\nwords, regardless of how the jth predictor is scaled, Xjˆβjwill remain the\\nsame. In contrast, the ridge regression coefficient estimates can change sub-\\nstantially when multiplying a given predictor by a constant. For instance,\\nconsider the incomevariable, which is measured in dollars. One could rea-\\nsonably have measured income in thousands of dollars, which would result\\ninareductionintheobservedvaluesof incomebyafactorof1,000.Nowdue\\nto the sum of squared coefficients term in the ridge regression formulation\\n(6.5), such a change in scale will not simply cause the ridge regression co-\\nefficient estimate for incometo change by a factor of 1,000. In other words,\\nXjˆβR\\nj,λwill depend not only on the value of λ, but also on the scaling of the\\njth predictor. In fact, the value of XjˆβR\\nj,λmay even depend on the scaling\\nof theotherpredictors! Therefore, it is best to apply ridge regression after\\nstandardizing the predictors , using the formula\\n˜xij=xij√\\n1\\nn∑n\\ni=1(xij−xj)2, (6.6)\\nso that they are all on the same scale. In ( 6.6), the denominator is the\\nestimated standard deviation of the jth predictor. Consequently, all of the\\nstandardized predictors will have a standard deviation of one. As a re-\\nsult the final fit will not depend on the scale on which the predictors are\\nmeasured. In Figure 6.4, they-axis displays the standardized ridge regres-\\nsion coefficient estimates—that is, the coefficient estimates that result from\\nperforming ridge regression using standardized predictors.\\nWhy Does Ridge Regression Improve Over Least Squares?\\nRidgeregression’sadvantageoverleastsquaresisrootedinthe bias-variance\\ntrade-off. Asλincreases, the flexibility of the ridge regression fit decreases,\\nleading to decreased variance but increased bias. This is illustrated in the\\nleft-hand panel of Figure 6.5, using a simulated data set containing p= 45\\npredictors and n= 50 observations. The green curve in the left-hand panel', metadata={'source': './documents/ISLP_website.pdf', 'page': 249}), Document(page_content='6.2 Shrinkage Methods 243\\n1e−01 1e+01 1e+030 10 20 30 40 50 60Mean Squared Error0.0 0.2 0.4 0.6 0.8 1.00 10 20 30 40 50 60Mean Squared Errorλ∥ˆβRλ∥2/∥ˆβ∥2FIGURE 6.5. Squared bias (black), variance (green), and test mean squared\\nerror (purple) for the ridge regression predictions on a simulated data set, as a\\nfunction of λand∥ˆβR\\nλ∥2/∥ˆβ∥2. The horizontal dashed lines indicate the minimum\\npossible MSE. The purple crosses indicate the ridge regression models for which\\nthe MSE is smallest.\\nof Figure 6.5displays the variance of the ridge regression predictions as a\\nfunction of λ. At the least squares coefficient estimates, which correspond\\nto ridge regression with λ=0, the variance is high but there is no bias. But\\nasλincreases, the shrinkage of the ridge coefficient estimates leads to a\\nsubstantial reduction in the variance of the predictions, at the expense of a\\nslight increase in bias. Recall that the test mean squared error (MSE), plot-\\nted in purple, is closely related to the variance plus the squared bias. For\\nvalues of λup to about 10, the variance decreases rapidly, with very little\\nincrease in bias, plotted in black. Consequently, the MSE drops consider-\\nably asλincreases from 0to10. Beyond this point, the decrease in variance\\ndue to increasing λslows, and the shrinkage on the coefficients causes them\\nto be significantly underestimated, resulting in a large increase in the bias.\\nThe minimum MSE is achieved at approximately λ= 30 . Interestingly,\\nbecause of its high variance, the MSE associated with the least squares\\nfit, when λ=0, is almost as high as that of the null model for which all\\ncoefficient estimates are zero, when λ=∞. However, for an intermediate\\nvalue ofλ, the MSE is considerably lower.\\nThe right-hand panel of Figure 6.5displays the same curves as the left-\\nhand panel, this time plotted against the ℓ2norm of the ridge regression\\ncoefficient estimates divided by the ℓ2norm of the least squares estimates.\\nNow as we move from left to right, the fits become more flexible, and so\\nthe bias decreases and the variance increases.\\nIn general, in situations where the relationship between the response\\nand the predictors is close to linear, the least squares estimates will have\\nlow bias but may have high variance. This means that a small change in\\nthe training data can cause a large change in the least squares coefficient\\nestimates. In particular, when the number of variables pis almost as large\\nas the number of observations n, as in the example in Figure 6.5, the\\nleast squares estimates will be extremely variable. And if p>n , then the\\nleast squares estimates do not even have a unique solution, whereas ridge\\nregression can still perform well by trading off a small increase in bias for a', metadata={'source': './documents/ISLP_website.pdf', 'page': 250}), Document(page_content='244 6. Linear Model Selection and Regularization\\nlarge decrease in variance. Hence, ridge regression works best in situations\\nwhere the least squares estimates have high variance.\\nRidge regression also has substantial computational advantages over best\\nsubset selection, which requires searching through 2pmodels. As we dis-\\ncussed previously, even for moderate values of p, such a search can be\\ncomputationally infeasible. In contrast, for any fixed value of λ, ridge re-\\ngression only fits a single model, and the model-fitting procedure can be\\nperformed quite quickly. In fact, one can show that the computations re-\\nquired to solve ( 6.5),simultaneously for all values of λ, are almost identical\\nto those for fitting a model using least squares.\\n6.2.2 The Lasso\\nRidge regression does have one obvious disadvantage. Unlike best subset,\\nforward stepwise, and backward stepwise selection, which will generally\\nselect models that involve just a subset of the variables, ridge regression\\nwill include all ppredictors in the final model. The penalty λ∑β2\\njin (6.5)\\nwillshrinkallofthecoefficientstowardszero,butitwillnotsetanyofthem\\nexactly to zero (unless λ=∞). This may not be a problem for prediction\\naccuracy, but it can create a challenge in model interpretation in settings in\\nwhich the number of variables pis quite large. For example, in the Credit\\ndata set, it appears that the most important variables are income,limit,\\nrating, andstudent. So we might wish to build a model including just\\nthese predictors. However, ridge regression will always generate a model\\ninvolving all ten predictors. Increasing the value of λwill tend to reduce\\nthe magnitudes of the coefficients, but will not result in exclusion of any of\\nthe variables.\\nThelassois a relatively recent alternative to ridge regression that over-lassocomes this disadvantage. The lasso coefficients, ˆβL\\nλ, minimize the quantity\\nn∑\\ni=1\\uf8eb\\n\\uf8edyi−β0−p∑\\nj=1βjxij\\uf8f6\\n\\uf8f82\\n+λp∑\\nj=1|βj|= RSS + λp∑\\nj=1|βj|. (6.7)\\nComparing ( 6.7) to (6.5), we see that the lasso and ridge regression have\\nsimilar formulations. The only difference is that the β2\\njterm in the ridge\\nregression penalty ( 6.5) has been replaced by |βj|in the lasso penalty ( 6.7).\\nIn statistical parlance, the lasso uses an ℓ1(pronounced “ell 1”) penalty\\ninstead of an ℓ2penalty. The ℓ1norm of a coefficient vector βis given by\\n∥β∥1=∑|βj|.\\nAs with ridge regression, the lasso shrinks the coefficient estimates to-\\nwards zero. However, in the case of the lasso, the ℓ1penalty has the effect\\nof forcing some of the coefficient estimates to be exactly equal to zero when\\nthetuningparameter λissufficientlylarge.Hence,muchlikebestsubsetse-\\nlection, the lasso performs variable selection . As a result, models generated\\nfrom the lasso are generally much easier to interpret than those produced\\nby ridge regression. We say that the lasso yields sparsemodels—that is,sparse\\nmodels that involve only a subset of the variables. As in ridge regression,\\nselecting a good value of λfor the lasso is critical; we defer this discussion\\nto Section 6.2.3, where we use cross-validation.', metadata={'source': './documents/ISLP_website.pdf', 'page': 251}), Document(page_content='6.2 Shrinkage Methods 245\\n20 50 100 200 500 2000 5000−200 0 100 200 300 400Standardized Coefficients0.0 0.2 0.4 0.6 0.8 1.0−300 −100 0 100 200 300 400Standardized CoefficientsIncomeLimitRatingStudentλ∥ˆβLλ∥1/∥ˆβ∥1FIGURE 6.6. The standardized lasso coefficients on the Creditdata set are\\nshown as a function of λand∥ˆβL\\nλ∥1/∥ˆβ∥1.\\nAs an example, consider the coefficient plots in Figure 6.6, which are gen-\\nerated from applying the lasso to the Creditdata set. When λ=0, then\\nthe lasso simply gives the least squares fit, and when λbecomes sufficiently\\nlarge, the lasso gives the null model in which all coefficient estimates equal\\nzero. However, in between these two extremes, the ridge regression and\\nlasso models are quite different from each other. Moving from left to right\\nin the right-hand panel of Figure 6.6, we observe that at first the lasso re-\\nsults in a model that contains only the ratingpredictor. Then studentand\\nlimitenter the model almost simultaneously, shortly followed by income.\\nEventually, the remaining variables enter the model. Hence, depending on\\nthe value of λ, the lasso can produce a model involving any number of vari-\\nables. In contrast, ridge regression will always include all of the variables in\\nthe model, although the magnitude of the coefficient estimates will depend\\nonλ.\\nAnother Formulation for Ridge Regression and the Lasso\\nOne can show that the lasso and ridge regression coefficient estimates solve\\nthe problems\\nminimize\\nβ\\uf8f1\\n\\uf8f4\\uf8f2\\n\\uf8f4\\uf8f3n∑\\ni=1\\uf8eb\\n\\uf8edyi−β0−p∑\\nj=1βjxij\\uf8f6\\n\\uf8f82\\uf8fc\\n\\uf8f4\\uf8fd\\n\\uf8f4\\uf8fesubject top∑\\nj=1|βj|≤s\\n(6.8)\\nand\\nminimize\\nβ\\uf8f1\\n\\uf8f4\\uf8f2\\n\\uf8f4\\uf8f3n∑\\ni=1\\uf8eb\\n\\uf8edyi−β0−p∑\\nj=1βjxij\\uf8f6\\n\\uf8f82\\uf8fc\\n\\uf8f4\\uf8fd\\n\\uf8f4\\uf8fesubject top∑\\nj=1β2\\nj≤s,\\n(6.9)\\nrespectively. In other words, for every value of λ, there is some ssuch that\\nthe Equations ( 6.7) and (6.8) will give the same lasso coefficient estimates.\\nSimilarly, for every value of λthere is a corresponding ssuch that Equa-\\ntions(6.5)and(6.9)willgivethesameridgeregressioncoefficientestimates.', metadata={'source': './documents/ISLP_website.pdf', 'page': 252}), Document(page_content='246 6. Linear Model Selection and Regularization\\nWhenp=2, then (6.8) indicates that the lasso coefficient estimates have\\nthe smallest RSS out of all points that lie within the diamond defined by\\n|β1|+|β2|≤s. Similarly, the ridge regression estimates have the smallest\\nRSS out of all points that lie within the circle defined by β2\\n1+β2\\n2≤s.\\nWecanthinkof( 6.8)asfollows.Whenweperformthelassowearetrying\\nto find the set of coefficient estimates that lead to the smallest RSS, subject\\nto the constraint that there is a budgetsfor how large∑p\\nj=1|βj|can be.\\nWhensis extremely large, then this budget is not very restrictive, and so\\nthe coefficient estimates can be large. In fact, if sis large enough that the\\nleast squares solution falls within the budget, then ( 6.8) will simply yield\\nthe least squares solution. In contrast, if sis small, then∑p\\nj=1|βj|must be\\nsmall in order to avoid violating the budget. Similarly, ( 6.9) indicates that\\nwhen we perform ridge regression, we seek a set of coefficient estimates\\nsuch that the RSS is as small as possible, subject to the requirement that∑p\\nj=1β2\\njnot exceed the budget s.\\nThe formulations ( 6.8) and (6.9) reveal a close connection between the\\nlasso, ridge regression, and best subset selection. Consider the problem\\nminimize\\nβ\\uf8f1\\n\\uf8f4\\uf8f2\\n\\uf8f4\\uf8f3n∑\\ni=1\\uf8eb\\n\\uf8edyi−β0−p∑\\nj=1βjxij\\uf8f6\\n\\uf8f82\\uf8fc\\n\\uf8f4\\uf8fd\\n\\uf8f4\\uf8fesubject top∑\\nj=1I(βj̸= 0) ≤s.\\n(6.10)\\nHereI(βj̸= 0) isanindicatorvariable:ittakesonavalueof1if βj̸=0,and\\nequals zero otherwise. Then ( 6.10) amounts to finding a set of coefficient\\nestimates such that RSS is as small as possible, subject to the constraint\\nthat no more than scoefficients can be nonzero. The problem ( 6.10) is\\nequivalent to best subset selection. Unfortunately, solving ( 6.10) is com-\\nputationally infeasible when pis large, since it requires considering all(p\\ns)\\nmodels containing spredictors. Therefore, we can interpret ridge regression\\nand the lasso as computationally feasible alternatives to best subset selec-\\ntion that replace the intractable form of the budget in ( 6.10) with forms\\nthat are much easier to solve. Of course, the lasso is much more closely\\nrelated to best subset selection, since the lasso performs feature selection\\nforssufficiently small in ( 6.8), while ridge regression does not.\\nThe Variable Selection Property of the Lasso\\nWhy is it that the lasso, unlike ridge regression, results in coefficient esti-\\nmates that are exactly equal to zero? The formulations ( 6.8) and (6.9) can\\nbe used to shed light on the issue. Figure 6.7illustrates the situation. The\\nleast squares solution is marked as ˆβ, while the blue diamond and circle\\nrepresent the lasso and ridge regression constraints in ( 6.8) and (6.9), re-\\nspectively. If sis sufficiently large, then the constraint regions will contain\\nˆβ, and so the ridge regression and lasso estimates will be the same as the\\nleast squares estimates. (Such a large value of scorresponds to λ=0 in\\n(6.5) and (6.7).) However, in Figure 6.7the least squares estimates lie out-\\nside of the diamond and the circle, and so the least squares estimates are\\nnot the same as the lasso and ridge regression estimates.\\nEach of the ellipses centered around ˆβrepresents a contour: this meanscontourthat all of the points on a particular ellipse have the same RSS value. As', metadata={'source': './documents/ISLP_website.pdf', 'page': 253}), Document(page_content='6.2 Shrinkage Methods 247\\nFIGURE 6.7. Contours of the error and constraint functions for the lasso\\n(left)and ridge regression (right). The solid blue areas are the constraint regions,\\n|β1|+|β2|≤sandβ2\\n1+β2\\n2≤s, while the red ellipses are the contours of the RSS.\\nthe ellipses expand away from the least squares coefficient estimates, the\\nRSS increases. Equations ( 6.8) and (6.9) indicate that the lasso and ridge\\nregression coefficient estimates are given by the first point at which an\\nellipse contacts the constraint region. Since ridge regression has a circular\\nconstraintwithnosharppoints,thisintersectionwillnotgenerallyoccuron\\nan axis, and so the ridge regression coefficient estimates will be exclusively\\nnon-zero. However, the lasso constraint has cornersat each of the axes, and\\nso the ellipse will often intersect the constraint region at an axis. When this\\noccurs, one of the coefficients will equal zero. In higher dimensions, many of\\nthe coefficient estimates may equal zero simultaneously. In Figure 6.7, the\\nintersection occurs at β1=0, and so the resulting model will only include\\nβ2.\\nIn Figure 6.7, we considered the simple case of p=2. When p=3,\\nthen the constraint region for ridge regression becomes a sphere, and the\\nconstraint region for the lasso becomes a polyhedron. When p>3, the\\nconstraint for ridge regression becomes a hypersphere, and the constraint\\nfor the lasso becomes a polytope. However, the key ideas depicted in Fig-\\nure6.7still hold. In particular, the lasso leads to feature selection when\\np>2due to the sharp corners of the polyhedron or polytope.\\nComparing the Lasso and Ridge Regression\\nIt is clear that the lasso has a major advantage over ridge regression, in\\nthat it produces simpler and more interpretable models that involve only a\\nsubset of the predictors. However, which method leads to better prediction\\naccuracy? Figure 6.8displays the variance, squared bias, and test MSE of\\nthe lasso applied to the same simulated data as in Figure 6.5. Clearly the\\nlasso leads to qualitatively similar behavior to ridge regression, in that as λ\\nincreases, the variance decreases and the bias increases. In the right-hand', metadata={'source': './documents/ISLP_website.pdf', 'page': 254}), Document(page_content='248 6. Linear Model Selection and Regularization\\n0.02 0.10 0.50 2.00 10.00 50.000 10 20 30 40 50 60Mean Squared Error0.0 0.2 0.4 0.6 0.8 1.00 10 20 30 40 50 60R2 on Training DataMean Squared ErrorλFIGURE 6.8. Left:Plots of squared bias (black), variance (green), and test\\nMSE (purple) for the lasso on a simulated data set. Right:Comparison of squared\\nbias, variance, and test MSE between lasso (solid) and ridge (dotted). Both are\\nplotted against their R2on the training data, as a common form of indexing. The\\ncrosses in both plots indicate the lasso model for which the MSE is smallest.\\npanel of Figure 6.8, the dotted lines represent the ridge regression fits.\\nHere we plot both against their R2on the training data. This is another\\nuseful way to index models, and can be used to compare models with\\ndifferent types of regularization, as is the case here. In this example, the\\nlasso and ridge regression result in almost identical biases. However, the\\nvariance of ridge regression is slightly lower than the variance of the lasso.\\nConsequently,theminimumMSEofridgeregressionisslightlysmallerthan\\nthat of the lasso.\\nHowever, the data in Figure 6.8were generated in such a way that all 45\\npredictorswererelatedtotheresponse—thatis,noneofthetruecoefficients\\nβ1,...,β45equaled zero. The lasso implicitly assumes that a number of the\\ncoefficients truly equal zero. Consequently, it is not surprising that ridge\\nregression outperforms the lasso in terms of prediction error in this setting.\\nFigure6.9illustrates a similar situation, except that now the response is a\\nfunction of only 2 out of 45predictors. Now the lasso tends to outperform\\nridge regression in terms of bias, variance, and MSE.\\nThese two examples illustrate that neither ridge regression nor the lasso\\nwill universally dominate the other. In general, one might expect the lasso\\nto perform better in a setting where a relatively small number of predictors\\nhave substantial coefficients, and the remaining predictors have coefficients\\nthat are very small or that equal zero. Ridge regression will perform better\\nwhen the response is a function of many predictors, all with coefficients of\\nroughly equal size. However, the number of predictors that is related to the\\nresponse is never known a priori for real data sets. A technique such as\\ncross-validation can be used in order to determine which approach is better\\non a particular data set.\\nAs with ridge regression, when the least squares estimates have exces-\\nsively high variance, the lasso solution can yield a reduction in variance\\nat the expense of a small increase in bias, and consequently can gener-\\nate more accurate predictions. Unlike ridge regression, the lasso performs\\nvariable selection, and hence results in models that are easier to interpret.', metadata={'source': './documents/ISLP_website.pdf', 'page': 255}), Document(page_content='6.2 Shrinkage Methods 249\\n0.02 0.10 0.50 2.00 10.00 50.000 20 40 60 80 100Mean Squared Error0.4 0.5 0.6 0.7 0.8 0.9 1.00 20 40 60 80 100R2 on Training DataMean Squared ErrorλFIGURE 6.9. Left:Plots of squared bias (black), variance (green), and test\\nMSE (purple) for the lasso. The simulated data is similar to that in Figure 6.8,\\nexcept that now only two predictors are related to the response. Right:Comparison\\nof squared bias, variance, and test MSE between lasso (solid) and ridge (dotted).\\nBoth are plotted against their R2on the training data, as a common form of\\nindexing. The crosses in both plots indicate the lasso model for which the MSE is\\nsmallest.\\nThereareveryefficientalgorithmsforfittingbothridgeandlassomodels;\\nin both cases the entire coefficient paths can be computed with about the\\nsame amount of work as a single least squares fit. We will explore this\\nfurther in the lab at the end of this chapter.\\nA Simple Special Case for Ridge Regression and the Lasso\\nIn order to obtain a better intuition about the behavior of ridge regression\\nand the lasso, consider a simple special case with n=p, andXa diag-\\nonal matrix with 1’s on the diagonal and 0’s in all off-diagonal elements.\\nTo simplify the problem further, assume also that we are performing regres-\\nsion without an intercept. With these assumptions, the usual least squares\\nproblem simplifies to finding β1,...,βpthat minimize\\np∑\\nj=1(yj−βj)2. (6.11)\\nIn this case, the least squares solution is given by\\nˆβj=yj.\\nAnd in this setting, ridge regression amounts to finding β1,...,βpsuch that\\np∑\\nj=1(yj−βj)2+λp∑\\nj=1β2\\nj (6.12)\\nis minimized, and the lasso amounts to finding the coefficients such that\\np∑\\nj=1(yj−βj)2+λp∑\\nj=1|βj| (6.13)', metadata={'source': './documents/ISLP_website.pdf', 'page': 256}), Document(page_content='250 6. Linear Model Selection and Regularization\\n−1.5 −0.5 0.0 0.5 1.0 1.5−1.5 −0.5 0.5 1.5Coefficient EstimateRidgeLeast Squares\\n−1.5 −0.5 0.0 0.5 1.0 1.5−1.5 −0.5 0.5 1.5Coefficient EstimateLassoLeast Squares\\nyjyjFIGURE 6.10. The ridge regression and lasso coefficient estimates for a simple\\nsetting with n=pandXa diagonal matrix with 1’s on the diagonal. Left:The\\nridge regression coefficient estimates are shrunken proportionally towards zero,\\nrelative to the least squares estimates. Right:The lasso coefficient estimates are\\nsoft-thresholded towards zero.\\nis minimized. One can show that in this setting, the ridge regression esti-\\nmates take the form\\nˆβR\\nj=yj/(1 + λ), (6.14)\\nand the lasso estimates take the form\\nˆβL\\nj=\\uf8f1\\n\\uf8f4\\uf8f2\\n\\uf8f4\\uf8f3yj−λ/2ifyj>λ/2;\\nyj+λ/2ifyj<−λ/2;\\n0 if|yj|≤λ/2.(6.15)\\nFigure6.10displays the situation. We can see that ridge regression and\\nthe lasso perform two very different types of shrinkage. In ridge regression,\\neach least squares coefficient estimate is shrunken by the same proportion.\\nIn contrast, the lasso shrinks each least squares coefficient towards zero by\\na constant amount, λ/2; the least squares coefficients that are less than\\nλ/2in absolute value are shrunken entirely to zero. The type of shrink-\\nage performed by the lasso in this simple setting ( 6.15) is known as soft-\\nthresholding . The fact that some lasso coefficients are shrunken entirely tosoft-\\nthresholdingzero explains why the lasso performs feature selection.\\nIn the case of a more general data matrix X, the story is a little more\\ncomplicated than what is depicted in Figure 6.10, but the main ideas still\\nhold approximately: ridge regression more or less shrinks every dimension\\nof the data by the same proportion, whereas the lasso more or less shrinks\\nall coefficients toward zero by a similar amount, and sufficiently small co-\\nefficients are shrunken all the way to zero.\\nBayesian Interpretation of Ridge Regression and the Lasso\\nWe now show that one can view ridge regression and the lasso through\\na Bayesian lens. A Bayesian viewpoint for regression assumes that the\\ncoefficient vector βhas some priordistribution, say p(β), where β=\\n(β0,β1,...,βp)T. The likelihood of the data can be written as f(Y|X,β),', metadata={'source': './documents/ISLP_website.pdf', 'page': 257}), Document(page_content='6.2 Shrinkage Methods 251\\n−3 −2 −1 0 1 2 30.0 0.1 0.2 0.3 0.4 0.5 0.6 0.7−3 −2 −1 0 1 2 30.0 0.1 0.2 0.3 0.4 0.5 0.6 0.7βjβj\\ng(βj)g(βj)\\nFIGURE 6.11. Left:Ridge regression is the posterior mode for βunder a Gaus-\\nsian prior. Right:The lasso is the posterior mode for βunder a double-exponential\\nprior.\\nwhereX=(X1,...,X p). Multiplying the prior distribution by the likeli-\\nhood gives us (up to a proportionality constant) the posterior distribution ,posterior\\ndistributionwhich takes the form\\np(β|X,Y)∝f(Y|X,β)p(β|X)=f(Y|X,β)p(β),\\nwhere the proportionality above follows from Bayes’ theorem, and the\\nequality above follows from the assumption that Xis fixed.\\nWe assume the usual linear model,\\nY=β0+X1β1+···+Xpβp+ϵ,\\nand suppose that the errors are independent and drawn from a normal dis-\\ntribution. Furthermore, assume that p(β)=∏p\\nj=1g(βj), for some density\\nfunction g. It turns out that ridge regression and the lasso follow naturally\\nfrom two special cases of g:\\n• Ifgis a Gaussian distribution with mean zero and standard deviation\\na function of λ, then it follows that the posterior mode forβ—thatposterior\\nmodeis, the most likely value for β, given the data—is given by the ridge\\nregression solution. (In fact, the ridge regression solution is also the\\nposterior mean.)\\n• Ifgis a double-exponential (Laplace) distribution with mean zero\\nand scale parameter a function of λ, then it follows that the posterior\\nmode for βis the lasso solution. (However, the lasso solution is not\\nthe posterior mean, and in fact, the posterior mean does not yield a\\nsparse coefficient vector.)\\nThe Gaussian and double-exponential priors are displayed in Figure 6.11.\\nTherefore, from a Bayesian viewpoint, ridge regression and the lasso follow\\ndirectly from assuming the usual linear model with normal errors, together\\nwith a simple prior distribution for β. Notice that the lasso prior is steeply\\npeaked at zero, while the Gaussian is flatter and fatter at zero. Hence, the\\nlasso expects a priori that many of the coefficients are (exactly) zero, while\\nridge assumes the coefficients are randomly distributed about zero.', metadata={'source': './documents/ISLP_website.pdf', 'page': 258}), Document(page_content='252 6. Linear Model Selection and Regularization\\n5e−03 5e−02 5e−01 5e+0025.0 25.2 25.4 25.6Cross−Validation Error5e−03 5e−02 5e−01 5e+00−300 −100 0 100 300Standardized CoefficientsλλFIGURE 6.12. Left:Cross-validation errors that result from applying ridge\\nregression to the Creditdata set with various values of λ.Right:The coefficient\\nestimates as a function of λ. The vertical dashed lines indicate the value of λ\\nselected by cross-validation.\\n6.2.3 Selecting the Tuning Parameter\\nJust as the subset selection approaches considered in Section 6.1require\\na method to determine which of the models under consideration is best,\\nimplementing ridge regression and the lasso requires a method for selecting\\na value for the tuning parameter λin (6.5) and (6.7), or equivalently, the\\nvalue of the constraint sin (6.9) and (6.8). Cross-validation provides a sim-\\nple way to tackle this problem. We choose a grid of λvalues, and compute\\nthe cross-validation error for each value of λ, as described in Chapter 5.W e\\nthen select the tuning parameter value for which the cross-validation error\\nis smallest. Finally, the model is re-fit using all of the available observations\\nand the selected value of the tuning parameter.\\nFigure6.12displays the choice of λthat results from performing leave-\\none-out cross-validation on the ridge regression fits from the Creditdata\\nset. The dashed vertical lines indicate the selected value of λ. In this case\\nthe value is relatively small, indicating that the optimal fit only involves a\\nsmallamountofshrinkagerelativetotheleastsquaressolution.Inaddition,\\nthe dip is not very pronounced, so there is rather a wide range of values\\nthat would give a very similar error. In a case like this we might simply use\\nthe least squares solution.\\nFigure6.13provides an illustration of ten-fold cross-validation applied to\\nthe lasso fits on the sparse simulated data from Figure 6.9. The left-hand\\npanelofFigure 6.13displaysthecross-validationerror,whiletheright-hand\\npanel displays the coefficient estimates. The vertical dashed lines indicate\\nthe point at which the cross-validation error is smallest. The two colored\\nlines in the right-hand panel of Figure 6.13represent the two predictors\\nthat are related to the response, while the grey lines represent the unre-\\nlated predictors; these are often referred to as signalandnoisevariables,signalrespectively. Not only has the lasso correctly given much larger coeffi-\\ncient estimates to the two signal predictors, but also the minimum cross-\\nvalidation error corresponds to a set of coefficient estimates for which only\\nthe signal variables are non-zero. Hence cross-validation together with the\\nlasso has correctly identified the two signal variables in the model, even\\nthough this is a challenging setting, with p= 45 variables and only n= 50', metadata={'source': './documents/ISLP_website.pdf', 'page': 259}), Document(page_content='6.3 Dimension Reduction Methods 253\\n0.0 0.2 0.4 0.6 0.8 1.00 200 600 1000 1400Cross−Validation Error0.0 0.2 0.4 0.6 0.8 1.0−5 0 5 10 15Standardized Coefficients∥ˆβLλ∥1/∥ˆβ∥1∥ˆβLλ∥1/∥ˆβ∥1FIGURE 6.13. Left: Ten-fold cross-validation MSE for the lasso, applied to\\nthe sparse simulated data set from Figure 6.9.Right:The corresponding lasso\\ncoefficient estimates are displayed. The two signal variables are shown in color,\\nand the noise variables are in gray. The vertical dashed lines indicate the lasso\\nfit for which the cross-validation error is smallest.\\nobservations. In contrast, the least squares solution—displayed on the far\\nright of the right-hand panel of Figure 6.13—assigns a large coefficient\\nestimate to only one of the two signal variables.\\n6.3 Dimension Reduction Methods\\nThe methods that we have discussed so far in this chapter have controlled\\nvariance in two different ways, either by using a subset of the original vari-\\nables, or by shrinking their coefficients toward zero. All of these methods\\nare defined using the original predictors, X1,X2,...,X p. We now explore\\na class of approaches that transform the predictors and then fit a least\\nsquares model using the transformed variables. We will refer to these tech-\\nniques as dimension reduction methods.dimension\\nreductionLetZ1,Z2,...,Z Mrepresent M<p linear combinations of our original\\nlinear\\ncombinationppredictors. That is,\\nZm=p∑\\nj=1φjmXj (6.16)\\nfor some constants φ1m,φ2m...,φpm,m=1,...,M . We can then fit the\\nlinear regression model\\nyi=θ0+M∑\\nm=1θmzim+ϵi,i=1, . . . , n, (6.17)\\nusing least squares. Note that in ( 6.17), the regression coefficients are given\\nbyθ0,θ1,...,θM.Iftheconstants φ1m,φ2m,...,φpmarechosenwisely,then\\nsuch dimension reduction approaches can often outperform least squares\\nregression. In other words, fitting ( 6.17) using least squares can lead to\\nbetter results than fitting ( 6.1) using least squares.\\nThe term dimension reduction comes from the fact that this approach\\nreduces the problem of estimating the p+1coefficients β0,β1,...,βpto the', metadata={'source': './documents/ISLP_website.pdf', 'page': 260}), Document(page_content='254 6. Linear Model Selection and Regularization\\n102030405060700 5 10 15 20 25 30 35PopulationAd SpendingFIGURE 6.14. The population size ( pop) and ad spending ( ad) for100different\\ncities are shown as purple circles. The green solid line indicates the first principal\\ncomponent, and the blue dashed line indicates the second principal component.\\nsimpler problem of estimating the M+1coefficients θ0,θ1,...,θM, where\\nM<p . In other words, the dimension of the problem has been reduced\\nfromp+1toM+1.\\nNotice that from ( 6.16),\\nM∑\\nm=1θmzim=M∑\\nm=1θmp∑\\nj=1φjmxij=p∑\\nj=1M∑\\nm=1θmφjmxij=p∑\\nj=1βjxij,\\nwhere\\nβj=M∑\\nm=1θmφjm. (6.18)\\nHence (6.17) can be thought of as a special case of the original linear\\nregression model given by ( 6.1). Dimension reduction serves to constrain\\nthe estimated βjcoefficients, since now they must take the form ( 6.18).\\nThis constraint on the form of the coefficients has the potential to bias the\\ncoefficient estimates. However, in situations where pis large relative to n,\\nselectingavalueof M≪pcansignificantlyreducethevarianceofthefitted\\ncoefficients. If M=p, and all the Zmare linearly independent, then ( 6.18)\\nposes no constraints. In this case, no dimension reduction occurs, and so\\nfitting (6.17) is equivalent to performing least squares on the original p\\npredictors.\\nAll dimension reduction methods work in two steps. First, the trans-\\nformed predictors Z1,Z2,...,Z Mare obtained. Second, the model is fit\\nusing these Mpredictors. However, the choice of Z1,Z2,...,Z M, or equiv-\\nalently, the selection of the φjm’s, can be achieved in different ways. In this\\nchapter,wewillconsidertwoapproachesforthistask: principal components\\nandpartial least squares .\\n6.3.1 Principal Components Regression\\nPrincipal components analysis (PCA) is a popular approach for derivingprincipal\\ncomponents\\nanalysis', metadata={'source': './documents/ISLP_website.pdf', 'page': 261}), Document(page_content='6.3 Dimension Reduction Methods 255\\na low-dimensional set of features from a large set of variables. PCA is\\ndiscussed in greater detail as a tool for unsupervised learning in Chapter 12.\\nHere we describe its use as a dimension reduction technique for regression.\\nAn Overview of Principal Components Analysis\\nPCA is a technique for reducing the dimension of an n×pdata matrix X.\\nThefirst principal component direction of the data is that along which the\\nobservations vary the most . For instance, consider Figure 6.14, which shows\\npopulation size ( pop) in tens of thousands of people, and ad spending for a\\nparticular company ( ad) in thousands of dollars, for 100cities.6The green\\nsolid line represents the first principal component direction of the data. We\\ncan see by eye that this is the direction along which there is the greatest\\nvariability in the data. That is, if we projected the 100 observations onto\\nthis line (as shown in the left-hand panel of Figure 6.15), then the resulting\\nprojected observations would have the largest possible variance; projecting\\nthe observations onto any other line would yield projected observations\\nwith lower variance. Projecting a point onto a line simply involves finding\\nthe location on the line which is closest to the point.\\nThe first principal component is displayed graphically in Figure 6.14, but\\nhow can it be summarized mathematically? It is given by the formula\\nZ1=0.839×(pop−pop)+0 .544×(ad−ad). (6.19)\\nHereφ11=0.839andφ21=0.544are the principal component loadings,\\nwhich define the direction referred to above. In ( 6.19),pop indicates the\\nmean of all popvalues in this data set, and adindicates the mean of all ad-\\nvertisingspending. The idea is that out of everypossible linear combination\\nofpopandadsuch that φ2\\n11+φ2\\n21=1, this particular linear combination\\nyields the highest variance: i.e. this is the linear combination for which\\nVar(φ11×(pop−pop)+φ21×(ad−ad))is maximized. It is necessary to\\nconsider only linear combinations of the form φ2\\n11+φ2\\n21=1, since otherwise\\nwe could increase φ11andφ21arbitrarily in order to blow up the variance.\\nIn (6.19), the two loadings are both positive and have similar size, and so\\nZ1is almost an averageof the two variables.\\nSincen= 100 ,popandadare vectors of length 100, and so is Z1in\\n(6.19). For instance,\\nzi1=0.839×(pop i−pop)+0 .544×(adi−ad). (6.20)\\nThe values of z11,...,z n1are known as the principal component scores , and\\ncan be seen in the right-hand panel of Figure 6.15.\\nThere is also another interpretation of PCA: the first principal compo-\\nnent vector defines the line that is as close as possible to the data. For\\ninstance, in Figure 6.14, the first principal component line minimizes the\\nsum of the squared perpendicular distances between each point and the\\nline. These distances are plotted as dashed line segments in the left-hand\\n6This dataset is distinct from the Advertising data discussed in Chapter 3.', metadata={'source': './documents/ISLP_website.pdf', 'page': 262}), Document(page_content='256 6. Linear Model Selection and Regularization\\n20 30 40 505 10 15 20 25 30PopulationAd Spending−20 −10 0 10 20−10 −5 0 5 101st Principal Component2nd Principal ComponentFIGURE 6.15. A subset of the advertising data. The mean popandadbudgets\\nare indicated with a blue circle. Left:The first principal component direction is\\nshown in green. It is the dimension along which the data vary the most, and it also\\ndefines the line that is closest to all nof the observations. The distances from each\\nobservation to the principal component are represented using the black dashed line\\nsegments. The blue dot represents (pop,ad).Right:The left-hand panel has been\\nrotated so that the first principal component direction coincides with the x-axis.\\npanel of Figure 6.15, in which the crosses represent the projection of each\\npoint onto the first principal component line. The first principal component\\nhas been chosen so that the projected observations are as close as possible\\nto the original observations.\\nIn the right-hand panel of Figure 6.15, the left-hand panel has been\\nrotated so that the first principal component direction coincides with the\\nx-axis. It is possible to show that the first principal component score for\\ntheith observation, given in ( 6.20), is the distance in the x-direction of the\\nith cross from zero. So for example, the point in the bottom-left corner of\\nthe left-hand panel of Figure 6.15has a large negative principal component\\nscore,zi1=−26.1, while the point in the top-right corner has a large\\npositive score, zi1= 18 .7. These scores can be computed directly using\\n(6.20).\\nWe can think of the values of the principal component Z1as single-\\nnumber summaries of the joint popandadbudgets for each location. In\\nthis example, if zi1=0.839×(pop i−pop)+0 .544×(adi−ad)<0,\\nthen this indicates a city with below-average population size and below-\\naverage ad spending. A positive score suggests the opposite. How well can a\\nsingle number represent both popandad? In this case, Figure 6.14indicates\\nthatpopandadhave approximately a linear relationship, and so we might\\nexpect that a single-number summary will work well. Figure 6.16displays\\nzi1versus both popandad.7The plots show a strong relationship between\\nthe first principal component and the two features. In other words, the first\\nprincipal component appears to capture most of the information contained\\nin thepopandadpredictors.\\nSo far we have concentrated on the first principal component. In gen-\\neral, one can construct up to pdistinct principal components. The second\\n7The principal components were calculated after first standardizing both popandad,\\na common approach. Hence, the x-axes on Figures 6.15and6.16are not on the same\\nscale.', metadata={'source': './documents/ISLP_website.pdf', 'page': 263}), Document(page_content='6.3 Dimension Reduction Methods 257\\n−3 −2 −1 0 1 2 320 30 40 50 601st Principal ComponentPopulation−3 −2 −1 0 1 2 35 10 15 20 25 301st Principal ComponentAd SpendingFIGURE 6.16. Plots of the first principal component scores zi1versuspopand\\nad. The relationships are strong.\\nprincipal component Z2is a linear combination of the variables that is un-\\ncorrelated with Z1, and has largest variance subject to this constraint. The\\nsecond principal component direction is illustrated as a dashed blue line in\\nFigure6.14. It turns out that the zero correlation condition of Z1withZ2\\nis equivalent to the condition that the direction must be perpendicular , orperpen-\\ndicular orthogonal , to the first principal component direction. The second principal\\northogonal component is given by the formula\\nZ2=0.544×(pop−pop)−0.839×(ad−ad).\\nSince the advertising data has two predictors, the first two principal com-\\nponents contain all of the information that is in popandad. However, by\\nconstruction, the first component will contain the most information. Con-\\nsider, for example, the much larger variability of zi1(thex-axis) versus\\nzi2(they-axis) in the right-hand panel of Figure 6.15. The fact that the\\nsecond principal component scores are much closer to zero indicates that\\nthis component captures far less information. As another illustration, Fig-\\nure6.17displays zi2versuspopandad. There is little relationship between\\nthe second principal component and these two predictors, again suggesting\\nthat in this case, one only needs the first principal component in order to\\naccurately represent the popandadbudgets.\\nWith two-dimensional data, such as in our advertising example, we can\\nconstruct at most two principal components. However, if we had other\\npredictors, such as population age, income level, education, and so forth,\\nthen additional components could be constructed. They would successively\\nmaximize variance, subject to the constraint of being uncorrelated with the\\npreceding components.\\nThe Principal Components Regression Approach\\nTheprincipal components regression (PCR) approach involves construct-principal\\ncomponents\\nregressioning the first Mprincipal components, Z1,...,Z M, and then using these\\ncomponents as the predictors in a linear regression model that is fit us-\\ning least squares. The key idea is that often a small number of principal\\ncomponents suffice to explain most of the variability in the data, as well\\nas the relationship with the response. In other words, we assume that the', metadata={'source': './documents/ISLP_website.pdf', 'page': 264}), Document(page_content='258 6. Linear Model Selection and Regularization\\n−1.0 −0.5 0.0 0.5 1.020 30 40 50 602nd Principal ComponentPopulation−1.0 −0.5 0.0 0.5 1.05 10 15 20 25 302nd Principal ComponentAd SpendingFIGURE 6.17. Plots of the second principal component scores zi2versuspop\\nandad. The relationships are weak.\\n0 10 20 30 400 10 20 30 40 50 60 70Number of ComponentsMean Squared Error0 10 20 30 400 50 100 150Number of ComponentsMean Squared ErrorSquared BiasTest MSEVariance\\nFIGURE 6.18. PCR was applied to two simulated data sets. In each panel, the\\nhorizontal dashed line represents the irreducible error. Left:Simulated data from\\nFigure6.8.Right:Simulated data from Figure 6.9.\\ndirections in which X1,...,X pshow the most variation are the directions\\nthat are associated with Y. While this assumption is not guaranteed to be\\ntrue, it often turns out to be a reasonable enough approximation to give\\ngood results.\\nIf the assumption underlying PCR holds, then fitting a least squares\\nmodel to Z1,...,Z Mwill lead to better results than fitting a least squares\\nmodel to X1,...,X p, since most or all of the information in the data that\\nrelates to the response is contained in Z1,...,Z M, and by estimating only\\nM≪pcoefficients we can mitigate overfitting. In the advertising data, the\\nfirst principal component explains most of the variance in both popandad,\\nso a principal component regression that uses this single variable to predict\\nsome response of interest, such as sales, will likely perform quite well.\\nFigure6.18displays the PCR fits on the simulated data sets from Fig-\\nures6.8and6.9. Recall that both data sets were generated using n= 50\\nobservations and p= 45 predictors. However, while the response in the\\nfirst data set was a function of all the predictors, the response in the sec-\\nond data set was generated using only two of the predictors. The curves are\\nplotted as a function of M, the number of principal components used as\\npredictors in the regression model. As more principal components are used', metadata={'source': './documents/ISLP_website.pdf', 'page': 265}), Document(page_content='6.3 Dimension Reduction Methods 259\\n0 10 20 30 400 10 20 30 40 50 60 70PCR\\nNumber of ComponentsMean Squared ErrorSquared BiasTest MSEVariance\\n0.0 0.2 0.4 0.6 0.8 1.00 10 20 30 40 50 60 70Ridge Regression and Lasso\\nShrinkage FactorMean Squared ErrorFIGURE 6.19. PCR, ridge regression, and the lasso were applied to a simu-\\nlated data set in which the first five principal components of Xcontain all the\\ninformation about the response Y. In each panel, the irreducible error Var(ϵ)is\\nshown as a horizontal dashed line. Left:Results for PCR. Right:Results for lasso\\n(solid) and ridge regression (dotted). The x-axis displays the shrinkage factor\\nof the coefficient estimates, defined as the ℓ2norm of the shrunken coefficient\\nestimates divided by the ℓ2norm of the least squares estimate.\\nin the regression model, the bias decreases, but the variance increases. This\\nresults in a typical U-shape for the mean squared error. When M=p= 45 ,\\nthen PCR amounts simply to a least squares fit using all of the original\\npredictors. The figure indicates that performing PCR with an appropriate\\nchoice of Mcan result in a substantial improvement over least squares, es-\\npecially in the left-hand panel. However, by examining the ridge regression\\nand lasso results in Figures 6.5,6.8, and6.9, we see that PCR does not\\nperform as well as the two shrinkage methods in this example.\\nThe relatively worse performance of PCR in Figure 6.18is a consequence\\nof the fact that the data were generated in such a way that many princi-\\npal components are required in order to adequately model the response.\\nIn contrast, PCR will tend to do well in cases when the first few principal\\ncomponents are sufficient to capture most of the variation in the predictors\\nas well as the relationship with the response. The left-hand panel of Fig-\\nure6.19illustrates the results from another simulated data set designed to\\nbe more favorable to PCR. Here the response was generated in such a way\\nthat it depends exclusively on the first five principal components. Now the\\nbias drops to zero rapidly as M, the number of principal components used\\nin PCR, increases. The mean squared error displays a clear minimum at\\nM=5. The right-hand panel of Figure 6.19displays the results on these\\ndata using ridge regression and the lasso. All three methods offer a signif-\\nicant improvement over least squares. However, PCR and ridge regression\\nslightly outperform the lasso.\\nWe note that even though PCR provides a simple way to perform re-\\ngression using M<p predictors, it is nota feature selection method. This\\nis because each of the Mprincipal components used in the regression is a\\nlinear combination of all pof theoriginalfeatures. For instance, in ( 6.19),\\nZ1was a linear combination of both popandad. Therefore, while PCR of-\\nten performs quite well in many practical settings, it does not result in the', metadata={'source': './documents/ISLP_website.pdf', 'page': 266}), Document(page_content='260 6. Linear Model Selection and Regularization\\n2 4 6 8 10−300 −100 0 100 200 300 400Number of ComponentsStandardized CoefficientsIncomeLimitRatingStudent\\n2 4 6 8 1020000 40000 60000 80000Number of ComponentsCross−Validation MSEFIGURE 6.20. Left:PCR standardized coefficient estimates on the Creditdata\\nset for different values of M.Right:The ten-fold cross-validation MSE obtained\\nusing PCR, as a function of M.\\ndevelopment of a model that relies upon a small set of the original features.\\nIn this sense, PCR is more closely related to ridge regression than to the\\nlasso. In fact, one can show that PCR and ridge regression are very closely\\nrelated. One can even think of ridge regression as a continuous version of\\nPCR!8\\nIn PCR, the number of principal components, M, is typically chosen by\\ncross-validation. The results of applying PCR to the Creditdata set are\\nshown in Figure 6.20; the right-hand panel displays the cross-validation er-\\nrorsobtained,asafunctionof M.Onthesedata,thelowestcross-validation\\nerror occurs when there are M= 10 components; this corresponds to al-\\nmost no dimension reduction at all, since PCR with M= 11 is equivalent\\nto simply performing least squares.\\nWhen performing PCR, we generally recommend standardizing each pre-\\ndictor, using ( 6.6), prior to generating the principal components. This stan-\\ndardization ensures that all variables are on the same scale. In the absence\\nof standardization, the high-variance variables will tend to play a larger\\nrole in the principal components obtained, and the scale on which the vari-\\nables are measured will ultimately have an effect on the final PCR model.\\nHowever, if the variables are all measured in the same units (say, kilograms,\\nor inches), then one might choose not to standardize them.\\n6.3.2 Partial Least Squares\\nThePCRapproachthatwejustdescribedinvolvesidentifyinglinearcombi-\\nnations, or directions , that best represent the predictors X1,...,X p. These\\ndirections are identified in an unsupervised way, since the response Yis\\nnot used to help determine the principal component directions. That is,\\nthe response does not supervise the identification of the principal compo-\\nnents. Consequently, PCR suffers from a drawback: there is no guarantee\\n8More details can be found in Section 3.5 of The Elements of Statistical Learning by\\nHastie, Tibshirani, and Friedman.', metadata={'source': './documents/ISLP_website.pdf', 'page': 267}), Document(page_content='6.3 Dimension Reduction Methods 261\\n20304050605 10 15 20 25 30PopulationAd SpendingFIGURE 6.21. For the advertising data, the first PLS direction (solid line)\\nand first PCR direction (dotted line) are shown.\\nthat the directions that best explain the predictors will also be the best\\ndirections to use for predicting the response. Unsupervised methods are\\ndiscussed further in Chapter 12.\\nWe now present partial least squares (PLS), a supervised alternative topartial least\\nsquaresPCR.LikePCR,PLSisadimensionreductionmethod,whichfirstidentifies\\nanewsetoffeatures Z1,...,Z Mthatarelinearcombinationsoftheoriginal\\nfeatures, and then fits a linear model via least squares using these Mnew\\nfeatures. But unlike PCR, PLS identifies these new features in a supervised\\nway—that is, it makes use of the response Yin order to identify new\\nfeatures that not only approximate the old features well, but also that are\\nrelated to the response . Roughly speaking, the PLS approach attempts to\\nfind directions that help explain both the response and the predictors.\\nWe now describe how the first PLS direction is computed. After stan-\\ndardizing the ppredictors, PLS computes the first direction Z1by setting\\neachφj1in (6.16) equal to the coefficient from the simple linear regression\\nofYontoXj. One can show that this coefficient is proportional to the cor-\\nrelation between YandXj. Hence, in computing Z1=∑p\\nj=1φj1Xj, PLS\\nplaces the highest weight on the variables that are most strongly related to\\nthe response.\\nFigure6.21displays an example of PLS on a synthetic dataset with Sales\\nin each of 100 regions as the response, and two predictors; Population Size\\nand Advertising Spending. The solid green line indicates the first PLS di-\\nrection, while the dotted line shows the first principal component direction.\\nPLS has chosen a direction that has less change in the addimension per\\nunit change in the popdimension, relative to PCA. This suggests that pop\\nis more highly correlated with the response than is ad. The PLS direction\\ndoes not fit the predictors as closely as does PCA, but it does a better job\\nexplaining the response.\\nTo identify the second PLS direction we first adjusteach of the variables\\nforZ1, by regressing each variable on Z1and taking residuals . These resid-\\nuals can be interpreted as the remaining information that has not been\\nexplained by the first PLS direction. We then compute Z2using this or-', metadata={'source': './documents/ISLP_website.pdf', 'page': 268}), Document(page_content='262 6. Linear Model Selection and Regularizationthogonalizeddata in exactly the same fashion asZ1was computed basedon the original data. This iterative approach can be repeatedMtimes toidentify multiple PLS componentsZ1,...,ZM. Finally, at the end of thisprocedure, we use least squares to fit a linear model to predictYusingZ1,...,ZMin exactly the same fashion as for PCR.As with PCR, the numberMof partial least squares directions used inPLS is a tuning parameter that is typically chosen by cross-validation. Wegenerally standardize the predictors and response before performing PLS.PLS is popular in the field of chemometrics, where many variables arisefrom digitized spectrometry signals. In practice it often performs no betterthan ridge regression or PCR. While the supervised dimension reductionof PLS can reduce bias, it also has the potential to increase variance, sothat the overall benefit of PLS relative to PCR is a wash.6.4 Considerations in High Dimensions6.4.1 High-Dimensional DataMost traditional statistical techniques for regression and classification areintended for thelow-dimensionalsetting in whichn, the number of ob-low-dimensionalservations, is much greater thanp, the number of features. This is due inpart to the fact that throughout most of the field’s history, the bulk of sci-entific problems requiring the use of statistics have been low-dimensional.For instance, consider the task of developing a model to predict a patient’sblood pressure on the basis of his or her age, sex, and body mass index(BMI). There are three predictors, or four if an intercept is included in themodel, and perhaps several thousand patients for whom blood pressureand age, sex, and BMI are available. Hencen≫p, and so the problem islow-dimensional. (By dimension here we are referring to the size ofp.)In the past 20 years, new technologies have changed the way that dataare collected in fields as diverse as finance, marketing, and medicine. It isnow commonplace to collect an almost unlimited number of feature mea-surements (pvery large). Whilepcan be extremely large, the number ofobservationsnis often limited due to cost, sample availability, or otherconsiderations. Two examples are as follows:1. Rather than predicting blood pressure on the basis of just age, sex,and BMI, one might also collect measurements for half a millionsin-gle nucleotide polymorphisms(SNPs; these are individual DNA mu-tations that are relatively common in the population) for inclusion inthe predictive model. Thenn≈200andp≈500,000.2. Amarketinganalystinterestedinunderstandingpeople’sonlineshop-ping patterns could treat as features all of the search terms enteredby users of a search engine. This is sometimes known as the “bag-of-words” model. The same researcher might have access to the searchhistories of only a few hundred or a few thousand search engine userswho have consented to share their information with the researcher.For a given user, each of thepsearch terms is scored present (0) or', metadata={'source': './documents/ISLP_website.pdf', 'page': 269}), Document(page_content='6.4 Considerations in High Dimensions 263\\nabsent (1), creating a large binary feature vector. Then n≈1,000\\nandpis much larger.\\nData sets containing more features than observations are often referred\\nto ashigh-dimensional . Classical approaches such as least squares linearhigh-\\ndimensionalregression are not appropriate in this setting. Many of the issues that arise\\nin the analysis of high-dimensional data were discussed earlier in this book,\\nsince they apply also when n>p : these include the role of the bias-variance\\ntrade-off and the danger of overfitting. Though these issues are always rele-\\nvant, they can become particularly important when the number of features\\nis very large relative to the number of observations.\\nWe have defined the high-dimensional setting as the case where the num-\\nber of features pis larger than the number of observations n. But the con-\\nsiderations that we will now discuss certainly also apply if pis slightly\\nsmaller than n, and are best always kept in mind when performing super-\\nvised learning.\\n6.4.2 What Goes Wrong in High Dimensions?\\nIn order to illustrate the need for extra care and specialized techniques\\nfor regression and classification when p>n , we begin by examining what\\ncan go wrong if we apply a statistical technique not intended for the high-\\ndimensional setting. For this purpose, we examine least squares regression.\\nButthesameconceptsapplytologisticregression,lineardiscriminantanal-\\nysis, and other classical statistical approaches.\\nWhen the number of features pis as large as, or larger than, the number\\nof observations n, least squares as described in Chapter 3cannot (or rather,\\nshould not ) be performed. The reason is simple: regardless of whether or\\nnot there truly is a relationship between the features and the response,\\nleast squares will yield a set of coefficient estimates that result in a perfect\\nfit to the data, such that the residuals are zero.\\nAnexampleisshowninFigure 6.22withp=1feature(plusanintercept)\\nin two cases: when there are 20 observations, and when there are only\\ntwo observations. When there are 20 observations, n>p and the least\\nsquaresregressionlinedoesnotperfectlyfitthedata;instead,theregression\\nline seeks to approximate the 20 observations as well as possible. On the\\nother hand, when there are only two observations, then regardless of the\\nvalues of those observations, the regression line will fit the data exactly.\\nThis is problematic because this perfect fit will almost certainly lead to\\noverfitting of the data. In other words, though it is possible to perfectly fit\\nthe training data in the high-dimensional setting, the resulting linear model\\nwill perform extremely poorly on an independent test set, and therefore\\ndoes not constitute a useful model. In fact, we can see that this happened\\nin Figure 6.22: the least squares line obtained in the right-hand panel will\\nperform very poorly on a test set comprised of the observations in the left-\\nhand panel. The problem is simple: when p>n orp≈n, a simple least\\nsquares regression line is too flexibleand hence overfits the data.\\nFigure6.23further illustrates the risk of carelessly applying least squares\\nwhen the number of features pis large. Data were simulated with n= 20\\nobservations, and regression was performed with between 1and20features,', metadata={'source': './documents/ISLP_website.pdf', 'page': 270}), Document(page_content='264 6. Linear Model Selection and Regularization\\n−1.5 −1.0 −0.5 0.0 0.5 1.0−10 −5 0 5 10−1.5 −1.0 −0.5 0.0 0.5 1.0−10 −5 0 5 10XXYY\\nFIGURE 6.22.Left:Least squares regression in the low-dimensional setting.Right:Least squares regression withn=2observations and two parameters to beestimated (an intercept and a coefficient).\\n5 10 150.2 0.4 0.6 0.8 1.0Number of VariablesR2\\n5 10 150.0 0.2 0.4 0.6 0.8Number of VariablesTraining MSE5 10 151 5 50 500Number of VariablesTest MSEFIGURE 6.23.On a simulated example withn= 20training observations,features that are completely unrelated to the outcome are added to the model.Left:TheR2increases to 1 as more features are included.Center:The trainingset MSE decreases to 0 as more features are included.Right:The test set MSEincreases as more features are included.each of which was completely unrelated to the response. As shown in thefigure,themodelR2increasesto1asthenumberoffeaturesincludedinthemodel increases, and correspondingly the training set MSE decreases to 0as the number of features increases,even though the features are completelyunrelated to the response. On the other hand, the MSE on anindependenttest setbecomes extremely large as the number of features included in themodel increases, because including the additional predictors leads to a vastincrease in the variance of the coefficient estimates. Looking at the testset MSE, it is clear that the best model contains at most a few variables.However, someone who carelessly examines only theR2or the training setMSE might erroneously conclude that the model with the greatest numberof variables is best. This indicates the importance of applying extra carewhen analyzing data sets with a large number of variables, and of alwaysevaluating model performance on an independent test set.', metadata={'source': './documents/ISLP_website.pdf', 'page': 271}), Document(page_content='6.4 Considerations in High Dimensions 265\\n116 21012345128 51012345170 111012345p= 20p= 50p= 2000\\nDegrees of FreedomDegrees of FreedomDegrees of FreedomFIGURE 6.24. The lasso was performed with n= 100 observations and three\\nvalues of p, the number of features. Of the pfeatures, 20 were associated with\\nthe response. The boxplots show the test MSEs that result using three different\\nvalues of the tuning parameter λin (6.7). For ease of interpretation, rather than\\nreporting λ, thedegrees of freedom are reported; for the lasso this turns out\\nto be simply the number of estimated non-zero coefficients. When p= 20 , the\\nlowest test MSE was obtained with the smallest amount of regularization. When\\np= 50 , the lowest test MSE was achieved when there is a substantial amount\\nof regularization. When p=2,000the lasso performed poorly regardless of the\\namount of regularization, due to the fact that only 20 of the 2,000 features truly\\nare associated with the outcome.\\nInSection 6.1.3,wesawanumberofapproachesforadjustingthetraining\\nset RSS or R2in order to account for the number of variables used to fit\\na least squares model. Unfortunately, the Cp, AIC, and BIC approaches\\nare not appropriate in the high-dimensional setting, because estimating ˆσ2\\nis problematic. (For instance, the formula for ˆσ2from Chapter 3yields an\\nestimate ˆσ2=0in this setting.) Similarly, problems arise in the application\\nof adjusted R2in the high-dimensional setting, since one can easily obtain\\na model with an adjusted R2value of 1. Clearly, alternative approaches\\nthat are better-suited to the high-dimensional setting are required.\\n6.4.3 Regression in High Dimensions\\nIt turns out that many of the methods seen in this chapter for fitting\\nless flexible least squares models, such as forward stepwise selection, ridge\\nregression, the lasso, and principal components regression, are particularly\\nusefulforperformingregressioninthehigh-dimensionalsetting.Essentially,\\nthese approaches avoid overfitting by using a less flexible fitting approach\\nthan least squares.\\nFigure6.24illustrates the performance of the lasso in a simple simulated\\nexample. There are p= 20 ,50, or2,000features, of which 20 are truly\\nassociated with the outcome. The lasso was performed on n= 100 training\\nobservations, and the mean squared error was evaluated on an independent\\ntest set. As the number of features increases, the test set error increases.\\nWhenp= 20 , the lowest validation set error was achieved when λin', metadata={'source': './documents/ISLP_website.pdf', 'page': 272}), Document(page_content='266 6. Linear Model Selection and Regularization\\n(6.7) was small; however, when pwas larger then the lowest validation\\nset error was achieved using a larger value of λ. In each boxplot, rather\\nthan reporting the values of λused, the degrees of freedom of the resulting\\nlasso solution is displayed; this is simply the number of non-zero coefficient\\nestimates in the lasso solution, and is a measure of the flexibility of the\\nlasso fit. Figure 6.24highlights three important points: (1) regularization\\nor shrinkage plays a key role in high-dimensional problems, (2) appropriate\\ntuning parameter selection is crucial for good predictive performance, and\\n(3) the test error tends to increase as the dimensionality of the problem\\n(i.e. the number of features or predictors) increases, unless the additional\\nfeatures are truly associated with the response.\\nThe third point above is in fact a key principle in the analysis of high-\\ndimensionaldata,whichisknownasthe curse of dimensionality .Onemightcurse of di-\\nmensionalitythink that as the number of features used to fit a model increases, the\\nquality of the fitted model will increase as well. However, comparing the\\nleft-hand and right-hand panels in Figure 6.24, we see that this is not\\nnecessarily the case: in this example, the test set MSE almost doubles as\\npincreases from 20 to 2,000. In general, adding additional signal features\\nthat are truly associated with the response will improve the fitted model ,\\nin the sense of leading to a reduction in test set error. However, adding\\nnoise features that are not truly associated with the response will lead\\nto a deterioration in the fitted model, and consequently an increased test\\nset error. This is because noise features increase the dimensionality of the\\nproblem, exacerbating the risk of overfitting (since noise features may be\\nassigned nonzero coefficients due to chance associations with the response\\non the training set) without any potential upside in terms of improved test\\nset error. Thus, we see that new technologies that allow for the collection\\nof measurements for thousands or millions of features are a double-edged\\nsword: they can lead to improved predictive models if these features are in\\nfact relevant to the problem at hand, but will lead to worse results if the\\nfeatures are not relevant. Even if they are relevant, the variance incurred\\nin fitting their coefficients may outweigh the reduction in bias that they\\nbring.\\n6.4.4 Interpreting Results in High Dimensions\\nWhen we perform the lasso, ridge regression, or other regression proce-\\ndures in the high-dimensional setting, we must be quite cautious in the way\\nthat we report the results obtained. In Chapter 3, we learned about multi-\\ncollinearity , the concept that the variables in a regression might be corre-\\nlated with each other. In the high-dimensional setting, the multicollinearity\\nproblem is extreme: any variable in the model can be written as a linear\\ncombination of all of the other variables in the model. Essentially, this\\nmeans that we can never know exactly which variables (if any) truly are\\npredictive of the outcome, and we can never identify the bestcoefficients\\nfor use in the regression. At most, we can hope to assign large regression\\ncoefficients to variables that are correlated with the variables that truly are\\npredictive of the outcome.', metadata={'source': './documents/ISLP_website.pdf', 'page': 273}), Document(page_content='6.5 Lab: Linear Models and Regularization Methods 267\\nFor instance, suppose that we are trying to predict blood pressure on the\\nbasis of half a million SNPs, and that forward stepwise selection indicates\\nthat 17 of those SNPs lead to a good predictive model on the training data.\\nIt would be incorrect to conclude that these 17 SNPs predict blood pressure\\nmore effectively than the other SNPs not included in the model. There are\\nlikely to be many sets of 17 SNPs that would predict blood pressure just\\nas well as the selected model. If we were to obtain an independent data set\\nand perform forward stepwise selection on that data set, we would likely\\nobtain a model containing a different, and perhaps even non-overlapping,\\nset of SNPs. This does not detract from the value of the model obtained—\\nfor instance, the model might turn out to be very effective in predicting\\nblood pressure on an independent set of patients, and might be clinically\\nuseful for physicians. But we must be careful not to overstate the results\\nobtained, and to make it clear that what we have identified is simply one\\nof many possible models for predicting blood pressure, and that it must be\\nfurther validated on independent data sets.\\nItisalsoimportanttobeparticularlycarefulinreportingerrorsandmea-\\nsures of model fit in the high-dimensional setting. We have seen that when\\np>n , it is easy to obtain a useless model that has zero residuals. There-\\nfore, one should neveruse sum of squared errors, p-values, R2statistics, or\\nother traditional measures of model fit on the training data as evidence of\\na good model fit in the high-dimensional setting. For instance, as we saw\\nin Figure 6.23, one can easily obtain a model with R2=1 whenp>n .\\nReporting this fact might mislead others into thinking that a statistically\\nvalid and useful model has been obtained, whereas in fact this provides\\nabsolutely no evidence of a compelling model. It is important to instead\\nreport results on an independent test set, or cross-validation errors. For\\ninstance, the MSE or R2on an independent test set is a valid measure of\\nmodel fit, but the MSE on the training set certainly is not.\\n6.5 Lab: Linear Models and Regularization\\nMethods\\nIn this lab we implement many of the techniques discussed in this chapter.\\nWe import some of our libraries at this top level.\\nIn[1]: import numpy as np\\nimport pandas as pd\\nfrom matplotlib.pyplot import subplots\\nfrom statsmodels.api import OLS\\nimport sklearn.model_selection as skm\\nimport sklearn.linear_model as skl\\nfrom sklearn.preprocessing import StandardScaler\\nfrom ISLP import load_data\\nfrom ISLP.models import ModelSpec as MS\\nfrom functools import partial\\nWe again collect the new imports needed for this lab.\\nIn[2]: from sklearn.pipeline import Pipeline\\nfrom sklearn.decomposition import PCA', metadata={'source': './documents/ISLP_website.pdf', 'page': 274}), Document(page_content='268 6. Linear Model Selection and Regularization\\nfrom sklearn.cross_decomposition import PLSRegression\\nfrom ISLP.models import \\\\\\n(Stepwise,\\nsklearn_selected,\\nsklearn_selection_path)\\n!pip install l0bnb\\nfrom l0bnb import fit_path\\nWe have installed the package l0bnbon the fly. Note the escaped !pip\\ninstall— this is run as a separate system command.\\n6.5.1 Subset Selection Methods\\nHere we implement methods that reduce the number of parameters in a\\nmodel by restricting the model to a subset of the input variables.\\nForward Selection\\nWe will apply the forward-selection approach to the Hittersdata. We wish\\nto predict a baseball player’s Salaryon the basis of various statistics asso-\\nciated with performance in the previous year.\\nFirst of all, we note that the Salaryvariable is missing for some of the\\nplayers. The np.isnan() function can be used to identify the missing ob-np.isnan()servations. It returns an array of the same shape as the input vector, with\\naTruefor any elements that are missing, and a Falsefor non-missing el-\\nements. The sum()method can then be used to count all of the missingsum()elements.\\nIn[3]: Hitters = load_data( \\'Hitters\\' )\\nnp.isnan(Hitters[ \\'Salary\\' ]).sum()\\nOut[3]: 59\\nWe see that Salaryis missing for 59 players. The dropna() method of data\\nframes removes all of the rows that have missing values in any variable (by\\ndefault — see Hitters.dropna? ).\\nIn[4]: Hitters = Hitters.dropna();\\nHitters.shape\\nOut[4]: (263, 20)\\nWe first choose the best model using forward selection based on Cp(6.2).\\nThis score is not built in as a metric to sklearn. We therefore define a\\nfunction to compute it ourselves, and use it as a scorer. By default, sklearn\\ntries to maximize a score, hence our scoring function computes the negative\\nCpstatistic.\\nIn[5]: def nCp(sigma2, estimator, X, Y):\\n\"Negative Cp statistic\"\\nn, p = X.shape\\nYhat = estimator.predict(X)\\nRSS = np.sum((Y - Yhat)**2)\\nreturn -(RSS + 2 * p * sigma2) / n', metadata={'source': './documents/ISLP_website.pdf', 'page': 275}), Document(page_content=\"6.5 Lab: Linear Models and Regularization Methods 269\\nWe need to estimate the residual variance σ2, which is the first argument\\nin our scoring function above. We will fit the biggest model, using all the\\nvariables, and estimate σ2based on its MSE.\\nIn[6]: design = MS(Hitters.columns.drop( 'Salary' )).fit(Hitters)\\nY=n p . a r r a y ( H i t t e r s [ 'Salary' ])\\nX=d e s i g n . t r a n s f o r m ( H i t t e r s )\\nsigma2 = OLS(Y,X).fit().scale\\nThe function sklearn_selected() expects a scorer with just three argu-\\nments — the last three in the definition of nCp()above. We use the function\\npartial() first seen in Section 5.3.3to freeze the first argument with our\\nestimate of σ2.\\nIn[7]: neg_Cp = partial(nCp, sigma2)\\nWe can now use neg_Cp() as a scorer for model selection.\\nAlong with a score we need to specify the search strategy. This is done\\nthrough the object Stepwise() in theISLP.models package. The method\\nStepwise.first_peak() runs forward stepwise until any further additions\\nto the model do not result in an improvement in the evaluation score.\\nSimilarly, the method Stepwise.fixed_steps() runs a fixed number of steps\\nof stepwise search.\\nIn[8]: strategy = Stepwise.first_peak(design,\\ndirection= 'forward' ,\\nmax_terms=len(design.terms))\\nWe now fit a linear regression model with Salaryas outcome using for-\\nward selection. To do so, we use the function sklearn_selected() from thesklearn_\\nselected()ISLP.models package. This takes a model from statsmodels along with a\\nsearch strategy and selects a model with its fitmethod. Without specify-\\ning ascoringargument, the score defaults to MSE, and so all 19 variables\\nwill be selected (output not shown).\\nIn[9]: hitters_MSE = sklearn_selected(OLS,\\nstrategy)\\nhitters_MSE.fit(Hitters, Y)\\nhitters_MSE.selected_state_\\nUsingneg_Cpresults in a smaller model, as expected, with just 10 vari-\\nables selected.\\nIn[10]: hitters_Cp = sklearn_selected(OLS,\\nstrategy,\\nscoring=neg_Cp)\\nhitters_Cp.fit(Hitters, Y)\\nhitters_Cp.selected_state_\\nOut[10]: ('Assists' ,\\n'AtBat' ,\\n'CAtBat' ,\\n'CRBI',\\n'CRuns' ,\\n'CWalks' ,\\n'Division' ,\", metadata={'source': './documents/ISLP_website.pdf', 'page': 276}), Document(page_content=\"270 6. Linear Model Selection and Regularization'Hits','PutOuts','Walks')Choosing Among Models Using the Validation Set Approach andCross-ValidationAs an alternative to usingCp, we might try cross-validation to select amodel in forward selection. For this, we need a method that stores thefull path of models found in forward selection, and allows predictions foreach of these. This can be done with thesklearn_selection_path()estima-sklearn_selection_path()tor fromISLP.models. The functioncross_val_predict()fromISLP.modelscross_val_predict()computes the cross-validated predictions for each of the models along thepath, which we can use to evaluate the cross-validated MSE along the path.Here we define a strategy that fits the full forward selection path. Whilethere are various parameter choices forsklearn_selection_path(), we usethe defaults here, which selects the model at each step based on the biggestreduction in RSS.In[11]:strategy = Stepwise.fixed_steps(design,len(design.terms),direction='forward')full_path = sklearn_selection_path(OLS, strategy)We now fit the full forward-selection path on theHittersdata and com-pute the fitted values.In[12]:full_path.fit(Hitters, Y)Yhat_in = full_path.predict(Hitters)Yhat_in.shapeOut[12]:(263, 20)This gives us an array of fitted values — 20 steps in all, including thefitted mean for the null model — which we can use to evaluate in-sampleMSE. As expected, the in-sample MSE improves each step we take, indicat-ing we must use either the validation or cross-validation approach to selectthe number of steps. We fix the y-axis to range from 50,000 to 250,000to compare to the cross-validation and validation set MSE below, as wellas other methods such as ridge regression, lasso and principal componentsregression.In[13]:mse_fig, ax = subplots(figsize=(8,8))insample_mse = ((Yhat_in - Y[:,None])**2).mean(0)n_steps = insample_mse.shape[0]ax.plot(np.arange(n_steps),insample_mse,'k',#c o l o rb l a c klabel='In-sample')ax.set_ylabel('MSE',fontsize=20)ax.set_xlabel('# steps of forward stepwise',fontsize=20)ax.set_xticks(np.arange(n_steps)[::2])ax.legend()\", metadata={'source': './documents/ISLP_website.pdf', 'page': 277}), Document(page_content='6.5 Lab: Linear Models and Regularization Methods 271\\nax.set_ylim([50000,250000]);\\nNotice the expression NoneinY[:,None] above. This adds an axis (dimen-\\nsion) to the one-dimensional array Y, which allows it to be recycled when\\nsubtracted from the two-dimensional Yhat_in.\\nWe are now ready to use cross-validation to estimate test error along\\nthe model path. We must use only the training observations to perform\\nall aspects of model-fitting — including variable selection. Therefore, the\\ndetermination of which model of a given size is best must be made using\\nonly the training observations in each training fold. This point is subtle but\\nimportant. If the full data set is used to select the best subset at each step,\\nthen the validation set errors and cross-validation errors that we obtain\\nwill not be accurate estimates of the test error.\\nWe now compute the cross-validated predicted values using 5-fold cross-\\nvalidation.\\nIn[14]: K=5\\nkfold = skm.KFold(K,\\nrandom_state=0,\\nshuffle=True)\\nYhat_cv = skm.cross_val_predict(full_path,\\nHitters,\\nY,\\ncv=kfold)\\nYhat_cv.shape\\nskm.KFold()\\nskm.cross_\\nval_predict()Out[14]: (263, 20)\\nThe prediction matrix Yhat_cvis the same shape as Yhat_in; the difference\\nis that the predictions in each row, corresponding to a particular sample\\nindex, were made from models fit on a training fold that did not include\\nthat row.\\nAt each model along the path, we compute the MSE in each of the cross-\\nvalidation folds. These we will average to get the mean MSE, and can also\\nuse the individual values to compute a crude estimate of the standard error\\nof the mean.9Hence we must know the test indices for each cross-validation\\nsplit. This can be found by using the split()method of kfold. Because we\\nfixed the random state above, whenever we split any array with the same\\nnumber of rows as Ywe recover the same training and test indices, though\\nwe simply ignore the training indices below.\\nIn[15]: cv_mse = []\\nfor train_idx, test_idx in kfold.split(Y):\\nerrors = (Yhat_cv[test_idx] - Y[test_idx,None])**2\\ncv_mse.append(errors.mean(0)) #c o l u m nm e a n s\\ncv_mse = np.array(cv_mse).T\\ncv_mse.shape\\nOut[15]: (20, 5)\\n9The estimate is crude because the five error estimates are based on overlapping\\ntraining sets, and hence are not independent.', metadata={'source': './documents/ISLP_website.pdf', 'page': 278}), Document(page_content=\"272 6. Linear Model Selection and RegularizationWe now add the cross-validation error estimates to our MSE plot. Weincludethemeanerroracrossthefivefolds,andtheestimateofthestandarderror of the mean.In[16]:ax.errorbar(np.arange(n_steps),cv_mse.mean(1),cv_mse.std(1) / np.sqrt(K),label='Cross-validated',c='r')#c o l o rr e dax.set_ylim([50000,250000])ax.legend()mse_figTo repeat the above using the validation set approach, we simply changeourcvargument to a validation set: one random split of the data into atest and training. We choose a test size of 20%, similar to the size of eachtest set in 5-fold cross-validation.skm.ShuffleSplit()In[17]:validation = skm.ShuffleSplit(n_splits=1,test_size=0.2,random_state=0)for train_idx, test_idx in validation.split(Y):full_path.fit(Hitters.iloc[train_idx],Y[train_idx])Yhat_val = full_path.predict(Hitters.iloc[test_idx])errors = (Yhat_val - Y[test_idx,None])**2validation_mse = errors.mean(0)Asforthein-sampleMSEcase,thevalidationsetapproachdoesnotprovidestandard errors.In[18]:ax.plot(np.arange(n_steps),validation_mse,'b--',#c o l o rb l u e ,b r o k e nl i n elabel='Validation')ax.set_xticks(np.arange(n_steps)[::2])ax.set_ylim([50000,250000])ax.legend()mse_figBest Subset SelectionForward stepwise is agreedyselection procedure; at each step it augmentsthe current set by including one additional variable. We now apply bestsubset selection to theHittersdata, which for every subset size, searchesfor the best set of predictors.We will use a package calledl0bnbto perform best subset selection.Insteadofconstrainingthesubsettobeagivensize,thispackageproducesapath of solutions using the subset size as a penalty rather than a constraint.Although the distinction is subtle, the difference comes when we cross-validate.In[19]:D=d e s i g n . f i t _ t r a n s f o r m ( H i t t e r s )D=D . d r o p ('intercept',a x i s = 1 )X=n p . a s a r r a y ( D )\", metadata={'source': './documents/ISLP_website.pdf', 'page': 279}), Document(page_content=\"6.5 Lab: Linear Models and Regularization Methods 273\\nHere we excluded the first column corresponding to the intercept, as l0bnb\\nwill fit the intercept separately. We can find a path using the fit_path()\\nfunction.\\nIn[20]: path = fit_path(X,\\nY,\\nmax_nonzeros=X.shape[1])\\nThe function fit_path() returns a list whose values include the fitted\\ncoefficients as B, an intercept as B0, as well as a few other attributes related\\nto the particular path algorithm used. Such details are beyond the scope\\nof this book.\\nIn[21]: path[3]\\nOut[21]: {'B':a r r a y ( [ 0 . ,3 . 2 5 4 8 4 4 ,0 . ,0 . ,0 . ,\\n0. , 0. , 0. , 0. , 0. ,\\n0. , 0.677753 , 0. , 0. , 0. ,\\n0. , 0. , 0. , 0. ]),\\n'B0':- 3 8 . 9 8 2 1 6 7 3 9 5 5 5 4 9 4 ,\\n'lambda_0' :0 . 0 1 1 4 1 6 2 4 8 0 2 7 4 5 0 1 9 4 ,\\n'M':0 . 5 8 2 9 8 6 1 7 3 3 3 8 2 0 1 1 ,\\n'Time_exceeded' :F a l s e }\\nIn the example above, we see that at the fourth step in the path, we have\\ntwo nonzero coefficients in 'B', corresponding to the value 0.114for the\\npenaltyparameter lambda_0 .Wecouldmakepredictionsusingthissequence\\nof fits on a validation set as a function of lambda_0 , or with more work using\\ncross-validation.\\n6.5.2 Ridge Regression and the Lasso\\nWe will use the sklearn.linear_model package (for which we use sklas\\nshorthand below) to fit ridge and lasso regularized linear models on the\\nHitters data. We start with the model matrix X(without an intercept)\\nthat we computed in the previous section on best subset regression.\\nRidge Regression\\nWewillusethefunction skl.ElasticNet() tofitbothridgeandthelasso.Toskl.Elastic\\nNet() fit apathof ridge regressions models, we use skl.ElasticNet.path() , which\\nskl.Elastic\\nNet.path()can fit both ridge and lasso, as well as a hybrid mixture; ridge regression\\ncorresponds to l1_ratio=0 . It is good practice to standardize the columns\\nofXin these applications, if the variables are measured in different units.\\nSinceskl.ElasticNet() does no normalization, we have to take care of that\\nourselves. Since we standardize first, in order to find coefficient estimates\\non the original scale, we must unstandardize the coefficient estimates. The\\nparameter λin (6.5) and (6.7) is called alphasinsklearn. In order to be\\nconsistent with the rest of this chapter, we use lambdasrather than alphas\\nin what follows.10\\n10At the time of publication, ridge fits like the one in code chunk [22] issue unwarranted\\nconvergence warning messages; we expect these to disappear as this package matures.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 280}), Document(page_content=\"274 6. Linear Model Selection and Regularization\\nIn[22]: Xs = X - X.mean(0)[None,:]\\nX_scale = X.std(0)\\nXs = Xs / X_scale[None,:]\\nlambdas = 10**np.linspace(8, -2, 100) / Y.std()\\nsoln_array = skl.ElasticNet.path(Xs,\\nY,\\nl1_ratio=0.,\\nalphas=lambdas)[1]\\nsoln_array.shape\\nOut[22]: (19, 100)\\nHere we extract the array of coefficients corresponding to the solutions\\nalong the regularization path. By default the skl.ElasticNet.path method\\nfits a path along an automatically selected range of λvalues, except for\\nthe case when l1_ratio=0 , which results in ridge regression (as is the case\\nhere).11So here we have chosen to implement the function over a grid of\\nvalues ranging from λ= 108toλ= 10−2scaled by the standard deviation\\nofy, essentially covering the full range of scenarios from the null model\\ncontaining only the intercept, to the least squares fit.\\nAssociated with each value of λis a vector of ridge regression coefficients,\\nthat can be accessed by a column of soln_array . In this case, soln_array is\\na19×100matrix, with 19 rows (one for each predictor) and 100 columns\\n(one for each value of λ).\\nWe transpose this matrix and turn it into a data frame to facilitate\\nviewing and plotting.\\nIn[23]: soln_path = pd.DataFrame(soln_array.T,\\ncolumns=D.columns,\\nindex=-np.log(lambdas))\\nsoln_path.index.name = 'negative log(lambda)'\\nsoln_path\\nOut[23]: AtBat Hits HmRun Runs ...\\nnegative\\nlog(lambda)\\n-12.310855 0.000800 0.000889 0.000695 0.000851 ...\\n-12.078271 0.001010 0.001122 0.000878 0.001074 ...\\n-11.845686 0.001274 0.001416 0.001107 0.001355 ...\\n-11.613102 0.001608 0.001787 0.001397 0.001710 ...\\n-11.380518 0.002029 0.002255 0.001763 0.002158 ...\\n... ... ... ... ... ...\\n100 rows × 19 columns\\nWe plot the paths to get a sense of how the coefficients vary with λ.T o\\ncontrol the location of the legend we first set legendtoFalsein the plot\\nmethod, adding it afterward with the legend() method of ax.\\nIn[24]: path_fig, ax = subplots(figsize=(8,8))\\nsoln_path.plot(ax=ax, legend=False)\\nax.set_xlabel( '$-\\\\log(\\\\lambda)$' ,f o n t s i z e = 2 0 )\\n11The reason is rather technical; for all models except ridge, we can find the smallest\\nvalue of λfor which all coefficients are zero. For ridge this value is ∞.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 281}), Document(page_content=\"6.5 Lab: Linear Models and Regularization Methods 275ax.set_ylabel('Standardized coefficients',f o n t s i z e = 2 0 )ax.legend(loc='upper left');(We have usedlatexformatting in the horizontal label, in order to formatthe Greekλappropriately.) We expect the coefficient estimates to be muchsmaller, in terms ofℓ2norm, when a large value ofλis used, as comparedto when a small value ofλis used. (Recall that theℓ2norm is the squareroot of the sum of squared coefficient values.) We display the coefficientsat the40th step, whereλis 25.535.In[25]:beta_hat = soln_path.loc[soln_path.index[39]]lambdas[39], beta_hatOut[25]:(25.535,AtBat5.433750Hits6.223582HmRun4.585498Runs5.880855RBI6.195921Walks6.277975Years5.299767......Let’s compute theℓ2norm of the standardized coefficients.In[26]:np.linalg.norm(beta_hat)Out[26]:24.17In contrast, here is theℓ2norm whenλis 2.44e-01. Note the much largerℓ2norm of the coefficients associated with this smaller value ofλ.In[27]:beta_hat = soln_path.loc[soln_path.index[59]]lambdas[59], np.linalg.norm(beta_hat)Out[27]:(0.2437, 160.4237)Above we normalizedXupfront, and fit the ridge model usingXs. ThePipeline()object insklearnprovides a clear way to separate feature nor-malization from the fitting of the ridge model itself.In[28]:ridge = skl.ElasticNet(alpha=lambdas[59], l1_ratio=0)scaler = StandardScaler(with_mean=True, with_std=True)pipe = Pipeline(steps=[('scaler',s c a l e r ) ,('ridge',r i d g e ) ] )pipe.fit(X, Y)We show that it gives the sameℓ2norm as in our previous fit on thestandardized data.In[29]:np.linalg.norm(ridge.coef_)Out[29]:160.4237Notice that the operationpipe.fit(X, Y)above has changed theridgeobject, and in particular has added attributes such ascoef_that were notthere before.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 282}), Document(page_content=\"276 6. Linear Model Selection and RegularizationEstimating Test Error of Ridge RegressionChoosing ana priorivalue ofλfor ridge regression is difficult if not im-possible. We will want to use the validation method or cross-validation toselect the tuning parameter. The reader may not be surprised that thePipeline()approach can be used inskm.cross_validate()with either avalidation method (i.e.validation) ork-fold cross-validation.We fix the random state of the splitter so that the results obtained willbe reproducible.In[30]:validation = skm.ShuffleSplit(n_splits=1,test_size=0.5,random_state=0)ridge.alpha = 0.01results = skm.cross_validate(ridge,X,Y,scoring='neg_mean_squared_error',cv=validation)-results['test_score']Out[30]:array([134214.0])The test MSE is 1.342e+05. Note that if we had instead simply fit a modelwith just an intercept, we would have predicted each test observation usingthe mean of the training observations. We can get the same result by fittinga ridge regression model with averylarge value ofλ. Note that1e10means1010.In[31]:ridge.alpha = 1e10results = skm.cross_validate(ridge,X,Y,scoring='neg_mean_squared_error',cv=validation)-results['test_score']Out[31]:array([231788.32])Obviously choosingλ=0.01is arbitrary, so we will use cross-validation orthe validation-set approach to choose the tuning parameterλ. The objectGridSearchCV()allows exhaustive grid search to choose such a parameter.GridSearchCV()We first use the validation set method to chooseλ.In[32]:param_grid = {'ridge__alpha':l a m b d a s }grid = skm.GridSearchCV(pipe,param_grid,cv=validation,scoring='neg_mean_squared_error')grid.fit(X, Y)grid.best_params_['ridge__alpha']grid.best_estimator_Out[32]:Pipeline(steps=[('scaler',S t a n d a r d S c a l e r ( ) ) ,('ridge',E l a s t i c N e t ( a l p h a = 0 . 0 0 5 8 9 9 ,l 1 _ r a t i o = 0 ) ) ] )\", metadata={'source': './documents/ISLP_website.pdf', 'page': 283}), Document(page_content=\"6.5 Lab: Linear Models and Regularization Methods 277\\nAlternatively, we can use 5-fold cross-validation.\\nIn[33]: grid = skm.GridSearchCV(pipe,\\nparam_grid,\\ncv=kfold,\\nscoring= 'neg_mean_squared_error' )\\ngrid.fit(X, Y)\\ngrid.best_params_[ 'ridge__alpha' ]\\ngrid.best_estimator_\\nRecall we set up the kfoldobject for 5-fold cross-validation on page 271.\\nWe now plot the cross-validated MSE as a function of −log(λ), which has\\nshrinkage decreasing from left to right.\\nIn[34]: ridge_fig, ax = subplots(figsize=(8,8))\\nax.errorbar(-np.log(lambdas),\\n-grid.cv_results_[ 'mean_test_score' ],\\nyerr=grid.cv_results_[ 'std_test_score' ]/n p . s q r t ( K ) )\\nax.set_ylim([50000,250000])\\nax.set_xlabel( '$-\\\\log(\\\\lambda)$' ,f o n t s i z e = 2 0 )\\nax.set_ylabel( 'Cross-validated MSE' ,f o n t s i z e = 2 0 ) ;\\nOne can cross-validate different metrics to choose a parameter. The de-\\nfault metric for skl.ElasticNet() is testR2. Let’s compare R2to MSE for\\ncross-validation here.\\nIn[35]: grid_r2 = skm.GridSearchCV(pipe,\\nparam_grid,\\ncv=kfold)\\ngrid_r2.fit(X, Y)\\nFinally, let’s plot the results for cross-validated R2.\\nIn[36]: r2_fig, ax = subplots(figsize=(8,8))\\nax.errorbar(-np.log(lambdas),\\ngrid_r2.cv_results_[ 'mean_test_score' ],\\nyerr=grid_r2.cv_results_[ 'std_test_score' ]/n p . s q r t ( K )\\n)\\nax.set_xlabel( '$-\\\\log(\\\\lambda)$' ,f o n t s i z e = 2 0 )\\nax.set_ylabel( 'Cross-validated $R^2$' ,f o n t s i z e = 2 0 ) ;\\nFast Cross-Validation for Solution Paths\\nThe ridge, lasso, and elastic net can be efficiently fit along a sequence of\\nλvalues, creating what is known as a solution path orregularization path .\\nHence there is specialized code to fit such paths, and to choose a suitable\\nvalue ofλusing cross-validation. Even with identical splits the results will\\nnot agree exactlywith our gridabove because the standardization of each\\nfeature in gridis carried out on each fold, while in pipeCVbelow it is carried\\nout only once. Nevertheless, the results are similar as the normalization is\\nrelatively stable across folds.\\nIn[37]: ridgeCV = skl.ElasticNetCV(alphas=lambdas,\\nl1_ratio=0,\\ncv=kfold)\\npipeCV = Pipeline(steps=[( 'scaler' ,s c a l e r ) ,\", metadata={'source': './documents/ISLP_website.pdf', 'page': 284}), Document(page_content=\"278 6. Linear Model Selection and Regularization\\n('ridge' ,r i d g e C V ) ] )\\npipeCV.fit(X, Y)\\nLet’s produce a plot again of the cross-validation error to see that it is\\nsimilar to using skm.GridSearchCV .\\nIn[38]: tuned_ridge = pipeCV.named_steps[ 'ridge' ]\\nridgeCV_fig, ax = subplots(figsize=(8,8))\\nax.errorbar(-np.log(lambdas),\\ntuned_ridge.mse_path_.mean(1),\\nyerr=tuned_ridge.mse_path_.std(1) / np.sqrt(K))\\nax.axvline(-np.log(tuned_ridge.alpha_), c= 'k',l s ='--')\\nax.set_ylim([50000,250000])\\nax.set_xlabel( '$-\\\\log(\\\\lambda)$' ,f o n t s i z e = 2 0 )\\nax.set_ylabel( 'Cross-validated MSE' ,f o n t s i z e = 2 0 ) ;\\nWe see that the value of λthat results in the smallest cross-validation\\nerror is 1.19e-02, available as the value tuned_ridge.alpha_ . What is the\\ntest MSE associated with this value of λ?\\nIn[39]: np.min(tuned_ridge.mse_path_.mean(1))\\nOut[39]: 115526.71\\nThis represents a further improvement over the test MSE that we got using\\nλ=4. Finally, tuned_ridge.coef_ has the coefficients fit on the entire data\\nset at this value of λ.\\nIn[40]: tuned_ridge.coef_\\nOut[40]: array([-222.80877051, 238.77246614, 3.21103754, -2.93050845,\\n3.64888723, 108.90953869, -50.81896152, -105.15731984,\\n122.00714801, 57.1859509 , 210.35170348, 118.05683748,\\n-150.21959435, 30.36634231, -61.62459095, 77.73832472,\\n40.07350744, -25.02151514, -13.68429544])\\nAs expected, none of the coefficients are zero—ridge regression does not\\nperform variable selection!\\nEvaluating Test Error of Cross-Validated Ridge\\nChoosing λusing cross-validation provides a single regression estimator,\\nsimilar to fitting a linear regression model as we saw in Chapter 3. It is\\ntherefore reasonable to estimate what its test error is. We run into a prob-\\nlem here in that cross-validation will have touchedall of its data in choosing\\nλ, hence we have no further data to estimate test error. A compromise is\\nto do an initial split of the data into two disjoint sets: a training set and a\\ntest set. We then fit a cross-validation tuned ridge regression on the train-\\ning set, and evaluate its performance on the test set. We might call this\\ncross-validation nested within the validation set approach. A priori there\\nis no reason to use half of the data for each of the two sets in validation.\\nBelow, we use 75% for training and 25% for test, with the estimator being\\nridge regression tuned using 5-fold cross-validation. This can be achieved\\nin code as follows:\", metadata={'source': './documents/ISLP_website.pdf', 'page': 285}), Document(page_content=\"6.5 Lab: Linear Models and Regularization Methods 279In[41]:outer_valid = skm.ShuffleSplit(n_splits=1,test_size=0.25,random_state=1)inner_cv = skm.KFold(n_splits=5,shuffle=True,random_state=2)ridgeCV = skl.ElasticNetCV(alphas=lambdas,l1_ratio=0,cv=inner_cv)pipeCV = Pipeline(steps=[('scaler',s c a l e r ) ,('ridge',r i d g e C V ) ] ) ;In[42]:results = skm.cross_validate(pipeCV,X,Y,cv=outer_valid,scoring='neg_mean_squared_error')-results['test_score']Out[42]:array([132393.84])The LassoWe saw that ridge regression with a wise choice ofλcan outperform leastsquares as well as the null model on theHittersdata set. We now askwhether the lasso can yield either a more accurate or a more interpretablemodel than ridge regression. In order to fit a lasso model, we once againuse theElasticNetCV()function; however, this time we use the argumentl1_ratio=1. Other than that change, we proceed just as we did in fitting aridge model.In[43]:lassoCV = skl.ElasticNetCV(n_alphas=100,l1_ratio=1,cv=kfold)pipeCV = Pipeline(steps=[('scaler',s c a l e r ) ,('lasso',l a s s o C V ) ] )pipeCV.fit(X, Y)tuned_lasso = pipeCV.named_steps['lasso']tuned_lasso.alpha_Out[43]:3.147In[44]:lambdas, soln_array = skl.Lasso.path(Xs,Y,l1_ratio=1,n_alphas=100)[:2]soln_path = pd.DataFrame(soln_array.T,columns=D.columns,index=-np.log(lambdas))We can see from the coefficient plot of the standardized coefficients thatdepending on the choice of tuning parameter, some of the coefficients willbe exactly equal to zero.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 286}), Document(page_content=\"280 6. Linear Model Selection and Regularization\\nIn[45]: path_fig, ax = subplots(figsize=(8,8))\\nsoln_path.plot(ax=ax, legend=False)\\nax.legend(loc= 'upper left' )\\nax.set_xlabel( '$-\\\\log(\\\\lambda)$' ,f o n t s i z e = 2 0 )\\nax.set_ylabel( 'Standardized coefficiients' ,f o n t s i z e = 2 0 ) ;\\nThe smallest cross-validated error is lower than the test set MSE of the null\\nmodel and of least squares, and very similar to the test MSE of 115526.71\\nof ridge regression (page 278) withλchosen by cross-validation.\\nIn[46]: np.min(tuned_lasso.mse_path_.mean(1))\\nOut[46]: 114690.73\\nLet’s again produce a plot of the cross-validation error.\\nIn[47]: lassoCV_fig, ax = subplots(figsize=(8,8))\\nax.errorbar(-np.log(tuned_lasso.alphas_),\\ntuned_lasso.mse_path_.mean(1),\\nyerr=tuned_lasso.mse_path_.std(1) / np.sqrt(K))\\nax.axvline(-np.log(tuned_lasso.alpha_), c= 'k',l s ='--')\\nax.set_ylim([50000,250000])\\nax.set_xlabel( '$-\\\\log(\\\\lambda)$' ,f o n t s i z e = 2 0 )\\nax.set_ylabel( 'Cross-validated MSE' ,f o n t s i z e = 2 0 ) ;\\nHowever, the lasso has a substantial advantage over ridge regression in\\nthat the resulting coefficient estimates are sparse. Here we see that 6 of the\\n19 coefficient estimates are exactly zero. So the lasso model with λchosen\\nby cross-validation contains only 13 variables.\\nIn[48]: tuned_lasso.coef_\\nOut[48]: array([-210.01008773, 243.4550306 , 0. , 0. ,\\n0. , 97.69397357, -41.52283116, -0. ,\\n0. , 39.62298193, 205.75273856, 124.55456561,\\n-126.29986768, 15.70262427, -59.50157967, 75.24590036,\\n21.62698014, -12.04423675, -0. ])\\nAs in ridge regression, we could evaluate the test error of cross-validated\\nlasso by first splitting into test and training sets and internally running\\ncross-validation on the training set. We leave this as an exercise.\\n6.5.3 PCR and PLS Regression\\nPrincipal Components Regression\\nPrincipal components regression (PCR) can be performed using PCA()fromPCA()thesklearn.decomposition module. We now apply PCR to the Hitters\\ndata, in order to predict Salary. Again, ensure that the missing values\\nhave been removed from the data, as described in Section 6.5.1.\\nWe use LinearRegression() to fit the regression model here. Note thatLinear\\nRegression() it fits an intercept by default, unlike the OLS()function seen earlier in\\nSection6.5.1.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 287}), Document(page_content=\"6.5 Lab: Linear Models and Regularization Methods 281In[49]:pca = PCA(n_components=2)linreg = skl.LinearRegression()pipe = Pipeline([('pca',p c a ) ,('linreg',l i n r e g ) ] )pipe.fit(X, Y)pipe.named_steps['linreg'].coef_Out[49]:array([0.09846131, 0.4758765 ])When performing PCA, the results vary depending on whether the datahas beenstandardizedor not. As in the earlier examples, this can be ac-complished by including an additional step in the pipeline.In[50]:pipe = Pipeline([('scaler',s c a l e r ) ,('pca',p c a ) ,('linreg',l i n r e g ) ] )pipe.fit(X, Y)pipe.named_steps['linreg'].coef_Out[50]:array([106.36859204, -21.60350456])We can of course use CV to choose the number of components, byusingskm.GridSearchCV, in this case fixing the parameters to vary then_components.In[51]:param_grid = {'pca__n_components':r a n g e ( 1 ,2 0 ) }grid = skm.GridSearchCV(pipe,param_grid,cv=kfold,scoring='neg_mean_squared_error')grid.fit(X, Y)Let’s plot the results as we have for other methods.In[52]:pcr_fig, ax = subplots(figsize=(8,8))n_comp = param_grid['pca__n_components']ax.errorbar(n_comp,-grid.cv_results_['mean_test_score'],grid.cv_results_['std_test_score']/n p . s q r t ( K ) )ax.set_ylabel('Cross-validated MSE',f o n t s i z e = 2 0 )ax.set_xlabel('# principal components',f o n t s i z e = 2 0 )ax.set_xticks(n_comp[::2])ax.set_ylim([50000,250000]);We see that the smallest cross-validation error occurs when 17 compo-nents are used. However, from the plot we also see that the cross-validationerror is roughly the same when only one component is included in themodel. This suggests that a model that uses just a small number of com-ponents might suffice.The CV score is provided for each possible number of components from1 to 19 inclusive. ThePCA()method complains if we try to fit an interceptonly withn_components=0so we also compute the MSE for just the nullmodel with these splits.In[53]:Xn = np.zeros((X.shape[0], 1))cv_null = skm.cross_validate(linreg,\", metadata={'source': './documents/ISLP_website.pdf', 'page': 288}), Document(page_content=\"282 6. Linear Model Selection and Regularization\\nXn,\\nY,\\ncv=kfold,\\nscoring= 'neg_mean_squared_error' )\\n-cv_null[ 'test_score' ].mean()\\nOut[53]: 204139.31\\nTheexplained_variance_ratio_ attribute of our PCAobject provides the\\npercentage of variance explained in the predictors and in the response using\\ndifferent numbers of components. This concept is discussed in greater detail\\nin Section 12.2.\\nIn[54]: pipe.named_steps[ 'pca'].explained_variance_ratio_\\nOut[54]: array([0.3831424 , 0.21841076])\\nBriefly, we can think of this as the amount of information about the predic-\\ntors that is captured using Mprincipal components. For example, setting\\nM=1only captures 38.31% of the variance, while M=2captures an ad-\\nditional 21.84%, for a total of 60.15% of the variance. By M=6it increases\\nto 88.63%. Beyond this the increments continue to diminish, until we use\\nallM=p= 19components, which captures all 100% of the variance.\\nPartial Least Squares\\nPartialleastsquares(PLS)isimplementedinthe PLSRegression() function.PLS\\nRegression()In[55]: pls = PLSRegression(n_components=2,\\nscale=True)\\npls.fit(X, Y)\\nAs was the case in PCR, we will want to use CV to choose the number\\nof components.\\nIn[56]: param_grid = { 'n_components' :range(1, 20)}\\ngrid = skm.GridSearchCV(pls,\\nparam_grid,\\ncv=kfold,\\nscoring= 'neg_mean_squared_error' )\\ngrid.fit(X, Y)\\nAs for our other methods, we plot the MSE.\\nIn[57]: pls_fig, ax = subplots(figsize=(8,8))\\nn_comp = param_grid[ 'n_components' ]\\nax.errorbar(n_comp,\\n-grid.cv_results_[ 'mean_test_score' ],\\ngrid.cv_results_[ 'std_test_score' ]/n p . s q r t ( K ) )\\nax.set_ylabel( 'Cross-validated MSE' ,f o n t s i z e = 2 0 )\\nax.set_xlabel( '# principal components' ,f o n t s i z e = 2 0 )\\nax.set_xticks(n_comp[::2])\\nax.set_ylim([50000,250000]);\\nCV error is minimized at 12, though there is little noticable difference\\nbetween this point and a much lower number like 2 or 3 components.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 289}), Document(page_content='6.6 Exercises 2836.6 ExercisesConceptual1. We perform best subset, forward stepwise, and backward stepwiseselection on a single data set. For each approach, we obtainp+1models, containing0,1,2,...,ppredictors. Explain your answers:(a) Which of the three models withkpredictors has the smallesttrainingRSS?(b) Which of the three models withkpredictors has the smallesttestRSS?(c) True or False:i. The predictors in thek-variable model identified by forwardstepwise are a subset of the predictors in the(k+1)-variablemodel identified by forward stepwise selection.ii. The predictors in thek-variable model identified by back-ward stepwise are a subset of the predictors in the(k+ 1)-variable model identified by backward stepwise selection.iii. The predictors in thek-variable model identified by back-ward stepwise are a subset of the predictors in the(k+ 1)-variable model identified by forward stepwise selection.iv. The predictors in thek-variable model identified by forwardstepwise are a subset of the predictors in the(k+1)-variablemodel identified by backward stepwise selection.v. The predictors in thek-variable model identified by bestsubset are a subset of the predictors in the(k+ 1)-variablemodel identified by best subset selection.2. For parts (a) through (c), indicate which of i. through iv. is correct.Justify your answer.(a) The lasso, relative to least squares, is:i. More flexible and hence will give improved prediction ac-curacy when its increase in bias is less than its decrease invariance.ii. More flexible and hence will give improved prediction accu-racy when its increase in variance is less than its decreasein bias.iii. Less flexible and hence will give improved prediction accu-racy when its increase in bias is less than its decrease invariance.iv. Less flexible and hence will give improved prediction accu-racy when its increase in variance is less than its decreasein bias.(b) Repeat (a) for ridge regression relative to least squares.(c) Repeat (a) for non-linear methods relative to least squares.', metadata={'source': './documents/ISLP_website.pdf', 'page': 290}), Document(page_content='284 6. Linear Model Selection and Regularization3. Suppose we estimate the regression coefficients in a linear regressionmodel by minimizingn∑i=1\\uf8eb\\uf8edyi−β0−p∑j=1βjxij\\uf8f6\\uf8f82subject top∑j=1|βj|≤sfor a particular value ofs. For parts (a) through (e), indicate whichof i. through v. is correct. Justify your answer.(a) As we increasesfrom 0, the training RSS will:i. Increase initially, and then eventually start decreasing in aninverted U shape.ii. Decrease initially, and then eventually start increasing in aU shape.iii. Steadily increase.iv. Steadily decrease.v. Remain constant.(b) Repeat (a) for test RSS.(c) Repeat (a) for variance.(d) Repeat (a) for (squared) bias.(e) Repeat (a) for the irreducible error.4. Suppose we estimate the regression coefficients in a linear regressionmodel by minimizingn∑i=1\\uf8eb\\uf8edyi−β0−p∑j=1βjxij\\uf8f6\\uf8f82+λp∑j=1β2jfor a particular value ofλ. For parts (a) through (e), indicate whichof i. through v. is correct. Justify your answer.(a) As we increaseλfrom 0, the training RSS will:i. Increase initially, and then eventually start decreasing in aninverted U shape.ii. Decrease initially, and then eventually start increasing in aU shape.iii. Steadily increase.iv. Steadily decrease.v. Remain constant.(b) Repeat (a) for test RSS.(c) Repeat (a) for variance.(d) Repeat (a) for (squared) bias.(e) Repeat (a) for the irreducible error.', metadata={'source': './documents/ISLP_website.pdf', 'page': 291}), Document(page_content='6.6 Exercises 285\\n5. It is well-known that ridge regression tends to give similar coefficient\\nvalues to correlated variables, whereas the lasso may give quite dif-\\nferent coefficient values to correlated variables. We will now explore\\nthis property in a very simple setting.\\nSuppose that n=2,p=2,x11=x12,x21=x22. Furthermore,\\nsuppose that y1+y2=0andx11+x21=0andx12+x22=0, so that\\nthe estimate for the intercept in a least squares, ridge regression, or\\nlasso model is zero: ˆβ0=0.\\n(a) Write out the ridge regression optimization problem in this set-\\nting.\\n(b) Argue that in this setting, the ridge coefficient estimates satisfy\\nˆβ1=ˆβ2.\\n(c) Write out the lasso optimization problem in this setting.\\n(d) Argue that in this setting, the lasso coefficients ˆβ1andˆβ2are\\nnot unique—in other words, there are many possible solutions\\nto the optimization problem in (c). Describe these solutions.\\n6. We will now explore ( 6.12) and (6.13) further.\\n(a) Consider ( 6.12) withp=1. For some choice of y1andλ>0,\\nplot (6.12) as a function of β1. Your plot should confirm that\\n(6.12) is solved by ( 6.14).\\n(b) Consider ( 6.13) withp=1. For some choice of y1andλ>0,\\nplot (6.13) as a function of β1. Your plot should confirm that\\n(6.13) is solved by ( 6.15).\\n7. We will now derive the Bayesian connection to the lasso and ridge\\nregression discussed in Section 6.2.2.\\n(a) Suppose that yi=β0+∑p\\nj=1xijβj+ϵiwhereϵ1,...,ϵnare inde-\\npendentandidenticallydistributedfroma N(0,σ2)distribution.\\nWrite out the likelihood for the data.\\n(b) Assume the following prior for β:β1,...,βpare independent\\nand identically distributed according to a double-exponential\\ndistribution with mean 0 and common scale parameter b: i.e.\\np(β)=1\\n2bexp(−|β|/b). Write out the posterior for βin this\\nsetting.\\n(c) Argue that the lasso estimate is the modeforβunder this pos-\\nterior distribution.\\n(d) Nowassumethefollowingpriorfor β:β1,...,βpareindependent\\nand identically distributed according to a normal distribution\\nwith mean zero and variance c. Write out the posterior for βin\\nthis setting.\\n(e) Argue that the ridge regression estimate is both the modeand\\nthemeanforβunder this posterior distribution.', metadata={'source': './documents/ISLP_website.pdf', 'page': 292}), Document(page_content='286 6. Linear Model Selection and RegularizationApplied8. In this exercise, we will generate simulated data, and will then usethis data to perform forward and backward stepwise selection.(a) Create a random number generator and use itsnormal()methodto generate a predictorXof lengthn= 100, as well as a noisevectorϵof lengthn= 100.(b) Generate a response vectorYof lengthn= 100according tothe modelY=β0+β1X+β2X2+β3X3+ϵ,whereβ0,β1,β2, andβ3are constants of your choice.(c) Useforwardstepwiseselectioninordertoselectamodelcontain-ing the predictorsX,X2,...,X10. What is the model obtainedaccording toCp? Report the coefficients of the model obtained.(d) Repeat (c), using backwards stepwise selection. How does youranswer compare to the results in (c)?(e) Now fit a lasso model to the simulated data, again usingX,X2,...,X10as predictors. Use cross-validation to select the optimalvalueofλ.Createplotsofthecross-validationerrorasafunctionofλ. Report the resulting coefficient estimates, and discuss theresults obtained.(f) Now generate a response vectorYaccording to the modelY=β0+β7X7+ϵ,and perform forward stepwise selection and the lasso. Discussthe results obtained.9. In this exercise, we will predict the number of applications receivedusing the other variables in theCollegedata set.(a) Split the data set into a training set and a test set.(b) Fit a linear model using least squares on the training set, andreport the test error obtained.(c) Fit a ridge regression model on the training set, withλchosenby cross-validation. Report the test error obtained.(d) Fit a lasso model on the training set, withλchosen by cross-validation. Report the test error obtained, along with the num-ber of non-zero coefficient estimates.(e) Fit a PCR model on the training set, withMchosen by cross-validation. Report the test error obtained, along with the valueofMselected by cross-validation.(f) Fit a PLS model on the training set, withMchosen by cross-validation. Report the test error obtained, along with the valueofMselected by cross-validation.', metadata={'source': './documents/ISLP_website.pdf', 'page': 293}), Document(page_content='6.6 Exercises 287(g) Comment on the results obtained. How accurately can we pre-dict the number of college applications received? Is there muchdifference among the test errors resulting from these five ap-proaches?10. Wehaveseenthatasthenumberoffeaturesusedinamodelincreases,thetrainingerrorwillnecessarilydecrease,butthetesterrormaynot.We will now explore this in a simulated data set.(a) Generate a data set withp= 20features,n=1,000observa-tions, and an associated quantitative response vector generatedaccording to the modelY=Xβ+ϵ,whereβhas some elements that are exactly equal to zero.(b) Splityourdatasetintoatrainingsetcontaining100observationsand a test set containing 900 observations.(c) Perform best subset selection on the training set, and plot thetraining set MSE associated with the best model of each size.(d) Plot the test set MSE associated with the best model of eachsize.(e) For which model size does the test set MSE take on its minimumvalue?Commentonyourresults.Ifittakesonitsminimumvaluefor a model containing only an intercept or a model containingall of the features, then play around with the way that you aregenerating the data in (a) until you come up with a scenario inwhich the test set MSE is minimized for an intermediate modelsize.(f) How does the model at which the test set MSE is minimizedcompare to the true model used to generate the data? Commenton the coefficient values.(g) Create a plot displaying√∑pj=1(βj−ˆβrj)2for a range of valuesofr, whereˆβrjis thejth coefficient estimate for the best modelcontainingrcoefficients. Comment on what you observe. Howdoes this compare to the test MSE plot from (d)?11. We will now try to predict per capita crime rate in theBostondataset.(a) Try out some of the regression methods explored in this chapter,such as best subset selection, the lasso, ridge regression, andPCR. Present and discuss results for the approaches that youconsider.(b) Propose a model (or set of models) that seem to perform well onthis data set, and justify your answer. Make sure that you areevaluating model performance using validation set error, cross-validation, or some other reasonable alternative, as opposed tousing training error.', metadata={'source': './documents/ISLP_website.pdf', 'page': 294}), Document(page_content='288 6. Linear Model Selection and Regularization(c) Does your chosen model involve all of the features in the dataset? Why or why not?', metadata={'source': './documents/ISLP_website.pdf', 'page': 295}), Document(page_content='7\\nMoving Beyond Linearity\\nSo far in this book, we have mostly focused on linear models. Linear models\\nare relatively simple to describe and implement, and have advantages over\\nother approaches in terms of interpretation and inference. However, stan-\\ndard linear regression can have significant limitations in terms of predic-\\ntive power. This is because the linearity assumption is almost always an\\napproximation, and sometimes a poor one. In Chapter 6we see that we can\\nimprove upon least squares using ridge regression, the lasso, principal com-\\nponents regression, and other techniques. In that setting, the improvement\\nis obtained by reducing the complexity of the linear model, and hence the\\nvariance of the estimates. But we are still using a linear model, which can\\nonly be improved so far! In this chapter we relax the linearity assumption\\nwhile still attempting to maintain as much interpretability as possible. We\\ndo this by examining very simple extensions of linear models like polyno-\\nmialregressionandstepfunctions,aswellasmoresophisticatedapproaches\\nsuch as splines, local regression, and generalized additive models.\\n•Polynomial regression extends the linear model by adding extra pre-\\ndictors, obtained by raising each of the original predictors to a power.\\nFor example, a cubicregression uses three variables, X,X2, andX3,\\nas predictors. This approach provides a simple way to provide a non-\\nlinear fit to data.\\n•Step functions cut the range of a variable into Kdistinct regions in\\norder to produce a qualitative variable. This has the effect of fitting\\na piecewise constant function.\\n•Regression splines are more flexible than polynomials and step func-\\ntions, and in fact are an extension of the two. They involve dividing\\nthe range of XintoKdistinct regions. Within each region, a poly-\\nnomial function is fit to the data. However, these polynomials are\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning , Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_7289', metadata={'source': './documents/ISLP_website.pdf', 'page': 296}), Document(page_content='290 7. Moving Beyond Linearity\\nconstrained so that they join smoothly at the region boundaries, or\\nknots. Provided that the interval is divided into enough regions, this\\ncan produce an extremely flexible fit.\\n•Smoothing splines are similar to regression splines, but arise in a\\nslightly different situation. Smoothing splines result from minimizing\\na residual sum of squares criterion subject to a smoothness penalty.\\n•Local regression is similar to splines, but differs in an important way.\\nThe regions are allowed to overlap, and indeed they do so in a very\\nsmooth way.\\n•Generalized additive models allow us to extend the methods above to\\ndeal with multiple predictors.\\nIn Sections 7.1–7.6, we present a number of approaches for modeling the\\nrelationship between a response Yand a single predictor Xin a flexible\\nway. In Section 7.7, we show that these approaches can be seamlessly in-\\ntegrated in order to model a response Yas a function of several predictors\\nX1,...,X p.\\n7.1 Polynomial Regression\\nHistorically, the standard way to extend linear regression to settings in\\nwhich the relationship between the predictors and the response is non-\\nlinear has been to replace the standard linear model\\nyi=β0+β1xi+ϵi\\nwith a polynomial function\\nyi=β0+β1xi+β2x2\\ni+β3x3\\ni+···+βdxd\\ni+ϵi, (7.1)\\nwhereϵiis the error term. This approach is known as polynomial regression ,polynomial\\nregressionand in fact we saw an example of this method in Section 3.3.2. For large\\nenough degree d, a polynomial regression allows us to produce an extremely\\nnon-linearcurve.Noticethatthecoefficientsin( 7.1)canbeeasilyestimated\\nusing least squares linear regression because this is just a standard linear\\nmodel with predictors xi,x2\\ni,x3\\ni,...,xd\\ni. Generally speaking, it is unusual\\nto usedgreater than 3or4because for large values of d, the polynomial\\ncurve can become overly flexible and can take on some very strange shapes.\\nThis is especially true near the boundary of the Xvariable.\\nThe left-hand panel in Figure 7.1is a plot of wageagainst agefor the\\nWagedata set, which contains income and demographic information for\\nmales who reside in the central Atlantic region of the United States. We\\nsee the results of fitting a degree-4 polynomial using least squares (solid\\nblue curve). Even though this is a linear regression model like any other,\\nthe individual coefficients are not of particular interest. Instead, we look at\\nthe entire fitted function across a grid of 63 values for agefrom 18 to 80 in\\norder to understand the relationship between ageandwage.', metadata={'source': './documents/ISLP_website.pdf', 'page': 297}), Document(page_content='7.1 Polynomial Regression 291\\n20 30 40 50 60 70 8050 100 150 200 250 300AgeWageDegree−4 Polynomial\\n20 30 40 50 60 70 800.00 0.05 0.10 0.15 0.20Age|||| | |||||||| |||||||| |||||\\n|||||||| ||||| | | ||||||\\n||||||\\n|| |||||||||| | | |||||| | | ||\\n||||||||||||||||||||| |||||||| |||| |||| | | | ||\\n|| | |||||| |||||||||||||| |||||||| ||||||||| |||||||| | | ||||||||||| |||| |||||||||||||||||||||||\\n|||| | |||||||||||||||||||||||||\\n|| ||||| | ||| | ||| |||||||||||| | | | | | |||| | ||||||||||||| ||| ||||||||||||| |||||||| | | | | ||| |||||||||||||||||||||||||||| ||\\n||||||||||| ||||||| ||||\\n||||\\n|||||||||||||||||||||| |||| ||||| | |||||||||| | ||||||||||||| |||||||| | | |||\\n||||||||| | | ||\\n|||||||||||| | | ||||||||||||||||| |||\\n|||||||||| ||||\\n|||| | | ||||||\\n||||| || | | | ||||||||||\\n|| ||||||||||||||||||| ||||| ||\\n|| |||| | | |||||||||\\n|||| ||||||||||||||||||| ||||||||||| ||||||| | |||||| ||||||||||||||\\n||||||||| | |||||\\n|| |||||| |||||| | | |||||||| | | | | |||||||||| |||||||\\n|| | | | ||||||||||||| |||||| | | ||\\n||||||||||| | ||||||||||||||||||\\n|| ||| | ||||||| |||||||||||||||||||\\n|| |||||||||||||||||||||||||||| | | ||||||\\n||||||||||||||||||||| | |||||||||| |||||||| ||| |||||||| |||| |||||||||| | | | | ||\\n||||| | |||| |||||\\n|| | |||||||| | |||||||| ||| ||||| ||||| | |||||\\n|||| ||| ||| | ||||| |||| |||||||||||||||| |||||||||||||||| ||||||||\\n|||||||||||| | ||||||||||||||||||||| ||| | | |||| |||||| |||||||||| ||||||||||\\n|| |||||||||||||||| |||| |||| ||\\n|||||||||||||||\\n|||||| ||||||| | | ||||||| || ||||| |||| | |||||| |||||||||||||||| |||||||||| |||| | ||||| ||||||| | ||||||||||||||||||\\n||| | |||||| |||| | ||||||||||||||| ||||||||||||\\n|||||||||||\\n||||| ||| | ||\\n||||||||||||||||\\n||||||||||||||||\\n|||||||| | | |||| | | |||||||||| |||||| | ||||||| ||||||\\n|| | | | | | | | | | ||||||| |||||||| ||||||| ||||| |||||||||||||||||||| ||||||||||||| |||| |||||||||||||||||||||| |||||||||||||||\\n|||| | | |||||||||||| | | | | | |||| ||||||||| | ||| ||||||| ||||||| ||| |||||||| | |||||||||||||||||| | | | |||| ||||||||||\\n|||||||||| |||||||||||||||||||||||||| | ||||||||||||||\\n|||||||||| | |||||||||||||| ||| |||| |||||||||||||||||||\\n||||||||||||||||||||||||||||||||||||| |||||||||||| ||||||| |||||||||||\\n|||| | |||| |||||||||||||| | | | ||||||||||||||||||||||| ||||||||| | ||||||||| |||| |||\\n|||||\\n|| ||| | ||||||||||||| | |||||||||||||||| | ||||||||||||\\n||| | | | | |||||| ||||||||||||||||||||| |||||||| ||||| | |||||||||||||||||||| ||||| ||||||||||||||||||| | |||||||||\\n||||||||||||| ||| ||| | | | ||||| ||| | ||||\\n|||||||\\n|| ||| ||| | ||||||||||||||||||||||||||||| ||| | | |||||||||||||||||||||| |||||||||| | || ||||||\\n|| ||||\\n||| | ||||||||||||| | ||||||||||||||\\n||||||||||||||||||||||| |||||||| |||||||||| | | | |||||||||||| ||||||||||||||||||| | ||||||||||||||||||| | ||||| | | ||||| | | ||||||\\n||||||||\\n|| |||||||| ||||\\n|| | ||||||||| ||||| ||||||| |||| |||| |||| | |||||||| ||||||||||||||||||||\\n|||| | |||||| |||||||\\n|| ||| |||||||\\n|| | | ||||||| ||||\\n|||||| ||||||| | |||||||||||| | |||||||| | ||||||||||||||||||| |||||||||||||||| | | ||\\n||||||\\n||||||||| | | |||| ||||||| |||| | | | | | | | |||||| | ||||||\\n||| | | ||||\\n|||||||| ||||||||||||| | | | | | |||||||||||||||||||||| | ||| | ||||| |||||| |||||| | ||| |||||| |||||| | | | ||||||| | |||||||| |||| |||| | | | | ||||||||||||||| ||||||||||||||||\\n|||||\\n||||\\n|||\\n||\\n|| | ||||| | ||\\n|||||||\\n||||||||||||||| | | | | ||| ||||| ||||||||| | ||||||||||||||| ||||| |||||||||||| || |||||||||| |||\\n|||||||| |||||||| | | ||||||||||||||||||||\\n|||||| ||||| | ||| |||||||| | | ||||||||||| | |||||||||\\n|||||||| |||||\\n|| ||||||| |||||||||||||| ||| | |||||\\n||| |||||||||| ||| ||||| |||||||||||||||||||||||||||| |||| |||||||||||||| | |||| ||Pr(Wage>250|Age)\\nFIGURE 7.1. TheWagedata.Left:The solid blue curve is a degree-4 polynomial\\nofwage(in thousands of dollars) as a function of age, fit by least squares. The\\ndashed curves indicate an estimated 95% confidence interval. Right:We model the\\nbinary event wage>250 using logistic regression, again with a degree-4 polynomial.\\nThe fitted posterior probability of wageexceeding $250,000is shown in blue, along\\nwith an estimated 95% confidence interval.\\nIn Figure 7.1, a pair of dashed curves accompanies the fit; these are ( 2×)\\nstandard error curves. Let’s see how these arise. Suppose we have computed\\nthe fit at a particular value of age,x0:\\nˆf(x0)=ˆβ0+ˆβ1x0+ˆβ2x2\\n0+ˆβ3x3\\n0+ˆβ4x4\\n0. (7.2)\\nWhatisthevarianceofthefit,i.e.Var ˆf(x0)?Leastsquaresreturnsvariance\\nestimates for each of the fitted coefficients ˆβj, as well as the covariances\\nbetween pairs of coefficient estimates. We can use these to compute the\\nestimated variance of ˆf(x0).1The estimated pointwise standard error of\\nˆf(x0)is the square-root of this variance. This computation is repeated\\nat each reference point x0, and we plot the fitted curve, as well as twice\\nthe standard error on either side of the fitted curve. We plot twice the\\nstandard error because, for normally distributed error terms, this quantity\\ncorresponds to an approximate 95% confidence interval.\\nIt seems like the wages in Figure 7.1are from two distinct populations:\\nthere appears to be a high earners group earning more than $250 ,000per\\nannum, as well as a low earners group. We can treat wageas a binary\\nvariable by splitting it into these two groups. Logistic regression can then\\nbe used to predict this binary response, using polynomial functions of age\\n1IfˆCis the5×5covariance matrix of the ˆβj, and if ℓT\\n0= (1 ,x0,x2\\n0,x3\\n0,x4\\n0), then\\nVar[ˆf(x0)] = ℓT\\n0ˆCℓ0.', metadata={'source': './documents/ISLP_website.pdf', 'page': 298}), Document(page_content='292 7. Moving Beyond Linearity\\nas predictors. In other words, we fit the model\\nPr(yi>250|xi)=exp(β0+β1xi+β2x2\\ni+···+βdxd\\ni)\\n1 + exp( β0+β1xi+β2x2\\ni+···+βdxd\\ni). (7.3)\\nThe result is shown in the right-hand panel of Figure 7.1. The gray marks\\non the top and bottom of the panel indicate the ages of the high earners\\nand the low earners. The solid blue curve indicates the fitted probabilities\\nof being a high earner, as a function of age. The estimated 95% confidence\\ninterval is shown as well. We see that here the confidence intervals are fairly\\nwide, especially on the right-hand side. Although the sample size for this\\ndata set is substantial ( n=3,000), there are only 79high earners, which\\nresults in a high variance in the estimated coefficients and consequently\\nwide confidence intervals.\\n7.2 Step Functions\\nUsing polynomial functions of the features as predictors in a linear model\\nimposes a globalstructure on the non-linear function of X. We can instead\\nusestep functions in order to avoid imposing such a global structure. Herestep\\nfunction we break the range of Xintobins, and fit a different constant in each bin.\\nThisamountstoconvertingacontinuousvariableintoan orderedcategorical\\nvariable.ordered\\ncategorical\\nvariableIn greater detail, we create cutpoints c1,c2,...,c Kin the range of X,\\nand then construct K+1new variables\\nC0(X)= I(X<c 1),\\nC1(X)= I(c1≤X<c 2),\\nC2(X)= I(c2≤X<c 3),\\n...\\nCK−1(X)= I(cK−1≤X<c K),\\nCK(X)= I(cK≤X),(7.4)\\nwhereI(·)is anindicator function that returns a 1if the condition is true,indicator\\nfunctionand returns a 0otherwise. For example, I(cK≤X)equals1ifcK≤X, and\\nequals0otherwise. These are sometimes called dummyvariables. Notice\\nthat for any value of X,C0(X)+C1(X)+···+CK(X)=1 , sinceXmust\\nbe in exactly one of the K+1intervals. We then use least squares to fit a\\nlinear model using C1(X),C2(X),...,C K(X)as predictors2:\\nyi=β0+β1C1(xi)+β2C2(xi)+···+βKCK(xi)+ϵi. (7.5)\\nFor a given value of X, at most one of C1,C2,...,C Kcan be non-zero.\\nNote that when X<c 1, all of the predictors in ( 7.5) are zero, so β0can\\n2We exclude C0(X)as a predictor in ( 7.5) because it is redundant with the intercept.\\nThis is similar to the fact that we need only two dummy variables to code a qualitative\\nvariable with three levels, provided that the model will contain an intercept. The decision\\nto exclude C0(X)instead of some other Ck(X)in (7.5) is arbitrary. Alternatively, we\\ncould include C0(X),C1(X),...,C K(X), and exclude the intercept.', metadata={'source': './documents/ISLP_website.pdf', 'page': 299}), Document(page_content='7.3 Basis Functions 293\\n20 30 40 50 60 70 8050 100 150 200 250 300AgeWagePiecewise Constant\\n20 30 40 50 60 70 800.00 0.05 0.10 0.15 0.20Age||||||||||||||||||||||||||\\n|||||||||||||||||||||\\n||||||\\n||||||||||||||||||||||||\\n||||||||||||||||||||||||||||||||||||||||||\\n||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\\n||||||||||||||||||||||||||||||\\n|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\\n||||||||||||||||||||||\\n||||\\n|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\\n|||||||||||||\\n||||||||||||||||||||||||||||||||||\\n||||||||||||||\\n||||||||||||\\n||||||||||||||||||||\\n||||||||||||||||||||||||||||\\n|||||||||||||||||\\n||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\\n|||||||||||||||\\n|||||||||||||||||||||||||||||||||||||||||||||\\n||||||||||||||||||||||||||||\\n||||||||||||||||||||||||||||||\\n||||||||||||||||||||||||||||||||\\n||||||||||||||||||||||||||||||||||||||\\n|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\\n|||||||||||||||\\n|||||||||||||||||||||||||||||||||||||||\\n||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\\n|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\\n||||||||||||||||||||||||||||\\n|||||||||||||||\\n||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\\n||||||||||||||||||||||||||||||||||||||||||\\n|||||||||||\\n|||||||||||\\n||||||||||||||||\\n||||||||||||||||\\n||||||||||||||||||||||||||||||||||||||||||||||\\n||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\\n|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\\n|||||||||||||||||||||||||||||||||||||||||||||||||||\\n|||||||||||||||||||||||||||||||||||||||||||||||||||\\n|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\\n|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\\n|||||\\n|||||||||||||||||||||||||||||||||||||||||||||||||\\n||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\\n|||||||||||||||||||||||||||||||||||\\n|||||||\\n||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\\n||||||\\n||||||||||||||||||||||||||||||||\\n||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\\n||||||||\\n||||||||||||||\\n|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\\n||||||||||||||||||\\n||||||||||||\\n|||||||||||||||\\n|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\\n||||||\\n||||||||||||||||||||||||||||||||||||||||||||||\\n|||||||||\\n||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\\n|||||\\n||||\\n|||\\n||\\n|||||||||||\\n|||||||\\n||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\\n||||||||||||||||||||||||||||||||||||||\\n||||||||||||||||||||||||||||||||||||||||||||||\\n|||||||||||||\\n||||||||||||||||||||||||||||||||\\n||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||Pr(Wage>250|Age)\\nFIGURE 7.2. TheWagedata.Left:The solid curve displays the fitted value from\\na least squares regression of wage(in thousands of dollars) using step functions\\nofage. The dashed curves indicate an estimated 95% confidence interval. Right:\\nWe model the binary event wage>250 using logistic regression, again using step\\nfunctions of age. The fitted posterior probability of wageexceeding $250,000is\\nshown, along with an estimated 95% confidence interval.\\nbe interpreted as the mean value of YforX<c 1. By comparison, ( 7.5)\\npredictsaresponseof β0+βjforcj≤X<c j+1,soβjrepresentstheaverage\\nincrease in the response for Xincj≤X<c j+1relative to X<c 1.\\nAn example of fitting step functions to the Wagedata from Figure 7.1is\\nshown in the left-hand panel of Figure 7.2. We also fit the logistic regression\\nmodel\\nPr(yi>250|xi)=exp(β0+β1C1(xi)+···+βKCK(xi))\\n1 + exp( β0+β1C1(xi)+···+βKCK(xi))(7.6)\\nin order to predict the probability that an individual is a high earner on the\\nbasis of age. The right-hand panel of Figure 7.2displays the fitted posterior\\nprobabilities obtained using this approach.\\nUnfortunately, unless there are natural breakpoints in the predictors,\\npiecewise-constant functions can miss the action. For example, in the left-\\nhand panel of Figure 7.2, the first bin clearly misses the increasing trend\\nofwagewithage. Nevertheless, step function approaches are very popular\\nin biostatistics and epidemiology, among other disciplines. For example,\\n5-year age groups are often used to define the bins.\\n7.3 Basis Functions\\nPolynomial and piecewise-constant regression models are in fact special\\ncases of a basis function approach. The idea is to have at hand a fam-basis\\nfunction', metadata={'source': './documents/ISLP_website.pdf', 'page': 300}), Document(page_content='294 7. Moving Beyond Linearity\\nily of functions or transformations that can be applied to a variable X:\\nb1(X),b2(X),...,b K(X). Instead of fitting a linear model in X, we fit the\\nmodel\\nyi=β0+β1b1(xi)+β2b2(xi)+β3b3(xi)+···+βKbK(xi)+ϵi.(7.7)\\nNote that the basis functions b1(·),b2(·),...,b K(·)are fixed and known.\\n(In other words, we choose the functions ahead of time.) For polynomial\\nregression, the basis functions are bj(xi)=xj\\ni, and for piecewise constant\\nfunctions they are bj(xi)= I(cj≤xi<c j+1). We can think of ( 7.7) as\\na standard linear model with predictors b1(xi),b2(xi),...,b K(xi). Hence,\\nwe can use least squares to estimate the unknown regression coefficients\\nin (7.7). Importantly, this means that all of the inference tools for linear\\nmodels that are discussed in Chapter 3, such as standard errors for the\\ncoefficient estimates and F-statistics for the model’s overall significance,\\nare available in this setting.\\nThus far we have considered the use of polynomial functions and piece-\\nwise constant functions for our basis functions; however, many alternatives\\nare possible. For instance, we can use wavelets or Fourier series to construct\\nbasis functions. In the next section, we investigate a very common choice\\nfor a basis function: regression splines .regression\\nspline\\n7.4 Regression Splines\\nNow we discuss a flexible class of basis functions that extends upon the\\npolynomial regression and piecewise constant regression approaches that\\nwe have just seen.\\n7.4.1 Piecewise Polynomials\\nInsteadoffittingahigh-degreepolynomialovertheentirerangeof X,piece-\\nwise polynomial regression involves fitting separate low-degree polynomialspiecewise\\npolynomial\\nregressionover different regions of X. For example, a piecewise cubic polynomial\\nworks by fitting a cubic regression model of the form\\nyi=β0+β1xi+β2x2\\ni+β3x3\\ni+ϵi, (7.8)\\nwhere the coefficients β0,β1,β2, andβ3differ in different parts of the range\\nofX. The points where the coefficients change are called knots.knotFor example, a piecewise cubic with no knots is just a standard cubic\\npolynomial, as in ( 7.1) withd=3. A piecewise cubic polynomial with a\\nsingle knot at a point ctakes the form\\nyi={\\nβ01+β11xi+β21x2\\ni+β31x3\\ni+ϵiifxi<c\\nβ02+β12xi+β22x2\\ni+β32x3\\ni+ϵiifxi≥c.\\nIn other words, we fit two different polynomial functions to the data, one\\non the subset of the observations with xi<c, and one on the subset of\\nthe observations with xi≥c. The first polynomial function has coefficients', metadata={'source': './documents/ISLP_website.pdf', 'page': 301}), Document(page_content='7.4 Regression Splines 295\\n20 30 40 50 60 7050 100 150 200 250AgeWagePiecewise Cubic\\n20 30 40 50 60 7050 100 150 200 250AgeWageContinuous Piecewise Cubic\\n20 30 40 50 60 7050 100 150 200 250AgeWageCubic Spline\\n20 30 40 50 60 7050 100 150 200 250AgeWageLinear Spline\\nFIGURE 7.3. Various piecewise polynomials are fit to a subset of the Wage\\ndata, with a knot at age=50.Top Left: The cubic polynomials are unconstrained.\\nTop Right: The cubic polynomials are constrained to be continuous at age=50.\\nBottom Left: The cubic polynomials are constrained to be continuous, and to have\\ncontinuous first and second derivatives. Bottom Right: A linear spline is shown,\\nwhich is constrained to be continuous.\\nβ01,β11,β21,andβ31, and the second has coefficients β02,β12,β22,andβ32.\\nEach of these polynomial functions can be fit using least squares applied\\nto simple functions of the original predictor.\\nUsing more knots leads to a more flexible piecewise polynomial. In gen-\\neral, if we place Kdifferent knots throughout the range of X, then we\\nwill end up fitting K+1different cubic polynomials. Note that we do not\\nneed to use a cubic polynomial. For example, we can instead fit piecewise\\nlinear functions. In fact, our piecewise constant functions of Section 7.2are\\npiecewise polynomials of degree 0!\\nThe top left panel of Figure 7.3shows a piecewise cubic polynomial fit to\\na subset of the Wagedata, with a single knot at age=50. We immediately see\\na problem: the function is discontinuous and looks ridiculous! Since each\\npolynomial has four parameters, we are using a total of eight degrees of\\nfreedomin fitting this piecewise polynomial model.degrees of\\nfreedom', metadata={'source': './documents/ISLP_website.pdf', 'page': 302}), Document(page_content='296 7. Moving Beyond Linearity\\n7.4.2 Constraints and Splines\\nThe top left panel of Figure 7.3looks wrong because the fitted curve is just\\ntoo flexible. To remedy this problem, we can fit a piecewise polynomial\\nunder the constraint that the fitted curve must be continuous. In other\\nwords,therecannotbeajumpwhen age=50.ThetoprightplotinFigure 7.3\\nshows the resulting fit. This looks better than the top left plot, but the V-\\nshaped join looks unnatural.\\nInthelowerleftplot,wehaveaddedtwoadditionalconstraints:nowboth\\nthe first and second derivatives of the piecewise polynomials are continuousderivativeatage=50. In other words, we are requiring that the piecewise polynomial\\nbe not only continuous when age=50, but also very smooth. Each constraint\\nthat we impose on the piecewise cubic polynomials effectively frees up one\\ndegree of freedom, by reducing the complexity of the resulting piecewise\\npolynomial fit. So in the top left plot, we are using eight degrees of free-\\ndom, but in the bottom left plot we imposed three constraints (continuity,\\ncontinuity of the first derivative, and continuity of the second derivative)\\nand so are left with five degrees of freedom. The curve in the bottom left\\nplot is called a cubic spline .3In general, a cubic spline with Kknots usescubic splinea total of 4+Kdegrees of freedom.\\nIn Figure 7.3, the lower right plot is a linear spline , which is continuouslinear splineatage=50. The general definition of a degree- dspline is that it is a piecewise\\ndegree-dpolynomial, with continuity in derivatives up to degree d−1at\\neach knot. Therefore, a linear spline is obtained by fitting a line in each\\nregion of the predictor space defined by the knots, requiring continuity at\\neach knot.\\nIn Figure 7.3, there is a single knot at age=50. Of course, we could add\\nmore knots, and impose continuity at each.\\n7.4.3 The Spline Basis Representation\\nThe regression splines that we just saw in the previous section may have\\nseemed somewhat complex: how can we fit a piecewise degree- dpolynomial\\nunder the constraint that it (and possibly its first d−1derivatives) be\\ncontinuous? It turns out that we can use the basis model ( 7.7) to represent\\na regression spline. A cubic spline with Kknots can be modeled as\\nyi=β0+β1b1(xi)+β2b2(xi)+···+βK+3bK+3(xi)+ϵi, (7.9)\\nfor an appropriate choice of basis functions b1,b2,...,b K+3. The model\\n(7.9) can then be fit using least squares.\\nJust as there were several ways to represent polynomials, there are also\\nmany equivalent ways to represent cubic splines using different choices of\\nbasis functions in ( 7.9). The most direct way to represent a cubic spline\\nusing (7.9) is to start off with a basis for a cubic polynomial—namely,\\nx, x2,andx3—and then add one truncated power basis function per knot.truncated\\npower basis\\n3Cubic splines are popular because most human eyes cannot detect the discontinuity\\nat the knots.', metadata={'source': './documents/ISLP_website.pdf', 'page': 303}), Document(page_content='7.4 Regression Splines 297\\n20 30 40 50 60 7050 100 150 200 250AgeWageNatural Cubic SplineCubic Spline\\nFIGURE 7.4. A cubic spline and a natural cubic spline, with three knots, fit to\\na subset of the Wagedata. The dashed lines denote the knot locations.\\nA truncated power basis function is defined as\\nh(x,ξ)=( x−ξ)3\\n+={\\n(x−ξ)3ifx>ξ\\n0otherwise ,(7.10)\\nwhereξis the knot. One can show that adding a term of the form β4h(x,ξ)\\nto the model ( 7.8) for a cubic polynomial will lead to a discontinuity in\\nonly the third derivative at ξ; the function will remain continuous, with\\ncontinuous first and second derivatives, at each of the knots.\\nInotherwords,inordertofitacubicsplinetoadatasetwith Kknots,we\\nperform least squares regression with an intercept and 3+Kpredictors, of\\nthe form X,X2,X3,h(X,ξ1),h(X,ξ2),...,h (X,ξK), whereξ1,...,ξKare\\nthe knots. This amounts to estimating a total of K+4regression coeffi-\\ncients; for this reason, fitting a cubic spline with Kknots uses K+4degrees\\nof freedom.\\nUnfortunately, splines can have high variance at the outer range of the\\npredictors—that is, when Xtakes on either a very small or very large\\nvalue. Figure 7.4shows a fit to the Wagedata with three knots. We see that\\nthe confidence bands in the boundary region appear fairly wild. A natu-\\nral spline is a regression spline with additional boundary constraints : thenatural\\nsplinefunction is required to be linear at the boundary (in the region where Xis\\nsmaller than the smallest knot, or larger than the largest knot). This addi-\\ntional constraint means that natural splines generally produce more stable\\nestimates at the boundaries. In Figure 7.4, a natural cubic spline is also\\ndisplayed as a red line. Note that the corresponding confidence intervals\\nare narrower.\\n7.4.4 Choosing the Number and Locations of the Knots\\nWhen we fit a spline, where should we place the knots? The regression\\nspline is most flexible in regions that contain a lot of knots, because in\\nthose regions the polynomial coefficients can change rapidly. Hence, one', metadata={'source': './documents/ISLP_website.pdf', 'page': 304}), Document(page_content='298 7. Moving Beyond Linearity\\n20 30 40 50 60 70 8050 100 150 200 250 300AgeWageNatural Cubic Spline\\n20 30 40 50 60 70 800.00 0.05 0.10 0.15 0.20Age|||| | |||||||| |||||||| |||||\\n|||||||| ||||| | | ||||||\\n||||||\\n|| |||||||||| | | |||||| | | ||\\n||||||||||||||||||||| |||||||| |||| |||| | | | ||\\n|| | |||||| |||||||||||||| |||||||| ||||||||| |||||||| | | ||||||||||| |||| |||||||||||||||||||||||\\n|||| | |||||||||||||||||||||||||\\n|| ||||| | ||| | ||| |||||||||||| | | | | | |||| | ||||||||||||| ||| ||||||||||||| |||||||| | | | | ||| |||||||||||||||||||||||||||| ||\\n||||||||||| ||||||| ||||\\n||||\\n|||||||||||||||||||||| |||| ||||| | |||||||||| | ||||||||||||| |||||||| | | |||\\n||||||||| | | ||\\n|||||||||||| | | ||||||||||||||||| |||\\n|||||||||| ||||\\n|||| | | ||||||\\n||||| || | | | ||||||||||\\n|| ||||||||||||||||||| ||||| ||\\n|| |||| | | |||||||||\\n|||| ||||||||||||||||||| ||||||||||| ||||||| | |||||| ||||||||||||||\\n||||||||| | |||||\\n|| |||||| |||||| | | |||||||| | | | | |||||||||| |||||||\\n|| | | | ||||||||||||| |||||| | | ||\\n||||||||||| | ||||||||||||||||||\\n|| ||| | ||||||| |||||||||||||||||||\\n|| |||||||||||||||||||||||||||| | | ||||||\\n||||||||||||||||||||| | |||||||||| |||||||| ||| |||||||| |||| |||||||||| | | | | ||\\n||||| | |||| |||||\\n|| | |||||||| | |||||||| ||| ||||| ||||| | |||||\\n|||| ||| ||| | ||||| |||| |||||||||||||||| |||||||||||||||| ||||||||\\n|||||||||||| | ||||||||||||||||||||| ||| | | |||| |||||| |||||||||| ||||||||||\\n|| |||||||||||||||| |||| |||| ||\\n|||||||||||||||\\n|||||| ||||||| | | ||||||| || ||||| |||| | |||||| |||||||||||||||| |||||||||| |||| | ||||| ||||||| | ||||||||||||||||||\\n||| | |||||| |||| | ||||||||||||||| ||||||||||||\\n|||||||||||\\n||||| ||| | ||\\n||||||||||||||||\\n||||||||||||||||\\n|||||||| | | |||| | | |||||||||| |||||| | ||||||| ||||||\\n|| | | | | | | | | | ||||||| |||||||| ||||||| ||||| |||||||||||||||||||| ||||||||||||| |||| |||||||||||||||||||||| |||||||||||||||\\n|||| | | |||||||||||| | | | | | |||| ||||||||| | ||| ||||||| ||||||| ||| |||||||| | |||||||||||||||||| | | | |||| ||||||||||\\n|||||||||| |||||||||||||||||||||||||| | ||||||||||| |||\\n|||||||||| | |||||||||||||| ||| |||| |||||||||||||||||||\\n||||||||||||||||||||||||||||||||||||| |||||||||||| ||||||| |||||||||||\\n|||| | |||| |||||||||||||| | | | ||||||||||||||||||||||| ||||||||| | ||||||||| |||| |||\\n|||||\\n|| ||| | ||||||||||||| | ||||||||| ||||||| | ||||||||||||\\n||| | | | | |||||| ||||||||||||||||||||| |||||||| ||||| | |||||||||||||||||||| | |||| ||||||||||||||||||| | |||||||||\\n||||||||||||| ||| ||| | | | ||||| ||| | ||||\\n|||||||\\n|| ||| ||| | |||||||||||||||||||||| ||||||| ||| | | |||||||||||||||||||||| |||||||||| | || ||||||\\n|| ||||\\n||| | ||||||||||||| | ||||||||||||||\\n||||||||||||||||||||||| |||||||| |||||||||| | | | |||||||||||| ||||||||||||||||||| | ||||||||||||||||||| | ||||| | | ||||| | | ||||||\\n||||||||\\n|| |||||||| ||||\\n|| | ||||||||| ||||| ||||||| |||| |||| |||| | |||||||| ||||||||||||||||||||\\n|||| | |||||| |||||||\\n|| ||| |||||||\\n|| | | ||||||| ||||\\n|||||| ||||||| | |||||||||||| | |||||||| | ||||||||||||||||||| |||||||||||||||| | | ||\\n||||||\\n||||||||| | | |||| ||||||| |||| | | | | | | | |||||| | ||||||\\n||| | | ||||\\n|||||||| ||||||||||||| | | | | | |||||||||||||||||||||| | ||| | ||||| |||||| |||||| | ||| |||||| |||||| | | | ||||||| | |||||||| |||| |||| | | | | ||||||||||||||| ||||||||||||||||\\n|||||\\n||||\\n|||\\n||\\n|| | ||||| | ||\\n|||||||\\n||||||||||||||| | | | | ||| ||||| ||||||||| | ||||||||||||||| ||||| |||||||||||| || |||||||||| |||\\n|||||||| |||||||| | | ||||||||||||||||||||\\n|||||| ||||| | ||| |||||||| | | ||||||||||| | |||||||||\\n|||||||| |||||\\n|| ||||||| |||||||||||||| ||| | |||||\\n||| |||||||||| ||| ||||| |||||||||||||||||||||||||||| |||| |||||||||||||| | |||| ||Pr(Wage>250|Age)\\nFIGURE 7.5. A natural cubic spline function with four degrees of freedom is\\nfit to the Wagedata.Left:A spline is fit to wage(in thousands of dollars) as\\na function of age.Right:Logistic regression is used to model the binary event\\nwage>250 as a function of age. The fitted posterior probability of wageexceeding\\n$250,000is shown. The dashed lines denote the knot locations.\\noption is to place more knots in places where we feel the function might\\nvary most rapidly, and to place fewer knots where it seems more stable.\\nWhile this option can work well, in practice it is common to place knots in\\na uniform fashion. One way to do this is to specify the desired degrees of\\nfreedom, and then have the software automatically place the corresponding\\nnumber of knots at uniform quantiles of the data.\\nFigure7.5shows an example on the Wagedata. As in Figure 7.4,w e\\nhave fit a natural cubic spline with three knots, except this time the knot\\nlocations were chosen automatically as the 25th, 50th, and 75th percentiles\\nofage. This was specified by requesting four degrees of freedom. The ar-\\ngument by which four degrees of freedom leads to three interior knots is\\nsomewhat technical.4\\nHow many knots should we use, or equivalently how many degrees of\\nfreedom should our spline contain? One option is to try out different num-\\nbers of knots and see which produces the best looking curve. A somewhat\\nmore objective approach is to use cross-validation, as discussed in Chap-\\nters5and6. With this method, we remove a portion of the data (say 10%),\\nfit a spline with a certain number of knots to the remaining data, and then\\nuse the spline to make predictions for the held-out portion. We repeat this\\nprocess multiple times until each observation has been left out once, and\\n4There are actually five knots, including the two boundary knots. A cubic spline with\\nfive knots has nine degrees of freedom. But natural cubic splines have two additional\\nnaturalconstraints at each boundary to enforce linearity, resulting in 9−4=5 degrees\\nof freedom. Since this includes a constant, which is absorbed in the intercept, we count\\nit as four degrees of freedom.', metadata={'source': './documents/ISLP_website.pdf', 'page': 305}), Document(page_content='7.4 Regression Splines 299\\n2 4 6 8 101600 1620 1640 1660 1680Degrees of Freedom of Natural SplineMean Squared Error2 4 6 8 101600 1620 1640 1660 1680Degrees of Freedom of Cubic SplineMean Squared ErrorFIGURE 7.6. Ten-fold cross-validated mean squared errors for selecting the\\ndegrees of freedom when fitting splines to the Wagedata. The response is wage\\nand the predictor age.Left:A natural cubic spline. Right:A cubic spline.\\nthen compute the overall cross-validated RSS. This procedure can be re-\\npeated for different numbers of knots K. Then the value of Kgiving the\\nsmallest RSS is chosen.\\nFigure7.6shows ten-fold cross-validated mean squared errors for splines\\nwith various degrees of freedom fit to the Wagedata. The left-hand panel\\ncorresponds to a natural cubic spline and the right-hand panel to a cu-\\nbic spline. The two methods produce almost identical results, with clear\\nevidence that a one-degree fit (a linear regression) is not adequate. Both\\ncurves flatten out quickly, and it seems that three degrees of freedom for\\nthe natural spline and four degrees of freedom for the cubic spline are quite\\nadequate.\\nIn Section 7.7we fit additive spline models simultaneously on several\\nvariables at a time. This could potentially require the selection of degrees\\nof freedom for each variable. In cases like this we typically adopt a more\\npragmatic approach and set the degrees of freedom to a fixed number, say\\nfour, for all terms.\\n7.4.5 Comparison to Polynomial Regression\\nFigure7.7compares a natural cubic spline with 15degrees of freedom to a\\ndegree-15polynomial on the Wagedata set. The extra flexibility in the poly-\\nnomial produces undesirable results at the boundaries, while the natural\\ncubic spline still provides a reasonable fit to the data. Regression splines\\noften give superior results to polynomial regression. This is because unlike\\npolynomials, which must use a high degree (exponent in the highest mono-\\nmial term, e.g. X15) to produce flexible fits, splines introduce flexibility\\nby increasing the number of knots but keeping the degree fixed. Generally,\\nthis approach produces more stable estimates. Splines also allow us to place\\nmore knots, and hence flexibility, over regions where the function fseems\\nto be changing rapidly, and fewer knots where fappears more stable.', metadata={'source': './documents/ISLP_website.pdf', 'page': 306}), Document(page_content='300 7. Moving Beyond Linearity\\n20 30 40 50 60 70 8050 100 150 200 250 300AgeWageNatural Cubic SplinePolynomial\\nFIGURE 7.7. On theWagedata set, a natural cubic spline with 15 degrees\\nof freedom is compared to a degree- 15polynomial. Polynomials can show wild\\nbehavior, especially near the tails.\\n7.5 Smoothing Splines\\nIn the last section we discussed regression splines, which we create by spec-\\nifying a set of knots, producing a sequence of basis functions, and then\\nusing least squares to estimate the spline coefficients. We now introduce a\\nsomewhat different approach that also produces a spline.\\n7.5.1 An Overview of Smoothing Splines\\nIn fitting a smooth curve to a set of data, what we really want to do is\\nfind some function, say g(x), that fits the observed data well: that is, we\\nwantRSS =∑n\\ni=1(yi−g(xi))2to be small. However, there is a problem\\nwith this approach. If we don’t put any constraints on g(xi), then we can\\nalways make RSS zero simply by choosing gsuch that it interpolates all\\nof theyi. Such a function would woefully overfit the data—it would be far\\ntoo flexible. What we really want is a function gthat makes RSS small,\\nbut that is also smooth.\\nHow might we ensure that gis smooth? There are a number of ways to\\ndo this. A natural approach is to find the function gthat minimizes\\nn∑\\ni=1(yi−g(xi))2+λ∫\\ng′′(t)2dt (7.11)\\nwhereλis a nonnegative tuning parameter . The function gthat minimizes\\n(7.11) is known as a smoothing spline .smoothing\\nsplineWhat does ( 7.11) mean? Equation 7.11takes the “Loss+Penalty” for-\\nmulation that we encounter in the context of ridge regression and the lasso\\nin Chapter 6. The term∑n\\ni=1(yi−g(xi))2is aloss function that encour-loss functionagesgto fit the data well, and the term λ∫\\ng′′(t)2dtis apenalty term\\nthat penalizes the variability in g. The notation g′′(t)indicates the second\\nderivative of the function g. The first derivative g′(t)measures the slope', metadata={'source': './documents/ISLP_website.pdf', 'page': 307}), Document(page_content='7.5 Smoothing Splines 301\\nof a function at t, and the second derivative corresponds to the amount by\\nwhich the slope is changing. Hence, broadly speaking, the second derivative\\nof a function is a measure of its roughness : it is large in absolute value if\\ng(t)is very wiggly near t, and it is close to zero otherwise. (The second\\nderivative of a straight line is zero; note that a line is perfectly smooth.)\\nThe∫\\nnotation is an integral, which we can think of as a summation over\\nthe range of t. In other words,∫\\ng′′(t)2dtis simply a measure of the total\\nchange in the function g′(t), over its entire range. If gis very smooth, then\\ng′(t)will be close to constant and∫\\ng′′(t)2dtwill take on a small value.\\nConversely, if gis jumpy and variable then g′(t)will vary significantly and∫\\ng′′(t)2dtwill take on a large value. Therefore, in ( 7.11),λ∫\\ng′′(t)2dten-\\ncourages gto be smooth. The larger the value of λ, the smoother gwill be.\\nWhenλ=0, then the penalty term in ( 7.11) has no effect, and so the\\nfunction gwill be very jumpy and will exactly interpolate the training\\nobservations. When λ→∞,gwill be perfectly smooth—it will just be\\na straight line that passes as closely as possible to the training points.\\nIn fact, in this case, gwill be the linear least squares line, since the loss\\nfunction in ( 7.11) amounts to minimizing the residual sum of squares. For\\nan intermediate value of λ,gwill approximate the training observations\\nbut will be somewhat smooth. We see that λcontrols the bias-variance\\ntrade-off of the smoothing spline.\\nThe function g(x)that minimizes ( 7.11) can be shown to have some spe-\\ncial properties: it is a piecewise cubic polynomial with knots at the unique\\nvalues of x1,...,x n, and continuous first and second derivatives at each\\nknot. Furthermore, it is linear in the region outside of the extreme knots.\\nIn other words, the function g(x)that minimizes ( 7.11) is a natural cubic\\nspline with knots at x1,...,x n!However, it is not the same natural cubic\\nspline that one would get if one applied the basis function approach de-\\nscribed in Section 7.4.3with knots at x1,...,x n—rather, it is a shrunken\\nversion of such a natural cubic spline, where the value of the tuning pa-\\nrameterλin (7.11) controls the level of shrinkage.\\n7.5.2 Choosing the Smoothing Parameter λ\\nWe have seen that a smoothing spline is simply a natural cubic spline\\nwith knots at every unique value of xi. It might seem that a smoothing\\nspline will have far too many degrees of freedom, since a knot at each data\\npoint allows a great deal of flexibility. But the tuning parameter λcontrols\\nthe roughness of the smoothing spline, and hence the effective degrees of\\nfreedom. It is possible to show that as λincreases from 0to∞, the effectiveeffective\\ndegrees of\\nfreedomdegrees of freedom, which we write dfλ, decrease from nto2.\\nIn the context of smoothing splines, why do we discuss effective degrees\\nof freedom instead of degrees of freedom? Usually degrees of freedom refer\\nto the number of free parameters, such as the number of coefficients fit in a\\npolynomial or cubic spline. Although a smoothing spline has nparameters\\nand hence nnominal degrees of freedom, these nparameters are heavily\\nconstrained or shrunk down. Hence dfλis a measure of the flexibility of the\\nsmoothing spline—the higher it is, the more flexible (and the lower-bias but\\nhigher-variance) the smoothing spline. The definition of effective degrees of', metadata={'source': './documents/ISLP_website.pdf', 'page': 308}), Document(page_content='302 7. Moving Beyond Linearity\\nfreedom is somewhat technical. We can write\\nˆgλ=Sλy, (7.12)\\nwhereˆgλis the solution to ( 7.11) for a particular choice of λ—that is, it\\nis ann-vector containing the fitted values of the smoothing spline at the\\ntraining points x1,...,x n. Equation 7.12indicates that the vector of fitted\\nvalues when applying a smoothing spline to the data can be written as a\\nn×nmatrixSλ(for which there is a formula) times the response vector\\ny. Then the effective degrees of freedom is defined to be\\ndfλ=n∑\\ni=1{Sλ}ii, (7.13)\\nthe sum of the diagonal elements of the matrix Sλ.\\nIn fitting a smoothing spline, we do not need to select the number or\\nlocation of the knots—there will be a knot at each training observation,\\nx1,...,x n. Instead, we have another problem: we need to choose the value\\nofλ.Itshouldcomeasnosurprisethatonepossiblesolutiontothisproblem\\nis cross-validation. In other words, we can find the value of λthat makes\\nthe cross-validated RSS as small as possible. It turns out that the leave-\\none-outcross-validation error (LOOCV) can be computed very efficiently\\nfor smoothing splines, with essentially the same cost as computing a single\\nfit, using the following formula:\\nRSScv(λ)=n∑\\ni=1(yi−ˆg(−i)\\nλ(xi))2=n∑\\ni=1[yi−ˆgλ(xi)\\n1−{Sλ}ii]2\\n.\\nThe notation ˆg(−i)\\nλ(xi)indicates the fitted value for this smoothing spline\\nevaluated at xi, where the fit uses all of the training observations except\\nfor theith observation (xi,yi). In contrast, ˆgλ(xi)indicates the smoothing\\nspline function fit to all of the training observations and evaluated at xi.\\nThis remarkable formula says that we can compute each of these leave-\\none-outfits using only ˆgλ, the original fit to allof the data!5We have\\na very similar formula ( 5.2) on page 205in Chapter 5for least squares\\nlinear regression. Using ( 5.2), we can very quickly perform LOOCV for\\nthe regression splines discussed earlier in this chapter, as well as for least\\nsquares regression using arbitrary basis functions.\\nFigure7.8shows the results from fitting a smoothing spline to the Wage\\ndata. The red curve indicates the fit obtained from pre-specifying that we\\nwould like a smoothing spline with 16effective degrees of freedom. The blue\\ncurve is the smoothing spline obtained when λis chosen using LOOCV; in\\nthis case, the value of λchosen results in 6.8effective degrees of freedom\\n(computed using ( 7.13)). For this data, there is little discernible difference\\nbetween the two smoothing splines, beyond the fact that the one with 16\\ndegrees of freedom seems slightly wigglier. Since there is little difference\\nbetween the two fits, the smoothing spline fit with 6.8degrees of freedom\\n5The exact formulas for computing ˆg(xi)andSλare very technical; however, efficient\\nalgorithms are available for computing these quantities.', metadata={'source': './documents/ISLP_website.pdf', 'page': 309}), Document(page_content='7.6 Local Regression 303\\n20 30 40 50 60 70 800 50 100 200 300AgeWageSmoothing Spline16 Degrees of Freedom6.8 Degrees of Freedom (LOOCV)\\nFIGURE 7.8. Smoothing spline fits to the Wagedata. The red curve results\\nfrom specifying 16effective degrees of freedom. For the blue curve, λwas found\\nautomatically by leave-one-out cross-validation, which resulted in 6.8effective\\ndegrees of freedom.\\nis preferable, since in general simpler models are better unless the data\\nprovides evidence in support of a more complex model.\\n7.6 Local Regression\\nLocal regression is a different approach for fitting flexible non-linear func-local\\nregressiontions, which involves computing the fit at a target point x0using only the\\nnearby training observations. Figure 7.9illustrates the idea on some simu-\\nlated data, with one target point near 0.4, and another near the boundary\\nat0.05. In this figure the blue line represents the function f(x)from which\\nthe data were generated, and the light orange line corresponds to the local\\nregression estimate ˆf(x). Local regression is described in Algorithm 7.1.\\nNote that in Step 3 of Algorithm 7.1, the weights Ki0will differ for each\\nvalue ofx0. In other words, in order to obtain the local regression fit at a\\nnew point, we need to fit a new weighted least squares regression model by\\nminimizing ( 7.14) for a new set of weights. Local regression is sometimes\\nreferredtoasa memory-based procedure,becauselikenearest-neighbors,we\\nneed all the training data each time we wish to compute a prediction. We\\nwill avoid getting into the technical details of local regression here—there\\nare books written on the topic.\\nIn order to perform local regression, there are a number of choices to\\nbe made, such as how to define the weighting function K, and whether\\nto fit a linear, constant, or quadratic regression in Step 3. (Equation 7.14\\ncorresponds to a linear regression.) While all of these choices make some\\ndifference, the most important choice is the spans, which is the proportion\\nof points used to compute the local regression at x0, as defined in Step 1\\nabove. The span plays a role like that of the tuning parameter λin smooth-', metadata={'source': './documents/ISLP_website.pdf', 'page': 310}), Document(page_content='304 7. Moving Beyond Linearity\\n0.0 0.2 0.4 0.6 0.8 1.0−1.0 −0.5 0.0 0.5 1.0 1.5OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO\\n0.0 0.2 0.4 0.6 0.8 1.0−1.0 −0.5 0.0 0.5 1.0 1.5OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOLocal Regression\\nFIGURE 7.9. Local regression illustrated on some simulated data, where the\\nblue curve represents f(x)from which the data were generated, and the light\\norange curve corresponds to the local regression estimate ˆf(x). The orange colored\\npoints are local to the target point x0, represented by the orange vertical line. The\\nyellow bell-shape superimposed on the plot indicates weights assigned to each\\npoint, decreasing to zero with distance from the target point. The fit ˆf(x0)atx0\\nis obtained by fitting a weighted linear regression (orange line segment), and using\\nthe fitted value at x0(orange solid dot) as the estimate ˆf(x0).\\ning splines: it controls the flexibility of the non-linear fit. The smaller the\\nvalue of s, the more localand wiggly will be our fit; alternatively, a very\\nlarge value of swill lead to a global fit to the data using all of the train-\\ning observations. We can again use cross-validation to choose s, or we can\\nspecify it directly. Figure 7.10displays local linear regression fits on the\\nWagedata, using two values of s:0.7and0.2. As expected, the fit obtained\\nusings=0.7is smoother than that obtained using s=0.2.\\nThe idea of local regression can be generalized in many different ways.\\nIn a setting with multiple features X1,X2,...,X p, one very useful general-\\nization involves fitting a multiple linear regression model that is global in\\nsome variables, but local in another, such as time. Such varying coefficient\\nmodelsare a useful way of adapting a model to the most recently gatheredvarying\\ncoefficient\\nmodeldata. Local regression also generalizes very naturally when we want to fit\\nmodels that are local in a pair of variables X1andX2, rather than one.\\nWe can simply use two-dimensional neighborhoods, and fit bivariate linear\\nregression models using the observations that are near each target point\\nin two-dimensional space. Theoretically the same approach can be imple-\\nmented in higher dimensions, using linear regressions fit to p-dimensional\\nneighborhoods. However, local regression can perform poorly if pis much\\nlarger than about 3 or 4 because there will generally be very few training\\nobservations close to x0. Nearest-neighbors regression, discussed in Chap-\\nter3, suffers from a similar problem in high dimensions.', metadata={'source': './documents/ISLP_website.pdf', 'page': 311}), Document(page_content='7.7 Generalized Additive Models 305\\nAlgorithm 7.1 Local Regression At X=x0\\n1. Gather the fraction s=k/nof training points whose xiare closest\\ntox0.\\n2. Assign a weight Ki0=K(xi,x0)to each point in this neighborhood,\\nso that the point furthest from x0has weight zero, and the closest\\nhas the highest weight. All but these knearest neighbors get weight\\nzero.\\n3. Fit a weighted least squares regression of theyion thexiusing the\\naforementioned weights, by finding ˆβ0andˆβ1that minimize\\nn∑\\ni=1Ki0(yi−β0−β1xi)2. (7.14)\\n4. The fitted value at x0is given by ˆf(x0)=ˆβ0+ˆβ1x0.\\n20 30 40 50 60 70 800 50 100 200 300AgeWageLocal Linear RegressionSpan is 0.2  (16.4 Degrees of Freedom)Span is 0.7  (5.3 Degrees of Freedom)\\nFIGURE 7.10. Local linear fits to the Wagedata. The span specifies the fraction\\nof the data used to compute the fit at each target point.\\n7.7 Generalized Additive Models\\nIn Sections 7.1–7.6, we present a number of approaches for flexibly predict-\\ning a response Yon the basis of a single predictor X. These approaches can\\nbe seen as extensions of simple linear regression. Here we explore the prob-\\nlem of flexibly predicting Yon the basis of several predictors, X1,...,X p.\\nThis amounts to an extension of multiple linear regression.\\nGeneralized additive models (GAMs) provide a general framework forgeneralized\\nadditive\\nmodelextending a standard linear model by allowing non-linear functions of each\\nofthevariables,whilemaintaining additivity .Justlikelinearmodels,GAMs\\nadditivity can be applied with both quantitative and qualitative responses. We first', metadata={'source': './documents/ISLP_website.pdf', 'page': 312}), Document(page_content='306 7. Moving Beyond Linearity\\n2003 2005 2007 2009−30 −20 −10 0 10 20 3020 30 40 50 60 70 80−50 −40 −30 −20 −10 0 10 20\\n−30 −20 −10 0 10 20 30 40<HS HS <Coll Coll >Collf1(year)\\nf2(age)\\nf3(education)\\nyearageeducationFIGURE 7.11. For theWagedata, plots of the relationship between each feature\\nand the response, wage, in the fitted model ( 7.16). Each plot displays the fitted\\nfunction and pointwise standard errors. The first two functions are natural splines\\ninyearandage, with four and five degrees of freedom, respectively. The third\\nfunction is a step function, fit to the qualitative variable education .\\nexamine GAMs for a quantitative response in Section 7.7.1, and then for a\\nqualitative response in Section 7.7.2.\\n7.7.1 GAMs for Regression Problems\\nA natural way to extend the multiple linear regression model\\nyi=β0+β1xi1+β2xi2+···+βpxip+ϵi\\nin order to allow for non-linear relationships between each feature and the\\nresponse is to replace each linear component βjxijwith a (smooth) non-\\nlinear function fj(xij). We would then write the model as\\nyi=β0+p∑\\nj=1fj(xij)+ϵi\\n=β0+f1(xi1)+f2(xi2)+···+fp(xip)+ϵi. (7.15)\\nThis is an example of a GAM. It is called an additive model because we\\ncalculate a separate fjfor each Xj, and then add together all of their\\ncontributions.\\nIn Sections 7.1–7.6, we discuss many methods for fitting functions to a\\nsingle variable. The beauty of GAMs is that we can use these methods\\nas building blocks for fitting an additive model. In fact, for most of the\\nmethods that we have seen so far in this chapter, this can be done fairly\\ntrivially. Take, for example, natural splines, and consider the task of fitting\\nthe model\\nwage =β0+f1(year )+f2(age)+f3(education )+ϵ (7.16)\\non theWagedata. Here yearandageare quantitative variables, while the\\nvariable education is qualitative with five levels: <HS,HS,<Coll,Coll,>Coll,\\nreferring to the amount of high school or college education that an individ-\\nual has completed. We fit the first two functions using natural splines. We', metadata={'source': './documents/ISLP_website.pdf', 'page': 313}), Document(page_content='7.7 Generalized Additive Models 307\\n2003 2005 2007 2009−30 −20 −10 0 10 20 3020 30 40 50 60 70 80−50 −40 −30 −20 −10 0 10 20\\n−30 −20 −10 0 10 20 30 40<HS HS <Coll Coll >Collf1(year)\\nf2(age)\\nf3(education)\\nyearageeducationFIGURE 7.12. Details are as in Figure 7.11, but now f1andf2are smoothing\\nsplines with four and five degrees of freedom, respectively.\\nfit the third function using a separate constant for each level, via the usual\\ndummy variable approach of Section 3.3.1.\\nFigure7.11shows the results of fitting the model ( 7.16) using least\\nsquares. This is easy to do, since as discussed in Section 7.4, natural splines\\ncan be constructed using an appropriately chosen set of basis functions.\\nHence the entire model is just a big regression onto spline basis variables\\nand dummy variables, all packed into one big regression matrix.\\nFigure7.11can be easily interpreted. The left-hand panel indicates that\\nholding ageandeducation fixed,wagetends to increase slightly with year;\\nthis may be due to inflation. The center panel indicates that holding\\neducation andyearfixed,wagetends to be highest for intermediate val-\\nues ofage, and lowest for the very young and very old. The right-hand\\npanel indicates that holding yearandagefixed,wagetends to increase\\nwitheducation : the more educated a person is, the higher their salary, on\\naverage. All of these findings are intuitive.\\nFigure7.12shows a similar triple of plots, but this time f1andf2are\\nsmoothing splines with four and five degrees of freedom, respectively. Fit-\\ntingaGAMwithasmoothingsplineis notquiteassimpleasfittinga GAM\\nwith a natural spline, since in the case of smoothing splines, least squares\\ncannot be used. However, standard software such as the Pythonpackage\\npygamcan be used to fit GAMs using smoothing splines, via an approachpygam\\nknown as backfitting . This method fits a model involving multiple predic-backfittingtors by repeatedly updating the fit for each predictor in turn, holding the\\nothers fixed. The beauty of this approach is that each time we update a\\nfunction, we simply apply the fitting method for that variable to a partial\\nresidual.6\\nThe fitted functions in Figures 7.11and7.12look rather similar. In most\\nsituations, the differences in the GAMs obtained using smoothing splines\\nversus natural splines are small.\\n6A partial residual for X3, for example, has the form ri=yi−f1(xi1)−f2(xi2). If we\\nknowf1andf2, then we can fit f3by treating this residual as a response in a non-linear\\nregression on X3.', metadata={'source': './documents/ISLP_website.pdf', 'page': 314}), Document(page_content='308 7. Moving Beyond Linearity\\nWe do not have to use splines as the building blocks for GAMs: we can\\njust as well use local regression, polynomial regression, or any combination\\nof the approaches seen earlier in this chapter in order to create a GAM.\\nGAMsareinvestigatedinfurtherdetailinthelabattheendofthischapter.\\nPros and Cons of GAMs\\nBefore we move on, let us summarize the advantages and limitations of a\\nGAM.\\nLGAMs allow us to fit a non-linear fjto eachXj, so that we can\\nautomatically model non-linear relationships that standard linear re-\\ngression will miss. This means that we do not need to manually try\\nout many different transformations on each variable individually.\\nLThe non-linear fits can potentially make more accurate predictions\\nfor the response Y.\\nLBecause the model is additive, we can examine the effect of each Xj\\nonYindividually while holding all of the other variables fixed.\\nLThe smoothness of the function fjfor the variable Xjcan be sum-\\nmarized via degrees of freedom.\\nNThe main limitation of GAMs is that the model is restricted to be\\nadditive. With many variables, important interactions can be missed.\\nHowever, as with linear regression, we can manually add interaction\\nterms to the GAM model by including additional predictors of the\\nformXj×Xk. In addition we can add low-dimensional interaction\\nfunctions of the form fjk(Xj,Xk)into the model; such terms can\\nbe fit using two-dimensional smoothers such as local regression, or\\ntwo-dimensional splines (not covered here).\\nFor fully general models, we have to look for even more flexible approaches\\nsuchasrandomforestsandboosting,describedinChapter 8.GAMsprovide\\na useful compromise between linear and fully nonparametric models.\\n7.7.2 GAMs for Classification Problems\\nGAMs can also be used in situations where Yis qualitative. For simplicity,\\nhere we assume Ytakes on values 0or1, and let p(X) = Pr( Y=1|X)be\\nthe conditional probability (given the predictors) that the response equals\\none. Recall the logistic regression model ( 4.6):\\nlog(p(X)\\n1−p(X))\\n=β0+β1X1+β2X2+···+βpXp. (7.17)\\nTheleft-hand side is thelog of the oddsof P(Y=1|X)versusP(Y=0|X),\\nwhich (7.17) represents as a linear function of the predictors. A natural way\\nto extend ( 7.17) to allow for non-linear relationships is to use the model\\nlog(p(X)\\n1−p(X))\\n=β0+f1(X1)+f2(X2)+···+fp(Xp). (7.18)', metadata={'source': './documents/ISLP_website.pdf', 'page': 315}), Document(page_content='7.8 Lab: Non-Linear Modeling 309\\n2003 2005 2007 2009−4 −2 0 2 420 30 40 50 60 70 80−8 −6 −4 −2 0 2\\n−400 −200 0 200 400<HS HS <Coll Coll >Collf1(year)\\nf2(age)\\nf3(education)\\nyearageeducationFIGURE 7.13. For theWagedata, the logistic regression GAM given in ( 7.19)\\nis fit to the binary response I(wage>250) . Each plot displays the fitted function\\nand pointwise standard errors. The first function is linear in year, the second\\nfunction a smoothing spline with five degrees of freedom in age, and the third a\\nstep function for education . There are very wide standard errors for the first\\nlevel<HSofeducation .\\nEquation 7.18is a logistic regression GAM. It has all the same pros and\\ncons as discussed in the previous section for quantitative responses.\\nWe fit a GAM to the Wagedata in order to predict the probability that\\nan individual’s income exceeds $250 ,000per year. The GAM that we fit\\ntakes the form\\nlog(p(X)\\n1−p(X))\\n=β0+β1×year +f2(age)+f3(education ),(7.19)\\nwhere\\np(X) = Pr( wage >250|year ,age,education ).\\nOnce again f2is fit using a smoothing spline with five degrees of freedom,\\nandf3is fit as a step function, by creating dummy variables for each of the\\nlevels of education. The resulting fit is shown in Figure 7.13. The last panel\\nlooks suspicious, with very wide confidence intervals for level <HS. In fact,\\nnoresponsevaluesequaloneforthatcategory:noindividualswithlessthan\\na high school education make more than $250 ,000per year. Hence we refit\\nthe GAM, excluding the individuals with less than a high school education.\\nThe resulting model is shown in Figure 7.14. As in Figures 7.11and7.12,\\nall three panels have similar vertical scales. This allows us to visually assess\\nthe relative contributions of each of the variables. We observe that ageand\\neducation have a much larger effect than yearon the probability of being\\na high earner.\\n7.8 Lab: Non-Linear Modeling\\nIn this lab, we demonstrate some of the nonlinear models discussed in\\nthis chapter. We use the Wagedata as a running example, and show that\\nmany of the complex non-linear fitting procedures discussed can easily be\\nimplemented in Python.', metadata={'source': './documents/ISLP_website.pdf', 'page': 316}), Document(page_content=\"310 7. Moving Beyond Linearity\\n2003 2005 2007 2009−4 −2 0 2 420 30 40 50 60 70 80−8 −6 −4 −2 0 2\\n−4 −2 0 2 4HS <Coll Coll >Collf1(year)\\nf2(age)\\nf3(education)\\nyearageeducationFIGURE 7.14. The same model is fit as in Figure 7.13, this time excluding the\\nobservations for which education is<HS. Now we see that increased education\\ntends to be associated with higher salaries.\\nAs usual, we start with some of our standard imports.\\nIn[1]: import numpy as np, pandas as pd\\nfrom matplotlib.pyplot import subplots\\nimport statsmodels.api as sm\\nfrom ISLP import load_data\\nfrom ISLP.models import (summarize,\\npoly,\\nModelSpec as MS)\\nfrom statsmodels.stats.anova import anova_lm\\nWe again collect the new imports needed for this lab. Many of these are\\ndeveloped specifically for the ISLPpackage.\\nIn[2]: from pygam import (s as s_gam,\\nla sl _ g a m ,\\nfa sf _ g a m ,\\nLinearGAM,\\nLogisticGAM)\\nfrom ISLP.transforms import (BSpline,\\nNaturalSpline)\\nfrom ISLP.models import bs, ns\\nfrom ISLP.pygam import (approx_lam,\\ndegrees_of_freedom,\\nplot as plot_gam,\\nanova as anova_gam)\\n7.8.1 Polynomial Regression and Step Functions\\nWe start by demonstrating how Figure 7.1can be reproduced. Let’s begin\\nby loading the data.\\nIn[3]: Wage = load_data( 'Wage')\\ny=W a g e [ 'wage']\\nage = Wage[ 'age']\", metadata={'source': './documents/ISLP_website.pdf', 'page': 317}), Document(page_content=\"7.8 Lab: Non-Linear Modeling 311\\nThroughout most of this lab, our response is Wage['wage'] , which we\\nhave stored as yabove. As in Section 3.6.6, we will use the poly()function\\nto create a model matrix that will fit a 4th degree polynomial in age.\\nIn[4]: poly_age = MS([poly( 'age',d e g r e e = 4 ) ] ) . f i t ( W a g e )\\nM=s m . O L S ( y ,p o l y _ a g e . t r a n s f o r m ( W a g e ) ) . f i t ( )\\nsummarize(M)\\nOut[4]: coef std err t P>|t|\\nintercept 111.7036 0.729 153.283 0.000\\npoly(age, degree=4)[0] 447.0679 39.915 11.201 0.000\\npoly(age, degree=4)[1] -478.3158 39.915 -11.983 0.000\\npoly(age, degree=4)[2] 125.5217 39.915 3.145 0.002\\npoly(age, degree=4)[3] -77.9112 39.915 -1.952 0.051\\nThis polynomial is constructed using the function poly(), which cre-\\nates a special transformer Poly()(usingsklearnterminology for featuretransformertransformations such as PCA()seen in Section 6.5.3) which allows for easy\\nevaluation of the polynomial at new data points. Here poly()is referred\\nto as ahelperfunction, and sets up the transformation; Poly()is the ac-helpertual workhorse that computes the transformation. See also the discussion\\nof transformations on page 118.\\nIn the code above, the first line executes the fit()method using the\\ndataframe Wage. This recomputes and stores as attributes any parameters\\nneeded by Poly()on the training data, and these will be used on all sub-\\nsequent evaluations of the transform() method. For example, it is used on\\nthe second line, as well as in the plotting function developed below.\\nWe now create a grid of values for ageat which we want predictions.\\nIn[5]: age_grid = np.linspace(age.min(),\\nage.max(),\\n100)\\nage_df = pd.DataFrame({ 'age':a g e _ g r i d } )\\nFinally, we wish to plot the data and add the fit from the fourth-degree\\npolynomial. As we will make several similar plots below, we first write a\\nfunction to create all the ingredients and produce the plot. Our function\\ntakes in a model specification (here a basis specified by a transform), as\\nwell as a grid of agevalues. The function produces a fitted curve as well\\nas 95% confidence bands. By using an argument for basiswe can produce\\nand plot the results with several different transforms, such as the splines\\nwe will see shortly.\\nIn[6]: def plot_wage_fit(age_df,\\nbasis,\\ntitle):\\nX=b a s i s . t r a n s f o r m ( W a g e )\\nXnew = basis.transform(age_df)\\nM=s m . O L S ( y ,X ) . f i t ( )\\npreds = M.get_prediction(Xnew)\\nbands = preds.conf_int(alpha=0.05)\\nfig, ax = subplots(figsize=(8,8))\\nax.scatter(age,\\ny,\", metadata={'source': './documents/ISLP_website.pdf', 'page': 318}), Document(page_content=\"312 7. Moving Beyond Linearity\\nfacecolor= 'gray',\\nalpha=0.5)\\nfor val, ls in zip([preds.predicted_mean,\\nbands[:,0],\\nbands[:,1]],\\n['b','r--','r--']):\\nax.plot(age_df.values, val, ls, linewidth=3)\\nax.set_title(title, fontsize=20)\\nax.set_xlabel( 'Age',f o n t s i z e = 2 0 )\\nax.set_ylabel( 'Wage',f o n t s i z e = 2 0 ) ;\\nreturn ax\\nWe include an argument alphatoax.scatter() to add some transparency\\nto the points. This provides a visual indication of density. Notice the use\\nof thezip()function in the forloop above (see Section 2.3.8). We have\\nthree lines to plot, each with different colors and line types. Here zip()\\nconveniently bundles these together as iterators in the loop.7\\niteratorWe now plot the fit of the fourth-degree polynomial using this function.\\nIn[7]: plot_wage_fit(age_df,\\npoly_age,\\n'Degree-4 Polynomial' );\\nWith polynomial regression we must decide on the degree of the polyno-\\nmial to use. Sometimes we just wing it, and decide to use second or third\\ndegree polynomials, simply to obtain a nonlinear fit. But we can make such\\na decision in a more systematic way. One way to do this is through hypoth-\\nesis tests, which we demonstrate here. We now fit a series of models ranging\\nfrom linear (degree-one) to degree-five polynomials, and look to determine\\nthe simplest model that is sufficient to explain the relationship between\\nwageandage. We use the anova_lm() function, which performs a series of\\nANOVA tests. An analysis of variance or ANOVA tests the null hypothesisanalysis of\\nvariancethat a model M1is sufficient to explain the data against the alternative\\nhypothesis that a more complex model M2is required. The determination\\nis based on an F-test. To perform the test, the models M1andM2must\\nbenested: the space spanned by the predictors in M1must be a subspace\\nof the space spanned by the predictors in M2. In this case, we fit five dif-\\nferent polynomial models and sequentially compare the simpler model to\\nthe more complex model.\\nIn[8]: models = [MS([poly( 'age',d e g r e e = d ) ] )\\nfor d in range(1, 6)]\\nXs = [model.fit_transform(Wage) for model in models]\\nanova_lm(*[sm.OLS(y, X_).fit()\\nfor X_ in Xs])\\nOut[8]: df_resid ssr df_diff ss_diff F Pr(>F)\\n02 9 9 8 . 0 5 . 0 2 2 e + 0 6 0 . 0 NaN NaN NaN\\n12 9 9 7 . 0 4 . 7 9 3 e + 0 6 1 . 0 2 2 8 7 8 6 . 0 1 0 1 4 3 . 5 9 3 2 . 3 6 4 e - 3 2\\n22 9 9 6 . 0 4 . 7 7 8 e + 0 6 1 . 0 1 5 7 5 5 . 6 9 49 . 8 8 9 1 . 6 7 9 e - 0 3\\n32 9 9 5 . 0 4 . 7 7 2 e + 0 6 1 . 06 0 7 0 . 1 5 23 . 8 1 0 5 . 1 0 5 e - 0 2\\n7InPython speak, an “iterator” is an object with a finite number of values, that can\\nbe iterated on, as in a loop.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 319}), Document(page_content=\"7.8 Lab: Non-Linear Modeling 313\\n42 9 9 4 . 0 4 . 7 7 0 e + 0 6 1 . 01 2 8 2 . 5 6 30 . 8 0 5 3 . 6 9 7 e - 0 1\\nNotice the *in theanova_lm() line above. This function takes a variable\\nnumber of non-keyword arguments, in this case fitted models. When these\\nmodels are provided as a list (as is done here), it must be prefixed by *.\\nThe p-value comparing the linear models[0] to the quadratic models[1]\\nis essentially zero, indicating that a linear fit is not sufficient.8Similarly the\\np-value comparing the quadratic models[1] to the cubic models[2] is very\\nlow (0.0017), so the quadratic fit is also insufficient. The p-value comparing\\nthe cubic and degree-four polynomials, models[2] andmodels[3] , is approx-\\nimately 5%, while the degree-five polynomial models[4] seems unnecessary\\nbecause its p-value is 0.37. Hence, either a cubic or a quartic polynomial\\nappear to provide a reasonable fit to the data, but lower- or higher-order\\nmodels are not justified.\\nInthiscase,insteadofusingthe anova()function,wecouldhaveobtained\\nthese p-values more succinctly by exploiting the fact that poly()creates\\northogonal polynomials.\\nIn[9]: summarize(M)\\nOut[9]: coef std err t P>|t|\\nintercept 111.7036 0.729 153.283 0.000\\npoly(age, degree=4)[0] 447.0679 39.915 11.201 0.000\\npoly(age, degree=4)[1] -478.3158 39.915 -11.983 0.000\\npoly(age, degree=4)[2] 125.5217 39.915 3.145 0.002\\npoly(age, degree=4)[3] -77.9112 39.915 -1.952 0.051\\nNotice that the p-values are the same, and in fact the square of the\\nt-statistics are equal to the F-statistics from the anova_lm() function; for\\nexample:\\nIn[10]: (-11.983)**2\\nOut[10]: 143.59228\\nHowever, the ANOVA method works whether or not we used orthogonal\\npolynomials, provided the models are nested. For example, we can use\\nanova_lm() to compare the following three models, which all have a linear\\nterm in education and a polynomial in ageof different degrees:\\nIn[11]: models = [MS([ 'education' ,p o l y ( 'age',d e g r e e = d ) ] )\\nfor d in range(1, 4)]\\nXEs = [model.fit_transform(Wage)\\nfor model in models]\\nanova_lm(*[sm.OLS(y, X_).fit() for X_ in XEs])\\nOut[11]: df_resid ssr df_diff ss_diff F Pr(>F)\\n02 9 9 7 . 0 3 . 9 0 2 e + 0 6 0 . 0 NaN NaN NaN\\n12 9 9 6 . 0 3 . 7 5 9 e + 0 6 1 . 0 1 4 2 8 6 2 . 7 0 1 1 1 3 . 9 9 2 3 . 8 3 8 e - 2 6\\n22 9 9 5 . 0 3 . 7 5 4 e + 0 6 1 . 05 9 2 6 . 2 0 74 . 7 2 9 2 . 9 7 4 e - 0 2\\n8Indexing starting at zero is confusing for the polynomial degree example, since\\nmodels[1] is quadratic rather than linear!\", metadata={'source': './documents/ISLP_website.pdf', 'page': 320}), Document(page_content=\"314 7. Moving Beyond Linearity\\nAs an alternative to using hypothesis tests and ANOVA, we could choose\\nthe polynomial degree using cross-validation, as discussed in Chapter 5.\\nNext we consider the task of predicting whether an individual earns more\\nthan $250,000 per year. We proceed much as before, except that first we\\ncreate the appropriate response vector, and then apply the glm()function\\nusing the binomial family in order to fit a polynomial logistic regression\\nmodel.\\nIn[12]: X=p o l y _ a g e . t r a n s f o r m ( W a g e )\\nhigh_earn = Wage[ 'high_earn' ]=y>2 5 0 #s h o r t h a n d\\nglm = sm.GLM(y > 250,\\nX,\\nfamily=sm.families.Binomial())\\nB=g l m . f i t ( )\\nsummarize(B)\\nOut[12]: coef std err z P>|z|\\nintercept -4.3012 0.345 -12.457 0.000\\npoly(age, degree=4)[0] 71.9642 26.133 2.754 0.006\\npoly(age, degree=4)[1] -85.7729 35.929 -2.387 0.017\\npoly(age, degree=4)[2] 34.1626 19.697 1.734 0.083\\npoly(age, degree=4)[3] -47.4008 24.105 -1.966 0.049\\nOnce again, we make predictions using the get_prediction() method.\\nIn[13]: newX = poly_age.transform(age_df)\\npreds = B.get_prediction(newX)\\nbands = preds.conf_int(alpha=0.05)\\nWe now plot the estimated relationship.\\nIn[14]: fig, ax = subplots(figsize=(8,8))\\nrng = np.random.default_rng(0)\\nax.scatter(age +\\n0.2 * rng.uniform(size=y.shape[0]),\\nnp.where(high_earn, 0.198, 0.002),\\nfc='gray',\\nmarker= '|')\\nfor val, ls in zip([preds.predicted_mean,\\nbands[:,0],\\nbands[:,1]],\\n['b','r--','r--']):\\nax.plot(age_df.values, val, ls, linewidth=3)\\nax.set_title( 'Degree-4 Polynomial' ,f o n t s i z e = 2 0 )\\nax.set_xlabel( 'Age',f o n t s i z e = 2 0 )\\nax.set_ylim([0,0.2])\\nax.set_ylabel( 'P(Wage > 250)' ,f o n t s i z e = 2 0 ) ;\\nWe have drawn the agevalues corresponding to the observations with wage\\nvalues above 250 as gray marks on the top of the plot, and those with\\nwagevalues below 250 are shown as gray marks on the bottom of the plot.\\nWe added a small amount of noise to jitter the agevalues a bit so that\\nobservations with the same agevalue do not cover each other up. This type\\nof plot is often called a rug plot.rug plotIn order to fit a step function, as discussed in Section 7.2, we first use\\nthepd.qcut() function to discretize agebased on quantiles. Then we usepd.qcut()\", metadata={'source': './documents/ISLP_website.pdf', 'page': 321}), Document(page_content=\"7.8 Lab: Non-Linear Modeling 315\\npd.get_dummies() to create the columns of the model matrix for this cate-pd.get_\\ndummies()gorical variable. Note that this function will include allcolumns for a given\\ncategorical, rather than the usual approach which drops one of the levels.\\nIn[15]: cut_age = pd.qcut(age, 4)\\nsummarize(sm.OLS(y, pd.get_dummies(cut_age)).fit())\\nOut[15]: coef std err t P>|t|\\n(17.999, 33.75] 94.1584 1.478 63.692 0.0\\n(33.75, 42.0] 116.6608 1.470 79.385 0.0\\n(42.0, 51.0] 119.1887 1.416 84.147 0.0\\n(51.0, 80.0] 116.5717 1.559 74.751 0.0\\nHerepd.qcut() automatically picked the cutpoints based on the quan-\\ntiles 25%, 50% and 75%, which results in four regions. We could also have\\nspecified our own quantiles directly instead of the argument 4. For cuts\\nnot based on quantiles we would use the pd.cut() function. The functionpd.cut()pd.qcut() (andpd.cut() ) returns an ordered categorical variable. The re-\\ngression model then creates a set of dummy variables for use in the regres-\\nsion. Since ageis the only variable in the model, the value $94,158.40 is the\\naverage salary for those under 33.75 years of age, and the other coefficients\\nare the average salary for those in the other age groups. We can produce\\npredictions and plots just as we did in the case of the polynomial fit.\\n7.8.2 Splines\\nIn order to fit regression splines, we use transforms from the ISLPpackage.\\nThe actual spline evaluation functions are in the scipy.interpolate pack-\\nage; we have simply wrapped them as transforms similar to Poly()and\\nPCA().\\nIn Section 7.4, we saw that regression splines can be fit by constructing\\nan appropriate matrix of basis functions. The BSpline() function generatesBSpline()the entire matrix of basis functions for splines with the specified set of\\nknots. By default, the B-splines produced are cubic. To change the degree,\\nuse the argument degree.\\nIn[16]: bs_ = BSpline(internal_knots=[25,40,60], intercept=True).fit(age)\\nbs_age = bs_.transform(age)\\nbs_age.shape\\nOut[16]: (3000, 7)\\nThisresultsinaseven-columnmatrix,whichiswhatisexpectedforacubic-\\nspline basis with 3 interior knots. We can form this same matrix using the\\nbs()object, which facilitates adding this to a model-matrix builder (as in\\npoly()versus its workhorse Poly()) described in Section 7.8.1.\\nWe now fit a cubic spline model to the Wagedata.\\nIn[17]: bs_age = MS([bs( 'age',i n t e r n a l _ k n o t s = [ 2 5 , 4 0 , 6 0 ] ) ] )\\nXbs = bs_age.fit_transform(Wage)\\nM=s m . O L S ( y ,X b s ) . f i t ( )\\nsummarize(M)\", metadata={'source': './documents/ISLP_website.pdf', 'page': 322}), Document(page_content=\"316 7. Moving Beyond LinearityOut[17]:coef std err ...intercept 60.494 9.460 ...bs(age, internal_knots=[25, 40, 60])[0] 3.980 12.538 ...bs(age, internal_knots=[25, 40, 60])[1] 44.631 9.626 ...bs(age, internal_knots=[25, 40, 60])[2] 62.839 10.755 ...bs(age, internal_knots=[25, 40, 60])[3] 55.991 10.706 ...bs(age, internal_knots=[25, 40, 60])[4] 50.688 14.402 ...bs(age, internal_knots=[25, 40, 60])[5] 16.606 19.126 ...The column names are a little cumbersome, and have caused us to trun-cate the printed summary. They can be set on construction using thenameargument as follows.In[18]:bs_age = MS([bs('age',internal_knots=[25,40,60],name='bs(age)')])Xbs = bs_age.fit_transform(Wage)M=s m . O L S ( y ,X b s ) . f i t ( )summarize(M)Out[18]:coef std err t P>|t|intercept 60.494 9.460 6.394 0.000bs(age, knots)[0] 3.981 12.538 0.317 0.751bs(age, knots)[1] 44.631 9.626 4.636 0.000bs(age, knots)[2] 62.839 10.755 5.843 0.000bs(age, knots)[3] 55.991 10.706 5.230 0.000bs(age, knots)[4] 50.688 14.402 3.520 0.000bs(age, knots)[5] 16.606 19.126 0.868 0.385Notice that there are 6 spline coefficients rather than 7. This is because, bydefault,bs()assumesintercept=False, since we typically have an overallintercept in the model. So it generates the spline basis with the given knots,and then discards one of the basis functions to account for the intercept.We could also use thedf(degrees of freedom) option to specify the com-plexity of the spline. We see above that with 3 knots, the spline basis has6 columns or degrees of freedom. When we specifydf=6rather than theactual knots,bs()will produce a spline with 3 knots chosen at uniformquantiles of the training data. We can see these chosen knots most easilyusingBspline()directly:In[19]:BSpline(df=6).fit(age).internal_knots_Out[19]:array([33.75, 42.0, 51.0])When asking for six degrees of freedom, the transform chooses knots atages 33.75, 42.0, and 51.0, which correspond to the 25th, 50th, and 75thpercentiles ofage.When using B-splines we need not limit ourselves to cubic polynomials(i.e.degree=3). For instance, usingdegree=0results in piecewise constantfunctions, as in our example withpd.qcut()above.In[20]:bs_age0 = MS([bs('age',df=3,degree=0)]).fit(Wage)Xbs0 = bs_age0.transform(Wage)summarize(sm.OLS(y, Xbs0).fit())\", metadata={'source': './documents/ISLP_website.pdf', 'page': 323}), Document(page_content=\"7.8 Lab: Non-Linear Modeling 317\\nOut[20]: coef std err t P>|t|\\nintercept 94.158 1.478 63.687 0.0\\nbs(age, df=3, degree=0)[0] 22.349 2.152 10.388 0.0\\nbs(age, df=3, degree=0)[1] 24.808 2.044 12.137 0.0\\nbs(age, df=3, degree=0)[2] 22.781 2.087 10.917 0.0\\nThis fit should be compared with cell [15] where we use qcut()to create\\nfour bins by cutting at the 25%, 50% and 75% quantiles of age. Since we\\nspecified df=3for degree-zero splines here, there will also be knots at the\\nsame three quantiles. Although the coefficients appear different, we see that\\nthis is a result of the different coding. For example, the first coefficient is\\nidentical in both cases, and is the mean response in the first bin. For the\\nsecond coefficient, we have 94.158 + 22 .349 = 116 .507≈116.611, the latter\\nbeing the mean in the second bin in cell [15]. Here the intercept is coded by\\na column of ones, so the second, third and fourth coefficients are increments\\nfor those bins. Why is the sum not exactly the same? It turns out that the\\nqcut()uses≤, whilebs()uses<when deciding bin membership.\\nIn order to fit a natural spline, we use the NaturalSpline() transformNatural\\nSpline() with the corresponding helper ns(). Here we fit a natural spline with five\\ndegrees of freedom (excluding the intercept) and plot the results.\\nIn[21]: ns_age = MS([ns( 'age',d f = 5 ) ] ) . f i t ( W a g e )\\nM_ns = sm.OLS(y, ns_age.transform(Wage)).fit()\\nsummarize(M_ns)\\nOut[21]: coef std err tP > | t |\\nintercept 60.475 4.708 12.844 0.000\\nns(age, df=5)[0] 61.527 4.709 13.065 0.000\\nns(age, df=5)[1] 55.691 5.717 9.741 0.000\\nns(age, df=5)[2] 46.818 4.948 9.463 0.000\\nns(age, df=5)[3] 83.204 11.918 6.982 0.000\\nns(age, df=5)[4] 6.877 9.484 0.725 0.468\\nWe now plot the natural spline using our plotting function.\\nIn[22]: plot_wage_fit(age_df,\\nns_age,\\n'Natural spline, df=5' );\\n7.8.3 Smoothing Splines and GAMs\\nA smoothing spline is a special case of a GAM with squared-error loss\\nand a single feature. To fit GAMs in Pythonwe will use the pygampackagepygam\\nwhich can be installed via pip install pygam . The estimator LinearGAM()LinearGAM()uses squared-error loss. The GAM is specified by associating each column\\nof a model matrix with a particular smoothing operation: sfor smoothing\\nspline;lfor linear, and ffor factor or categorical variables. The argument 0\\npassedto sbelowindicatesthatthissmootherwillapplytothefirstcolumn\\nof a feature matrix. Below, we pass it a matrix with a single column: X_age.\\nThe argument lamis the penalty parameter λas discussed in Section 7.5.2.\\nIn[23]: X_age = np.asarray(age).reshape((-1,1))\\ngam = LinearGAM(s_gam(0, lam=0.6))\\ngam.fit(X_age, y)\", metadata={'source': './documents/ISLP_website.pdf', 'page': 324}), Document(page_content=\"318 7. Moving Beyond LinearityOut[23]:LinearGAM(callbacks=[Deviance(), Diffs()], fit_intercept=True,max_iter=100, scale=None, terms=s(0) + intercept, tol=0.0001,verbose=False)Thepygamlibrary generally expects a matrix of features so we reshapeageto be a matrix (a two-dimensional array) instead of a vector (i.e. a one-dimensional array). The-1in the call to thereshape()method tellsnumpyto impute the size of that dimension based on the remaining entries of theshape tuple.Let’s investigate how the fit changes with the smoothing parameterlam.The functionnp.logspace()is similar tonp.linspace()but spaces pointsnp.logspace()evenly on the log-scale. Below we varylamfrom10−2to106.In[24]:fig, ax = subplots(figsize=(8,8))ax.scatter(age, y, facecolor='gray',a l p h a = 0 . 5 )for lam in np.logspace(-2, 6, 5):gam = LinearGAM(s_gam(0, lam=lam)).fit(X_age, y)ax.plot(age_grid,gam.predict(age_grid),label='{:.1e}'.format(lam),linewidth=3)ax.set_xlabel('Age',f o n t s i z e = 2 0 )ax.set_ylabel('Wage',f o n t s i z e = 2 0 ) ;ax.legend(title='$\\\\lambda$');Thepygampackage can perform a search for an optimal smoothing pa-rameter.In[25]:gam_opt = gam.gridsearch(X_age, y)ax.plot(age_grid,gam_opt.predict(age_grid),label='Grid search',linewidth=4)ax.legend()figAlternatively, we can fix the degrees of freedom of the smoothing splineusing a function included in theISLP.pygampackage. Below we find a valueofλthat gives us roughly four degrees of freedom. We note here that thesedegrees of freedom include the unpenalized intercept and linear term of thesmoothing spline, hence there are at least two degrees of freedom.In[26]:age_term = gam.terms[0]lam_4 = approx_lam(X_age, age_term, 4)age_term.lam = lam_4degrees_of_freedom(X_age, age_term)Out[26]:4.000000100004728Let’svarythedegreesoffreedominasimilarplottoabove.Wechoosethedegrees of freedom as the desired degrees of freedom plus one to accountfor the fact that these smoothing splines always have an intercept term.Hence, a value of one fordfis just a linear fit.In[27]:fig, ax = subplots(figsize=(8,8))ax.scatter(X_age,y,\", metadata={'source': './documents/ISLP_website.pdf', 'page': 325}), Document(page_content=\"7.8 Lab: Non-Linear Modeling 319\\nfacecolor= 'gray',\\nalpha=0.3)\\nfor df in [1,3,4,8,15]:\\nlam = approx_lam(X_age, age_term, df+1)\\nage_term.lam = lam\\ngam.fit(X_age, y)\\nax.plot(age_grid,\\ngam.predict(age_grid),\\nlabel='{:d}'.format(df),\\nlinewidth=4)\\nax.set_xlabel( 'Age',f o n t s i z e = 2 0 )\\nax.set_ylabel( 'Wage',f o n t s i z e = 2 0 ) ;\\nax.legend(title= 'Degrees of freedom' );\\nAdditive Models with Several Terms\\nThe strength of generalized additive models lies in their ability to fit mul-\\ntivariate regression models with more flexibility than linear models. We\\ndemonstrate two approaches: the first in a more manual fashion using nat-\\nural splines and piecewise constant functions, and the second using the\\npygampackage and smoothing splines.\\nWe now fit a GAM by hand to predict wageusing natural spline functions\\nofyearandage, treating education as a qualitative predictor, as in ( 7.16).\\nSince this is just a big linear regression model using an appropriate choice\\nof basis functions, we can simply do this using the sm.OLS() function.\\nWe will build the model matrix in a more manual fashion here, since we\\nwish to access the pieces separately when constructing partial dependence\\nplots.\\nIn[28]: ns_age = NaturalSpline(df=4).fit(age)\\nns_year = NaturalSpline(df=5).fit(Wage[ 'year'])\\nXs = [ns_age.transform(age),\\nns_year.transform(Wage[ 'year']),\\npd.get_dummies(Wage[ 'education' ]).values]\\nX_bh = np.hstack(Xs)\\ngam_bh = sm.OLS(y, X_bh).fit()\\nHere the function NaturalSpline() is the workhorse supporting the ns()\\nhelper function. We chose to use all columns of the indicator matrix for\\nthe categorical variable education , making an intercept redundant. Finally,\\nwe stacked the three component matrices horizontally to form the model\\nmatrixX_bh.\\nWe now show how to construct partial dependence plots for each of the\\nterms in our rudimentary GAM. We can do this by hand, given grids for\\nageandyear. We simply predict with new Xmatrices, fixing all but one\\nof the features at a time.\\nIn[29]: age_grid = np.linspace(age.min(),\\nage.max(),\\n100)\\nX_age_bh = X_bh.copy()[:100]\\nX_age_bh[:] = X_bh[:].mean(0)[None,:]\\nX_age_bh[:,:4] = ns_age.transform(age_grid)\\npreds = gam_bh.get_prediction(X_age_bh)\\nbounds_age = preds.conf_int(alpha=0.05)\", metadata={'source': './documents/ISLP_website.pdf', 'page': 326}), Document(page_content=\"320 7. Moving Beyond Linearity\\npartial_age = preds.predicted_mean\\ncenter = partial_age.mean()\\npartial_age -= center\\nbounds_age -= center\\nfig, ax = subplots(figsize=(8,8))\\nax.plot(age_grid, partial_age, 'b',l i n e w i d t h = 3 )\\nax.plot(age_grid, bounds_age[:,0], 'r--',l i n e w i d t h = 3 )\\nax.plot(age_grid, bounds_age[:,1], 'r--',l i n e w i d t h = 3 )\\nax.set_xlabel( 'Age')\\nax.set_ylabel( 'Effect on wage' )\\nax.set_title( 'Partial dependence of age on wage' ,f o n t s i z e = 2 0 ) ;\\nLet’s explain in some detail what we did above. The idea is to create a new\\nprediction matrix, where all but the columns belonging to ageare constant\\n(and set to their training-data means). The four columns for ageare filled\\nin with the natural spline basis evaluated at the 100 values in age_grid .\\n1. We made a grid of length 100 in age, and created a matrix X_age_bh\\nwith 100 rows and the same number of columns as X_bh.\\n2. We replaced every row of this matrix with the column means of the\\noriginal.\\n3. We then replace just the first four columns representing agewith the\\nnatural spline basis computed at the values in age_grid .\\nThe remaining steps should by now be familiar.\\nWe also look at the effect of yearonwage; the process is the same.\\nIn[30]: year_grid = np.linspace(2003, 2009, 100)\\nyear_grid = np.linspace(Wage[ 'year'].min(),\\nWage['year'].max(),\\n100)\\nX_year_bh = X_bh.copy()[:100]\\nX_year_bh[:] = X_bh[:].mean(0)[None,:]\\nX_year_bh[:,4:9] = ns_year.transform(year_grid)\\npreds = gam_bh.get_prediction(X_year_bh)\\nbounds_year = preds.conf_int(alpha=0.05)\\npartial_year = preds.predicted_mean\\ncenter = partial_year.mean()\\npartial_year -= center\\nbounds_year -= center\\nfig, ax = subplots(figsize=(8,8))\\nax.plot(year_grid, partial_year, 'b',l i n e w i d t h = 3 )\\nax.plot(year_grid, bounds_year[:,0], 'r--',l i n e w i d t h = 3 )\\nax.plot(year_grid, bounds_year[:,1], 'r--',l i n e w i d t h = 3 )\\nax.set_xlabel( 'Year')\\nax.set_ylabel( 'Effect on wage' )\\nax.set_title( 'Partial dependence of year on wage' ,f o n t s i z e = 2 0 ) ;\\nWe now fit the model ( 7.16) using smoothing splines rather than nat-\\nural splines. All of the terms in ( 7.16) are fit simultaneously, taking each\\nother into account to explain the response. The pygampackage only works\\nwith matrices, so we must convert the categorical series education to its\\narray representation, which can be found with the cat.codes attribute of\\neducation . Asyearonly has 7 unique values, we use only seven basis func-\\ntions for it.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 327}), Document(page_content=\"7.8 Lab: Non-Linear Modeling 321In[31]:gam_full = LinearGAM(s_gam(0) +s_gam(1, n_splines=7) +f_gam(2, lam=0))Xgam = np.column_stack([age,Wage['year'],Wage['education'].cat.codes])gam_full = gam_full.fit(Xgam, y)The twos_gam()terms result in smoothing spline fits, and use a defaultvalue forλ(lam=0.6), which is somewhat arbitrary. For the categorical termeducation, specified using af_gam()term, we specifylam=0to avoid anyshrinkage. We produce the partial dependence plot inageto see the effectof these choices.The values for the plot are generated by thepygampackage. We provideaplot_gam()function for partial-dependence plots inISLP.pygam, whichplot_gam()makes this job easier than in our last example with natural splines.In[32]:fig, ax = subplots(figsize=(8,8))plot_gam(gam_full, 0, ax=ax)ax.set_xlabel('Age')ax.set_ylabel('Effect on wage')ax.set_title('Partial dependence of age on wage - default lam=0.6',fontsize=20);We see that the function is somewhat wiggly. It is more natural to specifythedfthan a value forlam. We refit a GAM using four degrees of freedomeach forageandyear. Recall that the addition of one below takes intoaccount the intercept of the smoothing spline.In[33]:age_term = gam_full.terms[0]age_term.lam = approx_lam(Xgam, age_term, df=4+1)year_term = gam_full.terms[1]year_term.lam = approx_lam(Xgam, year_term, df=4+1)gam_full = gam_full.fit(Xgam, y)Note that updatingage_term.lamabove updates it ingam_full.terms[0]aswell! Likewise foryear_term.lam.Repeating the plot forage, we see that it is much smoother. We alsoproduce the plot foryear.In[34]:fig, ax = subplots(figsize=(8,8))plot_gam(gam_full,1,ax=ax)ax.set_xlabel('Year')ax.set_ylabel('Effect on wage')ax.set_title('Partial dependence of year on wage',f o n t s i z e = 2 0 )Finally we ploteducation, which is categorical. The partial dependenceplot is different, and more suitable for the set of fitted constants for eachlevel of this variable.In[35]:fig, ax = subplots(figsize=(8, 8))ax = plot_gam(gam_full, 2)ax.set_xlabel('Education')ax.set_ylabel('Effect on wage')\", metadata={'source': './documents/ISLP_website.pdf', 'page': 328}), Document(page_content=\"322 7. Moving Beyond Linearityax.set_title('Partial dependence of wage on education',fontsize=20);ax.set_xticklabels(Wage['education'].cat.categories, fontsize=8);ANOVA Tests for Additive ModelsInallofourmodels,thefunctionofyearlooksratherlinear.Wecanperforma series of ANOVA tests in order to determine which of these three modelsis best: a GAM that excludesyear(M1), a GAM that uses a linear functionofyear(M2), or a GAM that uses a spline function ofyear(M3).In[36]:gam_0 = LinearGAM(age_term + f_gam(2, lam=0))gam_0.fit(Xgam, y)gam_linear = LinearGAM(age_term +l_gam(1, lam=0) +f_gam(2, lam=0))gam_linear.fit(Xgam, y)Out[36]:LinearGAM(callbacks=[Deviance(), Diffs()], fit_intercept=True,max_iter=100, scale=None, terms=s(0) + l(1) + f(2) + intercept,tol=0.0001, verbose=False)Notice our use ofage_termin the expressions above. We do this becauseearlier we set the value forlamin this term to achieve four degrees offreedom.To directly assess the effect ofyearwe run an ANOVA on the threemodels fit above.In[37]:anova_gam(gam_0, gam_linear, gam_full)Out[37]:deviance df deviance_diff df_diff F pvalue03 7 1 4 3 6 2 . 3 6 6 2 9 9 1 . 0 0 4NaN NaN NaN NaN13 6 9 6 7 4 5 . 8 2 3 2 9 9 0 . 0 0 5 1 7 6 1 6 . 5 4 3 0 . 9 9 9 1 4 . 2 6 50 . 0 0 223 6 9 3 1 4 2 . 9 3 0 2 9 8 7 . 0 0 7 3 6 0 2 . 8 9 4 2 . 9 9 80 . 9 7 20 . 4 3 6WefindthatthereiscompellingevidencethataGAMwithalinearfunctioninyearis better than a GAM that does not includeyearat all (p-value=0.002). However, there is no evidence that a non-linear function ofyearis needed (p-value=0.435). In other words, based on the results of thisANOVA,M2is preferred.We can repeat the same process forageas well. We see there is very clearevidence that a non-linear term is required forage.In[38]:gam_0 = LinearGAM(year_term +f_gam(2, lam=0))gam_linear = LinearGAM(l_gam(0, lam=0) +year_term +f_gam(2, lam=0))gam_0.fit(Xgam, y)gam_linear.fit(Xgam, y)anova_gam(gam_0, gam_linear, gam_full)\", metadata={'source': './documents/ISLP_website.pdf', 'page': 329}), Document(page_content=\"7.8 Lab: Non-Linear Modeling 323Out[38]:deviance df deviance_diff df_diff F pvalue03 9 7 5 4 4 3 . 0 4 5 2 9 9 1 . 0 0 1NaN NaN NaN NaN13 8 5 0 2 4 6 . 9 0 8 2 9 9 0 . 0 0 1 1 2 5 1 9 6 . 1 3 7 1 . 0 0 0 1 0 1 . 2 7 00 . 0 0 023 6 9 3 1 4 2 . 9 3 0 2 9 8 7 . 0 0 7 1 5 7 1 0 3 . 9 7 8 2 . 9 9 34 2 . 4 4 80 . 0 0 0There is a (verbose)summary()method for the GAM fit. (We do notreproduce it here.)In[39]:gam_full.summary()Wecanmakepredictionsfromgamobjects,justlikefromlmobjects,usingthepredict()method for the classgam. Here we make predictions on thetraining set.In[40]:Yhat = gam_full.predict(Xgam)In order to fit a logistic regression GAM, we useLogisticGAM()fromLogisticGAM()pygam.In[41]:gam_logit = LogisticGAM(age_term +l_gam(1, lam=0) +f_gam(2, lam=0))gam_logit.fit(Xgam, high_earn)Out[41]:LogisticGAM(callbacks=[Deviance(), Diffs(), Accuracy()],fit_intercept=True, max_iter=100,terms=s(0) + l(1) + f(2) + intercept, tol=0.0001, verbose=False)In[42]:fig, ax = subplots(figsize=(8, 8))ax = plot_gam(gam_logit, 2)ax.set_xlabel('Education')ax.set_ylabel('Effect on wage')ax.set_title('Partial dependence of wage on education',fontsize=20);ax.set_xticklabels(Wage['education'].cat.categories, fontsize=8);The model seems to be very flat, with especially high error bars for thefirst category. Let’s look at the data a bit more closely.In[43]:pd.crosstab(Wage['high_earn'], Wage['education'])We see that there are no high earners in the first category of education,meaning that the model will have a hard time fitting. We will fit a logisticregression GAM excluding all observations falling into this category. Thisprovides more sensible results.To do so, we could subset the model matrix, though this will not removethe column fromXgam. While we can deduce which column corresponds tothis feature, for reproducibility’s sake we reform the model matrix on thissmaller subset.In[44]:only_hs = Wage['education']= ='1. < HS Grad'Wage_ = Wage.loc[∼only_hs]Xgam_ = np.column_stack([Wage_['age'],Wage_['year'],Wage_['education'].cat.codes-1])high_earn_ = Wage_['high_earn']\", metadata={'source': './documents/ISLP_website.pdf', 'page': 330}), Document(page_content=\"324 7. Moving Beyond LinearityIn the second-to-last line above, we subtract one from the codes of thecategory, due to a bug inpygam. It just relabels the education values andhence has no effect on the fit.We now fit the model.In[45]:gam_logit_ = LogisticGAM(age_term +year_term +f_gam(2, lam=0))gam_logit_.fit(Xgam_, high_earn_)Out[45]:LogisticGAM(callbacks=[Deviance(), Diffs(), Accuracy()],fit_intercept=True, max_iter=100,terms=s(0) + s(1) + f(2) + intercept, tol=0.0001, verbose=False)Let’s look at the effect ofeducation,yearandageon high earner statusnow that we’ve removed those observations.In[46]:fig, ax = subplots(figsize=(8, 8))ax = plot_gam(gam_logit_, 2)ax.set_xlabel('Education')ax.set_ylabel('Effect on wage')ax.set_title('Partial dependence of high earner status on education',f o n t s i z e = 2 0 ) ;ax.set_xticklabels(Wage['education'].cat.categories[1:],fontsize=8);In[47]:fig, ax = subplots(figsize=(8, 8))ax = plot_gam(gam_logit_, 1)ax.set_xlabel('Year')ax.set_ylabel('Effect on wage')ax.set_title('Partial dependence of high earner status on year',fontsize=20);In[48]:fig, ax = subplots(figsize=(8, 8))ax = plot_gam(gam_logit_, 0)ax.set_xlabel('Age')ax.set_ylabel('Effect on wage')ax.set_title('Partial dependence of high earner status on age',fontsize=20);7.8.4 Local RegressionWe illustrate the use of local regression using thelowess()function fromlowess()sm.nonparametric. Some implementations of GAMs allow terms to be localregression operators; this is not the case inpygam.Here we fit local linear regression models using spans of 0.2 and 0.5;that is, each neighborhood consists of 20% or 50% of the observations. Asexpected, using a span of 0.5 is smoother than 0.2.In[49]:lowess = sm.nonparametric.lowessfig, ax = subplots(figsize=(8,8))ax.scatter(age, y, facecolor='gray',a l p h a = 0 . 5 )for span in [0.2, 0.5]:fitted = lowess(y,\", metadata={'source': './documents/ISLP_website.pdf', 'page': 331}), Document(page_content=\"7.9 Exercises 325age,frac=span,xvals=age_grid)ax.plot(age_grid,fitted,label='{:.1f}'.format(span),linewidth=4)ax.set_xlabel('Age',f o n t s i z e = 2 0 )ax.set_ylabel('Wage',f o n t s i z e = 2 0 ) ;ax.legend(title='span',f o n t s i z e = 1 5 ) ;7.9 ExercisesConceptual1. It was mentioned in this chapter that a cubic regression spline withone knot atξcan be obtained using a basis of the formx,x2,x3,\\n(x−ξ)3+, where(x−ξ)3+=(x−ξ)3ifx>ξand equals 0 otherwise.We will now show that a function of the formf(x)=β0+β1x+β2x2+β3x3+β4(x−ξ)3+isindeedacubicregressionspline,regardlessofthevaluesofβ0,β1,β2,β3,β4.(a) Find a cubic polynomialf1(x)=a1+b1x+c1x2+d1x3such thatf(x)=f1(x)for allx≤ξ. Expressa1,b1,c1,d1interms ofβ0,β1,β2,β3,β4.(b) Find a cubic polynomialf2(x)=a2+b2x+c2x2+d2x3such thatf(x)=f2(x)for allx>ξ. Expressa2,b2,c2,d2interms ofβ0,β1,β2,β3,β4. We have now established thatf(x)isa piecewise polynomial.(c) Show thatf1(ξ)=f2(ξ). That is,f(x)is continuous atξ.(d) Show thatf′1(ξ)=f′2(ξ). That is,f′(x)is continuous atξ.(e) Show thatf′′1(ξ)=f′′2(ξ). That is,f′′(x)is continuous atξ.Therefore,f(x)is indeed a cubic spline.Hint: Parts (d) and (e) of this problem require knowledge of single-variable calculus. As a reminder, given a cubic polynomialf1(x)=a1+b1x+c1x2+d1x3,the first derivative takes the formf′1(x)=b1+2c1x+3d1x2\", metadata={'source': './documents/ISLP_website.pdf', 'page': 332}), Document(page_content='326 7. Moving Beyond Linearityand the second derivative takes the formf′′1(x)=2c1+6d1x.2. Suppose that a curveˆgis computed to smoothly fit a set ofnpointsusing the following formula:ˆg= arg ming(n∑i=1(yi−g(xi))2+λ∫[g(m)(x)]2dx),whereg(m)represents themth derivative ofg(andg(0)=g). Provideexample sketches ofˆgin each of the following scenarios.(a)λ=∞,m=0.(b)λ=∞,m=1.(c)λ=∞,m=2.(d)λ=∞,m=3.(e)λ=0,m=3.3. Suppose we fit a curve with basis functionsb1(X)=X,b2(X)=(X−1)2I(X≥1). (Note thatI(X≥1)equals1forX≥1and0otherwise.) We fit the linear regression modelY=β0+β1b1(X)+β2b2(X)+ϵ,and obtain coefficient estimatesˆβ0=1,ˆβ1=1,ˆβ2=−2. Sketch theestimated curve betweenX=−2andX=2. Note the intercepts,slopes, and other relevant information.4. Suppose we fit a curve with basis functionsb1(X)=I(0≤X≤2)−(X−1)I(1≤X≤2),b2(X)=(X−3)I(3≤X≤4) +I(4<X≤5).We fit the linear regression modelY=β0+β1b1(X)+β2b2(X)+ϵ,and obtain coefficient estimatesˆβ0=1,ˆβ1=1,ˆβ2=3. Sketch theestimated curve betweenX=−2andX=6. Note the intercepts,slopes, and other relevant information.5. Consider two curves,ˆg1andˆg2, defined byˆg1= arg ming(n∑i=1(yi−g(xi))2+λ∫[g(3)(x)]2dx),ˆg2= arg ming(n∑i=1(yi−g(xi))2+λ∫[g(4)(x)]2dx),whereg(m)represents themth derivative ofg.(a) Asλ→∞, willˆg1orˆg2have the smaller training RSS?(b) Asλ→∞, willˆg1orˆg2have the smaller test RSS?(c) Forλ=0, willˆg1orˆg2have the smaller training and test RSS?', metadata={'source': './documents/ISLP_website.pdf', 'page': 333}), Document(page_content='7.9 Exercises 327Applied6. In this exercise, you will further analyze theWagedata set consideredthroughout this chapter.(a) Perform polynomial regression to predictwageusingage. Usecross-validation to select the optimal degreedfor the polyno-mial. What degree was chosen, and how does this compare tothe results of hypothesis testing using ANOVA? Make a plot ofthe resulting polynomial fit to the data.(b) Fit a step function to predictwageusingage, and perform cross-validation to choose the optimal number of cuts. Make a plot ofthe fit obtained.7. TheWagedata set contains a number of other features not exploredin this chapter, such as marital status (maritl), job class (jobclass),and others. Explore the relationships between some of these otherpredictors andwage, and use non-linear fitting techniques in order tofit flexible models to the data. Create plots of the results obtained,and write a summary of your findings.8. Fit some of the non-linear models investigated in this chapter to theAutodata set. Is there evidence for non-linear relationships in thisdata set? Create some informative plots to justify your answer.9. This question uses the variablesdis(the weighted mean of distancesto five Boston employment centers) andnox(nitrogen oxides concen-tration in parts per 10 million) from theBostondata. We will treatdisas the predictor andnoxas the response.(a) Use thepoly()function from theISLP.modelsmodule to fit acubic polynomial regression to predictnoxusingdis. Report theregression output, and plot the resulting data and polynomialfits.(b) Plot the polynomial fits for a range of different polynomialdegrees (say, from 1 to 10), and report the associated residualsum of squares.(c) Perform cross-validation or another approach to select the opti-mal degree for the polynomial, and explain your results.(d) Use thebs()function from theISLP.modelsmodule to fit a re-gression spline to predictnoxusingdis. Report the output forthe fit using four degrees of freedom. How did you choose theknots? Plot the resulting fit.(e) Now fit a regression spline for a range of degrees of freedom, andplot the resulting fits and report the resulting RSS. Describe theresults obtained.(f) Perform cross-validation or another approach in order to selectthe best degrees of freedom for a regression spline on this data.Describe your results.', metadata={'source': './documents/ISLP_website.pdf', 'page': 334}), Document(page_content='328 7. Moving Beyond Linearity\\n10. This question relates to the Collegedata set.\\n(a) Splitthedataintoatrainingsetandatestset.Usingout-of-state\\ntuition as the response and the other variables as the predictors,\\nperform forward stepwise selection on the training set in order\\nto identify a satisfactory model that uses just a subset of the\\npredictors.\\n(b) Fit a GAM on the training data, using out-of-state tuition as\\nthe response and the features selected in the previous step as\\nthe predictors. Plot the results, and explain your findings.\\n(c) Evaluate the model obtained on the test set, and explain the\\nresults obtained.\\n(d) For which variables, if any, is there evidence of a non-linear\\nrelationship with the response?\\n11. In Section 7.7, it was mentioned that GAMs are generally fit using\\nabackfitting approach. The idea behind backfitting is actually quite\\nsimple. We will now explore backfitting in the context of multiple\\nlinear regression.\\nSuppose that we would like to perform multiple linear regression, but\\nwe do not have software to do so. Instead, we only have software\\nto perform simple linear regression. Therefore, we take the following\\niterative approach: we repeatedly hold all but one coefficient esti-\\nmate fixed at its current value, and update only that coefficient\\nestimateusingasimplelinearregression.Theprocessiscontinuedun-\\ntilconvergence —that is, until the coefficient estimates stop changing.\\nWe now try this out on a toy example.\\n(a) Generate a response Yand two predictors X1andX2, with\\nn= 100 .\\n(b) Write a function simple_reg() that takes two arguments outcome\\nandfeature, fits a simple linear regression model with this out-\\ncome and feature, and returns the estimated interceptand slope.\\n(c) Initialize beta1to take on a value of your choice. It does not\\nmatter what value you choose.\\n(d) Keeping beta1fixed, use your function simple_reg() to fit the\\nmodel:\\nY−beta1 ·X1=β0+β2X2+ϵ.\\nStore the resulting values as beta0andbeta2.\\n(e) Keeping beta2fixed, fit the model\\nY−beta2 ·X2=β0+β1X1+ϵ.\\nStore the result as beta0andbeta1(overwriting their previous\\nvalues).\\n(f) Write a for loop to repeat (c) and (d) 1,000 times. Report the\\nestimates of beta0,beta1, andbeta2at each iteration of the for\\nloop. Create a plot in which each of these values is displayed,\\nwithbeta0,beta1, andbeta2.', metadata={'source': './documents/ISLP_website.pdf', 'page': 335}), Document(page_content='7.9 Exercises 329(g) Compare your answer in (e) to the results of simply perform-ing multiple linear regression to predictYusingX1andX2.Useaxline()method to overlay those multiple linear regressioncoefficient estimates on the plot obtained in (e).(h) On this data set, how many backfitting iterations were requiredin order to obtain a “good” approximation to the multiple re-gression coefficient estimates?12. This problem is a continuation of the previous exercise. In a toyexample withp= 100, show that one can approximate the multiplelinearregressioncoefficientestimatesbyrepeatedlyperformingsimplelinear regression in a backfitting procedure. How many backfittingiterations are required in order to obtain a “good” approximation tothe multiple regression coefficient estimates? Create a plot to justifyyour answer.', metadata={'source': './documents/ISLP_website.pdf', 'page': 336}), Document(page_content='8\\nTree-Based Methods\\nIn this chapter, we describe tree-based methods for regression and classifi-\\ncation. These involve stratifying orsegmenting the predictor space into a\\nnumber of simple regions. In order to make a prediction for a given ob-\\nservation, we typically use the mean or the mode response value for the\\ntraining observations in the region to which it belongs. Since the set of\\nsplitting rules used to segment the predictor space can be summarized in\\na tree, these types of approaches are known as decision tree methods.decision treeTree-based methods are simple and useful for interpretation. However,\\nthey typically are not competitive with the best supervised learning ap-\\nproaches, such as those seen in Chapters 6and7, in terms of prediction\\naccuracy. Hence in this chapter we also introduce bagging,random forests ,\\nboosting, andBayesian additive regression trees . Each of these approaches\\ninvolves producing multiple trees which are then combined to yield a single\\nconsensus prediction. We will see that combining a large number of trees\\ncan often result in dramatic improvements in prediction accuracy, at the\\nexpense of some loss in interpretation.\\n8.1 The Basics of Decision Trees\\nDecisiontreescanbeappliedtobothregressionandclassificationproblems.\\nWe first consider regression problems, and then move on to classification.\\n8.1.1 Regression Trees\\nIn order to motivate regression trees , we begin with a simple example.regression\\ntree\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning , Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_8  331', metadata={'source': './documents/ISLP_website.pdf', 'page': 337}), Document(page_content='332 8. Tree-Based Methods|Years < 4.5\\nHits < 117.55.116.006.74FIGURE 8.1. For the Hitters data, a regression tree for predicting the log\\nsalary of a baseball player, based on the number of years that he has played in\\nthe major leagues and the number of hits that he made in the previous year. At a\\ngiven internal node, the label (of the form Xj<tk) indicates the left-hand branch\\nemanating from that split, and the right-hand branch corresponds to Xj≥tk.\\nFor instance, the split at the top of the tree results in two large branches. The\\nleft-hand branch corresponds to Years<4.5 , and the right-hand branch corresponds\\ntoYears>=4.5 . The tree has two internal nodes and three terminal nodes, or\\nleaves. The number in each leaf is the mean of the response for the observations\\nthat fall there.\\nPredicting Baseball Players’ Salaries Using Regression Trees\\nWe use the Hittersdata set to predict a baseball player’s Salarybased on\\nYears(the number of years that he has played in the major leagues) and\\nHits(thenumberofhitsthathemadeinthepreviousyear).Wefirstremove\\nobservations that are missing Salaryvalues, and log-transform Salaryso\\nthat its distribution has more of a typical bell-shape. (Recall that Salary\\nis measured in thousands of dollars.)\\nFigure8.1shows a regression tree fit to this data. It consists of a series\\nof splitting rules, starting at the top of the tree. The top split assigns\\nobservations having Years<4.5 to the left branch.1The predicted salary\\nfor these players is given by the mean response value for the players in\\nthe data set with Years<4.5 . For such players, the mean log salary is 5.107,\\nand so we make a prediction of e5.107thousands of dollars, i.e. $165,174, for\\nthese players. Players with Years>=4.5 are assigned to the right branch, and\\nthen that group is further subdivided by Hits. Overall, the tree stratifies\\nor segments the players into three regions of predictor space: players who\\nhave played for four or fewer years, players who have played for five or more\\nyears and who made fewer than 118 hits last year, and players who have\\nplayedforfiveormoreyearsandwhomadeatleast118hitslastyear.These\\nthree regions can be written as R1={X|Years<4.5 },R2={X|Years>=4.5 ,\\nHits<117.5 }, andR3={X|Years>=4.5 ,Hits>=117.5 }. Figure 8.2illustrates\\n1BothYearsandHitsare integers in these data; the function used to fit this tree\\nlabels the splits at the midpoint between two adjacent values.', metadata={'source': './documents/ISLP_website.pdf', 'page': 338}), Document(page_content='8.1 The Basics of Decision Trees 333\\nYearsHits\\n1117.5238\\n14.524R1R3R2FIGURE 8.2. The three-region partition for the Hitters data set from the\\nregression tree illustrated in Figure 8.1.\\nthe regions as a function of YearsandHits. The predicted salaries for these\\nthree groups are $1,000 ×e5.107=$165,174, $1,000 ×e5.999=$402,834, and\\n$1,000×e6.740=$845,346 respectively.\\nInkeepingwiththe treeanalogy,theregions R1,R2,andR3areknownas\\nterminal nodes orleavesof the tree. As is the case for Figure 8.1, decisionterminal\\nnode\\nleaftrees are typically drawn upside down , in the sense that the leaves are at\\nthe bottom of the tree. The points along the tree where the predictor space\\nis split are referred to as internal nodes . In Figure 8.1, the two internalinternal\\nnodenodes are indicated by the text Years<4.5 andHits<117.5 . We refer to the\\nsegments of the trees that connect the nodes as branches.branchWe might interpret the regression tree displayed in Figure 8.1as follows:\\nYearsis the most important factor in determining Salary, and players with\\nless experience earn lower salaries than more experienced players. Given\\nthat a player is less experienced, the number of hits that he made in the\\nprevious year seems to play little role in his salary. But among players who\\nhave been in the major leagues for five or more years, the number of hits\\nmade in the previous year does affect salary, and players who made more\\nhits last year tend to have higher salaries. The regression tree shown in\\nFigure8.1is likely an over-simplification of the true relationship between\\nHits,Years, andSalary. However, it has advantages over other types of\\nregression models (such as those seen in Chapters 3and6): it is easier to\\ninterpret, and has a nice graphical representation.\\nPrediction via Stratification of the Feature Space\\nWe now discuss the process of building a regression tree. Roughly speaking,\\nthere are two steps.\\n1. We divide the predictor space — that is, the set of possible values\\nforX1,X2,...,X p— intoJdistinct and non-overlapping regions,\\nR1,R2,...,R J.', metadata={'source': './documents/ISLP_website.pdf', 'page': 339}), Document(page_content='334 8. Tree-Based Methods\\n2. For every observation that falls into the region Rj, we make the same\\nprediction, which is simply the mean of the response values for the\\ntraining observations in Rj.\\nFor instance, suppose that in Step 1 we obtain two regions, R1andR2,\\nand that the response mean of the training observations in the first region\\nis10, while the response mean of the training observations in the second\\nregion is 20. Then for a given observation X=x, ifx∈R1we will predict\\na value of 10, and ifx∈R2we will predict a value of 20.\\nWe now elaborate on Step 1 above. How do we construct the regions\\nR1,...,R J? In theory, the regions could have any shape. However, we\\nchoose to divide the predictor space into high-dimensional rectangles, or\\nboxes, for simplicity and for ease of interpretation of the resulting predic-\\ntive model. The goal is to find boxes R1,...,R Jthat minimize the RSS,\\ngiven by\\nJ∑\\nj=1∑\\ni∈Rj(yi−ˆyRj)2, (8.1)\\nwhereˆyRjis the mean response for the training observations within the\\njth box. Unfortunately, it is computationally infeasible to consider every\\npossible partition of the feature space into Jboxes. For this reason, we take\\natop-down ,greedyapproachthatisknownas recursive binary splitting .Therecursive\\nbinary\\nsplittingapproachis top-down becauseitbeginsatthetopofthetree(atwhichpoint\\nall observations belong to a single region) and then successively splits the\\npredictor space; each split is indicated via two new branches further down\\non the tree. It is greedybecause at each step of the tree-building process,\\nthebestsplit is made at that particular step, rather than looking ahead\\nand picking a split that will lead to a better tree in some future step.\\nIn order to perform recursive binary splitting, we first select the pre-\\ndictorXjand the cutpoint ssuch that splitting the predictor space into\\nthe regions {X|Xj<s}and{X|Xj≥s}leads to the greatest possible\\nreduction in RSS. (The notation {X|Xj<s}meansthe region of predictor\\nspace in which Xjtakes on a value less than s.) That is, we consider all\\npredictors X1,...,X p, and all possible values of the cutpoint sfor each of\\nthe predictors, and then choose the predictor and cutpoint such that the\\nresulting tree has the lowest RSS. In greater detail, for any jands,w e\\ndefine the pair of half-planes\\nR1(j, s)={X|Xj<s}andR2(j, s)={X|Xj≥s}, (8.2)\\nand we seek the value of jandsthat minimize the equation\\n∑\\ni:xi∈R1(j,s)(yi−ˆyR1)2+∑\\ni:xi∈R2(j,s)(yi−ˆyR2)2, (8.3)\\nwhereˆyR1is the mean response for the training observations in R1(j, s),\\nandˆyR2is the mean response for the training observations in R2(j, s).\\nFinding the values of jandsthat minimize ( 8.3) can be done quite quickly,\\nespecially when the number of features pis not too large.\\nNext, we repeat the process, looking for the best predictor and best\\ncutpoint in order to split the data further so as to minimize the RSS within', metadata={'source': './documents/ISLP_website.pdf', 'page': 340}), Document(page_content='8.1 The Basics of Decision Trees 335\\n|t1t2t3t4\\nR1R1\\nR2R2\\nR3R3\\nR4R4\\nR5R5\\nX1X1X1\\nX2X2X2\\nX1≤t1X2≤t2X1≤t3X2≤t4FIGURE 8.3. Top Left: A partition of two-dimensional feature space that could\\nnot result from recursive binary splitting. Top Right: The output of recursive\\nbinary splitting on a two-dimensional example. Bottom Left: A tree corresponding\\nto the partition in the top right panel. Bottom Right: A perspective plot of the\\nprediction surface corresponding to that tree.\\neach of the resulting regions. However, this time, instead of splitting the\\nentire predictor space, we split one of the two previously identified regions.\\nWenowhavethreeregions.Again,welooktosplitoneofthesethreeregions\\nfurther, so as to minimize the RSS. The process continues until a stopping\\ncriterion is reached; for instance, we may continue until no region contains\\nmore than five observations.\\nOnce the regions R1,...,R Jhave been created, we predict the response\\nfor a given test observation using the mean of the training observations in\\nthe region to which that test observation belongs.\\nA five-region example of this approach is shown in Figure 8.3.\\nTree Pruning\\nThe process described above may produce good predictions on the training\\nset, but is likely to overfit the data, leading to poor test set performance.\\nThis is because the resulting tree might be too complex. A smaller tree', metadata={'source': './documents/ISLP_website.pdf', 'page': 341}), Document(page_content='336 8. Tree-Based Methods\\nwith fewer splits (that is, fewer regions R1,...,R J) might lead to lower\\nvariance and better interpretation at the cost of a little bias. One possible\\nalternative to the process described above is to build the tree only so long\\nas the decrease in the RSS due to each split exceeds some (high) threshold.\\nThis strategy will result in smaller trees, but is too short-sighted since a\\nseemingly worthless split early on in the tree might be followed by a very\\ngood split—that is, a split that leads to a large reduction in RSS later on.\\nTherefore, a better strategy is to grow a very large tree T0, and then\\npruneit back in order to obtain a subtree. How do we determine the bestprune\\nsubtreeway to prune the tree? Intuitively, our goal is to select a subtree that\\nleads to the lowest test error rate. Given a subtree, we can estimate its\\ntest error using cross-validation or the validation set approach. However,\\nestimatingthecross-validationerrorforeverypossiblesubtreewouldbetoo\\ncumbersome, since there is an extremely large number of possible subtrees.\\nInstead, we need a way to select a small set of subtrees for consideration.\\nCost complexity pruning —also known as weakest link pruning —gives uscost\\ncomplexity\\npruning\\nweakest link\\npruninga way to do just this. Rather than considering every possible subtree, we\\nconsider a sequence of trees indexed by a nonnegative tuning parameter α.\\nFor each value of αthere corresponds a subtree T⊂T0such that\\n|T|∑\\nm=1∑\\ni:xi∈Rm(yi−ˆyRm)2+α|T| (8.4)\\nis as small as possible. Here |T|indicates the number of terminal nodes\\nof the tree T,Rmis the rectangle (i.e. the subset of predictor space) cor-\\nresponding to the mth terminal node, and ˆyRmis the predicted response\\nassociated with Rm—that is, the mean of the training observations in Rm.\\nThe tuning parameter αcontrols a trade-off between the subtree’s com-\\nplexity and its fit to the training data. When α=0, then the subtree T\\nwill simply equal T0, because then ( 8.4) just measures the training error.\\nHowever, as αincreases, there is a price to pay for having a tree with\\nmany terminal nodes, and so the quantity ( 8.4) will tend to be minimized\\nfor a smaller subtree. Equation 8.4is reminiscent of the lasso ( 6.7) from\\nChapter 6, in which a similar formulation was used in order to control the\\ncomplexity of a linear model.\\nIt turns out that as we increase αfrom zero in ( 8.4), branches get pruned\\nfrom the tree in a nested and predictable fashion, so obtaining the whole\\nsequence of subtrees as a function of αis easy. We can select a value of\\nαusing a validation set or using cross-validation. We then return to the\\nfull data set and obtain the subtree corresponding to α. This process is\\nsummarized in Algorithm 8.1.\\nFigures8.4and8.5display the results of fitting and pruning a regression\\ntree on the Hittersdata, using nine of the features. First, we randomly\\ndivided the data set in half, yielding 132 observations in the training set\\nand 131 observations in the test set. We then built a large regression tree\\non the training data and varied αin (8.4) in order to create subtrees with\\ndifferent numbers of terminal nodes. Finally, we performed six-fold cross-\\nvalidation in order to estimate the cross-validated MSE of the trees as', metadata={'source': './documents/ISLP_website.pdf', 'page': 342}), Document(page_content='8.1 The Basics of Decision Trees 337\\nAlgorithm 8.1 Building a Regression Tree\\n1. Use recursive binary splitting to grow a large tree on the training\\ndata, stopping only when each terminal node has fewer than some\\nminimum number of observations.\\n2. Apply cost complexity pruning to the large tree in order to obtain a\\nsequence of best subtrees, as a function of α.\\n3. Use K-fold cross-validation to choose α. That is, divide the training\\nobservations into Kfolds. For each k=1,...,K :\\n(a) Repeat Steps 1 and 2 on all but the kth fold of the training data.\\n(b) Evaluate the mean squared prediction error on the data in the\\nleft-outkth fold, as a function of α.\\nAverage the results for each value of α, and pick αto minimize the\\naverage error.\\n4. Return the subtree from Step 2 that corresponds to the chosen value\\nofα.\\na function of α. (We chose to perform six-fold cross-validation because\\n132 is an exact multiple of six.) The unpruned regression tree is shown\\nin Figure 8.4. The green curve in Figure 8.5shows the CV error as a\\nfunction of the number of leaves,2while the orange curve indicates the\\ntest error. Also shown are standard error bars around the estimated errors.\\nFor reference, the training error curve is shown in black. The CV error\\nis a reasonable approximation of the test error: the CV error takes on its\\nminimum for a three-node tree, while the test error also dips down at the\\nthree-node tree (though it takes on its lowest value at the ten-node tree).\\nThe pruned tree containing three terminal nodes is shown in Figure 8.1.\\n8.1.2 Classification Trees\\nAclassification tree is very similar to a regression tree, except that it isclassification\\ntreeused to predict a qualitative response rather than a quantitative one. Re-\\ncall that for a regression tree, the predicted response for an observation is\\ngiven by the mean response of the training observations that belong to the\\nsame terminal node. In contrast, for a classification tree, we predict that\\neach observation belongs to the most commonly occurring class of training\\nobservations in the region to which it belongs. In interpreting the results of\\na classification tree, we are often interested not only in the class prediction\\ncorresponding to a particular terminal node region, but also in the class\\nproportions among the training observations that fall into that region.\\nThe task of growing a classification tree is quite similar to the task of\\ngrowing a regression tree. Just as in the regression setting, we use recursive\\n2Although CV error is computed as a function of α, it is convenient to display the\\nresult as a function of |T|, the number of leaves; this is based on the relationship between\\nαand|T|in the original tree grown to all the training data.', metadata={'source': './documents/ISLP_website.pdf', 'page': 343}), Document(page_content='338 8. Tree-Based Methods|Years < 4.5\\nRBI < 60.5Putouts < 82Years < 3.5Years < 3.5Hits < 117.5\\nWalks < 43.5Runs < 47.5Walks < 52.5RBI < 80.5Years < 6.55.4874.622 5.1835.394 6.1896.015 5.5716.4076.5496.459 7.0077.289FIGURE 8.4.Regression tree analysis for theHittersdata. The unpruned treethat results from top-down greedy splitting on the training data is shown.binary splitting to grow a classification tree. However, in the classificationsetting, RSS cannot be used as a criterion for making the binary splits.A natural alternative to RSS is theclassification error rate. Since we planclassificationerror rateto assign an observation in a given region to themost commonly occurringclassof training observations in that region, the classification error rate issimply the fraction of the training observations in that region that do notbelong to the most common class:E=1−maxk(ˆpmk).(8.5)Hereˆpmkrepresents the proportion of training observations in themthregion that are from thekth class. However, it turns out that classificationerror is not sufficiently sensitive for tree-growing, and in practice two othermeasures are preferable.TheGini indexis defined byGini indexG=K∑k=1ˆpmk(1−ˆpmk),(8.6)a measure of total variance across theKclasses. It is not hard to seethat the Gini index takes on a small value if all of theˆpmk’s are close tozero or one. For this reason the Gini index is referred to as a measure of', metadata={'source': './documents/ISLP_website.pdf', 'page': 344}), Document(page_content='8.1 The Basics of Decision Trees 339\\n2468100.0 0.2 0.4 0.6 0.8 1.0Tree SizeMean Squared ErrorTrainingCross−ValidationTest\\nFIGURE 8.5. Regression tree analysis for the Hitters data. The training,\\ncross-validation, and test MSE are shown as a function of the number of terminal\\nnodes in the pruned tree. Standard error bands are displayed. The minimum\\ncross-validation error occurs at a tree size of three.\\nnodepurity—a small value indicates that a node contains predominantly\\nobservations from a single class.\\nAn alternative to the Gini index is entropy, given byentropy\\nD=−K∑\\nk=1ˆpmklog ˆpmk. (8.7)\\nSince0≤ˆpmk≤1, it follows that 0≤−ˆpmklog ˆpmk. One can show that\\nthe entropy will take on a value near zero if the ˆpmk’s are all near zero or\\nnear one. Therefore, like the Gini index, the entropy will take on a small\\nvalue if the mth node is pure. In fact, it turns out that the Gini index and\\nthe entropy are quite similar numerically.\\nWhen building a classification tree, either the Gini index or the entropy\\nare typically used to evaluate the quality of a particular split, since these\\ntwo approaches are more sensitive to node purity than is the classification\\nerror rate. Any of these three approaches might be used when pruningthe\\ntree, but the classification error rate is preferable if prediction accuracy of\\nthe final pruned tree is the goal.\\nFigure8.6shows an example on the Heartdata set. These data con-\\ntain a binary outcome HDfor 303 patients who presented with chest pain.\\nAn outcome value of Yesindicates the presence of heart disease based on\\nan angiographic test, while Nomeans no heart disease. There are 13 predic-\\ntors including Age,Sex,Chol(a cholesterol measurement), and other heart\\nand lung function measurements. Cross-validation results in a tree with six\\nterminal nodes.\\nIn our discussion thus far, we have assumed that the predictor vari-\\nables take on continuous values. However, decision trees can be constructed\\neven in the presence of qualitative predictor variables. For instance, in the\\nHeartdata, some of the predictors, such as Sex,Thal(Thallium stress test),', metadata={'source': './documents/ISLP_website.pdf', 'page': 345}), Document(page_content='340 8. Tree-Based Methods|Thal:a\\nCa < 0.5MaxHR < 161.5RestBP < 157Chol < 244MaxHR < 156MaxHR < 145.5ChestPain:bcChol < 244 Sex < 0.5Ca < 0.5Slope < 1.5Age < 52Thal:bChestPain:aOldpeak < 1.1RestECG < 1No YesNoNoYesNoNo NoNo YesYes NoNoNoYesYes YesYes\\n510150.0 0.1 0.2 0.3 0.4 0.5 0.6Tree SizeErrorTrainingCross−ValidationTest|Thal:a\\nCa < 0.5MaxHR < 161.5ChestPain:bcCa < 0.5\\nNo NoNo YesYes YesFIGURE 8.6. Heartdata.Top:The unpruned tree. Bottom Left: Cross-valida-\\ntion error, training, and test error, for different sizes of the pruned tree. Bottom\\nRight:The pruned tree corresponding to the minimal cross-validation error.\\nandChestPain , are qualitative. Therefore, a split on one of these variables\\namounts to assigning some of the qualitative values to one branch and\\nassigning the remaining to the other branch. In Figure 8.6, some of the in-\\nternal nodes correspond to splitting qualitative variables. For instance, the\\ntop internal node corresponds to splitting Thal. The text Thal:aindicates\\nthat the left-hand branch coming out of that node consists of observations\\nwith the first value of the Thalvariable (normal), and the right-hand node\\nconsists of the remaining observations (fixed or reversible defects). The text\\nChestPain:bc twosplitsdownthetreeontheleftindicatesthattheleft-hand\\nbranch coming out of that node consists of observations with the second\\nand third values of the ChestPain variable, where the possible values are\\ntypical angina, atypical angina, non-anginal pain, and asymptomatic.\\nFigure8.6has a surprising characteristic: some of the splits yield two\\nterminal nodes that have the same predicted value . For instance, consider\\nthe split RestECG<1 near the bottom right of the unpruned tree. Regardless\\nof the value of RestECG, a response value of Yesis predicted for those ob-', metadata={'source': './documents/ISLP_website.pdf', 'page': 346}), Document(page_content='8.1 The Basics of Decision Trees 341\\nservations. Why, then, is the split performed at all? The split is performed\\nbecause it leads to increased node purity . That is, all 9 of the observations\\ncorresponding to the right-hand leaf have a response value of Yes, whereas\\n7/11of those corresponding to the left-hand leaf have a response value of\\nYes. Why is node purity important? Suppose that we have a test obser-\\nvation that belongs to the region given by that right-hand leaf. Then we\\ncan be pretty certain that its response value is Yes. In contrast, if a test\\nobservation belongs to the region given by the left-hand leaf, then its re-\\nsponse value is probably Yes, but we are much less certain. Even though\\nthe split RestECG<1 does not reduce the classification error, it improves the\\nGini index and the entropy, which are more sensitive to node purity.\\n8.1.3 Trees Versus Linear Models\\nRegressionandclassificationtreeshaveaverydifferentflavorfromthemore\\nclassicalapproachesforregressionandclassificationpresentedinChapters 3\\nand4. In particular, linear regression assumes a model of the form\\nf(X)=β0+p∑\\nj=1Xjβj, (8.8)\\nwhereas regression trees assume a model of the form\\nf(X)=M∑\\nm=1cm·1(X∈Rm) (8.9)\\nwhereR1,...,R Mrepresent a partition of feature space, as in Figure 8.3.\\nWhich model is better? It depends on the problem at hand. If the re-\\nlationship between the features and the response is well approximated by\\na linear model as in ( 8.8), then an approach such as linear regression will\\nlikely work well, and will outperform a method such as a regression tree\\nthat does not exploit this linear structure. If instead there is a highly non-\\nlinear and complex relationship between the features and the response as\\nindicated by model ( 8.9), then decision trees may outperform classical ap-\\nproaches. An illustrative example is displayed in Figure 8.7. The relative\\nperformances of tree-based and classical approaches can be assessed by es-\\ntimating the test error, using either cross-validation or the validation set\\napproach (Chapter 5).\\nOf course, other considerations beyond simply test error may come into\\nplay in selecting a statistical learning method; for instance, in certain set-\\ntings, prediction using a tree may be preferred for the sake of interpretabil-\\nity and visualization.\\n8.1.4 Advantages and Disadvantages of Trees\\nDecision trees for regression and classification have a number of advantages\\nover the more classical approaches seen in Chapters 3and4:\\nLTrees are very easy to explain to people. In fact, they are even easier\\nto explain than linear regression!', metadata={'source': './documents/ISLP_website.pdf', 'page': 347}), Document(page_content='342 8. Tree-Based Methods\\n−2 −1 0 1 2−2 −1 0 1 2X1X2\\n−2 −1 0 1 2−2 −1 0 1 2X1X2\\n−2 −1 0 1 2−2 −1 0 1 2X1X2\\n−2 −1 0 1 2−2 −1 0 1 2X1X2\\nFIGURE 8.7.Top Row:A two-dimensional classification example in which thetrue decision boundary is linear, and is indicated by the shaded regions. A classicalapproach that assumes a linear boundary (left) will outperform a decision treethat performs splits parallel to the axes (right).Bottom Row:Here the true de-cision boundary is non-linear. Here a linear model is unable to capture the truedecision boundary (left), whereas a decision tree is successful (right).LSome people believe that decision trees more closely mirror humandecision-making than do the regression and classification approachesseen in previous chapters.LTrees can be displayed graphically, and are easily interpreted even bya non-expert (especially if they are small).LTrees can easily handle qualitative predictors without the need tocreate dummy variables.MUnfortunately, trees generally do not have the same level of predictiveaccuracy as some of the other regression and classification approachesseen in this book.MAdditionally, trees can be very non-robust. In other words, a smallchange in the data can cause a large change in the final estimatedtree.However,byaggregatingmanydecisiontrees,usingmethodslikebagging,random forests, andboosting, the predictive performance of trees can besubstantially improved. We introduce these concepts in the next section.', metadata={'source': './documents/ISLP_website.pdf', 'page': 348}), Document(page_content='8.2Bagging,RandomForests,Boosting,andBayesianAdditiveRegressionTrees 343\\n8.2 Bagging, Random Forests, Boosting, and\\nBayesian Additive Regression Trees\\nAnensemble method is an approach that combines many simple “buildingensembleblock” models in order to obtain a single and potentially very powerful\\nmodel. These simple building block models are sometimes known as weak\\nlearners, since they may lead to mediocre predictions on their own.weak\\nlearnersWe will now discuss bagging, random forests, boosting, and Bayesian\\nadditive regression trees. These are ensemble methods for which the simple\\nbuilding block is a regression or a classification tree.\\n8.2.1 Bagging\\nThe bootstrap, introduced in Chapter 5, is an extremely powerful idea. It is\\nused in many situations in which it is hard or even impossible to directly\\ncompute the standard deviation of a quantity of interest. We see here that\\nthe bootstrap can be used in a completely different context, in order to\\nimprove statistical learning methods such as decision trees.\\nThe decision trees discussed in Section 8.1suffer from high variance .\\nThis means that if we split the training data into two parts at random,\\nand fit a decision tree to both halves, the results that we get could be\\nquite different. In contrast, a procedure with low variance will yield similar\\nresults if applied repeatedly to distinct data sets; linear regression tends\\nto have low variance, if the ratio of ntopis moderately large. Bootstrap\\naggregation , orbagging, is a general-purpose procedure for reducing thebaggingvariance of a statistical learning method; we introduce it here because it is\\nparticularly useful and frequently used in the context of decision trees.\\nRecall that given a set of nindependent observations Z1,...,Z n, each\\nwith variance σ2, the variance of the mean ¯Zof the observations is given\\nbyσ2/n. In other words, averaging a set of observations reduces variance .\\nHence a natural way to reduce the variance and increase the test set ac-\\ncuracy of a statistical learning method is to take many training sets from\\nthe population, build a separate prediction model using each training set,\\nand average the resulting predictions. In other words, we could calculate\\nˆf1(x),ˆf2(x),..., ˆfB(x)usingBseparate training sets, and average them\\nin order to obtain a single low-variance statistical learning model, given by\\nˆfavg(x)=1\\nBB∑\\nb=1ˆfb(x).\\nOf course, this is not practical because we generally do not have access\\nto multiple training sets. Instead, we can bootstrap, by taking repeated\\nsamples from the (single) training data set. In this approach we generate\\nBdifferent bootstrapped training data sets. We then train our method on\\nthebth bootstrapped training set in order to get ˆf∗b(x), and finally average\\nall the predictions, to obtain\\nˆfbag(x)=1\\nBB∑\\nb=1ˆf∗b(x).', metadata={'source': './documents/ISLP_website.pdf', 'page': 349}), Document(page_content='344 8. Tree-Based Methods\\n0 50 100 150 200 250 3000.100.150.200.250.30\\nNumber of TreesError\\nTest: BaggingTest: RandomForestOOB: BaggingOOB: RandomForestFIGURE 8.8. Bagging and random forest results for the Heartdata. The test\\nerror (black and orange) is shown as a function of B, the number of bootstrapped\\ntraining sets used. Random forests were applied with m=√p. The dashed line\\nindicates the test error resulting from a single classification tree. The green and\\nblue traces show the OOB error, which in this case is — by chance — considerably\\nlower.\\nThis is called bagging.\\nWhile bagging can improve predictions for many regression methods,\\nit is particularly useful for decision trees. To apply bagging to regression\\ntrees,wesimplyconstruct Bregressiontreesusing Bbootstrappedtraining\\nsets, and average the resulting predictions. These trees are grown deep,\\nand are not pruned. Hence each individual tree has high variance, but\\nlow bias. Averaging these Btrees reduces the variance. Bagging has been\\ndemonstrated to give impressive improvements in accuracy by combining\\ntogether hundreds or even thousands of trees into a single procedure.\\nThus far, we have described the bagging procedure in the regression\\ncontext,topredictaquantitativeoutcome Y.Howcanbaggingbeextended\\nto a classification problem where Yis qualitative? In that situation, there\\nareafewpossibleapproaches,butthesimplestisasfollows.Foragiventest\\nobservation, we can record the class predicted by each of the Btrees, and\\ntake amajority vote : the overall prediction is the most commonly occurringmajority\\nvoteclass among the Bpredictions.\\nFigure8.8shows the results from bagging trees on the Heartdata. The\\ntest error rate is shown as a function of B, the number of trees constructed\\nusing bootstrapped training data sets. We see that the bagging test error\\nrate is slightly lower in this case than the test error rate obtained from a\\nsingle tree. The number of trees Bis not a critical parameter with bagging;\\nusing a very large value of Bwill not lead to overfitting. In practice we', metadata={'source': './documents/ISLP_website.pdf', 'page': 350}), Document(page_content='8.2Bagging,RandomForests,Boosting,andBayesianAdditiveRegressionTrees 345\\nuse a value of Bsufficiently large that the error has settled down. Using\\nB= 100 is sufficient to achieve good performance in this example.\\nOut-of-Bag Error Estimation\\nIt turns out that there is a very straightforward way to estimate the test\\nerror of a bagged model, without the need to perform cross-validation or\\nthe validation set approach. Recall that the key to bagging is that trees are\\nrepeatedly fit to bootstrapped subsets of the observations. One can show\\nthat on average, each bagged tree makes use of around two-thirds of the\\nobservations.3The remaining one-third of the observations not used to fit a\\ngiven bagged tree are referred to as the out-of-bag (OOB) observations. Weout-of-bagcan predict the response for the ith observation using each of the trees in\\nwhich that observation was OOB. This will yield around B/3predictions\\nfor theith observation. In order to obtain a single prediction for the ith\\nobservation, we can average these predicted responses (if regression is the\\ngoal) or can take a majority vote (if classification is the goal). This leads\\nto a single OOB prediction for the ith observation. An OOB prediction\\ncan be obtained in this way for each of the nobservations, from which the\\noverall OOB MSE (for a regression problem) or classification error (for a\\nclassificationproblem)canbecomputed.TheresultingOOBerrorisavalid\\nestimate of the test error for the bagged model, since the response for each\\nobservation is predicted using only the trees that were not fit using that\\nobservation. Figure 8.8displays the OOB error on the Heartdata. It can\\nbe shown that with Bsufficiently large, OOB error is virtually equivalent\\nto leave-one-out cross-validation error. The OOB approach for estimating\\nthe test error is particularly convenient when performing bagging on large\\ndata sets for which cross-validation would be computationally onerous.\\nVariable Importance Measures\\nAs we have discussed, bagging typically results in improved accuracy over\\nprediction using a single tree. Unfortunately, however, it can be difficult to\\ninterpret the resulting model. Recall that one of the advantages of decision\\ntrees is the attractive and easily interpreted diagram that results, such as\\nthe one displayed in Figure 8.1. However, when we bag a large number of\\ntrees, it is no longer possible to represent the resulting statistical learning\\nprocedure using a single tree, and it is no longer clear which variables\\nare most important to the procedure. Thus, bagging improves prediction\\naccuracy at the expense of interpretability.\\nAlthoughthecollectionofbaggedtreesismuchmoredifficulttointerpret\\nthan a single tree, one can obtain an overall summary of the importance of\\neachpredictorusingtheRSS(forbaggingregressiontrees)ortheGiniindex\\n(for bagging classification trees). In the case of bagging regression trees, we\\ncan record the total amount that the RSS ( 8.1) is decreased due to splits\\nover a given predictor, averaged over all Btrees. A large value indicates\\nan important predictor. Similarly, in the context of bagging classification\\n3This relates to Exercise 2of Chapter 5.', metadata={'source': './documents/ISLP_website.pdf', 'page': 351}), Document(page_content='346 8. Tree-Based Methods\\nThalCaChestPainOldpeakMaxHRRestBPAgeCholSlopeSexExAngRestECGFbs\\n0 20 40 60 80 100Variable ImportanceFIGURE 8.9. A variable importance plot for the Heartdata. Variable impor-\\ntance is computed using the mean decrease in Gini index, and expressed relative\\nto the maximum.\\ntrees, we can add up the total amount that the Gini index ( 8.6) is decreased\\nby splits over a given predictor, averaged over all Btrees.\\nA graphical representation of the variable importances in theHeartdatavariable\\nimportanceisshowninFigure 8.9.WeseethemeandecreaseinGiniindexforeachvari-\\nable, relative to the largest. The variables with the largest mean decrease\\nin Gini index are Thal,Ca, andChestPain .\\n8.2.2 Random Forests\\nRandom forests provide an improvement over bagged trees by way of arandom\\nforestsmall tweak that decorrelates the trees. As in bagging, we build a number\\nofdecisiontreesonbootstrappedtrainingsamples.Butwhenbuildingthese\\ndecision trees, each time a split in a tree is considered, a random sample of\\nmpredictors is chosen as split candidates from the full set of ppredictors.\\nThe split is allowed to use only one of those mpredictors. A fresh sample of\\nmpredictors is taken at each split, and typically we choose m≈√p—that\\nis, the number of predictors considered at each split is approximately equal\\nto the square root of the total number of predictors (4 out of the 13 for the\\nHeartdata).\\nIn other words, in building a random forest, at each split in the tree,\\nthe algorithm is not even allowed to consider a majority of the available\\npredictors. This may sound crazy, but it has a clever rationale. Suppose\\nthat there is one very strong predictor in the data set, along with a num-\\nber of other moderately strong predictors. Then in the collection of bagged\\ntrees, most or all of the trees will use this strong predictor in the top split.\\nConsequently, all of the bagged trees will look quite similar to each other.', metadata={'source': './documents/ISLP_website.pdf', 'page': 352}), Document(page_content='8.2Bagging,RandomForests,Boosting,andBayesianAdditiveRegressionTrees 347\\nHence the predictions from the bagged trees will be highly correlated. Un-\\nfortunately, averaging many highly correlated quantities does not lead to\\nas large of a reduction in variance as averaging many uncorrelated quan-\\ntities. In particular, this means that bagging will not lead to a substantial\\nreduction in variance over a single tree in this setting.\\nRandom forests overcome this problem by forcing each split to consider\\nonly a subset of the predictors. Therefore, on average (p−m)/pof the\\nsplits will not even consider the strong predictor, and so other predictors\\nwill have more of a chance. We can think of this process as decorrelating\\nthe trees, thereby making the average of the resulting trees less variable\\nand hence more reliable.\\nThe main difference between bagging and random forests is the choice\\nof predictor subset size m. For instance, if a random forest is built using\\nm=p, then this amounts simply to bagging. On the Heartdata, random\\nforests using m=√pleads to a reduction in both test error and OOB error\\nover bagging (Figure 8.8).\\nUsing a small value of min building a random forest will typically be\\nhelpful when we have a large number of correlated predictors. We applied\\nrandom forests to a high-dimensional biological data set consisting of ex-\\npression measurements of 4,718 genes measured on tissue samples from 349\\npatients. There are around 20,000 genes in humans, and individual genes\\nhave different levels of activity, or expression, in particular cells, tissues,\\nand biological conditions. In this data set, each of the patient samples has\\na qualitative label with 15 different levels: either normal or 1 of 14 different\\ntypes of cancer. Our goal was to use random forests to predict cancer type\\nbased on the 500genes that have the largest variance in the training set.\\nWe randomly divided the observations into a training and a test set, and\\napplied random forests to the training set for three different values of the\\nnumber of splitting variables m. The results are shown in Figure 8.10. The\\nerror rate of a single tree is 45.7%, and the null rate is 75.4%.4We see that\\nusing400trees is sufficient to give good performance, and that the choice\\nm=√pgave a small improvement in test error over bagging ( m=p) in\\nthis example. As with bagging, random forests will not overfit if we increase\\nB, so in practice we use a value of Bsufficiently large for the error rate to\\nhave settled down.\\n8.2.3 Boosting\\nWe now discuss boosting, yet another approach for improving the predic-boostingtions resulting from a decision tree. Like bagging, boosting is a general\\napproach that can be applied to many statistical learning methods for re-\\ngression or classification. Here we restrict our discussion of boosting to the\\ncontext of decision trees.\\nRecallthatbagginginvolvescreatingmultiplecopiesoftheoriginaltrain-\\ning data set using the bootstrap, fitting a separate decision tree to each\\ncopy, and then combining all of the trees in order to create a single predic-\\n4The null rate results from simply classifying each observation to the dominant class\\noverall, which is in this case the normal class.', metadata={'source': './documents/ISLP_website.pdf', 'page': 353}), Document(page_content='348 8. Tree-Based Methods\\n0 100 200 300 400 5000.2 0.3 0.4 0.5Number of TreesTest Classification Errorm=pm=p/2m=p\\nFIGURE 8.10. Results from random forests for the 15-class gene expression\\ndata set with p= 500 predictors. The test error is displayed as a function of\\nthe number of trees. Each colored line corresponds to a different value of m, the\\nnumber of predictors available for splitting at each interior tree node. Random\\nforests ( m<p ) lead to a slight improvement over bagging ( m=p). A single\\nclassification tree has an error rate of 45.7%.\\ntive model. Notably, each tree is built on a bootstrap data set, independent\\nof the other trees. Boosting works in a similar way, except that the trees are\\ngrownsequentially : each tree is grown using information from previously\\ngrown trees. Boosting does not involve bootstrap sampling; instead each\\ntree is fit on a modified version of the original data set.\\nConsiderfirsttheregressionsetting.Likebagging,boostinginvolvescom-\\nbining a large number of decision trees, ˆf1,..., ˆfB. Boosting is described\\nin Algorithm 8.2.\\nWhat is the idea behind this procedure? Unlike fitting a single large deci-\\nsion tree to the data, which amounts to fitting the data hard and potentially\\noverfitting, the boosting approach instead learns slowly . Given the current\\nmodel, we fit a decision tree to the residuals from the model. That is, we\\nfit a tree using the current residuals, rather than the outcome Y, as the re-\\nsponse. We then add this new decision tree into the fitted function in order\\nto update the residuals. Each of these trees can be rather small, with just\\na few terminal nodes, determined by the parameter din the algorithm. By\\nfitting small trees to the residuals, we slowly improve ˆfin areas where it\\ndoes not perform well. The shrinkage parameter λslows the process down\\neven further, allowing more and different shaped trees to attack the resid-\\nuals. In general, statistical learning approaches that learn slowly tend to\\nperform well. Note that in boosting, unlike in bagging, the construction of\\neach tree depends strongly on the trees that have already been grown.\\nWe have just described the process of boosting regression trees. Boosting\\nclassification trees proceeds in a similar but slightly more complex way, and\\nthe details are omitted here.', metadata={'source': './documents/ISLP_website.pdf', 'page': 354}), Document(page_content='8.2Bagging,RandomForests,Boosting,andBayesianAdditiveRegressionTrees 349\\nAlgorithm 8.2 Boosting for Regression Trees\\n1. Setˆf(x)=0 andri=yifor alliin the training set.\\n2. Forb=1,2,...,B , repeat:\\n(a) Fit a tree ˆfbwithdsplits (d+1terminal nodes) to the training\\ndata(X,r).\\n(b) Update ˆfby adding in a shrunken version of the new tree:\\nˆf(x)←ˆf(x)+λˆfb(x). (8.10)\\n(c) Update the residuals,\\nri←ri−λˆfb(xi). (8.11)\\n3. Output the boosted model,\\nˆf(x)=B∑\\nb=1λˆfb(x). (8.12)\\nBoosting has three tuning parameters:\\n1. The number of trees B. Unlike bagging and random forests, boosting\\ncan overfit if Bis too large, although this overfitting tends to occur\\nslowly if at all. We use cross-validation to select B.\\n2. The shrinkage parameter λ, a small positive number. This controls\\ntherateatwhichboostinglearns.Typicalvaluesare 0.01or0.001,and\\nthe right choice can depend on the problem. Very small λcan require\\nusing a very large value of Bin order to achieve good performance.\\n3. The number dof splits in each tree, which controls the complexity\\nof the boosted ensemble. Often d=1works well, in which case each\\ntree is a stump, consisting of a single split. In this case, the boostedstumpensemble is fitting an additive model, since each term involves only a\\nsingle variable. More generally dis theinteraction depth , and controlsinteraction\\ndepththe interaction order of the boosted model, since dsplits can involve\\nat mostdvariables.\\nInFigure 8.11,weappliedboostingtothe15-classcancergeneexpression\\ndata set, in order to develop a classifier that can distinguish the normal\\nclass from the 14 cancer classes. We display the test error as a function of\\nthe total number of trees and the interaction depth d. We see that simple\\nstumps with an interaction depth of one perform well if enough of them\\nare included. This model outperforms the depth-two model, and both out-\\nperform a random forest. This highlights one difference between boosting\\nand random forests: in boosting, because the growth of a particular tree\\ntakes into account the other trees that have already been grown, smaller', metadata={'source': './documents/ISLP_website.pdf', 'page': 355}), Document(page_content='350 8. Tree-Based Methods\\n0 1000 2000 3000 4000 50000.05 0.10 0.15 0.20 0.25Number of TreesTest Classification ErrorBoosting: depth=1Boosting: depth=2RandomForest: m=p\\nFIGURE 8.11.Results from performing boosting and random forests on the15-class gene expressiondata set in order to predictcancerversusnormal. The testerror is displayed as a function of the number of trees. For the two boosted models,λ=0.01. Depth-1 trees slightly outperform depth-2 trees, and both outperformthe random forest, although the standard errors are around 0.02, making none ofthese differences significant. The test error rate for a single tree is 24%.trees are typically sufficient. Using smaller trees can aid in interpretabilityas well; for instance, using stumps leads to an additive model.8.2.4 Bayesian Additive Regression TreesFinally, we discussBayesian additive regression trees(BART), another en-Bayesianadditiveregressiontreessemblemethodthatusesdecisiontreesasitsbuildingblocks.Forsimplicity,we present BART for regression (as opposed to classification).Recall that bagging and random forests make predictions from an aver-age of regression trees, each of which is built using a random sample of dataand/or predictors. Each tree is built separately from the others. By con-trast, boosting uses a weighted sum of trees, each of which is constructedby fitting a tree to the residual of the current fit. Thus, each new tree at-tempts to capture signal that is not yet accounted for by the current setof trees. BART is related to both approaches: each tree is constructed ina random manner as in bagging and random forests, and each tree tries tocapture signal not yet accounted for by the current model, as in boosting.The main novelty in BART is the way in which new trees are generated.Before we introduce the BART algorithm, we define some notation. WeletKdenote the number of regression trees, andBthe number of iterationsfor which the BART algorithm will be run. The notationˆfbk(x)representsthe prediction atxfor thekth regression tree used in thebth iteration. Atthe end of each iteration, theKtrees from that iteration will be summed,i.e.ˆfb(x)=∑Kk=1ˆfbk(x)forb=1,...,B.In the first iteration of the BART algorithm, all trees are initialized tohaveasinglerootnode,withˆf1k(x)=1nK∑ni=1yi,themeanoftheresponse', metadata={'source': './documents/ISLP_website.pdf', 'page': 356}), Document(page_content='8.2Bagging,RandomForests,Boosting,andBayesianAdditiveRegressionTrees 351\\n(a):ˆfb−1\\nk(X) (b): Possibility #1 for ˆfb\\nk(X)|X < 169.17X < 114.305X < 140.35−0.5031 0.2667 −0.2470 0.4079|X < 169.17X < 114.305X < 140.35−0.5110 0.2693 −0.2649 0.4221(c): Possibility #2 for ˆfb\\nk(X) (d): Possibility #3 for ˆfb\\nk(X)|X < 169.17\\n−0.1218 0.4079|X < 169.17X < 114.305X < 106.755X < 140.35−0.05089 −1.03100 0.26670 −0.24700 0.40790FIGURE 8.12. A schematic of perturbed trees from the BART algorithm. (a):\\nThekth tree at the (b−1)st iteration, ˆfb−1\\nk(X), is displayed. Panels (b)–(d)\\ndisplay three of many possibilities for ˆfb\\nk(X), given the form of ˆfb−1\\nk(X).(b):One\\npossibility is that ˆfb\\nk(X)has the same structure as ˆfb−1\\nk(X), but with different\\npredictions at the terminal nodes. (c):Another possibility is that ˆfb\\nk(X)results\\nfrom pruning ˆfb−1\\nk(X).(d):Alternatively, ˆfb\\nk(X)may have more terminal nodes\\nthanˆfb−1\\nk(X).\\nvalues divided by the total number of trees. Thus, ˆf1(x)=∑K\\nk=1ˆf1\\nk(x)=\\n1\\nn∑n\\ni=1yi.\\nIn subsequent iterations, BART updates each of the Ktrees, one at a\\ntime. In the bth iteration, to update the kth tree, we subtract from each\\nresponse value the predictions from all but the kth tree, in order to obtain\\napartial residual\\nri=yi−∑\\nk′<kˆfb\\nk′(xi)−∑\\nk′>kˆfb−1\\nk′(xi)\\nfor theith observation, i=1,...,n . Rather than fitting a fresh tree to this\\npartial residual, BART randomly chooses a perturbation to the tree from\\nthe previous iteration ( ˆfb−1\\nk) from a set of possible perturbations, favoring\\nones that improve the fit to the partial residual. There are two components\\nto this perturbation:\\n1. Wemaychangethestructureofthetreebyaddingorpruningbranches.\\n2. We may change the prediction in each terminal node of the tree.\\nFigure8.12illustrates examples of possible perturbations to a tree.\\nThe output of BART is a collection of prediction models,\\nˆfb(x)=K∑\\nk=1ˆfb\\nk(x),forb=1,2,...,B.', metadata={'source': './documents/ISLP_website.pdf', 'page': 357}), Document(page_content='352 8. Tree-Based Methods\\nAlgorithm 8.3 Bayesian Additive Regression Trees\\n1. Letˆf1\\n1(x)=ˆf1\\n2(x)=···=ˆf1\\nK(x)=1\\nnK∑n\\ni=1yi.\\n2. Compute ˆf1(x)=∑K\\nk=1ˆf1\\nk(x)=1\\nn∑n\\ni=1yi.\\n3. Forb=2,...,B :\\n(a) Fork=1,2,...,K :\\ni. Fori=1,...,n , compute the current partial residual\\nri=yi−∑\\nk′<kˆfb\\nk′(xi)−∑\\nk′>kˆfb−1\\nk′(xi).\\nii. Fit a new tree, ˆfb\\nk(x), tori, by randomly perturbing the\\nkth tree from the previous iteration, ˆfb−1\\nk(x). Perturbations\\nthat improve the fit are favored.\\n(b) Compute ˆfb(x)=∑K\\nk=1ˆfb\\nk(x).\\n4. Compute the mean after Lburn-in samples,\\nˆf(x)=1\\nB−LB∑\\nb=L+1ˆfb(x).\\nWe typically throw away the first few of these prediction models, since\\nmodels obtained in the earlier iterations — known as the burn-inperiodburn-in— tend not to provide very good results. We can let Ldenote the num-\\nber of burn-in iterations; for instance, we might take L= 200 . Then, to\\nobtain a single prediction, we simply take the average after the burn-in\\niterations, ˆf(x)=1\\nB−L∑B\\nb=L+1ˆfb(x). However, it is also possible to com-\\npute quantities other than the average: for instance, the percentiles of\\nˆfL+1(x),..., ˆfB(x)provide a measure of uncertainty in the final predic-\\ntion. The overall BART procedure is summarized in Algorithm 8.3.\\nA key element of the BART approach is that in Step 3(a)ii., we do notfit\\na fresh tree to the current partial residual: instead, we try to improve the fit\\nto the current partial residual by slightly modifying the tree obtained in the\\nprevious iteration (see Figure 8.12). Roughly speaking, this guards against\\noverfitting since it limits how “hard” we fit the data in each iteration.\\nFurthermore, the individual trees are typically quite small. We limit the\\ntree size in order to avoid overfitting the data, which would be more likely\\nto occur if we grew very large trees.\\nFigure8.13shows the result of applying BART to the Heartdata, using\\nK= 200 trees, as the number of iterations is increased to 10,000. During\\nthe initial iterations, the test and training errors jump around a bit. After\\nthis initial burn-in period, the error rates settle down. We note that there\\nis only a small difference between the training error and the test error,\\nindicating that the tree perturbation process largely avoids overfitting.', metadata={'source': './documents/ISLP_website.pdf', 'page': 358}), Document(page_content='8.2Bagging,RandomForests,Boosting,andBayesianAdditiveRegressionTrees 353\\n5 10 50 100 50050000.0 0.1 0.2 0.3 0.4 0.5Number of IterationsErrorBART Training ErrorBART Test ErrorBoosting Training ErrorBoosting Test Error\\nFIGURE 8.13. BART and boosting results for the Heartdata. Both training\\nand test errors are displayed. After a burn-in period of 100iterations (shown in\\ngray), the error rates for BART settle down. Boosting begins to overfit after a\\nfew hundred iterations.\\nThetrainingandtesterrorsforboostingarealsodisplayedinFigure 8.13.\\nWe see that the test error for boosting approaches that of BART, but then\\nbegins to increase as the number of iterations increases. Furthermore, the\\ntraining error for boosting decreases as the number of iterations increases,\\nindicating that boosting has overfit the data.\\nThough the details are outside of the scope of this book, it turns out\\nthat the BART method can be viewed as a Bayesian approach to fitting an\\nensemble of trees: each time we randomly perturb a tree in order to fit the\\nresiduals, we are in fact drawing a new tree from a posterior distribution.\\n(Of course, this Bayesian connection is the motivation for BART’s name.)\\nFurthermore, Algorithm 8.3can be viewed as a Markov chain Monte CarloMarkov\\nchain Monte\\nCarloalgorithm for fitting the BART model.\\nWhenweapplyBART,wemustselectthenumberoftrees K,thenumber\\nofiterations B,andthenumberofburn-initerations L.Wetypicallychoose\\nlargevaluesfor BandK,andamoderatevaluefor L:forinstance, K= 200 ,\\nB=1,000, andL= 100 is a reasonable choice. BART has been shown to\\nhave very impressive out-of-box performance — that is, it performs well\\nwith minimal tuning.\\n8.2.5 Summary of Tree Ensemble Methods\\nTrees are an attractive choice of weak learner for an ensemble method\\nfor a number of reasons, including their flexibility and ability to handle', metadata={'source': './documents/ISLP_website.pdf', 'page': 359}), Document(page_content='354 8. Tree-Based Methodspredictors of mixed types (i.e. qualitative as well as quantitative). We havenow seen four approaches for fitting an ensemble of trees: bagging, randomforests, boosting, and BART.• Inbagging, the trees are grown independently on random samples ofthe observations. Consequently, the trees tend to be quite similar toeach other. Thus, bagging can get caught in local optima and can failto thoroughly explore the model space.• Inrandom forests, the trees are once again grown independently onrandom samples of the observations. However, each split on each treeis performed using a random subset of the features, thereby decorre-lating the trees, and leading to a more thorough exploration of modelspace relative to bagging.• Inboosting, we only use the original data, and do not draw any ran-dom samples. The trees are grown successively, using a “slow” learn-ing approach: each new tree is fit to the signal that is left over fromthe earlier trees, and shrunken down before it is used.• InBART, we once again only make use of the original data, and wegrow the trees successively. However, each tree is perturbed in orderto avoid local minima and achieve a more thorough exploration ofthe model space.8.3 Lab: Tree-Based MethodsWe import some of our usual libraries at this top level.In[1]:import numpy as npimport pandas as pdfrom matplotlib.pyplot import subplotsfrom statsmodels.datasets import get_rdatasetimport sklearn.model_selection as skmfrom ISLP import load_data, confusion_tablefrom ISLP.models import ModelSpec as MSWe also collect the new imports needed for this lab.In[2]:from sklearn.tree import (DecisionTreeClassifier as DTC,DecisionTreeRegressor as DTR,plot_tree,export_text)from sklearn.metrics import (accuracy_score,log_loss)from sklearn.ensemble import \\\\(RandomForestRegressor as RF,GradientBoostingRegressor as GBR)from ISLP.bart import BART', metadata={'source': './documents/ISLP_website.pdf', 'page': 360}), Document(page_content='8.3 Lab: Tree-Based Methods 355\\n8.3.1 Fitting Classification Trees\\nWe first use classification trees to analyze the Carseats data set. In these\\ndata,Salesis a continuous variable, and so we begin by recoding it as a\\nbinary variable. We use the where()function to create a variable, calledwhere()High, which takes on a value of Yesif theSalesvariable exceeds 8, and\\ntakes on a value of Nootherwise.\\nIn[3]: Carseats = load_data( \\'Carseats\\' )\\nHigh = np.where(Carseats.Sales > 8,\\n\"Yes\",\\n\"No\")\\nWe now use DecisionTreeClassifier() to fit a classification tree in orderDecisionTree\\nClassifier() to predict Highusing all variables but Sales. To do so, we must form a\\nmodel matrix as we did when fitting regression models.\\nIn[4]: model = MS(Carseats.columns.drop( \\'Sales\\' ), intercept=False)\\nD=m o d e l . f i t _ t r a n s f o r m ( C a r s e a t s )\\nfeature_names = list(D.columns)\\nX=n p . a s a r r a y ( D )\\nWe have converted Dfrom a data frame to an array X, which is needed in\\nsome of the analysis below. We also need the feature_names for annotating\\nour plots later.\\nThereareseveraloptionsneededtospecifytheclassifier,suchas max_depth\\n(how deep to grow the tree), min_samples_split (minimum number of ob-\\nservations in a node to be eligible for splitting) and criterion (whether to\\nuse Gini or cross-entropy as the split criterion). We also set random_state\\nfor reproducibility; ties in the split criterion are broken at random.\\nIn[5]: clf = DTC(criterion= \\'entropy\\' ,\\nmax_depth=3,\\nrandom_state=0)\\nclf.fit(X, High)\\nOut[5]: DecisionTreeClassifier(criterion= \\'entropy\\' ,m a x _ d e p t h = 3 )\\nIn our discussion of qualitative features in Section 3.3, we noted that for\\na linear regression model such a feature could be represented by including a\\nmatrix of dummy variables (one-hot-encoding) in the model matrix, using\\nthe formula notation of statsmodels . As mentioned in Section 8.1, there is\\na more natural way to handle qualitative features when building a decision\\ntree, that does not require such dummy variables; each split amounts to\\npartitioning the levels into two groups. However, the sklearnimplementa-\\ntion of decision trees does not take advantage of this approach; instead it\\nsimply treats the one-hot-encoded levels as separate variables.\\nIn[6]: accuracy_score(High, clf.predict(X))\\nOut[6]: 0.7275\\nWith only the default arguments, the training error rate is 21%. For clas-\\nsification trees, we can access the value of the deviance using log_loss() ,log_loss()', metadata={'source': './documents/ISLP_website.pdf', 'page': 361}), Document(page_content='356 8. Tree-Based Methods\\n−2∑\\nm∑\\nknmklog ˆpmk,\\nwherenmkis the number of observations in the mth terminal node that\\nbelong to the kth class.\\nIn[7]: resid_dev = np.sum(log_loss(High, clf.predict_proba(X)))\\nresid_dev\\nOut[7]: 0.4711\\nThis is closely related to the entropy, defined in ( 8.7). A small deviance\\nindicates a tree that provides a good fit to the (training) data.\\nOne of the most attractive properties of trees is that they can be graphi-\\ncallydisplayed.Hereweusethe plot()functiontodisplaythetreestructure\\n(not shown here).\\nIn[8]: ax = subplots(figsize=(12,12))[1]\\nplot_tree(clf,\\nfeature_names=feature_names,\\nax=ax);\\nThe most important indicator of Salesappears to be ShelveLoc .\\nWe can see a text representation of the tree using export_text() , whichexport_text()displays the split criterion (e.g. Price <= 92.5 ) for each branch. For leaf\\nnodesitshowstheoverallprediction( YesorNo).Wecanalsoseethenumber\\nof observations in that leaf that take on values of YesandNoby specifying\\nshow_weights=True .\\nIn[9]: print(export_text(clf,\\nfeature_names=feature_names,\\nshow_weights=True))\\nOut[9]: |--- ShelveLoc[Good] <= 0.50\\n|| - - - P r i c e < = 9 2 . 5 0\\n||| - - - I n c o m e < = 5 7 . 0 0\\n|||| - - - w e i g h t s : [ 7 . 0 0 , 3 . 0 0 ] c l a s s : N o\\n||| - - - I n c o m e > 5 7 . 0 0\\n|||| - - - w e i g h t s : [ 7 . 0 0 , 2 9 . 0 0 ] c l a s s : Y e s\\n|| - - - P r i c e > 9 2 . 5 0\\n||| - - - A d v e r t i s i n g < = 1 3 . 5 0\\n|||| - - - w e i g h t s : [ 1 8 3 . 0 0 , 4 1 . 0 0 ] c l a s s : N o\\n||| - - - A d v e r t i s i n g > 1 3 . 5 0\\n|||| - - - w e i g h t s : [ 2 0 . 0 0 , 2 5 . 0 0 ] c l a s s : Y e s\\n|--- ShelveLoc[Good] > 0.50\\n|| - - - P r i c e < = 1 3 5 . 0 0\\n||| - - - U S [ Y e s ] < = 0 . 5 0\\n|||| - - - w e i g h t s : [ 6 . 0 0 , 1 1 . 0 0 ] c l a s s : Y e s\\n||| - - - U S [ Y e s ] > 0 . 5 0\\n|||| - - - w e i g h t s : [ 2 . 0 0 , 4 9 . 0 0 ] c l a s s : Y e s\\n|| - - - P r i c e > 1 3 5 . 0 0\\n||| - - - I n c o m e < = 4 6 . 0 0\\n|||| - - - w e i g h t s : [ 6 . 0 0 , 0 . 0 0 ] c l a s s : N o\\n||| - - - I n c o m e > 4 6 . 0 0\\n|||| - - - w e i g h t s : [ 5 . 0 0 , 6 . 0 0 ] c l a s s : Y e s', metadata={'source': './documents/ISLP_website.pdf', 'page': 362}), Document(page_content=\"8.3 Lab: Tree-Based Methods 357\\nIn order to properly evaluate the performance of a classification tree on\\nthese data, we must estimate the test error rather than simply computing\\nthe training error. We split the observations into a training set and a test\\nset, build the tree using the training set, and evaluate its performance\\non the test data. This pattern is similar to that in Chapter 6, with the\\nlinear models replaced here by decision trees — the code for validation is\\nalmost identical. This approach leads to correct predictions for 68.5% of\\nthe locations in the test data set.\\nIn[10]: validation = skm.ShuffleSplit(n_splits=1,\\ntest_size=200,\\nrandom_state=0)\\nresults = skm.cross_validate(clf,\\nD,\\nHigh,\\ncv=validation)\\nresults[ 'test_score' ]\\nOut[10]: array([0.685])\\nNext, we consider whether pruning the tree might lead to improved clas-\\nsification performance. We first split the data into a training and test set.\\nWe will use cross-validation to prune the tree on the training set, and then\\nevaluate the performance of the pruned tree on the test set.\\nIn[11]: (X_train,\\nX_test,\\nHigh_train,\\nHigh_test) = skm.train_test_split(X,\\nHigh,\\ntest_size=0.5,\\nrandom_state=0)\\nWe first refit the full tree on the training set; here we do not set a max_depth\\nparameter, since we will learn that through cross-validation.\\nIn[12]: clf = DTC(criterion= 'entropy' ,r a n d o m _ s t a t e = 0 )\\nclf.fit(X_train, High_train)\\naccuracy_score(High_test, clf.predict(X_test))\\nOut[12]: 0.735\\nNext we use the cost_complexity_pruning_path() method of clfto extractcost_\\ncomplexity_\\npruning_\\npath()cost-complexity values.\\nIn[13]: ccp_path = clf.cost_complexity_pruning_path(X_train, High_train)\\nkfold = skm.KFold(10,\\nrandom_state=1,\\nshuffle=True)\\nThis yields a set of impurities and αvalues from which we can extract an\\noptimal one by cross-validation.\\nIn[14]: grid = skm.GridSearchCV(clf,\\n{'ccp_alpha' :c c p _ p a t h . c c p _ a l p h a s } ,\\nrefit=True,\", metadata={'source': './documents/ISLP_website.pdf', 'page': 363}), Document(page_content='358 8. Tree-Based Methodscv=kfold,scoring=\\'accuracy\\')grid.fit(X_train, High_train)grid.best_score_Out[14]:0.685Let’s take a look at the pruned true.In[15]:ax = subplots(figsize=(12, 12))[1]best_ = grid.best_estimator_plot_tree(best_,feature_names=feature_names,ax=ax);Thisisquiteabushytree.Wecouldcounttheleaves,orquerybest_instead.In[16]:best_.tree_.n_leavesOut[16]:30The tree with 30 terminal nodes results in the lowest cross-validation errorrate, with an accuracy of 68.5%. How well does this pruned tree performon the test data set? Once again, we apply thepredict()function.In[17]:print(accuracy_score(High_test,best_.predict(X_test)))confusion = confusion_table(best_.predict(X_test),High_test)confusionOut[17]:0.72Truth No YesPredictedNo 108 61Yes 10 21Now 72.0% of the test observations are correctly classified, which isslightly worse than the error for the full tree (with 35 leaves). So cross-validation has not helped us much here; it only pruned off 5 leaves, ata cost of a slightly worse error. These results would change if we were tochange the random number seeds above; even though cross-validation givesan unbiased approach to model selection, it does have variance.8.3.2 Fitting Regression TreesHere we fit a regression tree to theBostondata set. The steps are similarto those for classification trees.In[18]:Boston = load_data(\"Boston\")model = MS(Boston.columns.drop(\\'medv\\'), intercept=False)D=m o d e l . f i t _ t r a n s f o r m ( B o s t o n )feature_names = list(D.columns)X=n p . a s a r r a y ( D )', metadata={'source': './documents/ISLP_website.pdf', 'page': 364}), Document(page_content=\"8.3 Lab: Tree-Based Methods 359First, we split the data into training and test sets, and fit the tree to thetraining data. Here we use 30% of the data for the test set.In[19]:(X_train,X_test,y_train,y_test) = skm.train_test_split(X,Boston['medv'],test_size=0.3,random_state=0)Having formed our training and test data sets, we fit the regression tree.In[20]:reg = DTR(max_depth=3)reg.fit(X_train, y_train)ax = subplots(figsize=(12,12))[1]plot_tree(reg,feature_names=feature_names,ax=ax);The variablelstatmeasures the percentage of individuals with lowersocioeconomic status. The tree indicates that lower values oflstatcorre-spond to more expensive houses. The tree predicts a median house priceof $12,042 for small-sized homes (rm < 6.8), in suburbs in which residentshave low socioeconomic status (lstat > 14.4) and the crime-rate is mod-erate (crim > 5.8).Now we use the cross-validation function to see whether pruning the treewill improve performance.In[21]:ccp_path = reg.cost_complexity_pruning_path(X_train, y_train)kfold = skm.KFold(5,shuffle=True,random_state=10)grid = skm.GridSearchCV(reg,{'ccp_alpha':c c p _ p a t h . c c p _ a l p h a s } ,refit=True,cv=kfold,scoring='neg_mean_squared_error')G=g r i d . f i t ( X _ t r a i n ,y _ t r a i n )In keeping with the cross-validation results, we use the pruned tree tomake predictions on the test set.In[22]:best_ = grid.best_estimator_np.mean((y_test - best_.predict(X_test))**2)Out[22]:28.07In other words, the test set MSE associated with the regression tree is28.07. The square root of the MSE is therefore around 5.30, indicating thatthis model leads to test predictions that are within around $5300 of thetrue median home value for the suburb.Let’s plot the best tree to see how interpretable it is.In[23]:ax = subplots(figsize=(12,12))[1]plot_tree(G.best_estimator_,feature_names=feature_names,ax=ax);\", metadata={'source': './documents/ISLP_website.pdf', 'page': 365}), Document(page_content='360 8. Tree-Based Methods8.3.3 Bagging and Random ForestsHere we apply bagging and random forests to theBostondata, using theRandomForestRegressor()from thesklearn.ensemblepackage. Recall thatRandomForestRegressor()sklearn.ensemblebagging is simply a special case of a random forest withm=p. Therefore,theRandomForestRegressor()function can be used to perform both baggingand random forests. We start with bagging.In[24]:bag_boston = RF(max_features=X_train.shape[1], random_state=0)bag_boston.fit(X_train, y_train)Out[24]:RandomForestRegressor(max_features=12, random_state=0)The argumentmax_featuresindicates that all 12 predictors should beconsidered for each split of the tree — in other words, that bagging shouldbe done. How well does this bagged model perform on the test set?In[25]:ax = subplots(figsize=(8,8))[1]y_hat_bag = bag_boston.predict(X_test)ax.scatter(y_hat_bag, y_test)np.mean((y_test - y_hat_bag)**2)Out[25]:14.63The test set MSE associated with the bagged regression tree is 14.63, abouthalf that obtained using an optimally-pruned single tree. We could changethenumberoftreesgrownfromthedefaultof100byusingthen_estimatorsargument:In[26]:bag_boston = RF(max_features=X_train.shape[1],n_estimators=500,random_state=0).fit(X_train, y_train)y_hat_bag = bag_boston.predict(X_test)np.mean((y_test - y_hat_bag)**2)Out[26]:14.61There is not much change. Bagging and random forests cannot overfit byincreasing the number of trees, but can underfit if the number is too small.Growing a random forest proceeds in exactly the same way, exceptthat we use a smaller value of themax_featuresargument. By default,RandomForestRegressor()usespvariables when building a random forest ofregression trees (i.e. it defaults to bagging), andRandomForestClassifier()uses√pvariableswhenbuildingarandomforestofclassificationtrees.Herewe usemax_features=6.In[27]:RF_boston = RF(max_features=6,random_state=0).fit(X_train, y_train)y_hat_RF = RF_boston.predict(X_test)np.mean((y_test - y_hat_RF)**2)Out[27]:20.04The test set MSE is 20.04; this indicates that random forests did some-what worse than bagging in this case. Extracting thefeature_importances_values from the fitted model, we can view the importance of each variable.', metadata={'source': './documents/ISLP_website.pdf', 'page': 366}), Document(page_content=\"8.3 Lab: Tree-Based Methods 361\\nIn[28]: feature_imp = pd.DataFrame(\\n{'importance' :RF_boston.feature_importances_},\\nindex=feature_names)\\nfeature_imp.sort_values(by= 'importance' ,a s c e n d i n g = F a l s e )\\nOut[28]: importance\\nlstat 0.368683\\nrm 0.333842\\nptratio 0.057306\\nindus 0.053303\\ncrim 0.052426\\ndis 0.042493\\nnox 0.034410\\nage 0.024327\\ntax 0.022368\\nrad 0.005048\\nzn 0.003238\\nchas 0.002557\\nThisisarelativemeasure ofthetotaldecrease innodeimpuritythat results\\nfrom splits over that variable, averaged over all trees (this was plotted in\\nFigure8.9for a model fit to the Heartdata).\\nThe results indicate that across all of the trees considered in the random\\nforest, the wealth level of the community ( lstat) and the house size ( rm)\\nare by far the two most important variables.\\n8.3.4 Boosting\\nHere we use GradientBoostingRegressor() fromsklearn.ensemble to fitGradient\\nBoosting\\nRegressor()boosted regression trees to the Bostondata set. For classification we would\\nuseGradientBoostingClassifier() . The argument n_estimators=5000 indi-\\nGradient\\nBoosting\\nClassifier()cates that we want 5000 trees, and the option max_depth=3 limits the depth\\nof each tree. The argument learning_rate is theλmentioned earlier in the\\ndescription of boosting.\\nIn[29]: boost_boston = GBR(n_estimators=5000,\\nlearning_rate=0.001,\\nmax_depth=3,\\nrandom_state=0)\\nboost_boston.fit(X_train, y_train)\\nWe can see how the training error decreases with the train_score_ at-\\ntribute. To get an idea of how the test error decreases we can use the\\nstaged_predict() method to get the predicted values along the path.\\nIn[30]: test_error = np.zeros_like(boost_boston.train_score_)\\nfor idx, y_ in enumerate(boost_boston.staged_predict(X_test)):\\ntest_error[idx] = np.mean((y_test - y_)**2)\\nplot_idx = np.arange(boost_boston.train_score_.shape[0])\\nax = subplots(figsize=(8,8))[1]\\nax.plot(plot_idx,\\nboost_boston.train_score_,\\n'b',\\nlabel='Training' )\", metadata={'source': './documents/ISLP_website.pdf', 'page': 367}), Document(page_content=\"362 8. Tree-Based Methods\\nax.plot(plot_idx,\\ntest_error,\\n'r',\\nlabel='Test')\\nax.legend();\\nWe now use the boosted model to predict medvon the test set:\\nIn[31]: y_hat_boost = boost_boston.predict(X_test);\\nnp.mean((y_test - y_hat_boost)**2)\\nOut[31]: 14.48\\nThe test MSE obtained is 14.48, similar to the test MSE for bagging. If we\\nwant to, we can perform boosting with a different value of the shrinkage\\nparameter λin(8.10).Thedefaultvalueis0.001,butthisiseasilymodified.\\nHere we take λ=0.2.\\nIn[32]: boost_boston = GBR(n_estimators=5000,\\nlearning_rate=0.2,\\nmax_depth=3,\\nrandom_state=0)\\nboost_boston.fit(X_train,\\ny_train)\\ny_hat_boost = boost_boston.predict(X_test);\\nnp.mean((y_test - y_hat_boost)**2)\\nOut[32]: 14.50\\nIn this case, using λ=0.2leads to a almost the same test MSE as when\\nusingλ=0.001.\\n8.3.5 Bayesian Additive Regression Trees\\nIn this section we demonstrate a Pythonimplementation of BART found in\\ntheISLP.bart package. We fit a model to the Bostonhousing data set. This\\nBART()estimator is designed for quantitative outcome variables, thoughBART()other implementations are available for fitting logistic and probit models\\nto categorical outcomes.\\nIn[33]: bart_boston = BART(random_state=0, burnin=5, ndraw=15)\\nbart_boston.fit(X_train, y_train)\\nOut[33]: BART(burnin=5, ndraw=15, random_state=0)\\nOn this data set, with this split into test and training, we see that the\\ntest error of BART is similar to that of random forest.\\nIn[34]: yhat_test = bart_boston.predict(X_test.astype(np.float32))\\nnp.mean((y_test - yhat_test)**2)\\nOut[34]: 20.92\\nWe can check how many times each variable appeared in the collection\\nof trees. This gives a summary similar to the variable importance plot for\\nboosting and random forests.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 368}), Document(page_content='8.4 Exercises 363\\nIn[35]: var_inclusion = pd.Series(bart_boston.variable_inclusion_.mean(0),\\nindex=D.columns)\\nvar_inclusion\\nOut[35]: crim 25.333333\\nzn 27.000000\\nindus 21.266667\\nchas 20.466667\\nnox 25.400000\\nrm 32.400000\\nage 26.133333\\ndis 25.666667\\nrad 24.666667\\ntax 23.933333\\nptratio 25.000000\\nlstat 31.866667\\ndtype: float64\\n8.4 Exercises\\nConceptual\\n1. Draw an example (of your own invention) of a partition of two-\\ndimensional feature space that could result from recursive binary\\nsplitting. Your example should contain at least six regions. Draw a\\ndecision tree corresponding to this partition. Be sure to label all as-\\npects of your figures, including the regions R1,R2,..., the cutpoints\\nt1,t2,..., and so forth.\\nHint: Your result should look something like Figures 8.1and8.2.\\n2. It is mentioned in Section 8.2.3that boosting using depth-one trees\\n(orstumps) leads to an additivemodel: that is, a model of the form\\nf(X)=p∑\\nj=1fj(Xj).\\nExplain why this is the case. You can begin with ( 8.12) in\\nAlgorithm 8.2.\\n3. Consider the Gini index, classification error, and entropy in a simple\\nclassification setting with two classes. Create a single plot that dis-\\nplays each of these quantities as a function of ˆpm1. Thex-axis should\\ndisplayˆpm1, ranging from 0to1, and the y-axis should display the\\nvalue of the Gini index, classification error, and entropy.\\nHint: In a setting with two classes, ˆpm1=1−ˆpm2. You could make\\nthis plot by hand, but it will be much easier to make in R.\\n4. This question relates to the plots in Figure 8.14.', metadata={'source': './documents/ISLP_website.pdf', 'page': 369}), Document(page_content='364 8. Tree-Based Methods| X2 < 1 \\nX1 < 1 X1 < 0 X2 < 2 -1.80 -1.06  0.21  0.63  2.49  \\n \\n5 \\n15 \\n10 \\n0 \\n3 \\n0 \\n1 \\nX2 \\nX1 \\n0 \\n1 \\nFIGURE 8.14. Left: A partition of the predictor space corresponding to Exer-\\ncise4a.Right: A tree corresponding to Exercise 4b.\\n(a) Sketch the tree corresponding to the partition of the predictor\\nspace illustrated in the left-hand panel of Figure 8.14. The num-\\nbers inside the boxes indicate the mean of Ywithin each region.\\n(b) Create a diagram similar to the left-hand panel of Figure 8.14,\\nusing the tree illustrated in the right-hand panel of the same\\nfigure. You should divide up the predictor space into the correct\\nregions, and indicate the mean for each region.\\n5. Suppose we produce ten bootstrapped samples from a data set\\ncontaining red and green classes. We then apply a classification tree\\nto each bootstrapped sample and, for a specific value of X, produce\\n10estimates of P(Class is Red |X):\\n0.1,0.15,0.2,0.2,0.55,0.6,0.6,0.65,0.7,and0.75.\\nThere are two common ways to combine these results together into a\\nsingle class prediction. One is the majority vote approach discussed in\\nthis chapter. The second approach is to classify based on the average\\nprobability.Inthisexample,whatisthefinalclassificationundereach\\nof these two approaches?\\n6. Provide a detailed explanation of the algorithm that is used to fit a\\nregression tree.\\nApplied\\n7. In Section 8.3.3, we applied random forests to the Bostondata using\\nmax_features = 6 and using n_estimators = 100 andn_estimators =\\n500. Create a plot displaying the test error resulting from random\\nforests on this data set for a more comprehensive range of values\\nformax_features andn_estimators . You can model your plot after\\nFigure8.10. Describe the results obtained.\\n8. Inthelab,aclassificationtreewasappliedtothe Carseats datasetaf-\\nter converting Salesinto a qualitative response variable. Now we will\\nseek to predict Salesusing regression trees and related approaches,\\ntreating the response as a quantitative variable.', metadata={'source': './documents/ISLP_website.pdf', 'page': 370}), Document(page_content='8.4 Exercises 365(a) Split the data set into a training set and a test set.(b) Fit a regression tree to the training set. Plot the tree, and inter-pret the results. What test MSE do you obtain?(c) Use cross-validation in order to determine the optimal level oftree complexity. Does pruning the tree improve the test MSE?(d) Use the bagging approach in order to analyze this data. Whattest MSE do you obtain? Use thefeature_importance_values todetermine which variables are most important.(e) Use random forests to analyze this data. What test MSE doyou obtain? Use thefeature_importance_values to determinewhichvariablesaremostimportant.Describetheeffectofm,thenumber of variables considered at each split, on the error rateobtained.(f) Now analyze the data using BART, and report your results.9. This problem involves theOJdata set which is part of theISLPpackage.(a) Create a training set containing a random sample of800obser-vations, and a test set containing the remaining observations.(b) Fit a tree to the training data, withPurchaseas the responseand the other variables as predictors. What is the training errorrate?(c) Create a plot of the tree, and interpret the results. How manyterminal nodes does the tree have?(d) Use theexport_tree()function to produce a text summary ofthe fitted tree. Pick one of the terminal nodes, and interpret theinformation displayed.(e) Predict the response on the test data, and produce a confusionmatrix comparing the test labels to the predicted test labels.What is the test error rate?(f) Use cross-validation on the training set in order to determinethe optimal tree size.(g) Produce a plot with tree size on thex-axis and cross-validatedclassification error rate on they-axis.(h) Which tree size corresponds to the lowest cross-validated classi-fication error rate?(i) Produce a pruned tree corresponding to the optimal tree sizeobtained using cross-validation. If cross-validation does not leadto selection of a pruned tree, then create a pruned tree with fiveterminal nodes.(j) Compare the training error rates between the pruned and un-pruned trees. Which is higher?(k) Compare the test error rates between the pruned and unprunedtrees. Which is higher?', metadata={'source': './documents/ISLP_website.pdf', 'page': 371}), Document(page_content='366 8. Tree-Based Methods\\n10. We now use boosting to predict Salaryin theHittersdata set.\\n(a) Remove the observations for whom the salary information is\\nunknown, and then log-transform the salaries.\\n(b) Create a training set consisting of the first 200 observations, and\\na test set consisting of the remaining observations.\\n(c) Perform boosting on the training set with 1,000 trees for a range\\nof values of the shrinkage parameter λ. Produce a plot with\\ndifferent shrinkage values on the x-axis and the corresponding\\ntraining set MSE on the y-axis.\\n(d) Produce a plot with different shrinkage values on the x-axis and\\nthe corresponding test set MSE on the y-axis.\\n(e) Compare the test MSE of boosting to the test MSE that results\\nfrom applying two of the regression approaches seen in\\nChapters 3and6.\\n(f) Which variables appear to be the most important predictors in\\nthe boosted model?\\n(g) Now apply bagging to the training set. What is the test set MSE\\nfor this approach?\\n11. This question uses the Caravandata set.\\n(a) Create a training set consisting of the first 1,000 observations,\\nand a test set consisting of the remaining observations.\\n(b) Fit a boosting model to the training set with Purchase as the\\nresponse and the other variables as predictors. Use 1,000 trees,\\nand a shrinkage value of 0.01. Which predictors appear to be\\nthe most important?\\n(c) Use the boosting model to predict the response on the test data.\\nPredictthatapersonwillmakeapurchaseiftheestimatedprob-\\nability of purchase is greater than 20 % . Form a confusion ma-\\ntrix. What fraction of the people predicted to make a purchase\\ndo in fact make one? How does this compare with the results\\nobtained from applying KNN or logistic regression to this data\\nset?\\n12. Apply boosting, bagging, random forests, and BART to a data set\\nof your choice. Be sure to fit the models on a training set and to\\nevaluate their performance on a test set. How accurate are the results\\ncompared to simple methods like linear or logistic regression? Which\\nof these approaches yields the best performance?', metadata={'source': './documents/ISLP_website.pdf', 'page': 372}), Document(page_content='9\\nSupport Vector Machines\\nIn this chapter, we discuss the support vector machine (SVM), an approach\\nfor classification that was developed in the computer science community in\\nthe 1990s and that has grown in popularity since then. SVMs have been\\nshown to perform well in a variety of settings, and are often considered one\\nof the best “out of the box” classifiers.\\nThe support vector machine is a generalization of a simple and intu-\\nitive classifier called the maximal margin classifier , which we introduce in\\nSection9.1. Though it is elegant and simple, we will see that this classifier\\nunfortunately cannot be applied to most data sets, since it requires that\\nthe classes be separable by a linear boundary. In Section 9.2, we introduce\\nthesupport vector classifier , an extension of the maximal margin classifier\\nthat can be applied in a broader range of cases. Section 9.3introduces the\\nsupport vector machine , which is a further extension of the support vec-\\ntor classifier in order to accommodate non-linear class boundaries. Support\\nvector machines are intended for the binary classification setting in which\\nthere are two classes; in Section 9.4we discuss extensions of support vector\\nmachines to the case of more than two classes. In Section 9.5we discuss\\nthecloseconnectionsbetweensupportvectormachinesandotherstatistical\\nmethods such as logistic regression.\\nPeople often loosely refer to the maximal margin classifier, the support\\nvector classifier, and the support vector machine as “support vector\\nmachines”. To avoid confusion, we will carefully distinguish between these\\nthree notions in this chapter.\\n9.1 Maximal Margin Classifier\\nIn this section, we define a hyperplane and introduce the concept of an\\noptimal separating hyperplane.\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning , Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_9  367', metadata={'source': './documents/ISLP_website.pdf', 'page': 373}), Document(page_content='368 9. Support Vector Machines\\n9.1.1 What Is a Hyperplane?\\nIn ap-dimensional space, a hyperplane is a flat affine subspace ofhyperplanedimension p−1.1For instance, in two dimensions, a hyperplane is a flat\\none-dimensional subspace—in other words, a line. In three dimensions, a\\nhyperplane is a flat two-dimensional subspace—that is, a plane. In p>3\\ndimensions, it can be hard to visualize a hyperplane, but the notion of a\\n(p−1)-dimensional flat subspace still applies.\\nThe mathematical definition of a hyperplane is quite simple. In two di-\\nmensions, a hyperplane is defined by the equation\\nβ0+β1X1+β2X2=0 (9.1)\\nfor parameters β0,β1, andβ2. When we say that ( 9.1) “defines” the hyper-\\nplane, we mean that any X=(X1,X2)Tfor which ( 9.1) holds is a point\\non the hyperplane. Note that ( 9.1) is simply the equation of a line, since\\nindeed in two dimensions a hyperplane is a line.\\nEquation 9.1can be easily extended to the p-dimensional setting:\\nβ0+β1X1+β2X2+···+βpXp=0 (9.2)\\ndefines a p-dimensional hyperplane, again in the sense that if a point X=\\n(X1,X2,...,X p)Tinp-dimensional space (i.e. a vector of length p) satisfies\\n(9.2), thenXlies on the hyperplane.\\nNow, suppose that Xdoes not satisfy ( 9.2); rather,\\nβ0+β1X1+β2X2+···+βpXp>0. (9.3)\\nThen this tells us that Xlies to one side of the hyperplane. On the other\\nhand, if\\nβ0+β1X1+β2X2+···+βpXp<0, (9.4)\\nthenXlies on the other side of the hyperplane. So we can think of the\\nhyperplane as dividing p-dimensional space into two halves. One can easily\\ndetermineonwhichsideofthehyperplaneapointliesbysimplycalculating\\nthe sign of the left-hand side of ( 9.2). A hyperplane in two-dimensional\\nspace is shown in Figure 9.1.\\n9.1.2 Classification Using a Separating Hyperplane\\nNow suppose that we have an n×pdata matrix Xthat consists of n\\ntraining observations in p-dimensional space,\\nx1=\\uf8eb\\n\\uf8ec\\uf8edx11\\n...\\nx1p\\uf8f6\\n\\uf8f7\\uf8f8,...,x n=\\uf8eb\\n\\uf8ec\\uf8edxn1\\n...\\nxnp\\uf8f6\\n\\uf8f7\\uf8f8, (9.5)\\nand that these observations fall into two classes—that is, y1,...,y n∈\\n{−1,1}where−1represents one class and 1the other class. We also have a\\n1The word affineindicates that the subspace need not pass through the origin.', metadata={'source': './documents/ISLP_website.pdf', 'page': 374}), Document(page_content='9.1 Maximal Margin Classifier 369\\n−1.5 −1.0 −0.5 0.0 0.5 1.0 1.5−1.5 −1.0 −0.5 0.0 0.5 1.0 1.5X1X2\\nFIGURE 9.1. The hyperplane 1+2 X1+3X2=0is shown. The blue region is\\nthe set of points for which 1+2 X1+3X2>0, and the purple region is the set of\\npoints for which 1+2 X1+3X2<0.\\ntest observation, a p-vector of observed features x∗=(x∗\\n1... x∗\\np)T. Our\\ngoal is to develop a classifier based on the training data that will correctly\\nclassify the test observation using its feature measurements. We have seen\\na number of approaches for this task, such as linear discriminant analysis\\nand logistic regression in Chapter 4, and classification trees, bagging, and\\nboosting in Chapter 8. We will now see a new approach that is based upon\\nthe concept of a separating hyperplane .separating\\nhyperplaneSuppose that it is possible to construct a hyperplane that separates the\\ntraining observations perfectly according to their class labels. Examples\\nof three such separating hyperplanes are shown in the left-hand panel of\\nFigure9.2. We can label the observations from the blue class as yi=1and\\nthose from the purple class as yi=−1. Then a separating hyperplane has\\nthe property that\\nβ0+β1xi1+β2xi2+···+βpxip>0 ifyi=1, (9.6)\\nand\\nβ0+β1xi1+β2xi2+···+βpxip<0 ifyi=−1. (9.7)\\nEquivalently, a separating hyperplane has the property that\\nyi(β0+β1xi1+β2xi2+···+βpxip)>0 (9.8)\\nfor alli=1,...,n .\\nIfaseparatinghyperplaneexists,wecanuseittoconstructaverynatural\\nclassifier: a test observation is assigned a class depending on which side of\\nthe hyperplane it is located. The right-hand panel of Figure 9.2shows\\nan example of such a classifier. That is, we classify the test observation x∗\\nbasedonthesignof f(x∗)=β0+β1x∗\\n1+β2x∗\\n2+···+βpx∗\\np.Iff(x∗)ispositive,\\nthen we assign the test observation to class 1, and iff(x∗)is negative, then\\nwe assign it to class −1. We can also make use of the magnitude off(x∗). If', metadata={'source': './documents/ISLP_website.pdf', 'page': 375}), Document(page_content='370 9. Support Vector Machines\\n−1 0 1 2 3−1 0 1 2 3−1 0 1 2 3−1 0 1 2 3X1X1\\nX2X2\\nFIGURE 9.2. Left:There are two classes of observations, shown in blue and\\nin purple, each of which has measurements on two variables. Three separating\\nhyperplanes, out of many possible, are shown in black. Right:A separating hy-\\nperplane is shown in black. The blue and purple grid indicates the decision rule\\nmade by a classifier based on this separating hyperplane: a test observation that\\nfalls in the blue portion of the grid will be assigned to the blue class, and a test\\nobservation that falls into the purple portion of the grid will be assigned to the\\npurple class.\\nf(x∗)is far from zero, then this means that x∗lies far from the hyperplane,\\nand so we can be confident about our class assignment for x∗. On the other\\nhand,iff(x∗)isclosetozero,then x∗islocatednearthehyperplane,andso\\nwe are less certain about the class assignment for x∗. Not surprisingly, and\\nas we see in Figure 9.2, a classifier that is based on a separating hyperplane\\nleads to a linear decision boundary.\\n9.1.3 The Maximal Margin Classifier\\nIn general, if our data can be perfectly separated using a hyperplane, then\\nthere will in fact exist an infinite number of such hyperplanes. This is\\nbecauseagivenseparatinghyperplanecanusuallybeshiftedatinybitupor\\ndown,orrotated,withoutcomingintocontactwithanyoftheobservations.\\nThree possible separating hyperplanes are shown in the left-hand panel\\nof Figure 9.2. In order to construct a classifier based upon a separating\\nhyperplane, we must have a reasonable way to decide which of the infinite\\npossible separating hyperplanes to use.\\nA natural choice is the maximal margin hyperplane (also known as themaximal\\nmargin\\nhyperplaneoptimal separating hyperplane ), which is the separating hyperplane that\\noptimal\\nseparating\\nhyperplaneis farthest from the training observations. That is, we can compute the\\n(perpendicular) distance from each training observation to a given separat-\\ning hyperplane; the smallest such distance is the minimal distance from the\\nobservations to the hyperplane, and is known as the margin. The maximal\\nmargin margin hyperplane is the separating hyperplane for which the margin is\\nlargest—that is, it is the hyperplane that has the farthest minimum dis-\\ntance to the training observations. We can then classify a test observation\\nbasedonwhichsideofthemaximalmarginhyperplaneitlies.Thisisknown', metadata={'source': './documents/ISLP_website.pdf', 'page': 376}), Document(page_content='9.1 Maximal Margin Classifier 371\\n−10123−10123\\nX1X2\\nFIGURE 9.3. There are two classes of observations, shown in blue and in\\npurple. The maximal margin hyperplane is shown as a solid line. The margin\\nis the distance from the solid line to either of the dashed lines. The two blue\\npoints and the purple point that lie on the dashed lines are the support vectors,\\nand the distance from those points to the hyperplane is indicated by arrows. The\\npurple and blue grid indicates the decision rule made by a classifier based on this\\nseparating hyperplane.\\nas themaximal margin classifier . We hope that a classifier that has a largemaximal\\nmargin\\nclassifiermargin on the training data will also have a large margin on the test data,\\nand hence will classify the test observations correctly. Although the maxi-\\nmal margin classifier is often successful, it can also lead to overfitting when\\npis large.\\nIfβ0,β1,...,βpare the coefficients of the maximal margin hyperplane,\\nthen the maximal margin classifier classifies the test observation x∗based\\non the sign of f(x∗)=β0+β1x∗\\n1+β2x∗\\n2+···+βpx∗\\np.\\nFigure9.3shows the maximal margin hyperplane on the data set of\\nFigure9.2. Comparing the right-hand panel of Figure 9.2to Figure 9.3,\\nwe see that the maximal margin hyperplane shown in Figure 9.3does in-\\ndeed result in a greater minimal distance between the observations and the\\nseparating hyperplane—that is, a larger margin. In a sense, the maximal\\nmargin hyperplane represents the mid-line of the widest “slab” that we can\\ninsert between the two classes.\\nExaminingFigure 9.3,weseethatthreetrainingobservationsareequidis-\\ntant from the maximal margin hyperplane and lie along the dashed lines\\nindicating the width of the margin. These three observations are known as\\nsupport vectors ,sincetheyarevectorsin p-dimensionalspace(inFigure 9.3,support\\nvector p=2) and they “support” the maximal margin hyperplane in the sense\\nthat if these points were moved slightly then the maximal margin hyper-\\nplane would move as well. Interestingly, the maximal margin hyperplane\\ndepends directly on the support vectors, but not on the other observations:\\namovementtoanyoftheotherobservationswouldnotaffecttheseparating\\nhyperplane, provided that the observation’s movement does not cause it to', metadata={'source': './documents/ISLP_website.pdf', 'page': 377}), Document(page_content='372 9. Support Vector Machines\\ncross the boundary set by the margin. The fact that the maximal margin\\nhyperplane depends directly on only a small subset of the observations is\\nan important property that will arise later in this chapter when we discuss\\nthe support vector classifier and support vector machines.\\n9.1.4 Construction of the Maximal Margin Classifier\\nWe now consider the task of constructing the maximal margin hyperplane\\nbased on a set of ntraining observations x1,...,x n∈Rpand associated\\nclass labels y1,...,y n∈{−1,1}. Briefly, the maximal margin hyperplane\\nis the solution to the optimization problem\\nmaximize\\nβ0,β1,...,βp,MM (9.9)\\nsubject top∑\\nj=1β2\\nj=1, (9.10)\\nyi(β0+β1xi1+β2xi2+···+βpxip)≥M∀i=1, . . . , n. (9.11)\\nThis optimization problem ( 9.9)–(9.11) is actually simpler than it looks.\\nFirst of all, the constraint in ( 9.11) that\\nyi(β0+β1xi1+β2xi2+···+βpxip)≥M∀i=1,...,n\\nguarantees that each observation will be on the correct side of the hyper-\\nplane, provided that Mis positive. (Actually, for each observation to be\\non the correct side of the hyperplane we would simply need yi(β0+β1xi1+\\nβ2xi2+···+βpxip)>0, so the constraint in ( 9.11) in fact requires that each\\nobservation be on the correct side of the hyperplane, with some cushion,\\nprovided that Mis positive.)\\nSecond,notethat( 9.10)isnotreallyaconstraintonthehyperplane,since\\nifβ0+β1xi1+β2xi2+···+βpxip=0defines a hyperplane, then so does\\nk(β0+β1xi1+β2xi2+···+βpxip)=0 for anyk̸=0. However, ( 9.10) adds\\nmeaningto( 9.11);onecanshowthatwiththisconstrainttheperpendicular\\ndistance from the ith observation to the hyperplane is given by\\nyi(β0+β1xi1+β2xi2+···+βpxip).\\nTherefore, the constraints ( 9.10) and (9.11) ensure that each observation\\nis on the correct side of the hyperplane and at least a distance Mfrom the\\nhyperplane. Hence, Mrepresents the margin of our hyperplane, and the\\noptimization problem chooses β0,β1,...,βpto maximize M. This is exactly\\nthe definition of the maximal margin hyperplane! The problem ( 9.9)–(9.11)\\ncan be solved efficiently, but details of this optimization are outside of the\\nscope of this book.\\n9.1.5 The Non-separable Case\\nThe maximal margin classifier is a very natural way to perform classifi-\\ncation,if a separating hyperplane exists . However, as we have hinted, in\\nmany cases no separating hyperplane exists, and so there is no maximal', metadata={'source': './documents/ISLP_website.pdf', 'page': 378}), Document(page_content='9.2 Support Vector Classifiers 373\\n0123−1.0 −0.5 0.0 0.5 1.0 1.5 2.0X1X2\\nFIGURE 9.4. There are two classes of observations, shown in blue and in\\npurple. In this case, the two classes are not separable by a hyperplane, and so the\\nmaximal margin classifier cannot be used.\\nmargin classifier. In this case, the optimization problem ( 9.9)–(9.11) has no\\nsolution with M> 0. An example is shown in Figure 9.4. In this case, we\\ncannotexactlyseparate the two classes. However, as we will see in the next\\nsection, we can extend the concept of a separating hyperplane in order to\\ndevelop a hyperplane that almostseparates the classes, using a so-called\\nsoft margin . The generalization of the maximal margin classifier to the\\nnon-separable case is known as the support vector classifier .\\n9.2 Support Vector Classifiers\\n9.2.1 Overview of the Support Vector Classifier\\nIn Figure 9.4, we see that observations that belong to two classes are not\\nnecessarily separable by a hyperplane. In fact, even if a separating hyper-\\nplane does exist, then there are instances in which a classifier based on\\na separating hyperplane might not be desirable. A classifier based on a\\nseparating hyperplane will necessarily perfectly classify all of the training\\nobservations; this can lead to sensitivity to individual observations. An ex-\\nample is shown in Figure 9.5. The addition of a single observation in the\\nright-hand panel of Figure 9.5leads to a dramatic change in the maxi-\\nmal margin hyperplane. The resulting maximal margin hyperplane is not\\nsatisfactory—for one thing, it has only a tiny margin. This is problematic\\nbecause as discussed previously, the distance of an observation from the\\nhyperplane can be seen as a measure of our confidence that the obser-\\nvation was correctly classified. Moreover, the fact that the maximal mar-\\ngin hyperplane is extremely sensitive to a change in a single observation\\nsuggests that it may have overfit the training data.\\nIn this case, we might be willing to consider a classifier based on a hy-\\nperplane that does notperfectly separate the two classes, in the interest of', metadata={'source': './documents/ISLP_website.pdf', 'page': 379}), Document(page_content='374 9. Support Vector Machines\\n−1 0 1 2 3−1 0 1 2 3−1 0 1 2 3−1 0 1 2 3X1X1X2X2\\nFIGURE 9.5. Left:Two classes of observations are shown in blue and in\\npurple, along with the maximal margin hyperplane. Right:An additional blue\\nobservation has been added, leading to a dramatic shift in the maximal margin\\nhyperplane shown as a solid line. The dashed line indicates the maximal margin\\nhyperplane that was obtained in the absence of this additional point.\\n• Greater robustness to individual observations, and\\n• Better classification of mostof the training observations.\\nThat is, it could be worthwhile to misclassify a few training observations\\nin order to do a better job in classifying the remaining observations.\\nThesupport vector classifier , sometimes called a soft margin classifier ,support\\nvector\\nclassifier\\nsoft margin\\nclassifierdoes exactly this. Rather than seeking the largest possible margin so that\\nevery observation is not only on the correct side of the hyperplane but\\nalso on the correct side of the margin, we instead allow some observations\\nto be on the incorrect side of the margin, or even the incorrect side of\\nthe hyperplane. (The margin is softbecause it can be violated by some\\nof the training observations.) An example is shown in the left-hand panel\\nofFigure 9.6.Mostoftheobservationsareonthecorrectsideofthemargin.\\nHowever, a small subset of the observations are on the wrong side of the\\nmargin.\\nAn observation can be not only on the wrong side of the margin, but also\\non the wrong side of the hyperplane. In fact, when there is no separating\\nhyperplane,suchasituationisinevitable.Observationsonthewrongsideof\\nthehyperplanecorrespondtotrainingobservationsthataremisclassifiedby\\nthe support vector classifier. The right-hand panel of Figure 9.6illustrates\\nsuch a scenario.\\n9.2.2 Details of the Support Vector Classifier\\nThe support vector classifier classifies a test observation depending on\\nwhich side of a hyperplane it lies. The hyperplane is chosen to correctly\\nseparate most of the training observations into the two classes, but may', metadata={'source': './documents/ISLP_website.pdf', 'page': 380}), Document(page_content='9.2 Support Vector Classifiers 375\\n−0.50.0 0.5 1.0 1.5 2.0 2.5−1 0 1 2 3 412345678910\\n−0.50.0 0.5 1.0 1.5 2.0 2.5−1 0 1 2 3 4123456789101112X1X1\\nX2X2\\nFIGURE 9.6. Left:A support vector classifier was fit to a small data set. The\\nhyperplane is shown as a solid line and the margins are shown as dashed lines.\\nPurple observations: Observations 3,4,5, and6are on the correct side of the\\nmargin, observation 2is on the margin, and observation 1 is on the wrong side of\\nthe margin. Blue observations: Observations 7and10are on the correct side of\\nthe margin, observation 9is on the margin, and observation 8is on the wrong side\\nof the margin. No observations are on the wrong side of the hyperplane. Right:\\nSame as left panel with two additional points, 11and12. These two observations\\nare on the wrong side of the hyperplane and the wrong side of the margin.\\nmisclassifyafewobservations.Itisthesolutiontotheoptimizationproblem\\nmaximize\\nβ0,β1,...,βp,ϵ1,...,ϵn,MM (9.12)\\nsubject top∑\\nj=1β2\\nj=1, (9.13)\\nyi(β0+β1xi1+β2xi2+···+βpxip)≥M(1−ϵi),(9.14)\\nϵi≥0,n∑\\ni=1ϵi≤C, (9.15)\\nwhereCis a nonnegative tuning parameter. As in ( 9.11),Mis the width\\nof the margin; we seek to make this quantity as large as possible. In ( 9.14),\\nϵ1,...,ϵnareslack variables that allow individual observations to be onslack\\nvariablethe wrong side of the margin or the hyperplane; we will explain them in\\ngreater detail momentarily. Once we have solved ( 9.12)–(9.15), we classify\\na test observation x∗as before, by simply determining on which side of the\\nhyperplane it lies. That is, we classify the test observation based on the\\nsign off(x∗)=β0+β1x∗\\n1+···+βpx∗\\np.\\nThe problem ( 9.12)–(9.15) seems complex, but insight into its behavior\\ncan be made through a series of simple observations presented below. First\\nof all, the slack variable ϵitells us where the ith observation is located,\\nrelative to the hyperplane and relative to the margin. If ϵi=0then the ith\\nobservation is on the correct side of the margin, as we saw in Section 9.1.4.\\nIfϵi>0then the ith observation is on the wrong side of the margin, and\\nwe say that the ith observation has violatedthe margin. If ϵi>1then it is\\non the wrong side of the hyperplane.', metadata={'source': './documents/ISLP_website.pdf', 'page': 381}), Document(page_content='376 9. Support Vector Machines\\nWe now consider the role of the tuning parameter C. In (9.15),Cbounds\\nthe sum of the ϵi’s, and so it determines the number and severity of the vio-\\nlations to the margin (and to the hyperplane) that we will tolerate. We can\\nthink of Cas abudgetfor the amount that the margin can be violated\\nby thenobservations. If C=0 then there is no budget for violations to\\nthe margin, and it must be the case that ϵ1=···=ϵn=0, in which case\\n(9.12)–(9.15) simply amounts to the maximal margin hyperplane optimiza-\\ntion problem ( 9.9)–(9.11). (Of course, a maximal margin hyperplane exists\\nonly if the two classes are separable.) For C>0no more than Cobserva-\\ntions can be on the wrong side of the hyperplane, because if an observation\\nis on the wrong side of the hyperplane then ϵi>1, and (9.15) requires\\nthat∑n\\ni=1ϵi≤C. As the budget Cincreases, we become more tolerant of\\nviolations to the margin, and so the margin will widen. Conversely, as C\\ndecreases, we become less tolerant of violations to the margin and so the\\nmargin narrows. An example is shown in Figure 9.7.\\nInpractice, Cistreatedasatuningparameterthatisgenerallychosenvia\\ncross-validation. As with the tuning parameters that we have seen through-\\nout this book, Ccontrols the bias-variance trade-off of the statistical learn-\\ning technique. When Cis small, we seek narrow margins that are rarely\\nviolated; this amounts to a classifier that is highly fit to the data, which\\nmay have low bias but high variance. On the other hand, when Cis larger,\\nthe margin is wider and we allow more violations to it; this amounts to\\nfitting the data less hard and obtaining a classifier that is potentially more\\nbiased but may have lower variance.\\nThe optimization problem ( 9.12)–(9.15) has a very interesting property:\\nit turns out that only observations that either lie on the margin or that\\nviolate the margin will affect the hyperplane, and hence the classifier ob-\\ntained. In other words, an observation that lies strictly on the correct side\\nof the margin does not affect the support vector classifier! Changing the\\nposition of that observation would not change the classifier at all, provided\\nthat its position remains on the correct side of the margin. Observations\\nthat lie directly on the margin, or on the wrong side of the margin for\\ntheir class, are known as support vectors . These observations do affect the\\nsupport vector classifier.\\nThe fact that only support vectors affect the classifier is in line with our\\nprevious assertion that Ccontrols the bias-variance trade-off of the support\\nvector classifier. When the tuning parameter Cis large, then the margin is\\nwide, many observations violate the margin, and so there are many support\\nvectors. In this case, many observations are involved in determining the\\nhyperplane. The top left panel in Figure 9.7illustrates this setting: this\\nclassifier has low variance (since many observations are support vectors)\\nbut potentially high bias. In contrast, if Cis small, then there will be fewer\\nsupport vectors and hence the resulting classifier will have low bias but\\nhigh variance. The bottom right panel in Figure 9.7illustrates this setting,\\nwith only eight support vectors.\\nThe fact that the support vector classifier’s decision rule is based only\\non a potentially small subset of the training observations (the support vec-\\ntors) means that it is quite robust to the behavior of observations that\\nare far away from the hyperplane. This property is distinct from some of', metadata={'source': './documents/ISLP_website.pdf', 'page': 382}), Document(page_content='9.3 Support Vector Machines 377\\n−1012−3 −2 −1 0 1 2 3−1012−3 −2 −1 0 1 2 3\\n−1012−3 −2 −1 0 1 2 3−1012−3 −2 −1 0 1 2 3X1X1X1X1X2X2\\nX2X2\\nFIGURE 9.7. A support vector classifier was fit using four different values\\nof the tuning parameter Cin (9.12)–(9.15). The largest value of Cwas used\\nin the top left panel, and smaller values were used in the top right, bottom left,\\nand bottom right panels. When Cis large, then there is a high tolerance for\\nobservations being on the wrong side of the margin, and so the margin will be\\nlarge. As Cdecreases, the tolerance for observations being on the wrong side of\\nthe margin decreases, and the margin narrows.\\nthe other classification methods that we have seen in preceding chapters,\\nsuch as linear discriminant analysis. Recall that the LDA classification rule\\ndepends on the mean of allof the observations within each class, as well as\\nthe within-class covariance matrix computed using allof the observations.\\nIn contrast, logistic regression, unlike LDA, has very low sensitivity to ob-\\nservations far from the decision boundary. In fact we will see in Section 9.5\\nthat the support vector classifier and logistic regression are closely related.\\n9.3 Support Vector Machines\\nWe first discuss a general mechanism for converting a linear classifier into\\none that produces non-linear decision boundaries. We then introduce the\\nsupport vector machine, which does this in an automatic way.', metadata={'source': './documents/ISLP_website.pdf', 'page': 383}), Document(page_content='378 9. Support Vector Machines\\n−4 −2 0 2 4−4 −2 0 2 4−4 −2 0 2 4−4 −2 0 2 4X1X1X2X2\\nFIGURE 9.8. Left:The observations fall into two classes, with a non-lin-\\near boundary between them. Right:The support vector classifier seeks a linear\\nboundary, and consequently performs very poorly.\\n9.3.1 Classification with Non-Linear Decision Boundaries\\nThe support vector classifier is a natural approach for classification in the\\ntwo-class setting, if the boundary between the two classes is linear. How-\\never, in practice we are sometimes faced with non-linear class boundaries.\\nFor instance, consider the data in the left-hand panel of Figure 9.8. It is\\nclear that a support vector classifier or any linear classifier will perform\\npoorly here. Indeed, the support vector classifier shown in the right-hand\\npanel of Figure 9.8is useless here.\\nIn Chapter 7, we are faced with an analogous situation. We see there\\nthat the performance of linear regression can suffer when there is a non-\\nlinear relationship between the predictors and the outcome. In that case,\\nwe consider enlarging the feature space using functions of the predictors,\\nsuch as quadratic and cubic terms, in order to address this non-linearity.\\nIn the case of the support vector classifier, we could address the prob-\\nlem of possibly non-linear boundaries between classes in a similar way, by\\nenlarging the feature space using quadratic, cubic, and even higher-order\\npolynomial functions of the predictors. For instance, rather than fitting a\\nsupport vector classifier using pfeatures\\nX1,X2,...,X p,\\nwe could instead fit a support vector classifier using 2pfeatures\\nX1,X2\\n1,X2,X2\\n2,...,X p,X2\\np.', metadata={'source': './documents/ISLP_website.pdf', 'page': 384}), Document(page_content='9.3 Support Vector Machines 379\\nThen (9.12)–(9.15) would become\\nmaximize\\nβ0,β11,β12,...,βp1,βp2,ϵ1,...,ϵn,MM (9.16)\\nsubject to yi\\uf8eb\\n\\uf8edβ0+p∑\\nj=1βj1xij+p∑\\nj=1βj2x2\\nij\\uf8f6\\n\\uf8f8≥M(1−ϵi),\\nn∑\\ni=1ϵi≤C,ϵi≥0,p∑\\nj=12∑\\nk=1β2\\njk=1.\\nWhy does this lead to a non-linear decision boundary? In the enlarged\\nfeature space, the decision boundary that results from ( 9.16) is in fact lin-\\near. But in the original feature space, the decision boundary is of the form\\nq(x)=0 , whereqis a quadratic polynomial, and its solutions are gener-\\nally non-linear. One might additionally want to enlarge the feature space\\nwith higher-order polynomial terms, or with interaction terms of the form\\nXjXj′forj̸=j′. Alternatively, other functions of the predictors could\\nbe considered rather than polynomials. It is not hard to see that there\\nare many possible ways to enlarge the feature space, and that unless we\\nare careful, we could end up with a huge number of features. Then compu-\\ntations would become unmanageable. The support vector machine, which\\nwe present next, allows us to enlarge the feature space used by the support\\nvector classifier in a way that leads to efficient computations.\\n9.3.2 The Support Vector Machine\\nThesupport vector machine (SVM) is an extension of the support vectorsupport\\nvector\\nmachineclassifier that results from enlarging the feature space in a specific way,\\nusingkernels. We will now discuss this extension, the details of which are\\nkernel somewhat complex and beyond the scope of this book. However, the main\\nidea is described in Section 9.3.1: we may want to enlarge our feature space\\nin order to accommodate a non-linear boundary between the classes. The\\nkernel approach that we describe here is simply an efficient computational\\napproach for enacting this idea.\\nWe have not discussed exactly how the support vector classifier is com-\\nputed because the details become somewhat technical. However, it turns\\nout that the solution to the support vector classifier problem ( 9.12)–(9.15)\\ninvolves only the inner products of the observations (as opposed to the\\nobservations themselves). The inner product of two r-vectors aandbis\\ndefined as ⟨a, b⟩=∑r\\ni=1aibi. Thus the inner product of two observations\\nxi,xi′is given by\\n⟨xi,xi′⟩=p∑\\nj=1xijxi′j. (9.17)\\nIt can be shown that\\n• The linear support vector classifier can be represented as\\nf(x)=β0+n∑\\ni=1αi⟨x, x i⟩, (9.18)', metadata={'source': './documents/ISLP_website.pdf', 'page': 385}), Document(page_content='380 9. Support Vector Machines\\nwhere there are nparameters αi,i=1,...,n , one per training\\nobservation.\\n• To estimate the parameters α1,...,αnandβ0, all we need are the(n\\n2)\\ninner products ⟨xi,xi′⟩between all pairs of training observations.\\n(The notation(n\\n2)\\nmeansn(n−1)/2, and gives the number of pairs\\namong a set of nitems.)\\nNotice that in ( 9.18), in order to evaluate the function f(x), we need to\\ncomputetheinnerproductbetweenthenewpoint xandeachofthetraining\\npointsxi. However, it turns out that αiis nonzero only for the support\\nvectors in the solution—that is, if a training observation is not a support\\nvector, then its αiequals zero. So if Sis the collection of indices of these\\nsupport points, we can rewrite any solution function of the form ( 9.18) as\\nf(x)=β0+∑\\ni∈Sαi⟨x, x i⟩, (9.19)\\nwhich typically involves far fewer terms than in ( 9.18).2\\nTosummarize,inrepresentingthelinearclassifier f(x),andincomputing\\nits coefficients, all we need are inner products.\\nNow suppose that every time the inner product ( 9.17) appears in the\\nrepresentation ( 9.18), or in a calculation of the solution for the support\\nvector classifier, we replace it with a generalization of the inner product of\\nthe form\\nK(xi,xi′), (9.20)\\nwhereKis some function that we will refer to as a kernel. A kernel is akernelfunction that quantifies the similarity of two observations. For instance, we\\ncould simply take\\nK(xi,xi′)=p∑\\nj=1xijxi′j, (9.21)\\nwhich would just give us back the support vector classifier. Equation 9.21\\nis known as a linearkernel because the support vector classifier is linear\\nin the features; the linear kernel essentially quantifies the similarity of a\\npair of observations using Pearson (standard) correlation. But one could\\ninstead choose another form for ( 9.20). For instance, one could replace\\nevery instance of∑p\\nj=1xijxi′jwith the quantity\\nK(xi,xi′) = (1 +p∑\\nj=1xijxi′j)d. (9.22)\\nThis is known as a polynomial kernel of degree d, wheredis a positivepolynomial\\nkernelinteger. Using such a kernel with d> 1, instead of the standard linear\\nkernel(9.21),inthesupportvectorclassifieralgorithmleadstoamuchmore\\nflexibledecision boundary.Itessentiallyamountstofitting a support vector\\n2By expanding each of the inner products in ( 9.19), it is easy to see that f(x)is\\na linear function of the coordinates of x. Doing so also establishes the correspondence\\nbetween the αiand the original parameters βj.', metadata={'source': './documents/ISLP_website.pdf', 'page': 386}), Document(page_content='9.3 Support Vector Machines 381\\n−4 −2 0 2 4−4 −2 0 2 4      \\n      \\n−4 −2 0 2 4−4 −2 0 2 4    \\n    \\nX1X1X2X2\\nFIGURE 9.9. Left:An SVM with a polynomial kernel of degree 3 is applied to\\nthe non-linear data from Figure 9.8, resulting in a far more appropriate decision\\nrule.Right:An SVM with a radial kernel is applied. In this example, either kernel\\nis capable of capturing the decision boundary.\\nclassifier in a higher-dimensional space involving polynomials of degree d,\\nrather than in the original feature space. When the support vector classifier\\niscombinedwithanon-linearkernelsuchas( 9.22),theresultingclassifieris\\nknown as a support vector machine. Note that in this case the (non-linear)\\nfunction has the form\\nf(x)=β0+∑\\ni∈SαiK(x, x i). (9.23)\\nThe left-hand panel of Figure 9.9shows an example of an SVM with a\\npolynomial kernel applied to the non-linear data from Figure 9.8. The fit is\\na substantial improvement over the linear support vector classifier. When\\nd=1, then the SVM reduces to the support vector classifier seen earlier in\\nthis chapter.\\nThe polynomial kernel shown in ( 9.22) is one example of a possible\\nnon-linear kernel, but alternatives abound. Another popular choice is the\\nradial kernel , which takes the formradial kernel\\nK(xi,xi′) = exp( −γp∑\\nj=1(xij−xi′j)2). (9.24)\\nIn (9.24),γis a positive constant. The right-hand panel of Figure 9.9shows\\nan example of an SVM with a radial kernel on this non-linear data; it also\\ndoes a good job in separating the two classes.\\nHow does the radial kernel ( 9.24) actually work? If a given test obser-\\nvationx∗=(x∗\\n1,...,x∗\\np)Tis far from a training observation xiin terms of\\nEuclidean distance, then∑p\\nj=1(x∗\\nj−xij)2will be large, and so K(x∗,xi)=\\nexp(−γ∑p\\nj=1(x∗\\nj−xij)2)will be tiny. This means that in ( 9.23),xiwill\\nplay virtually no role in f(x∗). Recall that the predicted class label for the\\ntest observation x∗is based on the sign of f(x∗). In other words, training\\nobservations that are far from x∗will play essentially no role in the pre-\\ndicted class label for x∗. This means that the radial kernel has very local', metadata={'source': './documents/ISLP_website.pdf', 'page': 387}), Document(page_content='382 9. Support Vector Machines\\nFalse positive rateTrue positive rate0.0 0.2 0.4 0.6 0.8 1.00.0 0.2 0.4 0.6 0.8 1.0Support Vector ClassifierLDAFalse positive rateTrue positive rate0.0 0.2 0.4 0.6 0.8 1.00.0 0.2 0.4 0.6 0.8 1.0Support Vector ClassifierSVM: γ=10−3SVM: γ=10−2SVM: γ=10−1FIGURE 9.10. ROC curves for the Heartdata training set. Left:The support\\nvector classifier and LDA are compared. Right:The support vector classifier is\\ncompared to an SVM using a radial basis kernel with γ= 10−3,10−2, and10−1.\\nbehavior, in the sense that only nearby training observations have an effect\\non the class label of a test observation.\\nWhat is the advantage of using a kernel rather than simply enlarging\\nthe feature space using functions of the original features, as in ( 9.16)? One\\nadvantage is computational, and it amounts to the fact that using kernels,\\none need only compute K(xi,x′\\ni)for all(n\\n2)\\ndistinct pairs i, i′. This can be\\ndone without explicitly working in the enlarged feature space. This is im-\\nportant because in many applications of SVMs, the enlarged feature space\\nis so large that computations are intractable. For some kernels, such as the\\nradial kernel ( 9.24), the feature space is implicitand infinite-dimensional,\\nso we could never do the computations there anyway!\\n9.3.3 An Application to the Heart Disease Data\\nInChapter 8weapplydecisiontreesandrelatedmethodstothe Heartdata.\\nTheaim isto use 13predictorssuchas Age,Sex,andCholinorderto predict\\nwhether an individual has heart disease. We now investigate how an SVM\\ncompares to LDA on this data. After removing 6 missing observations, the\\ndata consist of 297 subjects, which we randomly split into 207 training and\\n90 test observations.\\nWe first fit LDA and the support vector classifier to the training data.\\nNotethatthesupportvectorclassifierisequivalenttoanSVMusingapoly-\\nnomial kernel of degree d=1. The left-hand panel of Figure 9.10displays\\nROC curves (described in Section 4.4.2) for the training set predictions for\\nboth LDA and the support vector classifier. Both classifiers compute scores\\nof the form ˆf(X)=ˆβ0+ˆβ1X1+ˆβ2X2+···+ˆβpXpfor each observation.\\nFor any given cutoff t, we classify observations into the heart disease or\\nno heart disease categories depending on whether ˆf(X)<torˆf(X)≥t.\\nThe ROC curve is obtained by forming these predictions and computing\\nthe false positive and true positive rates for a range of values of t. An opti-\\nmal classifier will hug the top left corner of the ROC plot. In this instance', metadata={'source': './documents/ISLP_website.pdf', 'page': 388}), Document(page_content='9.4 SVMs with More than Two Classes 383\\nFalse positive rateTrue positive rate0.0 0.2 0.4 0.6 0.8 1.00.0 0.2 0.4 0.6 0.8 1.0Support Vector ClassifierLDAFalse positive rateTrue positive rate0.0 0.2 0.4 0.6 0.8 1.00.0 0.2 0.4 0.6 0.8 1.0Support Vector ClassifierSVM: γ=10−3SVM: γ=10−2SVM: γ=10−1FIGURE 9.11. ROC curves for the test set of the Heartdata.Left:The support\\nvector classifier and LDA are compared. Right:The support vector classifier is\\ncompared to an SVM using a radial basis kernel with γ= 10−3,10−2, and10−1.\\nLDA and the support vector classifier both perform well, though there is a\\nsuggestion that the support vector classifier may be slightly superior.\\nThe right-hand panel of Figure 9.10displays ROC curves for SVMs using\\na radial kernel, with various values of γ. Asγincreases and the fit becomes\\nmore non-linear, the ROC curves improve. Using γ= 10−1appears to give\\nan almost perfect ROC curve. However, these curves represent training\\nerror rates, which can be misleading in terms of performance on new test\\ndata. Figure 9.11displays ROC curves computed on the 90test observa-\\ntions. We observe some differences from the training ROC curves. In the\\nleft-hand panel of Figure 9.11, the support vector classifier appears to have\\na small advantage over LDA (although these differences are not statisti-\\ncally significant). In the right-hand panel, the SVM using γ= 10−1, which\\nshowed the best results on the training data, produces the worst estimates\\non the test data. This is once again evidence that while a more flexible\\nmethod will often produce lower training error rates, this does not neces-\\nsarily lead to improved performance on test data. The SVMs with γ= 10−2\\nandγ= 10−3perform comparably to the support vector classifier, and all\\nthree outperform the SVM with γ= 10−1.\\n9.4 SVMs with More than Two Classes\\nSo far, our discussion has been limited to the case of binary classification:\\nthat is, classification in the two-class setting. How can we extend SVMs\\nto the more general case where we have some arbitrary number of classes?\\nIt turns out that the concept of separating hyperplanes upon which SVMs\\nare based does not lend itself naturally to more than two classes. Though\\na number of proposals for extending SVMs to the K-class case have been\\nmade, the two most popular are the one-versus-one andone-versus-all\\napproaches. We briefly discuss those two approaches here.', metadata={'source': './documents/ISLP_website.pdf', 'page': 389}), Document(page_content='384 9. Support Vector Machines\\n9.4.1 One-Versus-One Classification\\nSuppose that we would like to perform classification using SVMs, and there\\nareK> 2classes. A one-versus-one orall-pairs approach constructs(K\\n2)\\none-versus-\\none SVMs, each of which compares a pair of classes. For example, one such\\nSVM might compare the kth class, coded as +1, to thek′th class, coded\\nas−1. We classify a test observation using each of the(K\\n2)\\nclassifiers, and\\nwe tally the number of times that the test observation is assigned to each\\nof theKclasses. The final classification is performed by assigning the test\\nobservation to the class to which it was most frequently assigned in these(K\\n2)\\npairwise classifications.\\n9.4.2 One-Versus-All Classification\\nTheone-versus-all approach (also referred to as one-versus-rest ) is an al-one-versus-\\nall\\none-versus-\\nrestternative procedure for applying SVMs in the case of K> 2classes. We\\nfitKSVMs, each time comparing one of the Kclasses to the remaining\\nK−1classes. Let β0k,β1k,...,βpkdenote the parameters that result from\\nfitting an SVM comparing the kth class (coded as +1) to the others (coded\\nas−1). Letx∗denote a test observation. We assign the observation to the\\nclass for which β0k+β1kx∗\\n1+β2kx∗\\n2+···+βpkx∗\\npis largest, as this amounts\\nto a high level of confidence that the test observation belongs to the kth\\nclass rather than to any of the other classes.\\n9.5 Relationship to Logistic Regression\\nWhen SVMs were first introduced in the mid-1990s, they made quite a\\nsplash in the statistical and machine learning communities. This was due\\nin part to their good performance, good marketing, and also to the fact\\nthat the underlying approach seemed both novel and mysterious. The idea\\nof finding a hyperplane that separates the data as well as possible, while al-\\nlowing some violations to this separation, seemed distinctly different from\\nclassical approaches for classification, such as logistic regression and lin-\\near discriminant analysis. Moreover, the idea of using a kernel to expand\\nthe feature space in order to accommodate non-linear class boundaries ap-\\npeared to be a unique and valuable characteristic.\\nHowever, since that time, deep connections between SVMs and other\\nmore classical statistical methods have emerged. It turns out that one can\\nrewrite the criterion ( 9.12)–(9.15) for fitting the support vector classifier\\nf(X)=β0+β1X1+···+βpXpas\\nminimize\\nβ0,β1,...,βp\\uf8f1\\n\\uf8f2\\n\\uf8f3n∑\\ni=1max [0 ,1−yif(xi)] +λp∑\\nj=1β2\\nj\\uf8fc\\n\\uf8fd\\n\\uf8fe, (9.25)\\nwhereλis a nonnegative tuning parameter. When λis large then β1,...,βp\\nare small, more violations to the margin are tolerated, and a low-variance\\nbut high-bias classifier will result. When λis small then few violations\\nto the margin will occur; this amounts to a high-variance but low-bias', metadata={'source': './documents/ISLP_website.pdf', 'page': 390}), Document(page_content='9.5 Relationship to Logistic Regression 385\\nclassifier. Thus, a small value of λin (9.25) amounts to a small value of C\\nin (9.15). Note that the λ∑p\\nj=1β2\\njterm in ( 9.25) is the ridge penalty term\\nfrom Section 6.2.1, and plays a similar role in controlling the bias-variance\\ntrade-off for the support vector classifier.\\nNow(9.25)takesthe“Loss+Penalty”formthatwehaveseenrepeatedly\\nthroughout this book:\\nminimize\\nβ0,β1,...,βp{L(X,y,β)+λP(β)}. (9.26)\\nIn (9.26),L(X,y,β)is some loss function quantifying the extent to which\\nthe model, parametrized by β, fits the data (X,y), andP(β)is a penalty\\nfunction on the parameter vector βwhose effect is controlled by a nonneg-\\native tuning parameter λ. For instance, ridge regression and the lasso both\\ntake this form with\\nL(X,y,β)=n∑\\ni=1\\uf8eb\\n\\uf8edyi−β0−p∑\\nj=1xijβj\\uf8f6\\n\\uf8f82\\nand with P(β)=∑p\\nj=1β2\\njfor ridge regression and P(β)=∑p\\nj=1|βj|for\\nthe lasso. In the case of ( 9.25) the loss function instead takes the form\\nL(X,y,β)=n∑\\ni=1max [0 ,1−yi(β0+β1xi1+···+βpxip)].\\nThis is known as hinge loss , and is depicted in Figure 9.12. However, ithinge lossturns out that the hinge loss function is closely related to the loss function\\nused in logistic regression, also shown in Figure 9.12.\\nAn interesting characteristic of the support vector classifier is that only\\nsupport vectors play a role in the classifier obtained; observations on the\\ncorrect side of the margin do not affect it. This is due to the fact that the\\nloss function shown in Figure 9.12is exactly zero for observations for which\\nyi(β0+β1xi1+···+βpxip)≥1; these correspond to observations that are\\non the correct side of the margin.3In contrast, the loss function for logistic\\nregression shown in Figure 9.12is not exactly zero anywhere. But it is very\\nsmall for observations that are far from the decision boundary. Due to the\\nsimilarities between their loss functions, logistic regression and the support\\nvector classifier often give very similar results. When the classes are well\\nseparated, SVMs tend to behave better than logistic regression; in more\\noverlapping regimes, logistic regression is often preferred.\\nWhen the support vector classifier and SVM were first introduced, it was\\nthought that the tuning parameter Cin (9.15) was an unimportant “nui-\\nsance” parameter that could be set to some default value, like 1. However,\\nthe “Loss + Penalty” formulation ( 9.25) for the support vector classifier\\nindicates that this is not the case. The choice of tuning parameter is very\\nimportant and determines the extent to which the model underfits or over-\\nfits the data, as illustrated, for example, in Figure 9.7.\\n3With this hinge-loss + penalty representation, the margin corresponds to the value\\none, and the width of the margin is determined by∑β2\\nj.', metadata={'source': './documents/ISLP_website.pdf', 'page': 391}), Document(page_content='386 9. Support Vector Machines\\n−6 −4 −2 0 202468LossSVM LossLogistic Regression Loss\\nyi(β0+β1xi1+...+βpxip)FIGURE 9.12. The SVM and logistic regression loss functions are compared,\\nas a function of yi(β0+β1xi1+···+βpxip). Whenyi(β0+β1xi1+···+βpxip)is\\ngreater than 1, then the SVM loss is zero, since this corresponds to an observation\\nthat is on the correct side of the margin. Overall, the two loss functions have quite\\nsimilar behavior.\\nWe have established that the support vector classifier is closely related\\nto logistic regression and other preexisting statistical methods. Is the SVM\\nunique in its use of kernels to enlarge the feature space to accommodate\\nnon-linear class boundaries? The answer to this question is “no”. We could\\njust as well perform logistic regression or many of the other classification\\nmethods seen in this book using non-linear kernels; this is closely related\\nto some of the non-linear approaches seen in Chapter 7. However, for his-\\ntorical reasons, the use of non-linear kernels is much more widespread in\\nthe context of SVMs than in the context of logistic regression or other\\nmethods.\\nThough we have not addressed it here, there is in fact an extension\\nof the SVM for regression (i.e. for a quantitative rather than a qualita-\\ntive response), called support vector regression . In Chapter 3, we saw thatsupport\\nvector\\nregressionleast squares regression seeks coefficients β0,β1,...,βpsuch that the sum\\nof squared residuals is as small as possible. (Recall from Chapter 3that\\nresiduals are defined as yi−β0−β1xi1−···−βpxip.) Support vector\\nregression instead seeks coefficients that minimize a different type of loss,\\nwhere only residuals larger in absolute value than some positive constant\\ncontribute to the loss function. This is an extension of the margin used in\\nsupport vector classifiers to the regression setting.', metadata={'source': './documents/ISLP_website.pdf', 'page': 392}), Document(page_content=\"9.6 Lab: Support Vector Machines 3879.6 Lab: Support Vector MachinesIn this lab, we use thesklearn.svmlibrary to demonstrate the supportvector classifier and the support vector machine.We import some of our usual libraries.In[1]:import numpy as npfrom matplotlib.pyplot import subplots, cmimport sklearn.model_selection as skmfrom ISLP import load_data, confusion_tableWe also collect the new imports needed for this lab.In[2]:from sklearn.svm import SVCfrom ISLP.svm import plot as plot_svmfrom sklearn.metrics import RocCurveDisplayWe will use the functionRocCurveDisplay.from_estimator()to produceRocCurveDisplay.from_estimator()several ROC plots, using a shorthandroc_curve.In[3]:roc_curve = RocCurveDisplay.from_estimator#s h o r t h a n d9.6.1 Support Vector ClassifierWe now use theSupportVectorClassifier()function (abbreviatedSVC())SupportVectorClassifier()fromsklearnto fit the support vector classifier for a given value of theparameterC. TheCargument allows us to specify the cost of a violationto the margin. When thecostargument is small, then the margins will bewide and many support vectors will be on the margin or will violate themargin. When theCargument is large, then the margins will be narrow andthere will be few support vectors on the margin or violating the margin.Here we demonstrate the use ofSVC()on a two-dimensional example, sothat we can plot the resulting decision boundary. We begin by generatingthe observations, which belong to two classes, and checking whether theclasses are linearly separable.In[4]:rng = np.random.default_rng(1)X=r n g . s t a n d a r d _ n o r m a l ( ( 5 0 ,2 ) )y=n p . a r r a y ( [ - 1 ] * 2 5 + [ 1 ] * 2 5 )X[y==1] += 1fig, ax = subplots(figsize=(8,8))ax.scatter(X[:,0],X[:,1],c=y,cmap=cm.coolwarm);They are not. We now fit the classifier.In[5]:svm_linear = SVC(C=10, kernel='linear')svm_linear.fit(X, y)Out[5]:SVC(C=10, kernel='linear')The support vector classifier with two features can be visualized by plot-ting values of itsdecision function. We have included a function for this indecisionfunctiontheISLPpackage (inspired by a similar example in thesklearndocs).\", metadata={'source': './documents/ISLP_website.pdf', 'page': 393}), Document(page_content=\"388 9. Support Vector MachinesIn[6]:fig, ax = subplots(figsize=(8,8))plot_svm(X,y,svm_linear,ax=ax)The decision boundary between the two classes is linear (because weused the argumentkernel='linear'). The support vectors are marked with+and the remaining observations are plotted as circles.What if we instead used a smaller value of the cost parameter?In[7]:svm_linear_small = SVC(C=0.1, kernel='linear')svm_linear_small.fit(X, y)fig, ax = subplots(figsize=(8,8))plot_svm(X,y,svm_linear_small,ax=ax)With a smaller value of the cost parameter, we obtain a larger number ofsupport vectors, because the margin is now wider. For linear kernels, wecan extract the coefficients of the linear decision boundary as follows:In[8]:svm_linear.coef_Out[8]:array([[1.173 , 0.7734]])Since the support vector machine is an estimator insklearn, we can usethe usual machinery to tune it.In[9]:kfold = skm.KFold(5,random_state=0,shuffle=True)grid = skm.GridSearchCV(svm_linear,{'C':[0.001,0.01,0.1,1,5,10,100]},refit=True,cv=kfold,scoring='accuracy')grid.fit(X, y)grid.best_params_Out[9]:{'C':1 }Wecaneasilyaccessthecross-validationerrorsforeachofthesemodelsingrid.cv_results_. This prints out a lot of detail, so we extract the accuracyresults only.In[10]:grid.cv_results_[('mean_test_score')]Out[10]:array([0.46, 0.46, 0.72, 0.74, 0.74, 0.74, 0.74])We see thatC=1results in the highest cross-validation accuracy of 0.74,though the accuracy is the same for several values ofC. The classifiergrid.best_estimator_can be used to predict the class label on a set oftest observations. Let’s generate a test data set.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 394}), Document(page_content=\"9.6 Lab: Support Vector Machines 389In[11]:X_test = rng.standard_normal((20, 2))y_test = np.array([-1]*10+[1]*10)X_test[y_test==1] += 1Nowwepredicttheclasslabelsofthesetestobservations.Hereweusethebest model selected by cross-validation in order to make the predictions.In[12]:best_ = grid.best_estimator_y_test_hat = best_.predict(X_test)confusion_table(y_test_hat, y_test)Out[12]:Truth -1 1Predicted-1 8 4126Thus, with this value ofC, 70% of the test observations are correctly clas-sified. What if we had instead usedC=0.001?In[13]:svm_ = SVC(C=0.001,kernel='linear').fit(X, y)y_test_hat = svm_.predict(X_test)confusion_table(y_test_hat, y_test)Out[13]:Truth -1 1Predicted-1 2 018 1 0In this case 60% of test observations are correctly classified.We now consider a situation in which the two classes are linearly sepa-rable. Then we can find an optimal separating hyperplane using theSVC()estimator. We first further separate the two classes in our simulated dataso that they are linearly separable:In[14]:X[y==1] += 1.9;fig, ax = subplots(figsize=(8,8))ax.scatter(X[:,0], X[:,1], c=y, cmap=cm.coolwarm);Now the observations are just barely linearly separable.In[15]:svm_ = SVC(C=1e5, kernel='linear').fit(X, y)y_hat = svm_.predict(X)confusion_table(y_hat, y)Out[15]:Truth -1 1Predicted-1 25 010 2 5We fit the support vector classifier and plot the resulting hyperplane, usinga very large value ofCso that no observations are misclassified.In[16]:fig, ax = subplots(figsize=(8,8))plot_svm(X,y,svm_,ax=ax)\", metadata={'source': './documents/ISLP_website.pdf', 'page': 395}), Document(page_content='390 9. Support Vector Machines\\nIndeed no training errors were made and only three support vectors were\\nused. In fact, the large value of Calso means that these three support points\\nareon the margin , and define it. One may wonder how good the classifier\\ncould be on test data that depends on only three data points! We now try\\na smaller value of C.\\nIn[17]: svm_ = SVC(C=0.1, kernel= \\'linear\\' ).fit(X, y)\\ny_hat = svm_.predict(X)\\nconfusion_table(y_hat, y)\\nOut[17]: Truth -1 1\\nPredicted\\n-1 25 0\\n10 2 5\\nUsingC=0.1, we again do not misclassify any training observations, but we\\nalso obtain a much wider margin and make use of twelve support vectors.\\nThese jointly define the orientation of the decision boundary, and since\\nthere are more of them, it is more stable. It seems possible that this model\\nwill perform better on test data than the model with C=1e5(and indeed, a\\nsimple experiment with a large test set would bear this out).\\nIn[18]: fig, ax = subplots(figsize=(8,8))\\nplot_svm(X,\\ny,\\nsvm_,\\nax=ax)\\n9.6.2 Support Vector Machine\\nIn order to fit an SVM using a non-linear kernel, we once again use the\\nSVC()estimator. However, now we use a different value of the parameter\\nkernel. To fit an SVM with a polynomial kernel we use kernel=\"poly\" , and\\nto fit an SVM with a radial kernel we use kernel=\"rbf\" . In the former case\\nwe also use the degreeargument to specify a degree for the polynomial\\nkernel (this is din (9.22)), and in the latter case we use gammato specify a\\nvalue ofγfor the radial basis kernel ( 9.24).\\nWe first generate some data with a non-linear class boundary, as follows:\\nIn[19]: X=r n g . s t a n d a r d _ n o r m a l ( ( 2 0 0 ,2 ) )\\nX[:100] += 2\\nX[100:150] -= 2\\ny=n p . a r r a y ( [ 1 ] * 1 5 0 + [ 2 ] * 5 0 )\\nPlotting the data makes it clear that the class boundary is indeed non-\\nlinear.\\nIn[20]: fig, ax = subplots(figsize=(8,8))\\nax.scatter(X[:,0],\\nX[:,1],\\nc=y,\\ncmap=cm.coolwarm)\\nOut[20]: <matplotlib.collections.PathCollection at 0x7faa9ba52eb0>', metadata={'source': './documents/ISLP_website.pdf', 'page': 396}), Document(page_content='9.6 Lab: Support Vector Machines 391The data is randomly split into training and testing groups. We then fitthe training data using theSVC()estimator with a radial kernel andγ=1:In[21]:(X_train,X_test,y_train,y_test) = skm.train_test_split(X,y,test_size=0.5,random_state=0)svm_rbf = SVC(kernel=\"rbf\",g a m m a = 1 ,C = 1 )svm_rbf.fit(X_train, y_train)TheplotshowsthattheresultingSVMhasadecidedlynon-linearbound-ary.In[22]:fig, ax = subplots(figsize=(8,8))plot_svm(X_train,y_train,svm_rbf,ax=ax)We can see from the figure that there are a fair number of training errorsin this SVM fit. If we increase the value ofC, we can reduce the numberof training errors. However, this comes at the price of a more irregulardecision boundary that seems to be at risk of overfitting the data.In[23]:svm_rbf = SVC(kernel=\"rbf\",g a m m a = 1 ,C = 1 e 5 )svm_rbf.fit(X_train, y_train)fig, ax = subplots(figsize=(8,8))plot_svm(X_train,y_train,svm_rbf,ax=ax)We can perform cross-validation usingskm.GridSearchCV()to select thebest choice ofγandCfor an SVM with a radial kernel:In[24]:kfold = skm.KFold(5,random_state=0,shuffle=True)grid = skm.GridSearchCV(svm_rbf,{\\'C\\':[0.1,1,10,100,1000],\\'gamma\\':[0.5,1,2,3,4]},refit=True,cv=kfold,scoring=\\'accuracy\\');grid.fit(X_train, y_train)grid.best_params_Out[24]:{\\'C\\':1 0 0 ,\\'gamma\\':1 }The best choice of parameters under five-fold CV is achieved atC=1andgamma=0.5, though several other values also achieve the same value.In[25]:best_svm = grid.best_estimator_fig, ax = subplots(figsize=(8,8))plot_svm(X_train,', metadata={'source': './documents/ISLP_website.pdf', 'page': 397}), Document(page_content='392 9. Support Vector Machines\\ny_train,\\nbest_svm,\\nax=ax)\\ny_hat_test = best_svm.predict(X_test)\\nconfusion_table(y_hat_test, y_test)\\nOut[25]: Truth 1 2\\nPredicted\\n16 9 6\\n26 1 9\\nWith these parameters, 12% of test observations are misclassified by this\\nSVM.\\n9.6.3 ROC Curves\\nSVMsandsupportvectorclassifiersoutputclasslabelsforeachobservation.\\nHowever, it is also possible to obtain fitted values for each observation,\\nwhich are the numerical scores used to obtain the class labels. For instance,\\nin the case of a support vector classifier, the fitted value for an observation\\nX=(X1,X2,...,X p)Ttakes the form ˆβ0+ˆβ1X1+ˆβ2X2+...+ˆβpXp. For\\nan SVM with a non-linear kernel, the equation that yields the fitted value\\nis given in ( 9.23). The sign of the fitted value determines on which side\\nof the decision boundary the observation lies. Therefore, the relationship\\nbetween the fitted value and the class prediction for a given observation\\nis simple: if the fitted value exceeds zero then the observation is assigned\\nto one class, and if it is less than zero then it is assigned to the other.\\nBy changing this threshold from zero to some positive value, we skew the\\nclassifications in favor of one class versus the other. By considering a range\\nof these thresholds, positive and negative, we produce the ingredients for a\\nROC plot. We can access these values by calling the decision_function().function_\\ndecision() method of a fitted SVM estimator.\\nThe function ROCCurveDisplay.from_estimator() (which we have abbre-\\nviated to roc_curve() ) will produce a plot of a ROC curve. It takes a fittedroc_curve()estimator as its first argument, followed by a model matrix Xand labels y.\\nThe argument nameis used in the legend, while coloris used for the color\\nof the line. Results are plotted on our axis object ax.\\nIn[26]: fig, ax = subplots(figsize=(8,8))\\nroc_curve(best_svm,\\nX_train,\\ny_train,\\nname=\\'Training\\' ,\\ncolor=\\'r\\',\\nax=ax);\\nIn this example, the SVM appears to provide accurate predictions. By\\nincreasing γwe can produce a more flexible fit and generate further im-\\nprovements in accuracy.\\nIn[27]: svm_flex = SVC(kernel= \"rbf\",\\ngamma=50,', metadata={'source': './documents/ISLP_website.pdf', 'page': 398}), Document(page_content=\"9.6 Lab: Support Vector Machines 393\\nC=1)\\nsvm_flex.fit(X_train, y_train)\\nfig, ax = subplots(figsize=(8,8))\\nroc_curve(svm_flex,\\nX_train,\\ny_train,\\nname='Training $\\\\gamma=50$' ,\\ncolor='r',\\nax=ax);\\nHowever, these ROC curves are all on the training data. We are really\\nmore interested in the level of prediction accuracy on the test data. When\\nwe compute the ROC curves on the test data, the model with γ=0.5\\nappears to provide the most accurate results.\\nIn[28]: roc_curve(svm_flex,\\nX_test,\\ny_test,\\nname='Test $\\\\gamma=50$' ,\\ncolor='b',\\nax=ax)\\nfig;\\nLet’s look at our tuned SVM.\\nIn[29]: fig, ax = subplots(figsize=(8,8))\\nfor (X_, y_, c, name) in zip(\\n(X_train, X_test),\\n(y_train, y_test),\\n('r','b'),\\n('CV tuned on training' ,\\n'CV tuned on test' )):\\nroc_curve(best_svm,\\nX_,\\ny_,\\nname=name,\\nax=ax,\\ncolor=c)\\n9.6.4 SVM with Multiple Classes\\nIf the response is a factor containing more than two levels, then the SVC()\\nfunction will perform multi-class classification using either the one-versus-\\none approach (when decision_function_shape=='ovo' ) or one-versus-rest4\\n(whendecision_function_shape=='ovr' ). We explore that setting briefly\\nhere by generating a third class of observations.\\nIn[30]: rng = np.random.default_rng(123)\\nX=n p . v s t a c k ( [ X ,r n g . s t a n d a r d _ n o r m a l ( ( 5 0 ,2 ) ) ] )\\ny=n p . h s t a c k ( [ y ,[ 0 ] * 5 0 ] )\\nX[y==0,1] += 2\\nfig, ax = subplots(figsize=(8,8))\\nax.scatter(X[:,0], X[:,1], c=y, cmap=cm.coolwarm);\\n4One-versus-rest is also known as one-versus-all.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 399}), Document(page_content='394 9. Support Vector MachinesWe now fit an SVM to the data:In[31]:svm_rbf_3 = SVC(kernel=\"rbf\",C=10,gamma=1,decision_function_shape=\\'ovo\\');svm_rbf_3.fit(X, y)fig, ax = subplots(figsize=(8,8))plot_svm(X,y,svm_rbf_3,scatter_cmap=cm.tab10,ax=ax)Thesklearn.svmlibrary can also be used to perform support vector re-gression with a numerical response using the estimatorSupportVector-Regression().SupportVectorRegression()9.6.5 Application to Gene Expression DataWe now examine theKhandata set, which consists of a number of tissuesamples corresponding to four distinct types of small round blue cell tu-mors. For each tissue sample, gene expression measurements are available.The data set consists of training data,xtrainandytrain, and testing data,xtestandytest.We examine the dimension of the data:In[32]:Khan = load_data(\\'Khan\\')Khan[\\'xtrain\\'].shape, Khan[\\'xtest\\'].shapeOut[32]:((63, 2308), (20, 2308))This data set consists of expression measurements for 2,308 genes. Thetraining and test sets consist of 63 and 20 observations, respectively.We will use a support vector approach to predict cancer subtype usinggeneexpressionmeasurements.Inthisdataset,thereisaverylargenumberof features relative to the number of observations. This suggests that weshould use a linear kernel, because the additional flexibility that will resultfrom using a polynomial or radial kernel is unnecessary.In[33]:khan_linear = SVC(kernel=\\'linear\\',C = 1 0 )khan_linear.fit(Khan[\\'xtrain\\'], Khan[\\'ytrain\\'])confusion_table(khan_linear.predict(Khan[\\'xtrain\\']),Khan[\\'ytrain\\'])Out[33]:Truth 1 2 3 4Predicted18 0 0 020 2 3 0 030 0 1 2 040 0 0 2 0We see that there arenotraining errors. In fact, this is not surprising,becausethelargenumberofvariablesrelativetothenumberofobservationsimplies that it is easy to find hyperplanes that fully separate the classes.', metadata={'source': './documents/ISLP_website.pdf', 'page': 400}), Document(page_content=\"9.7 Exercises 395We are more interested in the support vector classifier’s performance onthe test observations.In[34]:confusion_table(khan_linear.predict(Khan['xtest']),Khan['ytest'])Out[34]:Truth 1 2 3 4Predicted13 0 0 020 6 2 030 0 4 040 0 0 5We see that usingC=10yields two test set errors on these data.9.7 ExercisesConceptual1. This problem involves hyperplanes in two dimensions.(a) Sketch the hyperplane1+3X1−X2=0. Indicate the set ofpoints for which1+3X1−X2>0, as well as the set of pointsfor which1+3X1−X2<0.(b) On the same plot, sketch the hyperplane−2+X1+2X2=0.Indicate the set of points for which−2+X1+2X2>0, as wellas the set of points for which−2+X1+2X2<0.2. We have seen that inp=2dimensions, a linear decision boundarytakes the formβ0+β1X1+β2X2=0. We now investigate a non-lineardecision boundary.(a) Sketch the curve(1 +X1)2+ (2−X2)2=4.(b) On your sketch, indicate the set of points for which(1 +X1)2+ (2−X2)2>4,as well as the set of points for which(1 +X1)2+ (2−X2)2≤4.(c) Suppose that a classifier assigns an observation to the blue classif(1 +X1)2+ (2−X2)2>4,and to the red class otherwise. To what class is the observation(0,0)classified?(−1,1)?(2,2)?(3,8)?(d) Argue that while the decision boundary in (c) is not linear interms ofX1andX2, it is linear in terms ofX1,X21,X2, andX22.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 401}), Document(page_content='396 9. Support Vector Machines\\n3. Here we explore the maximal margin classifier on a toy data set.\\n(a) We are given n=7observations in p=2dimensions. For each\\nobservation, there is an associated class label.\\nObs.X1X2 Y\\n1 3 4 Red\\n2 2 2 Red\\n3 4 4 Red\\n4 1 4 Red\\n5 2 1 Blue\\n6 4 3 Blue\\n7 4 1 Blue\\nSketch the observations.\\n(b) Sketchtheoptimalseparatinghyperplane,andprovidetheequa-\\ntion for this hyperplane (of the form ( 9.1)).\\n(c) Describe the classification rule for the maximal margin classifier.\\nIt should be something along the lines of “Classify to Red if\\nβ0+β1X1+β2X2>0, and classify to Blue otherwise.” Provide\\nthe values for β0,β1, andβ2.\\n(d) On your sketch, indicate the margin for the maximal margin\\nhyperplane.\\n(e) Indicate the support vectors for the maximal margin classifier.\\n(f) Argue that a slight movement of the seventh observation would\\nnot affect the maximal margin hyperplane.\\n(g) Sketch a hyperplane that is notthe optimal separating hyper-\\nplane, and provide the equation for this hyperplane.\\n(h) Draw an additional observation on the plot so that the two\\nclasses are no longer separable by a hyperplane.\\nApplied\\n4. Generate a simulated two-class data set with 100 observations and\\ntwo features in which there is a visible but non-linear separation be-\\ntween the two classes. Show that in this setting, a support vector\\nmachine with a polynomial kernel (with degree greater than 1) or a\\nradial kernel will outperform a support vector classifier on the train-\\ning data. Which technique performs best on the test data? Make\\nplots and report training and test error rates in order to back up\\nyour assertions.\\n5. WehaveseenthatwecanfitanSVMwithanon-linearkernelinorder\\ntoperformclassificationusinganon-lineardecisionboundary.Wewill\\nnow see that we can also obtain a non-linear decision boundary by\\nperforming logistic regression using non-linear transformations of the\\nfeatures.', metadata={'source': './documents/ISLP_website.pdf', 'page': 402}), Document(page_content='9.7 Exercises 397\\n(a) Generateadatasetwith n= 500 andp=2,suchthattheobser-\\nvationsbelong totwoclasses witha quadratic decision boundary\\nbetween them. For instance, you can do this as follows:\\nrng = np.random.default_rng(5)\\nx1 = rng.uniform(size=500) - 0.5\\nx2 = rng.uniform(size=500) - 0.5\\ny=x 1 * * 2-x 2 * * 2>0\\n(b) Plot the observations, colored according to their class labels.\\nYour plot should display X1on thex-axis, and X2on they-\\naxis.\\n(c) Fit a logistic regression model to the data, using X1andX2as\\npredictors.\\n(d) Apply this model to the training data in order to obtain a pre-\\ndicted class label for each training observation. Plot the ob-\\nservations, colored according to the predicted class labels. The\\ndecision boundary should be linear.\\n(e) Now fit a logistic regression model to the data using non-linear\\nfunctions of X1andX2as predictors (e.g. X2\\n1,X1×X2,log(X2),\\nand so forth).\\n(f) Apply this model to the training data in order to obtain a pre-\\ndicted class label for each training observation. Plot the ob-\\nservations, colored according to the predicted class labels. The\\ndecision boundary should be obviously non-linear. If it is not,\\nthen repeat (a)–(e) until you come up with an example in which\\nthe predicted class labels are obviously non-linear.\\n(g) Fit a support vector classifier to the data with X1andX2as\\npredictors. Obtain a class prediction for each training observa-\\ntion. Plot the observations, colored according to the predicted\\nclass labels .\\n(h) Fit a SVM using a non-linear kernel to the data. Obtain a class\\nprediction for each training observation. Plot the observations,\\ncolored according to the predicted class labels .\\n(i) Comment on your results.\\n6. AttheendofSection 9.6.1,itisclaimedthatinthecaseofdatathatis\\njust barely linearly separable, a support vector classifier with a small\\nvalue of Cthat misclassifies a couple of training observations may\\nperform better on test data than one with a huge value of Cthat does\\nnot misclassify any training observations. You will now investigate\\nthis claim.\\n(a) Generatetwo-classdatawith p=2insuchawaythattheclasses\\nare just barely linearly separable.\\n(b) Compute the cross-validation error rates for support vector\\nclassifiers with a range of Cvalues. How many training obser-\\nvations are misclassified for each value of Cconsidered, and how\\ndoes this relate to the cross-validation errors obtained?', metadata={'source': './documents/ISLP_website.pdf', 'page': 403}), Document(page_content='398 9. Support Vector Machines(c) Generate an appropriate test data set, and compute the testerrorscorrespondingtoeachofthevaluesofCconsidered.Whichvalue ofCleads to the fewest test errors, and how does thiscompare to the values ofCthat yield the fewest training errorsand the fewest cross-validation errors?(d) Discuss your results.7. In this problem, you will use support vector approaches in order topredict whether a given car gets high or low gas mileage based on theAutodata set.(a) Create a binary variable that takes on a1for cars with gasmileage above the median, and a0for cars with gas mileagebelow the median.(b) Fit a support vector classifier to the data with various values ofC, in order to predict whether a car gets high or low gas mileage.Report the cross-validation errors associated with different val-ues of this parameter. Comment on your results. Note you willneed to fit the classifier without the gas mileage variable to pro-duce sensible results.(c) Now repeat (b), this time using SVMs with radial and polyno-mial basis kernels, with different values ofgammaanddegreeandC. Comment on your results.(d) Make some plots to back up your assertions in (b) and (c).Hint: In the lab, we used theplot_svm()function for fitted SVMs.Whenp>2, you can use the keyword argumentfeaturestocreate plots displaying pairs of variables at a time.8. This problem involves theOJdata set which is part of theISLPpackage.(a) Create a training set containing a random sample of800observations, and a test set containing the remainingobservations.(b) Fit a support vector classifier to the training data usingC = 0.01, withPurchaseas the response and the other variablesas predictors. How many support points are there?(c) What are the training and test error rates?(d) Use cross-validation to select an optimalC. Consider values inthe range0.01to10.(e) Compute the training and test error rates using this new valueforC.(f) Repeat parts (b) through (e) using a support vector machinewith a radial kernel. Use the default value forgamma.(g) Repeat parts (b) through (e) using a support vector machinewith a polynomial kernel. Setdegree = 2.(h) Overall, which approach seems to give the best results on thisdata?', metadata={'source': './documents/ISLP_website.pdf', 'page': 404}), Document(page_content='10\\nDeep Learning\\nThis chapter covers the important topic of deep learning . At the time ofdeep\\nlearningwriting(2020),deeplearningisaveryactiveareaofresearchinthemachine\\nlearning and artificial intelligence communities. The cornerstone of deep\\nlearning is the neural network .neural\\nnetworkNeural networks rose to fame in the late 1980s. There was a lot of excite-\\nment and a certain amount of hype associated with this approach, and they\\nwere the impetus for the popular Neural Information Processing Systems\\nmeetings (NeurIPS, formerly NIPS) held every year, typically in exotic\\nplaces like ski resorts. This was followed by a synthesis stage, where the\\nproperties of neural networks were analyzed by machine learners, math-\\nematicians and statisticians; algorithms were improved, and the method-\\nology stabilized. Then along came SVMs, boosting, and random forests,\\nand neural networks fell somewhat from favor. Part of the reason was that\\nneural networks required a lot of tinkering, while the new methods were\\nmore automatic. Also, on many problems the new methods outperformed\\npoorly-trained neural networks. This was the status quo for the first decade\\nin the new millennium.\\nAll the while, though, a core group of neural-network enthusiasts were\\npushingtheirtechnologyharderonever-largercomputingarchitecturesand\\ndata sets. Neural networks resurfaced after 2010 with the new name deep\\nlearning, with new architectures, additional bells and whistles, and a string\\nof success stories on some niche problems such as image and video classifi-\\ncation, speech and text modeling. Many in the field believe that the major\\nreason for these successes is the availability of ever-larger training datasets,\\nmade possible by the wide-scale use of digitization in science and industry.\\nIn this chapter we discuss the basics of neural networks and deep learn-\\ning, and then go into some of the specializations for specific problems, such\\nas convolutional neural networks (CNNs) for image classification, and re-\\ncurrent neural networks (RNNs) for time series and other sequences. We\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning , Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_10 399', metadata={'source': './documents/ISLP_website.pdf', 'page': 405}), Document(page_content='400 10. Deep Learning\\nX1X2X3X4A1A2A3A4A5f(X)YHiddenLayerInputLayerOutputLayer\\nFIGURE 10.1. Neural network with a single hidden layer. The hidden layer\\ncomputes activations Ak=hk(X)that are nonlinear transformations of linear\\ncombinations of the inputs X1,X2,...,X p. Hence these Akare not directly ob-\\nserved. The functions hk(·)are not fixed in advance, but are learned during the\\ntraining of the network. The output layer is a linear model that uses these acti-\\nvationsAkas inputs, resulting in a function f(X).\\nwill also demonstrate these models using the Python torch package, along\\nwith a number of helper packages.\\nThe material in this chapter is slightly more challenging than elsewhere\\nin this book.\\n10.1 Single Layer Neural Networks\\nA neural network takes an input vector of pvariables X=(X1,X2,...,X p)\\nand builds a nonlinear function f(X)to predict the response Y. We have\\nbuilt nonlinear prediction models in earlier chapters, using trees, boosting\\nand generalized additive models. What distinguishes neural networks from\\nthese methods is the particular structure of the model. Figure 10.1shows\\na simple feed-forward neural network for modeling a quantitative responsefeed-forward\\nneural\\nnetworkusingp=4predictors. In the terminology of neural networks, the four fea-\\nturesX1,...,X 4make up the units in the input layer . The arrows indicate\\ninput layer that each of the inputs from the input layer feeds into each of the Khidden\\nunits(we get to pick K; here we chose 5). The neural network model hashidden unitsthe form\\nf(X)= β0+∑K\\nk=1βkhk(X)\\n=β0+∑K\\nk=1βkg(wk0+∑p\\nj=1wkjXj).(10.1)\\nIt is built up here in two steps. First the Kactivations Ak,k=1,...,K , inactivationsthehiddenlayerarecomputedasfunctionsoftheinputfeatures X1,...,X p,\\nAk=hk(X)=g(wk0+∑p\\nj=1wkjXj), (10.2)', metadata={'source': './documents/ISLP_website.pdf', 'page': 406}), Document(page_content='10.1 Single Layer Neural Networks 401\\n−4 −2 0240.0 0.2 0.4 0.6 0.8 1.0zg(z)sigmoidReLU\\nFIGURE 10.2. Activation functions. The piecewise-linear ReLUfunction is pop-\\nular for its efficiency and computability. We have scaled it down by a factor of\\nfive for ease of comparison.\\nwhereg(z)is a nonlinear activation function that is specified in advance.activation\\nfunctionWe can think of each Akas a different transformation hk(X)of the original\\nfeatures, much like the basis functions of Chapter 7. These Kactivations\\nfrom the hidden layer then feed into the output layer, resulting in\\nf(X)=β0+K∑\\nk=1βkAk, (10.3)\\na linear regression model in the K=5 activations. All the parameters\\nβ0,...,βKandw10,...,w Kpneed to be estimated from data. In the early\\ninstances of neural networks, the sigmoidactivation function was favored,sigmoid\\ng(z)=ez\\n1+ez=1\\n1+e−z, (10.4)\\nwhich is the same function used in logistic regression to convert a linear\\nfunction into probabilities between zero and one (see Figure 10.2). The\\npreferred choice in modern neural networks is the ReLU(rectified linearReLUunit) activation function, which takes the formrectified\\nlinear unit\\ng(z)=( z)+={0ifz<0\\nzotherwise .(10.5)\\nA ReLU activation can be computed and stored more efficiently than a\\nsigmoid activation. Although it thresholds at zero, because we apply it to a\\nlinear function ( 10.2) the constant term wk0will shift this inflection point.\\nSo in words, the model depicted in Figure 10.1derives five new features\\nby computing five different linear combinations of X, and then squashes\\neach through an activation function g(·)to transform it. The final model\\nis linear in these derived variables.\\nThename neuralnetwork originallyderivedfromthinkingofthesehidden\\nunits as analogous to neurons in the brain — values of the activations\\nAk=hk(X)close to one are firing, while those close to zero are silent\\n(using the sigmoid activation function).\\nThe nonlinearity in the activation function g(·)is essential, since without\\nit the model f(X)in (10.1) would collapse into a simple linear model in', metadata={'source': './documents/ISLP_website.pdf', 'page': 407}), Document(page_content='402 10. Deep Learning\\nX1,...,X p. Moreover, having a nonlinear activation function allows the\\nmodel to capture complex nonlinearities and interaction effects. Consider\\na very simple example with p=2 input variables X=(X1,X2), and\\nK=2hidden units h1(X)andh2(X)withg(z)=z2. We specify the other\\nparameters as\\nβ0=0,β1=1\\n4,β2=−1\\n4,\\nw10=0,w 11=1,w 12=1 ,\\nw20=0,w 21=1,w 22=−1.(10.6)\\nFrom (10.2), this means that\\nh1(X) = (0 + X1+X2)2,\\nh2(X) = (0 + X1−X2)2.(10.7)\\nThen plugging ( 10.7) into (10.1), we get\\nf(X)=0 +1\\n4·(0 + X1+X2)2−1\\n4·(0 + X1−X2)2\\n=1\\n4[\\n(X1+X2)2−(X1−X2)2]\\n=X1X2.(10.8)\\nSo the sum of two nonlinear transformations of linear functions can give\\nus an interaction! In practice we would not use a quadratic function for\\ng(z), since we would always get a second-degree polynomial in the original\\ncoordinates X1,...,X p. The sigmoid or ReLU activations do not have such\\na limitation.\\nFitting a neural network requires estimating the unknown parameters in\\n(10.1). For a quantitative response, typically squared-error loss is used, so\\nthat the parameters are chosen to minimize\\nn∑\\ni=1(yi−f(xi))2. (10.9)\\nDetailsabouthowtoperformthisminimizationareprovidedinSection 10.7.\\n10.2 Multilayer Neural Networks\\nModern neural networks typically have more than one hidden layer, and\\noften many units per layer. In theory a single hidden layer with a large\\nnumber of units has the ability to approximate most functions. However,\\nthe learning task of discovering a good solution is made much easier with\\nmultiple layers each of modest size.\\nWe will illustrate a large dense network on the famous and publicly\\navailable MNISThandwritten digit dataset.1Figure10.3shows examples of\\nthese digits. The idea is to build a model to classify the images into their\\ncorrect digit class 0–9. Every image has p= 28 ×28 = 784 pixels, each\\nof which is an eight-bit grayscale value between 0 and 255 representing\\n1See LeCun, Cortes, and Burges (2010) “The MNIST database of handwritten digits”,\\navailable at http://yann.lecun.com/exdb/mnist .', metadata={'source': './documents/ISLP_website.pdf', 'page': 408}), Document(page_content='10.2 Multilayer Neural Networks 403\\nFIGURE 10.3. Examples of handwritten digits from the MNISTcorpus. Each\\ngrayscale image has 28×28pixels, each of which is an eight-bit number (0–255)\\nwhich represents how dark that pixel is. The first 3, 5, and 8 are enlarged to show\\ntheir 784 individual pixel values.\\nthe relative amount of the written digit in that tiny square.2These pixels\\nare stored in the input vector X(in, say, column order). The output is\\nthe class label, represented by a vector Y=(Y0,Y1,...,Y 9)of 10 dummy\\nvariables, with a one in the position corresponding to the label, and zeros\\nelsewhere. In the machine learning community, this is known as one-hot\\nencoding. There are 60,000 training images, and 10,000 test images.one-hot\\nencodingOn a historical note, digit recognition problems were the catalyst that\\naccelerated the development of neural network technology in the late 1980s\\nat AT&T Bell Laboratories and elsewhere. Pattern recognition tasks of this\\nkind are relatively simple for humans. Our visual system occupies a large\\nfraction of our brains, and good recognition is an evolutionary force for\\nsurvival. These tasks are not so simple for machines, and it has taken more\\nthan 30 years to refine the neural-network architectures to match human\\nperformance.\\nFigure10.4shows a multilayer network architecture that works well for\\nsolving the digit-classification task. It differs from Figure 10.1in several\\nways:\\n• It has two hidden layers L1(256 units) and L2(128 units) rather\\nthan one. Later we will see a network with seven hidden layers.\\n• It has ten output variables, rather than one. In this case the ten vari-\\nables really represent a single qualitative variable and so are quite\\ndependent. (We have indexed them by the digit class 0–9 rather than\\n1–10, for clarity.) More generally, in multi-task learning one can pre-multi-task\\nlearningdict different responses simultaneously with a single network; they\\nall have a say in the formation of the hidden layers.\\n• The loss function used for training the network is tailored for the\\nmulticlass classification task.\\n2In the analog-to-digital conversion process, only part of the written numeral may\\nfall in the square representing a particular pixel.', metadata={'source': './documents/ISLP_website.pdf', 'page': 409}), Document(page_content='404 10. Deep LearningX1X2X3X4X5X6...XpA(1)1A(1)2A(1)3A(1)4...A(1)K1A(2)1A(2)2A(2)3...A(2)K2f0(X)Y0f1(X)Y1......f9(X)Y9HiddenlayerL2HiddenlayerL1InputlayerOutputlayer\\nW1W2BFIGURE 10.4. Neural network diagram with two hidden layers and multiple\\noutputs, suitable for the MNISThandwritten-digit problem. The input layer has\\np= 784 units, the two hidden layers K1= 256 andK2= 128 units respectively,\\nand the output layer 10units. Along with intercepts (referred to as biasesin the\\ndeep-learning community) this network has 235,146 parameters (referred to as\\nweights).\\nThe first hidden layer is as in ( 10.2), with\\nA(1)\\nk=h(1)\\nk(X)\\n=g(w(1)\\nk0+∑p\\nj=1w(1)\\nkjXj)(10.10)\\nfork=1,...,K 1. The second hidden layer treats the activations A(1)\\nkof\\nthe first hidden layer as inputs and computes new activations\\nA(2)\\nℓ=h(2)\\nℓ(X)\\n=g(w(2)\\nℓ0+∑K1\\nk=1w(2)\\nℓkA(1)\\nk)(10.11)\\nforℓ=1,...,K 2.Notice that each of the activations in the second layer\\nA(2)\\nℓ=h(2)\\nℓ(X)is a function of the input vector X. This is the case because\\nwhile they are explicitly a function of the activations A(1)\\nkfrom layer L1,\\nthese in turn are functions of X. This would also be the case with more\\nhidden layers. Thus, through a chain of transformations, the network is\\nable to build up fairly complex transformations of Xthat ultimately feed\\ninto the output layer as features.\\nWe have introduced additional superscript notation such as h(2)\\nℓ(X)and\\nw(2)\\nℓjin (10.10) and (10.11) to indicate to which layer the activations and\\nweights(coefficients) belong, in this case layer 2. The notation W1in Fig-weights', metadata={'source': './documents/ISLP_website.pdf', 'page': 410}), Document(page_content='10.2 Multilayer Neural Networks 405\\nure10.4represents the entire matrix of weights that feed from the input\\nlayer to the first hidden layer L1. This matrix will have 785×256 = 200 ,960\\nelements; there are 785 rather than 784 because we must account for the\\nintercept or biasterm.3\\nbiasEach element A(1)\\nkfeeds to the second hidden layer L2via the matrix of\\nweightsW2of dimension 257×128 = 32 ,896.\\nWe now get to the output layer, where we now have ten responses rather\\nthan one. The first step is to compute ten different linear models similar\\nto our single model ( 10.1),\\nZm=βm0+∑K2\\nℓ=1βmℓh(2)\\nℓ(X)\\n=βm0+∑K2\\nℓ=1βmℓA(2)\\nℓ,(10.12)\\nform=0,1,...,9. The matrix Bstores all 129×10 = 1 ,290of these\\nweights.\\nIf these were all separate quantitative responses, we would simply set\\neachfm(X)=Zmand be done. However, we would like our estimates to\\nrepresent class probabilities fm(X) = Pr( Y=m|X), just like in multi-\\nnomial logistic regression in Section 4.3.5. So we use the special softmaxsoftmaxactivation function (see ( 4.13) on page 145),\\nfm(X) = Pr( Y=m|X)=eZm\\n∑9\\nℓ=0eZℓ, (10.13)\\nform=0,1,...,9. This ensures that the 10 numbers behave like proba-\\nbilities (non-negative and sum to one). Even though the goal is to build\\na classifier, our model actually estimates a probability for each of the 10\\nclasses. The classifier then assigns the image to the class with the highest\\nprobability.\\nTo train this network, since the response is qualitative, we look for coef-\\nficient estimates that minimize the negative multinomial log-likelihood\\n−n∑\\ni=19∑\\nm=0yimlog(fm(xi)), (10.14)\\nalso known as the cross-entropy . This is a generalization of the crite-cross-\\nentropy rion (4.5) for two-class logistic regression. Details on how to minimize this\\nobjective are given in Section 10.7. If the response were quantitative, we\\nwould instead minimize squared-error loss as in ( 10.9).\\nTable10.1compares the test performance of the neural network with\\ntwo simple models presented in Chapter 4that make use of linear decision\\nboundaries:multinomiallogisticregressionandlineardiscriminantanalysis.\\nThe improvement of neural networks over both of these linear methods is\\ndramatic:thenetworkwithdropoutregularizationachievesatesterrorrate\\nbelow 2% on the 10,000test images. (We describe dropout regularization in\\nSection10.7.3.) In Section 10.9.2of the lab, we present the code for fitting\\nthis model, which runs in just over two minutes on a laptop computer.\\n3The use of “weights” for coefficients and “bias” for the intercepts wk0in (10.2) is\\npopular in the machine learning community; this use of bias is not to be confused with\\nthe “bias-variance” usage elsewhere in this book.', metadata={'source': './documents/ISLP_website.pdf', 'page': 411}), Document(page_content='406 10. Deep Learning\\nMethod Test Error\\nNeural Network + Ridge Regularization 2.3%\\nNeural Network + Dropout Regularization 1.8%\\nMultinomial Logistic Regression 7.2%\\nLinear Discriminant Analysis 12.7%\\nTABLE 10.1. Test error rate on the MNISTdata, for neural networks with two\\nforms of regularization, as well as multinomial logistic regression and linear dis-\\ncriminant analysis. In this example, the extra complexity of the neural network\\nleads to a marked improvement in test error.\\nFIGURE 10.5. A sample of images from the CIFAR100 database: a collection of\\nnatural images from everyday life, with 100 different classes represented.\\nAdding the number of coefficients in W1,W2andB, we get235,146in\\nall, more than 33 times the number 785×9=7,065needed for multinomial\\nlogistic regression. Recall that there are 60,000images in the training set.\\nWhile this might seem like a large training set, there are almost four times\\nasmanycoefficientsintheneuralnetworkmodelasthereareobservationsin\\nthe training set! To avoid overfitting, some regularization is needed. In this\\nexample, we used two forms of regularization: ridge regularization, which\\nis similar to ridge regression from Chapter 6, anddropoutregularization.dropoutWe discuss both forms of regularization in Section 10.7.\\n10.3 Convolutional Neural Networks\\nNeural networks rebounded around 2010 with big successes in image classi-\\nfication. Around that time, massive databases of labeled images were being\\naccumulated, with ever-increasing numbers of classes. Figure 10.5shows\\n75 images drawn from the CIFAR100 database.4This database consists of\\n60,000 images labeled according to 20 superclasses (e.g. aquatic mammals),\\nwith five classes per superclass (beaver, dolphin, otter, seal, whale). Each\\nimage has a resolution of 32×32pixels, with three eight-bit numbers per\\npixel representing red, green and blue. The numbers for each image are\\norganized in a three-dimensional array called a feature map . The first twofeature map\\n4See Chapter 3 of Krizhevsky (2009) “Learning multiple layers of fea-\\ntures from tiny images”, available at https://www.cs.toronto.edu/~kriz/\\nlearning-features-2009-TR.pdf .', metadata={'source': './documents/ISLP_website.pdf', 'page': 412}), Document(page_content='10.3 Convolutional Neural Networks 407\\nFIGURE 10.6. Schematic showing how a convolutional neural network classifies\\nan image of a tiger. The network takes in the image and identifies local features.\\nIt then combines the local features in order to create compound features, which in\\nthis example include eyes and ears. These compound features are used to output\\nthe label “tiger”.\\naxes are spatial (both are 32-dimensional), and the third is the channelchannelaxis,5representing the three colors. There is a designated training set of\\n50,000 images, and a test set of 10,000.\\nA special family of convolutional neural networks (CNNs) has evolved forconvolutional\\nneural\\nnetworksclassifying images such as these, and has shown spectacular success on a\\nwide range of problems. CNNs mimic to some degree how humans classify\\nimages, by recognizing specific features or patterns anywhere in the image\\nthat distinguish each particular object class. In this section we give a brief\\noverview of how they work.\\nFigure10.6illustrates the idea behind a convolutional neural network on\\na cartoon image of a tiger.6\\nThe network first identifies low-level features in the input image, such\\nas small edges, patches of color, and the like. These low-level features are\\nthen combined to form higher-level features, such as parts of ears, eyes,\\nand so on. Eventually, the presence or absence of these higher-level features\\ncontributes to the probability of any given output class.\\nHowdoesaconvolutionalneuralnetworkbuildupthishierarchy?Itcom-\\nbines two specialized types of hidden layers, called convolution layers and\\npoolinglayers. Convolution layers search for instances of small patterns in\\nthe image, whereas pooling layers downsample these to select a prominent\\nsubset. In order to achieve state-of-the-art results, contemporary neural-\\nnetwork architectures make use of many convolution and pooling layers.\\nWe describe convolution and pooling layers next.\\n10.3.1 Convolution Layers\\nAconvolution layer is made up of a large number of convolution filters , eachconvolution\\nlayer\\nconvolution\\nfilter5The term channel is taken from the signal-processing literature. Each channel is a\\ndistinct source of information.\\n6Thanks to Elena Tuzhilina for producing the diagram and https://www.\\ncartooning4kids.com/ for permission to use the cartoon tiger.', metadata={'source': './documents/ISLP_website.pdf', 'page': 413}), Document(page_content='408 10. Deep Learning\\nof which is a template that determines whether a particular local feature is\\npresent in an image. A convolution filter relies on a very simple operation,\\ncalled a convolution , which basically amounts to repeatedly multiplying\\nmatrix elements and then adding the results.\\nTo understand how a convolution filter works, consider a very simple\\nexample of a 4×3image:\\nOriginal Image =\\uf8ee\\n\\uf8ef\\uf8ef\\uf8f0abc\\ndef\\nghi\\njkl\\uf8f9\\n\\uf8fa\\uf8fa\\uf8fb.\\nNow consider a 2×2filter of the form\\nConvolution Filter =[αβ\\nγδ]\\n.\\nWhen we convolve the image with the filter, we get the result7\\nConvolved Image =\\uf8ee\\n\\uf8f0aα+bβ+dγ+eδbα+cβ+eγ+fδ\\ndα+eβ+gγ+hδeα+fβ+hγ+iδ\\ngα+hβ+jγ+kδhα+iβ+kγ+lδ\\uf8f9\\n\\uf8fb.\\nFor instance, the top-left element comes from multiplying each element in\\nthe2×2filter by the corresponding element in the top left 2×2portion\\nof the image, and adding the results. The other elements are obtained in a\\nsimilar way: the convolution filter is applied to every 2×2submatrix of the\\noriginal image in order to obtain the convolved image. If a 2×2submatrix\\nof the original image resembles the convolution filter, then it will have a\\nlargevalue in the convolved image; otherwise, it will have a smallvalue.\\nThus,the convolved image highlights regions of the original image that\\nresemble the convolution filter. We have used 2×2as an example; in\\ngeneral convolution filters are small ℓ1×ℓ2arrays, with ℓ1andℓ2small\\npositive integers that are not necessarily equal.\\nFigure10.7illustratestheapplicationoftwoconvolutionfilterstoa 192×\\n179image of a tiger, shown on the left-hand side.8Each convolution filter\\nis a15×15image containing mostly zeros (black), with a narrow strip\\nof ones (white) oriented either vertically or horizontally within the image.\\nWhen each filter is convolved with the image of the tiger, areas of the tiger\\nthat resemble the filter (i.e. that have either horizontal or vertical stripes or\\nedges) are givenlarge values,and areas of the tiger that do not resemble the\\nfeature are given small values. The convolved images are displayed on the\\nright-hand side. We see that the horizontal stripe filter picks out horizontal\\nstripes and edges in the original image, whereas the vertical stripe filter\\npicks out vertical stripes and edges in the original image.\\n7The convolved image is smaller than the original image because its dimension is\\ngiven by the number of 2×2submatrices in the original image. Note that 2×2is the\\ndimension of the convolution filter. If we want the convolved image to have the same\\ndimension as the original image, then padding can be applied.\\n8The tiger image used in Figures 10.7–10.9was obtained from the public domain\\nimage resource https://www.needpix.com/ .', metadata={'source': './documents/ISLP_website.pdf', 'page': 414}), Document(page_content='10.3 Convolutional Neural Networks 409\\nFIGURE 10.7. Convolution filters find local features in an image, such as edges\\nand small shapes. We begin with the image of the tiger shown on the left, and\\napply the two small convolution filters in the middle. The convolved images high-\\nlight areas in the original image where details similar to the filters are found.\\nSpecifically, the top convolved image highlights the tiger’s vertical stripes, whereas\\nthe bottom convolved image highlights the tiger’s horizontal stripes. We can think\\nof the original image as the input layer in a convolutional neural network, and\\nthe convolved images as the units in the first hidden layer.\\nWe have used a large image and two large filters in Figure 10.7for illus-\\ntration. For the CIFAR100 database there are 32×32color pixels per image,\\nand we use 3×3convolution filters.\\nIn a convolution layer, we use a whole bank of filters to pick out a variety\\nof differently-oriented edges and shapes in the image. Using predefined\\nfilters in this way is standard practice in image processing. By contrast,\\nwith CNNs the filters are learnedfor the specific classification task. We can\\nthink of the filter weights as the parameters going from an input layer to a\\nhidden layer, with one hidden unit for each pixel in the convolved image.\\nThis is in fact the case, though the parameters are highly structured and\\nconstrained (see Exercise 4for more details). They operate on localized\\npatches in the input image (so there are many structural zeros), and the\\nsameweightsina givenfilter are reusedfor allpossible patchesin the image\\n(so the weights are constrained).9\\nWe now give some additional details.\\n• Since the input image is in color, it has three channels represented\\nby a three-dimensional feature map (array). Each channel is a two-\\ndimensional ( 32×32) feature map — one for red, one for green, and\\none for blue. A single convolution filter will also have three channels,\\none per color, each of dimension 3×3, with potentially different filter\\nweights. The results of the three convolutions are summed to form\\na two-dimensional output feature map. Note that at this point the\\ncolor information has been used, and is not passed on to subsequent\\nlayers except through its role in the convolution.\\n9This used to be called weight sharing in the early years of neural networks.', metadata={'source': './documents/ISLP_website.pdf', 'page': 415}), Document(page_content='410 10. Deep Learning\\n• If we use Kdifferent convolution filters at this first hidden layer,\\nwe getKtwo-dimensional output feature maps, which together are\\ntreated as a single three-dimensional feature map. We view each of\\ntheKoutput feature maps as a separate channel of information, so\\nnow we have Kchannels in contrast to the three color channels of\\nthe original input feature map. The three-dimensional feature map is\\njust like the activations in a hidden layer of a simple neural network,\\nexcept organized and produced in a spatially structured way.\\n• We typically apply the ReLU activation function ( 10.5) to the con-\\nvolved image. This step is sometimes viewed as a separate layer in\\nthe convolutional neural network, in which case it is referred to as a\\ndetector layer .detector\\nlayer\\n10.3.2 Pooling Layers\\nApoolinglayer provides a way to condense a large image into a smallerpoolingsummary image. While there are a number of possible ways to perform\\npooling, the max pooling operation summarizes each non-overlapping 2×2\\nblock of pixels in an image using the maximum value in the block. This\\nreduces the size of the image by a factor of two in each direction, and it\\nalso provides some location invariance : i.e. as long as there is a large value\\nin one of the four pixels in the block, the whole block registers as a large\\nvalue in the reduced image.\\nHere is a simple example of max pooling:\\nMax pool\\uf8ee\\n\\uf8ef\\uf8ef\\uf8f01253\\n3012\\n2134\\n1120\\uf8f9\\n\\uf8fa\\uf8fa\\uf8fb→[35\\n24]\\n.\\n10.3.3 Architecture of a Convolutional Neural Network\\nSo far we have defined a single convolution layer — each filter produces a\\nnew two-dimensional feature map. The number of convolution filters in a\\nconvolution layer is akin to the number of units at a particular hidden layer\\nin a fully-connected neural network of the type we saw in Section 10.2.\\nThis number also defines the number of channels in the resulting three-\\ndimensional feature map. We have also described a pooling layer, which\\nreduces the first two dimensions of each three-dimensional feature map.\\nDeep CNNs havemanysuchlayers.Figure 10.8showsa typicalarchitecture\\nfor a CNN for the CIFAR100 image classification task.\\nAt the input layer, we see the three-dimensional feature map of a color\\nimage, where the channel axis represents each color by a 32×32two-\\ndimensional feature map of pixels. Each convolution filter produces a new\\nchannel at the first hidden layer, each of which is a 32×32feature map\\n(after some padding at the edges). After this first round of convolutions, we\\nnow have a new “image”; a feature map with considerably more channels\\nthan the three color input channels (six in the figure, since we used six\\nconvolution filters).', metadata={'source': './documents/ISLP_website.pdf', 'page': 416}), Document(page_content='3216328164322500100convolveconvolveconvolvepoolpoolpoolflatten8FIGURE 10.8. Architecture of a deep CNN for the CIFAR100 classification task.\\nConvolution layers are interspersed with 2×2max-pool layers, which reduce the\\nsize by a factor of 2 in both dimensions.\\nThis is followed by a max-pool layer, which reduces the size of the feature\\nmap in each channel by a factor of four: two in each dimension.\\nThisconvolve-then-poolsequenceisnowrepeatedforthenexttwolayers.\\nSome details are as follows:\\n• Each subsequent convolve layer is similar to the first. It takes as input\\nthe three-dimensional feature map from the previous layer and treats\\nit like a single multi-channel image. Each convolution filter learned\\nhas as many channels as this feature map.\\n• Since the channel feature maps are reduced in size after each pool\\nlayer, we usually increase the number of filters in the next convolve\\nlayer to compensate.\\n• Sometimes we repeat several convolve layers before a pool layer. This\\neffectively increases the dimension of the filter.\\nThese operations are repeated until the pooling has reduced each channel\\nfeature map down to just a few pixels in each dimension. At this point the\\nthree-dimensional feature maps are flattened — the pixels are treated as\\nseparate units — and fed into one or more fully-connected layers before\\nreaching the output layer, which is a softmax activation for the 100 classes\\n(as in (10.13)).\\nThere are many tuning parameters to be selected in constructing such a\\nnetwork, apart from the number, nature, and sizes of each layer. Dropout\\nlearning can be used at each layer, as well as lasso or ridge regularization\\n(see Section 10.7). The details of constructing a convolutional neural net-\\nwork can seem daunting. Fortunately, terrific software is available, with\\nextensive examples and vignettes that provide guidance on sensible choices\\nfor the parameters. For the CIFAR100 official test set, the best accuracy as\\nof this writing is just above 75%, but undoubtedly this performance will\\ncontinue to improve.\\n10.3.4 Data Augmentation\\nAn additional important trick used with image modeling is data augment-data aug-\\nmentationation. Essentially, each training image is replicated many times, with each\\nreplicate randomly distorted in a natural way such that human recognition\\nis unaffected. Figure 10.9shows some examples. Typical distortions are10.3 Convolutional Neural Networks 411', metadata={'source': './documents/ISLP_website.pdf', 'page': 417}), Document(page_content='412 10. Deep Learning\\nFIGURE 10.9. Data augmentation. The original image (leftmost) is distorted\\nin natural ways to produce different images with the same class label. These\\ndistortions do not fool humans, and act as a form of regularization when fitting\\nthe CNN.\\nzoom, horizontal and vertical shift, shear, small rotations, and in this case\\nhorizontal flips. At face value this is a way of increasing the training set\\nconsiderably with somewhat different examples, and thus protects against\\noverfitting. In fact we can see this as a form of regularization: we build a\\ncloud of images around each original image, all with the same label. This\\nkind of fattening of the data is similar in spirit to ridge regularization.\\nWe will see in Section 10.7.2that the stochastic gradient descent al-\\ngorithms for fitting deep learning models repeatedly process randomly-\\nselected batches of, say, 128 training images at a time. This works hand-in-\\nglove with augmentation, because we can distort each image in the batch\\non the fly, and hence do not have to store all the new images.\\n10.3.5 Results Using a Pretrained Classifier\\nHere we use an industry-level pretrained classifier to predict the class of\\nsome new images. The resnet50 classifier is a convolutional neural network\\nthat was trained using the imagenet data set, which consists of millions of\\nimages that belong to an ever-growing number of categories.10Figure10.10\\ndemonstrates the performance of resnet50 on six photographs (private col-\\nlection of one of the authors).11The CNN does a reasonable job classifying\\nthe hawk in the second image. If we zoom out as in the third image, it\\ngets confused and chooses the fountain rather than the hawk. In the final\\nimage a “jacamar” is a tropical bird from South and Central America with\\nsimilar coloring to the South African Cape Weaver. We give more details\\non this example in Section 10.9.4.\\nMuch of the work in fitting a CNN is in learning the convolution filters\\nat the hidden layers; these are the coefficients of a CNN. For models fit to\\nmassive corpora such as imagenet with many classes, the output of these\\nfilters can serve as features for general natural-image classification prob-\\nlems. One can use these pretrained hidden layers for new problems with\\nmuch smaller training sets (a process referred to as weight freezing ), andweight\\nfreezingjust train the last few layers of the network, which requires much less data.\\n10For more information about resnet50 , see He, Zhang, Ren, and Sun (2015) “Deep\\nresidual learning for image recognition”, https://arxiv.org/abs/1512.03385 . For de-\\ntails about imagenet , see Russakovsky, Deng, et al. (2015) “ImageNet Large Scale\\nVisual Recognition Challenge”, in International Journal of Computer Vision .\\n11Theseresnet results can change with time, since the publicly-trained model gets\\nupdated periodically.', metadata={'source': './documents/ISLP_website.pdf', 'page': 418}), Document(page_content='10.4 Document Classification 413\\nflamingo Cooper’s hawk Cooper’s hawk\\nflamingo 0.83 kite 0.60fountain 0.35\\nspoonbill 0.17 great grey owl 0.09 nail 0.12\\nwhite stork 0.00 robin 0.06hook 0.07\\nLhasa Apso cat Cape weaver\\nTibetan terrier 0.56 Old English sheepdog 0.82 jacamar 0.28\\nLhasa 0.32Shih-Tzu 0.04macaw 0.12\\ncocker spaniel 0.03 Persian cat 0.04robin 0.12\\nFIGURE 10.10. Classification of six photographs using the resnet50 CNN\\ntrained on the imagenet corpus. The table below the images displays the true\\n(intended) label at the top of each panel, and the top three choices of the classifier\\n(out of 100). The numbers are the estimated probabilities for each choice. (A kite\\nis a raptor, but not a hawk.)\\nThe vignettes and book12that accompany the keraspackage give more\\ndetails on such applications.\\n10.4 Document Classification\\nIn this section we introduce a new type of example that has important\\napplications in industry and science: predicting attributes of documents.\\nExamples of documents include articles in medical journals, Reuters news\\nfeeds, emails, tweets, and so on. Our example will be IMDb(Internet Movie\\nDatabase) ratings — short documents where viewers have written critiques\\nof movies.13The response in this case is the sentiment of the review, which\\nwill bepositiveornegative.\\n12Deep Learning with R by F. Chollet and J.J. Allaire, 2018, Manning Publications.\\n13For details, see Maas et al. (2011) “Learning word vectors for sentiment analysis”,\\ninProceedings of the 49th Annual Meeting of the Association for Computational Lin-\\nguistics: Human Language Technologies , pages 142–150.', metadata={'source': './documents/ISLP_website.pdf', 'page': 419}), Document(page_content='414 10. Deep Learning\\nHere is the beginning of a rather amusing negative review:\\nThis has to be one of the worst films of the 1990s. When my\\nfriends & I were watching this film (being the target audience it\\nwas aimed at) we just sat & watched the first half an hour with\\nour jaws touching the floor at how bad it really was. The rest\\nof the time, everyone else in the theater just started talking to\\neach other, leaving or generally crying into their popcorn ...\\nEach review can be a different length, include slang or non-words, have\\nspelling errors, etc. We need to find a way to featurize such a document.featurizeThis is modern parlance for defining a set of predictors.\\nThe simplest and most common featurization is the bag-of-words model.bag-of-wordsWe score each document for the presence or absence of each of the words in\\na language dictionary — in this case an English dictionary. If the dictionary\\ncontains Mwords,thatmeansforeachdocumentwecreateabinaryfeature\\nvector of length M, and score a 1for every word present, and 0otherwise.\\nThat can be a very wide feature vector, so we limit the dictionary — in\\nthis case to the 10,000 most frequently occurring words in the training\\ncorpus of 25,000 reviews. Fortunately there are nice tools for doing this\\nautomatically. Here is the beginning of a positive review that has been\\nredacted in this way:\\n⟨START⟩this film was just brilliant casting location scenery\\nstory direction everyone’s really suited the part they played and\\nyou could just imagine being there robert ⟨UNK⟩is an amazing\\nactor and now the same being director ⟨UNK⟩father came from\\nthe same scottish island as myself so i loved ...\\nHere we can see many words have been omitted, and some unknown words\\n(UNK) have been marked as such. With this reduction the binary feature\\nvector has length 10,000, and consists mostly of 0’s and a smattering of 1’s\\nin the positions corresponding to words that are present in the document.\\nWe have a training set and test set, each with 25,000 examples, and each\\nbalanced with regard to sentiment . The resulting training feature matrix X\\nhas dimension 25,000×10,000, but only 1.3% of the binary entries are non-\\nzero. We call such a matrix sparse, because most of the values are the same\\n(zero in this case); it can be stored efficiently in sparse matrix format .14\\nsparse\\nmatrix\\nformatThere are a variety of ways to account for the document length; here we\\nonly score a word as in or out of the document, but for example one could\\ninstead record the relative frequency of words. We split off a validation set\\nof size 2,000 from the 25,000 training observations (for model tuning), and\\nfit two model sequences:\\n• A lasso logistic regression using the glmnetpackage;\\n• A two-class neural network with two hidden layers, each with 16\\nReLU units.\\n14Rather than store the whole matrix, we can store instead the location and values for\\nthe nonzero entries. In this case, since the nonzero entries are all 1, just the locations\\nare stored.', metadata={'source': './documents/ISLP_website.pdf', 'page': 420}), Document(page_content='10.4 Document Classification 415\\nFIGURE 10.11. Accuracy of the lasso and a two-hidden-layer neural network\\non theIMDbdata. For the lasso, the x-axis displays −log(λ), while for the neural\\nnetwork it displays epochs (number of times the fitting algorithm passes through\\nthe training set). Both show a tendency to overfit, and achieve approximately the\\nsame test accuracy.\\nBoth methods produce a sequence of solutions. The lasso sequence is in-\\ndexed by the regularization parameter λ. The neural-net sequence is in-\\ndexed by the number of gradient-descent iterations used in the fitting,\\nas measured by training epochs or passes through the training set (Sec-\\ntion10.7). Notice that the training accuracy in Figure 10.11(black points)\\nincreases monotonically in both cases. We can use the validation error to\\npick a good solution from each sequence (blue points in the plots), which\\nwould then be used to make predictions on the test data set.\\nNote that a two-class neural network amounts to a nonlinear logistic\\nregression model. From ( 10.12) and (10.13) we can see that\\nlog(Pr(Y=1|X)\\nPr(Y=0|X))\\n=Z1−Z0 (10.15)\\n=( β10−β00)+K2∑\\nℓ=1(β1ℓ−β0ℓ)A(2)\\nℓ.\\n(This shows the redundancy in the softmax function; for Kclasses we\\nreally only need to estimate K−1sets of coefficients. See Section 4.3.5.) In\\nFigure10.11we show accuracy (fraction correct) rather than classificationaccuracy\\nerror (fraction incorrect), the former being more popular in the machine\\nlearning community. Both models achieve a test-set accuracy of about 88%.\\nThe bag-of-words model summarizes a document by the words present,\\nand ignores their context. There are at least two popular ways to take the\\ncontext into account:\\n• Thebag-of-n-gramsmodel. For example, a bag of 2-grams recordsbag-of-n-\\ngrams/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF\\n468 1 0 1 20.6 0.7 0.8 0.9 1.0Lasso\\n−log(λ)Accuracy\\n/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF\\n/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF\\n/uni25CF/uni25CF/uni25CFtrainvalidationtest/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF\\n51 0 1 5 2 00.6 0.7 0.8 0.9 1.0Neural Net\\nEpochsAccuracy/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF', metadata={'source': './documents/ISLP_website.pdf', 'page': 421}), Document(page_content='416 10. Deep Learningthe consecutive co-occurrence of every distinct pair of words. “Bliss-fully long” can be seen as a positive phrase in a movie review, while“blissfully short” a negative.• Treat the document as a sequence, taking account of all the words inthe context of those that preceded and those that follow.In the next section we explore models for sequences of data, which haveapplications in weather forecasting, speech recognition, language transla-tion, and time-series prediction, to name a few. We continue with thisIMDbexample there.10.5 Recurrent Neural NetworksMany data sources are sequential in nature, and call for special treatmentwhen building predictive models. Examples include:• Documents such as book and movie reviews, newspaper articles, andtweets. The sequence and relative positions of words in a documentcapture the narrative, theme and tone, and can be exploited in taskssuch as topic classification, sentiment analysis, and language transla-tion.• Time series of temperature, rainfall, wind speed, air quality, and soon. We may want to forecast the weather several days ahead, or cli-mate several decades ahead.• Financialtimeseries,wherewetrackmarketindices,tradingvolumes,stock and bond prices, and exchange rates. Here prediction is oftendifficult, but as we will see, certain indices can be predicted withreasonable accuracy.• Recorded speech, musical recordings, and other sound recordings. Wemay want to give a text transcription of a speech, or perhaps a lan-guage translation. We may want to assess the quality of a piece ofmusic, or assign certain attributes.• Handwriting, such as doctor’s notes, and handwritten digits such aszip codes. Here we want to turn the handwriting into digital text, orread the digits (optical character recognition).In arecurrent neural network(RNN), the input objectXis asequence.recurrentneuralnetworkConsider a corpus of documents, such as the collection ofIMDbmovie re-views. Each document can be represented as a sequence ofLwords, soX={X1,X2,...,XL}, where eachXℓrepresents a word. The order ofthe words, and closeness of certain words in a sentence, convey semanticmeaning. RNNs are designed to accommodate and take advantage of thesequential nature of such input objects, much like convolutional neural net-works accommodate the spatial structure of image inputs. The outputYcan also be a sequence (such as in language translation), but often is ascalar, like the binary sentiment label of a movie review document.', metadata={'source': './documents/ISLP_website.pdf', 'page': 422}), Document(page_content='10.5 Recurrent Neural Networks 417\\nA1A2A3AL-1AL=AℓOℓY\\nXℓO1X1O2X2O3X3OL-1XL-1OLY\\nXL...WUBWBWBWBWBWBUUUUFIGURE 10.12. Schematic of a simple recurrent neural network. The input is a\\nsequence of vectors {Xℓ}L\\n1, and here the target is a single response. The network\\nprocesses the input sequence Xsequentially; each Xℓfeeds into the hidden layer,\\nwhich also has as input the activation vector Aℓ−1from the previous element in\\nthe sequence, and produces the current activation vector Aℓ. The same collections\\nof weights W,UandBare used as each element of the sequence is processed. The\\noutput layer produces a sequence of predictions Oℓfrom the current activation\\nAℓ, but typically only the last of these, OL, is of relevance. To the left of the equal\\nsign is a concise representation of the network, which is unrolled into a more\\nexplicit version on the right.\\nFigure10.12illustratesthestructureofaverybasicRNNwithasequence\\nX={X1,X2,...,X L}as input, a simple output Y, and a hidden-layer\\nsequence {Aℓ}L\\n1={A1,A2,...,A L}. EachXℓis a vector; in the document\\nexample Xℓcould represent a one-hot encoding for the ℓth word based on\\nthe language dictionary for the corpus (see the top panel in Figure 10.13\\nfor a simple example). As the sequence is processed one vector Xℓat a\\ntime, the network updates the activations Aℓin the hidden layer, taking\\nas input the vector Xℓand the activation vector Aℓ−1from the previous\\nstep in the sequence. Each Aℓfeeds into the output layer and produces a\\nprediction OℓforY.OL, the last of these, is the most relevant.\\nIn detail, suppose each vector Xℓof the input sequence has pcomponents\\nXT\\nℓ=(Xℓ1,Xℓ2,...,X ℓp), and the hidden layer consists of KunitsAT\\nℓ=\\n(Aℓ1,Aℓ2,...,A ℓK). As in Figure 10.4, we represent the collection of K×\\n(p+1) shared weights wkjfor the input layer by a matrix W, and similarly\\nUis aK×Kmatrix of the weights uksfor the hidden-to-hidden layers,\\nandBis aK+1vector of weights βkfor the output layer. Then\\nAℓk=g(\\nwk0+p∑\\nj=1wkjXℓj+K∑\\ns=1uksAℓ−1,s)\\n, (10.16)\\nand the output Oℓis computed as\\nOℓ=β0+K∑\\nk=1βkAℓk (10.17)\\nfor a quantitative response, or with an additional sigmoid activation func-\\ntion for a binary response, for example. Here g(·)is an activation function\\nsuch as ReLU. Notice that the same weights W,UandBare used as we', metadata={'source': './documents/ISLP_website.pdf', 'page': 423}), Document(page_content='418 10. Deep Learning\\nprocess each element in the sequence, i.e. they are not functions of ℓ. This\\nis a form of weight sharing used by RNNs, and similar to the use of filtersweight\\nsharingin convolutional neural networks (Section 10.3.1.) As we proceed from be-\\nginning to end, the activations Aℓaccumulate a history of what has been\\nseen before, so that the learned context can be used for prediction.\\nFor regression problems the loss function for an observation (X,Y)is\\n(Y−OL)2, (10.18)\\nwhichonlyreferencesthefinaloutput OL=β0+∑K\\nk=1βkALk.ThusO1,O2,\\n...,O L−1arenotused.Whenwefitthemodel,eachelement Xℓoftheinput\\nsequence Xcontributes to OLvia the chain ( 10.16), and hence contributes\\nindirectly to learning the shared parameters W,UandBvia the loss\\n(10.18). Withninput sequence/response pairs (xi,yi),the parameters are\\nfound by minimizing the sum of squares\\nn∑\\ni=1(yi−oiL)2=n∑\\ni=1(\\nyi−(\\nβ0+K∑\\nk=1βkg(\\nwk0+p∑\\nj=1wkjxiLj+K∑\\ns=1uksai,L−1,s)))2\\n.\\n(10.19)\\nHere we use lowercase letters for the observed yiand vector sequences\\nxi={xi1,xi2,...,x iL},15as well as the derived activations.\\nSince the intermediate outputs Oℓare not used, one may well ask why\\nthey are there at all. First of all, they come for free, since they use the same\\noutputweights Bneededtoproduce OL,andprovideanevolvingprediction\\nfor the output. Furthermore, for some learning tasks the response is also a\\nsequence, and so the output sequence {O1,O2,...,O L}is explicitly needed.\\nWhen used at full strength, recurrent neural networks can be quite com-\\nplex. We illustrate their use in two simple applications. In the first, we\\ncontinue with the IMDbsentiment analysis of the previous section, where\\nwe process the words in the reviews sequentially. In the second application,\\nwe illustrate their use in a financial time series forecasting problem.\\n10.5.1 Sequential Models for Document Classification\\nHere we return to our classification task with the IMDbreviews. Our ap-\\nproach in Section 10.4was to use the bag-of-words model. Here the plan\\nis to use instead the sequence of words occurring in a document to make\\npredictions about the label for the entire document.\\nWe have, however, a dimensionality problem: each word in our document\\nis represented by a one-hot-encoded vector (dummy variable) with 10,000\\nelements (one per word in the dictionary)! An approach that has become\\npopular is to represent each word in a much lower-dimensional embeddingembeddingspace. This means that rather than representing each word by a binary\\nvector with 9,999 zeros and a single one in some position, we will represent\\nit instead by a set of mreal numbers, none of which are typically zero. Here\\nmis the embedding dimension, and can be in the low 100s, or even less.\\nThis means (in our case) that we need a matrix Eof dimension m×10,000,\\n15This is a sequence of vectors; each element xiℓis ap-vector.', metadata={'source': './documents/ISLP_website.pdf', 'page': 424}), Document(page_content='10.5 Recurrent Neural Networks 419thisisoneofthebestfilmsactuallythebestIhaveeverseenthefilmstartsonefalldayOne−hotEmbedFIGURE 10.13. Depiction of a sequence of 20words representing a single doc-\\nument: one-hot encoded using a dictionary of 16words (top panel) and embedded\\nin anm-dimensional space with m=5(bottom panel).\\nwhere each column is indexed by one of the 10,000 words in our dictionary,\\nand the values in that column give the mcoordinates for that word in the\\nembedding space.\\nFigure10.13illustrates the idea (with a dictionary of 16 rather than\\n10,000, and m=5). Where does Ecome from? If we have a large corpus\\nof labeled documents, we can have the neural network learnEas part\\nof the optimization. In this case Eis referred to as an embedding layer,embedding\\nlayerand a specialized Eis learned for the task at hand. Otherwise we can\\ninsert a precomputed matrix Ein the embedding layer, a process known\\nasweight freezing . Two pretrained embeddings, word2vec andGloVe, areweight\\nfreezing\\nword2vec\\nGloVewidely used.16These are built from a very large corpus of documents by\\na variant of principal components analysis (Section 12.2). The idea is that\\nthe positions of words in the embedding space preserve semantic meaning;\\ne.g. synonyms should appear near each other.\\nSo far, so good. Each document is now represented as a sequence of m-\\nvectors that represents the sequence of words. The next step is to limit\\neach document to the last Lwords. Documents that are shorter than L\\nget padded with zeros upfront. So now each document is represented by a\\nseries consisting of LvectorsX={X1,X2,...,X L}, and each Xℓin the\\nsequence has mcomponents.\\nWe now use the RNN structure in Figure 10.12. The training corpus\\nconsists of nseparate series (documents) of length L, each of which gets\\nprocessed sequentially from left to right. In the process, a parallel series of\\nhidden activation vectors Aℓ,ℓ=1,...,L is created as in ( 10.16) for each\\ndocument. Aℓfeedsintotheoutputlayertoproducetheevolvingprediction\\nOℓ. We use the final value OLto predict the response: the sentiment of the\\nreview.\\n16word2vec is described in Mikolov, Chen, Corrado, and Dean (2013), available\\nathttps://code.google.com/archive/p/word2vec .GloVeis described in Pennington,\\nSocher, and Manning (2014), available at https://nlp.stanford.edu/projects/glove .', metadata={'source': './documents/ISLP_website.pdf', 'page': 425}), Document(page_content='420 10. Deep Learning\\nThis is a simple RNN, and has relatively few parameters. If there are K\\nhidden units, the common weight matrix WhasK×(m+ 1) parameters,\\nthe matrix UhasK×Kparameters, and Bhas2(K+ 1) for the two-class\\nlogistic regression as in ( 10.15). These are used repeatedly as we process\\nthe sequence X={Xℓ}L\\n1from left to right, much like we use a single\\nconvolution filter to process each patch in an image (Section 10.3.1). If the\\nembedding layer Eis learned, that adds an additional m×Dparameters\\n(D= 10 ,000here), and is by far the biggest cost.\\nWe fit the RNN as described in Figure 10.12and the accompaying text to\\ntheIMDbdata. The model had an embedding matrix Ewithm= 32 (which\\nwas learned in training as opposed to precomputed), followed by a single\\nrecurrent layer with K= 32 hidden units. The model was trained with\\ndropout regularization on the 25,000 reviews in the designated training\\nset, and achieved a disappointing 76% accuracy on the IMDbtest data. A\\nnetwork using the GloVepretrained embedding matrix Eperformed slightly\\nworse.\\nFor ease of exposition we have presented a very simple RNN. More elab-\\norate versions use long term andshort term memory (LSTM). Two tracks\\nof hidden-layer activations are maintained, so that when the activation Aℓ\\nis computed, it gets input from hidden units both further back in time,\\nand closer in time — a so-called LSTM RNN . With long sequences, thisLSTM RNNovercomes the problem of early signals being washed out by the time they\\nget propagated through the chain to the final activation vector AL.\\nWhen we refit our model using the LSTM architecture for the hidden\\nlayer, the performance improved to 87% on the IMDbtest data. This is com-\\nparable with the 88% achieved by the bag-of-words model in Section 10.4.\\nWe give details on fitting these models in Section 10.9.6.\\nDespite this added LSTM complexity, our RNN is still somewhat “entry\\nlevel”. We could probably achieve slightly better results by changing the\\nsize of the model, changing the regularization, and including additional\\nhidden layers. However, LSTM models take a long time to train, which\\nmakes exploring many architectures and parameter optimization tedious.\\nRNNs provide a rich framework for modeling data sequences, and they\\ncontinue to evolve. There have been many advances in the development\\nof RNNs — in architecture, data augmentation, and in the learning algo-\\nrithms. At the time of this writing (early 2020) the leading RNN configura-\\ntions report accuracy above 95% on the IMDbdata. The details are beyond\\nthe scope of this book.17\\n10.5.2 Time Series Forecasting\\nFigure10.14shows historical trading statistics from the New York Stock\\nExchange. Shown are three daily time series covering the period December\\n3, 1962 to December 31, 1986:18\\n17AnIMDbleaderboard can be found at https://paperswithcode.com/sota/\\nsentiment-analysis-on-imdb .\\n18These data were assembled by LeBaron and Weigend (1998) IEEE Transactions on\\nNeural Networks , 9(1): 213–220.', metadata={'source': './documents/ISLP_website.pdf', 'page': 426}), Document(page_content='10.5 Recurrent Neural Networks 421Log(Trading Volume)−1.0 0.0 0.5 1.0Dow Jones Return−0.04 0.00 0.04\\n19651970197519801985−13 −11 −9 −8Log(Volatility)FIGURE 10.14. Historical trading statistics from the New York Stock Exchange.\\nDaily values of the normalized log trading volume, DJIA return, and log volatility\\nare shown for a 24-year period from 1962–1986. We wish to predict trading volume\\non any day, given the history on all earlier days. To the left of the red bar (January\\n2, 1980) is training data, and to the right test data.\\n•Log trading volume .Thisisthefractionofalloutstandingsharesthat\\nare traded on that day, relative to a 100-day moving average of past\\nturnover, on the log scale.\\n•Dow Jones return . This is the difference between the log of the Dow\\nJones Industrial Index on consecutive trading days.\\n•Log volatility . This is based on the absolute values of daily price\\nmovements.\\nPredicting stock prices is a notoriously hard problem, but it turns out that\\npredicting trading volume based on recent past history is more manageable\\n(and is useful for planning trading strategies).\\nAn observation here consists of the measurements (vt,rt,zt)on dayt, in\\nthis case the values for log_volume ,DJ_return andlog_volatility . There\\nareatotalof T=6,051suchtriples,eachofwhichisplottedasatimeseries\\nin Figure 10.14. One feature that strikes us immediately is that the day-\\nto-day observations are not independent of each other. The series exhibit\\nauto-correlation — in this case values nearby in time tend to be similarauto-\\ncorrelation to each other. This distinguishes time series from other data sets we have\\nencountered, in which observations can be assumed to be independent of', metadata={'source': './documents/ISLP_website.pdf', 'page': 427}), Document(page_content='422 10. Deep Learning\\n0 5 10 15 20 25 30 350.0 0.4 0.8Log( Trading Volume)\\nLagAutocorrelation FunctionFIGURE 10.15. The autocorrelation function for log_volume . We see that\\nnearby values are fairly strongly correlated, with correlations above 0.2as far as\\n20 days apart.\\neach other. To be clear, consider pairs of observations (vt,vt−ℓ),alagofℓlagdays apart. If we take all such pairs in the vtseries and compute their corre-\\nlation coefficient, this gives the autocorrelation at lag ℓ. Figure10.15shows\\nthe autocorrelation function for all lags up to 37, and we see considerable\\ncorrelation.\\nAnother interesting characteristic of this forecasting problem is that the\\nresponse variable vt—log_volume — is also a predictor! In particular, we\\nwill use the past values of log_volume to predict values in the future.\\nRNN forecaster\\nWe wish to predict a value vtfrom past values vt−1,vt−2,..., and also to\\nmake use of past values of the other series rt−1,rt−2,...andzt−1,zt−2,....\\nAlthough our combined data is quite a long series with 6,051 trading\\ndays, the structure of the problem is different from the previous document-\\nclassification example.\\n• We only have one series of data, not 25,000.\\n• We have an entire seriesof targets vt, and the inputs include past\\nvalues of this series.\\nHow do we represent this problem in terms of the structure displayed in\\nFigure10.12? The idea is to extract many short mini-series of input se-\\nquencesX={X1,X2,...,X L}with a predefined length L(called the laglagin this context), and a corresponding target Y. They have the form\\nX1=\\uf8eb\\n\\uf8edvt−L\\nrt−L\\nzt−L\\uf8f6\\n\\uf8f8,X2=\\uf8eb\\n\\uf8edvt−L+1\\nrt−L+1\\nzt−L+1\\uf8f6\\n\\uf8f8,···,XL=\\uf8eb\\n\\uf8edvt−1\\nrt−1\\nzt−1\\uf8f6\\n\\uf8f8,andY=vt.\\n(10.20)\\nSo here the target Yis the value of log_volume vtat a single timepoint t,\\nand the input sequence Xis the series of 3-vectors {Xℓ}L\\n1each consisting\\nof the three measurements log_volume ,DJ_return andlog_volatility from\\ndayt−L,t−L+1, up tot−1. Each value of tmakes a separate (X,Y)\\npair, for trunning from L+1toT. For the NYSEdata we will use the past', metadata={'source': './documents/ISLP_website.pdf', 'page': 428}), Document(page_content='10.5 Recurrent Neural Networks 423\\n1980198219841986−1.0 0.0 0.5 1.0Test Period: Observed and Predicted\\nYearlog(Trading Volume)FIGURE 10.16. RNN forecast of log_volume on theNYSEtest data. The black\\nlines are the true volumes, and the superimposed orange the forecasts. The fore-\\ncasted series accounts for 42% of the variance of log_volume .\\nfive trading days to predict the next day’s trading volume. Hence, we use\\nL=5. SinceT=6,051, we can create 6,046 such (X,Y)pairs. Clearly L\\nis a parameter that should be chosen with care, perhaps using validation\\ndata.\\nWe fit this model with K= 12 hidden units using the 4,281 training\\nsequences derived from the data before January 2, 1980 (see Figure 10.14),\\nand then used it to forecast the 1,770 values of log_volume after this date.\\nWe achieve an R2=0.42on the test data. Details are given in Sec-\\ntion10.9.6. As astraw man ,19using yesterday’s value for log_volume as\\nthe prediction for today has R2=0.18. Figure 10.16shows the forecast\\nresults. We have plotted the observed values of the daily log_volume for the\\ntest period 1980–1986 in black, and superimposed the predicted series in\\norange. The correspondence seems rather good.\\nIn forecasting the value of log_volume in the test period, we have to use\\nthe test data itself in forming the input sequences X. This may feel like\\ncheating, but in fact it is not; we are always using past data to predict the\\nfuture.\\nAutoregression\\nThe RNN we just fit has much in common with a traditional autoregressionauto-\\nregression (AR) linear model, which we present now for comparison. We first consider\\nthe response sequence vtalone, and construct a response vector yand a\\nmatrixMof predictors for least squares regression as follows:\\ny=\\uf8ee\\n\\uf8ef\\uf8ef\\uf8ef\\uf8ef\\uf8ef\\uf8f0vL+1\\nvL+2\\nvL+3\\n...\\nvT\\uf8f9\\n\\uf8fa\\uf8fa\\uf8fa\\uf8fa\\uf8fa\\uf8fbM=\\uf8ee\\n\\uf8ef\\uf8ef\\uf8ef\\uf8ef\\uf8ef\\uf8f01 vLvL−1··· v1\\n1vL+1 vL··· v2\\n1vL+2vL+1··· v3\\n...............\\n1vT−1vT−2···vT−L\\uf8f9\\n\\uf8fa\\uf8fa\\uf8fa\\uf8fa\\uf8fa\\uf8fb.(10.21)\\nMandyeach have T−Lrows, one per observation. We see that the\\npredictors for any given response vton daytare the previous Lvalues\\n19A straw man here refers to a simple and sensible prediction that can be used as a\\nbaseline for comparison.', metadata={'source': './documents/ISLP_website.pdf', 'page': 429}), Document(page_content='424 10. Deep Learning\\nof the same series. Fitting a regression of yonMamounts to fitting the\\nmodel\\nˆvt=ˆβ0+ˆβ1vt−1+ˆβ2vt−2+···+ˆβLvt−L, (10.22)\\nand is called an order- Lautoregressive model, or simply AR( L). For the\\nNYSEdata we can include lagged versions of DJ_return andlog_volatility ,\\nrtandzt, in the predictor matrix M, resulting in 3L+1columns. An AR\\nmodel with L=5 achieves a test R2of0.41, slightly inferior to the 0.42\\nachieved by the RNN.\\nOf course the RNN and AR models are very similar. They both use\\nthe same response Yand input sequences Xof length L=5 and dimen-\\nsionp=3 in this case. The RNN processes this sequence from left to\\nright with the same weights W(for the input layer), while the AR model\\nsimply treats all Lelements of the sequence equally as a vector of L×p\\npredictors — a process called flattening in the neural network literature.flatteningOf course the RNN also includes the hidden layer activations Aℓwhich\\ntransfer information along the sequence, and introduces additional nonlin-\\nearity. From ( 10.19) withK= 12 hidden units, we see that the RNN has\\n13 + 12 ×(1 + 3 + 12) = 205 parameters, compared to the 16 for the AR( 5)\\nmodel.\\nAn obvious extension of the AR model is to use the set of lagged predic-\\ntors as the input vector to an ordinary feedforward neural network ( 10.1),\\nand hence add more flexibility. This achieved a test R2=0.42, slightly\\nbetter than the linear AR, and the same as the RNN.\\nAll the models can be improved by including the variable day_of_week\\ncorresponding to the day tof the target vt(which can be learned from the\\ncalendar dates supplied with the data); trading volume is often higher on\\nMondays and Fridays. Since there are five trading days, this one-hot en-\\ncodes to five binary variables. The performance of the AR model improved\\ntoR2=0.46as did the RNN, and the nonlinear AR model improved to\\nR2=0.47.\\nWe used the most simple version of the RNN in our examples here.\\nAdditionalexperimentswiththeLSTMextensionoftheRNNyieldedsmall\\nimprovements, typically of up to 1% in R2in these examples.\\nWe give details of how we fit all three models in Section 10.9.6.\\n10.5.3 Summary of RNNs\\nWe have illustrated RNNs through two simple use cases, and have only\\nscratched the surface.\\nThere are many variations and enhancements of the simple RNN we\\nused for sequence modeling. One approach we did not discuss uses a one-\\ndimensional convolutional neural network, treating the sequence of vectors\\n(say words, as represented in the embedding space) as an image. The con-\\nvolution filter slides along the sequence in a one-dimensional fashion, with\\nthe potential to learn particular phrases or short subsequences relevant to\\nthe learning task.\\nOne can also have additional hidden layers in an RNN. For example,\\nwith two hidden layers, the sequence Aℓis treated as an input sequence to\\nthe next hidden layer in an obvious fashion.', metadata={'source': './documents/ISLP_website.pdf', 'page': 430}), Document(page_content='10.6 When to Use Deep Learning 425\\nThe RNN we used scanned the document from beginning to end; alter-\\nnativebidirectional RNNs scan the sequences in both directions.bidirectionalIn language translation the target is also a sequence of words, in a\\nlanguage different from that of the input sequence. Both the input se-\\nquence and the target sequence are represented by a structure similar to\\nFigure10.12, and they share the hidden units. In this so-called Seq2SeqSeq2Seqlearning, the hidden units are thought to capture the semantic meaning\\nof the sentences. Some of the big breakthroughs in language modeling and\\ntranslation resulted from the relatively recent improvements in such RNNs.\\nAlgorithms used to fit RNNs can be complex and computationally costly.\\nFortunately, good software protects users somewhat from these complexi-\\nties,andmakesspecifyingandfittingthesemodelsrelativelypainless.Many\\nof the models that we enjoy in daily life (like Google Translate ) use state-\\nof-the-art architectures developed by teams of highly skilled engineers, and\\nhave been trained using massive computational and data resources.\\n10.6 When to Use Deep Learning\\nThe performance of deep learning in this chapter has been rather impres-\\nsive. It nailed the digit classification problem, and deep CNNs have really\\nrevolutionized image classification. We see daily reports of new success sto-\\nries for deep learning. Many of these are related to image classification\\ntasks, such as machine diagnosis of mammograms or digital X-ray images,\\nophthalmology eye scans, annotations of MRI scans, and so on. Likewise\\nthere are numerous successes of RNNs in speech and language translation,\\nforecasting, and document modeling. The question that then begs an an-\\nswer is:should we discard all our older tools, and use deep learning on every\\nproblem with data? To address this question, we revisit our Hittersdataset\\nfrom Chapter 6.\\nThis is a regression problem, where the goal is to predict the Salaryof\\na baseball player in 1987 using his performance statistics from 1986. After\\nremoving players with missing responses, we are left with 263 players and\\n19 variables. We randomly split the data into a training set of 176 players\\n(twothirds),andatestsetof87players(onethird).Weusedthreemethods\\nfor fitting a regression model to these data.\\n• Alinearmodelwasusedtofitthetrainingdata,andmakepredictions\\non the test data. The model has 20 parameters.\\n• The same linear model was fit with lasso regularization. The tuning\\nparameter was selected by 10-fold cross-validation on the training\\ndata.Itselectedamodelwith12variableshavingnonzerocoefficients.\\n• A neural network with one hidden layer consisting of 64 ReLUunits\\nwas fit to the data. This model has 1,345 parameters.20\\n20The model was fit by stochastic gradient descent with a batch size of 32 for 1,000\\nepochs, and 10% dropout regularization. The test error performance flattened out and\\nstarted to slowly increase after 1,000 epochs. These fitting details are discussed in Sec-\\ntion10.7.', metadata={'source': './documents/ISLP_website.pdf', 'page': 431}), Document(page_content='426 10. Deep Learning\\nModel # Parameters Mean Abs. Error Test Set R2\\nLinear Regression 20 254.7 0.56\\nLasso 12 252.3 0.51\\nNeural Network 1345 257.4 0.54\\nTABLE 10.2. Prediction results on the Hitters test data for linear models fit\\nby ordinary least squares and lasso, compared to a neural network fit by stochastic\\ngradient descent with dropout regularization.\\nCoefficient Std. error t-statistic p-value\\nIntercept -226.67 86.26 -2.63 0.0103\\nHits 3.06 1.02 3.00 0.0036\\nWalks 0.181 2.04 0.09 0.9294\\nCRuns 0.859 0.12 7.09 <0.0001\\nPutOuts 0.465 0.13 3.60 0.0005\\nTABLE 10.3. Least squares coefficient estimates associated with the regres-\\nsion ofSalaryon four variables chosen by lasso on the Hitters data set. This\\nmodel achieved the best performance on the test data, with a mean absolute error\\nof 224.8. The results reported here were obtained from a regression on the test\\ndata, which was not used in fitting the lasso model.\\nTable10.2compares the results. We see similar performance for all three\\nmodels. We report the mean absolute error on the test data, as well as\\nthe test R2for each method, which are all respectable (see Exercise 5).\\nWe spent a fair bit of time fiddling with the configuration parameters of\\nthe neural network to achieve these results. It is possible that if we were to\\nspend more time, and got the form and amount of regularization just right,\\nthat we might be able to match or even outperform linear regression and\\nthe lasso. But with great ease we obtained linear models that work well.\\nLinear models are much easier to present and understand than the neural\\nnetwork, which is essentially a black box. The lasso selected 12 of the 19\\nvariables in making its prediction. So in cases like this we are much better\\noff following the Occam’s razor principle: when faced with several methodsOccam’s\\nrazorthat give roughly equivalent performance, pick the simplest.\\nAfter a bit more exploration with the lasso model, we identified an even\\nsimpler model with four variables. We then refit the linear model with these\\nfourvariablestothetrainingdata(theso-called relaxed lasso ),andachieved\\na test mean absolute error of 224.8, the overall winner! It is tempting to\\npresent the summary table from this fit, so we can see coefficients and p-\\nvalues; however, since the model was selected on the training data, there\\nwould be selection bias . Instead, we refit the model on the test data, which\\nwas not used in the selection. Table 10.3shows the results.\\nWehaveanumberofverypowerfultoolsatourdisposal,includingneural\\nnetworks, random forests and boosting, support vector machines and gen-\\neralized additive models, to name a few. And then we have linear models,\\nand simple variants of these. When faced with new data modeling and pre-\\ndiction problems, it’s tempting to always go for the trendy new methods.\\nOften they give extremely impressive results, especially when the datasets\\nare very large and can support the fitting of high-dimensional nonlinear\\nmodels. However, ifwe can produce models with the simpler tools that', metadata={'source': './documents/ISLP_website.pdf', 'page': 432}), Document(page_content='10.7 Fitting a Neural Network 427\\nperform as well, they are likely to be easier to fit and understand, and po-\\ntentially less fragile than the more complex approaches. Wherever possible,\\nit makes sense to try the simpler models as well, and then make a choice\\nbased on the performance/complexity tradeoff.\\nTypically we expect deep learning to be an attractive choice when the\\nsample size of the training set is extremely large, and when interpretability\\nof the model is not a high priority.\\n10.7 Fitting a Neural Network\\nFitting neural networks is somewhat complex, and we give a brief overview\\nhere. The ideas generalize to much more complex networks. Readers who\\nfind this material challenging can safely skip it. Fortunately, as we see in\\nthe lab at the end of this chapter, good software is available to fit neural\\nnetwork models in a relatively automated way, without worrying about the\\ntechnical details of the model-fitting procedure.\\nWe start with the simple network depicted in Figure 10.1in Section 10.1.\\nIn model ( 10.1) the parameters are β=(β0,β1,...,βK), as well as each of\\nthewk=(wk0,wk1,...,w kp),k=1,...,K . Givenobservations (xi,yi),i=\\n1, . . . , n, wecouldfitthemodelbysolvinganonlinearleastsquaresproblem\\nminimize\\n{wk}K\\n1,β1\\n2n∑\\ni=1(yi−f(xi))2, (10.23)\\nwhere\\nf(xi)=β0+K∑\\nk=1βkg(\\nwk0+p∑\\nj=1wkjxij)\\n. (10.24)\\nThe objective in ( 10.23) looks simple enough, but because of the nested\\narrangement of the parameters and the symmetry of the hidden units, it is\\nnot straightforward to minimize. The problem is nonconvex in the param-\\neters, and hence there are multiple solutions. As an example, Figure 10.17\\nshows a simple nonconvex function of a single variable θ; there are two\\nsolutions: one is a local minimum and the other is a global minimum . Fur-local\\nminimum\\nglobal\\nminimumthermore, ( 10.1) is the very simplest of neural networks; in this chapter we\\nhave presented much more complex ones where these problems are com-\\npounded. To overcome some of these issues and to protect from overfitting,\\ntwo general strategies are employed when fitting neural networks.\\n•Slow Learning: the model is fit in a somewhat slow iterative fash-\\nion, using gradient descent . The fitting process is then stopped whengradient\\ndescentoverfitting is detected.\\n•Regularization: penaltiesareimposedontheparameters,usuallylasso\\nor ridge as discussed in Section 6.2.\\nSuppose we represent all the parameters in one long vector θ. Then we\\ncan rewrite the objective in ( 10.23) as\\nR(θ)=1\\n2n∑\\ni=1(yi−fθ(xi))2, (10.25)', metadata={'source': './documents/ISLP_website.pdf', 'page': 433}), Document(page_content='428 10. Deep Learning\\nFIGURE 10.17. Illustration of gradient descent for one-dimensional θ. The\\nobjective function R(θ)is not convex, and has two minima, one at θ=−0.46\\n(local), the other at θ=1.02(global). Starting at some value θ0(typically ran-\\ndomly chosen), each step in θmoves downhill — against the gradient — until it\\ncannot go down any further. Here gradient descent reached the global minimum\\nin7steps.\\nwhere we make explicit the dependence of fon the parameters. The idea\\nof gradient descent is very simple.\\n1. Start with a guess θ0for all the parameters in θ, and set t=0.\\n2. Iterate until the objective ( 10.25) fails to decrease:\\n(a) Findavector δthatreflectsasmallchangein θ,suchthat θt+1=\\nθt+δreducesthe objective; i.e. such that R(θt+1)<R(θt).\\n(b) Sett←t+1.\\nOne can visualize (Figure 10.17) standing in a mountainous terrain, and\\nthe goal is to get to the bottom through a series of steps. As long as each\\nstep goes downhill, we must eventually get to the bottom. In this case we\\nwere lucky, because with our starting guess θ0we end up at the global\\nminimum. In general we can hope to end up at a (good) local minimum.\\n10.7.1 Backpropagation\\nHowdowefindthedirectionstomove θsoastodecreasetheobjective R(θ)\\nin (10.25)? Thegradient ofR(θ), evaluated at some current value θ=θm,gradientis the vector of partial derivatives at that point:\\n∇R(θm)=∂R(θ)\\n∂θ⏐⏐⏐\\nθ=θm. (10.26)\\nThe subscript θ=θmmeans that after computing the vector of derivatives,\\nwe evaluate it at the current guess, θm. This gives the direction in θ-space\\nin which R(θ)increases most rapidly. The idea of gradient descent is to\\nmoveθa little in the opposite direction (since we wish to go downhill):\\nθm+1←θm−ρ∇R(θm). (10.27)−1.0 −0.5 0.0 0.5 1.00123456\\nθR(θ)\\nθ0θ1θ2θ7/uni25CF/uni25CF/uni25CF/uni25CFR(θ0)R(θ1)R(θ2)R(θ7)', metadata={'source': './documents/ISLP_website.pdf', 'page': 434}), Document(page_content='10.7 Fitting a Neural Network 429\\nFor a small enough value of the learning rate ρ, this step will decrease thelearning rateobjective R(θ); i.e.R(θm+1)≤R(θm). If the gradient vector is zero, then\\nwe may have arrived at a minimum of the objective.\\nHow complicated is the calculation ( 10.26)? It turns out that it is quite\\nsimple here, and remains simple even for much more complex networks,\\nbecause of the chain rule of differentiation.chain ruleSinceR(θ)=∑n\\ni=1Ri(θ)=1\\n2∑n\\ni=1(yi−fθ(xi))2is a sum, its gradient\\nis also a sum over the nobservations, so we will just examine one of these\\nterms,\\nRi(θ)=1\\n2(\\nyi−β0−K∑\\nk=1βkg(\\nwk0+p∑\\nj=1wkjxij))2\\n. (10.28)\\nTo simplify the expressions to follow, we write zik=wk0+∑p\\nj=1wkjxij.\\nFirst we take the derivative with respect to βk:\\n∂Ri(θ)\\n∂βk=∂Ri(θ)\\n∂fθ(xi)·∂fθ(xi)\\n∂βk\\n=−(yi−fθ(xi))·g(zik). (10.29)\\nAnd now we take the derivative with respect to wkj:\\n∂Ri(θ)\\n∂wkj=∂Ri(θ)\\n∂fθ(xi)·∂fθ(xi)\\n∂g(zik)·∂g(zik)\\n∂zik·∂zik\\n∂wkj\\n=−(yi−fθ(xi))·βk·g′(zik)·xij. (10.30)\\nNotice that both these expressions contain the residual yi−fθ(xi). In\\n(10.29) we see that a fraction of that residual gets attributed to each of\\nthe hidden units according to the value of g(zik). Then in ( 10.30) we see\\na similar attribution to input jvia hidden unit k. So the act of differen-\\ntiation assigns a fraction of the residual to each of the parameters via the\\nchain rule — a process known as backpropagation in the neural networkbackprop-\\nagationliterature. Although these calculations are straightforward, it takes careful\\nbookkeeping to keep track of all the pieces.\\n10.7.2 Regularization and Stochastic Gradient Descent\\nGradient descent usually takes many steps to reach a local minimum. In\\npractice, there are a number of approaches for accelerating the process.\\nAlso, when nis large, instead of summing ( 10.29)–(10.30) over all nob-\\nservations, we can sample a small fraction or minibatch of them each timeminibatchwe compute a gradient step. This process is known as stochastic gradient\\ndescent(SGD) and is the state of the art for learning deep neural networks.stochastic\\ngradient\\ndescentFortunately, there is very good software for setting up deep learning mod-\\nels, and for fitting them to data, so most of the technicalities are hidden\\nfrom the user.\\nWe now turn to the multilayer network (Figure 10.4) used in the digit\\nrecognitionproblem.Thenetworkhasover235,000weights,whichisaround\\nfour times the number of training examples. Regularization is essential here', metadata={'source': './documents/ISLP_website.pdf', 'page': 435}), Document(page_content='430 10. Deep Learning\\n0 5 10 15 20 25 300.1 0.2 0.3 0.4EpochsValue of Objective FunctionTraining SetValidation Set\\n0 5 10 15 20 25 300.00 0.02 0.04 0.06 0.08 0.10 0.12EpochsClassification ErrorFIGURE 10.18. Evolution of training and validation errors for the MNISTneural\\nnetwork depicted in Figure 10.4, as a function of training epochs. The objective\\nrefers to the log-likelihood ( 10.14).\\nto avoid overfitting. The first row in Table 10.1uses ridge regularization on\\nthe weights. This is achieved by augmenting the objective function ( 10.14)\\nwith a penalty term:\\nR(θ;λ)=−n∑\\ni=19∑\\nm=0yimlog(fm(xi)) + λ∑\\njθ2\\nj. (10.31)\\nThe parameter λis often preset at a small value, or else it is found using the\\nvalidation-set approach of Section 5.3.1. We can also use different values of\\nλfor the groups of weights from different layers; in this case W1andW2\\nwere penalized, while the relatively few weights Bof the output layer were\\nnot penalized at all. Lasso regularization is also popular as an additional\\nform of regularization, or as an alternative to ridge.\\nFigure10.18shows some metrics that evolve during the training of the\\nnetwork on the MNISTdata. It turns out that SGD naturally enforces its\\nown form of approximately quadratic regularization.21Here the minibatch\\nsizewas128observationspergradientupdate.Theterm epochslabelingtheepochshorizontal axis in Figure 10.18counts the number of times an equivalent of\\nthe full training set has been processed. For this network, 20% of the 60,000\\ntraining observations were used as a validation set in order to determine\\nwhen training should stop. So in fact 48,000 observations were used for\\ntraining, and hence there are 48,000/128≈375minibatch gradient updates\\nper epoch. We see that the value of the validation objective actually starts\\nto increase by 30 epochs, so early stopping can also be used as an additionalearly\\nstoppingform of regularization.\\n21This and other properties of SGD for deep learning are the subject of much research\\nin the machine learning literature at the time of writing.', metadata={'source': './documents/ISLP_website.pdf', 'page': 436}), Document(page_content='10.7 Fitting a Neural Network 431\\nFIGURE 10.19. Dropout Learning. Left: a fully connected network. Right: net-\\nwork with dropout in the input and hidden layer. The nodes in grey are selected\\nat random, and ignored in an instance of training.\\n10.7.3 Dropout Learning\\nThe second row in Table 10.1is labeled dropout. This is a relatively newdropoutand efficient form of regularization, similar in some respects to ridge reg-\\nularization. Inspired by random forests (Section 8.2), the idea is to ran-\\ndomly remove a fraction φof the units in a layer when fitting the model.\\nFigure10.19illustrates this. This is done separately each time a training\\nobservation is processed. The surviving units stand in for those missing,\\nand their weights are scaled up by a factor of 1/(1−φ)to compensate.\\nThis prevents nodes from becoming over-specialized, and can be seen as\\na form of regularization. In practice dropout is achieved by randomly set-\\nting the activations for the “dropped out” units to zero, while keeping the\\narchitecture intact.\\n10.7.4 Network Tuning\\nThe network in Figure 10.4is considered to be relatively straightforward;\\nit nevertheless requires a number of choices that all have an effect on the\\nperformance:\\n•The number of hidden layers, and the number of units per layer.\\nModern thinking is that the number of units per hidden layer can\\nbe large, and overfitting can be controlled via the various forms of\\nregularization.\\n•Regularization tuning parameters. These include the dropout rate φ\\nand the strength λof lasso and ridge regularization, and are typically\\nset separately at each layer.\\n•Details of stochastic gradient descent. These include the batch size,\\nthe number of epochs, and if used, details of data augmentation (Sec-\\ntion10.3.4.)\\nChoices such as these can make a difference. In preparing this MNISTexam-\\nple, we achieved a respectable 1.8%misclassification error after some trial\\nand error. Finer tuning and training of a similar network can get under\\n1%error on these data, but the tinkering process can be tedious, and can\\nresult in overfitting if done carelessly.', metadata={'source': './documents/ISLP_website.pdf', 'page': 437}), Document(page_content='432 10. Deep Learning\\n251020500.0 0.5 1.0 1.5 2.0Degrees of FreedomErrorTraining ErrorTest Error\\nFIGURE 10.20. Double descent phenomenon, illustrated using error plots for a\\none-dimensional natural spline example. The horizontal axis refers to the number\\nof spline basis functions on the log scale. The training error hits zero when the\\ndegrees of freedom coincides with the sample size n= 20 , the “interpolation\\nthreshold”, and remains zero thereafter. The test error increases dramatically\\nat this threshold, but then descends again to a reasonable value before finally\\nincreasing again.\\n10.8 Interpolation and Double Descent\\nThroughoutthisbook,wehaverepeatedlydiscussedthebias-variancetrade-\\noff, first presented in Section 2.2.2. This trade-off indicates that statistical\\nlearning methods tend to perform the best, in terms of test-set error, for an\\nintermediate level of model complexity. In particular, if we plot “flexibil-\\nity” on the x-axis and error on the y-axis, then we generally expect to see\\nthat test error has a U-shape, whereas training error decreases monotoni-\\ncally. Two “typical” examples of this behavior can be seen in the right-hand\\npanel of Figure 2.9on page29, and in Figure 2.17on page39. One implica-\\ntion of the bias-variance trade-off is that it is generally not a good idea to\\ninterpolate the training data — that is, to get zero training error — sinceinterpolatethat will often result in very high test error.\\nHowever,itturnsoutthatincertainspecificsettingsitcanbepossiblefor\\na statistical learning method that interpolates the training data to perform\\nwell — or at least, better than a slightly less complex model that does not\\nquite interpolate the data. This phenomenon is known as double descent ,\\nand is displayed in Figure 10.20. “Double descent” gets its name from the\\nfact that the test error has a U-shape before the interpolation threshold is\\nreached, and then it descends again (for a while, at least) as an increasingly\\nflexible model is fit.\\nWe now describe the set-up that resulted in Figure 10.20. We simulated\\nn= 20 observations from the model\\nY= sin( X)+ϵ,\\nwhereX∼U[−5,5](uniform distribution), and ϵ∼N(0,σ2)withσ=0.3.\\nWe then fit a natural spline to the data, as described in Section 7.4, withd', metadata={'source': './documents/ISLP_website.pdf', 'page': 438}), Document(page_content='10.8 Interpolation and Double Descent 433\\n−4 −2 0 2 4−3 −2 −1 0 1 2 38 Degrees of Freedom\\nseq(−5, 5, len = 1000)−4 −2 0 2 4−3 −2 −1 0 1 2 320 Degrees of Freedom\\nseq(−5, 5, len = 1000)f(seq(−5, 5, len = 1000))\\n−4 −2 0 2 4−3 −2 −1 0 1 2 342 Degrees of Freedom\\n−4 −2 0 2 4−3 −2 −1 0 1 2 380 Degrees of Freedomf(seq(−5, 5, len = 1000))FIGURE 10.21. Fitted functions ˆfd(X)(orange), true function f(X)(black)\\nand the observed 20training data points. A different value of d(degrees of freedom)\\nis used in each panel. For d≥20the orange curves all interpolate the training\\npoints, and hence the training error is zero.\\ndegrees of freedom.22Recall from Section 7.4that fitting a natural spline\\nwithddegrees of freedom amounts to fitting a least-squares regression\\nof the response onto a set of dbasis functions. The upper-left panel of\\nFigure10.21shows the data, the true function f(X), andˆf8(X), the fitted\\nnatural spline with d=8degrees of freedom.\\nNext,wefitanaturalsplinewith d= 20 degreesoffreedom.Since n= 20 ,\\nthis means that n=d, and we have zero training error; in other words, we\\nhave interpolated the training data! We can see from the top-right panel of\\nFigure10.21thatˆf20(X)makes wild excursions, and hence the test error\\nwill be large.\\nWe now continue to fit natural splines to the data, with increasing values\\nofd. Ford>20, the least squares regression of Yontodbasis functions\\nis not unique: there are an infinite number of least squares coefficient es-\\ntimates that achieve zero error. To select among them, we choose the one\\nwith the smallest sum of squared coefficients,∑d\\nj=1ˆβ2\\nj. This is known as\\ntheminimum-norm solution.\\nThe two lower panels of Figure 10.21show the minimum-norm natural\\nspline fits with d= 42 andd= 80 degrees of freedom. Incredibly, ˆf42(X)\\nis quite a bit lessless wild than ˆf20(X),even though it makes use of more\\ndegrees of freedom . Andˆf80(X)is not much different. How can this be?\\nEssentially, ˆf20(X)isverywildbecausethereisjustasinglewaytointerpo-\\nlaten= 20 observations using d= 20 basis functions, and that single way\\nresults in a somewhat extreme fitted function. By contrast, there are an\\n22This implies the choice of dknots, here chosen at dequi-probability quantiles of the\\ntraining data. When d>n , the quantiles are found by interpolation.', metadata={'source': './documents/ISLP_website.pdf', 'page': 439}), Document(page_content='434 10. Deep Learning\\ninfinite number of ways to interpolate n= 20 observations using d= 42 or\\nd= 80 basis functions, and the smoothest of them — that is, the minimum\\nnorm solution — is much less wild than ˆf20(X)!\\nIn Figure 10.20, we display the training error and test error associated\\nwithˆfd(X), for a range of values of the degrees of freedom d. We see that\\nthe training error drops to zero once d= 20 and beyond; i.e. once the\\ninterpolation threshold is reached. By contrast, the test error shows a U-\\nshape for d≤20, grows extremely large around d= 20 , and then shows a\\nsecond region of descent for d>20. For this example the signal-to-noise\\nratio — Var(f(X))/σ2— is5.9, which is quite high (the data points are\\nclose to the true curve). So an estimate that interpolates the data and does\\nnot wander too far inbetween the observed data points will likely do well.\\nIn Figures 10.20and10.21, we have illustrated the double descent phe-\\nnomenon in a simple one-dimensional setting using natural splines. How-\\never, it turns out that the same phenomenon can arise for deep learning.\\nBasically, when we fit neural networks with a huge number of parameters,\\nwe are sometimes able to get good results with zero training error. This is\\nparticularlytrueinproblemswithhighsignal-to-noiseratio,suchasnatural\\nimage recognition and language translation, for example. This is because\\nthe techniques used to fit neural networks, including stochastic gradient\\ndescent, naturally lend themselves to selecting a “smooth” interpolating\\nmodel that has good test-set performance on these kinds of problems.\\nSome points are worth emphasizing:\\n•The double-descent phenomenon does not contradict the bias-variance\\ntrade-off, as presented in Section 2.2.2.Rather, the double-descent\\ncurve seen in the right-hand side of Figure 10.20is a consequence of\\nthe fact that the x-axis displays the number of spline basis functions\\nused, which does not properly capture the true “flexibility” of models\\nthat interpolate the training data. Stated another way, in this exam-\\nple, the minimum-norm natural spline with d= 42 has lower variance\\nthan the natural spline with d= 20 .\\n•Most of the statistical learning methods seen in this book do not exhibit\\ndouble descent. For instance, regularization approaches typically do\\nnot interpolate the training data, and thus double descent does not\\noccur. This is not a drawback of regularized methods: they can give\\ngreat results without interpolating the data !\\nIn particular, in the examples here, if we had fit the natural splines\\nusing ridge regression with an appropriately-chosen penalty rather\\nthan least squares, then we would not have seen double descent, and\\nin fact would have obtained better test error results.\\n•In Chapter 9, we saw that maximal margin classifiers and SVMs that\\nhave zero training error nonetheless often achieve very good test error.\\nThis is in part because those methods seek smooth minimum norm\\nsolutions. This is similar to the fact that the minimum-norm natural\\nspline can give good results with zero training error.\\n•The double-descent phenomenon has been used by the machine learn-\\ning community to explain the successful practice of using an over-', metadata={'source': './documents/ISLP_website.pdf', 'page': 440}), Document(page_content='10.9 Lab: Deep Learning 435\\nparametrized neural network (many layers, and many hidden units),\\nand then fitting all the way to zero training error. However, fitting\\nto zero error is not always optimal, and whether it is advisable de-\\npends on the signal-to-noise ratio. For instance, we may use ridge\\nregularization to avoid overfitting a neural network, as in ( 10.31). In\\nthis case, provided that we use an appropriate choice for the tuning\\nparameter λ, we will never interpolate the training data, and thus\\nwill not see the double descent phenomenon. Nonetheless we can get\\nvery good test-set performance, likely much better than we would\\nhave achieved had we interpolated the training data. Early stopping\\nduring stochastic gradient descent can also serve as a form of regular-\\nization that prevents us from interpolating the training data, while\\nstill getting very good results on test data.\\nTo summarize: though double descent can sometimes occur in neural net-\\nworks, we typically do not want to rely on this behavior. Moreover, it\\nis important to remember that the bias-variance trade-off always holds\\n(though it is possible that test error as a function of flexibility may not\\nexhibit a U-shape, depending on how we have parametrized the notion of\\n“flexibility” on the x-axis).\\n10.9 Lab: Deep Learning\\nIn this section we demonstrate how to fit the examples discussed in the\\ntext. We use the Python torch package, along with the pytorch_lightningtorch\\npytorch_\\nlightningpackage which provides utilities to simplify fitting and evaluating mod-\\nels. This code can be impressively fast with certain special processors,\\nsuch as Apple’s new M1 chip. The package is well-structured, flexible, and\\nwill feel comfortable to Pythonusers. A good companion is the site py-\\ntorch.org/tutorials . Much of our code is adapted from there, as well as the\\npytorch_lightning documentation.23\\nWe start with several standard imports that we have seen before.\\nIn[1]: import numpy as np, pandas as pd\\nfrom matplotlib.pyplot import subplots\\nfrom sklearn.linear_model import \\\\\\n(LinearRegression,\\nLogisticRegression,\\nLasso)\\nfrom sklearn.preprocessing import StandardScaler\\nfrom sklearn.model_selection import KFold\\nfrom sklearn.pipeline import Pipeline\\nfrom ISLP import load_data\\nfrom ISLP.models import ModelSpec as MS\\nfrom sklearn.model_selection import \\\\\\n(train_test_split,\\nGridSearchCV)\\n23The precise URLs at the time of writing are https://pytorch.org/tutorials/\\nbeginner/basics/intro.html andhttps://pytorch-lightning.readthedocs.io/en/\\nlatest/.', metadata={'source': './documents/ISLP_website.pdf', 'page': 441}), Document(page_content='436 10. Deep Learning\\nTorch-Specific Imports\\nThere are a number of imports for torch. (These are not included with\\nISLP, so must be installed separately.) First we import the main library\\nand essential tools used to specify sequentially-structured networks.\\nIn[2]: import torch\\nfrom torch import nn\\nfrom torch.optim import RMSprop\\nfrom torch.utils.data import TensorDataset\\nThere are several other helper packages for torch. For instance, the\\ntorchmetrics package has utilities to compute various metrics to evalu-torchmetricsate performance when fitting a model. The torchinfo package provides atorchinfouseful summary of the layers of a model. We use the read_image() functionread_image()when loading test images in Section 10.9.4.\\nIn[3]: from torchmetrics import (MeanAbsoluteError,\\nR2Score)\\nfrom torchinfo import summary\\nfrom torchvision.io import read_image\\nThe package pytorch_lightning is a somewhat higher-level interface to\\ntorchthat simplifies the specification and fitting of models by reducing the\\namount of boilerplate code needed (compared to using torchalone).\\nIn[4]: from pytorch_lightning import Trainer\\nfrom pytorch_lightning.loggers import CSVLogger\\nIn order to reproduce results we use seed_everything() . We will alsoseed_\\neverything() instruct torchto use deterministic algorithms where possible.\\nIn[5]: from pytorch_lightning.utilities.seed import seed_everything\\nseed_everything(0, workers=True)\\ntorch.use_deterministic_algorithms(True, warn_only=True)\\nWe will use several datasets shipped with torchvision for our examples:torchvisiona pretrained network for image classification, as well as some transforms\\nused for preprocessing.\\nIn[6]: from torchvision.datasets import MNIST, CIFAR100\\nfrom torchvision.models import (resnet50,\\nResNet50_Weights)\\nfrom torchvision.transforms import (Resize,\\nNormalize,\\nCenterCrop,\\nToTensor)\\nWe have provided a few utilities in ISLPspecifically for this lab. The\\nSimpleDataModule andSimpleModule are simple versions of objects used\\ninpytorch_lightning , the high-level module for fitting torchmodels. Al-\\nthough more advanced uses such as computing on graphical processing\\nunits (GPUs) and parallel data processing are possible in this module, we\\nwill not be focusing much on these in this lab. The ErrorTracker handles\\ncollections of targets and predictions over each mini-batch in the validation\\nor test stage, allowing computation of the metric over the entire validation\\nor test data set.', metadata={'source': './documents/ISLP_website.pdf', 'page': 442}), Document(page_content=\"10.9 Lab: Deep Learning 437\\nIn[7]: from ISLP.torch import (SimpleDataModule,\\nSimpleModule,\\nErrorTracker,\\nrec_num_workers)\\nIn addition we have included some helper functions to load the IMDb\\ndatabase, as well as a lookup that maps integers to particular keys in the\\ndatabase. We’ve included a slightly modified copy of the preprocessed IMDb\\ndata from keras, a separate package for fitting deep learning models. Thiskerassaves us significant preprocessing and allows us to focus on specifying and\\nfitting the models themselves.\\nIn[8]: from ISLP.torch.imdb import (load_lookup,\\nload_tensor,\\nload_sparse,\\nload_sequential)\\nFinally, we introduce some utility imports not directly related to torch.\\nTheglob()function from the globmodule is used to find all files matchingglob()wildcardcharacters,whichwewilluseinourexampleapplyingthe ResNet50\\nmodel to some of our own images. The jsonmodule will be used to load ajsonJSON file for looking up classes to identify the labels of the pictures in the\\nResNet50 example.\\nIn[9]: from glob import glob\\nimport json\\n10.9.1 Single Layer Network on Hitters Data\\nWe start by fitting the models in Section 10.6on theHittersdata.\\nIn[10]: Hitters = load_data( 'Hitters' ).dropna()\\nn=H i t t e r s . s h a p e [ 0 ]\\nWe will fit two linear models (least squares and lasso) and compare their\\nperformance to that of a neural network. For this comparison we will use\\nmean absolute error on a validation dataset.\\nMAE(y,ˆy)=1\\nnn∑\\ni=1|yi−ˆyi|.\\nWe set up the model matrix and the response.\\nIn[11]: model = MS(Hitters.columns.drop( 'Salary' ), intercept=False)\\nX=m o d e l . f i t _ t r a n s f o r m ( H i t t e r s ) . t o _ n u m p y ( )\\nY=H i t t e r s [ 'Salary' ].to_numpy()\\nTheto_numpy() method above converts pandasdata frames or series toto_numpy()numpyarrays. We do this because we will need to use sklearnto fit the\\nlasso model, and it requires this conversion. We also use a linear regres-\\nsion method from sklearn, rather than the method in Chapter 3 from\\nstatsmodels , to facilitate the comparisons.\\nWe now split the data into test and training, fixing the random state\\nused by sklearnto do the split.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 443}), Document(page_content=\"438 10. Deep Learning\\nIn[12]: (X_train,\\nX_test,\\nY_train,\\nY_test) = train_test_split(X,\\nY,\\ntest_size=1/3,\\nrandom_state=1)\\nLinear Models\\nWe fit the linear model and evaluate the test error directly.\\nIn[13]: hit_lm = LinearRegression().fit(X_train, Y_train)\\nYhat_test = hit_lm.predict(X_test)\\nnp.abs(Yhat_test - Y_test).mean()\\nOut[13]: 259.7153\\nNext we fit the lasso using sklearn. We are using mean absolute error\\nto select and evaluate a model, rather than mean squared error. The spe-\\ncialized solver we used in Section 6.5.2uses only mean squared error. So\\nhere, with a bit more work, we cre atea cross-validation grid and perform\\nthe cross-validation directly.\\nWe encode a pipeline with two steps: we first normalize the features\\nusing a StandardScaler() transform, and then fit the lasso without further\\nnormalization.\\nIn[14]: scaler = StandardScaler(with_mean=True, with_std=True)\\nlasso = Lasso(warm_start=True, max_iter=30000)\\nstandard_lasso = Pipeline(steps=[( 'scaler' ,s c a l e r ) ,\\n('lasso' ,l a s s o ) ] )\\nWe need to create a grid of values for λ. As is common practice, we\\nchoose a grid of 100 values of λ, uniform on the log scale from lam_max\\ndown to 0.01*lam_max . Herelam_maxis the smallest value of λwith an all-\\nzero solution. This value equals the largest absolute inner-product between\\nany predictor and the (centered) response.24\\nIn[15]: X_s = scaler.fit_transform(X_train)\\nn=X _ s . s h a p e [ 0 ]\\nlam_max = np.fabs(X_s.T.dot(Y_train - Y_train.mean())).max() / n\\nparam_grid = { 'alpha' :n p . e x p ( n p . l i n s p a c e ( 0 ,n p . l o g ( 0 . 0 1 ) ,1 0 0 ) )\\n*l a m _ m a x }\\nNote that we had to transform the data first, since the scale of the vari-\\nables impacts the choice of λ. We now perform cross-validation using this\\nsequence of λvalues.\\nIn[16]: cv = KFold(10,\\nshuffle=True,\\nrandom_state=1)\\ngrid = GridSearchCV(lasso,\\n24The derivation of this result is beyond the scope of this book.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 444}), Document(page_content=\"10.9 Lab: Deep Learning 439param_grid,cv=cv,scoring='neg_mean_absolute_error')grid.fit(X_train, Y_train);Weextractthelassomodelwithbestcross-validatedmeanabsoluteerror,and evaluate its performance onX_testandY_test, which were not usedin cross-validation.In[17]:trained_lasso = grid.best_estimator_Yhat_test = trained_lasso.predict(X_test)np.fabs(Yhat_test - Y_test).mean()Out[17]:257.2382This is similar to the results we got for the linear model fit by least squares.However, these results can vary a lot for different train/test splits; we en-courage the reader to try a different seed in code block 12 and rerun thesubsequent code up to this point.Specifying a Network: Classes and InheritanceTo fit the neural network, we first set up a model structure that describesthe network. Doing so requires us to define new classes specific to the modelwe wish to fit. Typically this is done inpytorchby sub-classing a genericrepresentation of a network, which is the approach we take here. Althoughthis example is simple, we will go through the steps in some detail, since itwill serve us well for the more complex examples to follow.In[18]:class HittersModel(nn.Module):def __init__(self, input_size):super(HittersModel, self).__init__()self.flatten = nn.Flatten()self.sequential = nn.Sequential(nn.Linear(input_size, 50),nn.ReLU(),nn.Dropout(0.4),nn.Linear(50, 1))def forward(self, x):x=s e l f . f l a t t e n ( x )return torch.flatten(self.sequential(x))Theclassstatement identifies the code chunk as a declaration for a classHittersModelthat inherits from the base classnn.Module. This base class isubiquitous intorchand represents the mappings in the neural networks.Indented beneath theclassstatement are the methods of this class: inthis case__init__andforward. The__init__method is called when aninstance of the class is created as in the cell below. In the methods,selfalways refers to an instance of the class. In the__init__method, we haveattached two objects toselfas attributes:flattenandsequential. Theseare used in theforwardmethod to describe the map that this moduleimplements.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 445}), Document(page_content=\"440 10. Deep LearningThere is one additional line in the__init__method, which is a call tosuper(). This function allows subclasses (i.e.HittersModel) to access meth-super()ods of the class they inherit from. For example, the classnn.Modulehas itsown__init__method, which is different from theHittersModel.__init__()method we’ve written above. Usingsuper()allows us to call the methodof the base class. Fortorchmodels, we will always be making thissuper()call as it is necessary for the model to be properly interpreted bytorch.Theobjectnn.Modulehasmoremethodsthansimply__init__andforward.These methods are directly accessible toHittersModelinstances because ofthis inheritance. One such method we will see shortly is theeval()method,used to disable dropout for when we want to evaluate the model on testdata.In[19]:hit_model = HittersModel(X.shape[1])The objectself.sequentialis a composition of four maps. The firstmaps the 19 features ofHittersto 50 dimensions, introducing50×19 + 50parameters for the weights andinterceptof the map (often called thebias).ThislayeristhenmappedtoaReLUlayerfollowedbya40%dropoutlayer,and finally a linear map down to 1 dimension, again with a bias. The totalnumber of trainable parameters is therefore50×19 + 50 + 50 + 1 = 1051.The packagetorchinfoprovides asummary()function that neatly sum-marizes this information. We specify the size of the input and see the sizeof each tensor as it passes through layers of the network.In[20]:summary(hit_model,input_size=X_train.shape,col_names=['input_size','output_size','num_params'])Out[20]:=====================================================================Layer (type:depth-idx) Input Shape Output Shape Param#=====================================================================HittersModel[175, 19] [175]--Flatten: 1-1[175, 19] [175, 19] --Sequential: 1-2[175, 19] [175, 1] --Linear: 2-1[175, 19] [175, 50] 1,000ReLU: 2-2[175, 50] [175, 50] --Dropout: 2-3[175, 50][175, 50]--Linear: 2-4[175, 50] [175, 1] 51=====================================================================Total params: 1,051Trainable params: 1,051We have truncated the end of the output slightly, here and in subsequentuses.We now need to transform our training data into a form accessible totorch. The basic datatype intorchis atensor, which is very similar toanndarrayfrom early chapters. We also note here thattorchtypicallyworks with 32-bit (single precision) rather than 64-bit (double precision)floating point numbers. We therefore convert our data tonp.float32beforeforming the tensor. TheXandYtensors are then arranged into aDatasetDataset\", metadata={'source': './documents/ISLP_website.pdf', 'page': 446}), Document(page_content='10.9 Lab: Deep Learning 441\\nrecognized by torchusingTensorDataset() .Tensor\\nDataset()In[21]: X_train_t = torch.tensor(X_train.astype(np.float32))\\nY_train_t = torch.tensor(Y_train.astype(np.float32))\\nhit_train = TensorDataset(X_train_t, Y_train_t)\\nWe do the same for the test data.\\nIn[22]: X_test_t = torch.tensor(X_test.astype(np.float32))\\nY_test_t = torch.tensor(Y_test.astype(np.float32))\\nhit_test = TensorDataset(X_test_t, Y_test_t)\\nFinally, this dataset is passed to a DataLoader() which ultimately passes\\ndata into our network. While this may seem like a lot of overhead, this\\nstructure is helpful for more complex tasks where data may live on different\\nmachines, or where data must be passed to a GPU. We provide a helper\\nfunction SimpleDataModule() inISLPto make this task easier for standardSimpleData\\nModule() usage. One of its arguments is num_workers , which indicates how many\\nprocesses we will use for loading the data. For small data like Hitters\\nthis will have little effect, but it does provide an advantage for the MNIST\\nandCIFAR100 examples below. The torchpackage will inspect the process\\nrunning and determine a maximum number of workers.25We’ve included a\\nfunction rec_num_workers() to compute this so we know how many workers\\nmight be reasonable (here the max was 16).\\nIn[23]: max_num_workers = rec_num_workers()\\nThe general training setup in pytorch_lightning involves training, vali-\\ndation and test data. These are each represented by different data loaders.\\nDuring each epoch, we run a training step to learn the model and a vali-\\ndation step to track the error. The test data is typically used at the end of\\ntraining to evaluate the model.\\nIn this case, as we had split only into test and training, we’ll use the\\ntest data as validation data with the argument validation=hit_test . The\\nvalidation argument can be a float between 0 and 1, an integer, or a\\nDataset. If a float (respectively, integer), it is interpreted as a percentage\\n(respectively number) of the trainingobservations to be used for validation.\\nIf it is a Dataset, it is passed directly to a data loader.\\nIn[24]: hit_dm = SimpleDataModule(hit_train,\\nhit_test,\\nbatch_size=32,\\nnum_workers=min(4, max_num_workers),\\nvalidation=hit_test)\\nNext we must provide a pytorch_lightning module that controls the\\nsteps performed during the training process. We provide methods for our\\nSimpleModule() that simply record the value of the loss function and any\\nadditionalmetricsattheendofeachepoch.Theseoperationsarecontrolled\\nby the methods SimpleModule.[training/test/validation]_step() , though\\nwe will not be modifying these in our examples.\\n25This depends on the computing hardware and the number of cores available.', metadata={'source': './documents/ISLP_website.pdf', 'page': 447}), Document(page_content=\"442 10. Deep Learning\\nIn[25]: hit_module = SimpleModule.regression(hit_model,\\nmetrics={ 'mae':MeanAbsoluteError()})\\nBy using the SimpleModule.regression() method, we indicate that weSimpleModule.\\nregression() will use squared-error loss as in ( 10.23). We have also asked for mean ab-\\nsolute error to be tracked as well in the metrics that are logged.\\nWe log our results via CSVLogger() , which in this case stores the results\\nin a CSV file within a directory logs/hitters . After the fitting is complete,\\nthis allows us to load the results as a pd.DataFrame() and visualize them\\nbelow. There are several ways to log the results within pytorch_lightning ,\\nthough we will not cover those here in detail.\\nIn[26]: hit_logger = CSVLogger( 'logs',n a m e = 'hitters' )\\nFinally we are ready to train our model and log the results. We use\\ntheTrainer() object from pytorch_lightning to do this work. The argu-\\nmentdatamodule=hit_dm tells the trainer how training/validation/test logs\\nare produced, while the first argument hit_module specifies the network\\narchitecture as well as the training/validation/test steps. The callbacks\\nargument allows for several tasks to be carried out at various points while\\ntraining a model. Here our ErrorTracker() callback will enable us to com-\\npute validation error while training and, finally, the test error. We now fit\\nthe model for 50 epochs.\\nIn[27]: hit_trainer = Trainer(deterministic=True,\\nmax_epochs=50,\\nlog_every_n_steps=5,\\nlogger=hit_logger,\\ncallbacks=[ErrorTracker()])\\nhit_trainer.fit(hit_module, datamodule=hit_dm)\\nAt each step of SGD, the algorithm randomly selects 32 training observa-\\ntions for the computation of the gradient. Recall from Section 10.7that an\\nepoch amounts to the number of SGD steps required to process nobserva-\\ntions. Since the training set has n= 175, and we specified a batch_size of\\n32 in the construction of hit_dm, an epoch is 175/32 = 5.5SGD steps.\\nAfter having fit the model, we can evaluate performance on our test data\\nusing the test()method of our trainer.\\nIn[28]: hit_trainer.test(hit_module, datamodule=hit_dm)\\nOut[28]: [{'test_loss' :1 0 4 0 9 8 . 5 4 6 9 , 'test_mae' :2 2 9 . 5 0 1 2 } ]\\nThe results of the fit have been logged into a CSV file. We can find the\\nresults specific to this run in the experiment.metrics_file_path attribute\\nof our logger. Note that each time the model is fit, the logger will output\\nresults into a new subdirectory of our directory logs/hitters .\\nWe now create a plot of the MAE (mean absolute error) as a function of\\nthe number of epochs. First we retrieve the logged summaries.\\nhit_results = pd.read_csv(hit_logger.experiment.metrics_file_path)\\nSince we will produce similar plots in later examples, we write a simple\\ngeneric function to produce this plot.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 448}), Document(page_content=\"10.9 Lab: Deep Learning 443In[29]:def summary_plot(results,ax,col='loss',valid_legend='Validation',training_legend='Training',ylabel='Loss',fontsize=20):for (column,color,label) in zip([f'train_{col}_epoch',f'valid_{col}'],['black','red'],[training_legend,valid_legend]):results.plot(x='epoch',y=column,label=label,marker='o',color=color,ax=ax)ax.set_xlabel('Epoch')ax.set_ylabel(ylabel)return axWe now set up our axes, and use our function to produce the MAE plot.In[30]:fig, ax = subplots(1, 1, figsize=(6, 6))ax = summary_plot(hit_results,ax,col='mae',ylabel='MAE',valid_legend='Validation (=Test)')ax.set_ylim([0, 400])ax.set_xticks(np.linspace(0, 50, 11).astype(int));We can predict directly from the final model, and evaluate its per-formance on the test data. Before fitting, we call theeval()method ofhit_model. This tellstorchto effectively consider this model to be fitted,sothatwecanuseittopredictonnewdata.Forourmodelhere,thebiggestchange is that the dropout layers will be turned off, i.e. no weights will berandomly dropped in predicting on new data.In[31]:hit_model.eval()preds = hit_module(X_test_t)torch.abs(Y_test_t - preds).mean()Out[31]:tensor(229.5012, grad_fn=<MeanBackward0>)CleanupIn setting up our data module, we had initiated several worker processesthat will remain running. We delete all references to the torch objects toensure these processes will be killed.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 449}), Document(page_content=\"444 10. Deep Learning\\nIn[32]: del(Hitters,\\nhit_model, hit_dm,\\nhit_logger,\\nhit_test, hit_train,\\nX, Y,\\nX_test, X_train,\\nY_test, Y_train,\\nX_test_t, Y_test_t,\\nhit_trainer, hit_module)\\n10.9.2 Multilayer Network on the MNIST Digit Data\\nThetorchvision packagecomeswithanumberofexampledatasets,includ-\\ning the MNISTdigit data. Our first step is to retrieve the training and test\\ndata sets; the MNIST()function within torchvision.datasets is provided forMNIST()this purpose. The data will be downloaded the first time this function is\\nexecuted, and stored in the directory data/MNIST .\\nIn[33]: (mnist_train,\\nmnist_test) = [MNIST(root= 'data',\\ntrain=train,\\ndownload=True,\\ntransform=ToTensor())\\nfor train in [True, False]]\\nmnist_train\\nOut[33]: Dataset MNIST\\nNumber of datapoints: 60000\\nRoot location: data\\nSplit: Train\\nStandardTransform\\nTransform: ToTensor()\\nThere are 60,000 images in the training data and 10,000 in the test data.\\nThe images are 28×28, and stored as a matrix of pixels. We need to\\ntransform each one into a vector.\\nNeural networks are somewhat sensitive to the scale of the inputs, much\\nas ridge and lasso regularization are affected by scaling. Here the inputs\\nare eight-bit grayscale values between 0 and 255, so we rescale to the unit\\ninterval.26This transformation, along with some reordering of the axes, is\\nperformed by the ToTensor() transform from the torchvision.transforms\\npackage.\\nAs in our Hittersexample, we form a data module from the training and\\ntest datasets, setting aside 20% of the training images for validation.\\nIn[34]: mnist_dm = SimpleDataModule(mnist_train,\\nmnist_test,\\nvalidation=0.2,\\nnum_workers=max_num_workers,\\nbatch_size=256)\\n26Note: eight bits means 28, which equals 256. Since the convention is to start at 0,\\nthe possible values range from 0to255.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 450}), Document(page_content=\"10.9 Lab: Deep Learning 445Let’s take a look at the data that will get fed into our network. We loopthrough the first few chunks of the test dataset, breaking after 2 batches:In[35]:for idx, (X_ ,Y_) in enumerate(mnist_dm.train_dataloader()):print('X: ',X _ . s h a p e )print('Y: ',Y _ . s h a p e )if idx >= 1:breakX: torch.Size([256, 1, 28, 28])Y: torch.Size([256])X: torch.Size([256, 1, 28, 28])Y: torch.Size([256])We see that theXfor each batch consists of 256 images of size1x28x28.Here the1indicates a single channel (greyscale). For RGB images such asCIFAR100below, we will see that the1in the size will be replaced by3forthe three RGB channels.Now we are ready to specify our neural network.In[36]:class MNISTModel(nn.Module):def __init__(self):super(MNISTModel, self).__init__()self.layer1 = nn.Sequential(nn.Flatten(),nn.Linear(28*28, 256),nn.ReLU(),nn.Dropout(0.4))self.layer2 = nn.Sequential(nn.Linear(256, 128),nn.ReLU(),nn.Dropout(0.3))self._forward = nn.Sequential(self.layer1,self.layer2,nn.Linear(128, 10))def forward(self, x):return self._forward(x)We see that in the first layer, each1x28x28image is flattened, thenmapped to 256 dimensions where we apply a ReLU activation with 40%dropout. A second layer maps the first layer’s output down to 128 di-mensions, applying a ReLU activation with 30% dropout. Finally, the 128dimensions are mapped down to 10, the number of classes in theMNISTdata.In[37]:mnist_model = MNISTModel()We can check that the model produces output of expected size based onour existing batchX_above.In[38]:mnist_model(X_).size()Out[38]:torch.Size([256, 10])Let’s take a look at the summary of the model. Instead of aninput_sizewe can pass a tensor of correct shape. In this case, we pass through thefinal batchedX_from above.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 451}), Document(page_content=\"446 10. Deep LearningIn[39]:summary(mnist_model,input_data=X_,col_names=['input_size','output_size','num_params'])Out[39]:=====================================================================Layer (type:depth-idx) Input Shape Output Shape Param#=====================================================================MNISTModel[256, 1, 28, 28] [256, 10] --Sequential: 1-1[256, 1, 28, 28] [256, 10] --Sequential: 2-1 [256, 1, 28, 28] [256, 256] --Flatten: 3-1 [256, 1, 28, 28] [256, 784] --Linear: 3-2 [256, 784] [256, 256] 200,960ReLU: 3-3 [256, 256] [256, 256] --Dropout: 3-4 [256, 256] [256, 256] --Sequential: 2-2 [256, 256] [256, 128] --Linear: 3-5 [256, 256] [256, 128] 32,896ReLU: 3-6 [256, 128] [256, 128] --Dropout: 3-7 [256, 128] [256, 128] --Linear: 2-3[256, 128] [256, 10] 1,290=====================================================================Total params: 235,146Trainable params: 235,146Having set up both the model and the data module, fitting this model isnow almost identical to theHittersexample. In contrast to our regressionmodel, here we will use theSimpleModule.classification()method whichSimpleModule.classifi-cation()uses the cross-entropy loss function instead of mean squared error.In[40]:mnist_module = SimpleModule.classification(mnist_model)mnist_logger = CSVLogger('logs',n a m e ='MNIST')Now we are ready to go. The final step is to supply training data, andfit the model.In[41]:mnist_trainer = Trainer(deterministic=True,max_epochs=30,logger=mnist_logger,callbacks=[ErrorTracker()])mnist_trainer.fit(mnist_module,datamodule=mnist_dm)We have suppressed the output here, which is a progress report on thefitting of the model, grouped by epoch. This is very useful, since on largedatasets fitting can take time. Fitting this model took 245 seconds on aMacBook Pro with an Apple M1 Pro chip with 10 cores and 16 GB ofRAM. Here we specified a validation split of 20%, so training is actuallyperformed on 80% of the 60,000 observations in the training set. This is analternativetoactuallysupplyingvalidationdata,likewedidfortheHittersdata. SGD uses batches of 256 observations in computing the gradient, anddoing the arithmetic, we see that an epoch corresponds to 188 gradientsteps.SimpleModule.classification()includes an accuracy metric by default.Other classification metrics can be added fromtorchmetrics. We will useoursummary_plot()function to display accuracy across epochs.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 452}), Document(page_content=\"10.9 Lab: Deep Learning 447\\nIn[42]: mnist_results = pd.read_csv(mnist_logger.experiment.\\nmetrics_file_path)\\nfig, ax = subplots(1, 1, figsize=(6, 6))\\nsummary_plot(mnist_results,\\nax,\\ncol='accuracy' ,\\nylabel= 'Accuracy' )\\nax.set_ylim([0.5, 1])\\nax.set_ylabel( 'Accuracy' )\\nax.set_xticks(np.linspace(0, 30, 7).astype(int));\\nOnce again we evaluate the accuracy using the test()method of our\\ntrainer. This model achieves 97% accuracy on the test data.\\nIn[43]: mnist_trainer.test(mnist_module,\\ndatamodule=mnist_dm)\\nOut[43]: [{'test_loss' :0 . 1 4 7 1 , 'test_accuracy' :0 . 9 6 8 1 } ]\\nTable10.1also reports the error rates resulting from LDA (Chapter 4) and\\nmulticlass logistic regression. For LDA we refer the reader to Section 4.7.3.\\nAlthough we could use the sklearnfunction LogisticRegression() to fit\\nmulticlass logistic regression, we are set up here to fit such a model with\\ntorch.Wejusthaveaninputlayerandanoutputlayer,andomitthehidden\\nlayers!\\nIn[44]: class MNIST_MLR(nn.Module):\\ndef __init__(self):\\nsuper(MNIST_MLR, self).__init__()\\nself.linear = nn.Sequential(nn.Flatten(),\\nnn.Linear(784, 10))\\ndef forward(self, x):\\nreturn self.linear(x)\\nmlr_model = MNIST_MLR()\\nmlr_module = SimpleModule.classification(mlr_model)\\nmlr_logger = CSVLogger( 'logs',n a m e = 'MNIST_MLR' )\\nIn[45]: mlr_trainer = Trainer(deterministic=True,\\nmax_epochs=30,\\ncallbacks=[ErrorTracker()])\\nmlr_trainer.fit(mlr_module, datamodule=mnist_dm)\\nWe fit the model just as before and compute the test results.\\nIn[46]: mlr_trainer.test(mlr_module,\\ndatamodule=mnist_dm)\\nOut[46]: [{'test_loss' :0 . 3 1 8 7 , 'test_accuracy' :0 . 9 2 4 1 } ]\\nThe accuracy is above 90% even for this pretty simple model.\\nAs in the Hitters example, we delete some of the objects we created\\nabove.\\nIn[47]: del(mnist_test,\\nmnist_train,\", metadata={'source': './documents/ISLP_website.pdf', 'page': 453}), Document(page_content='448 10. Deep Learningmnist_model,mnist_dm,mnist_trainer,mnist_module,mnist_results,mlr_model,mlr_module,mlr_trainer)10.9.3 Convolutional Neural NetworksIn this section we fit a CNN to theCIFAR100data, which is available in thetorchvisionpackage. It is arranged in a similar fashion as theMNISTdata.In[48]:(cifar_train,cifar_test) = [CIFAR100(root=\"data\",train=train,download=True)for train in [True, False]]In[49]:transform = ToTensor()cifar_train_X = torch.stack([transform(x) for x incifar_train.data])cifar_test_X = torch.stack([transform(x) for x incifar_test.data])cifar_train = TensorDataset(cifar_train_X,torch.tensor(cifar_train.targets))cifar_test = TensorDataset(cifar_test_X,torch.tensor(cifar_test.targets))TheCIFAR100datasetconsistsof50,000trainingimages,eachrepresentedby a three-dimensional tensor: each three-color image is represented as aset of three channels, each of which consists of32×32eight-bit pixels. Westandardize as we did for the digits, but keep the array structure. This isaccomplished with theToTensor()transform.Creating the data module is similar to theMNISTexample.In[50]:cifar_dm = SimpleDataModule(cifar_train,cifar_test,validation=0.2,num_workers=max_num_workers,batch_size=128)We again look at the shape of typical batches in our data loaders.In[51]:for idx, (X_ ,Y_) in enumerate(cifar_dm.train_dataloader()):print(\\'X: \\',X _ . s h a p e )print(\\'Y: \\',Y _ . s h a p e )if idx >= 1:breakX: torch.Size([128, 3, 32, 32])Y: torch.Size([128])X: torch.Size([128, 3, 32, 32])Y: torch.Size([128])', metadata={'source': './documents/ISLP_website.pdf', 'page': 454}), Document(page_content='10.9 Lab: Deep Learning 449\\nBefore we start, we look at some of the training images; similar code\\nproduced Figure 10.5on page406. The example below also illustrates that\\nTensorDataset objects can be indexed with integers — we are choosing ran-\\ndomimagesfromthetrainingdatabyindexing cifar_train .Inordertodis-\\nplay correctly, we must reorder the dimensions by a call to np.transpose() .\\nIn[52]: fig, axes = subplots(5, 5, figsize=(10,10))\\nrng = np.random.default_rng(4)\\nindices = rng.choice(np.arange(len(cifar_train)), 25,\\nreplace=False).reshape((5,5))\\nfor i in range(5):\\nfor j in range(5):\\nidx = indices[i,j]\\naxes[i,j].imshow(np.transpose(cifar_train[idx][0],\\n[1,2,0]),\\ninterpolation=None)\\naxes[i,j].set_xticks([])\\naxes[i,j].set_yticks([])\\nHere the imshow() method recognizes from the shape of its argument that.imshow()it is a 3-dimensional array, with the last dimension indexing the three RGB\\ncolor channels.\\nWe specify a moderately-sized CNN for demonstration purposes, simi-\\nlar in structure to Figure 10.8. We use several layers, each consisting of\\nconvolution, ReLU, and max-pooling steps. We first define a module that\\ndefines one of these layers. As in our previous examples, we overwrite the\\n__init__() andforward() methods of nn.Module . This user-defined module\\ncan now be used in ways just like nn.Linear() ornn.Dropout() .\\nIn[53]: class BuildingBlock(nn.Module):\\ndef __init__(self,\\nin_channels,\\nout_channels):\\nsuper(BuildingBlock, self).__init__()\\nself.conv = nn.Conv2d(in_channels=in_channels,\\nout_channels=out_channels,\\nkernel_size=(3,3),\\npadding= \\'same\\')\\nself.activation = nn.ReLU()\\nself.pool = nn.MaxPool2d(kernel_size=(2,2))\\ndef forward(self, x):\\nreturn self.pool(self.activation(self.conv(x)))\\nNotice that we used the padding = \"same\" argument to nn.Conv2d() ,\\nwhich ensures that the output channels have the same dimension as the\\ninput channels. There are 32 channels in the first hidden layer, in contrast\\nto the three channels in the input layer. We use a 3×3convolution fil-\\nter for each channel in all the layers. Each convolution is followed by a\\nmax-pooling layer over 2×2blocks.\\nIn forming our deep learning model for the CIFAR100 data, we use several\\nof ourBuildingBlock() modules sequentially. This simple example illus-\\ntrates some of the power of torch. Users can define modules of their own,', metadata={'source': './documents/ISLP_website.pdf', 'page': 455}), Document(page_content=\"450 10. Deep Learningwhich can be combined in other modules. Ultimately, everything is fit bya generic trainer.In[54]:class CIFARModel(nn.Module):def __init__(self):super(CIFARModel, self).__init__()sizes = [(3,32),(32,64),(64,128),(128,256)]self.conv = nn.Sequential(*[BuildingBlock(in_, out_)for in_, out_ in sizes])self.output = nn.Sequential(nn.Dropout(0.5),nn.Linear(2*2*256, 512),nn.ReLU(),nn.Linear(512, 100))def forward(self, x):val = self.conv(x)val = torch.flatten(val, start_dim=1)return self.output(val)We build the model and look at the summary. (We had created examplesofX_earlier.)In[55]:cifar_model = CIFARModel()summary(cifar_model,input_data=X_,col_names=['input_size','output_size','num_params'])Out[55]:======================================================================Layer (type:depth-idx) Input Shape Output Shape Param#======================================================================CIFARModel[128, 3, 32, 32] [128, 100]--Sequential: 1-1 [128, 3, 32, 32] [128, 256, 2, 2] --BuildingBlock: 2-1 [128, 3, 32, 32] [128, 32, 16, 16] --Conv2d: 3-1 [128, 3, 32, 32] [128, 32, 32, 32] 896ReLU: 3-2 [128, 32, 32, 32] [128, 32, 32, 32] --MaxPool2d: 3-3 [128, 32, 32, 32] [128, 32, 16, 16] --BuildingBlock: 2-2 [128, 32, 16, 16] [128, 64, 8, 8] --Conv2d: 3-4 [128, 32, 16, 16] [128, 64, 16, 16] 18,496ReLU: 3-5 [128, 64, 16, 16] [128, 64, 16, 16] --MaxPool2d: 3-6 [128, 64, 16, 16] [128, 64, 8, 8] --BuildingBlock: 2-3 [128, 64, 8, 8] [128, 128, 4, 4] --Conv2d: 3-7 [128, 64, 8, 8] [128, 128, 8, 8] 73,856ReLU: 3-8 [128, 128, 8, 8] [128, 128, 8, 8] --MaxPool2d: 3-9 [128, 128, 8, 8] [128, 128, 4, 4] --BuildingBlock: 2-4 [128, 128, 4, 4] [128, 256, 2, 2] --Conv2d: 3-10 [128, 128, 4, 4] [128, 256, 4, 4] 295,168ReLU: 3-11 [128, 256, 4, 4] [128, 256, 4, 4] --MaxPool2d: 3-12 [128, 256, 4, 4] [128, 256, 2, 2] --Sequential: 1-2 [128, 1024] [128, 100]--Dropout: 2-5 [128, 1024] [128, 1024] --Linear: 2-6 [128, 1024] [128, 512]524,800\", metadata={'source': './documents/ISLP_website.pdf', 'page': 456}), Document(page_content=\"10.9 Lab: Deep Learning 451ReLU: 2-7 [128, 512][128, 512]--Linear: 2-8 [128, 512][128, 100]51,300======================================================================Total params: 964,516Trainable params: 964,516The total number of trainable parameters is 964,516. By studying the sizeof the parameters, we can see that the channels halve in both dimensionsafter each of these max-pooling operations. After the last of these we havea layer with 256 channels of dimension2×2. These are then flattened toa dense layer of size 1,024; in other words, each of the2×2matrices isturned into a4-vector, and put side-by-side in one layer. This is followedby a dropout regularization layer, then another dense layer of size 512, andfinally, the output layer.Uptonow,wehavebeenusingadefaultoptimizerinSimpleModule().Forthese data, experiments show that a smaller learning rate performs betterthan the default 0.01. We use a custom optimizer here with a learning rateof 0.001. Besides this, the logging and training follow a similar patternto our previous examples. The optimizer takes an argumentparamsthatinforms the optimizer which parameters are involved in SGD (stochasticgradient descent).We saw earlier that entries of a module’s parameters are tensors. Inpassing the parameters to the optimizer we are doing more than simplypassing arrays; part of the structure of the graph is encoded in the tensorsthemselves.In[56]:cifar_optimizer = RMSprop(cifar_model.parameters(), lr=0.001)cifar_module = SimpleModule.classification(cifar_model,optimizer=cifar_optimizer)cifar_logger = CSVLogger('logs',n a m e ='CIFAR100')In[57]:cifar_trainer = Trainer(deterministic=True,max_epochs=30,logger=cifar_logger,callbacks=[ErrorTracker()])cifar_trainer.fit(cifar_module,datamodule=cifar_dm)This model takes 10 minutes or more to run and achieves about 42%accuracy on the test data. Although this is not terrible for 100-class data(a random classifier gets 1% accuracy), searching the web we see resultsaround 75%. Typically it takes a lot of architecture carpentry, fiddling withregularization, and time, to achieve such results.Let’s take a look at the validation and training accuracy across epochs.In[58]:log_path = cifar_logger.experiment.metrics_file_pathcifar_results = pd.read_csv(log_path)fig, ax = subplots(1, 1, figsize=(6, 6))summary_plot(cifar_results,ax,col='accuracy',ylabel='Accuracy')ax.set_xticks(np.linspace(0, 10, 6).astype(int))ax.set_ylabel('Accuracy')ax.set_ylim([0, 1]);\", metadata={'source': './documents/ISLP_website.pdf', 'page': 457}), Document(page_content=\"452 10. Deep Learning\\nFinally, we evaluate our model on our test data.\\nIn[59]: cifar_trainer.test(cifar_module,\\ndatamodule=cifar_dm)\\nOut[59]: [{'test_loss' :2 . 4 2 3 8 'test_accuracy' :0 . 4 2 0 6 } ]\\nHardware Acceleration\\nAs deep learning has become ubiquitous in machine learning, hardware\\nmanufacturers have produced special libraries that can often speed up the\\ngradient-descent steps.\\nFor instance, Mac OS devices with the M1 chip may have the Metalpro-\\ngramming framework enabled, which can speed up the torchcomputations.\\nWe present an example of how to use this acceleration.\\nThe main changes are to the Trainer() call as well as to the metrics that\\nwill be evaluated on the data. These metrics must be told where the data\\nwill be located at evaluation time. This is accomplished with a call to the\\nto()method of the metrics.\\nIn[60]: try:\\nfor name, metric in cifar_module.metrics.items():\\ncifar_module.metrics[name] = metric.to( 'mps')\\ncifar_trainer_mps = Trainer(accelerator= 'mps',\\ndeterministic=True,\\nmax_epochs=30)\\ncifar_trainer_mps.fit(cifar_module,\\ndatamodule=cifar_dm)\\ncifar_trainer_mps.test(cifar_module,\\ndatamodule=cifar_dm)\\nexcept:\\npass\\nThis yields approximately two- or three-fold acceleration for each epoch.\\nWe have protected this code block using try:andexcept: clauses; if it\\nworks, we get the speedup, if it fails, nothing happens.\\n10.9.4 Using Pretrained CNN Models\\nWe now show how to use a CNN pretrained on the imagenet database to\\nclassify natural images, and demonstrate how we produced Figure 10.10.\\nWe copied six JPEG images from a digital photo album into the direc-\\ntorybook_images . These images are available from the data section of www.\\nstatlearning.com , the ISLP book website. Download book_images.zip ;\\nwhen clicked it creates the book_images directory.\\nThe pretrained network we use is called resnet50 ; specification details\\ncan be found on the web. We will read in the images, and convert them into\\nthe array format expected by the torchsoftware to match the specifications\\ninresnet50 . The conversion involves a resize, a crop and then a predefined\\nstandardization for each of the three channels. We now read in the images\\nand preprocess them.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 458}), Document(page_content=\"10.9 Lab: Deep Learning 453\\nIn[61]: resize = Resize((232,232))\\ncrop = CenterCrop(224)\\nnormalize = Normalize([0.485,0.456,0.406],\\n[0.229,0.224,0.225])\\nimgfiles = sorted([f for f in glob( 'book_images/*' )])\\nimgs = torch.stack([torch.div(crop(resize(read_image(f))), 255)\\nfor f in imgfiles])\\nimgs = normalize(imgs)\\nimgs.size()\\nOut[61]: torch.Size([6, 3, 224, 224])\\nWe now set up the trained network with the weights we read in code\\nblock 6. The model has 50 layers, with a fair bit of complexity.\\nIn[62]: resnet_model = resnet50(weights=ResNet50_Weights.DEFAULT)\\nsummary(resnet_model,\\ninput_data=imgs,\\ncol_names=[ 'input_size' ,\\n'output_size' ,\\n'num_params' ])\\nWe set the mode to eval()to ensure that the model is ready to predict on\\nnew data.\\nIn[63]: resnet_model.eval()\\nInspecting the output above, we see that when setting up the resnet_model ,\\nthe authors defined a Bottleneck , much like our BuildingBlock module.\\nWe now feed our six images through the fitted network.\\nIn[64]: img_preds = resnet_model(imgs)\\nLet’s look at the predicted probabilities for each of the top 3 choices.\\nFirst we compute the probabilities by applying the softmax to the logits\\ninimg_preds . Note that we have had to call the detach() method on the\\ntensorimg_preds in order to convert it to our a more familiar ndarray.\\nIn[65]: img_probs = np.exp(np.asarray(img_preds.detach()))\\nimg_probs /= img_probs.sum(1)[:,None]\\nInordertoseetheclasslabels,wemustdownloadtheindexfileassociated\\nwithimagenet .27\\nIn[66]: labs = json.load(open( 'imagenet_class_index.json' ))\\nclass_labels = pd.DataFrame([(int(k), v[1]) for k, v in\\nlabs.items()],\\ncolumns=[ 'idx','label' ])\\nclass_labels = class_labels.set_index( 'idx')\\nclass_labels = class_labels.sort_index()\\nWe’ll now construct a data frame for each image file with the labels with\\nthe three highest probabilities as estimated by the model above.\\n27This is avalable from the book website and s3.amazonaws.com/deep-learning-\\nmodels/image-models/imagenet_class_index.json .\", metadata={'source': './documents/ISLP_website.pdf', 'page': 459}), Document(page_content=\"454 10. Deep Learning\\nIn[67]: for i, imgfile in enumerate(imgfiles):\\nimg_df = class_labels.copy()\\nimg_df[ 'prob']=i m g _ p r o b s [ i ]\\nimg_df = img_df.sort_values(by= 'prob',a s c e n d i n g = F a l s e ) [ : 3 ]\\nprint(f 'Image: {imgfile}' )\\nprint(img_df.reset_index().drop(columns=[ 'idx']))\\nImage: book_images/Cape_Weaver.jpg\\nlabel prob\\n0j a c a m a r 0 . 2 8 7 2 8 3\\n1b e e _ e a t e r0 . 0 4 6 7 6 8\\n2b u l b u l 0 . 0 3 7 5 0 7\\nImage: book_images/Flamingo.jpg\\nlabel prob\\n0f l a m i n g o 0 . 5 9 1 7 6 1\\n1s p o o n b i l l 0 . 0 1 2 3 8 6\\n2A m e r i c a n _ e g r e t0 . 0 0 2 1 0 5\\nImage: book_images/Hawk_Fountain.jpg\\nlabel prob\\n0g r e a t _ g r e y _ o w l0 . 2 8 7 9 5 9\\n1 kite 0.039478\\n2f o u n t a i n 0 . 0 2 9 3 8 4\\nImage: book_images/Hawk_cropped.jpg\\nlabel prob\\n0k i t e 0 . 3 0 1 8 3 0\\n1j a y 0 . 1 2 1 6 7 4\\n2m a g p i e0 . 0 1 5 5 1 3\\nImage: book_images/Lhasa_Apso.jpg\\nlabel prob\\n0 Lhasa 0.151143\\n1 Shih-Tzu 0.129850\\n2T i b e t a n _ t e r r i e r0 . 1 0 2 3 5 8\\nImage: book_images/Sleeping_Cat.jpg\\nlabel prob\\n0t a b b y 0 . 1 7 3 6 2 7\\n1t i g e r _ c a t0 . 1 1 0 4 1 4\\n2d o o r m a t 0 . 0 9 3 4 4 7\\nWe see that the model is quite confident about Flamingo.jpg , but a little\\nless so for the other images.\\nWe end this section with our usual cleanup.\\nIn[68]: del(cifar_test,\\ncifar_train,\\ncifar_dm,\\ncifar_module,\\ncifar_logger,\\ncifar_optimizer,\\ncifar_trainer)\\n10.9.5 IMDB Document Classification\\nWe now implement models for sentiment classification (Section 10.4) on\\ntheIMDBdataset. As mentioned above code block 8, we are using a prepro-\\ncessed version of the IMDBdataset found in the keraspackage. As kerasuses\", metadata={'source': './documents/ISLP_website.pdf', 'page': 460}), Document(page_content='10.9 Lab: Deep Learning 455tensorflow, a different tensor and deep learning library, we have convertedthe data to be suitable fortorch. The code used to convert fromkerasis available in the moduleISLP.torch._make_imdb. It requires some of thekeraspackages to run. These data use a dictionary of size 10,000.We have stored three different representations of the review data for thislab:•load_tensor(), a sparse tensor version usable bytorch;•load_sparse(), a sparse matrix version usable bysklearn, since wewill compare with a lasso fit;•load_sequential(), a padded version of the original sequence repre-sentation, limited to the last 500 words of each review.In[69]:(imdb_seq_train,imdb_seq_test) = load_sequential(root=\\'data/IMDB\\')padded_sample = np.asarray(imdb_seq_train.tensors[0][0])sample_review = padded_sample[padded_sample > 0][:12]sample_review[:12]Out[69]:array([ 1, 14, 22, 16, 43, 530, 973, 1622, 1385,65, 458, 4468], dtype=int32)The datasetsimdb_seq_trainandimdb_seq_testare both instances of theclassTensorDataset. The tensors used to construct them can be found inthetensorsattribute, with the first tensor the featuresXand the secondthe outcomeY. We have taken the first row of features and stored it aspadded_sample. In the preprocessing used to form these data, sequenceswere padded with 0s in the beginning if they were not long enough, hencewe remove this padding by restricting to entries wherepadded_sample > 0.We then provide the first 12 words of the sample review.WecanfindthesewordsinthelookupdictionaryfromtheISLP.torch.imdbmodule.In[70]:lookup = load_lookup(root=\\'data/IMDB\\')\\'\\'.join(lookup[i] for i in sample_review)Out[70]:\"<START> this film was just brilliant casting location scenerystory direction everyone\\'s\"For our first model, we have created a binary feature for each of the10,000 possible words in the dataset, with an entry of one in thei, jentryifwordjappearsinreviewi.Asmostreviewsarequiteshort,suchafeaturematrix has over 98% zeros. These data are accessed usingload_tensor()from theISLPlibrary.In[71]:max_num_workers=10(imdb_train,imdb_test) = load_tensor(root=\\'data/IMDB\\')imdb_dm = SimpleDataModule(imdb_train,imdb_test,validation=2000,num_workers=min(6, max_num_workers),batch_size=512)', metadata={'source': './documents/ISLP_website.pdf', 'page': 461}), Document(page_content=\"456 10. Deep Learning\\nWe’ll use a two-layer model for our first model.\\nIn[72]: class IMDBModel(nn.Module):\\ndef __init__(self, input_size):\\nsuper(IMDBModel, self).__init__()\\nself.dense1 = nn.Linear(input_size, 16)\\nself.activation = nn.ReLU()\\nself.dense2 = nn.Linear(16, 16)\\nself.output = nn.Linear(16, 1)\\ndef forward(self, x):\\nval = x\\nfor _map in [self.dense1,\\nself.activation,\\nself.dense2,\\nself.activation,\\nself.output]:\\nval = _map(val)\\nreturn torch.flatten(val)\\nWe now instantiate our model and look at a summary (not shown).\\nIn[73]: imdb_model = IMDBModel(imdb_test.tensors[0].size()[1])\\nsummary(imdb_model,\\ninput_size=imdb_test.tensors[0].size(),\\ncol_names=[ 'input_size' ,\\n'output_size' ,\\n'num_params' ])\\nWe’ll again use a smaller learning rate for these data, hence we pass an\\noptimizer to theSimpleModule . Since the reviews are classified into positive\\nor negative sentiment, we use SimpleModule.binary_classification() .28\\nIn[74]: imdb_optimizer = RMSprop(imdb_model.parameters(), lr=0.001)\\nimdb_module = SimpleModule.binary_classification(\\nimdb_model,\\noptimizer=imdb_optimizer)\\nHavingloadedthedatasetsintoadatamoduleandcreateda SimpleModule ,\\nthe remaining steps are familiar.\\nIn[75]: imdb_logger = CSVLogger( 'logs',n a m e = 'IMDB')\\nimdb_trainer = Trainer(deterministic=True,\\nmax_epochs=30,\\nlogger=imdb_logger,\\ncallbacks=[ErrorTracker()])\\nimdb_trainer.fit(imdb_module,\\ndatamodule=imdb_dm)\\nEvaluating the test error yields roughly 86% accuracy.\\nIn[76]: test_results = imdb_trainer.test(imdb_module, datamodule=imdb_dm)\\ntest_results\\n28Our use of binary_classification() instead of classification() is due to\\nsome subtlety in how torchmetrics.Accuracy() works, as well as the data type of\\nthe targets.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 462}), Document(page_content=\"10.9 Lab: Deep Learning 457\\nOut[76]: [{'test_loss' :1 . 0 8 6 3 , 'test_accuracy' :0 . 8 5 5 0 } ]\\nComparison to Lasso\\nWe now fit a lasso logistic regression model using LogisticRegression()\\nfromsklearn. Sincesklearndoes not recognize the sparse tensors of torch,\\nwe use a sparse matrix that is recognized by sklearn.\\nIn[77]: ((X_train, Y_train),\\n(X_valid, Y_valid),\\n(X_test, Y_test)) = load_sparse(validation=2000,\\nrandom_state=0,\\nroot='data/IMDB' )\\nSimilar to what we did in Section 10.9.1, we construct a series of 50\\nvalues for the lasso reguralization parameter λ.\\nIn[78]: lam_max = np.abs(X_train.T * (Y_train - Y_train.mean())).max()\\nlam_val = lam_max * np.exp(np.linspace(np.log(1),\\nnp.log(1e-4), 50))\\nWithLogisticRegression() the regularization parameter Cis specified as\\nthe inverse of λ. There are several solvers for logistic regression; here we\\nuseliblinear which works well with the sparse input format.\\nIn[79]: logit = LogisticRegression(penalty= 'l1',\\nC=1/lam_max,\\nsolver= 'liblinear' ,\\nwarm_start=True,\\nfit_intercept=True)\\nThe path of 50 values takes approximately 40 seconds to run.\\nIn[80]: coefs = []\\nintercepts = []\\nfor l in lam_val:\\nlogit.C = 1/l\\nlogit.fit(X_train, Y_train)\\ncoefs.append(logit.coef_.copy())\\nintercepts.append(logit.intercept_)\\nThe coefficient and intercepts have an extraneous dimension which can\\nbe removed by the np.squeeze() function.\\nIn[81]: coefs = np.squeeze(coefs)\\nintercepts = np.squeeze(intercepts)\\nWe’ll now make a plot to compare our neural network results with the\\nlasso.\\nIn[82]: %%capture\\nfig, axes = subplots(1, 2, figsize=(16, 8), sharey=True)\\nfor ((X_, Y_),\\ndata_,\\ncolor) in zip([(X_train, Y_train),\\n(X_valid, Y_valid),\\n(X_test, Y_test)],\", metadata={'source': './documents/ISLP_website.pdf', 'page': 463}), Document(page_content=\"458 10. Deep Learning\\n['Training' ,'Validation' ,'Test'],\\n['black' ,'red','blue']):\\nlinpred_ = X_ * coefs.T + intercepts[None,:]\\nlabel_ = np.array(linpred_ > 0)\\naccuracy_ = np.array([np.mean(Y_ == l) for l in label_.T])\\naxes[0].plot(-np.log(lam_val / X_train.shape[0]),\\naccuracy_,\\n'.--',\\ncolor=color,\\nmarkersize=13,\\nlinewidth=2,\\nlabel=data_)\\naxes[0].legend()\\naxes[0].set_xlabel( r'$-\\\\log(\\\\lambda)$' ,f o n t s i z e = 2 0 )\\naxes[0].set_ylabel( 'Accuracy' ,f o n t s i z e = 2 0 )\\nNoticethe useof %%capture ,whichsuppressesthe displayingof thepartially%%capturecompleted figure. This is useful when making a complex figure, since the\\nsteps can be spread across two or more cells. We now add a plot of the lasso\\naccuracy, and display the composed figure by simply entering its name at\\nthe end of the cell.\\nIn[83]: imdb_results = pd.read_csv(imdb_logger.experiment.metrics_file_path)\\nsummary_plot(imdb_results,\\naxes[1],\\ncol='accuracy' ,\\nylabel= 'Accuracy' )\\naxes[1].set_xticks(np.linspace(0, 30, 7).astype(int))\\naxes[1].set_ylabel( 'Accuracy' ,f o n t s i z e = 2 0 )\\naxes[1].set_xlabel( 'Epoch' ,f o n t s i z e = 2 0 )\\naxes[1].set_ylim([0.5, 1]);\\naxes[1].axhline(test_results[0][ 'test_accuracy' ],\\ncolor='blue',\\nlinestyle= '--',\\nlinewidth=3)\\nfig\\nFrom the graphs we see that the accuracy of the lasso logistic regression\\npeaks at about 0.88, as it does for the neural network.\\nOnce again, we end with a cleanup.\\nIn[84]: del(imdb_model,\\nimdb_trainer,\\nimdb_logger,\\nimdb_dm,\\nimdb_train,\\nimdb_test)\\n10.9.6 Recurrent Neural Networks\\nIn this lab we fit the models illustrated in Section 10.5.\\nSequential Models for Document Classification\\nHere we fit a simple LSTM RNN for sentiment prediction to the IMDb\\nmovie-review data, as discussed in Section 10.5.1. For an RNN we use\", metadata={'source': './documents/ISLP_website.pdf', 'page': 464}), Document(page_content=\"10.9 Lab: Deep Learning 459\\nthe sequence of words in a document, taking their order into account. We\\nloaded the preprocessed data at the beginning of Section 10.9.5. A script\\nthat details the preprocessing can be found in the ISLPlibrary. Notably,\\nsince more than 90% of the documents had fewer than 500 words, we set\\nthe document length to 500. For longer documents, we used the last 500\\nwords, and for shorter documents, we padded the front with blanks.\\nIn[85]: imdb_seq_dm = SimpleDataModule(imdb_seq_train,\\nimdb_seq_test,\\nvalidation=2000,\\nbatch_size=300,\\nnum_workers=min(6, max_num_workers)\\n)\\nThe first layer of the RNN is an embedding layer of size 32, which will\\nbe learned during training. This layer one-hot encodes each document as a\\nmatrix of dimension 500×10,003, and then maps these 10,003dimensions\\ndownto32.29Sinceeachwordisrepresentedbyaninteger,thisiseffectively\\nachieved by the creation of an embedding matrix of size 10,003×32; each\\nof the 500 integers in the document are then mapped to the appropriate\\n32 real numbers by indexing the appropriate rows of this matrix.\\nThe second layer is an LSTM with 32 units, and the output layer is a\\nsinglelogitforthebinaryclassificationtask.Inthelastlineofthe forward()\\nmethod below, we take the last 32-dimensional output of the LSTM and\\nmap it to our response.\\nIn[86]: class LSTMModel(nn.Module):\\ndef __init__(self, input_size):\\nsuper(LSTMModel, self).__init__()\\nself.embedding = nn.Embedding(input_size, 32)\\nself.lstm = nn.LSTM(input_size=32,\\nhidden_size=32,\\nbatch_first=True)\\nself.dense = nn.Linear(32, 1)\\ndef forward(self, x):\\nval, (h_n, c_n) = self.lstm(self.embedding(x))\\nreturn torch.flatten(self.dense(val[:,-1]))\\nWe instantiate and take a look at the summary of the model, using the\\nfirst 10 documents in the corpus.\\nIn[87]: lstm_model = LSTMModel(X_test.shape[-1])\\nsummary(lstm_model,\\ninput_data=imdb_seq_train.tensors[0][:10],\\ncol_names=[ 'input_size' ,\\n'output_size' ,\\n'num_params' ])\\nOut[87]: ====================================================================\\nLayer (type:depth-idx) Input Shape Output Shape Param #\\n====================================================================\\nLSTMModel [10, 500] [10] --\\n29The extra 3 dimensions correspond to commonly occurring non-word entries in the\\nreviews.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 465}), Document(page_content=\"460 10. Deep Learning\\nEmbedding: 1-1 [10, 500] [10, 500, 32] 320,096\\nLSTM: 1-2 [10, 500, 32] [10, 500, 32] 8,448\\nLinear: 1-3 [10, 32] [10, 1] 33\\n====================================================================\\nTotal params: 328,577\\nTrainable params: 328,577\\nThe 10,003 is suppressed in the summary, but we see it in the parameter\\ncount, since 10,003×32 = 320 ,096.\\nIn[88]: lstm_module = SimpleModule.binary_classification(lstm_model)\\nlstm_logger = CSVLogger( 'logs',n a m e = 'IMDB_LSTM' )\\nIn[89]: lstm_trainer = Trainer(deterministic=True,\\nmax_epochs=20,\\nlogger=lstm_logger,\\ncallbacks=[ErrorTracker()])\\nlstm_trainer.fit(lstm_module,\\ndatamodule=imdb_seq_dm)\\nThe rest is now similar to other networks we have fit. We track the test\\nperformance as the network is fit, and see that it attains 85% accuracy.\\nIn[90]: lstm_trainer.test(lstm_module, datamodule=imdb_seq_dm)\\nOut[90]: [{'test_loss' :0 . 8 1 7 8 , 'test_accuracy' :0 . 8 4 7 6 } ]\\nWe once again show the learning progress, followed by cleanup.\\nIn[91]: lstm_results = pd.read_csv(lstm_logger.experiment.metrics_file_path)\\nfig, ax = subplots(1, 1, figsize=(6, 6))\\nsummary_plot(lstm_results,\\nax,\\ncol='accuracy' ,\\nylabel= 'Accuracy' )\\nax.set_xticks(np.linspace(0, 20, 5).astype(int))\\nax.set_ylabel( 'Accuracy' )\\nax.set_ylim([0.5, 1])\\nIn[92]: del(lstm_model,\\nlstm_trainer,\\nlstm_logger,\\nimdb_seq_dm,\\nimdb_seq_train,\\nimdb_seq_test)\\nTime Series Prediction\\nWe now show how to fit the models in Section 10.5.2for time series pre-\\ndiction. We first load and standardize the data.\\nIn[93]: NYSE = load_data( 'NYSE')\\ncols = [ 'DJ_return' ,'log_volume' ,'log_volatility' ]\\nX=p d . D a t a F r a m e ( S t a n d a r d S c a l e r (\\nwith_mean=True,\\nwith_std=True).fit_transform(NYSE[cols]),\\ncolumns=NYSE[cols].columns,\\nindex=NYSE.index)\", metadata={'source': './documents/ISLP_website.pdf', 'page': 466}), Document(page_content='10.9 Lab: Deep Learning 461Next we set up the lagged versions of the data, dropping any rows withmissing values using thedropna()method.In[94]:for lag in range(1, 6):for col in cols:newcol = np.zeros(X.shape[0]) * np.nannewcol[lag:] = X[col].values[:-lag]X.insert(len(X.columns),\"{0}_{1}\".format(col, lag), newcol)X.insert(len(X.columns),\\'train\\',N Y S E [\\'train\\'])X=X . d r o p n a ( )Finally, we extract the response, training indicator, and drop the currentday’sDJ_returnandlog_volatilityto predict only from previous day’sdata.In[95]:Y, train = X[\\'log_volume\\'], X[\\'train\\']X=X . d r o p ( c o l u m n s = [\\'train\\']+c o l s )X.columnsOut[95]:Index([\\'DJ_return_1\\',\\'log_volume_1\\',\\'log_volatility_1\\',\\'DJ_return_2\\',\\'log_volume_2\\',\\'log_volatility_2\\',\\'DJ_return_3\\',\\'log_volume_3\\',\\'log_volatility_3\\',\\'DJ_return_4\\',\\'log_volume_4\\',\\'log_volatility_4\\',\\'DJ_return_5\\',\\'log_volume_5\\',\\'log_volatility_5\\'],dtype=\\'object\\')We first fit a simple linear model and compute theR2on the test datausing thescore()method.In[96]:M=L i n e a r R e g r e s s i o n ( )M.fit(X[train], Y[train])M.score(X[∼train], Y[∼train])Out[96]:0.4129We refit this model, including the factor variableday_of_week. For a cate-gorical series inpandas, we can form the indicators using theget_dummies()method.In[97]:X_day = pd.merge(X,pd.get_dummies(NYSE[\\'day_of_week\\']),on=\\'date\\')Note that we do not have to reinstantiate the linear regression model as itsfit()method accepts a design matrix and a response directly.In[98]:M.fit(X_day[train], Y[train])M.score(X_day[∼train], Y[∼train])Out[98]:0.4595This model achieves anR2of about 46%.To fit the RNN, we must reshape the data, as it will expect 5 laggedversions of each feature as indicated by theinput_shapeargument to thelayernn.RNN()below. We first ensure the columns of our data frame aresuch that a reshaped matrix will have the variables correctly lagged. Weuse thereindex()method to do this.', metadata={'source': './documents/ISLP_website.pdf', 'page': 467}), Document(page_content=\"462 10. Deep LearningFor an input shape(5,3), each row represents a lagged version of thethree variables. Thenn.RNN()layer also expects the first row of each obser-vation to be earliest in time, so we must reverse the current order. Hencewe loop overrange(5,0,-1)below, which is an example of using aslice()to index iterable objects. The general notation isstart:end:step.In[99]:ordered_cols = []for lag in range(5,0,-1):for col in cols:ordered_cols.append('{0}_{1}'.format(col, lag))X=X . r e i n d e x ( c o l u m n s = o r d e r e d _ c o l s )X.columnsOut[99]:Index(['DJ_return_5','log_volume_5','log_volatility_5','DJ_return_4','log_volume_4','log_volatility_4','DJ_return_3','log_volume_3','log_volatility_3','DJ_return_2','log_volume_2','log_volatility_2','DJ_return_1','log_volume_1','log_volatility_1'],dtype='object')We now reshape the data.In[100]:X_rnn = X.to_numpy().reshape((-1,5,3))X_rnn.shapeOut[100]:(6046, 5, 3)By specifying the first size as -1,numpy.reshape()deduces its size based onthe remaining arguments.Now we are ready to proceed with the RNN, which uses 12 hidden units,and 10% dropout. After passing through the RNN, we extract the finaltime point asval[:,-1]inforward()below. This gets passed through a10% dropout and then flattened through a linear layer.In[101]:class NYSEModel(nn.Module):def __init__(self):super(NYSEModel, self).__init__()self.rnn = nn.RNN(3,12,batch_first=True)self.dense = nn.Linear(12, 1)self.dropout = nn.Dropout(0.1)def forward(self, x):val, h_n = self.rnn(x)val = self.dense(self.dropout(val[:,-1]))return torch.flatten(val)nyse_model = NYSEModel()We fit the model in a similar fashion to previous networks. We supplythefitfunction with test data as validation data, so that when we monitorits progress and plot the history function we can see the progress on thetest data. Of course we should not use this as a basis for early stopping,since then the test performance would be biased.We form the training dataset similar to ourHittersexample.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 468}), Document(page_content=\"10.9 Lab: Deep Learning 463In[102]:datasets = []for mask in [train,∼train]:X_rnn_t = torch.tensor(X_rnn[mask].astype(np.float32))Y_t = torch.tensor(Y[mask].astype(np.float32))datasets.append(TensorDataset(X_rnn_t, Y_t))nyse_train, nyse_test = datasetsFollowing our usual pattern, we inspect the summary.In[103]:summary(nyse_model,input_data=X_rnn_t,col_names=['input_size','output_size','num_params'])Out[103]:====================================================================Layer (type:depth-idx) Input Shape Output Shape Param#====================================================================NYSEModel[1770, 5, 3] [1770]--RNN: 1-1[1770, 5, 3] [1770, 5, 12] 204Dropout: 1-2[1770, 12] [1770, 12] --Linear: 1-3[1770, 12] [1770, 1] 13====================================================================Total params: 217Trainable params: 217We again put the two datasets into a data module, with a batch size of 64.In[104]:nyse_dm = SimpleDataModule(nyse_train,nyse_test,num_workers=min(4, max_num_workers),validation=nyse_test,batch_size=64)We run some data through our model to be sure the sizes match up cor-rectly.In[105]:for idx, (x, y) in enumerate(nyse_dm.train_dataloader()):out = nyse_model(x)print(y.size(), out.size())if idx >= 2:breaktorch.Size([64]) torch.Size([64])torch.Size([64]) torch.Size([64])torch.Size([64]) torch.Size([64])We follow our previous example for setting up a trainer for a regressionproblem, requesting theR2metric to be be computed at each epoch.In[106]:nyse_optimizer = RMSprop(nyse_model.parameters(),lr=0.001)nyse_module = SimpleModule.regression(nyse_model,optimizer=nyse_optimizer,metrics={'r2':R2Score()})Fitting the model should by now be familiar. The results on the test dataare very similar to the linear AR model.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 469}), Document(page_content=\"464 10. Deep LearningIn[107]:nyse_trainer = Trainer(deterministic=True,max_epochs=200,callbacks=[ErrorTracker()])nyse_trainer.fit(nyse_module,datamodule=nyse_dm)nyse_trainer.test(nyse_module,datamodule=nyse_dm)Out[107]:[{'test_loss':0 . 6 1 4 1 ,'test_r2':0 . 4 1 7 2 } ]We could also fit a model without thenn.RNN()layer by just using ann.Flatten()layer instead. This would be a nonlinear AR model. If inaddition we excluded the hidden layer, this would be equivalent to ourearlier linear AR model.Instead we will fit a nonlinear AR model using the feature setX_daythatincludes theday_of_weekindicators. To do so, we must first create our testand training datasets and a corresponding data module. This may seem alittle burdensome, but is part of the general pipeline fortorch.In[108]:datasets = []for mask in [train,∼train]:X_day_t = torch.tensor(np.asarray(X_day[mask]).astype(np.float32))Y_t = torch.tensor(np.asarray(Y[mask]).astype(np.float32))datasets.append(TensorDataset(X_day_t, Y_t))day_train, day_test = datasetsCreating a data module follows a familiar pattern.In[109]:day_dm = SimpleDataModule(day_train,day_test,num_workers=min(4, max_num_workers),validation=day_test,batch_size=64)We build aNonLinearARModel()that takes as input the 20 features anda hidden layer with 32 units. The remaining steps are familiar.In[110]:class NonLinearARModel(nn.Module):def __init__(self):super(NonLinearARModel, self).__init__()self._forward = nn.Sequential(nn.Flatten(),nn.Linear(20, 32),nn.ReLU(),nn.Dropout(0.5),nn.Linear(32, 1))def forward(self, x):return torch.flatten(self._forward(x))In[111]:nl_model = NonLinearARModel()nl_optimizer = RMSprop(nl_model.parameters(),lr=0.001)nl_module = SimpleModule.regression(nl_model,optimizer=nl_optimizer,metrics={'r2':R2Score()})\", metadata={'source': './documents/ISLP_website.pdf', 'page': 470}), Document(page_content=\"10.10 Exercises 465\\nWe continue with the usual training steps, fit the model, and evaluate\\nthe test error. We see the test R2is a slight improvement over the linear\\nAR model that also includes day_of_week .\\nIn[112]: nl_trainer = Trainer(deterministic=True,\\nmax_epochs=20,\\ncallbacks=[ErrorTracker()])\\nnl_trainer.fit(nl_module, datamodule=day_dm)\\nnl_trainer.test(nl_module, datamodule=day_dm)\\nOut[112]: [{'test_loss' :0 . 5 6 2 5 , 'test_r2' :0 . 4 6 6 2 } ]\\n10.10 Exercises\\nConceptual\\n1. Consider a neural network with two hidden layers: p=4input units,\\n2units in the first hidden layer, 3units in the second hidden layer,\\nand a single output.\\n(a) Draw a picture of the network, similar to Figures 10.1or10.4.\\n(b) Write out an expression for f(X), assuming ReLU activation\\nfunctions. Be as explicit as you can!\\n(c) Now plug in some values for the coefficients and write out the\\nvalue off(X).\\n(d) How many parameters are there?\\n2. Consider the softmaxfunction in ( 10.13) (see also ( 4.13) on page 145)\\nfor modeling multinomial probabilities.\\n(a) In (10.13), show that if we add a constant cto each of the zℓ,\\nthen the probability is unchanged.\\n(b) In (4.13), show that if we add constants cj,j=0,1,...,p, to\\neachofthecorrespondingcoefficientsforeachoftheclasses,then\\nthe predictions at any new point xare unchanged.\\nThis shows that the softmax function is over-parametrized . However,over-\\nparametrized regularization and SGD typically constrain the solutions so that this\\nis not a problem.\\n3. Show that the negative multinomial log-likelihood ( 10.14) is equiva-\\nlent to the negative log of the likelihood expression ( 4.5) when there\\nareM=2classes.\\n4. Consider a CNN that takes in 32×32grayscale images and has a\\nsingle convolution layer with three 5×5convolution filters (without\\nboundary padding).\\n(a) Draw a sketch of the input and first hidden layer similar to\\nFigure10.8.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 471}), Document(page_content='466 10. Deep Learning\\n(b) How many parameters are in this model?\\n(c) Explain how this model can be thought of as an ordinary feed-\\nforward neural network with the individual pixels as inputs, and\\nwith constraints on the weights in the hidden units. What are\\nthe constraints?\\n(d) If there were no constraints, then how many weights would there\\nbe in the ordinary feed-forward neural network in (c)?\\n5. In Table 10.2on page 426, we see that the ordering of the three\\nmethods with respect to mean absolute error is different from the\\nordering with respect to test set R2. How can this be?\\nApplied\\n6. Consider the simple function R(β) = sin( β)+β/10.\\n(a) Draw a graph of this function over the range β∈[−6,6].\\n(b) What is the derivative of this function?\\n(c) Given β0=2.3, run gradient descent to find a local minimum\\nofR(β)using a learning rate of ρ=0.1. Show each of β0,β1,...\\nin your plot, as well as the final answer.\\n(d) Repeat with β0=1.4.\\n7. Fit a neural network to the Defaultdata. Use a single hidden layer\\nwith10units,anddropoutregularization.HavealookatLabs 10.9.1–\\n10.9.2for guidance. Compare the classification performance of your\\nmodel with that of linear logistic regression.\\n8. From your collection of personal photographs, pick 10 images of an-\\nimals (such as dogs, cats, birds, farm animals, etc.). If the subject\\ndoes not occupy a reasonable part of the image, then crop the image.\\nNow use a pretrained image classification CNN as in Lab 10.9.4to\\npredict the class of each of your images, and report the probabilities\\nfor the top five predicted classes for each image.\\n9. Fit a lag-5 autoregressive model to the NYSEdata, as described in\\nthe text and Lab 10.9.6. Refit the model with a 12-level factor repre-\\nsenting the month. Does this factor improve the performance of the\\nmodel?\\n10. In Section 10.9.6, we showed how to fit a linear AR model to the\\nNYSEdata using the LinearRegression() function. However, we also\\nmentioned that we can “flatten” the short sequences produced for\\nthe RNN model in order to fit a linear AR model. Use this latter\\napproach to fit a linear AR model to the NYSEdata. Compare the test\\nR2of this linear AR model to that of the linear AR model that we fit\\nin the lab. What are the advantages/disadvantages of each approach?\\n11. Repeat the previous exercise, but now fit a nonlinear AR model by\\n“flattening” the short sequences produced for the RNN model.', metadata={'source': './documents/ISLP_website.pdf', 'page': 472}), Document(page_content='10.10 Exercises 467\\n12. Consider the RNN fit to the NYSEdata in Section 10.9.6. Modify the\\ncode to allow inclusion of the variable day_of_week , and fit the RNN.\\nCompute the test R2.\\n13. Repeat the analysis of Lab 10.9.5on theIMDbdata using a similarly\\nstructured neural network. We used 16 hidden units at each of two\\nhidden layers. Explore the effect of increasing this to 32 and 64 units\\nper layer, with and without 30% dropout regularization.', metadata={'source': './documents/ISLP_website.pdf', 'page': 473}), Document(page_content='11\\nSurvival Analysis and Censored Data\\nIn this chapter, we will consider the topics of survival analysis andcensoredsurvival\\nanalysisdata. These arise in the analysis of a unique kind of outcome variable: the\\ncensored\\ndatatime until an event occurs .\\nFor example, suppose that we have conducted a five-year medical study,\\nin which patients have been treated for cancer. We would like to fit a model\\nto predict patient survival time, using features such as baseline health mea-\\nsurements or type of treatment. At first pass, this may sound like a regres-\\nsion problem of the kind discussed in Chapter 3. But there is an important\\ncomplication: hopefully some or many of the patients have survived until\\ntheendofthestudy.Suchapatient’ssurvivaltimeissaidtobe censored:w e\\nknow that it is at least five years, but we do not know its true value. We do\\nnot want to discard this subset of surviving patients, as the fact that they\\nsurvived at least five years amounts to valuable information. However, it is\\nnot clear how to make use of this information using the techniques covered\\nthus far in this textbook.\\nThough the phrase “survival analysis” evokes a medical study, the ap-\\nplications of survival analysis extend far beyond medicine. For example,\\nconsider a company that wishes to model churn, the process by which cus-\\ntomers cancel subscription to a service. The company might collect data on\\ncustomers over some time period, in order to model each customer’s time\\nto cancellation as a function of demographics or other predictors. However,\\npresumably not all customers will have canceled their subscription by the\\nend of this time period; for such customers, the time to cancellation is\\ncensored.\\nIn fact, survival analysis is relevant even in application areas that are\\nunrelatedtotime.Forinstance,supposewewishtomodelaperson’sweight\\nas a function of some covariates, using a dataset with measurements for a\\nlarge number of people. Unfortunately, the scale used to weigh those people\\nis unable to report weights above a certain number. Then, any weights that\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning , Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_11 469', metadata={'source': './documents/ISLP_website.pdf', 'page': 474}), Document(page_content='470 11. Survival Analysis and Censored Data\\nexceed that number are censored. The survival analysis methods presented\\nin this chapter could be used to analyze this dataset.\\nSurvival analysis is a very well-studied topic within statistics, due to its\\ncritical importance in a variety of applications, both in and out of medicine.\\nHowever, it has received relatively little attention in the machine learning\\ncommunity.\\n11.1 Survival and Censoring Times\\nFor each individual, we suppose that there is a true survival time ,T, as wellsurvival timeas a true censoring time ,C. (The survival time is also known as the failurecensoring\\ntimetimeor theevent time .) The survival time represents the time at which the\\nfailure time\\nevent timeevent of interest occurs: for instance, the time at which the patient dies,\\nor the customer cancels his or her subscription. By contrast, the censoring\\ntime is the time at which censoring occurs: for example, the time at which\\nthe patient drops out of the study or the study ends.\\nWe observe either the survival time Tor else the censoring time C.\\nSpecifically, we observe the random variable\\nY=min(T,C). (11.1)\\nIn other words, if the event occurs before censoring (i.e. T<C ) then we\\nobserve the true survival time T; however, if censoring occurs before the\\nevent (T>C ) then we observe the censoring time. We also observe a status\\nindicator,\\nδ={\\n1ifT≤C\\n0ifT>C .\\nThus,δ=1 if we observe the true survival time, and δ=0 if we instead\\nobserve the censoring time.\\nNow, suppose we observe n(Y,δ)pairs, which we denote as (y1,δ1),...,\\n(yn,δn). Figure 11.1displays an example from a (fictitious) medical study\\nin which we observe n=4 patients for a 365-day follow-up period. For\\npatients 1 and 3, we observe the time to event (such as death or disease\\nrelapse) T=ti. Patient 2 was alive when the study ended, and patient 4\\ndropped out of the study, or was “lost to follow-up”; for these patients we\\nobserveC=ci. Therefore, y1=t1,y3=t3,y2=c2,y4=c4,δ1=δ3=1,\\nandδ2=δ4=0.\\n11.2 A Closer Look at Censoring\\nIn order to analyze survival data, we need to make some assumptions about\\nwhycensoringhasoccurred.Forinstance,supposethatanumberofpatients\\ndropoutofacancerstudyearlybecausetheyareverysick.Ananalysisthat\\ndoes not take into consideration the reason why the patients dropped out\\nwill likely overestimate the true average survival time. Similarly, suppose\\nthat males who are very sick are more likely to drop out of the study than', metadata={'source': './documents/ISLP_website.pdf', 'page': 475}), Document(page_content='11.2 A Closer Look at Censoring 471\\n01002003001234Time in DaysPatient\\nFIGURE 11.1. Illustration of censored survival data. For patients 1 and 3, the\\nevent was observed. Patient 2 was alive when the study ended. Patient 4 dropped\\nout of the study.\\nfemales who are very sick. Then a comparison of male and female survival\\ntimes may wrongly suggest that males survive longer than females.\\nIn general, we need to assume that the censoring mechanism is indepen-\\ndent: conditional on the features, the event time Tis independent of the\\ncensoring time C. The two examples above violate the assumption of inde-\\npendent censoring. Typically, it is not possible to determine from the data\\nitself whether the censoring mechanism is independent. Instead, one has to\\ncarefully consider the data collection process in order to determine whether\\nindependent censoring is a reasonable assumption. In the remainder of this\\nchapter, we will assume that the censoring mechanism is independent.1\\nIn this chapter, we focus on right censoring , which occurs when T≥Y,\\ni.e. the true event time Tis at least as large as the observed time Y.\\n(Notice that T≥Yis a consequence of ( 11.1). Right censoring derives its\\nname from the fact that time is typically displayed from left to right, as in\\nFigure11.1.) However, other types of censoring are possible. For instance,\\ninleft censoring , the true event time Tis less than or equal to the observed\\ntimeY. For example, in a study of pregnancy duration, suppose that we\\nsurvey patients 250 days after conception, when some have already had\\ntheir babies. Then we know that for those patients, pregnancy duration is\\nless than 250 days. More generally, interval censoring refers to the setting\\nin which we do not know the exact event time, but we know that it falls\\nin some interval. For instance, this setting arises if we survey patients once\\nper week in order to determine whether the event has occurred. While left\\ncensoring and interval censoring can be accommodated using variants of\\nthe ideas presented in this chapter, in what follows we focus specifically on\\nright censoring.\\n1The assumption of independent censoring can be relaxed somewhat using the notion\\nofnon-informative censoring ; however, the definition of non-informative censoring is too\\ntechnical for this book.', metadata={'source': './documents/ISLP_website.pdf', 'page': 476}), Document(page_content='472 11. Survival Analysis and Censored Data\\n11.3 The Kaplan–Meier Survival Curve\\nThesurvival curve , orsurvival function , is defined assurvival\\ncurve\\nsurvival\\nfunctionS(t) = Pr( T>t ). (11.2)\\nThis decreasing function quantifies the probability of surviving past time\\nt. For example, suppose that a company is interested in modeling customer\\nchurn. Let Trepresent the time that a customer cancels a subscription to\\nthecompany’sservice.Then S(t)representstheprobabilitythatacustomer\\ncancels later than time t. The larger the value of S(t), the less likely that\\nthe customer will cancel before time t.\\nIn this section, we will consider the task of estimating the survival\\ncurve. Our investigation is motivated by the BrainCancer dataset, which\\ncontains the survival times for patients with primary brain tumors un-\\ndergoing treatment with stereotactic radiation methods.2The predictors\\naregtv(gross tumor volume, in cubic centimeters); sex(male or female);\\ndiagnosis (meningioma, LG glioma, HG glioma, or other); loc(the tumor\\nlocation: either infratentorial or supratentorial); ki(Karnofsky index); and\\nstereo(stereotactic method: either stereotactic radiosurgery or fraction-\\nated stereotactic radiotherapy, abbreviated as SRS and SRT, respectively).\\nOnly 53 of the 88 patients were still alive at the end of the study.\\nNow, we consider the task of estimating the survival curve ( 11.2) for\\nthese data. To estimate S(20) = Pr( T>20), the probability that a patient\\nsurvives for at least t= 20 months, it is tempting to simply compute the\\nproportion of patients who are known to have survived past 20 months, i.e.\\nthe proportion of patients for whom Y>20. This turns out to be 48/88,\\nor approximately 55%. However, this does not seem quite right, since Y\\nandTrepresent different quantities. In particular, 17 of the 40 patients\\nwho did not survive to 20months were actually censored, and this analysis\\nimplicitly assumes that T<20for all of those censored patients; of course,\\nwe do not know whether that is true.\\nAlternatively, to estimate S(20), we could consider computing the pro-\\nportion of patients for whom Y>20, out of the 71patients who were not\\ncensored by time t= 20 ; this comes out to 48/71, or approximately 68%.\\nHowever, this is not quite right either, since it amounts to completely ig-\\nnoring the patients who were censored before time t= 20 , even though the\\ntimeat which they are censored is potentially informative. For instance, a\\npatient who was censored at time t= 19 .9likely would have survived past\\nt= 20 had he or she not been censored.\\nWe have seen that estimating S(t)is complicated by the presence of\\ncensoring. We now present an approach to overcome these challenges. We\\nletd1<d 2<···<d Kdenote the Kunique death times among the non-\\ncensored patients, and we let qkdenote the number of patients who died\\nat timedk. Fork=1,...,K , we letrkdenote the number of patients alive\\n2This dataset is described in the following paper: Selingerová et al. (2016) Survival\\nof patients with primary brain tumors: Comparison of two statistical approaches. PLoS\\nOne, 11(2):e0148733.', metadata={'source': './documents/ISLP_website.pdf', 'page': 477}), Document(page_content='11.3 The Kaplan–Meier Survival Curve 473\\nand in the study just before dk; these are the at riskpatients. The set of\\npatients that are at risk at a given time are referred to as the risk set.risk setBy the law of total probability,3\\nPr(T>d k) = Pr( T>d k|T>d k−1) Pr( T>d k−1)\\n+ Pr( T>d k|T≤dk−1) Pr( T≤dk−1).\\nThe fact that dk−1<d kimplies that Pr(T>d k|T≤dk−1)=0 (it is\\nimpossible for a patient to survive past time dkif he or she did not survive\\nuntil an earlier time dk−1). Therefore,\\nS(dk) = Pr( T>d k) = Pr( T>d k|T>d k−1) Pr( T>d k−1).\\nPlugging in ( 11.2) again, we see that\\nS(dk) = Pr( T>d k|T>d k−1)S(dk−1).\\nThis implies that\\nS(dk) = Pr( T>d k|T>d k−1)×···×Pr(T>d 2|T>d 1) Pr( T>d 1).\\nWe now must simply plug in estimates of each of the terms on the right-\\nhand side of the previous equation. It is natural to use the estimator\\nˆPr(T>d j|T>d j−1)=( rj−qj)/rj,\\nwhich is the fraction of the risk set at time djwho survived past time dj.\\nThis leads to the Kaplan–Meier estimator of the survival curve:Kaplan–\\nMeier\\nestimatorˆS(dk)=k∏\\nj=1(rj−qj\\nrj)\\n. (11.3)\\nFor times tbetween dkanddk+1, we set ˆS(t)=ˆS(dk). Consequently, the\\nKaplan–Meier survival curve has a step-like shape.\\nThe Kaplan–Meier survival curve for the BrainCancer data is displayed\\nin Figure 11.2. Each point in the solid step-like curve shows the estimated\\nprobability of surviving past the time indicated on the horizontal axis. The\\nestimated probability of survival past 20 months is 71%, which is quite a\\nbit higher than the naive estimates of 55% and 68% presented earlier.\\nThe sequential construction of the Kaplan–Meier estimator — starting\\nat time zero and mapping out the observed events as they unfold in time —\\nis fundamental to many of the key techniques in survival analysis. These\\ninclude the log-rank test of Section 11.4, and Cox’s proportional hazard\\nmodel of Section 11.5.2.\\n3The law of total probability states that for any two events AandB,Pr(A)=\\nPr(A|B) Pr( B) + Pr( A|Bc) Pr( Bc), where Bcis the complement of the event B, i.e. it\\nis the event that Bdoes not hold.', metadata={'source': './documents/ISLP_website.pdf', 'page': 478}), Document(page_content='474 11. Survival Analysis and Censored Data\\n0204060800.0 0.2 0.4 0.6 0.8 1.0MonthsEstimated Probability of SurvivalFIGURE 11.2. For theBrainCancer data, we display the Kaplan–Meier survival\\ncurve (solid curve), along with standard error bands (dashed curves).\\n0204060800.0 0.2 0.4 0.6 0.8 1.0MonthsEstimated Probability of SurvivalFemaleMaleFIGURE 11.3. For the BrainCancer data, Kaplan–Meier survival curves for\\nmales and females are displayed.\\n11.4 The Log-Rank Test\\nWe now continue our analysis of the BrainCancer data introduced in Sec-\\ntion11.3. We wish to compare the survival of males to that of females.\\nFigure11.3shows the Kaplan–Meier survival curves for the two groups.\\nFemales seem to fare a little better up to about 50 months, but then the\\ntwo curves both level off to about 50%. How can we carry out a formal test\\nof equality of the two survival curves?\\nAt first glance, a two-sample t-test seems like an obvious choice: we could\\ntest whether the mean survival time among the females equals the mean\\nsurvival time among the males. But the presence of censoring again creates\\nacomplication.Toovercomethischallenge,wewillconducta log-rank test ,4\\nlog-rank test', metadata={'source': './documents/ISLP_website.pdf', 'page': 479}), Document(page_content='11.4 The Log-Rank Test 475\\nGroup 1 Group 2 Total\\nDied q1k q2k qk\\nSurvived r1k−q1kr2k−q2krk−qk\\nTotal r1k r2k rk\\nTABLE 11.1. Among the set of patients at risk at time dk, the number of patients\\nwho died and survived in each of two groups is reported.\\nwhich examines how the events in each group unfold sequentially in time.\\nRecall from Section 11.3thatd1<d 2<···<d Kare the unique death\\ntimes among the non-censored patients, rkis the number of patients at\\nrisk at time dk, andqkis the number of patients who died at time dk.W e\\nfurther define r1kandr2kto be the number of patients in groups 1 and 2,\\nrespectively, who are at risk at time dk. Similarly, we define q1kandq2kto\\nbe the number of patients in groups 1 and 2, respectively, who died at time\\ndk. Note that r1k+r2k=rkandq1k+q2k=qk.\\nAt each death time dk, we construct a 2×2table of counts of the form\\nshown in Table 11.1. Note that if the death times are unique (i.e. no two\\nindividuals die at the same time), then one of q1kandq2kequals one, and\\nthe other equals zero.\\nThe main idea behind the log-rank test statistic is as follows. In order\\nto testH0: E(X)= µfor some random variable X, one approach is to\\nconstruct a test statistic of the form\\nW=X−µ√\\nVar(X). (11.4)\\nTo construct the log-rank test statistic, we compute a quantity that takes\\nexactly the form ( 11.4), withX=∑K\\nk=1q1k, whereq1kis given in the top\\nleft of Table 11.1.\\nIn greater detail, if there is no difference in survival between the two\\ngroups, and conditioning on the row and column totals in Table 11.1, the\\nexpected value of q1kis\\nµk=r1k\\nrkqk. (11.5)\\nSo the expected value of X=∑K\\nk=1q1kisµ=∑K\\nk=1r1k\\nrkqk. Furthermore,\\nit can be shown5that the variance of q1kis\\nVar ( q1k)=qk(r1k/rk)(1−r1k/rk)(rk−qk)\\nrk−1. (11.6)\\nThoughq11,...,q 1Kmay be correlated, we nonetheless estimate\\nVar(K∑\\nk=1q1k)\\n≈K∑\\nk=1Var ( q1k)=K∑\\nk=1qk(r1k/rk)(1−r1k/rk)(rk−qk)\\nrk−1.\\n(11.7)\\n4The log-rank test is also known as the Mantel–Haenszel test orCochran–Mantel–\\nHaenszel test .\\n5For details, see Exercise 7at the end of this chapter.', metadata={'source': './documents/ISLP_website.pdf', 'page': 480}), Document(page_content='476 11. Survival Analysis and Censored Data\\nTherefore, to compute the log-rank test statistic, we simply proceed as\\nin (11.4), withX=∑K\\nk=1q1k, making use of ( 11.5) and (11.7). That is, we\\ncalculate\\nW=∑K\\nk=1(q1k−µk)√∑K\\nk=1Var ( q1k)=∑K\\nk=1(\\nq1k−qk\\nrkr1k)\\n√∑K\\nk=1qk(r1k/rk)(1−r1k/rk)(rk−qk)\\nrk−1.(11.8)\\nWhen the sample size is large, the log-rank test statistic Whas ap-\\nproximately a standard normal distribution; this can be used to compute\\nap-value for the null hypothesis that there is no difference between the\\nsurvival curves in the two groups.6\\nComparing the survival times of females and males on the BrainCancer\\ndata gives a log-rank test statistic of W=1.2, which corresponds to a two-\\nsidedp-value of 0.2using the theoretical null distribution, and a p-value\\nof0.25using the permutation null distribution with 1,000 permutations.\\nThus,wecannotrejectthenullhypothesisofnodifferenceinsurvivalcurves\\nbetween females and males.\\nThe log-rank test is closely related to Cox’s proportional hazards model,\\nwhich we discuss in Section 11.5.2.\\n11.5 Regression Models With a Survival Response\\nWe now consider the task of fitting a regression model to survival data.\\nAs in Section 11.1, the observations are of the form (Y,δ), where Y=\\nmin( T,C)is the (possibly censored) survival time, and δis an indicator\\nvariable that equals 1ifT≤C. Furthermore, X∈Rpis a vector of p\\nfeatures. We wish to predict the true survival time T.\\nSince the observed quantity Yis positive and may have a long right\\ntail, we might be tempted to fit a linear regression of log(Y)onX. But\\nas the reader will surely guess, censoring again creates a problem since\\nwe are actually interested in predicting Tand not Y. To overcome this\\ndifficulty, we instead make use of a sequential construction, similar to the\\nconstructions of the Kaplan–Meier survival curve in Section 11.3and the\\nlog-rank test in Section 11.4.\\n11.5.1 The Hazard Function\\nThehazard function orhazard rate — also known as the force of mortalityhazard\\nfunction— is formally defined as\\nh(t) = lim\\n∆t→0Pr(t<T ≤t+∆t|T>t )\\n∆t, (11.9)\\n6Alternatively, we can estimate the p-value via permutations, using ideas that will\\nbe presented in Section 13.5. The permutation distribution is obtained by randomly\\nswapping the labels for the observations in the two groups.', metadata={'source': './documents/ISLP_website.pdf', 'page': 481}), Document(page_content='11.5 Regression Models With a Survival Response 477\\nwhereTis the (unobserved) survival time. It is the death rate in the instant\\nafter time t, given survival past that time.7In (11.9), we take the limit as\\n∆tapproaches zero, so we can think of ∆tas being an extremely tiny\\nnumber. Thus, more informally, ( 11.9) implies that\\nh(t)≈Pr(t<T ≤t+∆t|T>t )\\n∆t\\nfor some arbitrarily small ∆t.\\nWhy should we care about the hazard function? First of all, it is closely\\nrelated to the survival curve ( 11.2), as we will see next. Second, it turns out\\nthat a key approach for modeling survival data as a function of covariates\\nrelies heavily on the hazard function; we will introduce this approach —\\nCox’s proportional hazards model — in Section 11.5.2.\\nWe now consider the hazard function h(t)in a bit more detail. Recall\\nthat for two events AandB, the probability of AgivenBcan be expressed\\nasPr(A|B) = Pr( A∩B)/Pr(B), i.e. the probability that AandBboth\\noccur divided by the probability that Boccurs. Furthermore, recall from\\n(11.2) thatS(t) = Pr( T>t ). Thus,\\nh(t) = lim\\n∆t→0Pr (( t<T ≤t+∆t)∩(T>t ))/∆t\\nPr(T>t )\\n= lim\\n∆t→0Pr(t<T ≤t+∆t)/∆t\\nPr(T>t )\\n=f(t)\\nS(t), (11.10)\\nwhere\\nf(t) = lim\\n∆t→0Pr(t<T ≤t+∆t)\\n∆t(11.11)\\nis theprobability density function associated with T, i.e. it is the instanta-probability\\ndensity\\nfunctionneous rate of death at time t. The second equality in ( 11.10) made use of\\nthe fact that if t<T ≤t+∆t, then it must be the case that T>t .\\nEquation 11.10implies a relationship between the hazard function h(t),\\nthe survival function S(t), and the probability density function f(t). In\\nfact, these are three equivalent ways8of describing the distribution of T.\\nThe likelihood associated with the ith observation is\\nLi={\\nf(yi)if theith observation is not censored\\nS(yi)if theith observation is censored\\n=f(yi)δiS(yi)1−δi. (11.12)\\nThe intuition behind ( 11.12) is as follows: if Y=yiand theith observation\\nis not censored, then the likelihood is the probability of dying in a tiny in-\\nterval around time yi. If theith observation is censored, then the likelihood\\n7Due to the ∆tin the denominator of ( 11.9), the hazard function is a rate of death,\\nrather than a probability of death. However, higher values of h(t)directly correspond\\nto a higher probability of death, just as higher values of a probability density function\\ncorrespond to more likely outcomes for a random variable. In fact, h(t)is the probability\\ndensity function for Tconditional on T>t .\\n8See Exercise 8.', metadata={'source': './documents/ISLP_website.pdf', 'page': 482}), Document(page_content='478 11. Survival Analysis and Censored Data\\nis the probability of surviving at least until time yi. Assuming that the n\\nobservations are independent, the likelihood for the data takes the form\\nL=n∏\\ni=1f(yi)δiS(yi)1−δi=n∏\\ni=1h(yi)δiS(yi), (11.13)\\nwhere the second equality follows from ( 11.10).\\nWenowconsiderthetaskofmodelingthesurvivaltimes.Ifweassumeex-\\nponential survival, i.e. that the probability density function of the survival\\ntimeTtakes the form f(t)=λexp(−λt), then estimating the parameter λ\\nby maximizing the likelihood in ( 11.13) is straightforward.9Alternatively,\\nwe could assume that the survival times are drawn from a more flexible\\nfamily of distributions, such as the Gamma or Weibull family. Another\\npossibility is to model the survival times non-parametrically, as was done\\nin Section 11.3using the Kaplan–Meier estimator.\\nHowever, what we would really like to do is model the survival time as\\na function of the covariates . To do this, it is convenient to work directly\\nwith the hazard function, instead of the probability density function.10\\nOne possible approach is to assume a functional form for the hazard func-\\ntionh(t|xi), such as h(t|xi) = exp(\\nβ0+∑p\\nj=1βjxij)\\n, where the exponent\\nfunction guarantees that the hazard function is non-negative. Note that\\nthe exponential hazard function is special, in that it does not vary with\\ntime.11Givenh(t|xi), we could calculate S(t|xi). Plugging these equations\\ninto (11.13), we could then maximize the likelihood in order to estimate the\\nparameter β=(β0,β1,...,βp)T. However, this approach is quite restric-\\ntive, in the sense that it requires us to make a very stringent assumption\\non the form of the hazard function h(t|xi). In the next section, we will\\nconsider a much more flexible approach.\\n11.5.2 Proportional Hazards\\nThe Proportional Hazards Assumption\\nTheproportional hazards assumption states thatproportional\\nhazards\\nassumptionh(t|xi)=h0(t) exp\\uf8eb\\n\\uf8edp∑\\nj=1xijβj\\uf8f6\\n\\uf8f8, (11.14)\\nwhereh0(t)≥0is an unspecified function, known as the baseline hazard .baseline\\nhazardIt is the hazard function for an individual with features xi1=···=xip=\\n0. The name “proportional hazards” arises from the fact that the hazard\\nfunction for an individual with feature vector xiis some unknown function\\n9See Exercise 9.\\n10Given the close relationship between the hazard function h(t)and the density func-\\ntionf(t)explored in Exercise 8, posing an assumption about the form of the hazard\\nfunction is closely related to posing an assumption about the form of the density func-\\ntion, as was done in the previous paragraph.\\n11The notation h(t|xi)indicates that we are now considering the hazard function for\\ntheith observation conditional on the values of the covariates, xi.', metadata={'source': './documents/ISLP_website.pdf', 'page': 483}), Document(page_content='11.5 Regression Models With a Survival Response 479\\n0.5 1.0 1.5 2.0−3 −2 −1 0TimeLog Hazard0.5 1.0 1.5 2.00.0 0.2 0.4 0.6 0.8 1.0TimeSurvival Probability\\n0.5 1.0 1.5 2.0−4 −2 0 1 2TimeLog Hazard0.5 1.0 1.5 2.00.0 0.2 0.4 0.6 0.8 1.0TimeSurvival ProbabilityFIGURE 11.4. Top:In a simple example with p=1 and a binary covariate\\nxi∈{0,1}, the log hazard and the survival function under the model (11.14)\\nare shown (green for xi=0 and black for xi=1). Because of the proportional\\nhazards assumption (11.14), the log hazard functions differ by a constant, and the\\nsurvival functions do not cross. Bottom: Again we have a single binary covariate\\nxi∈{0,1}. However, the proportional hazards assumption (11.14)does not hold.\\nThe log hazard functions cross, as do the survival functions.\\nh0(t)times the factor exp(∑p\\nj=1xijβj)\\n. The quantity exp(∑p\\nj=1xijβj)\\nis called the relative risk for the feature vector xi=(xi1,...,x ip)T, relative\\nto that for the feature vector xi= (0,...,0)T.\\nWhat does it mean that the baseline hazard function h0(t)in (11.14) is\\nunspecified? Basically, we make no assumptions about its functional form.\\nWe allow the instantaneous probability of death at time t, given that one\\nhas survived at least until time t, to take any form. This means that the\\nhazard function is very flexible and can model a wide range of relationships\\nbetween the covariates and survival time. Our only assumption is that a\\none-unit increase in xijcorresponds to an increase in h(t|xi)by a factor of\\nexp(βj).\\nAn illustration of the proportional hazards assumption ( 11.14) is given in\\nFigure11.4, in a simple setting with a single binary covariate xi∈{0,1}(so\\nthatp=1). In the top row, the proportional hazards assumption ( 11.14)\\nholds. Thus, the hazard functions of the two groups are a constant multiple\\nof each other, so that on the log scale, the gap between them is constant.\\nFurthermore, the survival curves never cross, and in fact the gap between\\nthe survival curves tends to (initially) increase over time. By contrast, in\\nthe bottom row, ( 11.14) does not hold. We see that the log hazard functions\\nfor the two groups cross, as do the survival curves.', metadata={'source': './documents/ISLP_website.pdf', 'page': 484}), Document(page_content='480 11. Survival Analysis and Censored Data\\nCox’s Proportional Hazards Model\\nBecause the form of h0(t)in the proportional hazards assumption ( 11.14)\\nis unknown, we cannot simply plug h(t|xi)into the likelihood ( 11.13) and\\nthen estimate β=(β1,...,βp)Tby maximum likelihood. The magic of\\nCox’s proportional hazards model lies in the fact that it is in fact possibleCox’s\\nproportional\\nhazards\\nmodelto estimate βwithout having to specify the form of h0(t).\\nTo accomplish this, we make use of the same “sequential in time” logic\\nthat we used to derive the Kaplan–Meier survival curve and the log-rank\\ntest. For simplicity, assume that there are no ties among the failure, or\\ndeath, times: i.e. each failure occurs at a distinct time. Assume that δi=\\n1, i.e. the ith observation is uncensored, and thus yiis its failure time.\\nThen the hazard function for the ith observation at time yiish(yi|xi)=\\nh0(yi) exp(∑p\\nj=1xijβj)\\n, and the total hazard at time yifor the at risk\\nobservations12is\\n∑\\ni′:yi′≥yih0(yi) exp\\uf8eb\\n\\uf8edp∑\\nj=1xi′jβj\\uf8f6\\n\\uf8f8.\\nTherefore, the probability that the ith observation is the one to fail at time\\nyi(as opposed to one of the other observations in the risk set) is\\nh0(yi) exp(∑p\\nj=1xijβj)\\n∑\\ni′:yi′≥yih0(yi) exp(∑p\\nj=1xi′jβj)=exp(∑p\\nj=1xijβj)\\n∑\\ni′:yi′≥yiexp(∑p\\nj=1xi′jβj).\\n(11.15)\\nNotice that the unspecified baseline hazard function h0(yi)cancels out of\\nthe numerator and denominator!\\nThepartial likelihood is simply the product of these probabilities over allpartial\\nlikelihoodof the uncensored observations,\\nPL(β)=∏\\ni:δi=1exp(∑p\\nj=1xijβj)\\n∑\\ni′:yi′≥yiexp(∑p\\nj=1xi′jβj). (11.16)\\nCritically, the partial likelihood is valid regardless of the true value of h0(t),\\nmaking the model very flexible and robust.13\\nTo estimate β, we simply maximize the partial likelihood ( 11.16) with\\nrespect to β. As was the case for logistic regression in Chapter 4, no closed-\\nform solution is available, and so iterative algorithms are required.\\nIn addition to estimating β, we can also obtain other model outputs that\\nwe saw in the context of least squares regression in Chapter 3and logistic\\nregression in Chapter 4. For example, we can obtain p-values corresponding\\n12Recall that the “at risk” observations at time yiare those that are still at risk of\\nfailure, i.e. those that have not yet failed or been censored before time yi.\\n13In general, the partial likelihood is used in settings where it is difficult to compute\\nthe full likelihood for all of the parameters. Instead, we compute a likelihood for just the\\nparameters of primary interest: in this case, β1,...,βp. It can be shown that maximizing\\n(11.16) provides good estimates for these parameters.', metadata={'source': './documents/ISLP_website.pdf', 'page': 485}), Document(page_content='11.5 Regression Models With a Survival Response 481\\nto particular null hypotheses (e.g. H0:βj=0), as well as confidence\\nintervals associated with the coefficients.\\nConnection With The Log-Rank Test\\nSuppose we have just a single predictor ( p=1), which we assume to be\\nbinary, i.e. xi∈{0,1}. In order to determine whether there is a difference\\nbetween the survival times of the observations in the group {i:xi=0}\\nand those in the group {i:xi=1}, we can consider taking two possible\\napproaches:\\nApproach #1: Fit a Cox proportional hazards model, and test the\\nnull hypothesis H0:β=0. (Sincep=1,βis a scalar.)\\nApproach #2: Perform a log-rank test to compare the two groups, as\\nin Section 11.4.\\nWhich one should we prefer?\\nIn fact, there is a close relationship between these two approaches. In\\nparticular, when taking Approach #1, there are a number of possible ways\\nto testH0. One way is known as a score test. It turns out that in the case of\\na single binary covariate, the score test for H0:β=0in Cox’s proportional\\nhazards model is exactly equal to the log-rank test. In other words, it does\\nnot matter whether we take Approach #1 or Approach #2!\\nAdditional Details\\nThe discussion of Cox’s proportional hazards model glossed over a few\\nsubtleties:\\n• There is no intercept in ( 11.14) nor in the equations that follow,\\nbecause an intercept can be absorbed into the baseline hazard h0(t).\\n• We have assumed that there are no tied failure times. In the case\\nof ties, the exact form of the partial likelihood ( 11.16) is a bit more\\ncomplicated, and a number of computational approximations must\\nbe used.\\n•(11.16) is known as the partiallikelihood because it is not exactly a\\nlikelihood. That is, it does not correspond exactly to the probability\\nof the data under the assumption ( 11.14). However, it is a very good\\napproximation.\\n• Wehavefocusedonlyonestimationofthecoefficients β=(β1,...,βp)T.\\nHowever, at times we may also wish to estimate the baseline hazard\\nh0(t),forinstancesothatwecanestimatethesurvivalcurve S(t|x)for\\nan individual with feature vector x. The details are beyond the scope\\nof this book. Estimation of h0(t)is implemented in the lifelines\\npackage in Python, which we will see in Section 11.8.', metadata={'source': './documents/ISLP_website.pdf', 'page': 486}), Document(page_content='482 11. Survival Analysis and Censored Data\\n11.5.3 Example: Brain Cancer Data\\nTable11.2shows the result of fitting the proportional hazards model to\\ntheBrainCancer data, which was originally described in Section 11.3. The\\ncoefficient column displays ˆβj. The results indicate, for instance, that the\\nestimated hazard for a male patient is e0.18=1.2times greater than for\\na female patient: in other words, with all other features held fixed, males\\nhave a 1.2 times greater chance of dying than females, at any point in time.\\nHowever, the p-value is 0.61, which indicates that this difference between\\nmales and females is not significant.\\nAs another example, we also see that each one-unit increase in the\\nKarnofsky index corresponds to a multiplier of exp(−0.05) = 0 .95in the\\ninstantaneous chance of dying. In other words, the higher the Karnofsky\\nindex, the lower the chance of dying at any given point in time. This effect\\nis highly significant, with a p-value of 0.0027 .\\nCoefficient Std. error z-statistic p-value\\nsex[Male] 0.18 0.36 0.51 0.61\\ndiagnosis[LG Glioma] 0.92 0.64 1.43 0.15\\ndiagnosis[HG Glioma] 2.15 0.45 4.78 0.00\\ndiagnosis[Other] 0.89 0.66 1.35 0.18\\nloc[Supratentorial] 0.44 0.70 0.63 0.53\\nki -0.05 0.02 -3.00 <0.01\\ngtv 0.03 0.02 1.54 0.12\\nstereo[SRT] 0.18 0.60 0.30 0.77\\nTABLE 11.2. Results for Cox’s proportional hazards model fit to the\\nBrainCancer data, which was first described in Section 11.3. The variable\\ndiagnosis is qualitative with four levels: meningioma, LG glioma, HG glioma,\\nor other. The variables sex,loc, andstereoare binary.\\n11.5.4 Example: Publication Data\\nNext, we consider the dataset Publication involving the time to publica-\\ntion of journal papers reporting the results of clinical trials funded by the\\nNational Heart, Lung, and Blood Institute.14For 244 trials, the time in\\nmonths until publication is recorded. Of the 244 trials, only 156 were pub-\\nlished during the study period; the remaining studies were censored. The\\ncovariatesincludewhetherthetrialfocusedonaclinicalendpoint( clinend),\\nwhether the trial involved multiple centers ( multi), the funding mechanism\\nwithintheNationalInstitutesofHealth( mech),trialsamplesize( sampsize ),\\nbudget ( budget), impact ( impact, related to the number of citations), and\\nwhether the trial produced a positive (significant) result ( posres). The last\\ncovariate is particularly interesting, as a number of studies have suggested\\nthat positive trials have a higher publication rate.\\n14This dataset is described in the following paper: Gordon et al. (2013) Publication of\\ntrials funded by the National Heart, Lung, and Blood Institute. New England Journal\\nof Medicine, 369(20):1926–1934.', metadata={'source': './documents/ISLP_website.pdf', 'page': 487}), Document(page_content='11.5 Regression Models With a Survival Response 483\\n0 20 40 60 80 100 1200.0 0.2 0.4 0.6 0.8 1.0MonthsProbability of Not Being PublishedNegative ResultPositive Result\\nFIGURE 11.5. Survival curves for time until publication for the Publication\\ndata described in Section 11.5.4, stratified by whether or not the study produced\\na positive result.\\nFigure11.5showstheKaplan–Meiercurvesforthetimeuntilpublication,\\nstratified by whether or not the study produced a positive result. We see\\nslightevidencethattimeuntilpublicationislowerforstudieswithapositive\\nresult.However,thelog-ranktestyieldsaveryunimpressive p-valueof 0.36.\\nWe now consider a more careful analysis that makes use of all of the\\navailablepredictors.TheresultsoffittingCox’sproportionalhazardsmodel\\nusing all of the available features are shown in Table 11.3. We find that the\\nchance of publication of a study with a positive result is e0.55=1.74times\\nhigher than the chance of publication of a study with a negative result\\nat any point in time, holding all other covariates fixed. The very small\\np-value associated with posresin Table 11.3indicates that this result is\\nhighly significant. This is striking, especially in light of our earlier finding\\nthat a log-rank test comparing time to publication for studies with positive\\nversus negative results yielded a p-value of 0.36. How can we explain this\\ndiscrepancy? The answer stems from the fact that the log-rank test did not\\nconsider any other covariates, whereas the results in Table 11.3are based\\non a Cox model using all of the available covariates. In other words, after\\nwe adjust for all of the other covariates, then whether or not the study\\nyielded a positive result is highly predictive of the time to publication.\\nIn order to gain more insight into this result, in Figure 11.6we display\\nestimates of the survival curves associated with positive and negative re-\\nsults, adjusting for the other predictors. To produce these survival curves,\\nwe estimated the underlying baseline hazard h0(t). We also needed to se-\\nlect representative values for the other predictors; we used the mean value\\nfor each predictor, except for the categorical predictor mech, for which we\\nused the most prevalent category ( R01). Adjusting for the other predictors,\\nwe now see a clear difference in the survival curves between studies with\\npositive versus negative results.\\nOther interesting insights can be gleaned from Table 11.3. For example,\\nstudies with a clinical endpoint are more likely to be published at any\\ngiven point in time than those with a non-clinical endpoint. The funding', metadata={'source': './documents/ISLP_website.pdf', 'page': 488}), Document(page_content='484 11. Survival Analysis and Censored Data\\nCoefficient Std. error z-statistic p-value\\nposres[Yes] 0.55 0.18 3.02 0.00\\nmulti[Yes] 0.15 0.31 0.47 0.64\\nclinend[Yes] 0.51 0.27 1.89 0.06\\nmech[K01] 1.05 1.06 1.00 0.32\\nmech[K23] -0.48 1.05 -0.45 0.65\\nmech[P01] -0.31 0.78 -0.40 0.69\\nmech[P50] 0.60 1.06 0.57 0.57\\nmech[R01] 0.10 0.32 0.30 0.76\\nmech[R18] 1.05 1.05 0.99 0.32\\nmech[R21] -0.05 1.06 -0.04 0.97\\nmech[R24,K24] 0.81 1.05 0.77 0.44\\nmech[R42] -14.78 3414.38 -0.00 1.00\\nmech[R44] -0.57 0.77 -0.73 0.46\\nmech[RC2] -14.92 2243.60 -0.01 0.99\\nmech[U01] -0.22 0.32 -0.70 0.48\\nmech[U54] 0.47 1.07 0.44 0.66\\nsampsize 0.00 0.00 0.19 0.85\\nbudget 0.00 0.00 1.67 0.09\\nimpact 0.06 0.01 8.23 0.00\\nTABLE 11.3. Results for Cox’s proportional hazards model fit to the\\nPublication data, using all of the available features. The features posres,multi,\\nandclinend are binary. The feature mechis qualitative with 14 levels; it is coded\\nso that the baseline level is Contract .\\nmechanism did not appear to be significantly associated with time until\\npublication.\\n11.6 Shrinkage for the Cox Model\\nIn this section, we illustrate that the shrinkage methods of Section 6.2\\ncan be applied to the survival data setting. In particular, motivated by\\nthe “loss+penalty” formulation of Section 6.2, we consider minimizing a\\npenalized version of the negative log partial likelihood in ( 11.16),\\n−log\\uf8eb\\n\\uf8ed∏\\ni:δi=1exp(∑p\\nj=1xijβj)\\n∑\\ni′:yi′≥yiexp(∑p\\nj=1xi′jβj)\\uf8f6\\n\\uf8f8+λP(β), (11.17)\\nwith respect to β=(β1,...,βp)T. We might take P(β)=∑p\\nj=1β2\\nj, which\\ncorresponds to a ridge penalty, or P(β)=∑p\\nj=1|βj|, which corresponds to\\na lasso penalty.\\nIn (11.17),λis a non-negative tuning parameter; typically we will mini-\\nmize it over a range of values of λ. Whenλ=0, then minimizing ( 11.17) is\\nequivalent to simply maximizing the usual Cox partial likelihood ( 11.16).\\nHowever, when λ>0, then minimizing ( 11.17) yields a shrunken version of\\nthe coefficient estimates. When λis large, then using a ridge penalty will\\ngive small coefficients that are not exactly equal to zero. By contrast, for a', metadata={'source': './documents/ISLP_website.pdf', 'page': 489}), Document(page_content='11.6 Shrinkage for the Cox Model 485\\n0 20 40 60 80 100 1200.0 0.2 0.4 0.6 0.8 1.0MonthsProbability of Not Being PublishedNegative ResultPositive Result\\nFIGURE 11.6. For thePublication data, we display survival curves for time\\nuntil publication, stratified by whether or not the study produced a positive result,\\nafter adjusting for all other covariates.\\nsufficiently large value of λ, using a lasso penalty will give some coefficients\\nthat are exactly equal to zero.\\nWenowapplythelasso-penalizedCoxmodeltothe Publication data,de-\\nscribed in Section 11.5.4. We first randomly split the 244 trials into equally-\\nsized training and test sets. The cross-validation results from the training\\nset are shown in Figure 11.7. The “partial likelihood deviance”, shown on\\nthey-axis, is twice the cross-validated negative log partial likelihood; it\\nplays the role of the cross-validation error.15Note the “U-shape” of the\\npartial likelihood deviance: just as we saw in previous chapters, the cross-\\nvalidation error is minimized for an intermediate level of model complexity.\\nSpecifically, this occurs when just two predictors, budgetandimpact, have\\nnon-zero estimated coefficients.\\nNow, how do we apply this model to the test set? This brings up an\\nimportant conceptual point: in essence, there is no simple way to compare\\npredicted survival times and true survival times on the test set. The first\\nproblem is that some of the observations are censored, and so the true sur-\\nvival times for those observations are unobserved. The second issue arises\\nfrom the fact that in the Cox model, rather than predicting a single sur-\\nvival time given a covariate vector x, we instead estimate an entire survival\\ncurve,S(t|x), as a function of t.\\nTherefore, to assess the model fit, we must take a different approach,\\nwhich involves stratifying the observations using the coefficient estimates.\\nIn particular, for each test observation, we compute the “risk” score\\nbudget i·ˆβbudget +impact i·ˆβimpact ,\\nwhereˆβbudget andˆβimpact are the coefficient estimates for these two features\\nfrom the training set. We then use these risk scores to categorize the obser-\\nvations based on their “risk”. For instance, the high risk group consists of\\nthe observations for which budget i·ˆβbudget +impact i·ˆβimpact is largest; by\\n15Cross-validation for the Cox model is more involved than for linear or logistic re-\\ngression, because the objective function is not a sum over the observations.', metadata={'source': './documents/ISLP_website.pdf', 'page': 490}), Document(page_content='486 11. Survival Analysis and Censored Data\\n2e−04 5e−04 1e−03 2e−03 5e−03 1e−02 2e−02 5e−02 1e−01 2e−019.0 9.1 9.2 9.3 9.4 9.5 9.6 9.7Partial Likelihood Deviance∥ˆβLλ∥1/|ˆβ∥1FIGURE 11.7. For thePublication data described in Section 11.5.4, cross-val-\\nidation results for the lasso-penalized Cox model are shown. The y-axis displays\\nthe partial likelihood deviance, which plays the role of the cross-validation error.\\nThex-axis displays the ℓ1norm (that is, the sum of the absolute values) of the\\ncoefficients of the lasso-penalized Cox model with tuning parameter λ, divided by\\ntheℓ1norm of the coefficients of the unpenalized Cox model. The dashed line\\nindicates the minimum cross-validation error.\\n(11.14), we see that these are the observations for which the instantaneous\\nprobability of being published at any moment in time is largest. In other\\nwords, the high risk group consists of the trials that are likely to be pub-\\nlished sooner. On the Publication data, we stratify the observations into\\ntertiles of low, medium, and high risk. The resulting survival curves for\\neach of the three strata are displayed in Figure 11.8. We see that there is\\nclear separation between the three strata, and that the strata are correctly\\nordered in terms of low, medium, and high risk of publication.\\n11.7 Additional Topics\\n11.7.1 Area Under the Curve for Survival Analysis\\nInChapter 4,weintroducedtheareaundertheROCcurve—oftenreferred\\ntoasthe“AUC”—asawaytoquantifytheperformanceofatwo-classclas-\\nsifier. Define the scorefor theith observation to be the classifier’s estimate\\nofPr(Y=1|X=xi). It turns out that if we consider all pairs consisting of\\none observation in Class 1 and one observation in Class 2, then the AUC\\nis the fraction of pairs for which the score for the observation in Class 1\\nexceeds the score for the observation in Class 2.\\nThis suggests a way to generalize the notion of AUC to survival anal-\\nysis. We calculate an estimated risk score, ˆηi=ˆβ1xi1+···+ˆβpxip, for\\ni=1,...,n , using the Cox model coefficients. If ˆηi′>ˆηi, then the model\\npredicts that the i′th observation has a larger hazard than the ith obser-\\nvation, and thus that the survival time tiwill begreaterthanti′. Thus, it\\nis tempting to try to generalize AUC by computing the proportion of ob-\\nservations for which ti>ti′andˆηi′>ˆηi. However, things are not quite so\\neasy, because recall that we do not observe t1,...,t n; instead, we observe', metadata={'source': './documents/ISLP_website.pdf', 'page': 491}), Document(page_content='11.7 Additional Topics 487\\n0 20 40 60 80 100 1200.0 0.2 0.4 0.6 0.8 1.0MonthsProbability of Not Being PublishedLow RiskMedium RiskHigh Risk\\nFIGURE 11.8. For the Publication data introduced in Section 11.5.4, we\\ncompute tertiles of “risk” in the test set using coefficients estimated on the training\\nset. There is clear separation between the resulting survival curves.\\nthe (possibly-censored) times y1,...,y n, as well as the censoring indicators\\nδ1,...,δn.\\nTherefore, Harrell’s concordance index (orC-index) computes the pro-Harrell’s\\nconcordance\\nindexportion of observation pairs for which ˆηi′>ˆηiandyi>yi′:\\nC=∑\\ni,i′:yi>yi′I(ˆηi′>ˆηi)δi′\\n∑\\ni,i′:yi>yi′δi′,\\nwhere the indicator variable I(ˆηi′>ˆηi)equals one if ˆηi′>ˆηi, and equals\\nzero otherwise. The numerator and denominator are multiplied by the sta-\\ntus indicator δi′, since if the i′th observation is uncensored (i.e. if δi′=1),\\nthenyi>y i′implies that ti>ti′. By contrast, if δi′=0, thenyi>y i′\\ndoes not imply that ti>ti′.\\nWe fit a Cox proportional hazards model on the training set of the\\nPublication data, and computed the C-index on the test set. This yielded\\nC=0.733. Roughly speaking, given two random papers from the test set,\\nthe model can predict with 73.3% accuracy which will be published first.\\n11.7.2 Choice of Time Scale\\nIn the examples considered thus far in this chapter, it has been fairly clear\\nhow to define time. For example, in the Publication example, time zero for\\neach paper was defined to be the calendar time at the end of the study,\\nand the failure time was defined to be the number of months that elapsed\\nfrom the end of the study until the paper was published.\\nHowever, in other settings, the definitions of time zero and failure time\\nmay be more subtle. For example, when examining the association between\\nrisk factors and disease occurrence in an epidemiological study, one might\\nuse the patient’s age to define time, so that time zero is the patient’s date\\nof birth. With this choice, the association between age and survival cannot\\nbe measured; however, there is no need to adjust for age in the analysis.\\nWhen examining covariates associated with disease-free survival (i.e. the', metadata={'source': './documents/ISLP_website.pdf', 'page': 492}), Document(page_content='488 11. Survival Analysis and Censored Data\\namount of time elapsed between treatment and disease recurrence), one\\nmight use the date of treatment as time zero.\\n11.7.3 Time-Dependent Covariates\\nA powerfulfeature of the proportional hazards model is its abilityto handle\\ntime-dependent covariates , predictors whose value may change over time.\\nFor example, suppose we measure a patient’s blood pressure every week\\nover the course of a medical study. In this case, we can think of the blood\\npressure for the ith observation not as xi, but rather as xi(t)at timet.\\nBecause the partial likelihood in ( 11.16) is constructed sequentially in\\ntime, dealing with time-dependent covariates is straightforward. In partic-\\nular, we simply replace xijandxi′jin (11.16) withxij(yi)andxi′j(yi),\\nrespectively; these are the current values of the predictors at time yi. By\\ncontrast, time-dependent covariates would pose a much greater challenge\\nwithin the context of a traditional parametric approach, such as ( 11.13).\\nOneexampleoftime-dependentcovariatesappearsintheanalysisofdata\\nfrom the Stanford Heart Transplant Program. Patients in need of a heart\\ntransplant were put on a waiting list. Some patients received a transplant,\\nbut others died while still on the waiting list. The primary objective of the\\nanalysis was to determine whether a transplant was associated with longer\\npatient survival.\\nA naïve approach would use a fixed covariate to represent transplant\\nstatus: that is, xi=1if theith patient ever received a transplant, and xi=\\n0otherwise. But this approach overlooks the fact that patients had to live\\nlong enough to get a transplant, and hence, on average, healthier patients\\nreceivedtransplants.Thisproblemcanbesolvedbyusingatime-dependent\\ncovariate for transplant: xi(t)=1 if the patient received a transplant by\\ntimet, andxi(t)=0 otherwise.\\n11.7.4 Checking the Proportional Hazards Assumption\\nWe have seen that Cox’s proportional hazards model relies on the propor-\\ntional hazards assumption ( 11.14). While results from the Cox model tend\\nto be fairly robust to violations of this assumption, it is still a good idea to\\ncheck whether it holds. In the case of a qualitative feature, we can plot the\\nlog hazard function for each level of the feature. If ( 11.14) holds, then the\\nlog hazard functions should just differ by a constant, as seen in the top-left\\npanel of Figure 11.4. In the case of a quantitative feature, we can take a\\nsimilar approach by stratifying the feature.\\n11.7.5 Survival Trees\\nInChapter 8,wediscussedflexibleandadaptivelearningproceduressuchas\\ntrees,randomforests,andboosting,whichweappliedinboththeregression\\nand classification settings. Most of these approaches can be generalized to\\nthe survival analysis setting. For example, survival trees are a modificationsurvival\\ntreesofclassificationandregressiontreesthatuseasplitcriterionthatmaximizes', metadata={'source': './documents/ISLP_website.pdf', 'page': 493}), Document(page_content=\"11.8 Lab: Survival Analysis 489\\nthe difference between the survival curves in the resulting daughter nodes.\\nSurvival trees can then be used to create random survival forests.\\n11.8 Lab: Survival Analysis\\nIn this lab, we perform survival analyses on three separate data sets. In\\nSection11.8.1we analyze the BrainCancer data that was first described\\nin Section 11.3. In Section 11.8.2, we examine the Publication data from\\nSection11.5.4. Finally, Section 11.8.3explores a simulated call-center data\\nset.\\nWe begin by importing some of our libraries at this top level. This makes\\nthe code more readable, as scanning the first few lines of the notebook tell\\nus what libraries are used in this notebook.\\nIn[1]: from matplotlib.pyplot import subplots\\nimport numpy as np\\nimport pandas as pd\\nfrom ISLP.models import ModelSpec as MS\\nfrom ISLP import load_data\\nWe also collect the new imports needed for this lab.\\nIn[2]: from lifelines import \\\\\\n(KaplanMeierFitter,\\nCoxPHFitter)\\nfrom lifelines.statistics import \\\\\\n(logrank_test,\\nmultivariate_logrank_test)\\nfrom ISLP.survival import sim_time\\n11.8.1 Brain Cancer Data\\nWe begin with the BrainCancer data set, contained in the ISLPpackage.\\nIn[3]: BrainCancer = load_data( 'BrainCancer' )\\nBrainCancer.columns\\nOut[3]: Index([ 'sex','diagnosis' ,'loc','ki','gtv','stereo' ,\\n'status' ,'time'],\\ndtype='object' )\\nThe rows index the 88 patients, while the 8 columns contain the predictors\\nand outcome variables. We first briefly examine the data.\\nIn[4]: BrainCancer[ 'sex'].value_counts()\\nOut[4]: Female 45\\nMale 43\\nName: sex, dtype: int64\\nIn[5]: BrainCancer[ 'diagnosis' ].value_counts()\", metadata={'source': './documents/ISLP_website.pdf', 'page': 494}), Document(page_content=\"490 11. Survival Analysis and Censored Data\\nOut[5]: Meningioma 42\\nHG glioma 22\\nOther 14\\nLG glioma 9\\nName: diagnosis, dtype: int64\\nIn[6]: BrainCancer[ 'status' ].value_counts()\\nOut[6]: 05 3\\n13 5\\nName: status, dtype: int64\\nBefore beginning an analysis, it is important to know how the status\\nvariable has been coded. Most software uses the convention that a status\\nof 1 indicates an uncensored observation (often death), and a statusof 0\\nindicatesacensoredobservation.Butsomescientistsmightusetheopposite\\ncoding. For the BrainCancer data set 35 patients died before the end of the\\nstudy, so we are using the conventional coding.\\nTo begin the analysis, we re-create the Kaplan-Meier survival curve\\nshown in Figure 11.2. The main package we will use for survival analy-\\nsis islifelines . The variable timecorresponds to yi, the time to the ithlifelinesevent (either censoring or death). The first argument to km.fitis the event\\ntime, and the second argument is the censoring variable, with a 1 indicat-\\ning an observed failure time. The plot()method produces a survival curve.plot()with pointwise confidence intervals. By default, these are 90% confidence\\nintervals, but this can be changed by setting the alphaargument to one\\nminus the desired confidence level.\\nIn[7]: fig, ax = subplots(figsize=(8,8))\\nkm = KaplanMeierFitter()\\nkm_brain = km.fit(BrainCancer[ 'time'], BrainCancer[ 'status' ])\\nkm_brain.plot(label= 'Kaplan Meier estimate' ,a x = a x )\\nNext we create Kaplan-Meier survival curves that are stratified by sex, in\\norder to reproduce Figure 11.3. We do this using the groupby() method of.groupby()a dataframe. This method returns a generator that can be iterated over in\\ntheforloop. In this case, the items in the forloop are 2-tuples representing\\nthe groups: the first entry is the value of the grouping column sexwhile\\nthe second value is the dataframe consisting of all rows in the dataframe\\nmatching that value of sex. We will want to use this data below in the log-\\nrank test, hence we store this information in the dictionary by_sex. Finally,\\nwe have also used the notion of string interpolation to automatically labelstring\\ninterpolationthe different lines in the plot. String interpolation is a powerful technique\\nto format strings — Pythonhas many ways to facilitate such operations.\\nIn[8]: fig, ax = subplots(figsize=(8,8))\\nby_sex = {}\\nfor sex, df in BrainCancer.groupby( 'sex'):\\nby_sex[sex] = df\\nkm_sex = km.fit(df[ 'time'], df['status' ])\\nkm_sex.plot(label= 'Sex=%s' %s e x ,a x = a x )\\nAs discussed in Section 11.4, we can perform a log-rank test to compare\\nthe survival of males to females. We use the logrank_test() function fromlogrank_\\ntest()\", metadata={'source': './documents/ISLP_website.pdf', 'page': 495}), Document(page_content=\"11.8 Lab: Survival Analysis 491thelifelines.statisticsmodule. The first two arguments are the eventtimes, with the second denoting the corresponding (optional) censoringindicators.In[9]:logrank_test(by_sex['Male']['time'],by_sex['Female']['time'],by_sex['Male']['status'],by_sex['Female']['status'])Out[9]:t_0-1null_distribution chi squareddegrees_of_freedom1test_name logrank_testtest_statistic p -log2(p)1.44 0.23 2.12The resultingp-value is0.23, indicating no evidence of a difference insurvival between the two sexes.Next, we use theCoxPHFitter()estimator fromlifelinesto fit CoxCoxPHFitter()proportional hazards models. To begin, we consider a model that usessexas the only predictor.In[10]:coxph = CoxPHFitter#s h o r t h a n dsex_df = BrainCancer[['time','status','sex']]model_df = MS(['time','status','sex'],intercept=False).fit_transform(sex_df)cox_fit = coxph().fit(model_df,'time','status')cox_fit.summary[['coef','se(coef)','p']]Out[10]:coef se(coef)pcovariatesex[Male] 0.407667 0.342004 0.233263The first argument tofitshould be a data frame containing at least theevent time (the second argumenttimein this case), as well as an op-tional censoring variable (the argumentstatusin this case). Note also thatthe Cox model does not include an intercept, which is why we used theintercept=Falseargument toModelSpecabove. Thesummary()method de-livers many columns; we chose to abbreviate its output here. It is possibleto obtain the likelihood ratio test comparing this model to the one with nofeatures as follows:In[11]:cox_fit.log_likelihood_ratio_test()Out[11]:null_distributionchi squareddegrees_freedom1test_name log-likelihood ratio testtest_statistic p -log2(p)1.44 0.23 2.12\", metadata={'source': './documents/ISLP_website.pdf', 'page': 496}), Document(page_content=\"492 11. Survival Analysis and Censored DataRegardless of which test we use, we see that there is no clear evidence fora difference in survival between males and females. As we learned in thischapter, the score test from the Cox model is exactly equal to the log ranktest statistic!Now we fit a model that makes use of additional predictors. We first notethat one of ourdiagnosisvalues is missing, hence we drop that observationbefore continuing.In[12]:cleaned = BrainCancer.dropna()all_MS = MS(cleaned.columns, intercept=False)all_df = all_MS.fit_transform(cleaned)fit_all = coxph().fit(all_df,'time','status')fit_all.summary[['coef','se(coef)','p']]Out[12]:coef se(coef)pcovariatesex[Male] 0.183748 0.360358 0.610119diagnosis[LG glioma] -1.239541 0.579557 0.032454diagnosis[Meningioma] -2.154566 0.450524 0.000002diagnosis[Other] -1.268870 0.617672 0.039949loc[Supratentorial] 0.441195 0.703669 0.530664ki -0.054955 0.018314 0.002693gtv 0.034293 0.022333 0.124660stereo[SRT] 0.177778 0.601578 0.767597Thediagnosisvariable has been coded so that the baseline corresponds toHG glioma. The results indicate that the risk associated with HG gliomais more than eight times (i.e.e2.15=8.62) the risk associated with menin-gioma. In other words, after adjusting for the other predictors, patientswith HG glioma have much worse survival compared to those with menin-gioma. In addition, larger values of the Karnofsky index,ki, are associatedwith lower risk, i.e. longer survival.Finally, we plot estimated survival curves for each diagnosis category,adjusting for the other predictors. To make these plots, we set the values ofthe other predictors equal to the mean for quantitative variables and equalto the mode for categorical. To do this, we use theapply()method acrossrows (i.e.axis=0) with a functionrepresentativethat checks if a columnis categorical or not.In[13]:levels = cleaned['diagnosis'].unique()def representative(series):if hasattr(series.dtype,'categories'):return pd.Series.mode(series)else:return series.mean()modal_data = cleaned.apply(representative, axis=0)We make four copies of the column means and assign thediagnosiscolumn to be the four different diagnoses.In[14]:modal_df = pd.DataFrame([modal_data.iloc[0] for _ in range(len(levels))])modal_df['diagnosis']=l e v e l smodal_df\", metadata={'source': './documents/ISLP_website.pdf', 'page': 497}), Document(page_content=\"11.8 Lab: Survival Analysis 493\\nOut[14]: sex diagnosis loc ki gtv stereo ...\\nFemale Meningioma Supratentorial 80.920 8.687 SRT ...\\nFemale HG glioma Supratentorial 80.920 8.687 SRT ...\\nFemale LG glioma Supratentorial 80.920 8.687 SRT ...\\nFemale Other Supratentorial 80.920 8.687 SRT ...\\nWe then construct the model matrix based on the model specification\\nall_MSused to fit the model, and name the rows according to the levels of\\ndiagnosis .\\nIn[15]: modal_X = all_MS.transform(modal_df)\\nmodal_X.index = levels\\nmodal_X\\nWe can use the predict_survival_function() method to obtain the esti-.predict_\\nsurvival_\\nfunction()mated survival function.\\nIn[16]: predicted_survival = fit_all.predict_survival_function(modal_X)\\npredicted_survival\\nOut[16]: Meningioma HG glioma LG glioma Other\\n0.070 0.998 0.982 0.995 0.995\\n1.180 0.998 0.982 0.995 0.995\\n1.410 0.996 0.963 0.989 0.990\\n1.540 0.996 0.963 0.989 0.990\\n... ... ... ... ...\\n67.380 0.689 0.040 0.394 0.405\\n73.740 0.689 0.040 0.394 0.405\\n78.750 0.689 0.040 0.394 0.405\\n82.560 0.689 0.040 0.394 0.405\\n85 rows × 4 columns\\nThis returns a data frame, whose plot methods yields the different survival\\ncurves. To avoid clutter in the plots, we do not display confidence intervals.\\nIn[17]: fig, ax = subplots(figsize=(8, 8))\\npredicted_survival.plot(ax=ax);\\n11.8.2 Publication Data\\nThePublication data presented in Section 11.5.4can be found in the\\nISLPpackage. We first reproduce Figure 11.5by plotting the Kaplan-Meier\\ncurves stratified on the posresvariable, which records whether the study\\nhad a positive or negative result.\\nIn[18]: fig, ax = subplots(figsize=(8,8))\\nPublication = load_data( 'Publication' )\\nby_result = {}\\nfor result, df in Publication.groupby( 'posres' ):\\nby_result[result] = df\\nkm_result = km.fit(df[ 'time'], df['status' ])\\nkm_result.plot(label= 'Result=%d' %r e s u l t ,a x = a x )\\nAs discussed previously, the p-values from fitting Cox’s proportional haz-\\nards model to the posresvariable are quite large, providing no evidence\\nof a difference in time-to-publication between studies with positive versus\\nnegative results.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 498}), Document(page_content=\"494 11. Survival Analysis and Censored Data\\nIn[19]: posres_df = MS([ 'posres' ,\\n'time',\\n'status' ],\\nintercept=False).fit_transform(Publication)\\nposres_fit = coxph().fit(posres_df,\\n'time',\\n'status' )\\nposres_fit.summary[[ 'coef','se(coef)' ,'p']]\\nOut[19]: coef se(coef) p\\ncovariate\\nposres 0.148076 0.161625 0.359578\\nHowever, the results change dramatically when we include other predic-\\ntors in the model. Here we exclude the funding mechanism variable.\\nIn[20]: model = MS(Publication.columns.drop( 'mech'),\\nintercept=False)\\ncoxph().fit(model.fit_transform(Publication),\\n'time',\\n'status' ).summary[[ 'coef','se(coef)' ,'p']]\\nOut[20]: coef se(coef) p\\ncovariate\\nposres 0.570774 0.175960 1.179606e-03\\nmulti -0.040863 0.251194 8.707727e-01\\nclinend 0.546180 0.262001 3.710099e-02\\nsampsize 0.000005 0.000015 7.506978e-01\\nbudget 0.004386 0.002464 7.511276e-02\\nimpact 0.058318 0.006676 2.426779e-18\\nWe see that there are a number of statistically significant variables, in-\\ncluding whether the trial focused on a clinical endpoint, the impact of the\\nstudy, and whether the study had positive or negative results.\\n11.8.3 Call Center Data\\nInthissection,wewillsimulatesurvivaldatausingtherelationshipbetween\\ncumulative hazard and the survival function explored in Exercise 8. Our\\nsimulated data will represent the observed wait times (in seconds) for 2,000\\ncustomers who have phoned a call center. In this context, censoring occurs\\nif a customer hangs up before his or her call is answered.\\nTherearethreecovariates: Operators (thenumberofcallcenteroperators\\navailable at the time of the call, which can range from 5to15),Center\\n(either A, B, or C), and Timeof day (Morning, Afternoon, or Evening). We\\ngenerate data for these covariates so that all possibilities are equally likely:\\nfor instance, morning, afternoon and evening calls are equally likely, and\\nany number of operators from 5to15is equally likely.\\nIn[21]: rng = np.random.default_rng(10)\\nN=2 0 0 0\\nOperators = rng.choice(np.arange(5, 16),\\nN,\\nreplace=True)\", metadata={'source': './documents/ISLP_website.pdf', 'page': 499}), Document(page_content=\"11.8 Lab: Survival Analysis 495\\nCenter = rng.choice([ 'A','B','C'],\\nN,\\nreplace=True)\\nTime = rng.choice([ 'Morn.' ,'After.' ,'Even.' ],\\nN,\\nreplace=True)\\nD=p d . D a t a F r a m e ( { 'Operators' :O p e r a t o r s ,\\n'Center' :p d . C a t e g o r i c a l ( C e n t e r ) ,\\n'Time':p d . C a t e g o r i c a l ( T i m e ) } )\\nWe then build a model matrix (omitting the intercept)\\nIn[22]: model = MS([ 'Operators' ,\\n'Center' ,\\n'Time'],\\nintercept=False)\\nX=m o d e l . f i t _ t r a n s f o r m ( D )\\nIt is worthwhile to take a peek at the model matrix X, so that we can be\\nsure that we understand how the variables have been coded. By default,\\nthe levels of categorical variables are sorted and, as usual, the first column\\nof the one-hot encoding of the variable is dropped.\\nIn[23]: X[:5]\\nOut[23]: Operators Center[B] Center[C] Time[Even.] Time[Morn.]\\n0 13 0.0 1.0 0.0 0.0\\n1 15 0.0 0.0 1.0 0.0\\n2 71 . 00 . 0 0.0 1.0\\n3 70 . 01 . 0 0.0 1.0\\n4 13 0.0 1.0 1.0 0.0\\nNext, we specify the coefficients and the hazard function.\\nIn[24]: true_beta = np.array([0.04, -0.3, 0, 0.2, -0.2])\\ntrue_linpred = X.dot(true_beta)\\nhazard = lambda t: 1e-5 * t\\nHere, we have set the coefficient associated with Operators to equal 0.04;\\nin other words, each additional operator leads to a e0.04=1.041-fold in-\\ncrease in the “risk” that the call will be answered, given the Centerand\\nTimecovariates. This makes sense: the greater the number of operators at\\nhand, the shorter the wait time! The coefficient associated with Center ==\\nBis−0.3, andCenter == A is treated as the baseline. This means that the\\nrisk of a call being answered at Center B is 0.74 times the risk that it will\\nbe answered at Center A; in other words, the wait times are a bit longer\\nat Center B.\\nRecall from Section 2.3.7the use of lambdafor creating short functions\\non the fly. We use the function sim_time() from the ISLP.survival pack-sim_time()age. This function uses the relationship between the survival function and\\ncumulative hazard S(t) = exp(−H(t))and the specific form of the cumu-\\nlative hazard function in the Cox model to simulate data based on values\\nof the linear predictor true_linpred and the cumulative hazard. We need\\nto provide the cumulative hazard function, which we do here.\\nIn[25]: cum_hazard = lambda t: 1e-5 * t**2 / 2\", metadata={'source': './documents/ISLP_website.pdf', 'page': 500}), Document(page_content='496 11. Survival Analysis and Censored DataWe are now ready to generate data under the Cox proportional hazardsmodel. We truncate the maximum time to 1000 seconds to keep simulatedwait times reasonable. The functionsim_time()takes a linear predictor, acumulative hazard function and a random number generator.In[26]:W=n p . a r r a y ( [ s i m _ t i m e ( l ,c u m _ h a z a r d ,r n g )for l in true_linpred])D[\\'Wait time\\']=n p . c l i p ( W ,0 ,1 0 0 0 )We now simulate our censoring variable, for which we assume 90% ofcalls were answered (Failed==1) before the customer hung up (Failed==0).In[27]:D[\\'Failed\\']=r n g . c h o i c e ( [ 1 ,0 ] ,N,p=[0.9, 0.1])D[:5]Out[27]:Operators Center Time Wait time Failed013 C After. 525.064979 1115 A Even. 254.677835 127B M o r n . 4 8 7 . 7 3 9 2 2 4 137C M o r n . 3 0 8 . 5 8 0 2 9 2 1413 C Even. 154.174608 1In[28]:D[\\'Failed\\'].mean()Out[28]:0.8985We now plot Kaplan-Meier survival curves. First, we stratify byCenter.In[29]:fig, ax = subplots(figsize=(8,8))by_center = {}for center, df in D.groupby(\\'Center\\'):by_center[center] = dfkm_center = km.fit(df[\\'Wait time\\'], df[\\'Failed\\'])km_center.plot(label=\\'Center=%s\\'%c e n t e r ,a x = a x )ax.set_title(\"Probability of Still Being on Hold\")Next, we stratify byTime.In[30]:fig, ax = subplots(figsize=(8,8))by_time = {}for time, df in D.groupby(\\'Time\\'):by_time[time] = dfkm_time = km.fit(df[\\'Wait time\\'], df[\\'Failed\\'])km_time.plot(label=\\'Time=%s\\'%t i m e ,a x = a x )ax.set_title(\"Probability of Still Being on Hold\")It seems that calls at Call Center B take longer to be answered than callsat Centers A and C. Similarly, it appears that wait times are longest in themorning and shortest in the evening hours. We can use a log-rank test todetermine whether these differences are statistically significant using thefunctionmultivariate_logrank_test().In[31]:multivariate_logrank_test(D[\\'Wait time\\'],D[\\'Center\\'],D[\\'Failed\\'])', metadata={'source': './documents/ISLP_website.pdf', 'page': 501}), Document(page_content=\"11.8 Lab: Survival Analysis 497Out[31]:t_0-1null_distributionchi squareddegrees_of_freedom2test_name multivariate_logrank_testtest_statistic p -log2(p)20.30 <0.005 14.65Next, we consider the effect ofTime.In[32]:multivariate_logrank_test(D['Wait time'],D['Time'],D['Failed'])Out[32]:t_0-1null_distributionchi squareddegrees_of_freedom2test_name multivariate_logrank_testtest_statistic p -log2(p)49.90 <0.005 35.99As in the case of a categorical variable with 2 levels, these results aresimilartothelikelihoodratiotestfromtheCoxproportionalhazardsmodel.First, we look at the results forCenter.In[33]:X=M S ( ['Wait time','Failed','Center'],intercept=False).fit_transform(D)F=c o x p h ( ) . f i t ( X ,'Wait time','Failed')F.log_likelihood_ratio_test()Out[33]:null_distributionchi squareddegrees_freedom2test_name log-likelihood ratio testtest_statistic p -log2(p)20.58 <0.005 14.85Next, we look at the results forTime.In[34]:X=M S ( ['Wait time','Failed','Time'],intercept=False).fit_transform(D)F=c o x p h ( ) . f i t ( X ,'Wait time','Failed')F.log_likelihood_ratio_test()Out[34]:null_distributionchi squareddegrees_freedom2test_name log-likelihood ratio testtest_statistic p -log2(p)48.12 <0.005 34.71We find that differences between centers are highly significant, as aredifferences between times of day.Finally, we fit Cox’s proportional hazards model to the data.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 502}), Document(page_content=\"498 11. Survival Analysis and Censored DataIn[35]:X=M S ( D . c o l u m n s ,intercept=False).fit_transform(D)fit_queuing = coxph().fit(X,'Wait time','Failed')fit_queuing.summary[['coef','se(coef)','p']]Out[35]:coef se(coef)pcovariateOperators 0.043934 0.007520 5.143677e-09Center[B] -0.236059 0.058113 4.864734e-05Center[C] 0.012231 0.057518 8.316083e-01Time[Even.] 0.268845 0.057797 3.294914e-06Time[Morn.] -0.148215 0.057334 9.734378e-03Thep-values for Center B and evening time are very small. It is alsoclear that the hazard — that is, the instantaneous risk that a call will beanswered — increases with the number of operators. Since we generatedthe data ourselves, we know that the true coefficients forOperators,Center=B,Center = C,Time = Even.andTime = Morn.are0.04,−0.3,0,0.2, and−0.2, respectively. The coefficient estimates from the fitted Cox model arefairly accurate.11.9 ExercisesConceptual1. For each example, state whether or not the censoring mechanism isindependent. Justify your answer.(a) In a study of disease relapse, due to a careless research scientist,all patients whose phone numbers begin with the number “2”are lost to follow up.(b) In a study of longevity, a formatting error causes all patient agesthat exceed 99 years to be lost (i.e. we know that those patientsare more than 99 years old, but we do not know their exactages).(c) Hospital A conducts a study of longevity. However, very sickpatients tend to be transferred to Hospital B, and are lost tofollow up.(d) In a study of unemployment duration, the people who find workearlier are less motivated to stay in touch with study investiga-tors, and therefore are more likely to be lost to follow up.(e) In a study of pregnancy duration, women who deliver their ba-bies pre-term are more likely to do so away from their usualhospital, and thus are more likely to be censored, relative towomen who deliver full-term babies.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 503}), Document(page_content='11.9 Exercises 499\\n(f) A researcher wishes to model the number of years of education\\nof the residents of a small town. Residents who enroll in college\\nout of town are more likely to be lost to follow up, and are\\nalso more likely to attend graduate school, relative to those who\\nattend college in town.\\n(g) Researchers conduct a study of disease-free survival (i.e. time\\nuntil disease relapse following treatment). Patients who have\\nnot relapsed within five years are considered to be cured, and\\nthus their survival time is censored at five years.\\n(h) Wewishtomodelthefailuretimeforsomeelectricalcomponent.\\nThis component can be manufactured in Iowa or in Pittsburgh,\\nwith no difference in quality. The Iowa factory opened five years\\nago, and so components manufactured in Iowa are censored at\\nfiveyears.ThePittsburghfactoryopenedtwoyearsago,sothose\\ncomponents are censored at two years.\\n(i) We wish to model the failure time of an electrical component\\nmade in two different factories, one of which opened before the\\nother. We have reason to believe that the components manufac-\\ntured in the factory that opened earlier are of higher quality.\\n2. We conduct a study with n=4participants who have just purchased\\ncell phones, in order to model the time until phone replacement. The\\nfirst participant replaces her phone after 1.2 years. The second par-\\nticipant still has not replaced her phone at the end of the two-year\\nstudy period. The third participant changes her phone number and is\\nlost to follow up (but has not yet replaced her phone) 1.5 years into\\nthe study. The fourth participant replaces her phone after 0.2 years.\\nFor each of the four participants ( i=1,...,4), answer the following\\nquestions using the notation introduced in Section 11.1:\\n(a) Is the participant’s cell phone replacement time censored?\\n(b) Is the value of ciknown, and if so, then what is it?\\n(c) Is the value of tiknown, and if so, then what is it?\\n(d) Is the value of yiknown, and if so, then what is it?\\n(e) Is the value of δiknown, and if so, then what is it?\\n3. For the example in Exercise 2, report the values of K,d1,...,d K,\\nr1,...,r K, andq1,...,q K, where this notation was defined in Sec-\\ntion11.3.\\n4. This problem makes use of the Kaplan-Meier survival curve displayed\\nin Figure 11.9. The raw data that went into plotting this survival\\ncurve is given in Table 11.4. The covariate column of that table is\\nnot needed for this problem.\\n(a) What is the estimated probability of survival past 50 days?', metadata={'source': './documents/ISLP_website.pdf', 'page': 504}), Document(page_content='500 11. Survival Analysis and Censored Data\\nObservation ( Y) Censoring Indicator ( δ) Covariate ( X)\\n26.5 1 0.1\\n37.2 1 11\\n57.3 1 -0.3\\n90.8 0 2.8\\n20.2 0 1.8\\n89.8 0 0.4\\nTABLE 11.4. Data used in Exercise 4.\\n(b) Write out an analytical expression for the estimated survival\\nfunction. For instance, your answer might be something along\\nthe lines of\\nˆS(t)=\\uf8f1\\n\\uf8f4\\uf8f2\\n\\uf8f4\\uf8f30.8 ift<31\\n0.5 if31≤t<77\\n0.22 if77≤t.\\n(The previous equation is for illustration only: it is not the cor-\\nrect answer!)\\n5. Sketch the survival function given by the equation\\nˆS(t)=\\uf8f1\\n\\uf8f4\\uf8f2\\n\\uf8f4\\uf8f30.8 ift<31\\n0.5 if31≤t<77\\n0.22 if77≤t.\\nYour answer should look something like Figure 11.9.\\n0204060800.0 0.2 0.4 0.6 0.8 1.0Time in DaysEstimated Probability of SurvivalFIGURE 11.9. A Kaplan-Meier survival curve used in Exercise 4.\\n6. This problem makes use of the data displayed in Figure 11.1. In\\ncompleting this problem, you can refer to the observation times as\\ny1,...,y 4. The ordering of these observation times can be seen from\\nFigure11.1; their exact values are not required.\\n(a) Report the values of δ1,...,δ4,K,d1,...,d K,r1,...,r K, and\\nq1,...,q K. The relevant notation is defined in Sections 11.1and\\n11.3.', metadata={'source': './documents/ISLP_website.pdf', 'page': 505}), Document(page_content='11.9 Exercises 501\\n(b) Sketch the Kaplan-Meier survival curve corresponding to this\\ndata set. (You do not need to use any software to do this — you\\ncan sketch it by hand using the results obtained in (a).)\\n(c) Based on the survival curve estimated in (b), what is the proba-\\nbility that the event occurs within 200days? What is the prob-\\nability that the event does not occur within 310 days?\\n(d) Write out an expression for the estimated survival curve from\\n(b).\\n7. In this problem, we will derive ( 11.5) and (11.6), which are needed\\nfor the construction of the log-rank test statistic ( 11.8). Recall the\\nnotation in Table 11.1.\\n(a) Assumethatthereisnodifferencebetweenthesurvivalfunctions\\nof the two groups. Then we can think of q1kas the number of\\nfailures if we draw r1kobservations, without replacement, from\\na risk set of rkobservations that contains a total of qkfailures.\\nArgue that q1kfollows a hypergeometric distribution . Write thehyper-\\ngeometric\\ndistributionparameters of this distribution in terms of r1k,rk, andqk.\\n(b) Given your previous answer, and the properties of the hyper-\\ngeometric distribution, what are the mean and variance of q1k?\\nCompare your answer to ( 11.5) and (11.6).\\n8. Recall that the survival function S(t), the hazard function h(t), and\\nthe density function f(t)are defined in ( 11.2), (11.9), and (11.11),\\nrespectively. Furthermore, define F(t)=1 −S(t). Show that the\\nfollowing relationships hold:\\nf(t)= dF(t)/dt\\nS(t) = exp(\\n−∫t\\n0h(u)du)\\n.\\n9. In this exercise, we will explore the consequences of assuming that\\nthe survival times follow an exponential distribution.\\n(a) Suppose that a survival time follows an Exp (λ)distribution,\\nso that its density function is f(t)= λexp(−λt). Using the\\nrelationships provided in Exercise 8, show that S(t) = exp( −λt).\\n(b) Now suppose that each of nindependent survival times follows\\nan Exp(λ)distribution. Write out an expression for the likeli-\\nhood function ( 11.13).\\n(c) Show that the maximum likelihood estimator for λis\\nˆλ=n∑\\ni=1δi/n∑\\ni=1yi.\\n(d) Use your answer to (c) to derive an estimator of the mean sur-\\nvival time.\\nHint: For (d), recall that the mean of an Exp( λ)random variable is\\n1/λ.', metadata={'source': './documents/ISLP_website.pdf', 'page': 506}), Document(page_content='502 11. Survival Analysis and Censored Data\\nApplied\\n10. This exercise focuses on the brain tumor data, which is included in\\ntheISLPlibrary.\\n(a) Plot the Kaplan-Meier survival curve with ±1standard error\\nbands,usingthe KaplanMeierFitter() estimatorinthe lifelines\\npackage.\\n(b) Draw a bootstrap sample of size n= 88 from the pairs (yi,δi),\\nand compute the resulting Kaplan-Meier survival curve. Repeat\\nthisprocess B= 200 times.Usetheresultstoobtainanestimate\\nof the standard error of the Kaplan-Meier survival curve at each\\ntimepoint. Compare this to the standard errors obtained in (a).\\n(c) Fit a Cox proportional hazards model that uses all of the pre-\\ndictors to predict survival. Summarize the main findings.\\n(d) Stratify the data by the value of ki. (Since only one observation\\nhaski==40,youcangroupthatobservationtogetherwiththeob-\\nservations that have ki==60.) Plot Kaplan-Meier survival curves\\nfor each of the five strata, adjusted for the other predictors.\\n11. This exercise makes use of the data in Table 11.4.\\n(a) Create two groups of observations. In Group 1, X<2, whereas\\nin Group 2, X≥2. Plot the Kaplan-Meier survival curves corre-\\nsponding to the two groups. Be sure to label the curves so that\\nit is clear which curve corresponds to which group. By eye, does\\nthere appear to be a difference between the two groups’ survival\\ncurves?\\n(b) Fit Cox’s proportional hazards model, using the group indicator\\nas a covariate. What is the estimated coefficient? Write a sen-\\ntence providing the interpretation of this coefficient, in terms\\nof the hazard or the instantaneous probability of the event. Is\\nthere evidence that the true coefficient value is non-zero?\\n(c) Recall from Section 11.5.2that in the case of a single binary\\ncovariate, the log-ran k test statistic should be identical to the\\nscore statistic for the Cox model. Conduct a log-rank test to de-\\ntermine whether there is a difference between the survival curves\\nfor the two groups. How does the p-value for the log-rank test\\nstatistic compare to the p-value for the score statistic for the\\nCox model from (b)?', metadata={'source': './documents/ISLP_website.pdf', 'page': 507}), Document(page_content='12\\nUnsupervised Learning\\nMost of this book concerns supervised learning methods such as\\nregressionandclassification.Inthesupervisedlearningsetting,wetypically\\nhave access to a set of pfeatures X1,X2,...,X p, measured on nobser-\\nvations, and a response Yalso measured on those same nobservations.\\nThe goal is then to predict YusingX1,X2,...,X p.\\nThis chapter will instead focus on unsupervised learning , a set of sta-\\ntistical tools intended for the setting in which we have only a set of fea-\\nturesX1,X2,...,X pmeasured on nobservations. We are not interested\\nin prediction, because we do not have an associated response variable Y.\\nRather, the goal is to discover interesting things about the measurements\\nonX1,X2,...,X p. Is there an informative way to visualize the data? Can\\nwe discover subgroups among the variables or among the observations?\\nUnsupervised learning refers to a diverse set of techniques for answering\\nquestions such as these. In this chapter, we will focus on two particu-\\nlar types of unsupervised learning: principal components analysis , a tool\\nused for data visualization or data pre-processing before supervised tech-\\nniques are applied, and clustering , a broad class of methods for discovering\\nunknown subgroups in data.\\n12.1 The Challenge of Unsupervised Learning\\nSupervised learning is a well-understood area. In fact, if you have read\\nthe preceding chapters in this book, then you should by now have a good\\ngrasp of supervised learning. For instance, if you are asked to predict a\\nbinary outcome from a data set, you have a very well developed set of tools\\nat your disposal (such as logistic regression, linear discriminant analysis,\\nclassification trees, support vector machines, and more) as well as a clear\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning , Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_12 503', metadata={'source': './documents/ISLP_website.pdf', 'page': 508}), Document(page_content='504 12. Unsupervised Learning\\nunderstanding of how to assess the quality of the results obtained (using\\ncross-validation, validation on an independent test set, and so forth).\\nIn contrast, unsupervised learning is often much more challenging. The\\nexercise tends to be more subjective, and there is no simple goal for the\\nanalysis, such as prediction of a response. Unsupervised learning is often\\nperformed as part of an exploratory data analysis . Furthermore, it can beexploratory\\ndata\\nanalysishard to assess the results obtained from unsupervised learning methods,\\nsince there is no universally accepted mechanism for performing cross-\\nvalidation or validating results on an independent data set. The reason\\nfor this difference is simple. If we fit a predictive model using a supervised\\nlearning technique, then it is possible to check our work by seeing how\\nwell our model predicts the response Yon observations not used in fitting\\nthe model. However, in unsupervised learning, there is no way to check our\\nworkbecause wedon’t knowthe trueanswer—theproblemis unsupervised.\\nTechniques for unsupervised learning are of growing importance in a\\nnumber of fields. A cancer researcher might assay gene expression levels in\\n100 patients with breast cancer. He or she might then look for subgroups\\namong the breast cancer samples, or among the genes, in order to obtain\\na better understanding of the disease. An online shopping site might try\\nto identify groups of shoppers with similar browsing and purchase histo-\\nries, as well as items that are of particular interest to the shoppers within\\neach group. Then an individual shopper can be preferentially shown the\\nitems in which he or she is particularly likely to be interested, based on\\nthe purchase histories of similar shoppers. A search engine might choose\\nwhich search results to display to a particular individual based on the click\\nhistories of other individuals with similar search patterns. These statistical\\nlearningtasks,and manymore, canbeperformed viaunsupervised learning\\ntechniques.\\n12.2 Principal Components Analysis\\nPrincipal components are discussed in Section 6.3.1in the context of\\nprincipal components regression. When faced with a large set of corre-\\nlated variables, principal components allow us to summarize this set with\\na smaller number of representative variables that collectively explain most\\nof the variability in the original set. The principal component directions\\nare presented in Section 6.3.1as directions in feature space along which\\nthe original data are highly variable . These directions also define lines and\\nsubspaces that are as close as possible to the data cloud. To perform\\nprincipal components regression, we simply use principal components as\\npredictors in a regression model in place of the original larger set of vari-\\nables.\\nPrincipal components analysis (PCA)referstotheprocessbywhichprin-principal\\ncomponents\\nanalysiscipal components are computed, and the subsequent use of these compo-\\nnents in understanding the data. PCA is an unsupervised approach, since\\nit involves only a set of features X1,X2,...,X p, and no associated response\\nY. Apart from producing derived variables for use in supervised learning\\nproblems, PCA also serves as a tool for data visualization (visualization of', metadata={'source': './documents/ISLP_website.pdf', 'page': 509}), Document(page_content='12.2 Principal Components Analysis 505the observations or visualization of the variables). It can also be used as atool for data imputation — that is, for filling in missing values in a datamatrix.We now discuss PCA in greater detail, focusing on the use of PCA asa tool for unsupervised data exploration, in keeping with the topic of thischapter.12.2.1 What Are Principal Components?Suppose that we wish to visualizenobservations with measurements on aset ofpfeatures,X1,X2,...,Xp, as part of an exploratory data analysis.We could do this by examining two-dimensional scatterplots of the data,each of which contains thenobservations’ measurements on two of thefeatures.However,there are(p2)=p(p−1)/2suchscatterplots; for example,withp= 10there are 45 plots! Ifpis large, then it will certainly not bepossible to look at all of them; moreover, most likely none of them willbe informative since they each contain just a small fraction of the totalinformation present in the data set. Clearly, a better method is required tovisualize thenobservations whenpis large. In particular, we would like tofind a low-dimensional representation of the data that captures as much oftheinformationaspossible.Forinstance,ifwecanobtainatwo-dimensionalrepresentation of the data that captures most of the information, then wecan plot the observations in this low-dimensional space.PCA provides a tool to do just this. It finds a low-dimensional represen-tation of a data set that contains as much as possible of the variation. Theidea is that each of thenobservations lives inp-dimensional space, but notall of these dimensions are equally interesting. PCA seeks a small numberof dimensions that are as interesting as possible, where the concept ofin-terestingis measured by the amount that the observations vary along eachdimension. Each of the dimensions found by PCA is a linear combinationof thepfeatures. We now explain the manner in which these dimensions,orprincipal components, are found.Thefirst principal componentof a set of featuresX1,X2,...,Xpis thenormalized linear combination of the featuresZ1=φ11X1+φ21X2+···+φp1Xp(12.1)that has the largest variance. Bynormalized, we mean that∑pj=1φ2j1=1.We refer to the elementsφ11,...,φp1as theloadingsof the first principalloadingcomponent; together, the loadings make up the principal component load-ing vector,φ1=(φ11φ21...φp1)T. We constrain the loadings so thattheir sum of squares is equal to one, since otherwise setting these elementsto be arbitrarily large in absolute value could result in an arbitrarily largevariance.Given ann×pdata setX, how do we compute the first principal com-ponent? Since we are only interested in variance, we assume that each ofthe variables inXhas been centered to have mean zero (that is, the col-umn means ofXare zero). We then look for the linear combination of thesample feature values of the formzi1=φ11xi1+φ21xi2+···+φp1xip(12.2)', metadata={'source': './documents/ISLP_website.pdf', 'page': 510}), Document(page_content='506 12. Unsupervised Learning\\nthathaslargestsamplevariance,subjecttotheconstraintthat∑p\\nj=1φ2\\nj1=1.\\nIn other words, the first principal component loading vector solves the op-\\ntimization problem\\nmaximize\\nφ11,...,φp1\\uf8f1\\n\\uf8f4\\uf8f2\\n\\uf8f4\\uf8f31\\nnn∑\\ni=1\\uf8eb\\n\\uf8edp∑\\nj=1φj1xij\\uf8f6\\n\\uf8f82\\uf8fc\\n\\uf8f4\\uf8fd\\n\\uf8f4\\uf8fesubject top∑\\nj=1φ2\\nj1=1. (12.3)\\nFrom (12.2) we can write the objective in ( 12.3) as1\\nn∑n\\ni=1z2\\ni1. Since\\n1\\nn∑n\\ni=1xij=0, the average of the z11,...,z n1will be zero as well. Hence\\nthe objective that we are maximizing in ( 12.3) is just the sample variance of\\nthenvalues of zi1. We refer to z11,...,z n1as thescoresof the first princi-score\\npal component. Problem ( 12.3) can be solved via an eigen decomposition ,eigen decom-\\npositiona standard technique in linear algebra, but the details are outside of the\\nscope of this book.1\\nThere is a nice geometric interpretation of the first principal component.\\nThe loading vector φ1with elements φ11,φ21,...,φp1defines a direction in\\nfeature space along which the data vary the most. If we project the ndata\\npointsx1,...,x nonto this direction, the projected values are the princi-\\npal component scores z11,...,z n1themselves. For instance, Figure 6.14on\\npage254displays the first principal component loading vector (green solid\\nline) on an advertising data set. In these data, there are only two features,\\nand so the observations as well as the first principal component loading\\nvector can be easily displayed. As can be seen from ( 6.19), in that data set\\nφ11=0.839andφ21=0.544.\\nAfter the first principal component Z1of the features has been deter-\\nmined, we can find the second principal component Z2. The second princi-\\npal component is the linear combination of X1,...,X pthat has maximal\\nvariance out of all linear combinations that are uncorrelated withZ1. The\\nsecond principal component scores z12,z22,...,z n2take the form\\nzi2=φ12xi1+φ22xi2+···+φp2xip, (12.4)\\nwhereφ2is the second principal component loading vector, with elements\\nφ12,φ22,...,φp2. It turns out that constraining Z2to be uncorrelated with\\nZ1is equivalent to constraining the direction φ2to be orthogonal (perpen-\\ndicular) to the direction φ1. In the example in Figure 6.14, the observations\\nlie in two-dimensional space (since p=2), and so once we have found φ1,\\nthere is only one possibility for φ2, which is shown as a blue dashed line.\\n(From Section 6.3.1, we know that φ12=0.544andφ22=−0.839.) But in\\na larger data set with p>2variables, there are multiple distinct principal\\ncomponents, and they are defined in a similar manner. To find φ2, we solve\\na problem similar to ( 12.3) withφ2replacing φ1, and with the additional\\nconstraint that φ2is orthogonal to φ1.2\\n1As an alternative to the eigen decomposition, a related technique called the singular\\nvalue decomposition can be used. This will be explored in the lab at the end of this\\nchapter.\\n2On a technical note, the principal component directions φ1,φ2,φ3,...are given\\nby the ordered sequence of eigenvectors of the matrix XTX, and the variances of the\\ncomponents are the eigenvalues. There are at most min( n−1,p)principal components.', metadata={'source': './documents/ISLP_website.pdf', 'page': 511}), Document(page_content='12.2 Principal Components Analysis 507\\n−3 −2 −1 0123−3 −2 −1 0123\\nFirst Principal ComponentSecond Principal ComponentAlabamaAlaskaArizona\\nArkansasCaliforniaColoradoConnecticutDelawareFloridaGeorgiaHawaii\\nIdahoIllinoisIndianaIowaKansasKentuckyLouisianaMaineMarylandMassachusettsMichiganMinnesota\\nMississippiMissouriMontanaNebraskaNevadaNew HampshireNew Jersey\\nNew MexicoNew York\\nNorth CarolinaNorth DakotaOhioOklahomaOregonPennsylvaniaRhode Island\\nSouth CarolinaSouth DakotaTennesseeTexasUtah\\nVermontVirginiaWashington\\nWest VirginiaWisconsinWyoming−0.50.00.5\\n−0.50.00.5\\nMurderAssaultUrbanPop\\nRape\\nFIGURE 12.1.The first two principal components for theUSArrestsdata. Theblue state names represent the scores for the first two principal components. Theorange arrows indicate the first two principal component loading vectors (with axeson the top and right). For example, the loading forRapeon the first componentis0.54, and its loading on the second principal component0.17(the wordRapeis centered at the point(0.54,0.17)). This figure is known as a biplot, because itdisplays both the principal component scores and the principal component loadings.Once we have computed the principal components, we can plot themagainst each other in order to produce low-dimensional views of the data.For instance, we can plot the score vectorZ1againstZ2,Z1againstZ3,Z2againstZ3, and so forth. Geometrically, this amounts to projecting theoriginal data down onto the subspace spanned byφ1,φ2, andφ3, andplotting the projected points.We illustrate the use of PCA on theUSArrestsdata set. For each of the50 states in the United States, the data set contains the number of arrestsper100,000residents for each of three crimes:Assault,Murder, andRape.We also recordUrbanPop(the percent of the population in each state livingin urban areas). The principal component score vectors have lengthn= 50,and the principal component loading vectors have lengthp=4. PCA wasperformedafterstandardizingeachvariabletohavemeanzeroandstandard', metadata={'source': './documents/ISLP_website.pdf', 'page': 512}), Document(page_content='508 12. Unsupervised Learning\\nPC1 PC2\\nMurder 0.5358995 −0.4181809\\nAssault 0.5831836 −0.1879856\\nUrbanPop 0.2781909 0.8728062\\nRape 0.5434321 0.1673186\\nTABLE 12.1. The principal component loading vectors, φ1andφ2, for the\\nUSArrests data. These are also displayed in Figure 12.1.\\ndeviation one. Figure 12.1plots the first two principal components of these\\ndata. The figure represents both the principal component scores and the\\nloading vectors in a single biplotdisplay. The loadings are also given inbiplotTable12.2.1.\\nIn Figure 12.1, we see that the first loading vector places approximately\\nequal weight on Assault,Murder, andRape, but with much less weight on\\nUrbanPop .Hencethiscomponentroughlycorrespondstoameasureofoverall\\nrates of serious crimes. The second loading vector places most of its weight\\nonUrbanPop and much less weight on the other three features. Hence, this\\ncomponent roughly corresponds to the level of urbanization of the state.\\nOverall, we see that the crime-related variables ( Murder,Assault, andRape)\\narelocatedclosetoeachother,andthatthe UrbanPop variableisfarfromthe\\nother three. This indicates that the crime-related variables are correlated\\nwith each other—states with high murder rates tend to have high assault\\nand rape rates—and that the UrbanPop variable is less correlated with the\\nother three.\\nWe can examine differences between the states via the two principal com-\\nponent score vectors shown in Figure 12.1. Our discussion of the loading\\nvectors suggests that states with large positive scores on the first compo-\\nnent, such as California, Nevada and Florida, have high crime rates, while\\nstates like North Dakota, with negative scores on the first component, have\\nlow crime rates. California also has a high score on the second component,\\nindicating a high level of urbanization, while the opposite is true for states\\nlike Mississippi. States close to zero on both components, such as Indiana,\\nhave approximately average levels of both crime and urbanization.\\n12.2.2 Another Interpretation of Principal Components\\nThe first two principal component loading vectors in a simulated three-\\ndimensional data set are shown in the left-hand panel of Figure 12.2; these\\ntwo loading vectors span a plane along which the observations have the\\nhighest variance.\\nIn the previous section, we describe the principal component loading vec-\\ntors as the directions in feature space along which the data vary the most,\\nand the principal component scores as projections along these directions.\\nHowever, an alternative interpretation of principal components can also be', metadata={'source': './documents/ISLP_website.pdf', 'page': 513}), Document(page_content='12.2 Principal Components Analysis 509\\nFirst principal componentSecond principal component−1.0 −0.5 0.0 0.5 1.0−1.0 −0.5 0.0 0.5 1.0••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••\\nFIGURE 12.2. Ninety observations simulated in three dimensions. The obser-\\nvations are displayed in color for ease of visualization. Left:the first two principal\\ncomponent directions span the plane that best fits the data. The plane is positioned\\nto minimize the sum of squared distances to each point. Right:the first two prin-\\ncipal component score vectors give the coordinates of the projection of the 90\\nobservations onto the plane.\\nuseful: principal components provide low-dimensional linear surfaces that\\nareclosestto the observations. We expand upon that interpretation here.3\\nThe first principal component loading vector has a very special property:\\nit is the line in p-dimensional space that is closestto thenobservations\\n(using average squared Euclidean distance as a measure of closeness). This\\ninterpretation can be seen in the left-hand panel of Figure 6.15; the dashed\\nlines indicate the distance between each observation and the line defined\\nby the first principal component loading vector. The appeal of this inter-\\npretation is clear: we seek a single dimension of the data that lies as close\\nas possible to all of the data points, since such a line will likely provide a\\ngood summary of the data.\\nThe notion of principal components as the dimensions that are clos-\\nest to the nobservations extends beyond just the first principal com-\\nponent. For instance, the first two principal components of a data set\\nspan the plane that is closest to the nobservations, in terms of average\\nsquared Euclidean distance. An example is shown in the left-hand panel\\nof Figure 12.2. The first three principal components of a data set span\\nthe three-dimensional hyperplane that is closest to the nobservations, and\\nso forth.\\nUsing this interpretation, together the first Mprincipal component score\\nvectors and the first Mprincipal component loading vectors provide the\\nbestM-dimensional approximation (in terms of Euclidean distance) to\\n3In this section, we continue to assume that each column of the data matrix Xhas\\nbeen centered to have mean zero—that is, the column mean has been subtracted from\\neach column.', metadata={'source': './documents/ISLP_website.pdf', 'page': 514}), Document(page_content='510 12. Unsupervised Learning\\ntheith observation xij. This representation can be written as\\nxij≈M∑\\nm=1zimφjm. (12.5)\\nWe can state this more formally by writing down an optimization prob-\\nlem. Suppose the data matrix Xis column-centered. Out of all approxima-\\ntions of the form xij≈∑M\\nm=1aimbjm, we could ask for the one with the\\nsmallest residual sum of squares:\\nminimize\\nA∈Rn×M,B∈Rp×M\\uf8f1\\n\\uf8f2\\n\\uf8f3p∑\\nj=1n∑\\ni=1(\\nxij−M∑\\nm=1aimbjm)2\\uf8fc\\n\\uf8fd\\n\\uf8fe. (12.6)\\nHere,Ais ann×Mmatrix whose (i, m)element is aim, andBis ap×M\\nelement whose (j, m)element is bjm.\\nIt can be shown that for any value of M, the columns of the matrices\\nˆAandˆBthat solve ( 12.6) are in fact the first Mprincipal components\\nscore and loading vectors. In other words, if ˆAandˆBsolve (12.6), then\\nˆaim=zimandˆbjm=φjm.4This means that the smallest possible value of\\nthe objective in ( 12.6) is\\np∑\\nj=1n∑\\ni=1(\\nxij−M∑\\nm=1zimφjm)2\\n. (12.7)\\nIn summary, together the Mprincipal component score vectors and M\\nprincipal component loading vectors can give a good approximation to the\\ndata when Mis sufficiently large. When M= min( n−1,p), then the\\nrepresentation is exact: xij=∑M\\nm=1zimφjm.\\n12.2.3 The Proportion of Variance Explained\\nIn Figure 12.2, we performed PCA on a three-dimensional data set (left-\\nhand panel) and projected the data onto the first two principal component\\nloading vectors in order to obtain a two-dimensional view of the data (i.e.\\nthe principal component score vectors; right-hand panel). We see that this\\ntwo-dimensional representation of the three-dimensional data does success-\\nfully capture the major pattern in the data: the orange, green, and cyan\\nobservations that are near each other in three-dimensional space remain\\nnearby in the two-dimensional representation. Similarly, we have seen on\\ntheUSArrests data set that we can summarize the 50 observations and 4\\nvariables using just the first two principal component score vectors and the\\nfirst two principal component loading vectors.\\nWe can now ask a natural question: how much of the information in\\na given data set is lost by projecting the observations onto the first few\\nprincipal components? That is, how much of the variance in the data is not\\ncontained in the first few principal components? More generally, we are\\ninterested in knowing the proportion of variance explained (PVE) by eachproportion\\nof variance\\nexplained 4Technically, the solution to ( 12.6) is not unique. Thus, it is more precise to state\\nthat any solution to ( 12.6) can be easily transformed to yield the principal components.', metadata={'source': './documents/ISLP_website.pdf', 'page': 515}), Document(page_content='12.2 Principal Components Analysis 511\\nprincipal component. The total variance present in a data set (assuming\\nthat the variables have been centered to have mean zero) is defined as\\np∑\\nj=1Var(Xj)=p∑\\nj=11\\nnn∑\\ni=1x2\\nij, (12.8)\\nand the variance explained by the mth principal component is\\n1\\nnn∑\\ni=1z2\\nim=1\\nnn∑\\ni=1\\uf8eb\\n\\uf8edp∑\\nj=1φjmxij\\uf8f6\\n\\uf8f82\\n. (12.9)\\nTherefore, the PVE of the mth principal component is given by\\n∑n\\ni=1z2\\nim∑p\\nj=1∑n\\ni=1x2\\nij=∑n\\ni=1(∑p\\nj=1φjmxij)2\\n∑p\\nj=1∑n\\ni=1x2\\nij. (12.10)\\nThe PVE of each principal component is a positive quantity. In order to\\ncompute the cumulative PVE of the first Mprincipal components, we can\\nsimply sum ( 12.10) over each of the first MPVEs. In total, there are\\nmin( n−1,p)principal components, and their PVEs sum to one.\\nIn Section 12.2.2, we showed that the first Mprincipal component load-\\ning and score vectors can be interpreted as the best M-dimensional approx-\\nimation to the data, in terms of residual sum of squares. It turns out that\\nthe variance of the data can be decomposed into the variance of the first M\\nprincipal components plus the mean squared error of this M-dimensional\\napproximation, as follows:\\np∑\\nj=11\\nnn∑\\ni=1x2\\nij\\n\\ued19\\ued18\\ued17 \\ued1a\\nVar. of data=M∑\\nm=11\\nnn∑\\ni=1z2\\nim\\n\\ued19\\ued18\\ued17 \\ued1a\\nVar. of first MPCs+1\\nnp∑\\nj=1n∑\\ni=1(\\nxij−M∑\\nm=1zimφjm)2\\n\\ued19 \\ued18\\ued17 \\ued1a\\nMSE of M-dimensional approximation(12.11)\\nThe three terms in this decomposition are discussed in ( 12.8), (12.9), and\\n(12.7), respectively. Since the first term is fixed, we see that by maximizing\\nthe variance of the first Mprincipal components, we minimize the mean\\nsquarederrorofthe M-dimensionalapproximation,andviceversa.Thisex-\\nplains why principal components can be equivalently viewed as minimizing\\nthe approximation error (as in Section 12.2.2) or maximizing the variance\\n(as in Section 12.2.1).\\nMoreover, we can use ( 12.11) to see that the PVE defined in ( 12.10)\\nequals\\n1−∑p\\nj=1∑n\\ni=1(\\nxij−∑M\\nm=1zimφjm)2\\n∑p\\nj=1∑n\\ni=1x2\\nij=1−RSS\\nTSS,\\nwhere TSS represents the total sum of squared elements of X, and RSS\\nrepresents the residual sum of squares of the M-dimensional approxima-\\ntion given by the principal components. Recalling the definition of R2from\\n(3.17), this means that we can interpret the PVE as the R2of the approx-\\nimation for Xgiven by the first Mprincipal components.', metadata={'source': './documents/ISLP_website.pdf', 'page': 516}), Document(page_content='512 12. Unsupervised Learning\\n1.0 1.5 2.0 2.5 3.0 3.5 4.00.0 0.2 0.4 0.6 0.8 1.0Principal ComponentProp. Variance Explained1.0 1.5 2.0 2.5 3.0 3.5 4.00.0 0.2 0.4 0.6 0.8 1.0Principal ComponentCumulative Prop. Variance ExplainedFIGURE 12.3. Left:a scree plot depicting the proportion of variance explained\\nby each of the four principal components in the USArrests data.Right:the cu-\\nmulative proportion of variance explained by the four principal components in the\\nUSArrests data.\\nIn theUSArrests data, the first principal component explains 62.0% of\\nthe variance in the data, and the next principal component explains 24.7%\\nofthevariance.Together,thefirsttwoprincipalcomponentsexplainalmost\\n87% of the variance in the data, and the last two principal components\\nexplain only 13% of the variance. This means that Figure 12.1provides a\\npretty accurate summary of the data using just two dimensions. The PVE\\nof each principal component, as well as the cumulative PVE, is shown\\nin Figure 12.3. The left-hand panel is known as a scree plot , and will bescree plotdiscussed later in this chapter.\\n12.2.4 More on PCA\\nScaling the Variables\\nWe have already mentioned that before PCA is performed, the variables\\nshould be centered to have mean zero. Furthermore, the results obtained\\nwhen we perform PCA will also depend on whether the variables have\\nbeen individually scaled (each multiplied by a different constant). This is in\\ncontrast to some other supervised and unsupervised learning techniques,\\nsuch as linear regression, in which scaling the variables has no effect. (In\\nlinear regression, multiplying a variable by a factor of cwill simply lead to\\nmultiplication of the corresponding coefficient estimate by a factor of 1/c,\\nand thus will have no substantive effect on the model obtained.)\\nFor instance, Figure 12.1was obtained after scaling each of the variables\\nto have standard deviation one. This is reproduced in the left-hand plot in\\nFigure12.4. Why does it matter that we scaled the variables? In these data,\\nthe variables are measured in different units; Murder,Rape, andAssaultare\\nreported as the number of occurrences per 100,000people, and UrbanPop is\\nthe percentage of the state’s population that lives in an urban area. These\\nfour variables have variances of 18.97,87.73,6945 .16, and209.5, respec-\\ntively. Consequently, if we perform PCA on the unscaled variables, then', metadata={'source': './documents/ISLP_website.pdf', 'page': 517}), Document(page_content='12.2 Principal Components Analysis 513\\n−3 −2 −1 0 1 2 3−3 −2 −1 0 1 2 3First Principal ComponentSecond Principal Component***************************************\\n***********−0.5 0.0 0.5\\n−0.5 0.0 0.5MurderAssaultUrbanPopRapeScaled\\n−100 −50 0 50 100 150−100 −50 0 50 100 150First Principal ComponentSecond Principal Component**************************************************−0.5 0.0 0.5 1.0\\n−0.5 0.0 0.5 1.0MurderAssaultUrbanPopRapeUnscaled\\nFIGURE 12.4. Two principal component biplots for the USArrests data.Left:\\nthe same as Figure 12.1, with the variables scaled to have unit standard deviations.\\nRight:principal components using unscaled data. Assault has by far the largest\\nloading on the first principal component because it has the highest variance among\\nthe four variables. In general, scaling the variables to have standard deviation one\\nis recommended.\\nthe first principal component loading vector will have a very large loading\\nforAssault, since that variable has by far the highest variance. The right-\\nhand plot in Figure 12.4displays the first two principal components for the\\nUSArrests data set, without scaling the variables to have standard devia-\\ntion one. As predicted, the first principal component loading vector places\\nalmost all of its weight on Assault, while the second principal component\\nloading vector places almost all of its weight on UrbanPop . Comparing this\\nto the left-hand plot, we see that scaling does indeed have a substantial\\neffect on the results obtained.\\nHowever, this result is simply a consequence of the scales on which the\\nvariables were measured. For instance, if Assaultwere measured in units\\nof the number of occurrences per 100people (rather than number of oc-\\ncurrences per 100,000people), then this would amount to dividing all of\\nthe elements of that variable by 1,000. Then the variance of the variable\\nwould be tiny, and so the first principal component loading vector would\\nhave a very small value for that variable. Because it is undesirable for the\\nprincipal components obtained to depend on an arbitrary choice of scaling,\\nwe typically scale each variable to have standard deviation one before we\\nperform PCA.\\nIn certain settings, however, the variables may be measured in the same\\nunits. In this case, we might not wish to scale the variables to have stan-\\ndard deviation one before performing PCA. For instance, suppose that the\\nvariables in a given data set correspond to expression levels for pgenes.\\nThen since expression is measured in the same “units” for each gene, we\\nmight choose not to scale the genes to each have standard deviation one.', metadata={'source': './documents/ISLP_website.pdf', 'page': 518}), Document(page_content='514 12. Unsupervised Learning\\nUniqueness of the Principal Components\\nWhile in theory the principal components need not be unique, in almost all\\npractical settings they are (up to sign flips). This means that two different\\nsoftware packages will yield the same principal component loading vectors,\\nalthough the signs of those loading vectors may differ. The signs may differ\\nbecause each principal component loading vector specifies a direction in p-\\ndimensional space: flipping the sign has no effect as the direction does not\\nchange. (Consider Figure 6.14—the principal component loading vector is\\na line that extends in either direction, and flipping its sign would have no\\neffect.) Similarly, the score vectors are unique up to a sign flip, since the\\nvariance of Zis the same as the variance of −Z. It is worth noting that\\nwhen we use ( 12.5) to approximate xijwe multiply zimbyφjm. Hence, if\\nthe sign is flipped on both the loading and score vectors, the final product\\nof the two quantities is unchanged.\\nDeciding How Many Principal Components to Use\\nIn general, an n×pdata matrix Xhasmin( n−1,p)distinct principal\\ncomponents. However, we usually are not interested in all of them; rather,\\nwe would like to use just the first few principal components in order to\\nvisualize or interpret the data. In fact, we would like to use the smallest\\nnumberofprincipalcomponentsrequiredtogeta goodunderstandingofthe\\ndata. How many principal components are needed? Unfortunately, there is\\nno single (or simple!) answer to this question.\\nWe typically decide on the number of principal components required\\nto visualize the data by examining a scree plot , such as the one shown\\nin the left-hand panel of Figure 12.3. We choose the smallest number of\\nprincipal components that are required in order to explain a sizable amount\\nof the variation in the data. This is done by eyeballing the scree plot, and\\nlooking for a point at which the proportion of variance explained by each\\nsubsequent principal component drops off. This drop is often referred to\\nas anelbowin the scree plot. For instance, by inspection of Figure 12.3,\\none might conclude that a fair amount of variance is explained by the first\\ntwo principal components, and that there is an elbow after the second\\ncomponent. After all, the third principal component explains less than ten\\npercent of the variance in the data, and the fourth principal component\\nexplains less than half that and so is essentially worthless.\\nHowever, this type of visual analysis is inherently ad hoc. Unfortunately,\\nthere is no well-accepted objective way to decide how many principal com-\\nponents are enough. In fact, the question of how many principal compo-\\nnents are enough is inherently ill-defined, and will depend on the specific\\narea of application and the specific data set. In practice, we tend to look\\nat the first few principal components in order to find interesting patterns\\nin the data. If no interesting patterns are found in the first few principal\\ncomponents, then further principal components are unlikely to be of inter-\\nest. Conversely, if the first few principal components are interesting, then\\nwe typically continue to look at subsequent principal components until no\\nfurther interesting patterns are found. This is admittedly a subjective ap-', metadata={'source': './documents/ISLP_website.pdf', 'page': 519}), Document(page_content='12.3 Missing Values and Matrix Completion 515\\nproach, and is reflective of the fact that PCA is generally used as a tool for\\nexploratory data analysis.\\nOn the other hand, if we compute principal components for use in a\\nsupervised analysis, such as the principal components regression presented\\nin Section 6.3.1, then there is a simple and objective way to determine how\\nmany principal components to use: we can treat the number of principal\\ncomponent score vectors to be used in the regression as a tuning parameter\\nto be selected via cross-validation or a related approach. The comparative\\nsimplicity of selecting the number of principal components for a supervised\\nanalysis is one manifestation of the fact that supervised analyses tend to\\nbe more clearly defined and more objectively evaluated than unsupervised\\nanalyses.\\n12.2.5 Other Uses for Principal Components\\nWe saw in Section 6.3.1that we can perform regression using the principal\\ncomponent score vectors as features. In fact, many statistical techniques,\\nsuch as regression, classification, and clustering, can be easily adapted to\\nuse then×Mmatrix whose columns are the first M≪pprincipal com-\\nponent score vectors, rather than using the full n×pdata matrix. This\\ncan lead to less noisy results, since it is often the case that the signal (as\\nopposed to the noise) in a data set is concentrated in its first few principal\\ncomponents.\\n12.3 Missing Values and Matrix Completion\\nOften datasets have missing values, which can be a nuisance. For example,\\nsuppose that we wish to analyze the USArrests data, and discover that 20\\nof the 200 values have been randomly corrupted and marked as missing.\\nUnfortunately, the statistical learning methods that we have seen in this\\nbook cannot handle missing values. How should we proceed?\\nWe could remove the rows that contain missing observations and per-\\nform our data analysis on the complete rows. But this seems wasteful, and\\ndepending on the fraction missing, unrealistic. Alternatively, if xijis miss-\\ning, then we could replace it by the mean of the jth column (using the\\nnon-missing entries to compute the mean). Although this is a common and\\nconvenient strategy, often we can do better by exploiting the correlation\\nbetween the variables.\\nIn this section we show how principal components can be used to imputeimpute\\nimputationthe missing values, through a process known as matrix completion . The\\nmatrix\\ncompletioncompleted matrix can then be used in a statistical learning method, such\\nas linear regression or LDA.\\nThisapproachforimputingmissingdataisappropriateifthemissingness\\nisrandom.Forexample,itissuitableifapatient’sweightismissingbecausemissing at\\nrandomthe battery of the electronic scale was flat at the time of his exam. By\\ncontrast, if the weight is missing because the patient was too heavy to\\nclimb on the scale, then this is not missing at random; the missingness is', metadata={'source': './documents/ISLP_website.pdf', 'page': 520}), Document(page_content='516 12. Unsupervised Learning\\ninformative, and the approach described here for handling missing data is\\nnot suitable.\\nSometimes data is missing by necessity. For example, if we form a matrix\\nof the ratings (on a scale from 1to5) thatncustomers have given to the\\nentire Netflix catalog of pmovies, then most of the matrix will be missing,\\nsince no customer will have seen and rated more than a tiny fraction of the\\ncatalog. If we can impute the missing values well, then we will have an idea\\nof what each customer will think of movies they have not yet seen. Hence\\nmatrix completion can be used to power recommender systems .recommender\\nsystems\\nPrincipal Components with Missing Values\\nIn Section 12.2.2, we showed that the first Mprincipal component score\\nand loading vectors provide the “best” approximation to the data matrix\\nX, in the sense of ( 12.6). Suppose that some of the observations xijare\\nmissing. We now show how one can both impute the missing values and\\nsolve the principal component problem at the same time. We return to a\\nmodified form of the optimization problem ( 12.6),\\nminimize\\nA∈Rn×M,B∈Rp×M\\uf8f1\\n\\uf8f2\\n\\uf8f3∑\\n(i,j)∈O(\\nxij−M∑\\nm=1aimbjm)2\\uf8fc\\n\\uf8fd\\n\\uf8fe, (12.12)\\nwhereOis the set of all observed pairs of indices (i, j), a subset of the\\npossible n×ppairs.\\nOnce we solve this problem:\\n• we can estimate a missing observation xijusingˆxij=∑M\\nm=1ˆaimˆbjm,\\nwhereˆaimandˆbjmare the(i, m)and(j, m)elements, respectively,\\nof the matrices ˆAandˆBthat solve ( 12.12); and\\n• we can (approximately) recover the Mprincipal component scores\\nand loadings, as we did when the data were complete.\\nIt turns out that solving ( 12.12) exactly is difficult, unlike in the case of\\ncomplete data: the eigen decomposition no longer applies. But the sim-\\nple iterative approach in Algorithm 12.1, which is demonstrated in Sec-\\ntion12.5.2, typically provides a good solution.56\\nWe illustrate Algorithm 12.1on theUSArrests data. There are p=4\\nvariables and n= 50 observations (states). We first standardized the data\\nso each variable has mean zero and standard deviation one. We then ran-\\ndomly selected 20 of the 50 states, and then for each of these we randomly\\nset one of the four variables to be missing. Thus, 10% of the elements of the\\ndata matrix were missing. We applied Algorithm 12.1withM=1principal\\ncomponent. Figure 12.5shows that the recovery of the missing elements\\n5This algorithm is referred to as “Hard-Impute” in Mazumder, Hastie, and Tibshi-\\nrani (2010) “Spectral regularization algorithms for learning large incomplete matrices”,\\npublished in Journal of Machine Learning Research , pages 2287–2322.\\n6Each iteration of Step 2 of this algorithm decreases the objective ( 12.14). However,\\nthe algorithm is not guaranteed to achieve the global optimum of ( 12.12).', metadata={'source': './documents/ISLP_website.pdf', 'page': 521}), Document(page_content='12.3 Missing Values and Matrix Completion 517\\nAlgorithm 12.1 Iterative Algorithm for Matrix Completion\\n1. Create a complete data matrix ˜Xof dimension n×pof which the\\n(i, j)element equals\\n˜xij={xijif(i, j)∈O\\n¯xjif(i, j)/∈O,\\nwhere¯xjis the average of the observed values for the jth variable in\\nthe incomplete data matrix X. Here,Oindexes the observations that\\nare observed in X.\\n2. Repeat steps (a)–(c) until the objective ( 12.14) fails to decrease:\\n(a) Solve\\nminimize\\nA∈Rn×M,B∈Rp×M\\uf8f1\\n\\uf8f2\\n\\uf8f3p∑\\nj=1n∑\\ni=1(\\n˜xij−M∑\\nm=1aimbjm)2\\uf8fc\\n\\uf8fd\\n\\uf8fe(12.13)\\nby computing the principal components of ˜X.\\n(b) For each element (i, j)/∈O, set˜xij←∑M\\nm=1ˆaimˆbjm.\\n(c) Compute the objective\\n∑\\n(i,j)∈O(\\nxij−M∑\\nm=1ˆaimˆbjm)2\\n. (12.14)\\n3. Return the estimated missing entries ˜xij,(i, j)/∈O.\\nis pretty accurate. Over 100 random runs of this experiment, the average\\ncorrelation between the true and imputed values of the missing elements\\nis 0.63, with a standard deviation of 0.11. Is this good performance? To\\nanswer this question, we can compare this correlation to what we would\\nhave gotten if we had estimated these 20 values using the complete data\\n— that is, if we had simply computed ˆxij=zi1φj1, wherezi1andφj1are\\nelements of the first principal component score and loading vectors of the\\ncomplete data.7Using the complete data in this way results in an average\\ncorrelation of 0.79 between the true and estimated values for these 20 el-\\nements, with a standard deviation of 0.08. Thus, our imputation method\\ndoes worse than the method that uses all of the data ( 0.63±0.11versus\\n0.79±0.08), but its performance is still pretty good. (And of course, the\\nmethod that uses all of the data cannot be applied in a real-world setting\\nwith missing data.)\\nFigure12.6further indicates that Algorithm 12.1performs fairly well on\\nthis dataset.\\n7This is an unattainable gold standard, in the sense that with missing data, we of\\ncourse cannot compute the principal components of the complete data.', metadata={'source': './documents/ISLP_website.pdf', 'page': 522}), Document(page_content='518 12. Unsupervised Learning\\nFIGURE 12.5. Missing value imputation on the USArrests data. Twenty values\\n(10% of the total number of matrix elements) were artificially set to be missing,\\nand then imputed via Algorithm 12.1withM=1. The figure displays the true\\nvaluexijand the imputed value ˆxijfor all twenty missing values. For each of the\\ntwenty missing values, the color indicates the variable, and the label indicates the\\nstate. The correlation between the true and imputed values is around 0.63.\\nWe close with a few observations:\\n• TheUSArrests data has only four variables, which is on the low end\\nfor methods like Algorithm 12.1to work well. For this reason, for this\\ndemonstration we randomly set at most one variable per state to be\\nmissing, and only used M=1principal component.\\n• In general, in order to apply Algorithm 12.1, we must select M, the\\nnumber of principal components to use for the imputation. One ap-\\nproach is to randomly leave out a few additional elements from the\\nmatrix, and select Mbased on how well those known values are re-\\ncovered. This is closely related to the validation-set approach seen in\\nChapter 5.\\nRecommender Systems\\nDigital streaming services like Netflix and Amazon use data about the con-\\ntent that a customer has viewed in the past, as well as data from other\\ncustomers, to suggest other content for the customer. As a concrete ex-\\nample, some years back, Netflix had customers rate each movie that they\\nhad seen with a score from 1–5. This resulted in a very big n×pmatrix\\nfor which the (i, j)element is the rating given by the ith customer to the−1.5 −1.0 −0.5 0.0 0.5 1.0 1.5−1.5 −1.0 −0.5 0.0 0.5 1.0 1.5\\nOriginal ValueImputed Value/uni25CF/uni25CF/uni25CF\\n/uni25CF/uni25CF/uni25CF\\n/uni25CF\\n/uni25CF/uni25CF\\n/uni25CF/uni25CF/uni25CF\\n/uni25CF/uni25CF/uni25CF\\n/uni25CF/uni25CF\\n/uni25CF/uni25CF/uni25CFORWATN\\nPAMOIDMD\\nVAALTX\\nWYMNAK\\nUTGA\\nMTMANY\\n/uni25CF/uni25CF/uni25CF/uni25CFMurderAssaultUrbanPopRape', metadata={'source': './documents/ISLP_website.pdf', 'page': 523}), Document(page_content='12.3 Missing Values and Matrix Completion 519\\nFIGURE 12.6. As described in the text, in each of 100 trials, we left out 20\\nelements of the USArrests dataset. In each trial, we applied Algorithm 12.1with\\nM=1 to impute the missing elements and compute the principal com ponents.\\nLeft:For each of the 50 states, the imputed first principal component scores\\n(averaged over 100 trials, and displayed with a standard deviation bar) are plotted\\nagainst the first principal component scores computed using all the data. Right:\\nThe imputed principal component loadings (averaged over 100 trials, and displayed\\nwith a standard deviation bar) are plotted against the true principal component\\nloadings.\\njth movie. One specific early example of this matrix had n= 480 ,189cus-\\ntomersand p= 17 ,770movies.However,onaverageeachcustomerhadseen\\naround 200 movies, so 99% of the matrix had missing elements. Table 12.2\\nillustrates the setup.\\nIn order to suggest a movie that a particular customer might like, Netflix\\nneededawaytoimputethemissingvaluesofthisdatamatrix.Thekeyidea\\nis as follows: the set of movies that the ith customer has seen will overlap\\nwith those that other customers have seen. Furthermore, some of those\\nother customers will have similar movie preferences to the ith customer.\\nThus, it should be possible to use similar customers’ ratings of movies that\\ntheith customer has not seen to predict whether the ith customer will like\\nthose movies.\\nMore concretely, by applying Algorithm 12.1, we can predict the ith cus-\\ntomer’s rating for the jth movie using ˆxij=∑M\\nm=1ˆaimˆbjm. Furthermore,\\nwe can interpret the Mcomponents in terms of “cliques” and “genres”:\\n•ˆaimrepresents the strength with which the ith user belongs to the\\nmth clique, where a cliqueis a group of customers that enjoys movies\\nof themth genre;\\n•ˆbjmrepresents the strength with which the jth movie belongs to the\\nmthgenre.\\nExamples of genres include Romance, Western, and Action.\\nPrincipal component models similar to Algorithm 12.1are at the heart\\nof many recommender systems. Although the data m atrices involved are−3−2−10 1 2 3−3−2−10 1 2 3\\nTrue First Principal ComponentImputed First Principal Component/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF/uni25CF\\n2468 1 0 1 22468 1 0 1 2True PC VariancesImputed PC Variances/uni25CF\\n/uni25CF/uni25CF/uni25CF', metadata={'source': './documents/ISLP_website.pdf', 'page': 524}), Document(page_content='520 12. Unsupervised LearningJerry MaguireOceansRoad to PerditionA Fortunate ManCatch Me If You CanDriving Miss DaisyThe Two PopesThe LaundromatCode 8The Social Network···Customer 1••••4••••• · · ·Customer 2••3•••3••3···Customer 3•2•4••••2•· · ·Customer 43••••••••• · · ·Customer 551••4••••• · · ·Customer 6•••••24••• · · ·Customer 7••5••••3•• · · ·Customer 8•••••••••• · · ·Customer 93•••5••1•• · · ·....................................TABLE 12.2.Excerpt of the Netflix movie rating data. The movies are ratedfrom 1 (worst) to 5 (best). The symbol•represents a missing value: a movie thatwas not rated by the corresponding customer.typically massive, algorithms have been developed that can exploit the highlevel of missingness in order to perform efficient computations.12.4 Clustering MethodsClusteringrefers to a very broad set of techniques for findingsubgroups, orclusteringclusters, in a data set. When we cluster the observations of a data set, weseek to partition them into distinct groups so that the observations withineach group are quite similar to each other, while observations in differentgroups are quite different from each other. Of course, to make this concrete,we must define what it means for two or more observations to besimilarordifferent. Indeed, this is often a domain-specific consideration that mustbe made based on knowledge of the data being studied.For instance, suppose that we have a set ofnobservations, each withpfeatures.Thenobservationscouldcorrespondtotissuesamplesforpatientswith breast cancer, and thepfeatures could correspond to measurementscollected for each tissue sample; these could be clinical measurements, suchas tumor stage or grade, or they could be gene expression measurements.We may have a reason to believe that there is some heterogeneity amongthentissue samples; for instance, perhaps there are a few differentun-knownsubtypes of breast cancer. Clustering could be used to find thesesubgroups. This is an unsupervised problem because we are trying to dis-cover structure—in this case, distinct clusters—on the basis of a data set.The goal in supervised problems, on the other hand, is to try to predictsome outcome vector such as survival time or response to drug treatment.Both clustering and PCA seek to simplify the data via a small numberof summaries, but their mechanisms are different:', metadata={'source': './documents/ISLP_website.pdf', 'page': 525}), Document(page_content='12.4 Clustering Methods 521\\n• PCA looks to find a low-dimensional representation of the observa-\\ntions that explain a good fraction of the variance;\\n• Clustering looks to find homogeneous subgroups among the observa-\\ntions.\\nAnother application of clustering arises in marketing. We may have ac-\\ncess to a large number of measurements (e.g. median household income,\\noccupation, distance from nearest urban area, and so forth) for a large\\nnumber of people. Our goal is to perform market segmentation by identify-\\ning subgroups of people who might be more receptive to a particular form\\nof advertising, or more likely to purchase a particular product. The task of\\nperforming market segmentation amounts to clustering the people in the\\ndata set.\\nSince clustering is popular in many fields, there exist a great num-\\nber of clustering methods. In this section we focus on perhaps the two\\nbest-known clustering approaches: K-means clustering andhierarchicalK-means\\nclusteringclustering . InK-means clustering, we seek to partition the observations\\nhierarchical\\nclusteringinto a pre-specified number of clusters. On the other hand, in hierarchical\\nclustering, we do not know in advance how many clusters we want; in fact,\\nwe end up with a tree-like visual representation of the observations, called\\nadendrogram , that allows us to view at once the clusterings obtained fordendrogrameach possible number of clusters, from 1ton. There are advantages and\\ndisadvantages to each of these clustering approaches, which we highlight in\\nthis chapter.\\nIn general, we can cluster observations on the basis of the features in\\norder to identify subgroups among the observations, or we can cluster fea-\\ntures on the basis of the observations in order to discover subgroups among\\nthe features. In what follows, for simplicity we will discuss clustering obser-\\nvations on the basis of the features, though the converse can be performed\\nby simply transposing the data matrix.\\n12.4.1K-Means Clustering\\nK-means clustering is a simple and elegant approach for partitioning a\\ndata set into Kdistinct, non-overlapping clusters. To perform K-means\\nclustering, we must first specify the desired number of clusters K; then the\\nK-means algorithm will assign each observation to exactly one of the K\\nclusters. Figure 12.7shows the results obtained from performing K-means\\nclustering on a simulated example consisting of 150observations in two\\ndimensions, using three different values of K.\\nTheK-means clustering procedure results from a simple and intuitive\\nmathematicalproblem.Webeginbydefiningsomenotation.Let C1,...,C K\\ndenote sets containing the indices of the observations in each cluster. These\\nsets satisfy two properties:\\n1.C1∪C2∪···∪CK={1,...,n }. In other words, each observation\\nbelongs to at least one of the Kclusters.\\n2.Ck∩Ck′=∅for allk̸=k′. In other words, the clusters are non-\\noverlapping: no observation belongs to more than one cluster.', metadata={'source': './documents/ISLP_website.pdf', 'page': 526}), Document(page_content='522 12. Unsupervised LearningK=2K=3K=4\\nFIGURE 12.7. A simulated data set with 150 observations in two-dimensional\\nspace. Panels show the results of applying K-means clustering with different values\\nofK, the number of clusters. The color of each observation indicates the cluster\\nto which it was assigned using the K-means clustering algorithm. Note that there\\nis no ordering of the clusters, so the cluster coloring is arbitrary. These cluster\\nlabels were not used in clustering; instead, they are the outputs of the clustering\\nprocedure.\\nFor instance, if the ith observation is in the kth cluster, then i∈Ck. The\\nideabehind K-meansclusteringisthata goodclusteringisoneforwhichthe\\nwithin-cluster variation is as small as possible. The within-cluster variation\\nfor cluster Ckis a measure W(Ck)of the amount by which the observations\\nwithin a cluster differ from each other. Hence we want to solve the problem\\nminimize\\nC1,...,C K{K∑\\nk=1W(Ck)}\\n. (12.15)\\nIn words, this formula says that we want to partition the observations into\\nKclusters such that the total within-cluster variation, summed over all K\\nclusters, is as small as possible.\\nSolving ( 12.15) seems like a reasonable idea, but in order to make it\\nactionable we need to define the within-cluster variation. There are many\\npossible ways to define this concept, but by far the most common choice\\ninvolves squared Euclidean distance . That is, we define\\nW(Ck)=1\\n|Ck|∑\\ni,i′∈Ckp∑\\nj=1(xij−xi′j)2, (12.16)\\nwhere|Ck|denotes the number of observations in the kth cluster. In other\\nwords, the within-cluster variation for the kth cluster is the sum of all of\\nthe pairwise squared Euclidean distances between the observations in the\\nkth cluster, divided by the total number of observations in the kth cluster.\\nCombining ( 12.15) and (12.16) gives the optimization problem that defines', metadata={'source': './documents/ISLP_website.pdf', 'page': 527}), Document(page_content='12.4 Clustering Methods 523\\nK-means clustering,\\nminimize\\nC1,...,C K\\uf8f1\\n\\uf8f2\\n\\uf8f3K∑\\nk=11\\n|Ck|∑\\ni,i′∈Ckp∑\\nj=1(xij−xi′j)2\\uf8fc\\n\\uf8fd\\n\\uf8fe. (12.17)\\nNow, we would like to find an algorithm to solve ( 12.17)—that is, a\\nmethodtopartitiontheobservationsinto Kclusterssuchthattheobjective\\nof (12.17) is minimized. This is in fact a very difficult problem to solve\\nprecisely,since there are almost Knwaysto partition nobservations into K\\nclusters. This is a huge number unless Kandnare tiny! Fortunately, a very\\nsimple algorithm can be shown to provide a local optimum—a pretty good\\nsolution—to the K-means optimization problem ( 12.17). This approach is\\nlaid out in Algorithm 12.2.\\nAlgorithm 12.2 K-Means Clustering\\n1. Randomly assign a number, from 1toK, to each of the observations.\\nThese serve as initial cluster assignments for the observations.\\n2. Iterate until the cluster assignments stop changing:\\n(a) For each of the Kclusters, compute the cluster centroid. The\\nkth cluster centroid is the vector of the pfeature means for the\\nobservations in the kth cluster.\\n(b) Assign each observation to the cluster whose centroid is closest\\n(whereclosestis defined using Euclidean distance).\\nAlgorithm 12.2isguaranteedtodecreasethevalueoftheobjective( 12.17)\\nat each step. To understand why, the following identity is illuminating:\\n1\\n|Ck|∑\\ni,i′∈Ckp∑\\nj=1(xij−xi′j)2=2∑\\ni∈Ckp∑\\nj=1(xij−¯xkj)2, (12.18)\\nwhere¯xkj=1\\n|Ck|∑\\ni∈Ckxijis the mean for feature jin cluster Ck.\\nIn Step 2(a) the cluster means for each feature are the constants that\\nminimize the sum-of-squared deviations, and in Step 2(b), reallocating the\\nobservations can only improve ( 12.18). This means that as the algorithm\\nis run, the clustering obtained will continually improve until the result no\\nlonger changes; the objective of ( 12.17) will never increase. When the result\\nno longer changes, a local optimum has been reached. Figure 12.8shows\\nthe progression of the algorithm on the toy example from Figure 12.7.\\nK-means clustering derives its name from the fact that in Step 2(a), the\\ncluster centroids are computed as the mean of the observations assigned to\\neach cluster.\\nBecause the K-means algorithm finds a local rather than a global opti-\\nmum, the results obtained will depend on the initial (random) cluster as-\\nsignment of each observation in Step 1 of Algorithm 12.2. For this reason,\\nit is important to run the algorithm multiple times from different random', metadata={'source': './documents/ISLP_website.pdf', 'page': 528}), Document(page_content='524 12. Unsupervised LearningDataStep 1Iteration 1, Step 2a\\nIteration 1, Step 2bIteration 2, Step 2aFinal Results\\nFIGURE 12.8. The progress of the K-means algorithm on the example of\\nFigure12.7withK=3.Top left: the observations are shown. Top center: in\\nStep 1 of the algorithm, each observation is randomly assigned to a cluster. Top\\nright:in Step 2(a), the cluster centroids are computed. These are shown as large\\ncolored disks. Initially the centroids are almost completely overlapping because\\nthe initial cluster assignments were chosen at random. Bottom left: in Step 2(b),\\neach observation is assigned to the nearest centroid. Bottom center: Step 2(a) is\\nonce again performed, leading to new cluster centroids. Bottom right: the results\\nobtained after ten iterations.\\ninitial configurations. Then one selects the bestsolution, i.e. that for which\\nthe objective ( 12.17) is smallest. Figure 12.9shows the local optima ob-\\ntained by running K-means clustering six times using six different initial\\ncluster assignments, using the toy data from Figure 12.7. In this case, the\\nbest clustering is the one with an objective value of 235.8.\\nAs we have seen, to perform K-means clustering, we must decide how\\nmany clusters we expect in the data. The problem of selecting Kis far from\\nsimple. This issue, along with other practical considerations that arise in\\nperforming K-means clustering, is addressed in Section 12.4.3.', metadata={'source': './documents/ISLP_website.pdf', 'page': 529}), Document(page_content='12.4 Clustering Methods 525320.9235.8235.8\\n235.8235.8310.9\\nFIGURE 12.9. K-means clustering performed six times on the data from\\nFigure12.7withK=3, each time with a different random assignment of the\\nobservations in Step 1 of the K-means algorithm. Above each plot is the value\\nof the objective ( 12.17). Three different local optima were obtained, one of which\\nresulted in a smaller value of the objective and provides better separation between\\nthe clusters. Those labeled in red all achieved the same best solution, with an\\nobjective value of 235.8.\\n12.4.2 Hierarchical Clustering\\nOne potential disadvantage of K-means clustering is that it requires us to\\npre-specify the number of clusters K.Hierarchical clustering is an alter-\\nnative approach which does not require that we commit to a particular\\nchoice of K. Hierarchical clustering has an added advantage over K-means\\nclustering in that it results in an attractive tree-based representation of the\\nobservations, called a dendrogram .\\nIn this section, we describe bottom-up oragglomerative clustering.bottom-up\\nagglomerativeThis is the most common type of hierarchical clustering, and refers to\\nthe fact that a dendrogram (generally depicted as an upside-down tree; see\\nFigure12.11) is built starting from the leaves and combining clusters up to\\nthetrunk.Wewillbeginwithadiscussionofhowtointerpretadendrogram', metadata={'source': './documents/ISLP_website.pdf', 'page': 530}), Document(page_content='526 12. Unsupervised Learning\\n−6 −4 −202−2024\\nX1X2\\nFIGURE 12.10. Forty-five observations generated in two-dimensional space.\\nIn reality there are three distinct classes, shown in separate colors. However, we\\nwill treat these class labels as unknown and will seek to cluster the observations\\nin order to discover the classes from the data.\\nand then discuss how hierarchical clustering is actually performed—that is,\\nhow the dendrogram is built.\\nInterpreting a Dendrogram\\nWe begin with the simulated data set shown in Figure 12.10, consisting of\\n45observations in two-dimensional space. The data were generated from a\\nthree-class model; the true class labels for each observation are shown in\\ndistinct colors. However, suppose that the data were observed without the\\nclass labels, and that we wanted to perform hierarchical clustering of the\\ndata. Hierarchical clustering (with complete linkage, to be discussed later)\\nyields the result shown in the left-hand panel of Figure 12.11. How can we\\ninterpret this dendrogram?\\nIn the left-hand panel of Figure 12.11, eachleafof the dendrogram rep-\\nresents one of the 45observations in Figure 12.10. However, as we move\\nup the tree, some leaves begin to fuseinto branches. These correspond to\\nobservations that are similar to each other. As we move higher up the tree,\\nbranches themselves fuse, either with leaves or other branches. The earlier\\n(lower in the tree) fusions occur, the more similar the groups of observa-\\ntions are to each other. On the other hand, observations that fuse later\\n(near the top of the tree) can be quite different. In fact, this statement\\ncan be made precise: for any two observations, we can look for the point in\\nthe tree where branches containing those two observations are first fused.\\nThe height of this fusion, as measured on the vertical axis, indicates how\\ndifferent the two observations are. Thus, observations that fuse at the very\\nbottom of the tree are quite similar to each other, whereas observations\\nthat fuse close to the top of the tree will tend to be quite different.\\nThis highlights a very important point in interpreting dendrograms that\\nis often misunderstood. Consider the left-hand panel of Figure 12.12, which\\nshows a simple dendrogram obtained from hierarchically clustering nine', metadata={'source': './documents/ISLP_website.pdf', 'page': 531}), Document(page_content='12.4 Clustering Methods 5270246810\\n0246810\\n0246810\\nFIGURE 12.11. Left:dendrogram obtained from hierarchically clustering the\\ndata from Figure 12.10with complete linkage and Euclidean distance. Center:the\\ndendrogram from the left-hand panel, cut at a height of nine (indicated by the\\ndashed line). This cut results in two distinct clusters, shown in different colors.\\nRight:the dendrogram from the left-hand panel, now cut at a height of five. This\\ncut results in three distinct clusters, shown in different colors. Note that the colors\\nwere not used in clustering, but are simply used for display purposes in this figure.\\nobservations. One can see that observations 5and7are quite similar to\\neach other, since they fuse at the lowest point on the dendrogram. Obser-\\nvations1and6are also quite similar to each other. However, it is tempting\\nbut incorrect to conclude from the figure that observations 9and2are\\nquite similar to each other on the basis that they are located near each\\nother on the dendrogram. In fact, based on the information contained in\\nthe dendrogram, observation 9is no more similar to observation 2than it\\nis to observations 8,5,and7. (This can be seen from the right-hand panel\\nof Figure 12.12, in which the raw data are displayed.) To put it mathe-\\nmatically, there are 2n−1possible reorderings of the dendrogram, where n\\nis the number of leaves. This is because at each of the n−1points where\\nfusions occur, the positions of the two fused branches could be swapped\\nwithout affecting the meaning of the dendrogram. Therefore, we cannot\\ndraw conclusions about the similarity of two observations based on their\\nproximity along the horizontal axis . Rather, we draw conclusions about\\nthe similarity of two observations based on the location on the vertical axis\\nwhere branches containing those two observations first are fused.\\nNow that we understand how to interpret the left-hand panel of Fig-\\nure12.11, we can move on to the issue of identifying clusters on the basis\\nof a dendrogram. In order to do this, we make a horizontal cut across the\\ndendrogram, as shown in the center and right-hand panels of Figure 12.11.\\nThe distinct sets of observations beneath the cut can be interpreted as clus-\\nters.InthecenterpanelofFigure 12.11,cuttingthedendrogramataheight\\nof nine results in two clusters, shown in distinct colors. In the right-hand\\npanel, cutting the dendrogram at a height of five results in three clusters.\\nFurther cuts can be made as one descends the dendrogram in order to ob-\\ntain any number of clusters, between 1 (corresponding to no cut) and n', metadata={'source': './documents/ISLP_website.pdf', 'page': 532}), Document(page_content='528 12. Unsupervised Learning3416928570.0 0.5 1.0 1.5 2.0 2.5 3.0123456789\\n−1.5 −1.0 −0.5 0.0 0.5 1.0−1.5 −1.0 −0.5 0.0 0.5X1X2\\nFIGURE 12.12. An illustration of how to properly interpret a dendrogram\\nwith nine observations in two-dimensional space. Left:a dendrogram generated\\nusing Euclidean distance and complete linkage. Observations 5and7are quite\\nsimilar to each other, as are observations 1and6. However, observation 9isno\\nmore similar to observation 2than it is to observations 8,5,and7, even though\\nobservations 9and2are close together in terms of horizontal distance. This is\\nbecause observations 2,8,5,and7all fuse with observation 9at the same height,\\napproximately 1.8.Right:the raw data used to generate the dendrogram can be\\nused to confirm that indeed, observation 9is no more similar to observation 2\\nthan it is to observations 8,5,and7.\\n(corresponding to a cut at height 0, so that each observation is in its own\\ncluster). In other words, the height of the cut to the dendrogram serves\\nthe same role as the KinK-means clustering: it controls the number of\\nclusters obtained.\\nFigure12.11therefore highlights a very attractive aspect of hierarchical\\nclustering: one single dendrogram can be used to obtain any number of\\nclusters. In practice, people often look at the dendrogram and select by eye\\na sensible number of clusters, based on the heights of the fusion and the\\nnumber of clusters desired. In the case of Figure 12.11, one might choose\\nto select either two or three clusters. However, often the choice of where to\\ncut the dendrogram is not so clear.\\nThe term hierarchical refers to the fact that clusters obtained by cutting\\nthe dendrogram at a given height are necessarily nested within the clusters\\nobtained by cutting the dendrogram at any greater height. However, on\\nan arbitrary data set, this assumption of hierarchical structure might be\\nunrealistic. For instance, suppose that our observations correspond to a\\ngroup of men and women, evenly split among Americans, Japanese, and\\nFrench. We can imagine a scenario in which the best division into two\\ngroups might split these people by gender, and the best division into three\\ngroups might split them by nationality. In this case, the true clusters are\\nnot nested, in the sense that the best division into three groups does not\\nresult from taking the best division into two groups and splitting up one\\nof those groups. Consequently, this situation could not be well-represented\\nby hierarchical clustering. Due to situations such as this one, hierarchical\\nclustering can sometimes yield worse(i.e. less accurate) results than K-\\nmeans clustering for a given number of clusters.', metadata={'source': './documents/ISLP_website.pdf', 'page': 533}), Document(page_content='12.4 Clustering Methods 529\\nAlgorithm 12.3 Hierarchical Clustering\\n1. Begin with nobservations and a measure (such as Euclidean dis-\\ntance) of all the(n\\n2)\\n=n(n−1)/2pairwise dissimilarities. Treat each\\nobservation as its own cluster.\\n2. Fori=n, n−1,...,2:\\n(a) Examine all pairwise inter-cluster dissimilarities among the i\\nclusters and identify the pair of clusters that are least dissimilar\\n(that is, most similar). Fuse these two clusters. The dissimilarity\\nbetween these two clusters indicates the height in the dendro-\\ngram at which the fusion should be placed.\\n(b) Compute the new pairwise inter-cluster dissimilarities among\\nthei−1remaining clusters.\\nThe Hierarchical Clustering Algorithm\\nThe hierarchical clustering dendrogram is obtained via an extremely simple\\nalgorithm.Webeginbydefiningsomesortof dissimilarity measurebetween\\neach pair of observations. Most often, Euclidean distance is used; we will\\ndiscuss the choice of dissimilarity measure later in this chapter. The algo-\\nrithm proceeds iteratively. Starting out at the bottom of the dendrogram,\\neach of the nobservations is treated as its own cluster. The two clusters\\nthat are most similar to each other are then fusedso that there now are\\nn−1clusters. Next the two clusters that are most similar to each other are\\nfused again, so that there now are n−2clusters. The algorithm proceeds\\nin this fashion until all of the observations belong to one single cluster, and\\nthe dendrogram is complete. Figure 12.13depicts the first few steps of the\\nalgorithm, for the data from Figure 12.12. To summarize, the hierarchical\\nclustering algorithm is given in Algorithm 12.3.\\nThis algorithm seems simple enough, but one issue has not been ad-\\ndressed. Consider the bottom right panel in Figure 12.13. How did we\\ndetermine that the cluster {5,7}should be fused with the cluster {8}?\\nWe have a concept of the dissimilarity between pairs of observations, but\\nhow do we define the dissimilarity between two clusters if one or both of\\nthe clusters contains multiple observations? The concept of dissimilarity\\nbetween a pair of observations needs to be extended to a pair of groups\\nof observations . This extension is achieved by developing the notion of\\nlinkage, which defines the dissimilarity between two groups of observa-linkagetions. The four most common types of linkage— complete ,average,single,\\nandcentroid—are briefly described in Table 12.3. Average, complete, and\\nsingle linkage are most popular among statisticians. Average and complete\\nlinkage are generally preferred over single linkage, as they tend to yield\\nmore balanced dendrograms. Centroid linkage is often used in genomics,\\nbut suffers from a major drawback in that an inversion can occur, wherebyinversiontwo clusters are fused at a height beloweither of the individual clusters in\\nthedendrogram.Thiscanleadtodifficultiesinvisualizationaswellasinin-\\nterpretation of the dendrogram. The dissimilarities computed in Step 2(b)', metadata={'source': './documents/ISLP_website.pdf', 'page': 534}), Document(page_content='530 12. Unsupervised Learning\\nLinkage Description\\nCompleteMaximal intercluster dissimilarity. Compute all pairwise\\ndissimilarities between the observations in cluster A and the\\nobservations in cluster B, and record the largestof these dis-\\nsimilarities.\\nSingleMinimal intercluster dissimilarity. Compute all pairwise dis-\\nsimilarities between the observations in cluster A and the\\nobservations in cluster B, and record the smallest of these\\ndissimilarities. Single linkage can result in extended, trailing\\nclustersinwhichsingleobservationsarefusedone-at-a-time.\\nAverageMean intercluster dissimilarity. Compute all pairwise dis-\\nsimilarities between the observations in cluster A and the\\nobservations in cluster B, and record the averageof these\\ndissimilarities.\\nCentroidDissimilarity between the centroid for cluster A (a mean\\nvector of length p) and the centroid for cluster B. Centroid\\nlinkage can result in undesirable inversions .\\nTABLE 12.3. A summary of the four most commonly-used types of linkage in\\nhierarchical clustering.\\nof the hierarchical clustering algorithm will depend on the type of linkage\\nused, as well as on the choice of dissimilarity measure. Hence, the resulting\\ndendrogram typically depends quite strongly on the type of linkage used,\\nas is shown in Figure 12.14.\\nChoice of Dissimilarity Measure\\nThus far, the examples in this chapter have used Euclidean distance as the\\ndissimilarity measure. But sometimes other dissimilarity measures might\\nbe preferred. For example, correlation-based distance considers two obser-\\nvations to be similar if their features are highly correlated, even though the\\nobserved values may be far apart in terms of Euclidean distance. This is\\nan unusual use of correlation, which is normally computed between vari-\\nables; here it is computed between the observation profiles for each pair\\nof observations. Figure 12.15illustrates the difference between Euclidean\\nand correlation-based distance. Correlation-based distance focuses on the\\nshapes of observation profiles rather than their magnitudes.\\nThe choice of dissimilarity measure is very important, as it has a strong\\neffect on the resulting dendrogram. In general, careful attention should be\\npaid to the type of data being clustered and the scientific question at hand.\\nThese considerations should determine what type of dissimilarity measure\\nis used for hierarchical clustering.\\nFor instance, consider an online retailer interested in clustering shoppers\\nbased on their past shopping histories. The goal is to identify subgroups\\nofsimilarshoppers, so that shoppers within each subgroup can be shown\\nitems and advertisements that are particularly likely to interest them. Sup-\\npose the data takes the form of a matrix where the rows are the shoppers\\nand the columns are the items available for purchase; the elements of the\\ndata matrix indicate the number of times a given shopper has purchased a', metadata={'source': './documents/ISLP_website.pdf', 'page': 535}), Document(page_content='12.4 Clustering Methods 531\\n123456789\\n−1.5 −1.0 −0.5 0.0 0.5 1.0−1.5 −1.0 −0.5 0.0 0.5123456789\\n−1.5 −1.0 −0.5 0.0 0.5 1.0−1.5 −1.0 −0.5 0.0 0.5\\n123456789\\n−1.5 −1.0 −0.5 0.0 0.5 1.0−1.5 −1.0 −0.5 0.0 0.5123456789\\n−1.5 −1.0 −0.5 0.0 0.5 1.0−1.5 −1.0 −0.5 0.0 0.5X1X1X1X1X2X2\\nX2X2\\nFIGURE 12.13. An illustration of the first few steps of the hierarchical\\nclustering algorithm, using the data from Figure 12.12, with complete linkage\\nand Euclidean distance. Top Left: initially, there are nine distinct clusters,\\n{1},{2},...,{9}.Top Right: the two clusters that are closest together, {5}and\\n{7}, are fused into a single cluster. Bottom Left: the two clusters that are closest\\ntogether, {6}and{1}, are fused into a single cluster. Bottom Right: the two clus-\\nters that are closest together using complete linkage ,{8}and the cluster {5,7},\\nare fused into a single cluster.\\ngiven item (i.e. a 0if the shopper has never purchased this item, a 1if the\\nshopper has purchased it once, etc.) What type of dissimilarity measure\\nshould be used to cluster the shoppers? If Euclidean distance is used, then\\nshoppers who have bought very few items overall (i.e. infrequent users of\\nthe online shopping site) will be clustered together. This may not be desir-\\nable.Ontheotherhand,ifcorrelation-baseddistanceisused,thenshoppers\\nwith similar preferences (e.g. shoppers who have bought items A and B but\\nnever items C or D) will be clustered together, even if some shoppers with\\nthese preferences are higher-volume shoppers than others. Therefore, for\\nthis application, correlation-based distance may be a better choice.\\nInadditiontocarefullyselectingthedissimilaritymeasureused,onemust\\nalso consider whether or not the variables should be scaled to have stan-\\ndarddeviationonebeforethedissimilaritybetweentheobservationsiscom-\\nputed. To illustrate this point, we continue with the online shopping ex-', metadata={'source': './documents/ISLP_website.pdf', 'page': 536}), Document(page_content='532 12. Unsupervised LearningAverage LinkageComplete LinkageSingle Linkage\\nFIGURE 12.14. Average, complete, and single linkage applied to an example\\ndata set. Average and complete linkage tend to yield more balanced clusters.\\nample just described. Some items may be purchased more frequently than\\nothers; for instance, a shopper might buy ten pairs of socks a year, but a\\ncomputer very rarely. High-frequency purchases like socks therefore tend\\nto have a much larger effect on the inter-shopper dissimilarities, and hence\\non the clustering ultimately obtained, than rare purchases like computers.\\nThis may not be desirable. If the variables are scaled to have standard de-\\nviation one before the inter-observation dissimilarities are computed, then\\neach variable will in effect be given equal importance in the hierarchical\\nclustering performed. We might also want to scale the variables to have\\nstandard deviation one if they are measured on different scales; otherwise,\\nthe choice of units (e.g. centimeters versus kilometers) for a particular vari-\\nable will greatly affect the dissimilarity measure obtained. It should come\\nas no surprise that whether or not it is a good decision to scale the variables\\nbefore computing the dissimilarity measure depends on the application at\\nhand. An example is shown in Figure 12.16. We note that the issue of\\nwhether or not to scale the variables before performing clustering applies\\ntoK-means clustering as well.\\n12.4.3 Practical Issues in Clustering\\nClustering can be a very useful tool for data analysis in the unsupervised\\nsetting. However, there are a number of issues that arise in performing\\nclustering. We describe some of these issues here.\\nSmall Decisions with Big Consequences\\nIn order to perform clustering, some decisions must be made.', metadata={'source': './documents/ISLP_website.pdf', 'page': 537}), Document(page_content='12.4 Clustering Methods 533\\n510152005 10 15 20\\nVariable IndexObservation 1Observation 2Observation 3\\n123FIGURE 12.15.Three observations with measurements on 20 variables areshown. Observations 1 and 3 have similar values for each variable and so thereis a small Euclidean distance between them. But they are very weakly correlated,so they have a large correlation-based distance. On the other hand, observations1 and 2 have quite different values for each variable, and so there is a largeEuclidean distance between them. But they are highly correlated, so there is asmall correlation-based distance between them.• Shouldtheobservationsorfeaturesfirstbestandardizedinsomeway?For instance, maybe the variables should be scaled to have standarddeviation one.• In the case of hierarchical clustering,–What dissimilarity measure should be used?–What type of linkage should be used?–Whereshouldwecutthedendrograminordertoobtainclusters?• In the case ofK-means clustering, how many clusters should we lookfor in the data?Each of these decisions can have a strong impact on the results obtained.In practice, we try several different choices, and look for the one withthe most useful or interpretable solution. With these methods, there is nosingle right answer—any solution that exposes some interesting aspects ofthe data should be considered.Validating the Clusters ObtainedAny time clustering is performed on a data set we will find clusters. But wereally want to know whether the clusters that have been found representtrue subgroups in the data, or whether they are simply a result ofclusteringthe noise. For instance, if we were to obtain an independent set of observa-tions, then would those observations also display the same set of clusters?This is a hard question to answer. There exist a number of techniques forassigning a p-value to a cluster in order to assess whether there is more', metadata={'source': './documents/ISLP_website.pdf', 'page': 538}), Document(page_content='534 12. Unsupervised Learning\\nSocks Computers0 2 4 6 8 10Socks Computers0.0 0.2 0.4 0.6 0.8 1.0 1.2Socks Computers0500 1000 1500FIGURE 12.16. An eclectic online retailer sells two items: socks and computers.\\nLeft:the number of pairs of socks, and computers, purchased by eight online shop-\\npers is displayed. Each shopper is shown in a different color. If inter-observation\\ndissimilarities are computed using Euclidean distance on the raw variables, then\\nthe number of socks purchased by an individual will drive the dissimilarities ob-\\ntained, and the number of computers purchased will have little effect. This might\\nbe undesirable, since (1) computers are more expensive than socks and so the\\nonline retailer may be more interested in encouraging shoppers to buy computers\\nthan socks, and (2) a large difference in the number of socks purchased by two\\nshoppers may be less informative about the shoppers’ overall shopping preferences\\nthan a small difference in the number of computers purchased. Center:the same\\ndata are shown, after scaling each variable by its standard deviation. Now the\\ntwo products will have a comparable effect on the inter-observation dissimilarities\\nobtained. Right:the same data are displayed, but now the y-axis represents the\\nnumber of dollars spent by each online shopper on socks and on computers. Since\\ncomputers are much more expensive than socks, now computer purchase history\\nwill drive the inter-observation dissimilarities obtained.\\nevidence for the cluster than one would expect due to chance. However,\\nthere has been no consensus on a single best approach. More details can\\nbe found in ESL.8\\nOther Considerations in Clustering\\nBothK-means and hierarchical clustering will assign each observation to\\na cluster. However, sometimes this might not be appropriate. For instance,\\nsuppose that most of the observations truly belong to a small number of\\n(unknown) subgroups, and a small subset of the observations are quite\\ndifferent from each other and from all other observations. Then since K-\\nmeans and hierarchical clustering force everyobservation into a cluster, the\\nclusters found may be heavily distorted due to the presence of outliers that\\ndo not belong to any cluster. Mixture models are an attractive approach\\nfor accommodating the presence of such outliers. These amount to a soft\\nversion of K-means clustering, and are described in ESL.\\n8ESL:The Elements of Statistical Learning by Hastie, Tibshirani and Friedman.', metadata={'source': './documents/ISLP_website.pdf', 'page': 539}), Document(page_content='12.5 Lab: Unsupervised Learning 535In addition, clustering methods generally are not very robust to pertur-bations to the data. For instance, suppose that we clusternobservations,and then cluster the observations again after removing a subset of thenobservations at random. One would hope that the two sets of clusters ob-tained would be quite similar, but often this is not the case!A Tempered Approach to Interpreting the Results of ClusteringWe have described some of the issues associated with clustering. However,clustering can be a very useful and valid statistical tool if used properly. Wementioned that small decisions in how clustering is performed, such as howthe data are standardized and what type of linkage is used, can have a largeeffect on the results. Therefore, we recommend performing clustering withdifferent choices of these parameters, and looking at the full set of resultsin order to see what patterns consistently emerge. Since clustering can benon-robust, we recommend clustering subsets of the data in order to get asense of the robustness of the clusters obtained. Most importantly, we mustbe careful about how the results of a clustering analysis are reported. Theseresults should not be taken as the absolute truth about a data set. Rather,they should constitute a starting point for the development of a scientifichypothesis and further study, preferably on an independent data set.12.5 Lab: Unsupervised LearningIn this lab we demonstrate PCA and clustering on several datasets. As inother labs, we import some of our libraries at this top level. This makesthe code more readable, as scanning the first few lines of the notebook tellus what libraries are used in this notebook.In[1]:import numpy as npimport pandas as pdimport matplotlib.pyplot as pltfrom statsmodels.datasets import get_rdatasetfrom sklearn.decomposition import PCAfrom sklearn.preprocessing import StandardScalerfrom ISLP import load_dataWe also collect the new imports needed for this lab.In[2]:from sklearn.cluster import \\\\(KMeans,AgglomerativeClustering)from scipy.cluster.hierarchy import \\\\(dendrogram,cut_tree)from ISLP.cluster import compute_linkage12.5.1 Principal Components AnalysisIn this lab, we perform PCA onUSArrests, a data set in theRcomputingenvironment. We retrieve the data usingget_rdataset(), which can fetchget_rdataset()', metadata={'source': './documents/ISLP_website.pdf', 'page': 540}), Document(page_content=\"536 12. Unsupervised Learningdata from many standardRpackages.The rows of the data set contain the 50 states, in alphabetical order.In[3]:USArrests = get_rdataset('USArrests').dataUSArrestsOut[3]:Murder Assault UrbanPop RapeAlabama 13.223658 21.2Alaska 10.026348 44.5Arizona 8.129480 31.0... ............Wisconsin 2.65366 10.8Wyoming 6.816160 15.6The columns of the data set contain the four variables.In[4]:USArrests.columnsOut[4]:Index(['Murder','Assault','UrbanPop','Rape'],dtype='object')We first briefly examine the data. We notice that the variables havevastly different means.In[5]:USArrests.mean()Out[5]:Murder 7.788Assault 170.760UrbanPop 65.540Rape21.232dtype: float64Dataframeshaveseveralusefulmethodsforcomputingcolumn-wisesum-maries. We can also examine the variance of the four variables using thevar()method.In[6]:USArrests.var()Out[6]:Murder 18.970465Assault 6945.165714UrbanPop 209.518776Rape87.729159dtype: float64Not surprisingly, the variables also have vastly different variances. TheUrbanPopvariable measures the percentage of the population in each stateliving in an urban area, which is not a comparable number to the number ofrapes in each state per 100,000 individuals. PCA looks for derived variablesthat account for most of the variance in the data set. If we do not scale thevariables before performing PCA, then the principal components wouldmostly be driven by theAssaultvariable, since it has by far the largestvariance. So if the variables are measured in different units or vary widelyin scale, it is recommended to standardize the variables to have standarddeviation one before performing PCA. Typically we set the means to zeroas well.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 541}), Document(page_content=\"12.5 Lab: Unsupervised Learning 537This scaling can be done via theStandardScaler()transform importedabove. We firstfitthe scaler, which computes the necessary means andstandard deviations and then apply it to our data using thetransformmethod. As before, we combine these steps using thefit_transform()method.In[7]:scaler = StandardScaler(with_std=True,with_mean=True)USArrests_scaled = scaler.fit_transform(USArrests)Having scaled the data, we can then perform principal components analysisusing thePCA()transform from thesklearn.decompositionpackage.PCA()In[8]:pcaUS = PCA()(By default, thePCA()transform centers the variables to have mean zerothough it does not scale them.) The transformpcaUScan be used to findthe PCAscoresreturned byfit(). Once thefitmethod has been called,thepcaUSobject also contains a number of useful quantities.In[9]:pcaUS.fit(USArrests_scaled)After fitting, themean_attribute corresponds to the means of the vari-ables. In this case, since we centered and scaled the data withscaler()themeans will all be 0.In[10]:pcaUS.mean_Out[10]:array([-0., 0., -0., 0.])The scores can be computed using thetransform()method ofpcaUSafterit has been fit.In[11]:scores = pcaUS.transform(USArrests_scaled)We will plot these scores a bit further down. Thecomponents_attributeprovides the principal component loadings: each row ofpcaUS.components_contains the corresponding principal component loading vector.In[12]:pcaUS.components_Out[12]:array([[ 0.53589947, 0.58318363, 0.27819087, 0.54343209],[0 . 4 1 8 1 8 0 8 7 , 0 . 1 8 7 9 8 5 6,- 0 . 8 7 2 8 0 6 1 9 ,- 0 . 1 6 7 3 1 8 6 4 ] ,[-0.34123273, -0.26814843, -0.37801579, 0.81777791],[0 . 6 4 9 2 2 7 8,- 0 . 7 4 3 4 0 7 4 8 , 0 . 1 3 3 8 7 7 7 3 , 0 . 0 8 9 0 2 4 3 2 ] ] )Thebiplotis a common visualization method used with PCA. It is notbuilt in as a standard part ofsklearn, though there are python packagesthat do produce such plots. Here we make a simple biplot manually.In[13]:i, j = 0, 1#w h i c hc o m p o n e n t sfig, ax = plt.subplots(1, 1, figsize=(8, 8))ax.scatter(scores[:,0], scores[:,1])ax.set_xlabel('PC%d'%( i + 1 ) )ax.set_ylabel('PC%d'%( j + 1 ) )for k in range(pcaUS.components_.shape[1]):\", metadata={'source': './documents/ISLP_website.pdf', 'page': 542}), Document(page_content=\"538 12. Unsupervised Learning\\nax.arrow(0, 0, pcaUS.components_[i,k], pcaUS.components_[j,k])\\nax.text(pcaUS.components_[i,k],\\npcaUS.components_[j,k],\\nUSArrests.columns[k])\\nNotice that this figure is a reflection of Figure 12.1through the y-axis.\\nRecall that the principal components are only unique up to a sign change,\\nso we can reproduce that figure by flipping the signs of the second set of\\nscores and loadings. We also increase the length of the arrows to emphasize\\nthe loadings.\\nIn[14]: scale_arrow = s_ = 2\\nscores[:,1] *= -1\\npcaUS.components_[1] *= -1 #f l i pt h ey - a x i s\\nfig, ax = plt.subplots(1, 1, figsize=(8, 8))\\nax.scatter(scores[:,0], scores[:,1])\\nax.set_xlabel( 'PC%d' %( i + 1 ) )\\nax.set_ylabel( 'PC%d' %( j + 1 ) )\\nfor k in range(pcaUS.components_.shape[1]):\\nax.arrow(0, 0, s_*pcaUS.components_[i,k], s_*pcaUS.components_[\\nj,k])\\nax.text(s_*pcaUS.components_[i,k],\\ns_*pcaUS.components_[j,k],\\nUSArrests.columns[k])\\nThe standard deviations of the principal component scores are as follows:\\nIn[15]: scores.std(0, ddof=1)\\nOut[15]: array([1.5909, 1.0050, 0.6032, 0.4207])\\nThevarianceofeachscorecanbeextracteddirectlyfromthe pcaUSobject\\nvia theexplained_variance_ attribute.\\nIn[16]: pcaUS.explained_variance_\\nOut[16]: array([2.5309, 1.01 , 0.3638, 0.177 ])\\nThe proportion of variance explained by each principal component (PVE)\\nis stored as explained_variance_ratio_ :\\nIn[17]: pcaUS.explained_variance_ratio_\\nOut[17]: array([0.6201, 0.2474, 0.0891, 0.0434])\\nWe see that the first principal component explains 62.0% of the variance\\nin the data, the next principal component explains 24.7% of the variance,\\nand so forth. We can plot the PVE explained by each component, as well\\nas the cumulative PVE. We first plot the proportion of variance explained.\\nIn[18]: %%capture\\nfig, axes = plt.subplots(1, 2, figsize=(15, 6))\\nticks = np.arange(pcaUS.n_components_)+1\\nax = axes[0]\\nax.plot(ticks,\\npcaUS.explained_variance_ratio_,\\nmarker= 'o')\", metadata={'source': './documents/ISLP_website.pdf', 'page': 543}), Document(page_content=\"12.5 Lab: Unsupervised Learning 539\\nax.set_xlabel( 'Principal Component' );\\nax.set_ylabel( 'Proportion of Variance Explained' )\\nax.set_ylim([0,1])\\nax.set_xticks(ticks)\\nNoticethe useof %%capture ,whichsuppressesthe displayingof thepartially\\ncompleted figure.\\nIn[19]: ax = axes[1]\\nax.plot(ticks,\\npcaUS.explained_variance_ratio_.cumsum(),\\nmarker= 'o')\\nax.set_xlabel( 'Principal Component' )\\nax.set_ylabel( 'Cumulative Proportion of Variance Explained' )\\nax.set_ylim([0, 1])\\nax.set_xticks(ticks)\\nfig\\nThe result is similar to that shown in Figure 12.3. Note that the method\\ncumsum() computes the cumulative sum of the elements of a numeric vector.cumsum()For instance:\\nIn[20]: a=n p . a r r a y ( [ 1 , 2 , 8 , - 3 ] )\\nnp.cumsum(a)\\nOut[20]: array([ 1, 3, 11, 8])\\n12.5.2 Matrix Completion\\nWe now re-create the analysis carried out on the USArrests data in Sec-\\ntion12.3.\\nWe saw in Section 12.2.2that solving the optimization problem ( 12.6) on\\na centered data matrix Xis equivalent to computing the first Mprincipal\\ncomponents of the data. We use our scaled and centered USArrests data as\\nXbelow. The singular value decomposition (SVD) is a general algorithmsingular\\nvalue de-\\ncomposition\\nsvd()for solving ( 12.6).\\nIn[21]: X=U S A r r e s t s _ s c a l e d\\nU, D, V = np.linalg.svd(X, full_matrices=False)\\nU.shape, D.shape, V.shape\\nOut[21]: ((50, 4), (4,), (4, 4))\\nThenp.linalg.svd() function returns three components, U,DandV. Thenp.linalg.\\nsvd() matrixVis equivalent to the loading matrix from principal components (up\\nto an unimportant sign flip). Using the full_matrices=False option ensures\\nthat for a tall matrix the shape of Uis the same as the shape of X.\\nIn[22]: V\\nOut[22]: array([[-0.53589947, -0.58318363, -0.27819087, -0.54343209],\\n[0 . 4 1 8 1 8 0 8 7 , 0 . 1 8 7 9 8 5 6,- 0 . 8 7 2 8 0 6 1 9 ,- 0 . 1 6 7 3 1 8 6 4 ] ,\\n[-0.34123273, -0.26814843, -0.37801579, 0.81777791],\\n[0 . 6 4 9 2 2 7 8,- 0 . 7 4 3 4 0 7 4 8 , 0 . 1 3 3 8 7 7 7 3 , 0 . 0 8 9 0 2 4 3 2 ] ] )\", metadata={'source': './documents/ISLP_website.pdf', 'page': 544}), Document(page_content='540 12. Unsupervised Learning\\nIn[23]: pcaUS.components_\\nOut[23]: array([[ 0.53589947, 0.58318363, 0.27819087, 0.54343209],\\n[0 . 4 1 8 1 8 0 8 7 , 0 . 1 8 7 9 8 5 6,- 0 . 8 7 2 8 0 6 1 9 ,- 0 . 1 6 7 3 1 8 6 4 ] ,\\n[-0.34123273, -0.26814843, -0.37801579, 0.81777791],\\n[0 . 6 4 9 2 2 7 8,- 0 . 7 4 3 4 0 7 4 8 , 0 . 1 3 3 8 7 7 7 3 , 0 . 0 8 9 0 2 4 3 2 ] ] )\\nThematrix Ucorrespondstoa standardized versionofthePCAscorematrix\\n(eachcolumnstandardizedtohavesum-of-squaresone).Ifwemultiplyeach\\ncolumn of Uby the corresponding element of D, we recover the PCA scores\\nexactly (up to a meaningless sign flip).\\nIn[24]: (U * D[None,:])[:3]\\nOut[24]: array([[-0.9856, 1.1334, -0.4443, 0.1563],\\n[-1.9501, 1.0732, 2.04 , -0.4386],\\n[-1.7632, -0.746 , 0.0548, -0.8347]])\\nIn[25]: scores[:3]\\nOut[25]: array([[ 0.9856, -1.1334, -0.4443, 0.1563],\\n[1 . 9 5 0 1 ,- 1 . 0 7 3 2 , 2 . 0 4 ,- 0 . 4 3 8 6 ] ,\\n[1 . 7 6 3 2 , 0 . 7 4 6, 0 . 0 5 4 8 ,- 0 . 8 3 4 7 ] ] )\\nWhile it would be possible to carry out this lab using the PCA()estimator,\\nhere we use the np.linalg.svd() function in order to illustrate its use.\\nWe now omit 20 entries in the 50×4data matrix at random. We do\\nso by first selecting 20 rows (states) at random, and then selecting one of\\nthe four entries in each row at random. This ensures that every row has at\\nleast three observed values.\\nIn[26]: n_omit = 20\\nnp.random.seed(15)\\nr_idx = np.random.choice(np.arange(X.shape[0]),\\nn_omit,\\nreplace=False)\\nc_idx = np.random.choice(np.arange(X.shape[1]),\\nn_omit,\\nreplace=True)\\nXna = X.copy()\\nXna[r_idx, c_idx] = np.nan\\nHere the array r_idxcontains 20 integers from 0 to 49; this represents\\nthe states (rows of X) that are selected to contain missing values. And c_idx\\ncontains 20 integers from 0 to 3, representing the features (columns in X)\\nthat contain the missing values for each of the selected states.\\nWe now write some code to implement Algorithm 12.1. We first write a\\nfunctionthattakesinamatrix,andreturnsanapproximationtothematrix\\nusing the svd()function. This will be needed in Step 2 of Algorithm 12.1.\\nIn[27]: def low_rank(X, M=1):\\nU, D, V = np.linalg.svd(X)\\nL=U [ : , : M ]*D [ N o n e , : M ]\\nreturn L.dot(V[:M])', metadata={'source': './documents/ISLP_website.pdf', 'page': 545}), Document(page_content='12.5 Lab: Unsupervised Learning 541\\nTo conduct Step 1 of the algorithm, we initialize Xhat— this is ˜Xin\\nAlgorithm 12.1— by replacing the missing values with the column means\\nof the non-missing entries. These are stored in Xbarbelow after running\\nnp.nanmean() over the row axis. We make a copy so that when we assignnp.nanmean()values to Xhatbelow we do not also overwrite the values in Xna.\\nIn[28]: Xhat = Xna.copy()\\nXbar = np.nanmean(Xhat, axis=0)\\nXhat[r_idx, c_idx] = Xbar[c_idx]\\nBefore we begin Step 2, we set ourselves up to measure the progress of\\nour iterations:\\nIn[29]: thresh = 1e-7\\nrel_err = 1\\ncount = 0\\nismiss = np.isnan(Xna)\\nmssold = np.mean(Xhat[ ∼ismiss]**2)\\nmss0 = np.mean(Xna[ ∼ismiss]**2)\\nHereismissis a logical matrix with the same dimensions as Xna; a given\\nelementis Trueifthecorrespondingmatrixelementismissing.Thenotation\\n∼ismissnegates this boolean vector. This is useful because it allows us to\\naccess both the missing and non-missing entries. We store the mean of the\\nsquared non-missing elements in mss0. We store the mean squared error\\nof the non-missing elements of the old version of Xhatinmssold(which\\ncurrently agrees with mss0). We plan to store the mean squared error of\\nthe non-missing elements of the current version of Xhatinmss, and will then\\niterate Step 2 of Algorithm 12.1until the relative error , defined as (mssold\\n- mss) / mss0 , falls below thresh = 1e-7 .9\\nIn Step 2(a) of Algorithm 12.1, we approximate Xhatusinglow_rank() ;\\nwe call this Xapp. In Step 2(b), we use Xappto update the estimates for\\nelements in Xhatthat are missing in Xna. Finally, in Step 2(c), we compute\\nthe relative error. These three steps are contained in the following while\\nloop:\\nIn[30]: while rel_err > thresh:\\ncount += 1\\n#S t e p2 ( a )\\nXapp = low_rank(Xhat, M=1)\\n#S t e p2 ( b )\\nXhat[ismiss] = Xapp[ismiss]\\n#S t e p2 ( c )\\nmss = np.mean(((Xna - Xapp)[ ∼ismiss])**2)\\nrel_err = (mssold - mss) / mss0\\nmssold = mss\\nprint(\"Iteration: {0}, MSS:{1:.3f}, Rel.Err {2:.2e}\"\\n.format(count, mss, rel_err))\\n9Algorithm 12.1tells us to iterate Step 2 until ( 12.14) is no longer decreasing. Deter-\\nmining whether ( 12.14) is decreasing requires us only to keep track of mssold - mss .\\nHowever, in practice, we keep track of (mssold - mss) / mss0 instead: this makes\\nit so that the number of iterations required for Algorithm 12.1to converge does not\\ndepend on whether we multiplied the raw data Xby a constant factor.', metadata={'source': './documents/ISLP_website.pdf', 'page': 546}), Document(page_content='542 12. Unsupervised Learning\\nIteration: 1, MSS:0.395, Rel.Err 5.99e-01\\nIteration: 2, MSS:0.382, Rel.Err 1.33e-02\\nIteration: 3, MSS:0.381, Rel.Err 1.44e-03\\nIteration: 4, MSS:0.381, Rel.Err 1.79e-04\\nIteration: 5, MSS:0.381, Rel.Err 2.58e-05\\nIteration: 6, MSS:0.381, Rel.Err 4.22e-06\\nIteration: 7, MSS:0.381, Rel.Err 7.65e-07\\nIteration: 8, MSS:0.381, Rel.Err 1.48e-07\\nIteration: 9, MSS:0.381, Rel.Err 2.95e-08\\nWe see that after eight iterations, the relative error has fallen below\\nthresh = 1e-7 , and so the algorithm terminates. When this happens, the\\nmean squared error of the non-missing elements equals 0.381.\\nFinally, we compute the correlation between the 20 imputed values and\\nthe actual values:\\nIn[31]: np.corrcoef(Xapp[ismiss], X[ismiss])[0,1]\\nOut[31]: 0.711\\nIn this lab, we implemented Algorithm 12.1ourselves for didactic pur-\\nposes. However, a reader who wishes to apply matrix completion to their\\ndata might look to more specialized Pythonimplementations.\\n12.5.3 Clustering\\nK-Means Clustering\\nThe estimator sklearn.cluster.KMeans() performs K-means clustering inKmeans()Python. We begin with a simple simulated example in which there truly are\\ntwo clusters in the data: the first 25 observations have a mean shift relative\\nto the next 25 observations.\\nIn[32]: np.random.seed(0);\\nX=n p . r a n d o m . s t a n d a r d _ n o r m a l ( ( 5 0 , 2 ) ) ;\\nX[:25,0] += 3;\\nX[:25,1] -= 4;\\nWe now perform K-means clustering with K=2.\\nIn[33]: kmeans = KMeans(n_clusters=2,\\nrandom_state=2,\\nn_init=20).fit(X)\\nWe specify random_state to make the results reproducible. The cluster as-\\nsignments of the 50 observations are contained in kmeans.labels_ .\\nIn[34]: kmeans.labels_\\nOut[34]: array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\\n1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\\n0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)\\nTheK-means clustering perfectly separated the observations into two clus-\\nters even though we did not supply any group information to KMeans() .W e\\ncan plot the data, with each observation colored according to its cluster\\nassignment.', metadata={'source': './documents/ISLP_website.pdf', 'page': 547}), Document(page_content='12.5 Lab: Unsupervised Learning 543\\nIn[35]: fig, ax = plt.subplots(1, 1, figsize=(8,8))\\nax.scatter(X[:,0], X[:,1], c=kmeans.labels_)\\nax.set_title( \"K-Means Clustering Results with K=2\" );\\nHeretheobservationscanbeeasilyplottedbecausetheyaretwo-dimensio-\\nnal. If there were more than two variables then we could instead perform\\nPCA and plot the first two principal component score vectors to represent\\nthe clusters.\\nIn this example, we knew that there really were two clusters because\\nwe generated the data. However, for real data, we do not know the true\\nnumber of clusters, nor whether they exist in any precise way. We could\\ninstead have performed K-means clustering on this example with K=3.\\nIn[36]: kmeans = KMeans(n_clusters=3,\\nrandom_state=3,\\nn_init=20).fit(X)\\nfig, ax = plt.subplots(figsize=(8,8))\\nax.scatter(X[:,0], X[:,1], c=kmeans.labels_)\\nax.set_title( \"K-Means Clustering Results with K=3\" );\\nWhenK=3,K-means clustering splits up the two clusters. We have used\\nthen_initargumenttorunthe K-meanswith20initialclusterassignments\\n(the default is 10). If a value of n_initgreater than one is used, then K-\\nmeans clustering will be performed using multiple random assignments in\\nStep 1 of Algorithm 12.2, and the KMeans() function will report only the\\nbest results. Here we compare using n_init=1 ton_init=20 .\\nIn[37]: kmeans1 = KMeans(n_clusters=3,\\nrandom_state=3,\\nn_init=1).fit(X)\\nkmeans20 = KMeans(n_clusters=3,\\nrandom_state=3,\\nn_init=20).fit(X);\\nkmeans1.inertia_, kmeans20.inertia_\\nOut[37]: (78.06, 75.04)\\nNote that kmeans.inertia_ is the total within-cluster sum of squares, which\\nwe seek to minimize by performing K-means clustering ( 12.17).\\nWestronglyrecommend always running K-means clustering with a large\\nvalue of n_init, such as 20 or 50, since otherwise an undesirable local\\noptimum may be obtained.\\nWhen performing K-means clustering, in addition to using multiple ini-\\ntial cluster assignments, it is also important to set a random seed using the\\nrandom_state argument to KMeans() . This way, the initial cluster assign-\\nments in Step 1 can be replicated, and the K-means output will be fully\\nreproducible.\\nHierarchical Clustering\\nTheAgglomerativeClustering() class from the sklearn.clustering pack-Agglomerative\\nClustering()age implements hierarchical clustering. As its name is long, we use the\\nshort hand HClustforhierarchical clustering . Note that this will not change', metadata={'source': './documents/ISLP_website.pdf', 'page': 548}), Document(page_content='544 12. Unsupervised Learningthe return type when using this method, so instances will still be of classAgglomerativeClustering. In the following example we use the data fromthe previous lab to plot the hierarchical clustering dendrogram using com-plete, single, and average linkage clustering with Euclidean distance as thedissimilarity measure. We begin by clustering observations using completelinkage.In[38]:HClust = AgglomerativeClusteringhc_comp = HClust(distance_threshold=0,n_clusters=None,linkage=\\'complete\\')hc_comp.fit(X)This computes the entire dendrogram. We could just as easily performhierarchical clustering with average or single linkage instead:In[39]:hc_avg = HClust(distance_threshold=0,n_clusters=None,linkage=\\'average\\');hc_avg.fit(X)hc_sing = HClust(distance_threshold=0,n_clusters=None,linkage=\\'single\\');hc_sing.fit(X);To use a precomputed distance matrix, we provide an additional argu-mentmetric=\"precomputed\".Inthecodebelow,thefirstfourlinescomputesthe50×50pairwise-distance matrix.In[40]:D=n p . z e r o s ( ( X . s h a p e [ 0 ] ,X . s h a p e [ 0 ] ) ) ;for i in range(X.shape[0]):x_ = np.multiply.outer(np.ones(X.shape[0]), X[i])D[i] = np.sqrt(np.sum((X - x_)**2, 1));hc_sing_pre = HClust(distance_threshold=0,n_clusters=None,metric=\\'precomputed\\',linkage=\\'single\\')hc_sing_pre.fit(D)We usedendrogram()fromscipy.cluster.hierarchyto plot the dendro-dendrogram()gram.However,dendrogram()expects a so-calledlinkage-matrix representa-tionof the clustering, which is not provided byAgglomerativeClustering(),but can be computed. The functioncompute_linkage()in theISLP.clustercompute_linkage()ISLP.clusterpackage is provided for this purpose.We can now plot the dendrograms. The numbers at the bottom of theplot identify each observation. Thedendrogram()function has a defaultmethod to color different branches of the tree that suggests a pre-definedcut of the tree at a particular depth. We prefer to overwrite this default bysetting this threshold to be infinite. Since we want this behavior for manydendrograms, we store these values in a dictionarycargsand pass this askeyword arguments using the notation**cargs.In[41]:cargs = {\\'color_threshold\\':-np.inf,\\'above_threshold_color\\':\\'black\\'}linkage_comp = compute_linkage(hc_comp)fig, ax = plt.subplots(1, 1, figsize=(8, 8))', metadata={'source': './documents/ISLP_website.pdf', 'page': 549}), Document(page_content='12.5 Lab: Unsupervised Learning 545\\ndendrogram(linkage_comp,\\nax=ax,\\n**cargs);\\nWe may want to color branches of the tree above and below a cut-\\nthresholddifferently.Thiscanbeachievedbychangingthe color_threshold .\\nLet’scutthetreeataheightof4,coloringlinksthatmergeabove4inblack.\\nIn[42]: fig, ax = plt.subplots(1, 1, figsize=(8, 8))\\ndendrogram(linkage_comp,\\nax=ax,\\ncolor_threshold=4,\\nabove_threshold_color= \\'black\\' );\\nTo determine the cluster labels for each observation associated with a\\ngiven cut of the dendrogram, we can use the cut_tree() function fromcut_tree()scipy.cluster.hierarchy :\\nIn[43]: cut_tree(linkage_comp, n_clusters=4).T\\nOut[43]: array([[0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 2, 0, 0, 0, 1, 1, 0, 0, 1,\\n0, 0, 2, 0, 2, 2, 3, 2, 3, 3, 3, 3, 2, 3, 3, 3, 3, 2, 3,\\n3, 3, 3, 2, 3, 3, 3, 3, 3, 3, 3, 3]])\\nThiscanalsobeachievedbyprovidinganargument n_clusters toHClust() ;\\nhowever each cut would require recomputing the clustering. Similarly, trees\\nmay be cut by distance threshold with an argument of distance_threshold\\ntoHClust() orheighttocut_tree() .\\nIn[44]: cut_tree(linkage_comp, height=5)\\nOut[44]: array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,\\n0, 0, 1, 0, 1, 1, 2, 1, 2, 2, 2, 2, 1, 2, 2, 2, 2, 1, 2,\\n2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2]])\\nTo scale the variables before performing hierarchical clustering of the\\nobservations, we use StandardScaler() as in our PCA example:\\nIn[45]: scaler = StandardScaler()\\nX_scale = scaler.fit_transform(X)\\nhc_comp_scale = HClust(distance_threshold=0,\\nn_clusters=None,\\nlinkage= \\'complete\\' ).fit(X_scale)\\nlinkage_comp_scale = compute_linkage(hc_comp_scale)\\nfig, ax = plt.subplots(1, 1, figsize=(8, 8))\\ndendrogram(linkage_comp_scale, ax=ax, **cargs)\\nax.set_title( \"Hierarchical Clustering with Scaled Features\" );\\nCorrelation-based distances between observations can be used for clus-\\ntering. The correlation between two observations measures the similarity\\nof their feature values.10Withnobservations, the n×ncorrelation matrix\\n10Suppose each observation has pfeatures, each a single numerical value. We measure\\nthe similarity of two such observations by computing the correlation of these ppairs of\\nnumbers.', metadata={'source': './documents/ISLP_website.pdf', 'page': 550}), Document(page_content='546 12. Unsupervised Learningcan then be used as a similarity (or affinity) matrix, i.e. so that one minusthe correlation matrix is the dissimilarity matrix used for clustering.Note that using correlation only makes sense for data with at least threefeatures since the absolute correlation between any two observations withmeasurements on two features is always one. Hence, we will cluster a three-dimensional data set.In[46]:X=n p . r a n d o m . s t a n d a r d _ n o r m a l ( ( 3 0 ,3 ) )corD = 1 - np.corrcoef(X)hc_cor = HClust(linkage=\\'complete\\',distance_threshold=0,n_clusters=None,metric=\\'precomputed\\')hc_cor.fit(corD)linkage_cor = compute_linkage(hc_cor)fig, ax = plt.subplots(1, 1, figsize=(8, 8))dendrogram(linkage_cor, ax=ax, **cargs)ax.set_title(\"Complete Linkage with Correlation-Based Dissimilarity\");12.5.4 NCI60 Data ExampleUnsupervised techniques are often used in the analysis of genomic data.In particular, PCA and hierarchical clustering are popular tools. We illus-trate these techniques on theNCI60cancer cell line microarray data, whichconsists of 6830 gene expression measurements on 64 cancer cell lines.In[47]:NCI60 = load_data(\\'NCI60\\')nci_labs = NCI60[\\'labels\\']nci_data = NCI60[\\'data\\']Each cell line is labeled with a cancer type. We do not make use of thecancer types in performing PCA and clustering, as these are unsupervisedtechniques. But after performing PCA and clustering, we will check tosee the extent to which these cancer types agree with the results of theseunsupervised techniques.The data has 64 rows and 6830 columns.In[48]:nci_data.shapeOut[48]:(64, 6830)We begin by examining the cancer types for the cell lines.In[49]:nci_labs.value_counts()Out[49]:labelNSCLC9RENAL9MELANOMA 8BREAST7COLON7LEUKEMIA 6OVARIAN 6', metadata={'source': './documents/ISLP_website.pdf', 'page': 551}), Document(page_content=\"12.5 Lab: Unsupervised Learning 547\\nCNS 5\\nPROSTATE 2\\nK562A-repro 1\\nK562B-repro 1\\nMCF7A-repro 1\\nMCF7D-repro 1\\nUNKNOWN 1\\ndtype: int64\\nPCA on the NCI60 Data\\nWefirstperformPCAonthedataafterscalingthevariables(genes)tohave\\nstandard deviation one, although here one could reasonably argue that it\\nis better not to scale the genes as they are measured in the same units.\\nIn[50]: scaler = StandardScaler()\\nnci_scaled = scaler.fit_transform(nci_data)\\nnci_pca = PCA()\\nnci_scores = nci_pca.fit_transform(nci_scaled)\\nWe now plot the first few principal component score vectors, in order to\\nvisualize the data. The observations (cell lines) corresponding to a given\\ncancer type will be plotted in the same color, so that we can see to what\\nextent the observations within a cancer type are similar to each other.\\nIn[51]: cancer_types = list(np.unique(nci_labs))\\nnci_groups = np.array([cancer_types.index(lab)\\nfor lab in nci_labs.values])\\nfig, axes = plt.subplots(1, 2, figsize=(15,6))\\nax = axes[0]\\nax.scatter(nci_scores[:,0],\\nnci_scores[:,1],\\nc=nci_groups,\\nmarker= 'o',\\ns=50)\\nax.set_xlabel( 'PC1'); ax.set_ylabel( 'PC2')\\nax = axes[1]\\nax.scatter(nci_scores[:,0],\\nnci_scores[:,2],\\nc=nci_groups,\\nmarker= 'o',\\ns=50)\\nax.set_xlabel( 'PC1'); ax.set_ylabel( 'PC3');\\nThe resulting plots are shown in Figure 12.17. On the whole, cell lines\\ncorresponding to a single cancer type do tend to have similar values on the\\nfirst few principal component score vectors. This indicates that cell lines\\nfrom the same cancer type tend to have pretty similar gene expression\\nlevels.\\nWe can also plot the percent variance explained by the principal compo-\\nnents as well as the cumulative percent variance explained. This is similar\\nto the plots we made earlier for the USArrests data.\\nIn[52]: fig, axes = plt.subplots(1, 2, figsize=(15,6))\\nax = axes[0]\\nticks = np.arange(nci_pca.n_components_)+1\", metadata={'source': './documents/ISLP_website.pdf', 'page': 552}), Document(page_content=\"548 12. Unsupervised Learning\\n−40 −20 0 20 40 60−60 −40 −20 0 20−40 −20 0 20 40 60−40 −20 0 20 40Z1Z1Z2\\nZ3\\nFIGURE 12.17. Projections of the NCI60cancer cell lines onto the first three\\nprincipal components (in other words, the scores for the first three principal com-\\nponents). On the whole, observations belonging to a single cancer type tend to\\nlie near each other in this low-dimensional space. It would not have been possible\\nto visualize the data without using a dimension reduction method such as PCA,\\nsince based on the full data set there are(6,830\\n2)\\npossible scatterplots, none of\\nwhich would have been particularly informative.\\nax.plot(ticks,\\nnci_pca.explained_variance_ratio_,\\nmarker= 'o')\\nax.set_xlabel( 'Principal Component' );\\nax.set_ylabel( 'PVE')\\nax = axes[1]\\nax.plot(ticks,\\nnci_pca.explained_variance_ratio_.cumsum(),\\nmarker= 'o');\\nax.set_xlabel( 'Principal Component' )\\nax.set_ylabel( 'Cumulative PVE' );\\nThe resulting plots are shown in Figure 12.18.\\nWeseethattogether,thefirstsevenprincipalcomponentsexplainaround\\n40% of the variance in the data. This is not a huge amount of the variance.\\nHowever, looking at the scree plot, we see that while each of the first seven\\nprincipal components explain a substantial amount of variance, there is a\\nmarkeddecreaseinthevarianceexplainedbyfurtherprincipalcomponents.\\nThat is, there is an elbowin the plot after approximately the seventh princi-\\npal component. This suggests that there may be little benefit to examining\\nmore than seven or so principal components (though even examining seven\\nprincipal components may be difficult).\\nClustering the Observations of the NCI60 Data\\nWe now perform hierarchical clustering of the cell lines in the NCI60data\\nusing complete, single, and average linkage. Once again, the goal is to find\\nout whether or not the observations cluster into distinct types of cancer.\\nEuclidean distance is used as the dissimilarity measure. We first write a\\nshort function to produce the three dendrograms.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 553}), Document(page_content=\"12.5 Lab: Unsupervised Learning 549\\n0 10 20 30 40 50 600 2 4 6 8 10Principal ComponentPVE\\n0 10 20 30 40 50 6020 40 60 80 100Principal ComponentCumulative PVEFIGURE 12.18. The PVE of the principal components of the NCI60cancer cell\\nline microarray data set. Left:the PVE of each principal component is shown.\\nRight:the cumulative PVE of the principal components is shown. Together, all\\nprincipal components explain 100,% of the variance.\\nIn[53]: def plot_nci(linkage, ax, cut=-np.inf):\\ncargs = { 'above_threshold_color' :'black' ,\\n'color_threshold' :cut}\\nhc = HClust(n_clusters=None,\\ndistance_threshold=0,\\nlinkage=linkage.lower()).fit(nci_scaled)\\nlinkage_ = compute_linkage(hc)\\ndendrogram(linkage_,\\nax=ax,\\nlabels=np.asarray(nci_labs),\\nleaf_font_size=10,\\n**cargs)\\nax.set_title( '%s Linkage' %l i n k a g e )\\nreturn hc\\nLet’s plot our results.\\nIn[54]: fig, axes = plt.subplots(3, 1, figsize=(15,30))\\nax = axes[0]; hc_comp = plot_nci( 'Complete' ,a x )\\nax = axes[1]; hc_avg = plot_nci( 'Average' ,a x )\\nax = axes[2]; hc_sing = plot_nci( 'Single' ,a x )\\nThe results are shown in Figure 12.19. We see that the choice of linkage\\ncertainly does affect the results obtained. Typically, single linkage will tend\\nto yieldtrailingclusters: very large clusters onto which individual observa-\\ntions attach one-by-one. On the other hand, complete and average linkage\\ntend to yield more balanced, attractive clusters. For this reason, complete\\nand average linkage are generally preferred to single linkage. Clearly cell\\nlines within a single cancer type do tend to cluster together, although the\\nclustering is not perfect. We will use complete linkage hierarchical cluster-\\ning for the analysis that follows.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 554}), Document(page_content='550 12. Unsupervised LearningBREASTBREASTCNSCNSRENALBREASTNSCLCRENALMELANOMAOVARIANOVARIANNSCLCOVARIANCOLONCOLONOVARIANPROSTATENSCLCNSCLCNSCLCPROSTATENSCLCMELANOMARENALRENALRENALOVARIANUNKNOWNOVARIANNSCLCCNSCNSCNSNSCLCRENALRENALRENALRENALNSCLCMELANOMAMELANOMAMELANOMAMELANOMAMELANOMAMELANOMABREASTBREASTCOLONCOLONCOLONCOLONCOLONBREASTMCF7A−reproBREASTMCF7D−reproLEUKEMIALEUKEMIALEUKEMIALEUKEMIAK562B−reproK562A−reproLEUKEMIALEUKEMIA40 80 120 160Complete LinkageLEUKEMIALEUKEMIALEUKEMIALEUKEMIALEUKEMIALEUKEMIAK562B−reproK562A−reproRENALNSCLCBREASTNSCLCBREASTMCF7A−reproBREASTMCF7D−reproCOLONCOLONCOLONRENALMELANOMAMELANOMABREASTBREASTMELANOMAMELANOMAMELANOMAMELANOMAMELANOMAOVARIANOVARIANNSCLCOVARIANUNKNOWNOVARIANNSCLCMELANOMACNSCNSCNSRENALRENALRENALRENALRENALRENALRENALPROSTATENSCLCNSCLCNSCLCNSCLCOVARIANPROSTATENSCLCCOLONCOLONOVARIANCOLONCOLONCNSCNSBREASTBREAST40 60 80 100 120Average LinkageLEUKEMIARENALBREASTLEUKEMIALEUKEMIACNSLEUKEMIALEUKEMIAK562B−reproK562A−reproNSCLCLEUKEMIAOVARIANNSCLCCNSBREASTNSCLCOVARIANCOLONBREASTMELANOMARENALMELANOMABREASTBREASTMELANOMAMELANOMAMELANOMAMELANOMAMELANOMABREASTOVARIANCOLONMCF7A−reproBREASTMCF7D−reproUNKNOWNOVARIANNSCLCNSCLCPROSTATEMELANOMACOLONOVARIANNSCLCRENALCOLONPROSTATECOLONOVARIANCOLONCOLONNSCLCNSCLCRENALNSCLCRENALRENALRENALRENALRENALCNSCNSCNS40 60 80 100Single Linkage\\nFIGURE 12.19.TheNCI60cancer cell line microarray data, clustered withaverage, complete, and single linkage, and using Euclidean distance as the dissim-ilarity measure. Complete and average linkage tend to yield evenly sized clusterswhereas single linkage tends to yield extended clusters to which single leaves arefused one by one.', metadata={'source': './documents/ISLP_website.pdf', 'page': 555}), Document(page_content=\"12.5 Lab: Unsupervised Learning 551\\nWe can cut the dendrogram at the height that will yield a particular\\nnumber of clusters, say four:\\nIn[55]: linkage_comp = compute_linkage(hc_comp)\\ncomp_cut = cut_tree(linkage_comp, n_clusters=4).reshape(-1)\\npd.crosstab(nci_labs[ 'label' ],\\npd.Series(comp_cut.reshape(-1), name= 'Complete' ))\\nThere are some clear patterns. All the leukemia cell lines fall in one\\ncluster, while the breast cancer cell lines are spread out over three different\\nclusters.\\nWe can plot a cut on the dendrogram that produces these four clusters:\\nIn[56]: fig, ax = plt.subplots(figsize=(10,10))\\nplot_nci( 'Complete' ,a x ,c u t = 1 4 0 )\\nax.axhline(140, c= 'r',l i n e w i d t h = 4 ) ;\\nTheaxhline() function draws a horizontal line line on top of any existing\\nset of axes. The argument 140plots a horizontal line at height 140 on the\\ndendrogram; this is a height that results in four distinct clusters. It is easy\\nto verify that the resulting clusters are the same as the ones we obtained\\nincomp_cut .\\nWe claimed earlier in Section 12.4.2thatK-means clustering and hier-\\narchical clustering with the dendrogram cut to obtain the same number\\nof clusters can yield very different results. How do these NCI60hierarchical\\nclusteringresultscomparetowhat wegetifweperform K-meansclustering\\nwithK=4?\\nIn[57]: nci_kmeans = KMeans(n_clusters=4,\\nrandom_state=0,\\nn_init=20).fit(nci_scaled)\\npd.crosstab(pd.Series(comp_cut, name= 'HClust' ),\\npd.Series(nci_kmeans.labels_, name= 'K-means' ))\\nOut[57]: K-means 0 1 2 3\\nHClust\\n02 8 3 9 0\\n17 0 0 0\\n20 0 0 8\\n30 9 0 0\\nWe see that the four clusters obtained using hierarchical clustering and\\nK-means clustering are somewhat different. First we note that the labels\\nin the two clusterings are arbitrary. That is, swapping the identifier of the\\ncluster does not change the clustering. We see here Cluster 3 in K-means\\nclustering is identical to cluster 2 in hierarchical clustering. However, the\\nother clusters differ: for instance, cluster 0 in K-means clustering contains\\na portion of the observations assigned to cluster 0 by hierarchical cluster-\\ning, as well as all of the observations assigned to cluster 1 by hierarchical\\nclustering.\\nRather than performing hierarchical clustering on the entire data ma-\\ntrix, we can also perform hierarchical clustering on the first few principal\\ncomponent score vectors, regarding these first few components as a less\\nnoisy version of the data.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 556}), Document(page_content='552 12. Unsupervised Learning\\nIn[58]: hc_pca = HClust(n_clusters=None,\\ndistance_threshold=0,\\nlinkage= \\'complete\\'\\n).fit(nci_scores[:,:5])\\nlinkage_pca = compute_linkage(hc_pca)\\nfig, ax = plt.subplots(figsize=(8,8))\\ndendrogram(linkage_pca,\\nlabels=np.asarray(nci_labs),\\nleaf_font_size=10,\\nax=ax,\\n**cargs)\\nax.set_title( \"Hier. Clust. on First Five Score Vectors\" )\\npca_labels = pd.Series(cut_tree(linkage_pca,\\nn_clusters=4).reshape(-1),\\nname=\\'Complete-PCA\\' )\\npd.crosstab(nci_labs[ \\'label\\' ], pca_labels)\\n12.6 Exercises\\nConceptual\\n1. This problem involves the K-means clustering algorithm.\\n(a) Prove ( 12.18).\\n(b) On the basis of this identity, argue that theK-means clusteringalgorithm (Algorithm12.2) decreases the objective (12.17) ateach iteration.2. Suppose that we have four observations, for which we compute adissimilarity matrix, given by\\uf8ee\\uf8ef\\uf8ef\\uf8f00.30.40.70.30.50.80.40.50.450.70.80.45\\uf8f9\\uf8fa\\uf8fa\\uf8fb.For instance, the dissimilarity between the first and second obser-vations is 0.3, and the dissimilarity between the second and fourthobservations is 0.8.(a) On the basis of this dissimilarity matrix, sketch the dendrogramthat results from hierarchically clustering these four observa-tions using complete linkage. Be sure to indicate on the plot theheight at which each fusion occurs, as well as the observationscorresponding to each leaf in the dendrogram.(b) Repeat (a), this time using single linkage clustering.(c) Suppose that we cut the dendrogram obtained in (a) such thattwo clusters result. Which observations are in each cluster?(d) Suppose that we cut the dendrogram obtained in (b) such thattwo clusters result. Which observations are in each cluster?', metadata={'source': './documents/ISLP_website.pdf', 'page': 557}), Document(page_content='12.6 Exercises 553\\n(e) It is mentioned in this chapter that at each fusion in the den-\\ndrogram, the position of the two clusters being fused can be\\nswappedwithoutchangingthemeaningofthedendrogram.Draw\\na dendrogram that is equivalent to the dendrogram in (a), for\\nwhich two or more of the leaves are repositioned, but for which\\nthe meaning of the dendrogram is the same.\\n3. In this problem, you will perform K-means clustering manually, with\\nK=2, on a small example with n=6 observations and p=2\\nfeatures. The observations are as follows.\\nObs.X1X2\\n114\\n213\\n304\\n451\\n562\\n640\\n(a) Plot the observations.\\n(b) Randomlyassignaclusterlabeltoeachobservation.Youcanuse\\nthenp.random.choice() function to do this. Report the cluster\\nlabels for each observation.\\n(c) Compute the centroid for each cluster.\\n(d) Assign each observation to the centroid to which it is closest, in\\nterms of Euclidean distance. Report the cluster labels for each\\nobservation.\\n(e) Repeat (c) and (d) until the answers obtained stop changing.\\n(f) In your plot from (a), color the observations according to the\\ncluster labels obtained.\\n4. Suppose that for a particular data set, we perform hierarchical clus-\\ntering using single linkage and using complete linkage. We obtain two\\ndendrograms.\\n(a) At a certain point on the single linkage dendrogram, the clus-\\nters{1,2,3}and{4,5}fuse. On the complete linkage dendro-\\ngram, the clusters {1,2,3}and{4,5}also fuse at a certain point.\\nWhich fusion will occur higher on the tree, or will they fuse at\\nthe same height, or is there not enough information to tell?\\n(b) At a certain point on the single linkage dendrogram, the clusters\\n{5}and{6}fuse. On the complete linkage dendrogram, the clus-\\nters{5}and{6}also fuse at a certain point. Which fusion will\\noccur higher on the tree, or will they fuse at the same height, or\\nis there not enough information to tell?\\n5. In words, describe the results that you would expect if you performed\\nK-means clustering of the eight shoppers in Figure 12.16, on the\\nbasis of their sock and computer purchases, with K=2. Give three\\nanswers, one for each of the variable scalings displayed. Explain.', metadata={'source': './documents/ISLP_website.pdf', 'page': 558}), Document(page_content='554 12. Unsupervised Learning\\n6. We saw in Section 12.2.2that the principal component loading and\\nscore vectors provide an approximation to a matrix, in the sense of\\n(12.5).Specifically,theprincipalcomponentscoreandloadingvectors\\nsolve the optimization problem given in ( 12.6).\\nNow,supposethatthe Mprincipalcomponentscorevectors zim,m=\\n1,...,M , are known. Using ( 12.6), explain that each of the first M\\nprincipal component loading vectors φjm,m=1,...,M , can be ob-\\ntained by performing pseparate least squares linear regressions. In\\neach regression, the principal component score vectors are the pre-\\ndictors, and one of the features of the data matrix is the response.\\nApplied\\n7. In this chapter, we mentioned the use of correlation-based distance\\nandEuclidean distance as dissimilaritymeasures for hierarchicalclus-\\ntering. It turns out that these two measures are almost equivalent: if\\neach observation has been centered to have mean zero and standard\\ndeviation one, and if we let rijdenote the correlation between the ith\\nandjth observations, then the quantity 1−rijis proportional to the\\nsquared Euclidean distance between the ith andjth observations.\\nOn the USArrests data, show that this proportionality holds.\\nHint: The Euclidean distance can be calculated using the\\npairwise_distances() function from the sklearn.metrics module, andpairwise_\\ndistances() correlations can be calculated using the np.corrcoef() function.\\n8. In Section 12.2.3, a formula for calculating PVE was given in Equa-\\ntion12.10. We also saw that the PVE can be obtained using the\\nexplained_variance_ratio_ attribute of a fitted PCA()estimator.\\nOn the USArrests data, calculate PVE in two ways:\\n(a) Using the explained_variance_ratio_ output of the fitted PCA()\\nestimator, as was done in Section 12.2.3.\\n(b) By applying Equation 12.10directly. The loadings are stored\\nas thecomponents_ attribute of the fitted PCA()estimator. Use\\nthose loadings in Equation 12.10to obtain the PVE.\\nThese two approaches should give the same results.\\nHint: You will only obtain the same results in (a) and (b) if the same\\ndata is used in both cases. For instance, if in (a) you performed PCA()\\nusing centered and scaled variables, then you must center and scale\\nthe variables before applying Equation 12.10in (b).\\n9. Consider the USArrests data. We will now perform hierarchical clus-\\ntering on the states.\\n(a) Using hierarchical clustering with complete linkage and\\nEuclidean distance, cluster the states.\\n(b) Cut the dendrogram at a height that results in three distinct\\nclusters. Which states belong to which clusters?', metadata={'source': './documents/ISLP_website.pdf', 'page': 559}), Document(page_content='12.6 Exercises 555(c) Hierarchically cluster the states using complete linkage and Eu-clidean distance,after scaling the variables to have standard de-viation one.(d) What effect does scaling the variables have on the hierarchicalclustering obtained? In your opinion, should the variables bescaled before the inter-observation dissimilarities are computed?Provide a justification for your answer.10. In this problem, you will generate simulated data, and then performPCA andK-means clustering on the data.(a) Generate a simulated data set with 20 observations in each ofthree classes (i.e. 60 observations total), and 50 variables.Hint: There are a number of functions inPythonthat you canuse to generate data. One example is thenormal()method oftherandom()function innumpy; theuniform()method is anotheroption. Be sure to add a mean shift to the observations in eachclass so that there are three distinct classes.(b) Perform PCA on the 60 observations and plot the first two prin-cipal component score vectors. Use a different color to indicatethe observations in each of the three classes. If the three classesappear separated in this plot, then continue on to part (c). Ifnot, then return to part (a) and modify the simulation so thatthere is greater separation between the three classes. Do notcontinue to part (c) until the three classes show at least someseparation in the first two principal component score vectors.(c) PerformK-means clustering of the observations withK=3.How well do the clusters that you obtained inK-means cluster-ing compare to the true class labels?Hint: You can use thepd.crosstab()function inPythonto com-pare the true class labels to the class labels obtained by cluster-ing. Be careful how you interpret the results:K-means clusteringwill arbitrarily number the clusters, so you cannot simply checkwhether the true class labels and clustering labels are the same.(d) PerformK-means clustering withK=2. Describe your results.(e) Now performK-means clustering withK=4, and describe yourresults.(f) Now performK-means clustering withK=3on the first twoprincipal component score vectors, rather than on the raw data.That is, performK-means clustering on the60×2matrix ofwhich the first column is the first principal component scorevector,andthesecondcolumnisthesecondprincipalcomponentscore vector. Comment on the results.(g) Using theStandardScaler()estimator, performK-means clus-tering withK=3on the dataafter scaling each variable to havestandard deviation one. How do these results compare to thoseobtained in (b)? Explain.', metadata={'source': './documents/ISLP_website.pdf', 'page': 560}), Document(page_content='556 12. Unsupervised Learning\\n11. Write a Pythonfunction to perform matrix completion as in Algo-\\nrithm12.1, and as outlined in Section 12.5.2. In each iteration, the\\nfunction should keep track of the relative error, as well as the itera-\\ntion count. Iterations should continue until the relative error is small\\nenough or until some maximum number of iterations is reached (set a\\ndefault value for this maximum number). Furthermore, there should\\nbe an option to print out the progress in each iteration.\\nTest your function on the Bostondata. First, standardize the features\\nto have mean zero and standard deviation one using the\\nStandardScaler() function. Run an experiment where you randomly\\nleave out an increasing (and nested) number of observations from 5%\\nto 30%, in steps of 5%. Apply Algorithm 12.1withM=1,2,...,8.\\nDisplay the approximation error as a function of the fraction of ob-\\nservations that are missing, and the value of M, averaged over 10\\nrepetitions of the experiment.\\n12. In Section 12.5.2, Algorithm 12.1was implemented using the\\nsvd()function from the np.linalg module. However, given the con-\\nnection between the svd()function and the PCA()estimator high-\\nlighted in the lab, we could have instead implemented the algorithm\\nusingPCA().\\nWriteafunctiontoimplementAlgorithm 12.1thatmakesuseof PCA()\\nrather than svd().\\n13. On the book website, www.statlearning.com , there is a gene expres-\\nsion data set ( Ch12Ex13.csv ) that consists of 40 tissue samples with\\nmeasurements on 1,000 genes. The first 20 samples are from healthy\\npatients, while the second 20 are from a diseased group.\\n(a) Load in the data using pd.read_csv() . You will need to select\\nheader = None .\\n(b) Apply hierarchical clustering to the samples using correlation-\\nbased distance, and plot the dendrogram. Do the genes separate\\nthe samples into the two groups? Do your results depend on the\\ntype of linkage used?\\n(c) Your collaborator wants to know which genes differ the most\\nacross the two groups. Suggest a way to answer this question,\\nand apply it here.', metadata={'source': './documents/ISLP_website.pdf', 'page': 561}), Document(page_content='13\\nMultiple Testing\\nThus far, this textbook has mostly focused on estimation and its close\\ncousin,prediction . In this chapter, we instead focus on hypothesis testing,\\nwhich is key to conducting inference . We remind the reader that inference\\nwas briefly discussed in Chapter 2.\\nWhile Section 13.1provides a brief review of null hypotheses, p-values,\\ntest statistics, and other key ideas in hypothesis testing, this chapter as-\\nsumes that the reader has had previous exposure to these topics. In par-\\nticular, we will not focus on whyorhowto conduct a hypothesis test — a\\ntopic on which entire books can be (and have been) written! Instead, we\\nwill assume that the reader is interested in testing some particular set of\\nnull hypotheses, and has a specific plan in mind for how to conduct the\\ntests and obtain p-values.\\nMuchoftheemphasisinclassicalstatisticsfocusesontestingasinglenull\\nhypothesis, such as H0: the expected blood pressure of mice in the control\\ngroup equals the expected blood pressure of mice in the treatment group . Of\\ncourse,wewouldprobablyliketodiscoverthatthere isadifferencebetween\\nthemeanbloodpressureinthetwogroups.Butforreasonsthatwillbecome\\nclear, we construct a null hypothesis corresponding to no difference.\\nIn contemporary settings, we are often faced with huge amounts of\\ndata, and consequently may wish to test a great many null hypotheses.\\nFor instance, rather than simply testing H0, we might want to test m\\nnull hypotheses, H01,...,H 0m, whereH0j: the expected value of the jth\\nbiomarker among mice in the control group equals the expected value of the\\njthbiomarker among mice in the treatment group . When conducting mul-\\ntiple testing , we need to be very careful about how we interpret the results,\\nin order to avoid erroneously rejecting far too many null hypotheses.\\nThis chapter discusses classical as well as more contemporary ways to\\nconduct multiple testing in a big-data setting. In Section 13.2, we highlight\\nthe challenges associated with multiple testing. Classical solutions to these\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning , Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0_13 557', metadata={'source': './documents/ISLP_website.pdf', 'page': 562}), Document(page_content='558 13. Multiple Testing\\nchallenges are presented in Section 13.3, and more contemporary solutions\\nin Sections 13.4and13.5.\\nIn particular, Section 13.4focuses on the false discovery rate. The no-\\ntion of the false discovery rate dates back to the 1990s. It quickly rose in\\npopularity in the early 2000s, when large-scale data sets began to come out\\nof genomics. These datasets were unique not only because of their large\\nsize,1but also because they were typically collected for exploratory pur-\\nposes: researchers collected these datasets in order to test a huge number\\nof null hypotheses, rather than just a very small number of pre-specified\\nnull hypotheses. Today, of course, huge datasets are collected without a\\npre-specified null hypothesis across virtually all fields. As we will see, the\\nfalse discovery rate is perfectly-suited for this modern-day reality.\\nThis chapter naturally centers upon the classical statistical technique of\\np-values, used to quantify the results of hypothesis tests. At the time of\\nwriting of this book (2020), p-values have recently been the topic of exten-\\nsive commentary in the social science research community, to the extent\\nthat some social science journals have gone so far as to ban the use of\\np-values altogether! We will simply comment that when properly under-\\nstood and applied, p-values provide a powerful tool for drawing inferential\\nconclusions from our data.\\n13.1 A Quick Review of Hypothesis Testing\\nHypothesis tests provide a rigorous statistical framework for answering\\nsimple “yes-or-no” questions about data, such as the following:\\n1. Is the true coefficient βjin a linear regression of YontoX1,...,X p\\nequal to zero?2\\n2. Is there a difference in the expected blood pressure of laboratory mice\\nin the control group and laboratory mice in the treatment group?3\\nIn Section 13.1.1, we briefly review the steps involved in hypothesis test-\\ning. Section 13.1.2discusses the different types of mistakes, or errors, that\\ncan occur in hypothesis testing.\\n13.1.1 Testing a Hypothesis\\nConducting a hypothesis test typically proceeds in four steps. First, we de-\\nfine the null and alternative hypotheses. Next, we construct a test statistic\\nthat summarizes the strength of evidence against the null hypothesis. We\\nthen compute a p-value that quantifies the probability of having obtained\\n1Microarraydatawasviewedas“bigdata”atthetime,althoughbytoday’sstandards,\\nthis label seems quaint: a microarray dataset can be (and typically was) stored in a\\nMicrosoft Excel spreadsheet!\\n2This hypothesis test was discussed on page 76of Chapter 3.\\n3The “treatment group” refers to the set of mice that receive an experimental treat-\\nment, and the “control group” refers to those that do not.', metadata={'source': './documents/ISLP_website.pdf', 'page': 563}), Document(page_content='13.1 A Quick Review of Hypothesis Testing 559\\na comparable or more extreme value of the test statistic under the null\\nhypothesis. Finally, based on the p-value, we decide whether to reject the\\nnull hypothesis. We now briefly discuss each of these steps in turn.\\nStep 1: Define the Null and Alternative Hypotheses\\nIn hypothesis testing, we divide the world into two possibilities: the null\\nhypothesis and thealternative hypothesis . The null hypothesis, denoted H0,null\\nhypothesis\\nalternative\\nhypothesisis the default state of belief about the world.4For instance, null hypotheses\\nassociatedwiththetwoquestionsposedearlierinthischapterareasfollows:\\n1. The true coefficient βjin a linear regression of YontoX1,...,X p\\nequals zero.\\n2. There is no difference between the expected blood pressure of mice\\nin the control and treatment groups.\\nThe null hypothesis is boring by construction: it may well be true, but we\\nmight hope that our data will tell us otherwise.\\nThe alternative hypothesis, denoted Ha, represents something different\\nand unexpected: for instance, that there isa difference between the ex-\\npected blood pressure of the mice in the two groups. Typically, the alter-\\nnative hypothesis simply posits that the null hypothesis does not hold: if\\nthe null hypothesis states that there is no difference between A and B , then\\nthe alternative hypothesis states that there is a difference between A and\\nB.\\nIt is important to note that the treatment of H0andHais asymmetric.\\nH0is treated as the default state of the world, and we focus on using data\\nto reject H0. If we reject H0, then this provides evidence in favor of Ha.W e\\ncan think of rejecting H0as making a discovery about our data: namely, we\\nare discovering that H0does not hold! By contrast, if we fail to reject H0,\\nthen our findings are more nebulous: we will not know whether we failed\\nto reject H0because our sample size was too small (in which case testing\\nH0again on a larger or higher-quality dataset might lead to rejection), or\\nwhether we failed to reject H0becauseH0really holds.\\nStep 2: Construct the Test Statistic\\nNext, we wish to use our data in order to find evidence for or against\\nthe null hypothesis. In order to do this, we must compute a test statistic ,test statisticdenoted T, which summarizes the extent to which our data are consistent\\nwithH0. The way in which we construct Tdepends on the nature of the\\nnull hypothesis that we are testing.\\nTo make things concrete, let xt\\n1,...,xt\\nntdenote the blood pressure mea-\\nsurementsforthe ntmiceinthetreatmentgroup,andlet xc\\n1,...,xc\\nncdenote\\nthe blood pressure measurements for the ncmice in the control group, and\\nµt= E( Xt),µc= E( Xc).TotestH0:µt=µc,wemakeuseofa two-sample\\nt-statistic ,5defined astwo-sample\\nt-statistic\\n4H0is pronounced “H naught” or “H zero”.', metadata={'source': './documents/ISLP_website.pdf', 'page': 564}), Document(page_content='560 13. Multiple Testing\\n−4−20240.0 0.1 0.2 0.3 0.4Value of Test StatisticProbability Density FunctionT=2.33\\nFIGURE 13.1. The density function for the N(0,1)distribution, with the ver-\\ntical line indicating a value of 2.33. 1% of the area under the curve falls to the\\nright of the vertical line, so there is only a 2% chance of observing a N(0,1)value\\nthat is greater than 2.33or less than −2.33. Therefore, if a test statistic has a\\nN(0,1)null distribution, then an observed test statistic of T=2.33leads to a\\np-value of 0.02.\\nT=ˆµt−ˆµc\\ns√\\n1\\nnt+1\\nnc(13.1)\\nwhereˆµt=1\\nnt∑nt\\ni=1xt\\ni,ˆµc=1\\nnc∑nc\\ni=1xc\\ni, and\\ns=√\\n(nt−1)s2\\nt+(nc−1)s2c\\nnt+nc−2(13.2)\\nis an estimator of the pooled standard deviation of the two samples.6Here,\\ns2\\ntands2\\ncare unbiased estimators of the variance of the blood pressure in\\nthe treatment and control groups, respectively. A large (absolute) value of\\nTprovides evidence against H0:µt=µc, and hence evidence in support\\nofHa:µt̸=µc.\\nStep 3: Compute the p-Value\\nIn the previous section, we noted that a large (absolute) value of a two-\\nsamplet-statisticprovidesevidenceagainst H0.Thisbegsthequestion: how\\nlarge is large? In other words, how much evidence against H0is provided\\nby a given value of the test statistic?\\nThe notion of a p-valueprovides us with a way to formalize as well asp-valueanswer this question. The p-value is defined as the probability of observing\\na test statistic equal to or more extreme than the observed statistic, under\\nthe assumption that H0is in fact true . Therefore, a small p-value provides\\nevidence againstH0.\\n5Thet-statistic derives its name from the fact that, under H0, it follows a t-\\ndistribution.\\n6Note that ( 13.2) assumes that the control and treatment groups have equal variance.\\nWithout this assumption, ( 13.2) would take a slightly different form.', metadata={'source': './documents/ISLP_website.pdf', 'page': 565}), Document(page_content='13.1 A Quick Review of Hypothesis Testing 561\\nTo make this concrete, suppose that T=2.33for the test statistic in\\n(13.1). Then, we can ask: what is the probability of having observed such\\na large value of T, if indeed H0holds? It turns out that under H0, the\\ndistribution of Tin (13.1) follows approximately a N(0,1)distribution7—\\nthat is, a normal distribution with mean 0and variance 1. This distribution\\nis displayed in Figure 13.1. We see that the vast majority — 98% — of the\\nN(0,1)distribution falls between −2.33and2.33. This means that under\\nH0, we would expect to see such a large value of |T|only 2% of the time.\\nTherefore, the p-value corresponding to T=2.33is0.02.\\nThe distribution of the test statistic under H0(also known as the test\\nstatistic’s null distribution ) will depend on the details of what type ofnull\\ndistributionnull hypothesis is being tested, and what type of test statistic is used. In\\ngeneral, most commonly-used test statistics follow a well-known statistical\\ndistribution under the null hypothesis — such as a normal distribution,\\nat-distribution, a χ2-distribution, or an F-distribution — provided that\\nthe sample size is sufficiently large and that some other assumptions hold.\\nTypically, the Rfunction that is used to compute a test statistic will make\\nuse of this null distribution in order to output a p-value. In Section 13.5,\\nwe will see an approach to estimate the null distribution of a test statistic\\nusing re-sampling; in many contemporary settings, this is a very attractive\\noption, as it exploits the availability of fast computers in order to avoid\\nhaving to make potentially problematic assumptions about the data.\\nThep-value is perhaps one of the most used and abused notions in all of\\nstatistics.Inparticular,itissometimessaidthatthe p-valueistheprobabil-\\nity thatH0holds, i.e., that the null hypothesis is true. This is not correct!\\nThe one and only correct interpretation of the p-value is as the fraction\\nof the time that we would expect to see such an extreme value of the test\\nstatistic8if we repeated the experiment many many times, provided H0\\nholds.\\nIn Step 2 we computed a test statistic, and noted that a large (absolute)\\nvalue of the test statistic provides evidence against H0. In Step 3 the test\\nstatistic was converted to a p-value, with small p-values providing evidence\\nagainstH0. What, then, did we accomplish by converting the test statistic\\nfrom Step 2 into a p-value in Step 3? To answer this question, suppose\\na data analyst conducts a statistical test, and reports a test statistic of\\nT= 17 .3. Does this provide strong evidence against H0? It’s impossible\\nto know, without more information: in particular, we would need to know\\n7More precisely, assuming that the observations are drawn from a normal distribution,\\nthenTfollows a t-distribution with nt+nc−2degrees of freedom. Provided that nt+\\nnc−2is larger than around 40, this is very well-approximated by a N(0,1)distribution.\\nIn Section 13.5, we will see an alternative and often more attractive way to approximate\\nthe null distribution of T, which avoids making stringent assumptions about the data.\\n8Aone-sided p-value is the probability of seeing such an extreme value of the test\\nstatistic; e.g. the probability of seeing a test statistic greater than or equal to T=2.33.\\nAtwo-sided p-value is the probability of seeing such an extreme value of the absolute\\ntest statistic; e.g. the probability of seeing a test statistic greater than or equal to 2.33\\nor less than or equal to −2.33. The default recommendation is to report a two-sided\\np-value rather than a one-sided p-value, unless there is a clear and compelling reason\\nthat only one direction of the test statistic is of scientific interest.', metadata={'source': './documents/ISLP_website.pdf', 'page': 566}), Document(page_content='562 13. Multiple Testing\\nTruth\\nH0 Ha\\nDecisionRejectH0 Type I Error Correct\\nDo Not Reject H0 Correct Type II Error\\nTABLE 13.1. A summary of the possible scenarios associated with testing the\\nnull hypothesis H0. Type I errors are also known as false positives, and Type II\\nerrors as false negatives.\\nwhatvalueoftheteststatisticshouldbeexpected,under H0.Thisisexactly\\nwhat ap-value gives us. In other words, a p-value allows us to transform\\nour test statistic, which is measured on some arbitrary and uninterpretable\\nscale, into a number between 0and1that can be more easily interpreted.\\nStep 4: Decide Whether to Reject the Null Hypothesis\\nOnce we have computed a p-value corresponding to H0, it remains for\\nus to decide whether or not to reject H0. (We do not usually talk about\\n“accepting” H0: instead, we talk about “failing to reject” H0.) A small p-\\nvalue indicates that such a large value of the test statistic is unlikely to\\noccur under H0, and thereby provides evidence against H0. If thep-value\\nis sufficiently small, then we will want to reject H0(and, therefore, make\\na “discovery”). But how small is small enough to reject H0?\\nIt turns out that the answer to this question is very much in the eyes\\nof the beholder, or more specifically, the data analyst. The smaller the p-\\nvalue, the stronger the evidence against H0. In some fields, it is typical to\\nrejectH0if thep-value is below 0.05; this means that, if H0holds, we would\\nexpect to see such a small p-value no more than 5% of the time.9However,\\nin other fields, a much higher burden of proof is required: for example, in\\nsome areas of physics, it is typical to reject H0only if the p-value is below\\n10−9!\\nIn the example displayed in Figure 13.1, if we use a threshold of 0.05as\\nour cut-off for rejecting the null hypothesis, then we will reject the null. By\\ncontrast, if we use a threshold of 0.01, then we will fail to reject the null.\\nThese ideas are formalized in the next section.\\n13.1.2 Type I and Type II Errors\\nIf the null hypothesis holds, then we say that it is a true null hypothesis ;true null\\nhypothesisotherwise, it is a false null hypothesis . For instance, if we test H0:µt=µc\\nfalse null\\nhypothesisas in Section 13.1.1, and there is indeed no difference in the population\\nmean blood pressure for mice in the treatment group and mice in the\\ncontrol group, then H0is true; otherwise, it is false. Of course, we do not\\nknowa prioriwhether H0is true or whether it is false: this is why we need\\nto conduct a hypothesis test!\\n9Though a threshold of 0.05to reject H0is ubiquitous in some areas of science, we\\nadvise against blind adherence to this arbitrary choice. Furthermore, a data analyst\\nshould typically report the p-value itself, rather than just whether or not it exceeds a\\nspecified threshold value.', metadata={'source': './documents/ISLP_website.pdf', 'page': 567}), Document(page_content='13.2 The Challenge of Multiple Testing 563\\nTable13.1summarizes the possible scenarios associated with testing the\\nnull hypothesis H0.10Once the hypothesis test is performed, the rowof the\\ntable is known (based on whether or not we have rejected H0); however, it\\nis impossible for us to know which columnwe are in. If we reject H0when\\nH0is false (i.e., when Hais true), or if we do not reject H0when it is true,\\nthen we arrived at the correct result. However, if we erroneously reject H0\\nwhenH0is in fact true, then we have committed a Type I error . TheType IType I errorerror rate is defined as the probability of making a Type I error given thatType I error\\nrateH0holds, i.e., the probability of incorrectly rejecting H0. Alternatively, if\\nwe do not reject H0whenH0is in fact false, then we have committed a\\nType II error . Thepowerof the hypothesis test is defined as the probabilityType II\\nerror\\npowerof not making a Type II error given that Haholds, i.e., the probability of\\ncorrectly rejecting H0.\\nIdeally we would like both the Type I and Type II error rates to be small.\\nBut in practice, this is hard to achieve! There typically is a trade-off: we\\ncan make the Type I error small by only rejecting H0if we are quite sure\\nthat it doesn’t hold; however, this will result in an increase in the Type II\\nerror. Alternatively, we can make the Type II error small by rejecting H0\\nin the presence of even modest evidence that it does not hold, but this will\\ncause the Type I error to be large. In practice, we typically view Type I\\nerrors as more “serious” than Type II errors, because the former involves\\ndeclaring a scientific finding that is not correct. Hence, when we perform\\nhypothesis testing, we typically require a low Type I error rate — e.g.,\\nat most α=0.05— while trying to make the Type II error small (or,\\nequivalently, the power large).\\nIt turns out that there is a direct correspondence between the p-value\\nthreshold that causes us to reject H0, and the Type I error rate. By only\\nrejecting H0when the p-value is below α, we ensure that the Type I error\\nrate will be less than or equal to α.\\n13.2 The Challenge of Multiple Testing\\nIn the previous section, we saw that rejecting H0if thep-value is below\\n(say)0.01provides us with a simple way to control the Type I error for H0\\nat level0.01: ifH0is true, then there is no more than a 1% probability that\\nwe will reject it. But now suppose that we wish to test mnull hypotheses,\\nH01,...,H 0m. Will it do to simply reject all null hypotheses for which the\\ncorresponding p-value falls below (say) 0.01? Stated another way, if we\\nreject all null hypotheses for which the p-value falls below 0.01, then how\\nmany Type I errors should we expect to make?\\nAs a first step towards answering this question, consider a stockbroker\\nwho wishes to drum up new clients by convincing them of her trading\\n10There are parallels between Table 13.1and Table 4.6, which has to do with the\\noutput of a binary classifier. In particular, recall from Table 4.6that a false positive\\nresults from predicting a positive (non-null) label when the true label is in fact negative\\n(null). This is closely related to a Type I error, which results from rejecting the null\\nhypothesis when in fact the null hypothesis holds.', metadata={'source': './documents/ISLP_website.pdf', 'page': 568}), Document(page_content='564 13. Multiple Testing\\nacumen. She tells 1,024 ( 1,024 = 210) potential new clients that she can\\ncorrectlypredictwhetherApple’sstockpricewillincreaseordecreasefor10\\ndays running. There are 210possibilities for how Apple’s stock price might\\nchange over the course of these 10 days. Therefore, she emails each client\\none of these 210possibilities. The vast majority of her potential clients\\nwill find that the stockbroker’s predictions are no better than chance (and\\nmany will find them to be even worse than chance). But a broken clock is\\nright twice a day, and one of her potential clients will be really impressed\\nto find that her predictions were correct for all 10 of the days! And so the\\nstockbroker gains a new client.\\nWhat happened here? Does the stockbroker have any actual insight into\\nwhether Apple’s stock price will increase or decrease? No. How, then, did\\nshe manage to predict Apple’s stock price perfectly for 10 days running?\\nThe answer is that she made a lot of guesses, and one of them happened\\nto be exactly right.\\nHow does this relate to multiple testing? Suppose that we flip 1,024 fair\\ncoins11ten times each. Then we would expect (on average) one coin to\\ncome up all tails. (There’s a 1/210=1/1,024chance that any single coin\\nwill come up all tails. So if we flip 1,024coins, then we expect one coin to\\ncome up all tails, on average.) If one of our coins comes up all tails, then\\nwe might therefore conclude that this particular coin is not fair. In fact, a\\nstandard hypothesis test for the null hypothesis that this particular coin\\nis fair would lead to a p-value below 0.002!12But it would be incorrect to\\nconclude that the coin is not fair: in fact, the null hypothesis holds, and we\\njust happen to have gotten ten tails in a row by chance.\\nThese examples illustrate the main challenge of multiple testing : whenmultiple\\ntestingtesting a huge number of null hypotheses, we are bound to get some very\\nsmallp-valuesbychance.Ifwemakeadecisionaboutwhethertorejecteach\\nnull hypothesis without accounting for the fact that we have performed a\\nvery large number of tests, then we may end up rejecting a great number\\nof true null hypotheses — that is, making a large number of Type I errors.\\nHow severe is the problem? Recall from the previous section that if we\\nreject a single null hypothesis, H0, if itsp-value is less than, say, α=0.01,\\nthen there is a 1% chance of making a false rejection if H0is in fact true.\\nNowwhat if wetest mnullhypotheses, H01,...,H 0m,all of whicharetrue?\\nThere’s a 1% chance of rejecting any individual null hypothesis; therefore,\\nwe expect to falsely reject approximately 0.01×mnull hypotheses. If m=\\n10,000, then that means that we expect to falsely reject 100null hypotheses\\nby chance! That is a lotof Type I errors.\\nThecruxoftheissueisasfollows:rejectinganullhypothesisifthe p-value\\nis below αcontrols the probability of falsely rejecting that null hypothesis\\nat levelα. However, if we do this for mnull hypotheses, then the chance of\\nfalsely rejecting at least one of the mnull hypotheses is quite a bit higher!\\n11Afair coin is one that has an equal chance of landing heads or tails.\\n12Recall that the p-value is the probability of observing data at least this extreme,\\nunder the null hypothesis. If the coin is fair, then the probability of observing at least\\nten tails is (1/2)10=1/1,024<0.001. Thep-value is therefore 2/1,024<0.002, since\\nthis is the probability of observing ten heads or ten tails.', metadata={'source': './documents/ISLP_website.pdf', 'page': 569}), Document(page_content='13.3 The Family-Wise Error Rate 565\\nH0is True H0is False Total\\nRejectH0 V S R\\nDo Not Reject H0 U W m−R\\nTotal m0 m−m0 m\\nTABLE 13.2. A summary of the results of testing mnull hypotheses. A given\\nnull hypothesis is either true or false, and a test of that null hypothesis can either\\nreject or fail to reject it. In practice, the individual values of V,S,U, andWare\\nunknown. However, we do have access to V+S=RandU+W=m−R, which\\nare the numbers of null hypotheses rejected and not rejected, respectively.\\nWe will investigate this issue in greater detail, and pose a solution to it, in\\nSection13.3.\\n13.3 The Family-Wise Error Rate\\nIn the following sections, we will discuss testing multiple hypotheses while\\ncontrolling the probability of making at least one Type I error.\\n13.3.1 What is the Family-Wise Error Rate?\\nRecall that the Type I error rate is the probability of rejecting H0ifH0is\\ntrue. The family-wise error rate (FWER) generalizes this notion to the set-family-wise\\nerror rateting ofmnull hypotheses, H01,...,H 0m, and is defined as the probability\\nof making at least one Type I error. To state this idea more formally, con-\\nsider Table 13.2, which summarizes the possible outcomes when performing\\nmhypothesis tests. Here, Vrepresents the number of Type I errors (also\\nknown as false positives or false discoveries), Sthe number of true posi-\\ntives,Uthe number of true negatives, and Wthe number of Type II errors\\n(also known as false negatives). Then the family-wise error rate is given by\\nFWER= Pr( V≥1). (13.3)\\nA strategy of rejecting any null hypothesis for which the p-value is below\\nα(i.e. controlling the Type I error for each null hypothesis at level α) leads\\nto a FWER of\\nFWER(α)=1 −Pr(V= 0)\\n=1 −Pr(do not falsely reject any null hypotheses )\\n=1 −Pr(⋂m\\nj=1{do not falsely reject H0j})\\n. (13.4)\\nRecall from basic probability that if two events AandBare independent,\\nthenPr(A∩B) = Pr( A) Pr( B). Therefore, if we make the additional rather\\nstrong assumptions that the mtests are independent and that all mnull\\nhypotheses are true, then\\nFWER(α)=1 −m∏\\nj=1(1−α)=1 −(1−α)m. (13.5)', metadata={'source': './documents/ISLP_website.pdf', 'page': 570}), Document(page_content='566 13. Multiple Testing\\nFIGURE 13.2. The family-wise error rate, as a function of the number of\\nhypotheses tested (displayed on the log scale), for three values of α:α=0.05\\n(orange), α=0.01(blue), and α=0.001(purple). The dashed line indicates\\n0.05. For example, in order to control the FWER at 0.05when testing m= 50\\nnull hypotheses, we must control the Type I error for each null hypothesis at level\\nα=0.001.\\nHence, if we test only one null hypothesis, then FWER (α)=1 −(1−α)1=\\nα,sotheTypeIerrorrateandtheFWERareequal.However,ifweperform\\nm= 100 independent tests, then FWER (α)=1 −(1−α)100. For instance,\\ntakingα=0.05leads to a FWER of 1−(1−0.05)100=0.994. In other\\nwords, we are virtually guaranteed to make at least one Type I error!\\nFigure13.2displays ( 13.5) for various values of m, the number of hy-\\npotheses, and α, the Type I error. We see that setting α=0.05results in\\na high FWER even for moderate m. Withα=0.01, we can test no more\\nthan five null hypotheses before the FWER exceeds 0.05. Only for very\\nsmall values, such as α=0.001, do we manage to ensure a small FWER,\\nat least for moderately-sized m.\\nWe now briefly return to the example in Section 13.1.1, in which we\\nconsider testing a single null hypothesis of the form H0:µt=µcusing a\\ntwo-sample t-statistic. Recall from Figure 13.1that in order to guarantee\\nthat the Type I error does not exceed 0.02, we decide whether or not to\\nrejectH0using a cutpoint of 2.33(i.e. we reject H0if|T|≥2.33). Now,\\nwhat if we wish to test 10 null hypotheses using two-sample t-statistics,\\ninstead of just one? We will see in Section 13.3.2that we can guarantee\\nthat the FWER does not exceed 0.02by rejecting only null hypotheses\\nfor which the p-value falls below 0.002. This corresponds to a much more\\nstringent cutpoint of 3.09(i.e. we should reject H0jonly if its test statistic\\n|Tj|≥3.09, forj=1,...,10). In other words, controlling the FWER at\\nlevelαamounts to a much higher bar, in terms of evidence required to\\nreject any given null hypothesis, than simply controlling the Type I error\\nfor each null hypothesis at level α.1 2 5 10 20 50 100 200 5000.0 0.2 0.4 0.6 0.8 1.0Number of HypothesesFamily−Wise Error Rateα=0.05α=0.01α=0.001', metadata={'source': './documents/ISLP_website.pdf', 'page': 571}), Document(page_content='13.3 The Family-Wise Error Rate 567\\nManager Mean,¯xStandard Deviation, st-statistic p-value\\nOne 3.0 7.4 2.86 0.006\\nTwo -0.1 6.9 -0.10 0.918\\nThree 2.8 7.5 2.62 0.012\\nFour 0.5 6.7 0.53 0.601\\nFive 0.3 6.8 0.31 0.756\\nTABLE 13.3. The first two columns correspond to the sample mean and sample\\nstandard deviation of the percentage excess return, over n= 50 months, for the\\nfirst five managers in the Funddataset. The last two columns provide the t-statistic\\n(√n·¯X/S) and associated p-value for testing H0j:µj=0, the null hypothesis\\nthat the (population) mean return for the jth hedge fund manager equals zero.\\n13.3.2 Approaches to Control the Family-Wise Error Rate\\nIn this section, we briefly survey some approaches to control the FWER.\\nWe will illustrate these approaches on the Funddataset, which records the\\nmonthly percentage excess returns for 2,000 fund managers over n= 50\\nmonths.13Table13.3provides relevant summary statistics for the first five\\nmanagers.\\nWe first present the Bonferroni method and Holm’s step-down proce-\\ndure, which are very general-purpose approaches for controlling the FWER\\nthat can be applied whenever mp-values have been computed, regardless\\nof the form of the null hypotheses, the choice of test statistics, or the\\n(in)dependence of the p-values. We then briefly discuss Tukey’s method\\nand Scheffé’s method in order to illustrate the fact that, in certain sit-\\nuations, more specialized approaches for controlling the FWER may be\\npreferable.\\nThe Bonferroni Method\\nAs in the previous section, suppose we wish to test H01,...,H 0m. LetAj\\ndenote the event that we make a Type I error for the jth null hypothesis,\\nforj=1,...,m . Then\\nFWER= Pr( falsely reject at least one null hypothesis )\\n= Pr( ∪m\\nj=1Aj)\\n≤m∑\\nj=1Pr(Aj). (13.6)\\nIn (13.6), the inequality results from the fact that for any two events A\\nandB,Pr(A∪B)≤Pr(A) + Pr( B), regardless of whether AandBare\\nindependent. The Bonferroni method , orBonferroni correction , sets the\\nthreshold for rejecting each hypothesis test to α/m, so that Pr(Aj)≤α/m.\\nEquation 13.6implies that\\nFWER(α/m)≤m×α\\nm=α,\\n13Excessreturnscorrespondtotheadditionalreturnthefundmanagerachievesbeyond\\nthe market’s overall return. So if the market increases by 5%during a given period and\\nthe fund manager achieves a 7%return, their excess return would be 7%−5% = 2% .', metadata={'source': './documents/ISLP_website.pdf', 'page': 572}), Document(page_content='568 13. Multiple Testing\\nso this procedure controls the FWER at level α. For instance, in order to\\ncontrol the FWER at level 0.1while testing m= 100 null hypotheses, the\\nBonferroni procedure requires us to control the Type I error for each null\\nhypothesis at level 0.1/100 = 0 .001, i.e. to reject all null hypotheses for\\nwhich the p-value is below 0.001.\\nWe now consider the Funddataset in Table 13.3. If we control the Type\\nI error at level α=0.05for each fund manager separately, then we will\\nconclude that the first and third managers have significantly non-zero ex-\\ncess returns; in other words, we will reject H01:µ1=0andH03:µ3=0.\\nHowever, as discussed in previous sections, this procedure does not account\\nfor the fact that we have tested multiple hypotheses, and therefore it will\\nlead to a FWER greater than 0.05. If we instead wish to control the FWER\\nat level0.05, then, using a Bonferroni correction, we must control the Type\\nI error for each individual manager at level α/m=0.05/5=0 .01. Conse-\\nquently, we will reject the null hypothesis only for the first manager, since\\nthep-values for all other managers exceed 0.01. The Bonferroni correction\\ngives us peace of mind that we have not falsely rejected too many null\\nhypotheses, but for a price: we reject few null hypotheses, and thus will\\ntypically make quite a few Type II errors.\\nThe Bonferroni correction is by far the best-known and most commonly-\\nused multiplicity correction in all of statistics. Its ubiquity is due in large\\npart to the fact that it is very easy to understand and simple to implement,\\nand also from the fact that it successfully controls Type I error regardless\\nof whether the mhypothesis tests are independent. However, as we will see,\\nit is typically neither the most powerful nor the best approach for multiple\\ntesting correction. In particular, the Bonferroni correction can be quite\\nconservative, in the sense that the true FWER is often quite a bit lower\\nthan the nominal (or target) FWER; this results from the inequality in\\n(13.6). By contrast, a less conservative procedure might allow us to control\\nthe FWER while rejecting more null hypotheses, and therefore making\\nfewer Type II errors.\\nHolm’s Step-Down Procedure\\nHolm’s method , also known as Holm’s step-down procedure or the Holm–Holm’s\\nmethodBonferroni method, is an alternative to the Bonferroni procedure. Holm’s\\nmethod controls the FWER, but it is less conservative than Bonferroni, in\\nthesensethatitwillrejectmorenullhypotheses,typicallyresultinginfewer\\nType II errors and hence greater power. The procedure is summarized in\\nAlgorithm 13.1. The proof that this method controls the FWER is similar\\nto, but slightly more complicated than, the argument in ( 13.6) that the\\nBonferroni method controls the FWER. It is worth noting that in Holm’s\\nprocedure, the threshold that we use to reject each null hypothesis — p(L)\\nin Step 5 — actually depends on the values of allmof thep-values. (See the\\ndefinition of Lin (13.7).) This is in contrast to the Bonferroni procedure,\\nin which to control the FWER at level α, we reject any null hypotheses for\\nwhich the p-value is below α/m, regardless of the other p-values. Holm’s\\nmethod makes no independence assumptions about the mhypothesis tests,\\nand is uniformly more powerful than the Bonferroni method — it will', metadata={'source': './documents/ISLP_website.pdf', 'page': 573}), Document(page_content='13.3 The Family-Wise Error Rate 569\\nAlgorithm 13.1 Holm’s Step-Down Procedure to Control the FWER\\n1. Specify α, the level at which to control the FWER.\\n2. Compute p-values,p1,...,p m, for the mnull hypotheses\\nH01,...,H 0m.\\n3. Order the mp-values so that p(1)≤p(2)≤···≤p(m).\\n4. Define\\nL= min{\\nj:p(j)>α\\nm+1−j}\\n. (13.7)\\n5. Reject all null hypotheses H0jfor which pj<p(L).\\nalways reject at least as many null hypotheses as Bonferroni — and so it\\nshould always be preferred.\\nWe now consider applying Holm’s method to the first five fund managers\\nin theFunddataset in Table 13.3, while controlling the FWER at level 0.05.\\nThe ordered p-values are p(1)=0.006,p(2)=0.012,p(3)=0.601,p(4)=\\n0.756 andp(5)=0.918. The Holm procedure rejects the first two null\\nhypotheses, because p(1)=0.006<0.05/(5 + 1 −1) = 0 .01andp(2)=\\n0.012<0.05/(5 + 1 −2) = 0 .0125 , butp(3)=0.601>0.05/(5 + 1 −3) =\\n0.0167 , which implies that L=3. We note that, in this setting, Holm is\\nmore powerful than Bonferroni: the former rejects the null hypotheses for\\nthe first and third managers, whereas the latter rejects the null hypothesis\\nonly for the first manager.\\nFigure13.3provides an illustration of the Bonferroni and Holm methods\\non three simulated data sets in a setting involving m= 10 hypothesis tests,\\nof which m0=2 of the null hypotheses are true. Each panel displays the\\nten corresponding p-values, ordered from smallest to largest, and plotted\\non a log scale. The eight red points represent the false null hypotheses, and\\nthe two black points represent the true null hypotheses. We wish to control\\nthe FWER at level 0.05. The Bonferroni procedure requires us to reject all\\nnull hypotheses for which the p-value is below 0.005; this is represented by\\nthe black horizontal line. The Holm procedure requires us to reject all null\\nhypotheses that fall below the blue line. The blue line always lies above the\\nblack line, so Holm will always reject more tests than Bonferroni; the region\\nbetween the two lines corresponds to the hypotheses that are only rejected\\nby Holm. In the left-hand panel, both Bonferroni and Holm successfully\\nreject seven of the eight false null hypotheses. In the center panel, Holm\\nsuccessfully rejects all eight of the false null hypotheses, while Bonferroni\\nfails to reject one. In the right-hand panel, Bonferroni only rejects three of\\nthe false null hypotheses, while Holm rejects all eight. Neither Bonferroni\\nnor Holm makes any Type I errors in these examples.\\nTwo Special Cases: Tukey’s Method and Scheffé’s Method\\nBonferroni’s method and Holm’s method can be used in virtually any set-\\nting in which we wish to control the FWER for mnull hypotheses: they', metadata={'source': './documents/ISLP_website.pdf', 'page': 574}), Document(page_content='570 13. Multiple Testing\\n2 4 6 8 101e−05 1e−04 1e−03 1e−02 1e−01Ordering of p−valuesp−values (log scale)2 4 6 8 101e−07 1e−05 1e−03 1e−01Ordering of p−valuesp−values (log scale)2 4 6 8 101e−05 1e−04 1e−03 1e−02 1e−01Ordering of p−valuesp−values (log scale)FIGURE 13.3. Each panel displays, for a separate simulation, the sorted\\np-values for tests of m= 10 null hypotheses. The p-values corresponding to\\nthem0=2 true null hypotheses are displayed in black, and the rest are in red.\\nWhen controlling the FWER at level 0.05, the Bonferroni procedure rejects all\\nnull hypotheses that fall below the black line, and the Holm procedure rejects all\\nnull hypotheses that fall below the blue line. The region between the blue and black\\nlines indicates null hypotheses that are rejected using the Holm procedure but not\\nusing the Bonferroni procedure. In the center panel, the Holm procedure rejects\\none more null hypothesis than the Bonferroni procedure. In the right-hand panel,\\nit rejects five more null hypotheses.\\nmake no assumptions about the nature of the null hypotheses, the type\\nof test statistic used, or the (in)dependence of the p-values. However, in\\ncertain very specific settings, we can achieve higher power by controlling\\nthe FWER using approaches that are more tailored to the task at hand.\\nTukey’s method and Scheffé’s method provide two such examples.\\nTable13.3indicates that for the Funddataset, Managers One and Two\\nhave the greatest difference in their sample mean returns. This finding\\nmight motivate us to test the null hypothesis H0:µ1=µ2, whereµjis the\\n(population) mean return for the jth fund manager. A two-sample t-test\\n(13.1) forH0yields ap-value of 0.0349 , suggesting modest evidence against\\nH0. However, this p-value is misleading, since we decided to compare the\\naverage returns of Managers One and Two only after having examined the\\nreturns for all five managers; this essentially amounts to having performed\\nm=5×(5−1)/2 = 10 hypothesis tests, and selecting the one with the\\nsmallest p-value. This suggests that in order to control the FWER at level\\n0.05, we should make a Bonferroni correction for m= 10 hypothesis tests,\\nand therefore should only reject a null hypothesis for which the p-value\\nis below 0.005. If we do this, then we will be unable to reject the null\\nhypothesis that Managers One and Two have identical performance.\\nHowever, in this setting, a Bonferroni correction is actually a bit too\\nstringent, since it fails to consider the fact that the m= 10 hypothesis\\ntests are all somewhat related: for instance, Managers Two and Five have\\nsimilar mean returns, as do Managers Two and Four; this guarantees that\\nthe mean returns of Managers Four and Five are similar. Stated another\\nway, the mp-values for the mpairwise comparisons are notindependent.\\nTherefore, it should be possible to control the FWER in a way that is', metadata={'source': './documents/ISLP_website.pdf', 'page': 575}), Document(page_content='13.3 The Family-Wise Error Rate 571\\n2 4 6 8 10 12 141e−04 1e−03 1e−02 1e−01 1e+00Ordering of p−valuesp−values (log scale)2 4 6 8 10 12 141e−06 1e−04 1e−02 1e+00Ordering of p−valuesp−values (log scale)2 4 6 8 10 12 141e−04 1e−03 1e−02 1e−01 1e+00Ordering of p−valuesp−values (log scale)FIGURE 13.4. Each panel displays, for a separate simulation, the sorted\\np-values for tests of m= 15 hypotheses, corresponding to pairwise tests for the\\nequality of G=6means. The m0= 10 true null hypotheses are displayed in black,\\nand the rest are in red. When controlling the FWER at level 0.05, the Bonferroni\\nprocedure rejects all null hypotheses that fall below the black line, whereas Tukey\\nrejects all those that fall below the blue line. Thus, Tukey’s method has slightly\\nhigher power than Bonferroni’s method. Controlling the Type I error without\\nadjusting for multiple testing involves rejecting all those that fall below the green\\nline.\\nless conservative. This is exactly the idea behind Tukey’s method : whenTukey’s\\nmethodperforming m=G(G−1)/2pairwise comparisons of Gmeans, it allows\\nus to control the FWER at level αwhile rejecting all null hypotheses for\\nwhich the p-value falls below αT, for some αT>α/m.\\nFigure13.4illustrates Tukey’s method on three simulated data sets in a\\nsetting with G=6means, with µ1=µ2=µ3=µ4=µ5̸=µ6. Therefore,\\nof them=G(G−1)/2 = 15 null hypotheses of the form H0:µj=µk,\\nten are true and five are false. In each panel, the true null hypotheses\\nare displayed in black, and the false ones are in red. The horizontal lines\\nindicate that Tukey’s method always results in at least as many rejections\\nas Bonferroni’s method. In the left-hand panel, Tukey correctly rejects two\\nmore null hypotheses than Bonferroni.\\nNow, suppose that we once again examine the data in Table 13.3, and no-\\nticethatManagersOneandThreehavehighermeanreturnsthanManagers\\nTwo, Four, and Five. This might motivate us to test the null hypothesis\\nH0:1\\n2(µ1+µ3)=1\\n3(µ2+µ4+µ5). (13.8)\\n(Recall that µjis the population mean return for the jth hedge fund man-\\nager.) It turns out that we could test ( 13.8) using a variant of the two-\\nsamplet-test presented in ( 13.1), leading to a p-value of 0.004. This sug-\\ngests strong evidence of a difference between Managers One and Three\\ncompared to Managers Two, Four, and Five. However, there is a problem:\\nwe decided to test the null hypothesis in ( 13.8) only after peeking at the\\ndata in Table 13.3. In a sense, this means that we have conducted multiple\\ntesting. In this setting, using Bonferroni to control the FWER at level α', metadata={'source': './documents/ISLP_website.pdf', 'page': 576}), Document(page_content='572 13. Multiple Testing\\nwould require a p-value threshold of α/m, for an extremely large value of\\nm14.\\nScheffé’s method is designed for exactly this setting. It allows us to com-Scheffé’s\\nmethodpute a value αSsuch that rejecting the null hypothesis H0in (13.8) if the\\np-value is below αSwill control the Type I error at level α. It turns out that\\nfor theFundexample, in order to control the Type I error at level α=0.05,\\nwe must set αS=0.002. Therefore, we are unable to reject H0in (13.8), de-\\nspite the apparently very small p-value of 0.004. An important advantage of\\nScheffé’s method is that we can use this same threshold of αS=0.002in or-\\nder to perform a pairwise comparison of any split of the managers into two\\ngroups: for instance, we could also test H0:1\\n3(µ1+µ2+µ3)=1\\n2(µ4+µ5)\\nandH0:1\\n4(µ1+µ2+µ3+µ4)= µ5using the same threshold of 0.002,\\nwithout needing to further adjust for multiple testing.\\nTo summarize, Holm’s procedure and Bonferroni’s procedure are very\\ngeneral approaches for multiple testing correction that can be applied un-\\nder all circumstances. However, in certain special cases, more powerful pro-\\ncedures for multiple testing correction may be available, in order to control\\nthe FWER while achieving higher power (i.e. committing fewer Type II\\nerrors) than would be possible using Holm or Bonferroni. In this section,\\nwe have illustrated two such examples.\\n13.3.3 Trade-Off Between the FWER and Power\\nIngeneral,thereisatrade-offbetweentheFWERthresholdthatwechoose,\\nand our powerto reject the null hypotheses. Recall that power is defined\\nas the number of false null hypotheses that we reject divided by the total\\nnumber of false null hypotheses, i.e. S/(m−m0)using the notation of Ta-\\nble13.2. Figure13.5illustrates the results of a simulation setting involving\\nmnull hypotheses, of which 90% are true and the remaining 10% are false;\\npower is displayed as a function of the FWER. In this particular simulation\\nsetting, when m= 10 , a FWER of 0.05corresponds to power of approxi-\\nmately 60%. However, as mincreases, the power decreases. With m= 500 ,\\nthe power is below 0.2at a FWER of 0.05, so that we successfully reject\\nonly 20% of the false null hypotheses.\\nFigure13.5indicates that it is reasonable to control the FWER when m\\ntakes on a small value, like 5or10. However, for m= 100 orm=1,000,\\nattempting to control the FWER will make it almost impossible to reject\\nanyofthefalsenullhypotheses.Inotherwords,thepowerwillbeextremely\\nlow.\\nWhy is this the case? Recall that, using the notation in Table 13.2, the\\nFWER is defined as Pr(V≥1)(13.3). In other other words, controlling the\\nFWER at level αguarantees that the data analyst is very unlikely (with\\nprobability no more than α) to reject anytrue null hypotheses, i.e. to have\\nany false positives. In order to make good on this guarantee when mis\\nlarge, the data analyst may be forced to reject very few null hypotheses, or\\nperhaps even none at all (since if R=0then also V=0; see Table 13.2).\\n14In fact, calculating the “correct” value of mis quite technical, and outside the scope\\nof this book.', metadata={'source': './documents/ISLP_website.pdf', 'page': 577}), Document(page_content='13.4 The False Discovery Rate 573\\n0.00.20.40.60.81.00.0 0.2 0.4 0.6 0.8 1.0Family−Wise Error RatePowerm = 10m = 100m = 500FIGURE 13.5. In a simulation setting in which 90% of the mnull hypotheses are\\ntrue, we display the power (the fraction of false null hypotheses that we successfully\\nreject) as a function of the family-wise error rate. The curves correspond to\\nm= 10 (orange) ,m= 100 (blue), andm= 500 (purple). As the value of m\\nincreases, the power decreases. The vertical dashed line indicates a FWER of\\n0.05.\\nThis is scientifically uninteresting, and typically results in very low power,\\nas in Figure 13.5.\\nIn practice, when mis large, we may be willing to tolerate a few false\\npositives, in the interest of making more discoveries, i.e. more rejections of\\nthe null hypothesis. This is the motivation behind the false discovery rate,\\nwhich we present next.\\n13.4 The False Discovery Rate\\n13.4.1 Intuition for the False Discovery Rate\\nAs we just discussed, when mis large, then trying to prevent anyfalse\\npositives (as in FWER control) is simply too stringent. Instead, we might\\ntry to make sure that the ratio of false positives ( V) to total positives ( V+\\nS=R) is sufficiently low, so that most of the rejected null hypotheses are\\nnot false positives. The ratio V/Ris known as the false discovery proportionfalse\\ndiscovery\\nproportion(FDP).\\nIt might be tempting to ask the data analyst to control the FDP: to\\nmake sure that no more than, say, 20% of the rejected null hypotheses are\\nfalse positives. However, in practice, controlling the FDP is an impossible\\ntask for the data analyst, since she has no way to be certain, on any par-\\nticular dataset, which hypotheses are true and which are false. This is very\\nsimilar to the fact that the data analyst can control the FWER, i.e. she\\ncan guarantee that Pr(V≥1)≤αfor any pre-specified α, but she cannot\\nguarantee that V=0 on any particular dataset (short of failing to reject\\nany null hypotheses, i.e. setting R=0).', metadata={'source': './documents/ISLP_website.pdf', 'page': 578}), Document(page_content='574 13. Multiple Testing\\nTherefore, we instead control the false discovery rate (FDR)15, definedfalse\\ndiscovery\\nrateas\\nFDR= E( FDP) = E( V/R). (13.9)\\nWhen we control the FDR at (say) level q= 20% , we are rejecting as many\\nnull hypotheses as possible while guaranteeing that no more than 20% of\\nthose rejected null hypotheses are false positives, on average .\\nIn the definition of the FDR in ( 13.9), the expectation is taken over the\\npopulation from which the data are generated. For instance, suppose we\\ncontrol the FDR for mnull hypotheses at q=0.2. This means that if we\\nrepeat this experiment a huge number of times, and each time control the\\nFDRatq=0.2,thenweshouldexpectthat,onaverage,20%oftherejected\\nnull hypotheses will be false positives. On a given dataset, the fraction of\\nfalse positives among the rejected hypotheses may be greater than or less\\nthan 20%.\\nThus far, we have motivated the use of the FDR from a pragmatic per-\\nspective, by arguing that when mis large, controlling the FWER is simply\\ntoo stringent, and will not lead to “enough” discoveries. An additional mo-\\ntivation for the use of the FDR is that it aligns well with the way that data\\nare often collected in contemporary applications. As datasets continue to\\ngrow in size across a variety of fields, it is increasingly common to conduct a\\nhuge number of hypothesis tests for exploratory, rather than confirmatory,\\npurposes. For instance, a genomic researcher might sequence the genomes\\nof individuals with and without some particular medical condition, and\\nthen, for each of 20,000 genes, test whether sequence variants in that gene\\nare associated with the medical condition of interest. This amounts to per-\\nformingm= 20 ,000hypothesis tests. The analysis is exploratory in nature,\\nin the sense that the researcher does not have any particular hypothesis\\nin mind; instead she wishes to see whether there is modest evidence for\\nthe association between each gene and the disease, with a plan to further\\ninvestigate any genes for which there is such evidence. She is likely willing\\nto tolerate some number of false positives in the set of genes that she will\\ninvestigate further; thus, the FWER is not an appropriate choice. How-\\never, some correction for multiple testing is required: it would not be a\\ngood idea for her to simply investigate allgenes with p-values less than\\n(say) 0.05, since we would expect 1,000 genes to have such small p-values\\nsimply by chance, even if no genes are associated with the disease (since\\n0.05×20,000 = 1 ,000). Controlling the FDR for her exploratory analysis\\nat 20% guarantees that — on average — no more than 20% of the genes\\nthat she investigates further are false positives.\\nIt is worth noting that unlike p-values, for which a threshold of 0.05\\nis typically viewed as the minimum standard of evidence for a “positive”\\nresult, and a threshold of 0.01 or even 0.001 is viewed as much more com-\\npelling, there is no standard accepted threshold for FDR control. Instead,\\nthechoiceofFDRthresholdistypicallycontext-dependent,orevendataset-\\ndependent. For instance, the genomic researcher in the previous example\\nmight seek to control the FDR at a threshold of 10% if the planned follow-\\n15IfR=0, then we replace the ratio V/R with0, to avoid computing 0/0. Formally,\\nFDR= E( V/R|R>0) Pr( R>0).', metadata={'source': './documents/ISLP_website.pdf', 'page': 579}), Document(page_content='13.4 The False Discovery Rate 575\\nup analysis is time-consuming or expensive. Alternatively, a much larger\\nthreshold of 30% might be suitable if she plans an inexpensive follow-up\\nanalysis.\\n13.4.2 The Benjamini–Hochberg Procedure\\nWe now focus on the task of controlling the FDR: that is, deciding which\\nnull hypotheses to reject while guaranteeing that the FDR, E(V/R), is less\\nthan or equal to some pre-specified value q. In order to do this, we need\\nsome way to connect the p-values,p1,...,p m, from the mnull hypotheses\\nto the desired FDR value, q. It turns out that a very simple procedure,\\noutlined in Algorithm 13.2, can be used to control the FDR.\\nAlgorithm 13.2 Benjamini–Hochberg Procedure to Control the FDR\\n1. Specify q, the level at which to control the FDR.\\n2. Compute p-values,p1,...,p m, for the mnull hypotheses\\nH01,...,H 0m.\\n3. Order the mp-values so that p(1)≤p(2)≤···≤p(m).\\n4. Define\\nL= max {j:p(j)< qj/m }. (13.10)\\n5. Reject all null hypotheses H0jfor which pj≤p(L).\\nAlgorithm 13.2is known as the Benjamini–Hochberg procedure . The cruxBenjamini–\\nHochberg\\nprocedureof this procedure lies in ( 13.10). For example, consider again the first five\\nmanagers in the Funddataset, presented in Table 13.3. (In this example,\\nm=5, although typically we control the FDR in settings involving a much\\ngreater number of null hypotheses.) We see that p(1)=0.006<0.05×1/5,\\np(2)=0.012 <0.05×2/5,p(3)=0.601 >0.05×3/5,p(4)=0.756 >\\n0.05×4/5, andp(5)=0.918>0.05×5/5. Therefore, to control the FDR\\nat5%, we reject the null hypotheses that the first and third fund managers\\nperform no better than chance.\\nAs long as the mp-values are independent or only mildly dependent,\\nthen the Benjamini–Hochberg procedure guarantees16that\\nFDR≤q.\\nIn other words, this procedure ensures that, on average, no more than a\\nfractionqof the rejected null hypotheses are false positives. Remarkably,\\nthis holds regardless of how many null hypotheses are true, and regardless\\nof the distribution of the p-values for the null hypotheses that are false.\\nTherefore, the Benjamini–Hochberg procedure gives us a very easy way to\\ndetermine, given a set of mp-values, which null hypotheses to reject in\\norder to control the FDR at any pre-specified level q.\\n16However, the proof is well beyond the scope of this book.', metadata={'source': './documents/ISLP_website.pdf', 'page': 580}), Document(page_content='576 13. Multiple Testing\\nα=0.05 α=0.1 α=0.3\\n1 5 50 5001e−051e−031e−01\\nIndexP−Value\\n1 5 50 5001e−051e−031e−01\\nIndexP−Value\\n1 5 50 5001e−051e−031e−01\\nIndexP−Value\\nFIGURE 13.6. Each panel displays the same set of m=2,000orderedp-values\\nfor theFunddata. The green lines indicate the p-value thresholds corresponding\\nto FWER control, via the Bonferroni procedure, at levels α=0.05(left),α=0.1\\n(center), andα=0.3(right). The orange lines indicate the p-value thresholds\\ncorresponding to FDR control, via Benjamini–Hochberg, at levels q=0.05(left),\\nq=0.1(center), andq=0.3(right). When the FDR is controlled at level q=0.1,\\n146 null hypotheses are rejected (center); the corresponding p-values are shown\\nin blue. When the FDR is controlled at level q=0.3, 279 null hypotheses are\\nrejected (right); the corresponding p-values are shown in blue.\\nThere is a fundamental difference between the Bonferroni procedure of\\nSection13.3.2and the Benjamini–Hochberg procedure. In the Bonferroni\\nprocedure, in order to control the FWER for mnull hypotheses at level\\nα, we must simply reject null hypotheses for which the p-value is below\\nα/m. This threshold of α/mdoes not depend on anything about the data\\n(beyond the value of m), and certainly does not depend on the p-values\\nthemselves. By contrast, the rejection threshold used in the Benjamini–\\nHochberg procedure is more complicated: we reject all null hypotheses for\\nwhich the p-value is less than or equal to the Lth smallest p-value, where\\nLis itself a function of all mp-values, as in ( 13.10). Therefore, when con-\\nducting the Benjamini–Hochberg procedure, we cannot plan out in advance\\nwhat threshold we will use to reject p-values; we need to first see our data.\\nFor instance, in the abstract, there is no way to know whether we will reject\\na null hypothesis corresponding to a p-value of 0.01 when using an FDR\\nthreshold of 0.1withm= 100 ; the answer depends on the values of the\\notherm−1p-values. This property of the Benjamini–Hochberg procedure\\nis shared by the Holm procedure, which also involves a data-dependent\\np-value threshold.\\nFigure13.6displaystheresultsofapplyingtheBonferroniandBenjamini–\\nHochberg procedures on the Funddata set, using the full set of m=2,000\\nfund managers, of which the first five were displayed in Table 13.3. When\\nthe FWER is controlled at level 0.3using Bonferroni, only one null hypoth-\\nesis is rejected; that is, we can conclude only that a single fund manager is\\nbeating the market. This is despite the fact that a substantial portion of', metadata={'source': './documents/ISLP_website.pdf', 'page': 581}), Document(page_content='13.5 A Re-Sampling Approach top-Values and False Discovery Rates 577them=2,000fund managers appear to have beaten the market withoutperforming correction for multiple testing — for instance, 13 of them havep-values below0.001. By contrast, when the FDR is controlled at level0.3,we can conclude that 279 fund managers are beating the market: we expectthatnomorethanaround279×0.3 = 83.7ofthesefundmanagershadgoodperformance only due to chance. Thus, we see that FDR control is muchmilder — and more powerful — than FWER control, in the sense that itallows us to reject many more null hypotheses, with a cost of substantiallymore false positives.The Benjamini–Hochberg procedure has been around since the mid-1990s.Whileagreatmanypapershavebeenpublishedsincethenproposingalternative approaches for FDR control that can perform better in partic-ular scenarios, the Benjamini–Hochberg procedure remains a very usefuland widely-applicable approach.13.5 A Re-Sampling Approach top-Values andFalse Discovery RatesThus far, the discussion in this chapter has assumed that we are interestedin testing a particular null hypothesisH0using a test statisticT, whichhas some known (or assumed) distribution underH0, such as a normaldistribution, at-distribution, aχ2-distribution, or anF-distribution. Thisis referred to as thetheoretical null distribution. We typically rely upontheoreticalnulldistributionthe availability of a theoretical null distribution in order to obtain ap-value associated with our test statistic. Indeed, for most of the types ofnull hypotheses that we might be interested in testing, a theoretical nulldistribution is available, provided that we are willing to make stringentassumptions about our data.However, if our null hypothesisH0or test statisticTis somewhat un-usual, then it may be the case that no theoretical null distribution is avail-able. Alternatively, even if a theoretical null distribution exists, then wemay be wary of relying upon it, perhaps because some assumption that isrequired for it to hold is violated. For instance, maybe the sample size istoo small.In this section, we present a framework for performing inference in thissetting, which exploits the availability of fast computers in order to approx-imate the null distribution ofT, and thereby to obtain ap-value. While thisframework is very general, it must be carefully instantiated for a specificproblem of interest. Therefore, in what follows, we consider a specific ex-ample in which we wish to test whether the means of two random variablesare equal, using a two-samplet-test.The discussion in this section is more challenging than the precedingsections in this chapter, and can be safely skipped by a reader who iscontent to use the theoretical null distribution to computep-values for hisor her test statistics.', metadata={'source': './documents/ISLP_website.pdf', 'page': 582}), Document(page_content='578 13. Multiple Testing\\n13.5.1 A Re-Sampling Approach to the p-Value\\nWereturntotheexampleofSection 13.1.1,inwhichwewishtotestwhether\\nthe mean of a random variable Xequals the mean of a random variable Y,\\ni.e.H0: E(X) = E( Y), against the alternative Ha: E(X)̸= E( Y). Given\\nnXindependent observations from XandnYindependent observations\\nfromY, the two-sample t-statistic takes the form\\nT=ˆµX−ˆµY\\ns√\\n1\\nnX+1\\nnY(13.11)\\nwhereˆµX=1\\nnX∑nX\\ni=1xi,ˆµY=1\\nnY∑nY\\ni=1yi,s=√\\n(nX−1)s2\\nX+(nY−1)s2\\nY\\nnX+nY−2,\\nands2\\nXands2\\nYare unbiased estimators of the variances in the two groups.\\nA large (absolute) value of Tprovides evidence against H0.\\nIfnXandnYare large, then Tin (13.11) approximately follows a N(0,1)\\ndistribution. But if nXandnYare small, then in the absence of a strong\\nassumption about the distribution of XandY, we do not know the the-\\noretical null distribution of T.17In this case, it turns out that we can\\napproximate the null distribution of Tusing are-sampling approach, orre-samplingmore specifically, a permutation approach.permutationTodothis,weconductathoughtexperiment.If H0holds,sothat E(X)=\\nE(Y), and we make the stronger assumption that the distributions of X\\nandYare the same, then the distribution of Tis invariant under swapping\\nobservations of Xwith observations of Y. That is, if we randomly swap\\nsome of the observations in Xwith the observations in Y, thenthe test\\nstatistic Tin(13.11)computed based on this swapped data has the same\\ndistribution as Tbased on the original data. This is true only if H0holds,\\nand the distributions of XandYare the same.\\nThis suggests that in order to approximate the null distribution of T,\\nwe can take the following approach. We randomly permute the nX+nY\\nobservations Btimes, for some large value of B, and each time we compute\\n(13.11). We let T∗1,...,T∗Bdenote the values of ( 13.11) on the permuted\\ndata. These can be viewed as an approximation of the null distribution\\nofTunderH0. Recall that by definition, a p-value is the probability of\\nobserving a test statistic at least this extreme under H0. Therefore, to\\ncompute a p-value for T, we can simply compute\\np-value=∑B\\nb=11(|T∗b|≥|T|)\\nB, (13.12)\\nthe fraction of permuted datasets for which the value of the test statistic\\nis at least as extreme as the value observed on the original data. This\\nprocedure is summarized in Algorithm 13.3.\\n17If we assume that XandYare normally distributed, then Tin (13.11) follows a\\nt-distribution with nX+nY−2degrees of freedom under H0. However, in practice, the\\ndistribution of random variables is rarely known, and so it can be preferable to perform\\na re-sampling approach instead of making strong and unjustified assumptions. If the\\nresults of the re-sampling approach disagree with the results of assuming a theoretical\\nnull distribution, then the results of the re-sampling approach are more trustworthy.', metadata={'source': './documents/ISLP_website.pdf', 'page': 583}), Document(page_content='13.5 A Re-Sampling Approach to p-Values and False Discovery Rates 579\\nAlgorithm 13.3 Re-Sampling p-Value for a Two-Sample t-Test\\n1. Compute T, defined in ( 13.11), on the original data x1,...,x nXand\\ny1,...,y nY.\\n2. Forb=1,...,B , whereBis a large number (e.g. B= 10 ,000):\\n(a) Permute the nX+nYobservations at random. Call the first nX\\npermuted observations x∗\\n1,...,x∗\\nnX, and call the remaining nY\\nobservations y∗\\n1,...,y∗\\nnY.\\n(b) Compute ( 13.11) on the permuted data x∗\\n1,...,x∗\\nnXand\\ny∗\\n1,...,y∗\\nnY, and call the result T∗b.\\n3. Thep-value is given by∑B\\nb=11(|T∗b|≥|T|)\\nB.\\nWe try out this procedure on the Khandataset, which consists of expres-\\nsion measurements for 2,308genes in four sub-types of small round blood\\ncell tumors, a type of cancer typically seen in children. This dataset is part\\nof theISLR2package. We restrict our attention to the two sub-types for\\nwhich the most observations are available: rhabdomyosarcoma ( nX= 29 )\\nand Burkitt’s lymphoma ( nY= 25 ).\\nA two-sample t-test for the null hypothesis that the 11th gene’s mean\\nexpression values are equal in the two groups yields T=−2.09. Using\\nthe theoretical null distribution, which is a t52distribution (since nX+\\nnY−2 = 52 ), we obtain a p-value of 0.041. (Note that a t52distribution\\nis virtually indistinguishable from a N(0,1)distribution.) If we instead\\napply Algorithm 13.3withB= 10 ,000, then we obtain a p-value of 0.042.\\nFigure13.7displays the theoretical null distribution, the re-sampling null\\ndistribution, and the actual value of the test statistic ( T=−2.09) for this\\ngene. In this example, we see very little difference between the p-values\\nobtained using the theoretical null distribution and the re-sampling null\\ndistribution.\\nBy contrast, Figure 13.8shows an analogous set of results for the 877th\\ngene. In this case, there is a substantial difference between the theoretical\\nand re-sampling null distributions, which results in a difference between\\ntheirp-values.\\nIn general, in settings with a smaller sample size or a more skewed data\\ndistribution (so that the theoretical null distribution is less accurate), the\\ndifference between the re-sampling and theoretical p-values will tend to\\nbe more pronounced. In fact, the substantial difference between the re-\\nsampling and theoretical null distributions in Figure 13.8is due to the\\nfact that a single observation in the 877th gene is very far from the other\\nobservations, leading to a very skewed distribution.\\n13.5.2 A Re-Sampling Approach to the False Discovery Rate\\nNow, suppose that we wish to control the FDR for mnull hypotheses,\\nH01,...,H 0m, in a setting in which either no theoretical null distribution\\nis available, or else we simply prefer to avoid the use of a theoretical null', metadata={'source': './documents/ISLP_website.pdf', 'page': 584}), Document(page_content='580 13. Multiple Testing\\nNull Distribution of Test Statistic for 11th Gene−4−20240 100 200 300 400T=−2.0936\\nFIGURE 13.7. The11th gene in the Khandataset has a test statistic of\\nT=−2.09. Its theoretical and re-sampling null distributions are almost identical.\\nThe theoretical p-value equals 0.041and the re-sampling p-value equals 0.042.\\nFIGURE 13.8. The877th gene in the Khandataset has a test statistic of\\nT=−0.57. Its theoretical and re-sampling null distributions are quite different.\\nThe theoretical p-value equals 0.571, and the re-sampling p-value equals 0.673.\\ndistribution.AsinSection 13.5.1,wemakeuseofatwo-sample t-statisticfor\\neach hypothesis, leading to the test statistics T1,...,T m. We could simply\\ncompute a p-value for each of the mnull hypotheses, as in Section 13.5.1,\\nand then apply the Benjamini–Hochberg procedure of Section 13.4.2to\\nthesep-values. However, it turns out that we can do this in a more direct\\nway, without even needing to compute p-values.\\nRecall from Section 13.4that the FDR is defined as E(V/R), using the\\nnotation in Table 13.2. In order to estimate the FDR via re-sampling, we\\nfirst make the following approximation:\\nFDR=E(V\\nR)\\n≈E(V)\\nR. (13.13)\\nNow suppose we reject any null hypothesis for which the test statistic\\nexceedscin absolute value. Then computing Rin the denominator on the\\nright-hand side of ( 13.13) is straightforward: R=∑m\\nj=11(|Tj|≥c).Null Distribution of Test Statistic for 877th Gene−4−20240 100 200 300 400T=−0.5696', metadata={'source': './documents/ISLP_website.pdf', 'page': 585}), Document(page_content='13.5 A Re-Sampling Approach to p-Values and False Discovery Rates 581\\nHowever, the numerator E(V)on the right-hand side of ( 13.13) is more\\nchallenging. This is the expected number of false positives associated with\\nrejecting any null hypothesis for which the test statistic exceeds cin abso-\\nlute value. At the risk of stating the obvious, estimating Vis challenging\\nbecause we do not know which of H01,...,H 0mare really true, and so we\\ndo not know which rejected hypotheses are false positives. To overcome this\\nproblem, we take a re-sampling approach, in which we simulate data under\\nH01,...,H 0m, and then compute the resulting test statistics. The number\\nof re-sampled test statistics that exceed cprovides an estimate of V.\\nIn greater detail, in the case of a two-sample t-statistic ( 13.11) for each\\nof the null hypotheses H01,...,H 0m, we can estimate E(V)as follows. Let\\nx(j)\\n1,...,x(j)\\nnXandy(j)\\n1,...,y(j)\\nnYdenote the data associated with the jth\\nnull hypothesis, j=1,...,m . We permute these nX+nYobservations at\\nrandom, and then compute the t-statistic on the permuted data. For this\\npermuted data, we know that all of the null hypotheses H01,...,H 0mhold;\\ntherefore,thenumberofpermuted t-statisticsthatexceedthethreshold cin\\nabsolute value provides an estimate for E(V). This estimate can be further\\nimproved by repeating the permutation process Btimes, for a large value\\nofB, and averaging the results.\\nAlgorithm 13.4details this procedure.18It provides what is known as a\\nplug-inestimate oftheFDR,becausetheapproximationin( 13.13)allowsus\\nto estimate the FDR by plugging Rinto the denominator and an estimate\\nforE(V)into the numerator.\\nWe apply the re-sampling approach to the FDR from Algorithm 13.4,\\nas well as the Benjamini–Hochberg approach from Algorithm 13.2using\\ntheoretical p-values, to the m=2,308genes in the Khandataset. Results are\\nshowninFigure 13.9.Weseethatforagivennumberofrejectedhypotheses,\\nthe estimated FDRs are almost identical for the two methods.\\nWe began this section by noting that in order to control the FDR for m\\nhypothesis tests using a re-sampling approach, we could simply compute m\\nre-sampling p-values as in Section 13.5.1, and then apply the Benjamini–\\nHochberg procedure of Section 13.4.2to thesep-values. It turns out that if\\nwe define the jth re-sampling p-value as\\npj=∑m\\nj′=1∑B\\nb=11(|T∗b\\nj′|≥|Tj|)\\nBm(13.14)\\nforj=1,...,m , instead of as in ( 13.12), then applying the Benjamini–\\nHochberg procedure to these re-sampled p-values is exactlyequivalent to\\nAlgorithm 13.4. Note that ( 13.14) is an alternative to ( 13.12) that pools\\nthe information across all mhypothesis tests in approximating the null\\ndistribution.\\n13.5.3 When Are Re-Sampling Approaches Useful?\\nIn Sections 13.5.1and13.5.2, we considered testing null hypotheses of the\\nformH0: E(X) = E( Y)using a two-sample t-statistic ( 13.11), for which we\\n18To implement Algorithm 13.4efficiently, the same set of permutations in Step 2(b)i.\\nshould be used for all mnull hypotheses.', metadata={'source': './documents/ISLP_website.pdf', 'page': 586}), Document(page_content='582 13. Multiple Testing\\nAlgorithm 13.4 Plug-In FDR for a Two-Sample T-Test\\n1. Select a threshold c, wherec>0.\\n2. Forj=1,...,m :\\n(a) Compute T(j), the two-sample t-statistic ( 13.11) for the null\\nhypothesis H0jon the basis of the original data, x(j)\\n1,...,x(j)\\nnX\\nandy(j)\\n1,...,y(j)\\nnY.\\n(b) Forb=1,...,B , whereBis a large number (e.g. B= 10 ,000):\\ni. Permute the nX+nYobservations at random. Call the first\\nnXobservations x∗(j)\\n1,...,x∗(j)\\nnX, and call the remaining ob-\\nservations y∗(j)\\n1,...,y∗(j)\\nnY.\\nii. Compute ( 13.11) on the permuted data x∗(j)\\n1,...,x∗(j)\\nnXand\\ny∗(j)\\n1,...,y∗(j)\\nnY, and call the result T(j),∗b.\\n3. Compute R=∑m\\nj=11(|T(j)|≥c).\\n4. Compute ˆV=∑B\\nb=1∑m\\nj=11(|T(j),∗b|≥c)\\nB.\\n5. The estimated FDR associated with the threshold cisˆV /R.\\napproximatedthenulldistributionviaare-samplingapproach.Wesawthat\\nusing the re-sampling approach gave us substantially different results from\\nusingthetheoretical p-valueapproachinFigure 13.8,butnotinFigure 13.7.\\nIn general, there are two settings in which a re-sampling approach is\\nparticularly useful:\\n1. Perhaps no theoretical null distribution is available. This may be the\\ncase if you are testing an unusual null hypothesis H0, or using an\\nunsual test statistic T.\\n2. Perhaps a theoretical null distribution isavailable, but the assump-\\ntions required for its validity do not hold. For instance, the two-\\nsamplet-statistic in ( 13.11) follows a tnX+nY−2distribution only if\\nthe observations are normally distributed. Furthermore, it follows a\\nN(0,1)distribution only if nXandnYare quite large. If the data are\\nnon-normal and nXandnYare small, then p-values that make use\\nof the theoretical null distribution will not be valid (i.e. they will not\\nproperly control the Type I error).\\nIn general, if you can come up with a way to re-sample or permute\\nyour observations in order to generate data that follow the null distribu-\\ntion, then you can compute p-values or estimate the FDR using variants\\nof Algorithms 13.3and13.4. In many real-world settings, this provides a\\npowerfultoolforhypothesistestingwhennoout-of-boxhypothesistestsare\\navailable, or when the key assumptions underlying those out-of-box tests\\nare violated.', metadata={'source': './documents/ISLP_website.pdf', 'page': 587}), Document(page_content='13.6 Lab: Multiple Testing 583\\n0500 1000 1500 20000.0 0.2 0.4 0.6 0.8 1.0Number of RejectionsFalse Discovery RateFIGURE 13.9. Forj=1,...,m =2,308, we tested the null hypothesis that for\\nthejth gene in the Khandataset, the meanexpression in Burkitt’s lymphoma equals\\nthemeanexpressioninrhabdomyosarcoma.Foreachvalueof kfrom1to2,308,the\\ny-axis displays the estimated FDR associated with rejecting the null hypotheses\\ncorresponding to the ksmallest p-values. The orange dashed curve shows the\\nFDR obtained using the Benjamini–Hochberg procedure, whereas the blue solid\\ncurve shows the FDR obtained using the re-sampling approach of Algorithm 13.4,\\nwithB= 10 ,000. There is very little difference between the two FDR estimates.\\nAccording to either estimate, rejecting the null hypothesis for the 500 genes with\\nthe smallest p-values corresponds to an FDR of around 17.7%.\\n13.6 Lab: Multiple Testing\\nWe include our usual imports seen in earlier labs.\\nIn[1]: import numpy as np\\nimport pandas as pd\\nimport matplotlib.pyplot as plt\\nimport statsmodels.api as sm\\nfrom ISLP import load_data\\nWe also collect the new imports needed for this lab.\\nIn[2]: from scipy.stats import \\\\\\n(ttest_1samp,\\nttest_rel,\\nttest_ind,\\nta st _ d b n )\\nfrom statsmodels.stats.multicomp import \\\\\\npairwise_tukeyhsd\\nfrom statsmodels.stats.multitest import \\\\\\nmultipletests as mult_test\\n13.6.1 Review of Hypothesis Tests\\nWe begin by performing some one-sample t-tests. First we create 100 vari-\\nables, each consisting of 10 observations. The first 50 variables have mean\\n0.5and variance 1, while the others have mean 0and variance 1.\\nIn[3]: rng = np.random.default_rng(12)\\nX=r n g . s t a n d a r d _ n o r m a l ( ( 1 0 ,1 0 0 ) )\\ntrue_mean = np.array([0.5]*50 + [0]*50)\\nX+ =t r u e _ m e a n [ N o n e , : ]', metadata={'source': './documents/ISLP_website.pdf', 'page': 588}), Document(page_content=\"584 13. Multiple Testing\\nTo begin, we use ttest_1samp() from the scipy.stats module to test H0:ttest_1samp()µ1=0, the null hypothesis that the first variable has mean zero.\\nIn[4]: result = ttest_1samp(X[:,0], 0)\\nresult.pvalue\\nOut[4]: 0.931\\nThep-value comes out to 0.931, which is not low enough to reject the null\\nhypothesis at level α=0.05. In this case, µ1=0.5, so the null hypothesis\\nis false. Therefore, we have made a Type II error by failing to reject the\\nnull hypothesis when the null hypothesis is false.\\nWe now test H0,j:µj=0forj=1,...,100. We compute the 100 p-\\nvalues, and then construct a vector recording whether the jthp-value is\\nless than or equal to 0.05, in which case we reject H0j, or greater than 0.05,\\nin which case we do not reject H0j, forj=1,...,100.\\nIn[5]: p_values = np.empty(100)\\nfor i in range(100):\\np_values[i] = ttest_1samp(X[:,i], 0).pvalue\\ndecision = pd.cut(p_values,\\n[0, 0.05, 1],\\nlabels=[ 'Reject H0' ,\\n'Do not reject H0' ])\\ntruth = pd.Categorical(true_mean == 0,\\ncategories=[True, False],\\nordered=True)\\nSince this is a simulated data set, we can create a 2×2table similar to\\nTable13.2.\\nIn[6]: pd.crosstab(decision,\\ntruth,\\nrownames=[ 'Decision' ],\\ncolnames=[ 'H0'])\\nOut[6]: H0 True False\\nDecision\\nReject H0 5 15\\nDo not reject H0 45 35\\nTherefore, at level α=0.05, we reject 15 of the 50 false null hypotheses,\\nand we incorrectly reject 5 of the true null hypotheses. Using the notation\\nfrom Section 13.3, we have V=5,S= 15,U= 45andW= 35. We have\\nsetα=0.05, which means that we expect to reject around 5% of the true\\nnull hypotheses. This is in line with the 2×2table above, which indicates\\nthat we rejected V=5of the50true null hypotheses.\\nIn the simulation above, for the false null hypotheses, the ratio of the\\nmean to the standard deviation was only 0.5/1=0.5. This amounts to\\nquite a weak signal, and it resulted in a high number of Type II errors.\\nLet’s instead simulate data with a stronger signal, so that the ratio of the\\nmean to the standard deviation for the false null hypotheses equals 1.W e\\nmake only 10 Type II errors.\", metadata={'source': './documents/ISLP_website.pdf', 'page': 589}), Document(page_content=\"13.6 Lab: Multiple Testing 585\\nIn[7]: true_mean = np.array([1]*50 + [0]*50)\\nX=r n g . s t a n d a r d _ n o r m a l ( ( 1 0 ,1 0 0 ) )\\nX+ =t r u e _ m e a n [ N o n e , : ]\\nfor i in range(100):\\np_values[i] = ttest_1samp(X[:,i], 0).pvalue\\ndecision = pd.cut(p_values,\\n[0, 0.05, 1],\\nlabels=[ 'Reject H0' ,\\n'Do not reject H0' ])\\ntruth = pd.Categorical(true_mean == 0,\\ncategories=[True, False],\\nordered=True)\\npd.crosstab(decision,\\ntruth,\\nrownames=[ 'Decision' ],\\ncolnames=[ 'H0'])\\nOut[7]: H0 True False\\nDecision\\nReject H0 2 40\\nDo not reject H0 48 10\\n13.6.2 Family-Wise Error Rate\\nRecall from ( 13.5) that if the null hypothesis is true for each of minde-\\npendent hypothesis tests, then the FWER is equal to 1−(1−α)m.W e\\ncan use this expression to compute the FWER for m=1,...,500and\\nα=0.05,0.01, and0.001. We plot the FWER for these values of αin order\\nto reproduce Figure 13.2.\\nIn[8]: m=n p . l i n s p a c e ( 1 ,5 0 1 )\\nfig, ax = plt.subplots()\\n[ax.plot(m,\\n1-( 1-a l p h a ) * * m ,\\nlabel=r'$\\\\alpha=%s$' %s t r ( a l p h a ) )\\nfor alpha in [0.05, 0.01, 0.001]]\\nax.set_xscale( 'log')\\nax.set_xlabel( 'Number of Hypotheses' )\\nax.set_ylabel( 'Family-Wise Error Rate' )\\nax.legend()\\nax.axhline(0.05, c= 'k',l s ='--');\\nAs discussed previously, even for moderate values of msuch as50, the\\nFWER exceeds 0.05unlessαis set to a very low value, such as 0.001. Of\\ncourse, the problem with setting αto such a low value is that we are likely\\nto make a number of Type II errors: in other words, our power is very low.\\nWe now conduct a one-sample t-test for each of the first five managers\\nin theFunddataset, in order to test the null hypothesis that the jth fund\\nmanager’s mean return equals zero, H0,j:µj=0.\\nIn[9]: Fund = load_data( 'Fund')\\nfund_mini = Fund.iloc[:,:5]\\nfund_mini_pvals = np.empty(5)\\nfor i in range(5):\", metadata={'source': './documents/ISLP_website.pdf', 'page': 590}), Document(page_content='586 13. Multiple Testingfund_mini_pvals[i] = ttest_1samp(fund_mini.iloc[:,i], 0).pvaluefund_mini_pvalsOut[9]:array([0.006, 0.918, 0.012, 0.601, 0.756])Thep-values are low for Managers One and Three, and high for the otherthree managers. However, we cannot simply rejectH0,1andH0,3, sincethis would fail to account for the multiple testing that we have performed.Instead,wewillconductBonferroni’smethodandHolm’smethodtocontrolthe FWER.To do this, we use themultipletests()function from thestatsmodelsmultiple-tests()module (abbreviated tomult_test()). Given thep-values, for methods likeHolm and Bonferroni the function outputsadjustedp-values, which can beadjustedp-valuesthought of as a new set ofp-values that have been corrected for multipletesting. If the adjustedp-value for a given hypothesis is less than or equaltoα, then that hypothesis can be rejected while maintaining a FWER ofno more thanα. In other words, for such methods, the adjustedp-valuesresulting from themultipletests()function can simply be compared tothe desired FWER in order to determine whether or not to reject eachhypothesis. We will later see that we can use the same function to controlFDR as well.Themult_test()function takesp-values and amethodargument, as wellas an optionalalphaargument. It returns the decisions (rejectbelow) aswell as the adjustedp-values (bonf).In[10]:reject, bonf = mult_test(fund_mini_pvals, method =\"bonferroni\")[:2]rejectOut[10]:array([ True, False, False, False, False])Thep-valuesbonfare simply thefund_mini_pvaluesmultiplied by 5 andtruncated to be less than or equal to 1.In[11]:bonf, np.minimum(fund_mini_pvals * 5, 1)Out[11]:(array([0.03, 1. , 0.06, 1. , 1. ]),array([0.03, 1. , 0.06, 1. , 1. ]))Therefore, using Bonferroni’s method, we are able to reject the null hy-pothesis only for Manager One while controlling FWER at0.05.By contrast, using Holm’s method, the adjustedp-values indicate thatwe can reject the null hypotheses for Managers One and Three at a FWERof0.05.In[12]:mult_test(fund_mini_pvals, method =\"holm\",a l p h a = 0 . 0 5 ) [ : 2 ]Out[12]:(array([ True, False, True, False, False]),array([0.03, 1. , 0.05, 1. , 1. ]))Asdiscussedpreviously,ManagerOneseemstoperformparticularlywell,whereas Manager Two has poor performance.', metadata={'source': './documents/ISLP_website.pdf', 'page': 591}), Document(page_content=\"13.6 Lab: Multiple Testing 587\\nIn[13]: fund_mini.mean()\\nOut[13]: Manager1 3.0\\nManager2 -0.1\\nManager3 2.8\\nManager4 0.5\\nManager5 0.3\\ndtype: float64\\nIsthereevidenceofameaningfuldifferenceinperformancebetweenthese\\ntwo managers? We can check this by performing a pairedt-testusing thepairedt-testttest_rel() function from scipy.stats :ttest_rel()\\nIn[14]: ttest_rel(fund_mini[ 'Manager1' ],\\nfund_mini[ 'Manager2' ]).pvalue\\nOut[14]: 0.038\\nThe test results in a p-value of 0.038, suggesting a statistically significant\\ndifference.\\nHowever, we decided to perform this test only after examining the data\\nand noting that Managers One and Two had the highest and lowest mean\\nperformances. In a sense, this means that we have implicitly performed(5\\n2)\\n= 5(5−1)/2 = 10hypothesis tests, rather than just one, as discussed\\nin Section 13.3.2. Hence, we use the pairwise_tukeyhsd() function frompairwise_\\ntukeyhsd()statsmodels.stats.multicomp to apply Tukey’s method in order to adjust\\nfor multiple testing. This function takes as input a fitted ANOVA regres-ANOVAsion model, which is essentially just a linear regression in which all of the\\npredictors are qualitative. In this case, the response consists of the monthly\\nexcess returns achieved by each manager, and the predictor indicates the\\nmanager to which each return corresponds.\\nIn[15]: returns = np.hstack([fund_mini.iloc[:,i] for i in range(5)])\\nmanagers = np.hstack([[i+1]*50 for i in range(5)])\\ntukey = pairwise_tukeyhsd(returns, managers)\\nprint(tukey.summary())\\nMultiple Comparison of Means - Tukey HSD, FWER=0.05\\n===================================================\\ngroup1 group2 meandiff p-adj lower upper reject\\n---------------------------------------------------\\n12 - 3 . 1 0 . 1 8 6 2 - 6 . 9 8 6 5 0 . 7 8 6 5 F a l s e\\n13 - 0 . 2 0 . 9 9 9 9 - 4 . 0 8 6 5 3 . 6 8 6 5 F a l s e\\n14 - 2 . 5 0 . 3 9 4 8 - 6 . 3 8 6 5 1 . 3 8 6 5 F a l s e\\n15 - 2 . 7 0 . 3 1 5 2 - 6 . 5 8 6 5 1 . 1 8 6 5 F a l s e\\n2 3 2.9 0.2453 -0.9865 6.7865 False\\n240 . 6 0 . 9 9 3 2 - 3 . 2 8 6 5 4 . 4 8 6 5 F a l s e\\n250 . 4 0 . 9 9 8 6 - 3 . 4 8 6 5 4 . 2 8 6 5 F a l s e\\n34 - 2 . 3 0 . 4 8 2 - 6 . 1 8 6 5 1 . 5 8 6 5 F a l s e\\n35 - 2 . 5 0 . 3 9 4 8 - 6 . 3 8 6 5 1 . 3 8 6 5 F a l s e\\n45 - 0 . 2 0 . 9 9 9 9 - 4 . 0 8 6 5 3 . 6 8 6 5 F a l s e\\n---------------------------------------------------\\nThepairwise_tukeyhsd() function provides confidence intervals for the\\ndifference between each pair of managers ( lowerandupper), as well as a\", metadata={'source': './documents/ISLP_website.pdf', 'page': 592}), Document(page_content='588 13. Multiple Testing\\nFIGURE 13.10. 95% confidence intervals for each manager on the Funddata,\\nusing Tukey’s method to adjust for multiple testing. All of the confidence intervals\\noverlap, so none of the differences among managers are statistically significant\\nwhen controlling FWER at level 0.05.\\np-value. All of these quantities have been adjusted for multiple testing.\\nNotice that the p-value for the difference between Managers One and Two\\nhas increased from 0.038to0.186, so there is no longer clear evidence of\\na difference between the managers’ performances. We can plot the confi-\\ndence intervals for the pairwise comparisons using the plot_simultaneous()\\nmethod of tukey. Any pair of intervals that don’t overlap indicates a sig-\\nnificant difference at the nominal level of 0.05. In this case, no differences\\nare considered significant as reported in the table above.\\nIn[16]: fig, ax = plt.subplots(figsize=(8,8))\\ntukey.plot_simultaneous(ax=ax);\\nThe result can be seen19in Figure 13.10.\\n13.6.3 False Discovery Rate\\nNow we perform hypothesis tests for all 2,000 fund managers in the Fund\\ndataset. We perform a one-sample t-test ofH0,j:µj=0, which states that\\nthejth fund manager’s mean return is zero.\\nIn[17]: fund_pvalues = np.empty(2000)\\nfor i, manager in enumerate(Fund.columns):\\nfund_pvalues[i] = ttest_1samp(Fund[manager], 0).pvalue\\nThere are far too many managers to consider trying to control the\\nFWER. Instead, we focus on controlling the FDR: that is, the expected\\nfraction of rejected null hypotheses that are actually false positives. The\\n19Traditionally this plot shows intervals for each paired difference. With many groups\\nit is more convenient and equivalent to display one interval per group, as is done here.\\nBy “differencing” all pairs of intervals displayed here you recover the traditional plot.', metadata={'source': './documents/ISLP_website.pdf', 'page': 593}), Document(page_content='13.6 Lab: Multiple Testing 589\\nmultipletests() function (abbreviated mult_test() ) can be used to carry\\nout the Benjamini–Hochberg procedure.\\nIn[18]: fund_qvalues = mult_test(fund_pvalues, method = \"fdr_bh\" )[1]\\nfund_qvalues[:10]\\nOut[18]: array([0.09, 0.99, 0.12, 0.92, 0.96, 0.08, 0.08, 0.08, 0.08,\\n0.08])\\nTheq-values output by the Benjamini–Hochberg procedure can be inter-q-valuespreted as the smallest FDR threshold at which we would reject a particular\\nnull hypothesis. For instance, a q-value of 0.1indicates that we can reject\\nthe corresponding null hypothesis at an FDR of 10% or greater, but that\\nwe cannot reject the null hypothesis at an FDR below 10%.\\nIf we control the FDR at 10%, then for how many of the fund managers\\ncan we reject H0,j:µj=0?\\nIn[19]: (fund_qvalues <= 0.1).sum()\\nOut[19]: 146\\nWe find that 146 of the 2,000 fund managers have a q-value below 0.1;\\ntherefore, we are able to conclude that 146 of the fund managers beat the\\nmarket at an FDR of 10%. Only about 15 (10% of 146) of these fund\\nmanagers are likely to be false discoveries.\\nBy contrast, if we had instead used Bonferroni’s method to control the\\nFWER at level α=0.1, then we would have failed to reject any null\\nhypotheses!\\nIn[20]: (fund_pvalues <= 0.1 / 2000).sum()\\nOut[20]: 0\\nFigure13.6displays the ordered p-values,p(1)≤p(2)≤···≤p(2000), for\\ntheFunddataset, as well as the threshold for rejection by the Benjamini–\\nHochberg procedure. Recall that the Benjamini–Hochberg procedure iden-\\ntifies the largest p-value such that p(j)< qj/m, and rejects all hypotheses\\nfor which the p-value is less than or equal to p(j). In the code below, we im-\\nplement the Benjamini–Hochberg procedure ourselves, in order to illustrate\\nhow it works. We first order the p-values. We then identify all p-values that\\nsatisfyp(j)< qj/m (sorted_set_ ). Finally, selected_ is a boolean array\\nindicating which p-values are less than or equal to the largest p-value in\\nsorted_[sorted_set_] . Therefore, selected_ indexes the p-values rejected\\nby the Benjamini–Hochberg procedure.\\nIn[21]: sorted_ = np.sort(fund_pvalues)\\nm=f u n d _ p v a l u e s . s h a p e [ 0 ]\\nq=0 . 1\\nsorted_set_ = np.where(sorted_ < q * np.linspace(1, m, m) / m)[0]\\nif sorted_set_.shape[0] > 0:\\nselected_ = fund_pvalues < sorted_[sorted_set_].max()\\nsorted_set_ = np.arange(sorted_set_.max())\\nelse:\\nselected_ = []\\nsorted_set_ = []', metadata={'source': './documents/ISLP_website.pdf', 'page': 594}), Document(page_content=\"590 13. Multiple Testing\\nWe now reproduce the middle panel of Figure 13.6.\\nIn[22]: fig, ax = plt.subplots()\\nax.scatter(np.arange(0, sorted_.shape[0]) + 1,\\nsorted_, s=10)\\nax.set_yscale( 'log')\\nax.set_xscale( 'log')\\nax.set_ylabel( 'P-Value' )\\nax.set_xlabel( 'Index' )\\nax.scatter(sorted_set_+1, sorted_[sorted_set_], c= 'r',s = 2 0 )\\nax.axline((0, 0), (1,q/m), c= 'k',l s ='--',l i n e w i d t h = 3 ) ;\\n13.6.4 A Re-Sampling Approach\\nHere, we implement the re-sampling approach to hypothesis testing using\\ntheKhandataset, which we investigated in Section 13.5. First, we merge\\nthe training and testing data, which results in observations on 83 patients\\nfor 2,308 genes.\\nIn[23]: Khan = load_data( 'Khan')\\nD=p d . c o n c a t ( [ K h a n [ 'xtrain' ], Khan[ 'xtest' ]])\\nD['Y']=p d . c o n c a t ( [ K h a n [ 'ytrain' ], Khan[ 'ytest' ]])\\nD['Y'].value_counts()\\nOut[23]: 22 9\\n42 5\\n31 8\\n11 1\\nName: Y, dtype: int64\\nThere are four classes of cancer. For each gene, we compare the mean ex-\\npression in the second class (rhabdomyosarcoma) to the mean expression in\\nthe fourth class (Burkitt’s lymphoma). Performing a standard two-sample\\nt-test using ttest_ind() fromscipy.stats on the11th gene produces attest_ind()test-statistic of -2.09 and an associated p-value of 0.0412, suggesting mod-\\nestevidenceofadifferenceinmeanexpressionlevelsbetweenthetwocancer\\ntypes.\\nIn[24]: D2 = D[lambda df:df[ 'Y']= =2 ]\\nD4 = D[lambda df:df[ 'Y']= =4 ]\\ngene_11 = 'G0011'\\nobservedT, pvalue = ttest_ind(D2[gene_11],\\nD4[gene_11],\\nequal_var=True)\\nobservedT, pvalue\\nOut[24]: (-2.094, 0.041)\\nHowever, this p-value relies on the assumption that under the null hy-\\npothesis of no difference between the two groups, the test statistic follows\\nat-distribution with 29 + 25−2 = 52degrees of freedom. Instead of us-\\ning this theoretical null distribution, we can randomly split the 54 patients\\ninto two groups of 29 and 25, and compute a new test statistic. Under the\\nnull hypothesis of no difference between the groups, this new test statis-\\ntic should have the same distribution as our original one. Repeating this\", metadata={'source': './documents/ISLP_website.pdf', 'page': 595}), Document(page_content='13.6 Lab: Multiple Testing 591\\nprocess 10,000 times allows us to approximate the null distribution of the\\ntest statistic. We compute the fraction of the time that our observed test\\nstatistic exceeds the test statistics obtained via re-sampling.\\nIn[25]: B=1 0 0 0 0\\nTnull = np.empty(B)\\nD_ = np.hstack([D2[gene_11], D4[gene_11]])\\nn_ = D2[gene_11].shape[0]\\nD_null = D_.copy()\\nfor b in range(B):\\nrng.shuffle(D_null)\\nttest_ = ttest_ind(D_null[:n_],\\nD_null[n_:],\\nequal_var=True)\\nTnull[b] = ttest_.statistic\\n(np.abs(Tnull) > np.abs(observedT)).mean()\\nOut[25]: 0.0398\\nThisfraction,0.0398,isourre-sampling-based p-value.Itisalmostidenti-\\ncal to the p-value of 0.0412 obtained using the theoretical null distribution.\\nWe can plot a histogram of the re-sampling-based test statistics in order\\nto reproduce Figure 13.7.\\nIn[26]: fig, ax = plt.subplots(figsize=(8,8))\\nax.hist(Tnull,\\nbins=100,\\ndensity=True,\\nfacecolor= \\'y\\',\\nlabel=\\'Null\\')\\nxval = np.linspace(-4.2, 4.2, 1001)\\nax.plot(xval,\\nt_dbn.pdf(xval, D_.shape[0]-2),\\nc=\\'r\\')\\nax.axvline(observedT,\\nc=\\'b\\',\\nlabel=\\'Observed\\' )\\nax.legend()\\nax.set_xlabel( \"Null Distribution of Test Statistic\" );\\nThe re-sampling-based null distribution is almost identical to the theoret-\\nical null distribution, which is displayed in red.\\nFinally, we implement the plug-in re-sampling FDR approach outlined\\nin Algorithm 13.4. Depending on the speed of your computer, calculating\\nthe FDR for all 2,308 genes in the Khandataset may take a while. Hence,\\nwe will illustrate the approach on a random subset of 100 genes. For each\\ngene, we first compute the observed test statistic, and then produce 10,000\\nre-sampled test statistics. This may take a few minutes to run. If you are\\nin a rush, then you could set Bequal to a smaller value (e.g. B=500).\\nIn[27]: m, B = 100, 10000\\nidx = rng.choice(Khan[ \\'xtest\\' ].columns, m, replace=False)\\nT_vals = np.empty(m)\\nTnull_vals = np.empty((m, B))\\nfor j in range(m):\\ncol = idx[j]', metadata={'source': './documents/ISLP_website.pdf', 'page': 596}), Document(page_content='592 13. Multiple Testing\\nT_vals[j] = ttest_ind(D2[col],\\nD4[col],\\nequal_var=True).statistic\\nD_ = np.hstack([D2[col], D4[col]])\\nD_null = D_.copy()\\nfor b in range(B):\\nrng.shuffle(D_null)\\nttest_ = ttest_ind(D_null[:n_],\\nD_null[n_:],\\nequal_var=True)\\nTnull_vals[j,b] = ttest_.statistic\\nNext, we compute the number of rejected null hypotheses R, the esti-\\nmated number of false positives ˆV, and the estimated FDR, for a range\\nof threshold values cin Algorithm 13.4. The threshold values are chosen\\nusing the absolute values of the test statistics from the 100 genes.\\nIn[28]: cutoffs = np.sort(np.abs(T_vals))\\nFDRs, Rs, Vs = np.empty((3, m))\\nfor j in range(m):\\nR=n p . s u m ( n p . a b s ( T _ v a l s )> =c u t o f f s [ j ] )\\nV=n p . s u m ( n p . a b s ( T n u l l _ v a l s )> =c u t o f f s [ j ] )/B\\nRs[j] = R\\nVs[j] = V\\nFDRs[j] = V / R\\nNow, for any given FDR, we can find the genes that will be rejected.\\nFor example, with FDR controlled at 0.1, we reject 15 of the 100 null\\nhypotheses. On average, we would expect about one or two of these genes\\n(i.e. 10% of 15) to be false discoveries. At an FDR of 0.2, we can reject\\nthe null hypothesis for 28 genes, of which we expect around six to be false\\ndiscoveries.\\nThe variable idxstores which genes were included in our 100 randomly-\\nselected genes. Let’s look at the genes whose estimated FDR is less than\\n0.1.\\nIn[29]: sorted(idx[np.abs(T_vals) >= cutoffs[FDRs < 0.1].min()])\\nAtanFDRthresholdof0.2,moregenesareselected,atthecostofhaving\\na higher expected proportion of false discoveries.\\nIn[30]: sorted(idx[np.abs(T_vals) >= cutoffs[FDRs < 0.2].min()])\\nThe next line generates Figure 13.11, which is similar to Figure 13.9,\\nexcept that it is based on only a subset of the genes.\\nIn[31]: fig, ax = plt.subplots()\\nax.plot(Rs, FDRs, \\'b\\',l i n e w i d t h = 3 )\\nax.set_xlabel( \"Number of Rejections\" )\\nax.set_ylabel( \"False Discovery Rate\" );', metadata={'source': './documents/ISLP_website.pdf', 'page': 597}), Document(page_content='13.7 Exercises 593\\nFIGURE 13.11.The estimated false discovery rate versus the number of rejectednull hypotheses, for 100 genes randomly selected from theKhandataset.13.7 ExercisesConceptual1. Suppose we testmnull hypotheses, all of which are true. We controlthe Type I error for each null hypothesis at levelα. For each sub-problem, justify your answer.(a) In total, how many Type I errors do we expect to make?(b) Suppose that themtests that we perform are independent.Whatisthefamily-wiseerrorrateassociatedwiththesemtests?Hint: If two eventsAandBare independent, thenPr(A∩B)=Pr(A) Pr(B).(c) Suppose thatm=2, and that thep-values for the two tests arepositively correlated, so that if one is small then the other willtend to be small as well, and if one is large then the other willtend to be large. How does the family-wise error rate associatedwith thesem=2tests qualitatively compare to the answer in(b) withm=2?Hint: First, suppose that the twop-values are perfectly correlated.(d) Suppose again thatm=2, but that now thep-values for thetwo tests are negatively correlated, so that if one is large thenthe other will tend to be small. How does the family-wise errorrate associated with thesem=2tests qualitatively compare tothe answer in (b) withm=2?Hint: First, suppose that whenever onep-value is less thanα,then the other will be greater thanα. In other words, we cannever reject both null hypotheses.', metadata={'source': './documents/ISLP_website.pdf', 'page': 598}), Document(page_content='594 13. Multiple Testing\\n2. Suppose that we test mhypotheses, and control the Type I error for\\neach hypothesis at level α. Assume that all mp-values are indepen-\\ndent, and that all null hypotheses are true.\\n(a) Let the random variable Ajequal1if thejth null hypothesis is\\nrejected, and 0otherwise. What is the distribution of Aj?\\n(b) What is the distribution of∑m\\nj=1Aj?\\n(c) What is the standard deviation of the number of Type I errors\\nthat we will make?\\n3. Suppose we test mnull hypotheses, and control the Type I error for\\nthejth null hypothesis at level αj, forj=1,...,m . Argue that the\\nfamily-wise error rate is no greater than∑m\\nj=1αj.\\nNull Hypothesis p-value\\nH01 0.0011\\nH02 0.031\\nH03 0.017\\nH04 0.32\\nH05 0.11\\nH06 0.90\\nH07 0.07\\nH08 0.006\\nH09 0.004\\nH10 0.0009\\nTABLE 13.4. p-values for Exercise 4.\\n4. Suppose we test m= 10 hypotheses, and obtain the p-values shown\\nin Table 13.4.\\n(a) Suppose that we wish to control the Type I error for each null\\nhypothesis at level α=0.05. Which null hypotheses will we\\nreject?\\n(b) Now suppose that we wish to control the FWER at level α=\\n0.05. Which null hypotheses will we reject? Justify your answer.\\n(c) Now suppose that we wish to control the FDR at level q=0.05.\\nWhich null hypotheses will we reject? Justify your answer.\\n(d) Now suppose that we wish to control the FDR at level q=0.2.\\nWhich null hypotheses will we reject? Justify your answer.\\n(e) Of the null hypotheses rejected at FDR level q=0.2, approxi-\\nmately how many are false positives? Justify your answer.\\n5. For this problem, you will make up p-values that lead to a certain\\nnumber of rejections using the Bonferroni and Holm procedures.\\n(a) Give an example of five p-values(i.e. five numbers between 0and\\n1which, for the purpose of this problem, we will interpret as p-\\nvalues) for which both Bonferroni’s method and Holm’s method', metadata={'source': './documents/ISLP_website.pdf', 'page': 599}), Document(page_content='13.7 Exercises 595\\nreject exactly one null hypothesis when controlling the FWER\\nat level0.1.\\n(b) Now give an example of five p-values for which Bonferroni re-\\njects one null hypothesis and Holm rejects more than one null\\nhypothesis at level 0.1.\\n6. For each of the three panels in Figure 13.3, answer the following\\nquestions:\\n(a) How many false positives, false negatives, true positives, true\\nnegatives, Type I errors, and Type II errors result from applying\\nthe Bonferroni procedure to control the FWER at level α=\\n0.05?\\n(b) How many false positives, false negatives, true positives, true\\nnegatives, Type I errors, and Type II errors result from applying\\nthe Holm procedure to control the FWER at level α=0.05?\\n(c) What is the false discovery proportion associated with using the\\nBonferroni procedure to control the FWER at level α=0.05?\\n(d) What is the false discovery proportion associated with using the\\nHolm procedure to control the FWER at level α=0.05?\\n(e) How would the answers to (a) and (c) change if we instead used\\nthe Bonferroni procedure to control the FWER at level α=\\n0.001?\\nApplied\\n7. This problem makes use of the Carseats dataset in the ISLPpackage.\\n(a) For each quantitative variable in the dataset besides Sales, fit\\na linear model to predict Salesusing that quantitative variable.\\nReport the p-values associated with the coefficients for the vari-\\nables. That is, for each model of the form Y=β0+β1X+ϵ,\\nreport the p-value associated with the coefficient β1. Here,Y\\nrepresents SalesandXrepresents one of the other quantitative\\nvariables.\\n(b) Suppose we control the Type I error at level α=0.05for the\\np-values obtained in (a). Which null hypotheses do we reject?\\n(c) NowsupposewecontroltheFWERatlevel 0.05forthep-values.\\nWhich null hypotheses do we reject?\\n(d) Finally, suppose we control the FDR at level 0.2for thep-values.\\nWhich null hypotheses do we reject?\\n8. In this problem, we will simulate data from m= 100 fund managers.\\nrng = np.random.default_rng(1)\\nn, m = 20, 100\\nX=r n g . n o r m a l ( s i z e = ( n ,m ) )', metadata={'source': './documents/ISLP_website.pdf', 'page': 600}), Document(page_content='596 13. Multiple TestingThesedatarepresenteachfundmanager’spercentagereturnsforeachofn= 20months. We wish to test the null hypothesis that eachfund manager’s percentage returns have population mean equal tozero. Notice that we simulated the data in such a way that each fundmanager’s percentage returns do have population mean zero; in otherwords, allmnull hypotheses are true.(a) Conduct a one-samplet-test for each fund manager, and plot ahistogram of thep-values obtained.(b) If we control Type I error for each null hypothesis at levelα=0.05, then how many null hypotheses do we reject?(c) If we control the FWER at level0.05, then how many null hy-potheses do we reject?(d) If we control the FDR at level0.05, then how many null hy-potheses do we reject?(e) Now suppose we “cherry-pick” the10fund managers who per-form the best in our data. If we control the FWER for just these10 fund managers at level0.05, then how many null hypothe-ses do we reject? If we control the FDR for just these 10 fundmanagers at level0.05, then how many null hypotheses do wereject?(f) Explain why the analysis in (e) is misleading.Hint: The standard approaches for controlling the FWER andFDR assume thatalltested null hypotheses are adjusted for mul-tiplicity, and that no “cherry-picking” of the smallestp-valueshas occurred. What goes wrong if we cherry-pick?', metadata={'source': './documents/ISLP_website.pdf', 'page': 601}), Document(page_content='Index\\naccuracy, 415\\nactivation, 400\\nactivation function, 401\\nadditive, 11,94–98,110–111\\nadditivity, 305,306\\nadjusted R2,87,231,232,236–\\n238\\nAdvertising data set, 15,16,19,\\n69,71–73,77,78,80,82,\\n83,85,87–90,95,96,109–\\n111\\nagglomerative clustering, 525\\nAkaike information criterion, 87,\\n231,232,236–238\\nalternative hypothesis, 76,559\\nanalysis of variance, 312\\nANOVA, 587\\narea under the curve, 155,486–\\n487\\nargument, 40\\narray,42\\nattribute, 42\\nAUC,155\\nAutodataset, 12,66,98–101,129,\\n197,202–207,327,398\\nauto-correlation, 421\\nautoregression, 423\\naxes,48\\nbackfitting, 307,328backpropagation, 429\\nbackward stepwise selection, 87,\\n234–235\\nbag-of-n-grams,415\\nbag-of-words, 414\\nbagging, 11,24,331,343–346,354,\\n360–361\\nBART,343,350,353,354,362–\\n363\\nbaseline, 93,145,161\\nbasis function, 293–294,296\\nBayes\\nclassifier, 35–37,147\\ndecision boundary, 148\\nerror,35–37\\nBayes’ theorem, 146,250\\nBayesian, 250–251,353\\nBayesianadditiveregressiontrees,\\n331,343,350,350,353,\\n354,362–363\\nBayesianinformationcriterion, 87,\\n231,232,236–238\\nBenjamini–Hochbergprocedure, 575–\\n577\\nBernoulli distribution, 172\\nbest subset selection, 231,246\\nbias,31–34,74,90,159,405\\nbias-variance\\ndecomposition, 32\\n© Springer Nature Switzerland AG 2023 \\nG. James et al., An Introduction to Statistical Learning , Springer Texts in Statistics, \\nhttps://doi.org/10.1007/978-3-031-38747-0 597', metadata={'source': './documents/ISLP_website.pdf', 'page': 602}), Document(page_content='598 Index\\ntrade-off, 31–34,38,111–112,\\n157,159,163,164,242,\\n254,263,266,301,336,\\n376,385\\nbidirectional, 425\\nBikeshare data set, 12,167–172\\nbinary,27,138\\nbiplot,507,508\\nBonferroni method, 575–577,585\\nBoolean, 53,176\\nboosting, 11,24,331,343,347–\\n350,354,361–362\\nbootstrap, 11,201,212–214,343\\nBostondata set, 12,67,117,122,\\n133,199,227,287,327,\\n364,556\\nbottom-up clustering, 525\\nboxplot, 62\\nBrainCancer data set, 12,472–\\n474,476,482\\nbranch,333\\nburn-in, 352\\nC-index, 487\\nCaravan data set, 12,184,366\\nCarseats data set, 12,126,130,\\n364\\ncategorical, 2,27\\ncensored data, 469–502\\ncensoring\\nindependent, 471\\ninterval, 471\\nleft,471\\nmechanism, 471\\nnon-informative, 471\\nright,471\\ntime,470\\nchain rule, 429\\nchannel, 407\\nCIFAR100 data set, 406,409–411,\\n448,449\\nclassification, 2,11,27,34–39,135–\\n199,367–382\\nerror rate, 338\\ntree,337–341,355–358\\nclassifier, 135\\ncluster analysis, 25–26\\nclustering, 4,25–26,520–535\\nagglomerative, 525bottom-up, 525\\nhierarchical, 521,525–535\\nK-means,11,521–524\\nCochran–Mantel–Haenszeltest, 475\\ncoefficient, 71\\nCollege dataset, 12,65,286,328\\ncollinearity, 106–110\\nconcatenation, 41\\nconditional probability, 35\\nconfidenceinterval, 75–76,90,110,\\n292\\nconfounding, 144\\nconfusion matrix, 153,176\\ncontinuous, 2\\ncontour, 246\\ncontour plot, 50\\ncontrast, 94\\nconvenience function, 53\\nconvolution filter, 407\\nconvolution layer, 407\\nconvolutionalneuralnetwork, 406–\\n413\\ncorrelation, 79,82–83,530\\ncount data, 167,170\\nCox’sproportionalhazardsmodel,\\n480,483–486\\nCp,87,231,232,236–238\\nCreditdata set, 12,91,92,94,\\n97,98,106–109\\ncross-entropy, 405\\ncross-validation, 11,31,34,201–\\n211,231,252,270\\nk-fold,206–209\\nleave-one-out, 204–206\\ncurse of dimensionality, 115,193,\\n266\\ndata augmentation, 411\\ndata frame, 55\\nData sets\\nAdvertising ,15,16,19,69,\\n71–73,77,78,80,82,83,\\n85,87–90,95,96,109–\\n111\\nAuto,12,66,98–101,129,197,\\n202–207,327,398\\nBikeshare ,12,167–172', metadata={'source': './documents/ISLP_website.pdf', 'page': 603}), Document(page_content='Index 599\\nBoston,12,67,117,122,133,\\n199,227,287,327,364,\\n556\\nBrainCancer ,12,472–474,476,\\n482\\nCaravan,12,184,366\\nCarseats ,12,126,130,364\\nCIFAR100 ,406,409–411,448,\\n449\\nCollege,12,65,286,328\\nCredit,12,91,92,94,97,98,\\n106–109\\nDefault,12,136–139,141–\\n144,152–156,160,161,\\n225,226,466\\nFund,12,567–570,572,575,\\n576,585,588,589\\nHeart,339,340,344–347,352,\\n353,382,383\\nHitters,12,332,333,336,\\n338,339,366,425,426,\\n437,446\\nIMDb,413,415,416,418,420,\\n437,458,467\\nIncome,16–18,21–23\\nKhan,12,579–581,583,590,\\n593\\nMNIST,402–404,406,430,431,\\n441,444,445,448\\nNCI60,4,5,12,546,548–550\\nNYSE,12,422–424,466,467\\nOJ,12,365,398\\nPortfolio ,12\\nPublication ,12,482–487\\nSmarket,2,3,12,173,184,\\n196\\nUSArrests ,12,507,508,510,\\n512,513,515,516,518,\\n519\\nWage,1,2,8,9,12,290,291,\\n293,295,297–300,302–\\n306,309,315,327\\nWeekly,12,196,226\\ndata type, 42\\ndecision function, 387\\ndecision tree, 11,331–342\\ndeep learning, 399Default dataset, 12,136–139,141–\\n144,152–156,160,161,\\n225,226,466\\ndegrees of freedom, 30,266,295,\\n296,301\\ndendrogram, 521,525–530\\ndensity function, 146\\ndependent variable, 15\\nderivative, 296,300\\ndetector layer, 410\\ndeviance, 232\\ndictionary, 66\\ndimensionreduction, 230,253–262\\ndiscriminant function, 149\\ndiscriminant method, 146–161\\ndissimilarity, 530–532\\ndistance\\ncorrelation-based, 530–532,554\\nEuclidean, 509,522,523,529–\\n532\\ndouble descent, 431–435\\ndouble-exponentialdistribution, 251\\ndropout, 406,431\\ndummy variable, 91–94,138,142,\\n292\\nearly stopping, 430\\neffective degrees of freedom, 301\\neigen decomposition, 506,516\\nelbow,548\\nembedding, 418\\nembedding layer, 419\\nensemble, 343–354\\nentropy, 337–339,363\\nepochs,430\\nerror\\nirreducible, 17,30\\nrate,34\\nreducible, 17\\nterm,16\\nEuclideandistance, 509,522,523,\\n529–532,554\\nevent time, 470\\nexception, 45\\nexpected value, 18\\nexploratory data analysis, 504\\nexponential, 173\\nexponential family, 173\\nF-statistic, 84', metadata={'source': './documents/ISLP_website.pdf', 'page': 604}), Document(page_content='600 Index\\nfactor,92\\nfactorial, 170\\nfailure time, 470\\nfalse\\ndiscoveryproportion, 155,573\\ndiscovery rate, 558,573–577,\\n579–582\\nnegative, 155,562\\npositive, 155,562,563\\npositive rate, 155,156,382\\nfamily-wiseerrorrate, 565–573,577\\nfeature,15\\nfeature map, 406\\nfeature selection, 230\\nfeaturize, 414\\nfeed-forward neural network, 400\\nfigure,48\\nfit,21\\nfitted value, 101\\nflattening, 424\\nflexible,21\\nfloating point, 43\\nforwardstepwiseselection, 86,87,\\n233–234,268\\nfunction, 40\\nFunddata set, 12,567–570,572,\\n575,576,585,588,589\\nGamma, 173\\nGaussian(normal)distribution, 146,\\n147,150,172,561\\ngeneralized additive model, 5,24,\\n162,289,290,305–309,\\n319\\ngeneralized linear model, 5,135,\\n167–174,217\\ngenerative model, 146–161\\nGiniindex, 337–339,345,346,363\\nglobal minimum, 427\\ngradient, 428\\ngradient descent, 427\\nHarrell’s concordance index, 487\\nhazard function, 476–478\\nbaseline, 478\\nhazard rate, 476\\nHeartdataset, 339,340,344–347,\\n352,353,382,383\\nheatmap, 50\\nhelper,311heteroscedasticity, 103,168\\nhidden layer, 400\\nhidden units, 400\\nhierarchical clustering, 525–530\\ndendrogram, 525–528\\ninversion, 529\\nlinkage,529–530\\nhierarchical principle, 96\\nhigh-dimensional, 86,234,263\\nhinge loss, 385\\nHitters data set, 12,332,333,\\n336,338,339,366,425,\\n426,437,446\\nhold-out set, 202\\nHolm’s method, 568,576,585\\nhypergeometric distribution, 501\\nhyperparameter, 187\\nhyperplane, 367–372\\nhypothesis test, 76–77,84,103,\\n558–583\\nIMDbdata set, 413,415,416,418,\\n420,437,458,467\\nimputation, 515\\nIncomedata set, 16–18,21–23\\nincrement, 60\\nindependent variable, 15\\nindexable, 186\\nindicator function, 292\\ninference, 17,18\\ninner product, 379,380\\ninput layer, 400\\ninput variable, 15\\nintegral, 301\\ninteraction, 70,89,95–98,110–\\n111,308\\nintercept, 71,72\\ninterpolate, 432\\ninterpretability, 229\\ninversion, 529\\nirreducible error, 17,36,90,110\\niterator, 312\\njoint distribution, 158\\nK-means clustering, 11,521–524\\nK-nearestneighbors, 135,164–167\\nclassifier, 11,36–37\\nregression, 111–115', metadata={'source': './documents/ISLP_website.pdf', 'page': 605}), Document(page_content='Index 601\\nKaplan–Meiersurvivalcurve, 472–\\n474,483\\nkernel,379–382,384,394\\nlinear,380\\nnon-linear, 377–382\\npolynomial, 380,382\\nradial,381–383,390\\nkernel density estimator, 159\\nkeyword, 46\\nKhandata set, 12,579–581,583,\\n590,593\\nknot,290,294,296–299\\nℓ1norm,244\\nℓ2norm,242\\nlag,422\\nLaplace distribution, 251\\nlasso,11,24,244–251,265–266,\\n336,385,484\\nleaf,333,526\\nlearning rate, 429\\nleast squares, 5,21,71–72,140,\\n141,229\\nline,73\\nweighted, 103\\nlevel,92\\nleverage, 104–106\\nlikelihood function, 141\\nlinear,2,69–115\\nlinear combination, 128,230,253,\\n505\\nlineardiscriminantanalysis, 5,11,\\n135,138,147–155,164–\\n167,377,382\\nlinear kernel, 380\\nlinear model, 20,69–115\\nlinear regression, 5,11,69–115,\\n172–173\\nmultiple, 80–90\\nsimple,70–80\\nlink function, 172,173\\nlinkage,529–530,548\\naverage, 529–530\\ncentroid, 529–530\\ncomplete, 526,529–530\\nsingle,529–530\\nlist,41\\nlist comprehension, 123\\nlocal minimum, 427local regression, 290\\nlog odds, 145\\nlog-rank test, 474–476,483\\nlogistic function, 139\\nlogistic regression, 5,11,25,135,\\n138–144,164–167,172–\\n173,308–309,377,384–\\n385\\nmultinomial, 145,163\\nmultiple, 142–144\\nlogit,140\\nloss function, 300,385\\nlow-dimensional, 262\\nLSTM RNN, 420\\nmain effects, 96\\nmajority vote, 344\\nMallow’s Cp,87,231,232,236–\\n238\\nMantel–Haenszel test, 475\\nmargin,370,385\\nmarginal distribution, 158\\nMarkov chain Monte Carlo, 353\\nmatrix completion, 515\\nmatrix multiplication, 10\\nmaximal margin\\nclassifier, 367–372\\nhyperplane, 370\\nmaximumlikelihood, 139–141,143,\\n170\\nmean squared error, 28\\nmesh,53\\nmethod, 43\\nminibatch, 429\\nmisclassification error, 35\\nmissing at random, 515\\nmissing data, 56,515–520\\nmixed selection, 87\\nMNISTdataset, 402–404,406,430,\\n431,441,444,445,448\\nmodel assessment, 201\\nmodel selection, 201\\nmodule, 42\\nmulticollinearity, 108,266\\nmultinomiallogisticregression, 145,\\n163\\nmultiple testing, 557–583\\nmulti-task learning, 403\\nmultivariate Gaussian, 150', metadata={'source': './documents/ISLP_website.pdf', 'page': 606}), Document(page_content='602 Index\\nmultivariate normal, 150\\nnaive Bayes, 135,158–161,164–\\n167\\nnamespace, 116\\nnatural spline, 297,298,301,317\\nNCI60data set, 4,5,12,546,548–\\n550\\nnegative binomial, 173\\nnegativepredictivevalue, 155,156\\nneural network, 5,399\\nnode\\ninternal, 333\\npurity,337–339\\nterminal, 333\\nnoise,21,252\\nnon-linear, 2,11,289–329\\ndecision boundary, 377–382\\nkernel,377–382\\nnon-parametric, 20,22–23,111–\\n115,193\\nnormal(Gaussian)distribution, 146,\\n147,150,172,476,561\\nnotebook, 40\\nnull,152\\ndistribution, 561,578\\nhypothesis, 76,559\\nmodel,87,231,245\\nnull rate, 186\\nNYSEdata set, 12,422–424,466,\\n467\\nOccam’s razor, 426\\nodds,140,145,195\\nOJdata set, 12,365,398\\none-hot encoding, 92,126,403\\none-standard-error rule, 240\\none-versus-all, 384\\none-versus-one, 384\\none-versus-rest, 384\\noptimalseparatinghyperplane, 370\\noptimism of training error, 30\\nordered categorical variable, 315\\northogonal, 257,506\\nbasis,125\\nout-of-bag, 345\\noutlier,103–104\\noutput variable, 15\\nover-parametrized, 465\\noverdispersion, 172overfitting, 21,23,25,30–31,88,\\n152,233,371\\np-value,77,82,560–562,578–579\\nadjusted, 586\\npackage, 42\\nparameter, 71\\nparametric, 20–22,111–115\\npartialleastsquares, 254,260–262,\\n282\\npartial likelihood, 480\\npath algorithm, 249\\npermutation, 578\\npermutation approach, 577–582\\nperpendicular, 257\\nPoisson distribution, 169,172\\nPoisson regression, 135,167–173\\npolynomial\\nkernel,380,382\\nregression, 98–99,289–292,294–\\n295\\npooling, 410\\npopulation regression line, 73\\nPortfolio data set, 12\\npositivepredictivevalue, 155,156\\nposterior\\ndistribution, 251\\nmode,251\\nprobability, 147\\npower,108,155,563\\nprecision, 155\\nprediction, 17\\ninterval, 90,110\\npredictor, 15\\nprincipal components, 505\\nanalysis, 11,254–260,504–515\\nloading vector, 505,506\\nmissing values, 515–520\\nproportionofvarianceexplained,\\n510–515,547\\nregression, 11,254–260,280–\\n282,504,515\\nscore vector, 506\\nscree plot, 514–515\\nprior\\ndistribution, 251\\nprobability, 146\\nprobability density function, 477,\\n478', metadata={'source': './documents/ISLP_website.pdf', 'page': 607}), Document(page_content='Index 603\\nprojection, 230\\nproportionalhazardsassumption,\\n478\\npruning, 336\\ncost complexity, 336\\nweakest link, 336\\nPublication data set, 12,482–\\n487\\nPython objects and functions\\n%%capture ,458\\niloc[],58\\nloc[],57\\nAgglomerativeClustering() ,\\n543\\nanova(),313\\nanova_lm() ,125,129,312,\\n313\\naxhline() ,122,551\\naxline() ,121,129,329\\nBART(),362\\nbiplot,537\\nboot_SE() ,223\\nboxplot() ,62,66\\nbs(),315,327\\nBSpline() ,315\\nclone(),222\\ncolumns.drop() ,122\\ncompute_linkage() ,544\\nconfusion_table() ,176\\ncontour() ,50\\ncorr(),129,174\\ncost_complexity_pruning_path() ,\\n357\\nCoxPHFitter() ,491\\ncross_val_predict() ,270\\ncross_validate() ,218,219,\\n226\\ncumsum() ,539\\ncut_tree() ,545\\ndata.frame() ,227\\nDataset,440\\ndecision_function() ,392\\nDecisionTreeClassifier() ,\\n354,355\\nDecisionTreeRegressor() ,354\\ndef,121\\ndendrogram() ,544\\ndescribe() ,62,66\\ndir(),116drop(),179\\ndropna() ,56,268,461\\nDTC(),seeDecisionTreeClassifier()\\nDTR(),seeDecisionTreeRegressor()\\ndtype,43\\nElasticNetCV() ,279\\nenumerate() ,217\\nexport_text() ,356\\nexport_tree() ,365\\nfit(),118,181,218\\nfit_transform() ,119\\nfor,59\\nGaussianNB() ,182\\nGBR(),seeGradientBoosting-\\nRegressor()\\nget_dummies() ,461\\nget_influence() ,121\\nget_prediction() ,120,314\\nget_rdataset() ,535\\nglm(),313\\nglob(),437\\nGradientBoostingClassifier() ,\\n361\\nGradientBoostingRegressor() ,\\n354,361\\nGridSearchCV() ,276\\ngroupby() ,490\\nhist(),62\\niloc[],58,59\\nimport,42\\nimshow() ,50,449\\nISLP.bart ,362\\nISLP.cluster ,544\\njson,437\\nKaplanMeierFitter() ,502\\nkeras,437\\nKFold(),219\\nKMeans() ,542,543\\nKmeans() ,542\\nKNeighborsClassifier() ,183\\nlambda,58\\nLDA(),seeLinearDiscriminant-\\nAnalysis()\\nlegend() ,132\\nlifelines ,490\\nLinearDiscriminantAnalysis() ,\\n174,179\\nLinearGAM() ,317\\nLinearRegression() ,280', metadata={'source': './documents/ISLP_website.pdf', 'page': 608}), Document(page_content='604 Index\\nload_data() ,117\\nloc[],58,59,177\\nlog_loss() ,355\\nLogisticGAM() ,323\\nlogrank_test() ,490\\nlowess() ,324\\nmatplotlib ,48\\nmax(),66\\nmean(),48\\nmedian() ,197\\nmin(),66\\nMNIST(),444\\nModelSpec() ,116–118,122,\\n124,267\\nMS(),seeModelSpec()\\nmult_test() ,seemultipletests()\\nmultipletests() ,586\\nmultipletests() ,583,589\\nmultivariate_logrank_test() ,\\n496\\nNaturalSpline() ,317,319\\nndim,42\\nnn.RNN() ,461\\nnormal() ,132,286,555\\nnp,seenumpy\\nnp.all() ,54,180\\nnp.allclose() ,190\\nnp.any() ,54\\nnp.arange() ,51\\nnp.argmax() ,122\\nnp.array() ,42\\nnp.concatenate() ,133\\nnp.corrcoef() ,46,554\\nnp.empty() ,224\\nnp.isnan() ,268\\nnp.ix_() ,53\\nnp.linalg.svd() ,539\\nnp.linspace() ,50\\nnp.logspace() ,318\\nnp.mean() ,47,176\\nnp.nan,60\\nnp.nanmean() ,541\\nnp.percentile() ,228\\nnp.power() ,219\\nnp.random.choice() ,553\\nnp.random.default_rng() ,46,\\n47\\nnp.random.normal() ,45\\nnp.sqrt() ,45np.squeeze() ,457\\nnp.std() ,47\\nnp.sum() ,43\\nnp.var() ,47\\nnp.where() ,180\\nns(),317\\nnumpy,42,555\\nos.chdir() ,55\\nouter(),219\\npairwise_distances() ,554\\npairwise_tukeyhsd() ,587\\npandas,55\\nparams,175\\npartial() ,222,269\\nPCA(),280,537,540,554\\npd,seepandas\\npd.crosstab() ,555\\npd.cut() ,315\\npd.get_dummies() ,314\\npd.plotting.scatter_matrix() ,\\n62\\npd.qcut() ,314,315\\npd.read_csv() ,55,556\\npd.Series() ,62\\nPipeline() ,275\\nplot(),48,61,356,490\\nplot.scatter() ,120\\nplot_gam() ,321\\nplot_svm() ,398\\nPLSRegression() ,282\\npoly(),125,313,327\\npredict() ,175,178,181,216,\\n218,323,358\\npredict_survival_function() ,\\n493\\nprint(),40\\npvalues,175\\npygam,307,317\\npytorch_lightning ,435\\nQDA(),seeQuadraticDiscriminant-\\nAnalysis()\\nQuadraticDiscriminantAnalysis() ,\\n174,181\\nrandom() ,555\\nRandomForestRegressor() ,354,\\n360\\nread_image() ,436\\nreindex() ,461\\nreshape() ,43', metadata={'source': './documents/ISLP_website.pdf', 'page': 609}), Document(page_content='Index 605\\nreturn,198\\nRF(),seeRandomForestRegressor()\\nrng,seenp.random.default_rng()\\nrng.choice() ,60\\nrng.standard_normal() ,60\\nroc_curve() ,392\\nRocCurveDisplay.from_estimator() ,\\n387\\nsavefig() ,50\\nscatter() ,49,61\\nscipy.interpolate ,315\\nscore(),218,461\\nseed_everything() ,436\\nset_index() ,57\\nset_title() ,49\\nset_xlabel() ,49\\nset_xscale() ,198\\nset_ylabel() ,49\\nset_yscale() ,198\\nshape,43\\nShuffleSplit() ,219\\nsim_time() ,495\\nSimpleDataModule() ,441\\nSimpleModule.classification() ,\\n446\\nSimpleModule.regression() ,\\n442\\nskl,seesklearn.linear_model\\nskl.ElasticNet() ,273,277\\nskl.ElasticNet.path ,274\\nskl.ElasticNet.path() ,273\\nsklearn,118,181\\nsklearn.ensemble ,360\\nsklearn.linear_model ,267\\nsklearn.model_selection ,267\\nsklearn_selected() ,269\\nsklearn_selection_path() ,\\n270\\nsklearn_sm() ,218\\nskm,seesklearn.model_selection\\nskm.cross_val_predict() ,271\\nskm.KFold() ,271\\nskm.ShuffleSplit() ,272\\nslice(),51,462\\nsm,seestatsmodels\\nsm.GLM() ,174,192,226\\nsm.Logit() ,174\\nsm.OLS() ,118,129,174,319StandardScaler() ,185,438,\\n537,555\\nstatsmodels ,116,173\\nstd(),186\\nStepwise() ,269\\nstr.contains() ,59\\nsubplots() ,48\\nsum(),43,268\\nsummarize() ,118,129,223,\\n226\\nsummary() ,119,322,587\\nsuper(),440\\nSupportVectorClassifier() ,\\n387,389–391,393\\nSupportVectorRegression() ,\\n394\\nSVC(),seeSupportVector-\\nClassifier()\\nsvd(),539\\nSVR(),seeSupportVector-\\nRegression()\\nTensorDataset() ,441\\nto_numpy() ,437\\ntorch,435\\ntorchinfo ,436\\ntorchmetrics ,436\\ntorchvision ,436\\nToTensor() ,444\\ntrain_test_split() ,186,216\\ntransform() ,118,119\\nttest_1samp() ,584\\nttest_ind() ,590\\nttest_rel() ,587\\ntuple,43\\nuniform() ,555\\nvalue_counts() ,66\\nvar(),536\\nvariance_inflation_factor() ,\\n116,124\\nVIF(),seevariance_inflation-\\n_factor()\\nwhere(),355\\nzip(),60,312\\nq-values, 589\\nquadratic, 98\\nquadraticdiscriminantanalysis, 4,\\n135,156–157,164–167', metadata={'source': './documents/ISLP_website.pdf', 'page': 610}), Document(page_content='606 Index\\nqualitative, 2,27,91,135,167,\\n202\\nvariable, 91–94\\nquantitative, 2,27,91,135,167,\\n202\\nradial kernel, 381,383,390\\nrandom forest, 11,331,343,346–\\n347,354,360–361\\nrandom seed, 46\\nre-sampling, 577–582\\nrecall,155\\nreceiveroperatingcharacteristic(ROC),\\n154,382–383\\nrecommender systems, 516\\nrectified linear unit, 401\\nrecurrentneuralnetwork, 416–427\\nrecursivebinarysplitting, 334,337,\\n338\\nreducible error, 17,90\\nregression, 2,11,27\\nlocal,289,290,304–305\\npiecewisepolynomial, 294–295\\npolynomial, 289–292,299\\nspline,289,294\\ntree,331–337,358–360\\nregularization, 230,240,406,484–\\n486\\nReLU,401\\nresampling, 201–214\\nresidual, 71,81\\nplot,100\\nstandarderror, 75,77–78,88–\\n89,109\\nstudentized, 104\\nsum of squares, 71,79,81\\nresiduals, 263,348\\nresponse, 15\\nridgeregression, 11,240–244,385,\\n484\\nrisk set, 473\\nrobust,374,376,535\\nROC curve, 154,382–383,486–\\n487\\nR2,77–80,88,109,238\\nrug plot, 314\\nscale equivariant, 242\\nScheffé’s method, 572\\nscree plot, 512,514–515elbow,514\\nsemi-supervised learning, 27\\nsensitivity, 153,155,156\\nseparating hyperplane, 367–372\\nSeq2Seq, 425\\nsequence, 41\\nshrinkage, 230,240,484–486\\npenalty, 240\\nsigmoid, 401\\nsignal,252\\nsignature, 45\\nsingularvaluedecomposition, 539\\nslack variable, 375\\nslice,51\\nslope,71,72\\nSmarket data set, 2,3,12,173,\\n184,196\\nsmoother, 308\\nsmoothing spline, 290,300–303\\nsoft margin classifier, 372–374\\nsoft-thresholding, 250\\nsoftmax, 145,405\\nsparse,244,252\\nsparse matrix format, 414\\nsparsity, 244\\nspecificity, 153,155,156\\nspline,289,294–303\\ncubic,296\\nlinear,296\\nnatural, 297,301\\nregression, 289,294–299\\nsmoothing, 30,290,300–303\\nthin-plate, 22\\nstandard error, 75,101\\nstandardize, 185\\nstatistical model, 1\\nstep function, 111,289,292–293\\nstepwise model selection, 11,231,\\n233\\nstochastic gradient descent, 429\\nstring,41\\nstring interpolation, 490\\nstump,349\\nsubset selection, 230–240\\nsubtree, 336\\nsupervised learning, 25–27,261\\nsupport vector, 371,376,385\\nclassifier, 367,372–377\\nmachine, 5,11,24,377–386', metadata={'source': './documents/ISLP_website.pdf', 'page': 611}), Document(page_content='Index 607\\nregression, 386\\nsurvival\\nanalysis, 469–502\\ncurve,472,483\\nfunction, 472\\ntime,470\\nsynergy, 70,89,95–98,110–111\\nsystematic, 16\\nt-distribution, 77,165\\nt-statistic, 76\\nt-test\\none-sample, 583,584,588\\npaired,587\\ntwo-sample, 559,570,571,577–\\n581,584,590\\ntest\\nerror,35,37,176\\nMSE,28–32\\nobservations, 28\\nset,30\\nstatistic, 559\\ntheoretical null distribution, 577\\ntime series, 101\\ntotal sum of squares, 79\\ntracking, 102\\ntrain,21\\ntraining\\ndata,20\\nerror,35,37,176\\nMSE,28–31\\ntransformer, 311\\ntree,331–342\\ntree-based method, 331\\ntrue negative, 155\\ntrue positive, 155\\ntrue positive rate, 155,156,382\\ntruncated power basis, 296\\nTukey’s method, 571,585,587\\ntuning parameter, 187,240,484\\ntwo-sample t-test,474\\nType I error, 155,562–565\\nType I error rate, 563\\nType II error, 155,563,568,584unsupervisedlearning, 25–27,255,\\n260,503–552\\nUSArrests data set, 12,507,508,\\n510,512,513,515,516,\\n518,519\\nvalidation set, 202\\napproach, 202–204\\nvariable, 15\\ndependent, 15\\ndummy, 91–94,97–98\\nimportance, 346,360\\nindependent, 15\\nindicator, 35\\ninput,15\\noutput,15\\nqualitative, 91–94,97–98\\nselection, 86,230,244\\nvariance, 18,31–34,159\\ninflation factor, 108–110,123\\nvarying coefficient model, 305\\nWagedata set, 1,2,8,9,12,290,\\n291,293,295,297–300,\\n302–306,309,315,327\\nweak learner, 343\\nweakest link pruning, 336\\nWeeklydata set, 12,196,226\\nweight freezing, 412,419\\nweight sharing, 418\\nweighted least squares, 103,304\\nweights, 404\\nwith replacement, 214\\nwithin class covariance, 150\\nwrapper, 217', metadata={'source': './documents/ISLP_website.pdf', 'page': 612})]\n"
     ]
    }
   ],
   "source": [
    "print(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_openai import OpenAIEmbeddings\n",
    "from langchain_text_splitters import RecursiveCharacterTextSplitter\n",
    "from langchain_community.vectorstores import Chroma\n",
    "\n",
    "# quote= \"Artem Bisliouk is a hockey player from Mannheim. He is very good at data science and 22 years old.\"\n",
    "\n",
    "splitter = RecursiveCharacterTextSplitter(\n",
    "    chunk_size=200,\n",
    "    chunk_overlap=50,\n",
    "    separators=['.'])\n",
    "\n",
    "docs = splitter.split_documents(data) \n",
    "# docs = splitter.split_text(quote) \n",
    "\n",
    "# Embed the documents and store them in a Chroma DB\n",
    "embedding_model = OpenAIEmbeddings(openai_api_key=openai_api_key)\n",
    "docstorage = Chroma.from_documents(docs, embedding_model)\n",
    "# docstorage = Chroma.from_texts(docs, embedding_model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Artem Bisliouk is a hockey player from Mannheim. He is very good at data science and 22 years old.']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "docs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the Retrieval QA Chain to integrate the database and LLM\n",
    "from langchain.chains import RetrievalQA\n",
    "from langchain_openai import OpenAI\n",
    "\n",
    "qa = RetrievalQA.from_chain_type(\n",
    "    OpenAI(model_name=\"gpt-3.5-turbo-instruct\", temperature=0, openai_api_key=openai_api_key), chain_type=\"stuff\", retriever=docstorage.as_retriever())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " test error, cross-validation\n"
     ]
    }
   ],
   "source": [
    "# Run the chain on the query provided\n",
    "query = \"\"\"Therefore, a better strategy is to grow a very large tree T0, and then\n",
    "prune it back in order to obtain a subtree. How do we determine the best prune\n",
    "subtree way to prune the tree? Intuitively, our goal is to select a subtree that\n",
    "leads to the lowest test error rate. Given a subtree, we can estimate its\n",
    "[MASK] using [MASK] or the validation set approach. However,\n",
    "estimating the cross-validation error for every possible subtree would be too\n",
    "cumbersome, since there is an extremely large number of possible subtrees.\"\"\" \n",
    "print(qa.run(query))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
